From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 3 Dec 2023 05:33:24 -0500
Subject: [PATCH] Fix keepalive kicked name

Removed since 1.20.6

diff --git a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 9fa1cb2146bb8173a246981c0f73cd8c1bc44abe..487e24f1a26428cdaee3ca946ac8f5f1f599b260 100644
--- a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -288,7 +288,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         if (GaleGlobalConfiguration.get().misc.keepalive.sendMultiple) {
             if (elapsedTime >= 1000L) { // 1 second
                 if (!this.processedDisconnect && this.keepAlives.size() >= KEEPALIVE_LIMIT_IN_SECONDS) {
-                    LOGGER.warn("{} was kicked due to keepalive timeout!", this.player.getName());
+                    LOGGER.warn("{} was kicked due to keepalive timeout!", this.player.getScoreboardName()); // Leaf - Fix keepalive kicked name
                     disconnect(Component.translatable("disconnect.timeout"), org.bukkit.event.player.PlayerKickEvent.Cause.TIMEOUT);
                 } else {
                     this.keepAliveTime = currentTime; // hijack this field for 1 second intervals
