From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Tue, 12 Mar 2024 01:10:54 +0800
Subject: [PATCH] Use a shadow fork that supports Java 21

Removed since 1.20.6

diff --git a/build.gradle.kts b/build.gradle.kts
index 71f72777303cce3b94628f96c5877e75332cf4d8..78a543150da9e2db42ec8ce07f4fa7528467462d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,7 +3,7 @@ import io.papermc.paperweight.util.*
 plugins {
     java
     `maven-publish`
-    id("com.github.johnrengelman.shadow")
+    id("io.github.goooler.shadow") // Leaf - use a shadow fork that supports Java 21
 }
 
 val log4jPlugins = sourceSets.create("log4jPlugins")
