From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 4 Nov 2024 23:07:27 -0500
Subject: [PATCH] Leaf Commands


diff --git a/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java b/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java
index 7d66126c1f5957c109a2426d53f5d0072886309b..f623a1d9e580104e1ab8b7cf88b9d26212f8acd8 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java
@@ -6,6 +6,7 @@ import org.bukkit.util.permissions.DefaultPermissions;
 public final class CraftDefaultPermissions {
     private static final String ROOT = "minecraft";
     public static final String GALE_ROOT = "gale"; // Gale - set Gale permissions root
+    public static final String LEAF_ROOT = "leaf"; // Leaf - Leaf commands
 
     public static final String writeBooks = GALE_ROOT + ".writebooks"; // Gale - Pufferfish - make book writing configurable
 
