From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 29 Apr 2024 14:18:58 -0400
Subject: [PATCH] Fix Pufferfish and Purpur patches


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index 5b844d2fea387f2cb655169f458ec0ca627f1c65..f267abd6f905180cfa5cf4961cf79f57544a891c 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -32,9 +32,11 @@ public record ServerBuildInfoImpl(
 
     private static final String BRAND_PAPER_NAME = "Paper";
     private static final String BRAND_GALE_NAME = "Gale"; // Gale - branding changes
-    // Leaf start - Rebrand
+    // Leaf start - Rebrand & Fix Pufferfish and Purpur patches
+    private static final String BRAND_PUFFERFISH_NAME = "Pufferfish";
+    private static final String BRAND_PURPUR_NAME = "Purpur";
     private static final String BRAND_LEAF_NAME = "Leaf";
-    // Leaf end - Rebrand
+    // Leaf end - Rebrand & Fix Pufferfish and Purpur patches
 
     private static final String BUILD_DEV = "DEV";
 
@@ -65,11 +67,13 @@ public record ServerBuildInfoImpl(
 
     @Override
     public boolean isBrandCompatible(final @NotNull Key brandId) {
-        // Leaf start - Rebrand
+        // Leaf start - Rebrand & Fix Pufferfish and Purpur patches
         return brandId.equals(this.brandId)
             || brandId.equals(BRAND_PAPER_ID)
-            || brandId.equals(BRAND_GALE_ID); // Gale - branding changes
-            // Leaf end - Rebrand
+            || brandId.equals(BRAND_GALE_ID) // Gale - branding changes
+            || brandId.equals(BRAND_PUFFERFISH_ID)
+            || brandId.equals(BRAND_PURPUR_ID);
+            // Leaf end - Rebrand & Fix Pufferfish and Purpur patches
     }
 
     @Override
