From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 4 Nov 2024 23:07:27 -0500
Subject: [PATCH] Leaf Commands


diff --git a/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java b/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java
index 11c857e6d42f23994dbded70fbe052e15db2dc4b..ed275123642d929b51b93f7f2ea825858411a921 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java
@@ -6,6 +6,7 @@ import org.bukkit.util.permissions.DefaultPermissions;
 public final class CraftDefaultPermissions {
     private static final String ROOT = "minecraft";
     public static final String GALE_ROOT = "gale"; // Gale - set Gale permissions root
+    public static final String LEAF_ROOT = "leaf"; // Leaf - Leaf commands
 
     private CraftDefaultPermissions() {}
 
