From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 31 Jul 2024 22:05:21 +0800
Subject: [PATCH] Leaf Bootstrap

org.bukkit.craftbukkit.Main#main -> LeafBootstrap -> PaperBootstrap -> ...

diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 7675acdca2162e403f4ff523bcdd6efe0cb42d49..d10df0f81d922c3a5eca62afc376457fb8d4c197 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -239,7 +239,7 @@ public class Main {
                 System.setProperty("library.jansi.version", "Paper"); // Paper - set meaningless jansi version to prevent git builds from crashing on Windows
                 System.setProperty("jdk.console", "java.base"); // Paper - revert default console provider back to java.base so we can have our own jline
 
-                io.papermc.paper.PaperBootstrap.boot(options);
+                org.dreeam.leaf.LeafBootstrap.boot(options); // Leaf - Leaf Boostrap
             } catch (Throwable t) {
                 t.printStackTrace();
             }
