From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 27 Apr 2026 13:02:58 -0400
Subject: [PATCH] Virtual thread support for chat executor


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 0b35a293f3436bcf8a39c4f07a88340cd9d33bcb..d0a48f86544991ad6a23aedeefca15c017fc0a5b 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -312,9 +312,11 @@ public class GlobalConfiguration extends ConfigurationPart {
                     _chatExecutorMaxSize = _chatExecutorCoreSize;
                 }
 
+                if (!org.dreeam.leaf.config.modules.opt.VirtualThreadSupport.asyncChatExecutor) { // Leaf - Virtual thread support for chat executor
                 java.util.concurrent.ThreadPoolExecutor executor = (java.util.concurrent.ThreadPoolExecutor) net.minecraft.server.MinecraftServer.getServer().chatExecutor;
                 executor.setCorePoolSize(_chatExecutorCoreSize);
                 executor.setMaximumPoolSize(_chatExecutorMaxSize);
+                } // Leaf - Virtual thread support for chat executor
             }
         }
         public int maxJoinsPerTick = 5;
