From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 27 Apr 2026 13:02:58 -0400
Subject: [PATCH] Virtual thread support for chat executor


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 7d2b38ecf1478c9e53d033f41c4184864c8d32e4..0b174cba7f1a03cd38c8e2f208c81dedaac6e7f3 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -314,9 +314,11 @@ public class GlobalConfiguration extends ConfigurationPart {
                     _chatExecutorMaxSize = _chatExecutorCoreSize;
                 }
 
+                if (!org.dreeam.leaf.config.modules.opt.VirtualThreadSupport.asyncChatExecutor) { // Leaf - Virtual thread support for chat executor
                 java.util.concurrent.ThreadPoolExecutor executor = (java.util.concurrent.ThreadPoolExecutor) net.minecraft.server.MinecraftServer.getServer().chatExecutor;
                 executor.setCorePoolSize(_chatExecutorCoreSize);
                 executor.setMaximumPoolSize(_chatExecutorMaxSize);
+                } // Leaf - Virtual thread support for chat executor
             }
         }
         public int maxJoinsPerTick = 5;
