From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 27 Apr 2026 13:54:29 -0400
Subject: [PATCH] Virtual thread support for folia async scheduler


diff --git a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
index a1c153650b9146a245bf61ed3e89d407065da077..5959e059f5a06e0f53aa18bffef4e76f9b6d8e5a 100644
--- a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
@@ -21,27 +21,9 @@ import java.util.logging.Level;
 
 public final class FoliaAsyncScheduler implements AsyncScheduler {
 
-    private static final Logger LOGGER = LogUtils.getClassLogger();
+    public static final Logger LOGGER = LogUtils.getClassLogger(); // Leaf - Virtual thread support for folia async scheduler - private -> pubic
 
-    private final Executor executors = new ThreadPoolExecutor(Math.max(4, Runtime.getRuntime().availableProcessors() / 2), Integer.MAX_VALUE,
-        30L, TimeUnit.SECONDS, new SynchronousQueue<>(),
-        new ThreadFactory() {
-            private final AtomicInteger idGenerator = new AtomicInteger();
-
-            @Override
-            public Thread newThread(final Runnable run) {
-                final Thread ret = new Thread(run);
-
-                ret.setName("Folia Async Scheduler Thread #" + this.idGenerator.getAndIncrement());
-                ret.setPriority(Thread.NORM_PRIORITY - 1);
-                ret.setUncaughtExceptionHandler((final Thread thread, final Throwable thr) -> {
-                    LOGGER.error("Uncaught exception in thread: " + thread.getName(), thr);
-                });
-
-                return ret;
-            }
-        }
-    );
+    private final Executor executors = org.dreeam.leaf.async.ExecutorFactory.buildFoliaAsyncSchedulerExecutor(); // Leaf - Virtual thread support for folia async scheduler
 
     private final ScheduledExecutorService timerThread = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
         @Override
