From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 27 Apr 2026 14:03:48 -0400
Subject: [PATCH] Virtual thread support for paper configuration pool


diff --git a/src/main/java/io/papermc/paper/connection/PaperConfigurationTask.java b/src/main/java/io/papermc/paper/connection/PaperConfigurationTask.java
index 362056053164b2e4d682e04058b193c3078fd148..97dc88701b1e722262c71dc22d5fe0aac33cd1cd 100644
--- a/src/main/java/io/papermc/paper/connection/PaperConfigurationTask.java
+++ b/src/main/java/io/papermc/paper/connection/PaperConfigurationTask.java
@@ -13,10 +13,9 @@ import net.minecraft.server.network.ServerConfigurationPacketListenerImpl;
 import org.slf4j.Logger;
 
 public class PaperConfigurationTask implements ConfigurationTask {
-    private static final Logger LOGGER = LogUtils.getClassLogger();
+    public static final Logger LOGGER = LogUtils.getClassLogger(); // Leaf - Virtual thread support for paper configuration pool - private -> public
 
-    public static final ExecutorService CONFIGURATION_POOL = Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("Configuration Thread #%d")
-        .setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER)).build());
+    public static final ExecutorService CONFIGURATION_POOL = org.dreeam.leaf.async.ExecutorFactory.buildPaperConfigurationPool(); // Leaf - Virtual thread support for paper configuration pool
 
     public static final ConfigurationTask.Type TYPE = new ConfigurationTask.Type("paper_event_handling");
 
