From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <mc@jvavav.com>
Date: Sun, 27 Apr 2025 21:19:30 +0800
Subject: [PATCH] Async chunk sending


diff --git a/src/main/java/io/papermc/paper/antixray/BitStorageReader.java b/src/main/java/io/papermc/paper/antixray/BitStorageReader.java
index c27703775c845a8b0bd1b0cb8f05eb736d8a813c..7e72c910232b85b411ecba1fa0acdf7f81c85418 100644
--- a/src/main/java/io/papermc/paper/antixray/BitStorageReader.java
+++ b/src/main/java/io/papermc/paper/antixray/BitStorageReader.java
@@ -8,11 +8,18 @@ public final class BitStorageReader {
     private int longInBufferIndex;
     private int bitInLongIndex;
     private long current;
+    private int length; // Leaf
 
     public void setBuffer(byte[] buffer) {
         this.buffer = buffer;
     }
 
+    // Leaf start
+    public void setLength(int length) {
+        this.length = length;
+    }
+    // Leaf end
+
     public void setBits(int bits) {
         this.bits = bits;
         mask = (1 << bits) - 1;
@@ -25,7 +32,7 @@ public final class BitStorageReader {
     }
 
     private void init() {
-        if (buffer.length > longInBufferIndex + 7) {
+        if (length > longInBufferIndex + 7) { // Leaf
             current = ((((long) buffer[longInBufferIndex]) << 56)
                 | (((long) buffer[longInBufferIndex + 1] & 0xff) << 48)
                 | (((long) buffer[longInBufferIndex + 2] & 0xff) << 40)
diff --git a/src/main/java/io/papermc/paper/antixray/BitStorageWriter.java b/src/main/java/io/papermc/paper/antixray/BitStorageWriter.java
index 83412e0ddaade11eb7ac7b41bb8ae5b085802775..d521bd2275152575f5fe5038a817471026abccda 100644
--- a/src/main/java/io/papermc/paper/antixray/BitStorageWriter.java
+++ b/src/main/java/io/papermc/paper/antixray/BitStorageWriter.java
@@ -9,11 +9,18 @@ public final class BitStorageWriter {
     private int bitInLongIndex;
     private long current;
     private boolean dirty;
+    private int length; // Leaf
 
     public void setBuffer(byte[] buffer) {
         this.buffer = buffer;
     }
 
+    // Leaf start
+    public void setLength(int length) {
+        this.length = length;
+    }
+    // Leaf end
+
     public void setBits(int bits) {
         this.bits = bits;
         mask = (1L << bits) - 1;
@@ -26,7 +33,7 @@ public final class BitStorageWriter {
     }
 
     private void init() {
-        if (buffer.length > longInBufferIndex + 7) {
+        if (length > longInBufferIndex + 7) { // Leaf
             current = ((((long) buffer[longInBufferIndex]) << 56)
                 | (((long) buffer[longInBufferIndex + 1] & 0xff) << 48)
                 | (((long) buffer[longInBufferIndex + 2] & 0xff) << 40)
@@ -41,7 +48,7 @@ public final class BitStorageWriter {
     }
 
     public void flush() {
-        if (dirty && buffer.length > longInBufferIndex + 7) {
+        if (dirty && length > longInBufferIndex + 7) { // Leaf
             buffer[longInBufferIndex] = (byte) (current >> 56 & 0xff);
             buffer[longInBufferIndex + 1] = (byte) (current >> 48 & 0xff);
             buffer[longInBufferIndex + 2] = (byte) (current >> 40 & 0xff);
diff --git a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
index ee2d3a54d760f9c26542eab03c51651a30e279a0..bbf338e5e98a7b7d1c7cd986333df03f4aad4bb1 100644
--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
@@ -227,6 +227,8 @@ public final class ChunkPacketBlockControllerAntiXray extends ChunkPacketBlockCo
         boolean[] obfuscateTemp = null;
         bitStorageReader.setBuffer(chunkPacketInfoAntiXray.getBuffer());
         bitStorageWriter.setBuffer(chunkPacketInfoAntiXray.getBuffer());
+        bitStorageReader.setLength(chunkPacketInfoAntiXray.getLength());
+        bitStorageWriter.setLength(chunkPacketInfoAntiXray.getLength());
         int numberOfBlocks = presetBlockStateBits.length;
         // Keep the lambda expressions as simple as possible. They are used very frequently.
         LayeredIntSupplier random = numberOfBlocks == 1 ? (() -> 0) : engineMode == EngineMode.OBFUSCATE_LAYER ? new LayeredIntSupplier() {
diff --git a/src/main/java/io/papermc/paper/antixray/ChunkPacketInfo.java b/src/main/java/io/papermc/paper/antixray/ChunkPacketInfo.java
index a33a4d45d478ededff27244fcb910d3f369f2151..26f0e4b43c2adc27e0ca2bafb7a1bab3fcf8f6cc 100644
--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketInfo.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketInfo.java
@@ -13,6 +13,7 @@ public class ChunkPacketInfo<T> {
     private final int[] indexes;
     private final Object[][] presetValues;
     private byte[] buffer;
+    private int length; // Leaf
 
     public ChunkPacketInfo(ClientboundLevelChunkWithLightPacket chunkPacket, LevelChunk chunk) {
         this.chunkPacket = chunkPacket;
@@ -36,8 +37,19 @@ public class ChunkPacketInfo<T> {
         return buffer;
     }
 
+    // Leaf start
+    public int getLength() {
+        return length;
+    }
+    // Leaf end
+
     public void setBuffer(byte[] buffer) {
         this.buffer = buffer;
+        this.length = buffer.length; // Leaf
+    }
+
+    public void setLength(int length) {
+        this.length = length; // Leaf
     }
 
     public int getBits(int chunkSectionIndex) {
