From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 22 May 2025 20:56:06 +0900
Subject: [PATCH] Async target finding


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/set/OptimizedSmallEnumSet.java b/src/main/java/ca/spottedleaf/moonrise/common/set/OptimizedSmallEnumSet.java
index 4123edddc556c47f3f8d83523c125fd2e46b30e2..10dd3a4af3568d7aafe3a27113a0a8b6a8ac5a19 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/set/OptimizedSmallEnumSet.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/set/OptimizedSmallEnumSet.java
@@ -4,8 +4,8 @@ import java.util.Collection;
 
 public final class OptimizedSmallEnumSet<E extends Enum<E>> {
 
-    private final Class<E> enumClass;
-    private long backingSet;
+    // private final Class<E> enumClass; // Leaf
+    public long backingSet; // Leaf
 
     public OptimizedSmallEnumSet(final Class<E> clazz) {
         if (clazz == null) {
@@ -14,7 +14,7 @@ public final class OptimizedSmallEnumSet<E extends Enum<E>> {
         if (!clazz.isEnum()) {
             throw new IllegalArgumentException("Class must be enum, not " + clazz.getCanonicalName());
         }
-        this.enumClass = clazz;
+        // this.enumClass = clazz; // Leaf
     }
 
     public boolean addUnchecked(final E element) {
