From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 11 May 2025 00:37:44 +0200
Subject: [PATCH] Optimise player movement checks


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java b/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
index 7e85bb58b8a5de17fdab9ad3c574f2f80952e1ca..8e424b48c06e90893f0f2fde7c35410bd8271469 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
@@ -86,6 +86,11 @@ public abstract class SingleUserAreaMap<T> {
         if (fromX == NOT_SET) {
             return false;
         }
+        // Leaf start - Optimise player movement checks
+        if (org.dreeam.leaf.config.modules.opt.OptimizePlayerMovementProcessing.enabled && fromX == toX && fromZ == toZ && oldViewDistance == newViewDistance) {
+            return true;
+        }
+        // Leaf end - Optimise player movement checks
 
         this.lastChunkX = toX;
         this.lastChunkZ = toZ;
