From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <mc@jvavav.com>
Date: Fri, 2 May 2025 18:22:24 -0700
Subject: [PATCH] Async chunk sending


diff --git a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
index ca24f4cd7a50f0156d84e263c60f841cca95c669..eb78e79eef5fc92810b1b5b6f26fcc9cc66e6412 100644
--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
@@ -185,7 +185,7 @@ public final class ChunkPacketBlockControllerAntiXray extends ChunkPacketBlockCo
             return;
         }
 
-        if (!Bukkit.isPrimaryThread()) {
+        if (!Bukkit.isPrimaryThread() && !org.dreeam.leaf.async.chunk.AsyncChunkSend.POOL.isSameThread()) { // Leaf - Async chunk sending
             // Plugins?
             MinecraftServer.getServer().scheduleOnMain(() -> modifyBlocks(chunkPacket, chunkPacketInfo));
             return;
@@ -196,6 +196,7 @@ public final class ChunkPacketBlockControllerAntiXray extends ChunkPacketBlockCo
         int z = chunk.getPos().z;
         Level level = chunk.getLevel();
         ((ChunkPacketInfoAntiXray) chunkPacketInfo).setNearbyChunks(level.getChunkIfLoaded(x - 1, z), level.getChunkIfLoaded(x + 1, z), level.getChunkIfLoaded(x, z - 1), level.getChunkIfLoaded(x, z + 1));
+        if (org.dreeam.leaf.async.chunk.AsyncChunkSend.POOL.isSameThread()) { ((Runnable) chunkPacketInfo).run(); return; } // Leaf - Async chunk sending
         executor.execute((Runnable) chunkPacketInfo);
     }
 
