From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 3 Jun 2025 18:51:44 +0900
Subject: [PATCH] cache getBiome


diff --git a/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java b/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java
index f4c9f8fcc1dd0899186f7d940f4b68329c289b57..1dbbfa51226c69e61564d40a804a82c184d6f3ef 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/DelegatedLevelAccessor.java
@@ -265,6 +265,13 @@ public abstract class DelegatedLevelAccessor implements WorldGenLevel {
         return this.delegate.getBiome(pos);
     }
 
+    // Leaf start - cache getBiome
+    @Override
+    public Holder<Biome> getBiomeCached(net.minecraft.world.level.chunk.@org.jspecify.annotations.Nullable LevelChunk chunk, BlockPos pos) {
+        return this.delegate.getBiomeCached(chunk, pos);
+    }
+    // Leaf end - cache getBiome
+
     @Override
     public Stream<BlockState> getBlockStatesIfLoaded(AABB box) {
         return this.delegate.getBlockStatesIfLoaded(box);
