From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 31 Oct 2025 00:04:42 +0900
Subject: [PATCH] async chunk sender


diff --git a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockController.java b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockController.java
index 774a1b05212f3bdd9c68f1051d93b1e8dfcf3272..4a8141e3d91bd07cd2c7225044e5cdf938f6b2c5 100644
--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockController.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockController.java
@@ -36,6 +36,12 @@ public class ChunkPacketBlockController {
         chunkPacket.setReady(true);
     }
 
+    // Leaf start - async chunk sender
+    public void leaf$modifyBlocks(ClientboundLevelChunkWithLightPacket chunkPacket, ChunkPacketInfo<BlockState> chunkPacketInfo) {
+        chunkPacket.setReady(true);
+    }
+    // Leaf end - async chunk sender
+
     public void onBlockChange(Level level, BlockPos blockPos, BlockState newBlockState, BlockState oldBlockState, @Block.UpdateFlags int flags, int maxUpdateDepth) {
 
     }
diff --git a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
index 57e091df075809c3629ba22562053251037917b5..798723991cc500f88ef003b59f7bc7f9d546f9cd 100644
--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
@@ -200,6 +200,22 @@ public final class ChunkPacketBlockControllerAntiXray extends ChunkPacketBlockCo
         executor.execute((Runnable) chunkPacketInfo);
     }
 
+    // Leaf start - async chunk sender
+    @Override
+    public void leaf$modifyBlocks(ClientboundLevelChunkWithLightPacket chunkPacket, ChunkPacketInfo<BlockState> chunkPacketInfo) {
+        if (!(chunkPacketInfo instanceof ChunkPacketInfoAntiXray)) {
+            chunkPacket.setReady(true);
+            return;
+        }
+        LevelChunk chunk = chunkPacketInfo.getChunk();
+        int x = chunk.getPos().x();
+        int z = chunk.getPos().z();
+        Level level = chunk.getLevel();
+        ((ChunkPacketInfoAntiXray) chunkPacketInfo).setNearbyChunks(level.getChunkIfLoaded(x - 1, z), level.getChunkIfLoaded(x + 1, z), level.getChunkIfLoaded(x, z - 1), level.getChunkIfLoaded(x, z + 1));
+        ((Runnable) chunkPacketInfo).run();
+    }
+    // Leaf end - async chunk sender
+
     // Actually these fields should be variables inside the obfuscate method but in sync mode or with SingleThreadExecutor in async mode it's okay (even without ThreadLocal)
     // If an ExecutorService with multiple threads is used, ThreadLocal must be used here
     private final ThreadLocal<int[]> presetBlockStateBits = ThreadLocal.withInitial(() -> new int[getPresetBlockStatesFullLength()]);
