From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 7 Jun 2025 20:33:16 +0900
Subject: [PATCH] Dump pwt thread


diff --git a/src/main/java/org/spigotmc/WatchdogThread.java b/src/main/java/org/spigotmc/WatchdogThread.java
index 3e8b43794dd3a57eafb4b9f2fc206ae60d3a0b82..958e6d5dcf6899c5ec41f15f013dcaa208378217 100644
--- a/src/main/java/org/spigotmc/WatchdogThread.java
+++ b/src/main/java/org/spigotmc/WatchdogThread.java
@@ -124,6 +124,18 @@ public class WatchdogThread extends ca.spottedleaf.moonrise.common.util.TickThre
                 WatchdogThread.dumpThread(ManagementFactory.getThreadMXBean().getThreadInfo(MinecraftServer.getServer().serverThread.threadId(), Integer.MAX_VALUE), logger);
                 logger.log(Level.SEVERE, "------------------------------");
 
+                // Leaf start - dump pwt thread
+                if (org.dreeam.leaf.config.modules.async.SparklyPaperParallelWorldTicking.enabled) {
+                    logger.log(Level.SEVERE, "Parallel world ticking thread dump");
+                    for (Thread thread : org.apache.commons.lang3.ThreadUtils.getAllThreads()) {
+                        if (thread instanceof ServerLevelTickThread tickThread) {
+                            WatchdogThread.dumpThread(ManagementFactory.getThreadMXBean().getThreadInfo(tickThread.threadId(), Integer.MAX_VALUE), logger);
+                        }
+                    }
+                    logger.log(Level.SEVERE, "------------------------------");
+                }
+                // Leaf end - dump pwt thread
+
                 // Paper start - Only print full dump on long timeouts
                 if (isLongTimeout) {
                     logger.log(Level.SEVERE, "Entire Thread Dump:");
