From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 24 Apr 2025 16:36:16 -0400
Subject: [PATCH] Paw optimization

Some random optimizations

diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
index f4e3f9b27ede2df802de9912ce514a54bfe08e3c..c0a61b7ed4dd8c2e8862f00404d8ac97de01acdc 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
@@ -208,12 +208,32 @@ public class CraftBlockData implements BlockData {
         StringBuilder stateString = new StringBuilder(BuiltInRegistries.BLOCK.getKey(this.state.getBlock()).toString());
         if (!singleton) {
             stateString.append('[');
-            stateString.append(states.map(Property.Value::toString).collect(Collectors.joining(",")));
+            // Leaf start - paw optimization
+            java.util.Iterator<?> iterator = states.iterator();
+            while (iterator.hasNext()) {
+                Object val = iterator.next();
+                if (val == null) {
+                    stateString.append("<NULL>");
+                } else {
+                    stateString.append(val.toString());
+                }
+
+                if (iterator.hasNext()) {
+                    stateString.append(',');
+                }
+            }
+            // Leaf end - paw optimization
             stateString.append(']');
         }
         return stateString.toString();
     }
 
+    // Leaf start - paw optimization
+    private <T extends Comparable<T>> String getValueName(Property<T> property, Comparable<?> value) {
+        return property.getName((T) value);
+    }
+    // Leaf end - paw optimization
+
     public Map<String, String> toStates(boolean hideUnspecified) {
         Map<String, String> serializedStates = new HashMap<>();
         if (hideUnspecified && this.parsedStates != null) {
