From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 3 Jun 2025 18:51:44 +0900
Subject: [PATCH] cache getBiome


diff --git a/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java b/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java
index bb117482fe9d7ca5a6931cbf740cd553b40acda5..514ef7471b8da5fd8538f251c46ab215a3e7a48f 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java
@@ -266,6 +266,13 @@ public abstract class DelegatedGeneratorAccess implements WorldGenLevel {
         return this.delegate.getBiome(pos);
     }
 
+    // Leaf start - cache getBiome
+    @Override
+    public Holder<Biome> getBiomeCached(net.minecraft.world.level.chunk.@org.jspecify.annotations.Nullable LevelChunk chunk, BlockPos pos) {
+        return this.delegate.getBiomeCached(chunk, pos);
+    }
+    // Leaf end - cache getBiome
+
     @Override
     public Stream<BlockState> getBlockStatesIfLoaded(AABB box) {
         return this.delegate.getBlockStatesIfLoaded(box);
