From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sun, 25 Jan 2026 20:50:29 +0900
Subject: [PATCH] cache entity isAlive


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 1742395c42d7844ae3b66ca4bfab515a7ef52c08..3296eb75cc3ce0a97ad63337db0041514e056a04 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2695,6 +2695,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     public void setRealHealth(double health) {
         if (Double.isNaN(health)) return; // Paper - Check for NaN
         this.health = health;
+        this.getHandle().leaf$onAliveDirty(); // Leaf - cache entity isAlive
     }
 
     public void updateScaledHealth() {
@@ -2712,6 +2713,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         this.getHandle().getEntityData().set(net.minecraft.world.entity.LivingEntity.DATA_HEALTH_ID, (float) this.getScaledHealth());
 
         this.getHandle().maxHealthCache = this.getMaxHealth();
+        this.getHandle().leaf$onAliveDirty(); // Leaf - cache entity isAlive
     }
 
     @Override
