From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Tue, 8 Jul 2025 12:08:25 +0200
Subject: [PATCH] Toggleable async catcher


diff --git a/src/main/java/org/spigotmc/AsyncCatcher.java b/src/main/java/org/spigotmc/AsyncCatcher.java
index 7e7a0ed944961af5ab7a49bc659da2862d2e2c82..e584205c3e1412bf2f2c4e4eb114a29cebc73fad 100644
--- a/src/main/java/org/spigotmc/AsyncCatcher.java
+++ b/src/main/java/org/spigotmc/AsyncCatcher.java
@@ -5,6 +5,7 @@ import net.minecraft.server.MinecraftServer;
 public class AsyncCatcher {
 
     public static void catchOp(String reason) {
+        if (!org.dreeam.leaf.config.modules.misc.AsyncCatcherConfig.enabled) return; // Leaf - Toggleable async catcher
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) { // Paper - chunk system
             MinecraftServer.LOGGER.error("Thread {} failed main thread check: {}", Thread.currentThread().getName(), reason, new Throwable()); // Paper
             throw new IllegalStateException("Asynchronous " + reason + "!");
