From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 19 Jun 2025 07:37:19 +0800
Subject: [PATCH] Update spigot mapping to compat spigot reobf mapping plugins


diff --git a/src/main/java/io/papermc/paper/util/MappingEnvironment.java b/src/main/java/io/papermc/paper/util/MappingEnvironment.java
index d22e477e3b1cebb3ffc5b96261b8d4fb4ef5b240..b9d5ee2f72993f1a9d65a55988f512b6e9c5f1d8 100644
--- a/src/main/java/io/papermc/paper/util/MappingEnvironment.java
+++ b/src/main/java/io/papermc/paper/util/MappingEnvironment.java
@@ -11,7 +11,7 @@ import org.checkerframework.framework.qual.DefaultQualifier;
 @DefaultQualifier(NonNull.class)
 public final class MappingEnvironment {
     public static final boolean DISABLE_PLUGIN_REMAPPING = Boolean.getBoolean("paper.disablePluginRemapping");
-    public static final String LEGACY_CB_VERSION = "v1_21_R4";
+    public static final String LEGACY_CB_VERSION = "v1_21_R5"; // Leaf - Update spigot mapping to compat spigot reobf mapping plugins
     private static final @Nullable String MAPPINGS_HASH = readMappingsHash();
     private static final boolean REOBF = checkReobf();
 
