From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Thu, 4 Mar 2077 00:00:00 +0800
Subject: [PATCH] Leaf Test: Async Executor


diff --git a/src/test/java/gg/pufferfish/pufferfish/AsyncExecutorTest.java b/src/test/java/gg/pufferfish/pufferfish/AsyncExecutorTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..a057661103ff579acb2c613ddee6f1ed95226253
--- /dev/null
+++ b/src/test/java/gg/pufferfish/pufferfish/AsyncExecutorTest.java
@@ -0,0 +1,59 @@
+// Leaf Test - Async Executor
+
+package gg.pufferfish.pufferfish;
+
+import gg.pufferfish.pufferfish.util.AsyncExecutor;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import org.bukkit.support.environment.LeafTest;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+@LeafTest
+public class AsyncExecutorTest {
+
+    AsyncExecutor mobSpawnExecutor = new AsyncExecutor("Leaf Async Mob Spawn Test Thread");
+
+    @Test
+    public void testTaskSubmission() throws InterruptedException {
+        mobSpawnExecutor.start();
+        int submittedTaskCount = 0;
+        AtomicInteger completedTaskCount = new AtomicInteger(0);
+        System.out.println("Submitting 1");
+        submittedTaskCount++;
+        mobSpawnExecutor.submit(() -> {
+            try {
+                System.out.println(" -> 1...");
+                Thread.sleep(500);
+                System.out.println(" -> 1 done");
+                completedTaskCount.incrementAndGet();
+            } catch (InterruptedException ignored) {
+            }
+        });
+
+        Thread.sleep(100);
+
+        System.out.println("Submitting...");
+        submittedTaskCount++;
+        mobSpawnExecutor.submit(() -> {
+            System.out.println(" -> " + completedTaskCount.incrementAndGet());
+        });
+        submittedTaskCount++;
+        mobSpawnExecutor.submit(() -> {
+            System.out.println(" -> " + completedTaskCount.incrementAndGet());
+        });
+        submittedTaskCount++;
+        mobSpawnExecutor.submit(() -> {
+            System.out.println(" -> " + completedTaskCount.incrementAndGet());
+        });
+
+        Thread.sleep(2000);
+        mobSpawnExecutor.join(9999);
+
+        final int submittedTaskCountFinal = submittedTaskCount;
+        final int completedTaskCountFinal = completedTaskCount.get();
+
+        Assertions.assertEquals(submittedTaskCountFinal, completedTaskCountFinal, "Submitted (" + submittedTaskCountFinal + ") tasks, but only (" +  completedTaskCountFinal + ") tasks completed." );
+    }
+}
