From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 8 Jul 2025 16:36:53 +0900
Subject: [PATCH] 1


diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index bef8d6e47969a033a273395b0f5b0b9f5bfb7121..b9c159537c7db0ff85ebb0ff6b510e5b5061f5f9 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -42,12 +42,6 @@ class PaperEventManager {
         if (event.isAsynchronous() && this.server.isPrimaryThread()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
         } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
-            // Leaf start - Multithreaded tracker
-            if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled && org.dreeam.leaf.config.modules.async.MultithreadedTracker.allowRescheduleEvent && Thread.currentThread() instanceof org.dreeam.leaf.async.tracker.MultithreadedTracker.MultithreadedTrackerThread) {
-                net.minecraft.server.MinecraftServer.getServer().scheduleOnMain(event::callEvent);
-                return;
-            }
-            // Leaf end - Multithreaded tracker
             // Leaf start - Asynchronous locator
             if (org.dreeam.leaf.config.modules.async.AsyncLocator.enabled) {
                 net.minecraft.server.MinecraftServer.getServer().scheduleOnMain(event::callEvent);
