From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 8 Jul 2025 16:36:59 +0900
Subject: [PATCH] 1


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index acb5182a59859f65653cd309ca85b8301ace5383..cf53241af01140476ee3097960d5ff27f440b632 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1833,26 +1833,6 @@ public class CraftEventFactory {
     }
 
     public static boolean handleBlockFormEvent(Level world, BlockPos pos, net.minecraft.world.level.block.state.BlockState state, int flags, @Nullable Entity entity, boolean checkSetResult) {
-        // Leaf start - Multithreaded tracker
-        if (org.dreeam.leaf.config.modules.async.MultithreadedTracker.enabled && Thread.currentThread() instanceof org.dreeam.leaf.async.tracker.MultithreadedTracker.MultithreadedTrackerThread) {
-            java.util.concurrent.CompletableFuture<Boolean> future = new java.util.concurrent.CompletableFuture<>();
-            net.minecraft.server.MinecraftServer.getServer().scheduleOnMain(() -> {
-                boolean resultFlag = false;
-                CraftBlockState snapshot = CraftBlockStates.getBlockState(world, pos);
-                snapshot.setData(state);
-
-                BlockFormEvent event = (entity == null) ? new BlockFormEvent(snapshot.getBlock(), snapshot) : new EntityBlockFormEvent(entity.getBukkitEntity(), snapshot.getBlock(), snapshot);
-                if (event.callEvent()) {
-                    boolean result = snapshot.place(flags);
-                    resultFlag = !checkSetResult || result;
-                }
-
-                future.complete(resultFlag);
-            });
-
-            return future.join();
-        }
-        // Leaf end - Multithreaded tracker
         CraftBlockState snapshot = CraftBlockStates.getBlockState(world, pos);
         snapshot.setData(state);
 
