From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Github Actions <artyom030803@gmail.com>
Date: Wed, 25 Mar 2026 08:48:21 +0300
Subject: [PATCH] Use optimized collections in craftbukkit classes


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
index ac56dbb9db9b49cd856f2e2567a7ac1be55293ef..aacfea554dfde170b7039cd3d70f689aafdb3429 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
@@ -1,7 +1,10 @@
 package org.bukkit.craftbukkit.block;
 
 import com.google.common.base.Preconditions;
+
+import java.util.EnumMap;
 import java.util.HashMap;
+import java.util.IdentityHashMap;
 import java.util.Map;
 import java.util.function.BiFunction;
 import javax.annotation.Nullable;
@@ -68,7 +71,7 @@ public final class CraftBlockStates {
         }
     }
 
-    private static final Map<Material, BlockStateFactory<?>> FACTORIES = new HashMap<>();
+    private static final Map<Material, BlockStateFactory<?>> FACTORIES = new EnumMap<>(Material.class); // Leaf - use optimized collections
     private static final BlockStateFactory<?> DEFAULT_FACTORY = new BlockStateFactory<>(CraftBlockState.class) {
         @Override
         public CraftBlockState createBlockState(World world, BlockPos pos, net.minecraft.world.level.block.state.BlockState state, BlockEntity blockEntity) {
@@ -78,7 +81,7 @@ public final class CraftBlockStates {
         }
     };
 
-    private static final Map<BlockEntityType<?>, BlockStateFactory<?>> FACTORIES_BY_BLOCK_ENTITY_TYPE = new HashMap<>();
+    private static final Map<BlockEntityType<?>, BlockStateFactory<?>> FACTORIES_BY_BLOCK_ENTITY_TYPE = new IdentityHashMap<>(); // Leaf - use optimized collections
     private static void register(BlockEntityType<?> type, BlockStateFactory<?> factory) {
         FACTORIES_BY_BLOCK_ENTITY_TYPE.put(type, factory);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java b/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
index 8430d131bbc12e730b3d3ea16a152fcd6488a697..82920aad0ab769bc43ddfad5d4ddb001d09159e9 100644
--- a/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
+++ b/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
@@ -2,6 +2,8 @@ package org.bukkit.craftbukkit.boss;
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
+
+import java.util.EnumMap;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -47,7 +49,7 @@ public class CraftBossBar implements BossBar {
     }
 
     private void initialize() {
-        this.flags = new HashMap<>();
+        this.flags = new EnumMap<>(BarFlag.class); // Leaf - use optimized collections
         this.flags.put(BarFlag.DARKEN_SKY, new FlagContainer(this.handle::shouldDarkenScreen, this.handle::setDarkenScreen));
         this.flags.put(BarFlag.PLAY_BOSS_MUSIC, new FlagContainer(this.handle::shouldPlayBossMusic, this.handle::setPlayBossMusic));
         this.flags.put(BarFlag.CREATE_FOG, new FlagContainer(this.handle::shouldCreateWorldFog, this.handle::setCreateWorldFog));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
index db1a81d0661305acdf22a9a8ff07ef12ae1d7003..6b4c74ffd66649d4854661dda3e674c90cbd2ef7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
@@ -1,10 +1,8 @@
 package org.bukkit.craftbukkit.entity;
 
 import com.google.common.base.Preconditions;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
+
+import java.util.*;
 import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 import java.util.function.Function;
@@ -271,8 +269,9 @@ public final class CraftEntityTypes {
         }
     };
     // Paper end - respect randomizeData
-    private static final Map<Class<?>, EntityTypeData<?, ?>> CLASS_TYPE_DATA = new HashMap<>();
-    private static final Map<EntityType, EntityTypeData<?, ?>> ENTITY_TYPE_DATA = new HashMap<>();
+    // Leaf - use optimized collections
+    private static final Map<Class<?>, EntityTypeData<?, ?>> CLASS_TYPE_DATA = new IdentityHashMap<>();
+    private static final Map<EntityType, EntityTypeData<?, ?>> ENTITY_TYPE_DATA = new EnumMap<>(EntityType.class);
 
     static {
         // Living
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
index a456cc637e2a990354852d8537787af9cd7eb670..d52b4dae7aeaa1e06532051747855bc42bb0cfc7 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
@@ -1,17 +1,18 @@
 package org.bukkit.craftbukkit.inventory.util;
 
-import java.util.HashMap;
-import java.util.Map;
 import org.bukkit.event.inventory.InventoryType;
 import org.bukkit.inventory.Inventory;
 import org.bukkit.inventory.InventoryHolder;
 
+import java.util.EnumMap;
+import java.util.Map;
+
 public final class CraftInventoryCreator {
 
     public static final CraftInventoryCreator INSTANCE = new CraftInventoryCreator();
 
     private final CraftCustomInventoryConverter defaultConverter = new CraftCustomInventoryConverter();
-    private final Map<InventoryType, InventoryConverter> converterMap = new HashMap<>();
+    private final Map<InventoryType, InventoryConverter> converterMap = new EnumMap<>(InventoryType.class); // Leaf - use optimized collections
 
     private CraftInventoryCreator() {
         this.converterMap.put(InventoryType.CHEST, this.defaultConverter);
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 900a50ed18e931cd090d970634fa27726926af07..94a1d3d05fa85045ec67fdb3e5c6755e0f722ef0 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -19,13 +19,7 @@ import io.papermc.paper.registry.RegistryKey;
 import java.io.File;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.LinkedHashMap;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 import java.util.logging.Level;
 import java.util.stream.Stream;
 import net.kyori.adventure.text.event.HoverEvent;
@@ -162,10 +156,11 @@ public final class CraftMagicNumbers implements UnsafeValues {
     }
 
     // ========================================================================
-    private static final Map<Block, Material> BLOCK_MATERIAL = new HashMap<>();
-    private static final Map<Item, Material> ITEM_MATERIAL = new HashMap<>();
-    private static final Map<Material, Item> MATERIAL_ITEM = new HashMap<>();
-    private static final Map<Material, Block> MATERIAL_BLOCK = new HashMap<>();
+    // Leaf - use optimized collections
+    private static final Map<Block, Material> BLOCK_MATERIAL = new IdentityHashMap<>();
+    private static final Map<Item, Material> ITEM_MATERIAL = new IdentityHashMap<>();
+    private static final Map<Material, Item> MATERIAL_ITEM = new EnumMap<>(Material.class);
+    private static final Map<Material, Block> MATERIAL_BLOCK = new EnumMap<>(Material.class);
 
     static {
         for (Block block : BuiltInRegistries.BLOCK) {
