From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 11 Sep 2025 16:33:55 +0900
Subject: [PATCH] entity activation


diff --git a/src/main/java/io/papermc/paper/entity/activation/ActivationType.java b/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
index 493914c8d14a4f4b58e57ca0e2d3723e6fd4c364..88d36a5d54ab6aaeab1e41ec8372129a811a95d2 100644
--- a/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
+++ b/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
@@ -21,6 +21,7 @@ public enum ActivationType {
     RAIDER,
     MISC;
 
+    @Deprecated // Leaf - entity activation
     AABB boundingBox = new AABB(0, 0, 0, 0, 0, 0);
 
     /**
