From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 26 Aug 2025 19:59:41 +0900
Subject: [PATCH] optimize isAlive


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 1fd4bea31ed92903a3727b1f311dfaa37d5bc684..0d295eb55bfd3f4f93742f8174daf2c9ca8c5805 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2802,6 +2802,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     public void setRealHealth(double health) {
         if (Double.isNaN(health)) return; // Paper - Check for NaN
         this.health = health;
+        this.getHandle().leaf$onAliveDirty(); // Leaf - optimize isAlive
     }
 
     public void updateScaledHealth() {
@@ -2819,6 +2820,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         this.getHandle().getEntityData().set(net.minecraft.world.entity.LivingEntity.DATA_HEALTH_ID, (float) this.getScaledHealth());
 
         this.getHandle().maxHealthCache = this.getMaxHealth();
+        this.getHandle().leaf$onAliveDirty(); // Leaf - optimize isAlive
     }
 
     @Override
