From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Wed, 25 Mar 2026 08:48:21 +0300
Subject: [PATCH] Use optimized collections in CB classes


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
index ac56dbb9db9b49cd856f2e2567a7ac1be55293ef..116750abadb9102c309a6bf558a8d6047e2eda53 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockStates.java
@@ -68,7 +68,7 @@ public final class CraftBlockStates {
         }
     }
 
-    private static final Map<Material, BlockStateFactory<?>> FACTORIES = new HashMap<>();
+    private static final Map<Material, BlockStateFactory<?>> FACTORIES = new java.util.EnumMap<>(Material.class); // Leaf - Use optimized collections in CB classes
     private static final BlockStateFactory<?> DEFAULT_FACTORY = new BlockStateFactory<>(CraftBlockState.class) {
         @Override
         public CraftBlockState createBlockState(World world, BlockPos pos, net.minecraft.world.level.block.state.BlockState state, BlockEntity blockEntity) {
@@ -78,7 +78,7 @@ public final class CraftBlockStates {
         }
     };
 
-    private static final Map<BlockEntityType<?>, BlockStateFactory<?>> FACTORIES_BY_BLOCK_ENTITY_TYPE = new HashMap<>();
+    private static final Map<BlockEntityType<?>, BlockStateFactory<?>> FACTORIES_BY_BLOCK_ENTITY_TYPE = new java.util.IdentityHashMap<>(); // Leaf - Use optimized collections in CB classes
     private static void register(BlockEntityType<?> type, BlockStateFactory<?> factory) {
         FACTORIES_BY_BLOCK_ENTITY_TYPE.put(type, factory);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java b/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
index 8430d131bbc12e730b3d3ea16a152fcd6488a697..cc3dbe4f6ec338a2552b9496234415139510e8e2 100644
--- a/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
+++ b/src/main/java/org/bukkit/craftbukkit/boss/CraftBossBar.java
@@ -47,7 +47,7 @@ public class CraftBossBar implements BossBar {
     }
 
     private void initialize() {
-        this.flags = new HashMap<>();
+        this.flags = new java.util.EnumMap<>(BarFlag.class); // Leaf - Use optimized collections in craftbukkit classes
         this.flags.put(BarFlag.DARKEN_SKY, new FlagContainer(this.handle::shouldDarkenScreen, this.handle::setDarkenScreen));
         this.flags.put(BarFlag.PLAY_BOSS_MUSIC, new FlagContainer(this.handle::shouldPlayBossMusic, this.handle::setPlayBossMusic));
         this.flags.put(BarFlag.CREATE_FOG, new FlagContainer(this.handle::shouldCreateWorldFog, this.handle::setCreateWorldFog));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
index db1a81d0661305acdf22a9a8ff07ef12ae1d7003..8107780ce4950c37cf2bfddd50604a4e27cb05b8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityTypes.java
@@ -271,8 +271,10 @@ public final class CraftEntityTypes {
         }
     };
     // Paper end - respect randomizeData
-    private static final Map<Class<?>, EntityTypeData<?, ?>> CLASS_TYPE_DATA = new HashMap<>();
-    private static final Map<EntityType, EntityTypeData<?, ?>> ENTITY_TYPE_DATA = new HashMap<>();
+    // Leaf start - Use optimized collections in CB classes
+    private static final Map<Class<?>, EntityTypeData<?, ?>> CLASS_TYPE_DATA = new java.util.IdentityHashMap<>();
+    private static final Map<EntityType, EntityTypeData<?, ?>> ENTITY_TYPE_DATA = new java.util.EnumMap<>(EntityType.class);
+    // Leaf end - Use optimized collections in CB classes
 
     static {
         // Living
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
index a456cc637e2a990354852d8537787af9cd7eb670..82938d46c8e01461337aab483a9718e2679ddf7c 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/util/CraftInventoryCreator.java
@@ -11,7 +11,7 @@ public final class CraftInventoryCreator {
     public static final CraftInventoryCreator INSTANCE = new CraftInventoryCreator();
 
     private final CraftCustomInventoryConverter defaultConverter = new CraftCustomInventoryConverter();
-    private final Map<InventoryType, InventoryConverter> converterMap = new HashMap<>();
+    private final Map<InventoryType, InventoryConverter> converterMap = new java.util.EnumMap<>(InventoryType.class); // Leaf - Use optimized collections in CB classes
 
     private CraftInventoryCreator() {
         this.converterMap.put(InventoryType.CHEST, this.defaultConverter);
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 900a50ed18e931cd090d970634fa27726926af07..60be852fbecf054852b3cb9089976b7915138426 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -162,10 +162,12 @@ public final class CraftMagicNumbers implements UnsafeValues {
     }
 
     // ========================================================================
-    private static final Map<Block, Material> BLOCK_MATERIAL = new HashMap<>();
-    private static final Map<Item, Material> ITEM_MATERIAL = new HashMap<>();
-    private static final Map<Material, Item> MATERIAL_ITEM = new HashMap<>();
-    private static final Map<Material, Block> MATERIAL_BLOCK = new HashMap<>();
+    // Leaf start - Use optimized collections in CB classes
+    private static final Map<Block, Material> BLOCK_MATERIAL = new java.util.IdentityHashMap<>();
+    private static final Map<Item, Material> ITEM_MATERIAL = new java.util.IdentityHashMap<>();
+    private static final Map<Material, Item> MATERIAL_ITEM = new java.util.EnumMap<>(Material.class);
+    private static final Map<Material, Block> MATERIAL_BLOCK = new java.util.EnumMap<>(Material.class);
+    // Leaf end - Use optimized collections in CB classes
 
     static {
         for (Block block : BuiltInRegistries.BLOCK) {
