From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 7 Aug 2025 00:22:50 +0800
Subject: [PATCH] Bump netty to 4.2.x


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 748bd9650da4a209743b7a5dde584b2e19c5a578..3e972dc2160f953daaa727c2305ec71eed7981a5 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -21,6 +21,14 @@ public class Main {
     // Paper start - Reset loggers after shutdown
     static {
         System.setProperty("java.util.logging.manager", "io.papermc.paper.log.CustomLogManager");
+
+        // Leaf start - Bump netty to 4.2.x
+        // Fallback to original pooled allocator if no explicitly define adaptive allocator
+        // To follow https://github.com/netty/netty/wiki/Netty-4.2-Migration-Guide
+        if (System.getProperty("io.netty.allocator.type") == null) {
+            System.setProperty("io.netty.allocator.type", "pooled");
+        }
+        // Leaf end - Bump netty to 4.2.x
     }
     // Paper end - Reset loggers after shutdown
 
