From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 1 Aug 2026 08:54:06 +0900
Subject: [PATCH] cache biome zoom seed


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index bcf8ec090cfd1faed825bfa7bc012e719c7f8da4..22e244d63d13dffc1efa091ac8853a4fb224a958 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1296,7 +1296,7 @@ public final class CraftServer implements Server {
             net.minecraft.server.Main.forceUpgrade(this.console.storageSource, DataFixers.getDataFixer(), this.console.options.has("eraseCache"), () -> true, contextRegistryAccess, this.console.options.has("recreateRegionFiles"));
         }
 
-        long biomeZoomSeed = BiomeManager.obfuscateSeed(genSettingsFinal.options().seed());
+        long biomeZoomSeed = genSettingsFinal.options().biomeZoomSeed(); // Leaf - cache biome zoom seed
         LevelStem customStem = genSettingsFinal.dimensions().get(actualDimension).orElse(null);
         if (customStem == null) {
             customStem = levelStemRegistry.getValue(actualDimension);
