From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 29 Apr 2024 14:18:58 -0400
Subject: [PATCH] Fix Pufferfish and Purpur patches


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 374d39057f4913ff9586205ee2cf305a99279bae..9efc174b48afe8548f5be11805c334e50173cf4d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -424,14 +424,16 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         BlockPos pos = CraftLocation.toBlockPos(location);
         // Paper start
         MenuProvider menuProvider = Blocks.ENCHANTING_TABLE.defaultBlockState().getMenuProvider(this.getHandle().level(), pos);
-        if (menuProvider == null) {
-            if (!force) {
-                return null;
-            }
+        // Leaf start - Fix Pufferfish and Purpur patches - Fix persistent enchanting table lapis duplication
+        if (force) {
+            Component title = menuProvider == null ? Component.translatable("container.enchant") : menuProvider.getDisplayName();
             menuProvider = new net.minecraft.world.SimpleMenuProvider((syncId, inventory, player) -> {
-                return new net.minecraft.world.inventory.EnchantmentMenu(syncId, inventory, net.minecraft.world.inventory.ContainerLevelAccess.create(this.getHandle().level(), pos));
-            }, Component.translatable("container.enchant"));
+                return new net.minecraft.world.inventory.EnchantmentMenu(syncId, inventory, net.minecraft.world.inventory.ContainerLevelAccess.create(this.getHandle().level(), pos), false);
+            }, title);
+        } else if (menuProvider == null) {
+            return null;
         }
+        // Leaf end - Fix Pufferfish and Purpur patches - Fix persistent enchanting table lapis duplication
         this.getHandle().openMenu(menuProvider);
         // Paper end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/view/builder/CraftEnchantmentInventoryViewBuilder.java b/src/main/java/org/bukkit/craftbukkit/inventory/view/builder/CraftEnchantmentInventoryViewBuilder.java
index 6f502941596a0f909d28d85356373fea6adb7539..518339cf7f7e1b088c8f7209bef7c0120c400ddc 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/view/builder/CraftEnchantmentInventoryViewBuilder.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/view/builder/CraftEnchantmentInventoryViewBuilder.java
@@ -25,7 +25,7 @@ public class CraftEnchantmentInventoryViewBuilder extends CraftAbstractLocationI
         if (this.position == null) {
             this.position = player.blockPosition();
             super.defaultTitle = new EnchantingTableBlockEntity(this.position, Blocks.ENCHANTING_TABLE.defaultBlockState()).getDisplayName();
-            return new EnchantmentMenu(player.nextContainerCounter(), player.getInventory(), ContainerLevelAccess.create(this.world, this.position));
+            return new EnchantmentMenu(player.nextContainerCounter(), player.getInventory(), ContainerLevelAccess.create(this.world, this.position), this.checkReachable); // Leaf - Fix Pufferfish and Purpur patches - Fix persistent enchanting table lapis duplication
         }
 
         final BlockEntity entity = this.world.getBlockEntity(position);
@@ -35,6 +35,6 @@ public class CraftEnchantmentInventoryViewBuilder extends CraftAbstractLocationI
             super.defaultTitle = new EnchantingTableBlockEntity(this.position, Blocks.ENCHANTING_TABLE.defaultBlockState()).getDisplayName();
         }
 
-        return new EnchantmentMenu(player.nextContainerCounter(), player.getInventory(), ContainerLevelAccess.create(this.world, this.position));
+        return new EnchantmentMenu(player.nextContainerCounter(), player.getInventory(), ContainerLevelAccess.create(this.world, this.position), this.checkReachable); // Leaf - Fix Pufferfish and Purpur patches - Fix persistent enchanting table lapis duplication
     }
 }
