From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Tue, 28 Apr 2026 12:59:23 +0300
Subject: [PATCH] remove unnecessary getBukkitEntity calls


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 138d8c1f78dea434cb7f1b8b1e39bdf38e55e58a..0a0d772fb4739dcdc88e2b276d11c161e2afb34a 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -732,26 +732,32 @@ public class CraftEventFactory {
             event = CraftEventFactory.callCreatureSpawnEvent((net.minecraft.world.entity.LivingEntity) entity, spawnReason);
         } else if (entity instanceof ItemEntity) {
             event = CraftEventFactory.callItemSpawnEvent((ItemEntity) entity);
-        } else if (entity.getBukkitEntity() instanceof org.bukkit.entity.Projectile) {
-            // Not all projectiles extend EntityProjectile, so check for Bukkit interface instead
-            event = CraftEventFactory.callProjectileLaunchEvent(entity);
-        } else if (entity.getBukkitEntity() instanceof org.bukkit.entity.Vehicle) {
-            event = CraftEventFactory.callVehicleCreateEvent(entity);
-        } else if (entity.getBukkitEntity() instanceof org.bukkit.entity.LightningStrike) {
-            LightningStrikeEvent.Cause cause = switch (spawnReason) {
-                case COMMAND -> LightningStrikeEvent.Cause.COMMAND;
-                case CUSTOM -> LightningStrikeEvent.Cause.CUSTOM;
-                case SPAWNER -> LightningStrikeEvent.Cause.SPAWNER;
-                default -> LightningStrikeEvent.Cause.UNKNOWN;
-            };
-            // This event is called in nms-patches for common causes like Weather, Trap or Trident (SpawnReason.DEFAULT) then can ignore this cases for avoid two calls to this event
-            if (cause == LightningStrikeEvent.Cause.UNKNOWN && spawnReason == SpawnReason.DEFAULT) {
-                return true;
-            }
-            event = CraftEventFactory.callLightningStrikeEvent((LightningStrike) entity.getBukkitEntity(), cause);
+        // Leaf start - only get bukkit entity once
         } else if (!(entity instanceof ServerPlayer)) {
-            event = CraftEventFactory.callEntitySpawnEvent(entity);
+            org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity();
+
+            switch (bukkitEntity) {
+                case Projectile projectile ->
+                    // Not all projectiles extend EntityProjectile, so check for Bukkit interface instead
+                    event = CraftEventFactory.callProjectileLaunchEvent(projectile);
+                case Vehicle vehicle -> event = CraftEventFactory.callVehicleCreateEvent(vehicle);
+                case LightningStrike lightningStrike -> {
+                    LightningStrikeEvent.Cause cause = switch (spawnReason) {
+                        case COMMAND -> LightningStrikeEvent.Cause.COMMAND;
+                        case CUSTOM -> LightningStrikeEvent.Cause.CUSTOM;
+                        case SPAWNER -> LightningStrikeEvent.Cause.SPAWNER;
+                        default -> LightningStrikeEvent.Cause.UNKNOWN;
+                    };
+                    // This event is called in nms-patches for common causes like Weather, Trap or Trident (SpawnReason.DEFAULT) then can ignore this cases for avoid two calls to this event
+                    if (cause == LightningStrikeEvent.Cause.UNKNOWN && spawnReason == SpawnReason.DEFAULT) {
+                        return true;
+                    }
+                    event = CraftEventFactory.callLightningStrikeEvent(lightningStrike, cause);
+                }
+                default -> event = CraftEventFactory.callEntitySpawnEvent(bukkitEntity);
+            }
         }
+        // Leaf start - only get bukkit entity once
 
         if (event != null && (event.isCancelled() || entity.isRemoved())) {
             Entity vehicle = entity.getVehicle();
@@ -768,7 +774,7 @@ public class CraftEventFactory {
         // Spigot start - SPIGOT-7523: Merge after spawn event and only merge if the event was not cancelled (gets checked above)
         if (entity instanceof net.minecraft.world.entity.ExperienceOrb xp) {
             double radius = world.spigotConfig.expMerge;
-            event = CraftEventFactory.callEntitySpawnEvent(entity); // Call spawn event for ExperienceOrb entities
+            event = CraftEventFactory.callEntitySpawnEvent(xp.getBukkitEntity()); // Call spawn event for ExperienceOrb entities
             if (radius > 0 && !event.isCancelled() && !entity.isRemoved()) {
                 // Paper start - Maximum exp value when merging; Whole section has been tweaked, see comments for specifics
                 final long maxValue = world.paperConfig().entities.behavior.experienceMergeMaxValue;
@@ -800,9 +806,7 @@ public class CraftEventFactory {
         return true;
     }
 
-    public static EntitySpawnEvent callEntitySpawnEvent(Entity entity) {
-        org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity();
-
+    public static EntitySpawnEvent callEntitySpawnEvent(org.bukkit.entity.Entity bukkitEntity) { // Leaf - only get bukkit entity once
         EntitySpawnEvent event = new EntitySpawnEvent(bukkitEntity);
         bukkitEntity.getServer().getPluginManager().callEvent(event);
         return event;
@@ -1351,32 +1355,46 @@ public class CraftEventFactory {
         return event;
     }
 
+    // Leaf start - only get bukkit entity once
     public static FoodLevelChangeEvent callFoodLevelChangeEvent(net.minecraft.world.entity.player.Player entity, int level) {
         return CraftEventFactory.callFoodLevelChangeEvent(entity, level, null);
     }
 
+    public static FoodLevelChangeEvent callFoodLevelChangeEvent(org.bukkit.entity.HumanEntity humanEntity, int level) {
+        return CraftEventFactory.callFoodLevelChangeEvent(humanEntity, level, null);
+    }
+
     public static FoodLevelChangeEvent callFoodLevelChangeEvent(net.minecraft.world.entity.player.Player entity, int level, ItemStack item) {
-        FoodLevelChangeEvent event = new FoodLevelChangeEvent(entity.getBukkitEntity(), level, (item == null) ? null : CraftItemStack.asBukkitCopy(item));
-        entity.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        org.bukkit.entity.HumanEntity humanEntity = entity.getBukkitEntity();
+        return callFoodLevelChangeEvent(humanEntity, level, item);
+    }
+
+    public static FoodLevelChangeEvent callFoodLevelChangeEvent(org.bukkit.entity.HumanEntity humanEntity, int level, ItemStack item) {
+        FoodLevelChangeEvent event = new FoodLevelChangeEvent(humanEntity, level, (item == null) ? null : CraftItemStack.asBukkitCopy(item));
+        humanEntity.getServer().getPluginManager().callEvent(event);
         return event;
     }
+    // Leaf end
 
     public static PigZapEvent callPigZapEvent(Entity pig, Entity lightning, Entity pigzombie) {
-        PigZapEvent event = new PigZapEvent((Pig) pig.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), (PigZombie) pigzombie.getBukkitEntity());
-        pig.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        Pig bukkitPig = (Pig) pig.getBukkitEntity(); // Leaf - only get bukkit entity once
+        PigZapEvent event = new PigZapEvent(bukkitPig, (LightningStrike) lightning.getBukkitEntity(), (PigZombie) pigzombie.getBukkitEntity());
+        bukkitPig.getServer().getPluginManager().callEvent(event);
         return event;
     }
 
     public static boolean callHorseJumpEvent(Entity horse, float power) {
-        HorseJumpEvent event = new HorseJumpEvent((AbstractHorse) horse.getBukkitEntity(), power);
-        horse.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        AbstractHorse bukkitHorse = (AbstractHorse) horse.getBukkitEntity(); // Leaf - only get bukkit entity once
+        HorseJumpEvent event = new HorseJumpEvent(bukkitHorse, power);
+        bukkitHorse.getServer().getPluginManager().callEvent(event);
         return !event.isCancelled();
     }
 
     // Paper start
     public static com.destroystokyo.paper.event.entity.EntityZapEvent callEntityZapEvent(Entity entity, Entity lightning, Entity changedEntity) {
-        com.destroystokyo.paper.event.entity.EntityZapEvent event = new com.destroystokyo.paper.event.entity.EntityZapEvent(entity.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), changedEntity.getBukkitEntity());
-        entity.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity(); // Leaf - only get bukkit entity once
+        com.destroystokyo.paper.event.entity.EntityZapEvent event = new com.destroystokyo.paper.event.entity.EntityZapEvent(bukkitEntity, (LightningStrike) lightning.getBukkitEntity(), changedEntity.getBukkitEntity());
+        bukkitEntity.getServer().getPluginManager().callEvent(event);
         return event;
     }
     // Paper end
@@ -1395,20 +1413,23 @@ public class CraftEventFactory {
     }
 
     public static CreeperPowerEvent callCreeperPowerEvent(Entity creeper, Entity lightning, CreeperPowerEvent.PowerCause cause) {
-        CreeperPowerEvent event = new CreeperPowerEvent((Creeper) creeper.getBukkitEntity(), (LightningStrike) lightning.getBukkitEntity(), cause);
-        creeper.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        Creeper bukkitCreeper = (Creeper) creeper.getBukkitEntity(); // Leaf - only get bukkit entity once
+        CreeperPowerEvent event = new CreeperPowerEvent(bukkitCreeper, (LightningStrike) lightning.getBukkitEntity(), cause);
+        bukkitCreeper.getServer().getPluginManager().callEvent(event);
         return event;
     }
 
     public static EntityTargetEvent callEntityTargetEvent(Entity entity, Entity target, EntityTargetEvent.TargetReason reason) {
-        EntityTargetEvent event = new EntityTargetEvent(entity.getBukkitEntity(), (target == null) ? null : target.getBukkitEntity(), reason);
-        entity.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity(); // Leaf - only get bukkit entity once
+        EntityTargetEvent event = new EntityTargetEvent(bukkitEntity, (target == null) ? null : target.getBukkitEntity(), reason);
+        bukkitEntity.getServer().getPluginManager().callEvent(event);
         return event;
     }
 
     public static EntityTargetLivingEntityEvent callEntityTargetLivingEvent(Entity entity, net.minecraft.world.entity.LivingEntity target, EntityTargetEvent.TargetReason reason) {
-        EntityTargetLivingEntityEvent event = new EntityTargetLivingEntityEvent(entity.getBukkitEntity(), (target == null) ? null : (LivingEntity) target.getBukkitEntity(), reason);
-        entity.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity(); // Leaf - only get bukkit entity once
+        EntityTargetLivingEntityEvent event = new EntityTargetLivingEntityEvent(bukkitEntity, (target == null) ? null : (LivingEntity) target.getBukkitEntity(), reason);
+        bukkitEntity.getServer().getPluginManager().callEvent(event);
         return event;
     }
 
@@ -1493,8 +1514,7 @@ public class CraftEventFactory {
     }
     // Paper end
 
-    public static ProjectileLaunchEvent callProjectileLaunchEvent(Entity entity) {
-        Projectile bukkitEntity = (Projectile) entity.getBukkitEntity();
+    public static ProjectileLaunchEvent callProjectileLaunchEvent(Projectile bukkitEntity) { // Leaf - only get bukkit entity once
         ProjectileLaunchEvent event = new ProjectileLaunchEvent(bukkitEntity);
         Bukkit.getPluginManager().callEvent(event);
         return event;
@@ -1895,8 +1915,7 @@ public class CraftEventFactory {
         return event;
     }
 
-    public static VehicleCreateEvent callVehicleCreateEvent(Entity entity) {
-        Vehicle bukkitEntity = (Vehicle) entity.getBukkitEntity();
+    public static VehicleCreateEvent callVehicleCreateEvent(Vehicle bukkitEntity) { // Leaf - only get bukkit entity once
         VehicleCreateEvent event = new VehicleCreateEvent(bukkitEntity);
         Bukkit.getPluginManager().callEvent(event);
         return event;
@@ -2188,7 +2207,8 @@ public class CraftEventFactory {
     }
 
     public static EntityExplodeEvent callEntityExplodeEvent(Entity entity, List<Block> blocks, float yield, Explosion.BlockInteraction effect) {
-        EntityExplodeEvent event = new EntityExplodeEvent(entity.getBukkitEntity(), entity.getBukkitEntity().getLocation(), blocks, yield, CraftExplosionResult.toExplosionResult(effect));
+        org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity(); // Leaf - only get bukkit entity once
+        EntityExplodeEvent event = new EntityExplodeEvent(bukkitEntity, bukkitEntity.getLocation(), blocks, yield, CraftExplosionResult.toExplosionResult(effect));
         Bukkit.getPluginManager().callEvent(event);
         return event;
     }
