From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Petal: Async Pathfinding


diff --git a/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java b/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java
index efb22c6535f74273ae92418a787e509fd91a8131..9f74edf84d277fa1f677e9c2faa52fdcd6d6590f 100644
--- a/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java
+++ b/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java
@@ -43,8 +43,12 @@ public class PaperPathfinder implements com.destroystokyo.paper.entity.Pathfinde
     @Nullable
     @Override
     public PathResult getCurrentPath() {
-        Path path = this.entity.getNavigation().getPath();
-        return path != null && !path.isDone() ? new PaperPathResult(path) : null;
+        // Kaiiju start - petal - async path processing
+        net.minecraft.world.entity.ai.navigation.PathNavigation navigation = this.entity.getNavigation();
+        Path path = processPath(navigation.getPath());
+        return path != null && navigation.getPath() == path && navigation.leaf$preparePath() && navigation.getPath() == path
+            ? new PaperPathResult(path) : null;
+        // Kaiiju end - petal - async path processing
     }
 
     @Nullable
@@ -52,7 +56,7 @@ public class PaperPathfinder implements com.destroystokyo.paper.entity.Pathfinde
     public PathResult findPath(final Location loc, final int reachRange) {
         Preconditions.checkArgument(loc != null, "Location can not be null");
         Preconditions.checkArgument(reachRange >= 0, "Reach range can not be negative");
-        Path path = this.entity.getNavigation().createPath(loc.getX(), loc.getY(), loc.getZ(), reachRange);
+        Path path = processPath(this.entity.getNavigation().createPath(loc.getX(), loc.getY(), loc.getZ(), reachRange)); // Kaiiju - petal - async path processing
         return path != null ? new PaperPathResult(path) : null;
     }
 
@@ -61,10 +65,20 @@ public class PaperPathfinder implements com.destroystokyo.paper.entity.Pathfinde
     public PathResult findPath(final Entity target, final int reachRange) {
         Preconditions.checkArgument(target != null, "Target can not be null");
         Preconditions.checkArgument(reachRange >= 0, "Reach range can not be negative");
-        Path path = this.entity.getNavigation().createPath(((CraftEntity) target).getHandle(), reachRange);
+        Path path = processPath(this.entity.getNavigation().createPath(((CraftEntity) target).getHandle(), reachRange)); // Kaiiju - petal - async path processing
         return path != null ? new PaperPathResult(path) : null;
     }
 
+    // Kaiiju start - petal - async path processing
+    @Nullable
+    private static Path processPath(@Nullable Path path) {
+        if (path instanceof org.dreeam.leaf.async.path.AsyncPath asyncPath) {
+            asyncPath.process();
+        }
+        return path;
+    }
+    // Kaiiju end - petal - async path processing
+
     @Override
     public boolean moveTo(@Nonnull PathResult path, double speed) {
         Preconditions.checkArgument(path != null, "PathResult can not be null");
