From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 14 Jan 2024 05:14:09 -0500
Subject: [PATCH] Configurable server GUI name


diff --git a/net/minecraft/server/gui/MinecraftServerGui.java b/net/minecraft/server/gui/MinecraftServerGui.java
index 960454a167ef52eb867da40af1fc5b3de4ea307c..db4d0cc0149b54e5b8a9b3e646d690e6da7cf782 100644
--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -53,7 +53,7 @@ public class MinecraftServerGui extends JComponent {
         } catch (Exception var3) {
         }
 
-        final JFrame frame = new JFrame("Purpur Minecraft server"); // Purpur - Improve GUI
+        final JFrame frame = new JFrame(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName); // Purpur - Improve GUI // Leaf - Configurable server GUI name
         final MinecraftServerGui gui = new MinecraftServerGui(server);
         frame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
         frame.add(gui);
@@ -61,7 +61,7 @@ public class MinecraftServerGui extends JComponent {
         frame.setLocationRelativeTo(null);
         frame.setVisible(true);
         // Paper start - Improve ServerGUI
-        frame.setName("Purpur Minecraft server"); // Purpur - Improve GUI
+        frame.setName(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName); // Purpur - Improve GUI // Leaf - Configurable server GUI name
         try {
             frame.setIconImage(javax.imageio.ImageIO.read(java.util.Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("logo.png"))));
         } catch (java.io.IOException ignore) {
@@ -71,7 +71,7 @@ public class MinecraftServerGui extends JComponent {
             @Override
             public void windowClosing(final WindowEvent event) {
                 if (!gui.isClosing.getAndSet(true)) {
-                    frame.setTitle("Purpur Minecraft server - shutting down!"); // Purpur - Improve GUI
+                    frame.setTitle(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName + " - shutting down!"); // Purpur - Improve GUI // Leaf - Configurable server GUI name
                     server.halt(true);
                     gui.runFinalizers();
                 }
