From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 14 Aug 2025 14:12:38 +0900
Subject: [PATCH] optimize attribute


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 097ad442a4eed12412f6dbd3e6eb34b983237b4b..3ae5196e0a0fa5b8c8f590b3cf6345993c459e7c 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -438,7 +438,7 @@ public class ServerEntity {
                 this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundUpdateAttributesPacket(this.entity.getId(), attributes));
             }
 
-            attributes.clear();
+            //attributes.clear(); // Leaf - optimize attribute
         }
     }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index cd87d5ac79fd7c77118a91a73a1218e4c27e2923..98d6d2d6382a22c2a2e33575f562984eb8bfb603 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -392,7 +392,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             .add(Attributes.NAME_TAG_DISTANCE)
             .add(Attributes.BELOW_NAME_DISTANCE);
     }
-    public boolean shouldSendAttribute(Attribute attribute) { return true; } // Purpur - Ridables
+    public boolean shouldSendAttribute(Attribute attribute) { return true; } // Purpur - Ridables // Leaf - optimize attribute - diff on change
 
     @Override
     protected void checkFallDamage(final double ya, final boolean onGround, final BlockState onState, final BlockPos pos) {
@@ -1369,13 +1369,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     private void refreshDirtyAttributes() {
-        Set<AttributeInstance> attributesToUpdate = this.getAttributes().getAttributesToUpdate();
+        // Leaf start - optimize attribute
+        int[] attributesToUpdate = this.getAttributes().getAttributesToUpdateIds();
 
-        for (AttributeInstance changedAttributeInstance : attributesToUpdate) {
-            this.onAttributeUpdated(changedAttributeInstance.getAttribute());
+        for (int attribute : attributesToUpdate) {
+            this.onAttributeUpdated(org.dreeam.leaf.util.RegistryTypeManager.ATTRIBUTE[attribute]);
         }
-
-        attributesToUpdate.clear();
+        // Leaf end - optimize attribute
     }
 
     protected void onAttributeUpdated(final Holder<Attribute> attribute) {
diff --git a/net/minecraft/world/entity/ai/attributes/Attribute.java b/net/minecraft/world/entity/ai/attributes/Attribute.java
index ad3dc7f69eb6e5a167118c4fe16c2fe4083522dd..c2947dd00a7ddd0622eece451bacff477bd81772 100644
--- a/net/minecraft/world/entity/ai/attributes/Attribute.java
+++ b/net/minecraft/world/entity/ai/attributes/Attribute.java
@@ -16,10 +16,15 @@ public class Attribute {
     private boolean syncable;
     private final String descriptionId;
     public Attribute.Sentiment sentiment = Attribute.Sentiment.POSITIVE;
+    // Leaf start - optimize attribute
+    public final int id;
+    private static int size = 0;
+    // Leaf end - optimize attribute
 
     protected Attribute(final String descriptionId, final double defaultValue) {
         this.defaultValue = defaultValue;
         this.descriptionId = descriptionId;
+        this.id = size++; // Leaf - optimize attribute
     }
 
     public double getDefaultValue() {
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
index 82fb9b7092ee37bb4b46ad9a352df64821259ddc..084922a0630a24c8111f383703e893eed56f322b 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
@@ -51,16 +51,19 @@ public class AttributeInstance {
 
     @VisibleForTesting
     Map<Identifier, AttributeModifier> getModifiers(final AttributeModifier.Operation operation) {
-        return this.modifiersByOperation.computeIfAbsent(operation, key -> new Object2ObjectOpenHashMap<>());
+        return this.modifiersByOperation.computeIfAbsent(operation, key -> new Object2ObjectArrayMap<>()); // Leaf - optimize attribute
     }
 
+    // Leaf start - optimize attribute
+    private static final AttributeModifier[] EMPTY_ATTRIBUTE_MODIFIER_ARRAY = {};
     public Set<AttributeModifier> getModifiers() {
-        return ImmutableSet.copyOf(this.modifierById.values());
+        return this.modifierById.isEmpty() ? Set.of() : it.unimi.dsi.fastutil.objects.ObjectArraySet.ofUnchecked(this.modifierById.values().toArray(EMPTY_ATTRIBUTE_MODIFIER_ARRAY));
     }
 
     public Set<AttributeModifier> getPermanentModifiers() {
-        return ImmutableSet.copyOf(this.permanentModifiers.values());
+        return this.permanentModifiers.isEmpty() ? Set.of() : it.unimi.dsi.fastutil.objects.ObjectArraySet.ofUnchecked(this.permanentModifiers.values().toArray(EMPTY_ATTRIBUTE_MODIFIER_ARRAY));
     }
+    // Leaf end - optimize attribute
 
     public @Nullable AttributeModifier getModifier(final Identifier id) {
         return this.modifierById.get(id);
@@ -137,12 +140,15 @@ public class AttributeInstance {
     }
 
     public double getValue() {
-        if (this.dirty) {
-            this.cachedValue = this.calculateValue();
-            this.dirty = false;
+        // Leaf start - optimize attribute
+        if (!this.dirty) {
+            return this.cachedValue;
         }
-
-        return this.cachedValue;
+        double value = this.calculateValue();
+        this.cachedValue = value;
+        this.dirty = false;
+        return value;
+        // Leaf end - optimize attribute
     }
 
     private double calculateValue() {
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeMap.java b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
index c075423eaf52dfb84c87f335dcd5beeca6b73d60..76b093843b7738b0d91db601e204c323af5386e3 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -13,10 +13,12 @@ import net.minecraft.core.Holder;
 import net.minecraft.resources.Identifier;
 import org.jspecify.annotations.Nullable;
 
-public class AttributeMap {
-    public final Map<Holder<Attribute>, AttributeInstance> attributes = new Object2ObjectOpenHashMap<>();
-    private final Set<AttributeInstance> attributesToSync = new ObjectOpenHashSet<>();
-    private final Set<AttributeInstance> attributesToUpdate = new ObjectOpenHashSet<>();
+public class AttributeMap implements java.util.function.Consumer<AttributeInstance> { // Leaf - optimize attribute
+    // Leaf start - optimize attribute
+    public final Map<Holder<Attribute>, AttributeInstance> attributes = new org.dreeam.leaf.util.map.AttributeInstanceArrayMap();
+    private final Set<AttributeInstance> attributesToSync = new org.dreeam.leaf.util.map.AttributeInstanceSet((org.dreeam.leaf.util.map.AttributeInstanceArrayMap) attributes);
+    private final Set<AttributeInstance> attributesToUpdate = new org.dreeam.leaf.util.map.AttributeInstanceSet((org.dreeam.leaf.util.map.AttributeInstanceArrayMap) attributes);
+    // Leaf end - optimize attribute
     private final AttributeSupplier supplier;
     private final net.minecraft.world.entity.LivingEntity entity; // Purpur - Ridables
 
@@ -25,33 +27,107 @@ public class AttributeMap {
         this(supplier, null);
     }
     public AttributeMap(AttributeSupplier defaultAttributes, net.minecraft.world.entity.LivingEntity entity) {
-        this.entity = entity;
+        this.entity = entity instanceof net.minecraft.world.entity.ambient.Bat ? entity : null; // Leaf - optimize attribute - only check bat
         // Purpur end - Ridables
         this.supplier = defaultAttributes;
     }
 
     private void onAttributeModified(final AttributeInstance attributeInstance) {
-        this.attributesToUpdate.add(attributeInstance);
-        if (attributeInstance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(attributeInstance.getAttribute().value()))) { // Purpur - Ridables
-            this.attributesToSync.add(attributeInstance);
+        // Leaf start - optimize attribute
+        Attribute attribute = attributeInstance.getAttribute().value();
+        ((org.dreeam.leaf.util.map.AttributeInstanceSet) this.attributesToUpdate).addAttribute(attribute);
+        if (attribute.isClientSyncable() && (entity == null || entity.shouldSendAttribute(attribute))) { // Purpur - Ridables
+            ((org.dreeam.leaf.util.map.AttributeInstanceSet) this.attributesToSync).addAttribute(attribute);
         }
+        // Leaf end - optimize attribute
     }
 
+    // Leaf start - optimize attribute
+    @Override
+    public void accept(AttributeInstance instance) {
+        this.onAttributeModified(instance);
+    }
+    private static final AttributeInstance[] EMPTY_ATTRIBUTE_INSTANCE = new AttributeInstance[0];
+    @Deprecated
     public Set<AttributeInstance> getAttributesToSync() {
-        return this.attributesToSync;
+        if (attributesToSync.isEmpty()) {
+            return Set.of();
+        }
+        it.unimi.dsi.fastutil.objects.ReferenceArraySet<AttributeInstance> clone = it.unimi.dsi.fastutil.objects.ReferenceArraySet.ofUnchecked(attributesToSync.toArray(EMPTY_ATTRIBUTE_INSTANCE));
+        this.attributesToSync.clear();
+        return clone;
     }
 
+    @Deprecated
     public Set<AttributeInstance> getAttributesToUpdate() {
-        return this.attributesToUpdate;
+        if (attributesToUpdate.isEmpty()) {
+            return Set.of();
+        }
+        it.unimi.dsi.fastutil.objects.ReferenceArraySet<AttributeInstance> clone = it.unimi.dsi.fastutil.objects.ReferenceArraySet.ofUnchecked(attributesToUpdate.toArray(EMPTY_ATTRIBUTE_INSTANCE));
+        this.attributesToUpdate.clear();
+        return clone;
+    }
+
+    public boolean attributeDirty() {
+        return !attributesToSync.isEmpty();
+    }
+
+    public int[] getAttributesToUpdateIds() {
+        int[] clone = ((org.dreeam.leaf.util.map.AttributeInstanceSet) attributesToUpdate).inner.toIntArray();
+        this.attributesToUpdate.clear();
+        return clone;
+    }
+
+    public int[] getAttributesToSyncIds() {
+        int[] clone = ((org.dreeam.leaf.util.map.AttributeInstanceSet) attributesToSync).inner.toIntArray();
+        this.attributesToSync.clear();
+        return clone;
     }
 
     public Collection<AttributeInstance> getSyncableAttributes() {
-        return this.attributes.values().stream().filter(instance -> instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))).collect(Collectors.toList()); // Purpur - Ridables
+        List<AttributeInstance> list = new ArrayList<>(this.attributes.size());
+        for (AttributeInstance instance : this.attributes.values()) {
+            if (instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))) { // Purpur - Ridables
+                list.add(instance);
+            }
+        }
+        return list;
+    }
+
+    public List<net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot> getSyncableAttributesPacket() {
+        List<net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot> list = new ArrayList<>(attributes.size());
+        if (attributes instanceof org.dreeam.leaf.util.map.AttributeInstanceArrayMap map) {
+            for (AttributeInstance instance : map.elements()) {
+                if (instance != null && instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))) { // Purpur - Ridables
+                    list.add(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot(
+                        instance.getAttribute(), instance.getBaseValue(), instance.getModifiers()
+                    ));
+                }
+            }
+        } else {
+            for (AttributeInstance instance : attributes.values()) {
+                if (instance != null && instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))) { // Purpur - Ridables
+                    list.add(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot(
+                        instance.getAttribute(), instance.getBaseValue(), instance.getModifiers()
+                    ));
+                }
+            }
+        }
+        return list;
     }
 
     public @Nullable AttributeInstance getInstance(final Holder<Attribute> attribute) {
-        return this.attributes.computeIfAbsent(attribute, key -> this.supplier.createInstance(this::onAttributeModified, (Holder<Attribute>)key));
+        AttributeInstance v;
+        if ((v = this.attributes.get(attribute)) == null) {
+            AttributeInstance newValue;
+            if ((newValue = this.supplier.createInstance(this, attribute)) != null) {
+                attributes.put(attribute, newValue);
+                return newValue;
+            }
+        }
+        return v;
     }
+    // Leaf end - optimize attribute
 
     public boolean hasAttribute(final Holder<Attribute> attribute) {
         return this.attributes.get(attribute) != null || this.supplier.hasAttribute(attribute);
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java b/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
index 5d2c1cfbdd20c5c195790b5241c3a607c51aa72c..444ecf92bc521769611b62f649ab14f8fa65ab2c 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
@@ -11,7 +11,7 @@ public class AttributeSupplier {
     private final Map<Holder<Attribute>, AttributeInstance> instances;
 
     private AttributeSupplier(final Map<Holder<Attribute>, AttributeInstance> instances) {
-        this.instances = instances;
+        this.instances = new org.dreeam.leaf.util.map.AttributeInstanceArrayMap(instances); // Leaf - optimize attribute
     }
 
     public AttributeInstance getAttributeInstance(final Holder<Attribute> attribute) {
