From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 11 Aug 2022 04:13:27 +0800
Subject: [PATCH] Leaf config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9bab00ab10c78908090c8a1a12d4c84e9324b08b..e13b529dfa46ff2b4b6beaa198283b61c3639f93 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2387,6 +2387,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Leaf start - Leaf config - API
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLeafConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Leaf end - Leaf config - API
+
         /**
          * Sends the component to the player
          *
