From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 19 Feb 2025 00:34:16 -0500
Subject: [PATCH] PlayerInventoryOverflowEvent


diff --git a/src/main/java/org/bukkit/event/HandlerList.java b/src/main/java/org/bukkit/event/HandlerList.java
index ef9d9f3ddaf25aa990715139e1916672f76c35e5..13ab01b4fff89f43028e25d983cc522aac946c69 100644
--- a/src/main/java/org/bukkit/event/HandlerList.java
+++ b/src/main/java/org/bukkit/event/HandlerList.java
@@ -38,6 +38,8 @@ public class HandlerList {
      */
     private static final java.util.Set<String> EVENT_TYPES = java.util.concurrent.ConcurrentHashMap.newKeySet();
 
+    public static short isListeningInvOverflowCached = -1; // Leaf - PlayerInventoryOverflowEvent
+
     /**
      * Bake all handler lists. Best used just after all normal event
      * registration is complete, ie just after all plugins are loaded if
@@ -64,6 +66,7 @@ public class HandlerList {
                     h.handlers = null;
                 }
             }
+            isListeningInvOverflowCached = -1; // Leaf - PlayerInventoryOverflowEvent
         }
     }
 
@@ -144,6 +147,7 @@ public class HandlerList {
     public synchronized void unregister(@NotNull RegisteredListener listener) {
         if (handlerslots.get(listener.getPriority()).remove(listener)) {
             handlers = null;
+            isListeningInvOverflowCached = -1; // Leaf - PlayerInventoryOverflowEvent
         }
     }
 
@@ -162,7 +166,12 @@ public class HandlerList {
                 }
             }
         }
-        if (changed) handlers = null;
+        // Leaf start - PlayerInventoryOverflowEvent
+        if (changed) {
+            handlers = null;
+            isListeningInvOverflowCached = -1;
+        }
+        // Leaf end - PlayerInventoryOverflowEvent
     }
 
     /**
@@ -180,7 +189,12 @@ public class HandlerList {
                 }
             }
         }
-        if (changed) handlers = null;
+        // Leaf start - PlayerInventoryOverflowEvent
+        if (changed) {
+            handlers = null;
+            isListeningInvOverflowCached = -1;
+        }
+        // Leaf end - PlayerInventoryOverflowEvent
     }
 
     /**
diff --git a/src/main/java/org/dreeam/leaf/event/player/PlayerInventoryOverflowEvent.java b/src/main/java/org/dreeam/leaf/event/player/PlayerInventoryOverflowEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..eae76671190ef84529c0dd503263e43a15a74e8a
--- /dev/null
+++ b/src/main/java/org/dreeam/leaf/event/player/PlayerInventoryOverflowEvent.java
@@ -0,0 +1,65 @@
+package org.dreeam.leaf.event.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.ItemStack;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.Map;
+
+/**
+ * Called when a plugin uses {@link Inventory#addItem(ItemStack...)} to add items
+ * into player's inventory and the inventory becomes full.
+ * <p>
+ * Notice that, using this event is the worst option. It's better to handle
+ * those overflow / leftover items via {@link Inventory#addItem(ItemStack...)}
+ * logic directly.
+ * Also, See <a href="https://github.com/PaperMC/Paper/issues/11886">Paper#11886</a>
+ * for more.
+ */
+@NullMarked
+public class PlayerInventoryOverflowEvent extends PlayerEvent {
+
+    private static final HandlerList HANDLER_LIST = new HandlerList();
+
+    public static short isListeningInvOverflowCached = -1;
+
+    private final Inventory inventory;
+    private final Map<Integer, ItemStack> overflowItemStacks;
+
+    public PlayerInventoryOverflowEvent(final Player player, final Map<Integer, ItemStack> overflowItemStacks) {
+        super(player);
+        this.inventory = player.getInventory();
+        this.overflowItemStacks = overflowItemStacks;
+    }
+
+    /**
+     * Gets player's inventory, which is full
+     *
+     * @return The player inventory.
+     */
+    public Inventory getPlayerInventory() {
+        return this.inventory;
+    }
+
+    /**
+     * Gets overflow items which cannot be added into
+     * the full inventory
+     *
+     * @return The overflow items.
+     */
+    public Map<Integer, ItemStack> getOverflowItemStacks() {
+        return this.overflowItemStacks;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+}
