From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Fri, 9 Jan 2026 02:40:25 -0500
Subject: [PATCH] Cache namespacedKey toString and hash


diff --git a/src/main/java/org/bukkit/NamespacedKey.java b/src/main/java/org/bukkit/NamespacedKey.java
index 57d95e22db8c6b925fc96cf1a45a5a04366bb517..dc4c05f3a5e3f091b778cbf65182a328e7ba8d53 100644
--- a/src/main/java/org/bukkit/NamespacedKey.java
+++ b/src/main/java/org/bukkit/NamespacedKey.java
@@ -37,6 +37,10 @@ public final class NamespacedKey implements Key, com.destroystokyo.paper.Namespa
 
     private final String namespace;
     private final String key;
+    // Leaf start - Cache namespacedKey toString and hash
+    private final String keyStr;
+    private final int hash;
+    // Leaf - Cache namespacedKey toString and hash
 
     /**
      * Create a key in a specific namespace.
@@ -54,7 +58,15 @@ public final class NamespacedKey implements Key, com.destroystokyo.paper.Namespa
         this.namespace = namespace;
         this.key = key;
 
-        this.validate();
+        // Leaf start - Cache namespacedKey toString and hash
+        this.keyStr = this.namespace + ':' + this.key;
+
+        this.validate(keyStr.length());
+
+        int result = this.namespace.hashCode();
+        result = (31 * result) + this.key.hashCode();
+        this.hash = result;
+        // Leaf end - Cache namespacedKey toString and hash
     }
 
     /**
@@ -75,12 +87,21 @@ public final class NamespacedKey implements Key, com.destroystokyo.paper.Namespa
         this.namespace = plugin.namespace();
         this.key = key.toLowerCase(Locale.ROOT);
 
+        // Leaf start - Cache namespacedKey toString and hash
+        this.keyStr = this.namespace + ':' + this.key;
+
         // Check validity after normalization
-        this.validate();
+        this.validate(keyStr.length());
+
+        int result = this.namespace.hashCode();
+        result = (31 * result) + this.key.hashCode();
+        this.hash = result;
+        // Leaf end - Cache namespacedKey toString and hash
+
     }
 
-    private void validate() {
-        Preconditions.checkArgument(this.namespace.length() + 1 + this.key.length() <= Short.MAX_VALUE, "NamespacedKey must be less than 32768 characters");
+    private void validate(int keyStrLen) { // Leaf - Cache namespacedKey toString and hash
+        Preconditions.checkArgument(keyStrLen <= Short.MAX_VALUE, "NamespacedKey must be less than 32768 characters"); // Leaf - Cache namespacedKey toString and hash
         checkError("[a-z0-9_-.]", "namespace", this.namespace, Key.checkNamespace(this.namespace));
         checkError("[a-z0-9_-./]", "key", this.key, Key.checkValue(this.key));
     }
@@ -106,9 +127,7 @@ public final class NamespacedKey implements Key, com.destroystokyo.paper.Namespa
 
     @Override
     public int hashCode() {
-        int result = this.namespace.hashCode();
-        result = (31 * result) + this.key.hashCode();
-        return result;
+        return this.hash; // Leaf - Cache namespacedKey toString and hash
     }
 
     @Override
@@ -123,7 +142,7 @@ public final class NamespacedKey implements Key, com.destroystokyo.paper.Namespa
 
     @Override
     public String toString() {
-        return this.namespace + ':' + this.key;
+        return this.keyStr; // Leaf - Cache namespacedKey toString and hash
     }
 
     /**
