From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Mon, 27 Apr 2026 21:14:18 +0300
Subject: [PATCH] use fastutil for primitive keys in data maps


diff --git a/src/main/java/org/bukkit/ChatColor.java b/src/main/java/org/bukkit/ChatColor.java
index f283bcabff7fe6eede6cf4344537e43048c35166..496db92af789a1ceffc4549ad1d91f76ec05baf9 100644
--- a/src/main/java/org/bukkit/ChatColor.java
+++ b/src/main/java/org/bukkit/ChatColor.java
@@ -1,9 +1,13 @@
 package org.bukkit;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.Maps;
-import java.util.Map;
+
 import java.util.regex.Pattern;
+
+import it.unimi.dsi.fastutil.chars.Char2ObjectMap;
+import it.unimi.dsi.fastutil.chars.Char2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -246,8 +250,8 @@ public enum ChatColor {
     private final char code;
     private final boolean isFormat;
     private final String toString;
-    private static final Map<Integer, ChatColor> BY_ID = Maps.newHashMap();
-    private static final Map<Character, ChatColor> BY_CHAR = Maps.newHashMap();
+    private static final Int2ObjectMap<ChatColor> BY_ID = new Int2ObjectOpenHashMap<>();
+    private static final Char2ObjectMap<ChatColor> BY_CHAR = new Char2ObjectOpenHashMap<>();
 
     private ChatColor(char code, int intCode) {
         this(code, intCode, false);
diff --git a/src/main/java/org/bukkit/CoalType.java b/src/main/java/org/bukkit/CoalType.java
index ede570814963a0fb93e2e5affa54439880cd685b..24fec91bae13e52b323c750ff2699c3d907bf14d 100644
--- a/src/main/java/org/bukkit/CoalType.java
+++ b/src/main/java/org/bukkit/CoalType.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.jetbrains.annotations.Nullable;
 
 /**
@@ -13,7 +13,7 @@ public enum CoalType {
     CHARCOAL(0x1);
 
     private final byte data;
-    private static final Map<Byte, CoalType> BY_DATA = Maps.newHashMap();
+    private static final Byte2ObjectMap<CoalType> BY_DATA = new Byte2ObjectOpenHashMap<>();
 
     private CoalType(final int data) {
         this.data = (byte) data;
diff --git a/src/main/java/org/bukkit/CropState.java b/src/main/java/org/bukkit/CropState.java
index 9420e919088392c6aac4114356af7a2096478195..0f69bca96bd890ce260f3320b826ff5f7036df36 100644
--- a/src/main/java/org/bukkit/CropState.java
+++ b/src/main/java/org/bukkit/CropState.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.jetbrains.annotations.Nullable;
 
 /**
@@ -44,7 +44,7 @@ public enum CropState {
     RIPE(0x7);
 
     private final byte data;
-    private static final Map<Byte, CropState> BY_DATA = Maps.newHashMap();
+    private static final Byte2ObjectMap<CropState> BY_DATA = new Byte2ObjectOpenHashMap<>();
 
     private CropState(final int data) {
         this.data = (byte) data;
diff --git a/src/main/java/org/bukkit/Difficulty.java b/src/main/java/org/bukkit/Difficulty.java
index 427ce8cfd6f63e5c7ec7b264b15ab4111b947729..8460970c107c28c1ba1e5396e4b07cc4984805a2 100644
--- a/src/main/java/org/bukkit/Difficulty.java
+++ b/src/main/java/org/bukkit/Difficulty.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import org.jetbrains.annotations.Nullable;
 
 /**
@@ -34,7 +34,7 @@ public enum Difficulty implements net.kyori.adventure.translation.Translatable {
     HARD(3);
 
     private final int value;
-    private static final Map<Integer, Difficulty> BY_ID = Maps.newHashMap();
+    private static final Int2ObjectMap<Difficulty> BY_ID = new Int2ObjectOpenHashMap<>();
 
     private Difficulty(final int value) {
         this.value = value;
diff --git a/src/main/java/org/bukkit/Effect.java b/src/main/java/org/bukkit/Effect.java
index d2952c7750af9aa1ed762f3b49bd6b7b11c6c7a0..c82a1c0ceb12dda8465d53ab4f1ff45ab0f32a74 100644
--- a/src/main/java/org/bukkit/Effect.java
+++ b/src/main/java/org/bukkit/Effect.java
@@ -1,8 +1,9 @@
 package org.bukkit;
 
 import com.google.common.base.Enums;
-import com.google.common.collect.Maps;
-import java.util.Map;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import org.bukkit.block.BlockFace;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -448,7 +449,7 @@ public enum Effect {
     private final int id;
     private final Type type;
     private final java.util.List<Class<?>> data; // Paper - support multiple data types
-    private static final Map<Integer, Effect> BY_ID = Maps.newHashMap();
+    private static final Int2ObjectMap<Effect> BY_ID = new Int2ObjectOpenHashMap<>();
 
     Effect(int id, /*@NotNull*/ Type type) {
         this(id, type, (Class<?>[]) null); // Paper - support multiple data types
diff --git a/src/main/java/org/bukkit/GameMode.java b/src/main/java/org/bukkit/GameMode.java
index 796f75313255ecc8917819b28eaf12776f2e4ce3..52bb8d5fd0c3691eafb0c3e61fa0ff99f71b513b 100644
--- a/src/main/java/org/bukkit/GameMode.java
+++ b/src/main/java/org/bukkit/GameMode.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import org.bukkit.entity.HumanEntity;
 import org.jetbrains.annotations.Nullable;
 
@@ -34,7 +34,7 @@ public enum GameMode implements net.kyori.adventure.translation.Translatable { /
     SPECTATOR(3);
 
     private final int value;
-    private static final Map<Integer, GameMode> BY_ID = Maps.newHashMap();
+    private static final Int2ObjectMap<GameMode> BY_ID = new Int2ObjectOpenHashMap<>();
     // Paper start - translation keys
     private final String translationKey;
 
diff --git a/src/main/java/org/bukkit/GrassSpecies.java b/src/main/java/org/bukkit/GrassSpecies.java
index 43da0eb033f7cb65e3b2cacd67946bc9dc187678..3601d0202c856df7369cdf8ad1189a7d16d2096b 100644
--- a/src/main/java/org/bukkit/GrassSpecies.java
+++ b/src/main/java/org/bukkit/GrassSpecies.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.jetbrains.annotations.Nullable;
 
 /**
@@ -25,7 +25,7 @@ public enum GrassSpecies {
     FERN_LIKE(0x2);
 
     private final byte data;
-    private static final Map<Byte, GrassSpecies> BY_DATA = Maps.newHashMap();
+    private static final Byte2ObjectMap<GrassSpecies> BY_DATA = new Byte2ObjectOpenHashMap<>();
 
     private GrassSpecies(final int data) {
         this.data = (byte) data;
diff --git a/src/main/java/org/bukkit/Instrument.java b/src/main/java/org/bukkit/Instrument.java
index da5a41f9133a8c4b0c1eb959bc7cbb7259e39d77..7e58c2fd0dffe54d533f221beaff55d41ba82918 100644
--- a/src/main/java/org/bukkit/Instrument.java
+++ b/src/main/java/org/bukkit/Instrument.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.jetbrains.annotations.Nullable;
 
 public enum Instrument {
@@ -106,7 +106,7 @@ public enum Instrument {
 
     private final byte type;
     private final Sound sound;
-    private static final Map<Byte, Instrument> BY_DATA = Maps.newHashMap();
+    private static final Byte2ObjectMap<Instrument> BY_DATA = new Byte2ObjectOpenHashMap<>();
 
     // Paper - remove ctor (the server still uses the byte magic value)
 
diff --git a/src/main/java/org/bukkit/Note.java b/src/main/java/org/bukkit/Note.java
index aff858346776386f1288b648b221404f7f412399..a4982c4cfda2312ebddb6178b8c424465519af3a 100644
--- a/src/main/java/org/bukkit/Note.java
+++ b/src/main/java/org/bukkit/Note.java
@@ -1,8 +1,9 @@
 package org.bukkit;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.Maps;
-import java.util.Map;
+
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
@@ -26,7 +27,7 @@ public class Note {
         private final boolean sharpable;
         private final byte id;
 
-        private static final Map<Byte, Note.Tone> BY_DATA = Maps.newHashMap();
+        private static final Byte2ObjectMap<Tone> BY_DATA = new Byte2ObjectOpenHashMap<>();
         /** The number of tones including sharped tones. */
         public static final byte TONES_COUNT = 12;
 
diff --git a/src/main/java/org/bukkit/SandstoneType.java b/src/main/java/org/bukkit/SandstoneType.java
index d2ea2c81f94093414399512e289cea6c12855ec2..fcebf82f84bdd760a1f66d99f6a634701fef617f 100644
--- a/src/main/java/org/bukkit/SandstoneType.java
+++ b/src/main/java/org/bukkit/SandstoneType.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.jetbrains.annotations.Nullable;
 
 /**
@@ -15,7 +15,7 @@ public enum SandstoneType {
     SMOOTH(0x2);
 
     private final byte data;
-    private static final Map<Byte, SandstoneType> BY_DATA = Maps.newHashMap();
+    private static final Byte2ObjectMap<SandstoneType> BY_DATA = new Byte2ObjectOpenHashMap<>();
 
     private SandstoneType(final int data) {
         this.data = (byte) data;
diff --git a/src/main/java/org/bukkit/TreeSpecies.java b/src/main/java/org/bukkit/TreeSpecies.java
index abbe7c3b2984dc03098780d5678553685ea78dc4..1d1bcd30dd4896543a31992bfce7101d4f89b2ff 100644
--- a/src/main/java/org/bukkit/TreeSpecies.java
+++ b/src/main/java/org/bukkit/TreeSpecies.java
@@ -1,7 +1,7 @@
 package org.bukkit;
 
-import com.google.common.collect.Maps;
-import java.util.Map;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.jetbrains.annotations.Nullable;
 
 /**
@@ -39,7 +39,7 @@ public enum TreeSpecies {
     ;
 
     private final byte data;
-    private static final Map<Byte, TreeSpecies> BY_DATA = Maps.newHashMap();
+    private static final Byte2ObjectMap<TreeSpecies> BY_DATA = new Byte2ObjectOpenHashMap<>();
 
     private TreeSpecies(final int data) {
         this.data = (byte) data;
diff --git a/src/main/java/org/bukkit/material/types/MushroomBlockTexture.java b/src/main/java/org/bukkit/material/types/MushroomBlockTexture.java
index ca1b5d270848aebea3fb5c87daa5dc11ff1a877c..89cb0531cf318d7bc0726b7ad75a0fba66deaaad 100644
--- a/src/main/java/org/bukkit/material/types/MushroomBlockTexture.java
+++ b/src/main/java/org/bukkit/material/types/MushroomBlockTexture.java
@@ -1,7 +1,10 @@
 package org.bukkit.material.types;
 
-import com.google.common.collect.Maps;
+import java.util.EnumMap;
 import java.util.Map;
+
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
 import org.bukkit.block.BlockFace;
 import org.jetbrains.annotations.Nullable;
 
@@ -65,8 +68,8 @@ public enum MushroomBlockTexture {
      * Stem texture on all faces.
      */
     ALL_STEM(15, null);
-    private static final Map<Byte, MushroomBlockTexture> BY_DATA = Maps.newHashMap();
-    private static final Map<BlockFace, MushroomBlockTexture> BY_BLOCKFACE = Maps.newHashMap();
+    private static final Byte2ObjectMap<MushroomBlockTexture> BY_DATA = new Byte2ObjectOpenHashMap<>();
+    private static final Map<BlockFace, MushroomBlockTexture> BY_BLOCKFACE = new EnumMap<>(BlockFace.class);
 
     private final Byte data;
     private final BlockFace capFace;
