From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Limit max size of subscription codec

Removed since Leaf 26.2, fixed in vanilla

diff --git a/net/minecraft/network/protocol/game/ServerboundDebugSubscriptionRequestPacket.java b/net/minecraft/network/protocol/game/ServerboundDebugSubscriptionRequestPacket.java
index e619e37f3e5bb95cd4809859bdc0c02507216457..abd21a4b48ef90764de626be1db901f49fa2e654 100644
--- a/net/minecraft/network/protocol/game/ServerboundDebugSubscriptionRequestPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundDebugSubscriptionRequestPacket.java
@@ -11,10 +11,13 @@ import net.minecraft.network.protocol.PacketType;
 import net.minecraft.util.debug.DebugSubscription;
 
 public record ServerboundDebugSubscriptionRequestPacket(Set<DebugSubscription<?>> subscriptions) implements Packet<ServerGamePacketListener> {
-    private static final StreamCodec<RegistryFriendlyByteBuf, Set<DebugSubscription<?>>> SET_STREAM_CODEC = ByteBufCodecs.registry(
-            Registries.DEBUG_SUBSCRIPTION
-        )
-        .apply(ByteBufCodecs.collection(ReferenceOpenHashSet::new));
+    // Leaf start - Limit max size of subscription codec
+    private static final StreamCodec<RegistryFriendlyByteBuf, Set<DebugSubscription<?>>> SET_STREAM_CODEC = ByteBufCodecs.collection(
+        ReferenceOpenHashSet::new,
+        ByteBufCodecs.registry(Registries.DEBUG_SUBSCRIPTION),
+        64 // Check total count of subscriptions in DebugSubscriptions every mc update
+    );
+    // Leaf end - Limit max size of subscription codec
     public static final StreamCodec<RegistryFriendlyByteBuf, ServerboundDebugSubscriptionRequestPacket> STREAM_CODEC = SET_STREAM_CODEC.map(
         ServerboundDebugSubscriptionRequestPacket::new, ServerboundDebugSubscriptionRequestPacket::subscriptions
     );
