--- a/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java
@@ -10,9 +_,11 @@
 
 public abstract class MinecartBehavior {
     protected final AbstractMinecart minecart;
+    protected final boolean hasVehicleEntityCollisionEvent;
 
     protected MinecartBehavior(AbstractMinecart minecart) {
         this.minecart = minecart;
+        this.hasVehicleEntityCollisionEvent = org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0; // Leaf - Do not process VehicleEntityCollisionEvent if no listeners
     }
 
     public InterpolationHandler getInterpolation() {
