--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -525,13 +_,16 @@
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
-                        // CraftBukkit start
-                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                            entity.getBukkitEntity()
-                        );
-                        if (!collisionEvent.callEvent()) continue;
-                        // CraftBukkit end
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (hasVehicleEntityCollisionEvent) {
+                            // CraftBukkit start
+                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                            );
+                            if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
+                        }
+                        // Leaf end
                         boolean flag = entity.startRiding(this.minecart);
                         if (flag) {
                             return true;
@@ -557,13 +_,16 @@
                         || entity.isPassenger()) {
                         // CraftBukkit start
                         if (!this.minecart.isPassengerOfSameVehicle(entity)) {
-                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                                entity.getBukkitEntity()
-                            );
-                            if (!collisionEvent.callEvent()) {
-                                continue;
+                            // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                            if (hasVehicleEntityCollisionEvent) {
+                                // CraftBukkit start
+                                org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                    (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                                );
+                                if (!collisionEvent.callEvent()) continue;
+                                // CraftBukkit end
                             }
+                            // Leaf end
                         }
                         // CraftBukkit end
                         entity.push(this.minecart);
@@ -574,15 +_,16 @@
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, box)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
-                    // CraftBukkit start
-                    org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                        (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                        entity1.getBukkitEntity()
-                    );
-                    if (!collisionEvent.callEvent()) {
-                        continue;
+                    // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                    if (hasVehicleEntityCollisionEvent) {
+                        // CraftBukkit start
+                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
+                        );
+                        if (!collisionEvent.callEvent()) continue;
+                        // CraftBukkit end
                     }
-                    // CraftBukkit end
+                    // Leaf end
                     entity1.push(this.minecart);
                     flag = true;
                 }
