--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -376,22 +_,28 @@
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
-                        // CraftBukkit start
-                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
-                        );
-                        if (!collisionEvent.callEvent()) continue;
-                        // CraftBukkit end
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (hasVehicleEntityCollisionEvent) {
+                            // CraftBukkit start
+                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                            );
+                            if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
+                        }
+                        // Leaf end
                         entity.startRiding(this.minecart);
                     } else {
-                        // CraftBukkit start
-                        if (!this.minecart.isPassengerOfSameVehicle(entity)) {
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (hasVehicleEntityCollisionEvent) {
+                            // CraftBukkit start
                             org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                                 (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
                             );
                             if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
                         }
-                        // CraftBukkit end
+                        // Leaf end
                         entity.push(this.minecart);
                     }
                 }
@@ -399,12 +_,16 @@
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, aabb)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
-                    // CraftBukkit start
-                    org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                        (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
-                    );
-                    if (!collisionEvent.callEvent()) continue;
-                    // CraftBukkit end
+                    // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                    if (hasVehicleEntityCollisionEvent) {
+                        // CraftBukkit start
+                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
+                        );
+                        if (!collisionEvent.callEvent()) continue;
+                        // CraftBukkit end
+                    }
+                    // Leaf end
                     entity1.push(this.minecart);
                 }
             }
