From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Sun, 12 Jun 2022 10:57:51 -0500
Subject: [PATCH] Leaf config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d73b7d72d00fa0edd6542226348af14c5203cf8d..18158411949971973ddc738e903316da605e7f93 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1949,6 +1949,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Leaf start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLeafConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Leaf end
+
         /**
          * Sends the component to the player
          *
