From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 24 Feb 2024 01:16:07 -0500
Subject: [PATCH] Including 5s in getTPS()


diff --git a/net/minecraft/server/gui/StatsComponent.java b/net/minecraft/server/gui/StatsComponent.java
index f485e83fa6cbf510472bcffd18d29a41e177e856..7dd342f0c01f1ea4acad8ae50e1cd0bb5cb68a5f 100644
--- a/net/minecraft/server/gui/StatsComponent.java
+++ b/net/minecraft/server/gui/StatsComponent.java
@@ -34,6 +34,7 @@ public class StatsComponent extends JComponent {
         long l = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
         // Paper start - Improve ServerGUI
         double[] tps = org.bukkit.Bukkit.getTPS();
+        String tpsMsg = org.dreeam.leaf.config.modules.misc.Including5sIngetTPS.enabled ? "TPS from last 5s, 1m, 5m, 15m: " : "TPS from last 1m, 5m, 15m: "; // Leaf - Including 5s in getTPS()
         String[] tpsAvg = new String[tps.length];
 
         for (int g = 0; g < tps.length; g++) {
@@ -41,7 +42,7 @@ public class StatsComponent extends JComponent {
         }
         this.msgs[0] = "Memory use: " + l / 1024L / 1024L + " mb (" + Runtime.getRuntime().freeMemory() * 100L / Runtime.getRuntime().maxMemory() + "% free)";
         this.msgs[1] = "Avg tick: " + DECIMAL_FORMAT.format((double)this.server.getAverageTickTimeNanos() / TimeUtil.NANOSECONDS_PER_MILLISECOND) + " ms";
-        this.msgs[2] = "TPS from last 1m, 5m, 15m: " + String.join(", ", tpsAvg);
+        this.msgs[2] = tpsMsg + String.join(", ", tpsAvg); // Leaf - Including 5s in getTPS()
         // Paper end - Improve ServerGUI
         this.values[this.vp++ & 0xFF] = (int)(l * 100L / Runtime.getRuntime().maxMemory());
         this.repaint();
