From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 14 Jan 2024 05:14:09 -0500
Subject: [PATCH] Configurable server GUI name


diff --git a/net/minecraft/server/gui/MinecraftServerGui.java b/net/minecraft/server/gui/MinecraftServerGui.java
index f5c52f0aa85d8ae66215fbbe94773b16e06a72a5..b135dfbbbd5392b0e2c96c7142f6d5bf76ed4b7c 100644
--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -51,7 +51,7 @@ public class MinecraftServerGui extends JComponent {
         } catch (Exception var3) {
         }
 
-        final JFrame frame = new JFrame("Purpur Minecraft server"); // Purpur - Improve GUI
+        final JFrame frame = new JFrame(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName); // Purpur - Improve GUI // Leaf - Configurable server GUI name
         final MinecraftServerGui gui = new MinecraftServerGui(server);
         frame.setDefaultCloseOperation(2);
         frame.add(gui);
@@ -59,7 +59,7 @@ public class MinecraftServerGui extends JComponent {
         frame.setLocationRelativeTo(null);
         frame.setVisible(true);
         // Paper start - Improve ServerGUI
-        frame.setName("Purpur Minecraft server"); // Purpur - Improve GUI
+        frame.setName(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName); // Purpur - Improve GUI // Leaf - Configurable server GUI name
         try {
             frame.setIconImage(javax.imageio.ImageIO.read(java.util.Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("logo.png"))));
         } catch (java.io.IOException ignore) {
@@ -69,7 +69,7 @@ public class MinecraftServerGui extends JComponent {
             @Override
             public void windowClosing(final WindowEvent event) {
                 if (!gui.isClosing.getAndSet(true)) {
-                    frame.setTitle("Purpur Minecraft server - shutting down!"); // Purpur - Improve GUI
+                    frame.setTitle(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName + " - shutting down!"); // Purpur - Improve GUI // Leaf - Configurable server GUI name
                     server.halt(true);
                     gui.runFinalizers();
                 }
