From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 12 Oct 2022 14:48:45 -0400
Subject: [PATCH] Remove Spigot check for broken BungeeCord configurations


diff --git a/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index 0eaa9c9bae27ca91c3685c7b49a8cc82a6b9bd40..94967a2a735180f7a4860ef4de3a9ad9fcc680db 100644
--- a/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -168,7 +168,7 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                 if (split.length == 4) {
                     this.connection.spoofedProfile = ServerHandshakePacketListenerImpl.gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
                 }
-            } else if ((split.length == 3 || split.length == 4) && (ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher(split[1]).matches())) {
+            } else if (!org.dreeam.leaf.config.modules.misc.RemoveSpigotCheckBungee.enabled && (split.length == 3 || split.length == 4) && (ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher(split[1]).matches())) { // Leaf - Remove Spigot check for broken BungeeCord configurations
                 Component message = Component.literal("Unknown data in login hostname, did you forget to enable BungeeCord in spigot.yml?");
                 this.connection.send(new ClientboundLoginDisconnectPacket(message));
                 this.connection.disconnect(message);
