From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 4cac7a519fbf30ba137db87606f915df4dbccff8..96badfebc4f9550976d797bf63af886854d7f801 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1289,6 +1289,8 @@ public abstract class PlayerList {
             return uuidStatsFile;
         }
 
+        // Leaf start - Remove useless creating stats json bases on player name logic
+        /*
         String playerNameStatsFile = gameProfile.name() + ".json";
         if (FileUtil.isValidPathSegment(playerNameStatsFile)) {
             Path playerNameStatsPath = statFolder.resolve(playerNameStatsFile);
@@ -1301,6 +1303,8 @@ public abstract class PlayerList {
                 }
             }
         }
+        */
+        // Leaf end - Remove useless creating stats json bases on player name logic
 
         return uuidStatsFile;
     }
