From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 3e7ed544fbc2e4cd05e5208c5a8420236d974a1e..663d090bc29c219089b0f38f7c22300d1840c345 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1452,6 +1452,8 @@ public abstract class PlayerList {
             return uuidStatsFile;
         }
 
+        // Leaf start - Remove useless creating stats json bases on player name logic
+        /*
         String playerNameStatsFile = gameProfile.name() + ".json";
         if (FileUtil.isValidPathSegment(playerNameStatsFile)) {
             Path playerNameStatsPath = statFolder.resolve(playerNameStatsFile);
@@ -1464,6 +1466,8 @@ public abstract class PlayerList {
                 }
             }
         }
+        */
+        // Leaf end - Remove useless creating stats json bases on player name logic
 
         return uuidStatsFile;
     }
