From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 47db522a285b43dd526d097b3b4107f7e39ece5e..3e78dba5d890e4011422a0685c9f6854ee55739c 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1439,6 +1439,8 @@ public abstract class PlayerList {
             return uuidStatsFile;
         }
 
+        // Leaf start - Remove useless creating stats json bases on player name logic
+        /*
         String playerNameStatsFile = gameProfile.name() + ".json";
         if (FileUtil.isValidPathSegment(playerNameStatsFile)) {
             Path playerNameStatsPath = statFolder.resolve(playerNameStatsFile);
@@ -1451,6 +1453,8 @@ public abstract class PlayerList {
                 }
             }
         }
+        */
+        // Leaf end - Remove useless creating stats json bases on player name logic
 
         return uuidStatsFile;
     }
