From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index d6aaf0326a287a8936329bc1de61ab92b76e705a..bb6017af2dad42fe9789ac7ea74c8fdf7c7eb576 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1442,6 +1442,8 @@ public abstract class PlayerList {
             return uuidStatsFile;
         }
 
+        // Leaf start - Remove useless creating stats json bases on player name logic
+        /*
         String playerNameStatsFile = gameProfile.name() + ".json";
         if (FileUtil.isValidPathSegment(playerNameStatsFile)) {
             Path playerNameStatsPath = statFolder.resolve(playerNameStatsFile);
@@ -1454,6 +1456,8 @@ public abstract class PlayerList {
                 }
             }
         }
+        */
+        // Leaf end - Remove useless creating stats json bases on player name logic
 
         return uuidStatsFile;
     }
