From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 4eb9f5f9bcd94425fb49efca2d289e0d5b7c47d7..b34477282e4b033ef4bb3ae3ce23027c79c6d6ce 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1280,6 +1280,8 @@ public abstract class PlayerList {
         if (Files.exists(path)) {
             return path;
         } else {
+            // Leaf start - Remove useless creating stats json bases on player name logic
+            /*
             String string = profile.name() + ".json";
             if (FileUtil.isValidPathSegment(string)) {
                 Path path1 = worldPath.resolve(string);
@@ -1292,6 +1294,8 @@ public abstract class PlayerList {
                     }
                 }
             }
+            */
+            // Leaf end - Remove useless creating stats json bases on player name logic
 
             return path;
         }
