From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 7 Jun 2024 17:43:43 +0800
Subject: [PATCH] Remove stream in BlockBehaviour cache blockstate


diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 37a0140e9b1add79f9ffdb4292a4fc553d0df61e..b488e754e834e28cbf6b0ad83b610d438995bc40 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -1100,8 +1100,14 @@ public abstract class BlockBehaviour implements FeatureElement {
                     );
                 }
 
-                this.largeCollisionShape = Arrays.stream(Direction.Axis.values())
-                    .anyMatch(axis -> this.collisionShape.min(axis) < 0.0 || this.collisionShape.max(axis) > 1.0);
+                // Leaf start - Remove stream in BlockBehaviour cache blockstate
+                for (Direction.Axis axis : net.minecraft.core.Direction.Axis.VALUES) {
+                    if (this.collisionShape.min(axis) < 0.0 || this.collisionShape.max(axis) > 1.0) {
+                        this.largeCollisionShape = true;
+                        break;
+                    }
+                }
+                // Leaf end - Remove stream in BlockBehaviour cache blockstate
                 this.faceSturdy = new boolean[DIRECTIONS.length * SUPPORT_TYPE_COUNT];
 
                 for (Direction direction : DIRECTIONS) {
