From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 6 Apr 2025 11:22:35 +0200
Subject: [PATCH] Virtual thread for download pool


diff --git a/net/minecraft/util/Util.java b/net/minecraft/util/Util.java
index 98e96a7cfdbc6b7999a1c3bc55f08aa41914268a..2c6bb1b11b31758438bc4bc763407d1c92e1e0ad 100644
--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -99,7 +99,7 @@ public class Util {
     private static final TracingExecutor IO_POOL = makeIoExecutor("IO-Worker-", false);
     public static final TracingExecutor DIMENSION_DATA_IO_POOL = makeExtraIoExecutor("Dimension-Data-IO-Worker-"); // Paper - Separate dimension data IO pool
     // Paper start - Limit download pool size
-    private static final TracingExecutor DOWNLOAD_POOL = new TracingExecutor(Executors.newFixedThreadPool(4, new java.util.concurrent.ThreadFactory() {
+    private static final TracingExecutor DOWNLOAD_POOL = new TracingExecutor(Executors.newFixedThreadPool(4, createDownloadPoolFactory()/*new java.util.concurrent.ThreadFactory() { // Leaf - Virtual thread for download pool
         private final AtomicInteger count = new AtomicInteger();
 
         @Override
@@ -112,8 +112,37 @@ public class Util {
             });
             return ret;
         }
-    }));
+    }*/)); // Leaf - Virtual thread for download pool
     // Paper end - Limit download pool size
+
+    // Leaf start - Virtual thread for download pool
+    private static java.util.concurrent.ThreadFactory createDownloadPoolFactory() {
+        if (org.dreeam.leaf.config.modules.opt.VT4ProfileExecutor.enabled) {
+            return Thread.ofVirtual()
+                .name("Download-", 0)
+                .uncaughtExceptionHandler((Thread thread, Throwable throwable) -> {
+                    LOGGER.error("Uncaught exception in thread {}", thread.getName(), throwable);
+                })
+                .factory();
+        } else {
+            return new java.util.concurrent.ThreadFactory() {
+                private final AtomicInteger count = new AtomicInteger();
+
+                @Override
+                public Thread newThread(Runnable run) {
+                    Thread ret = new Thread(run);
+                    ret.setDaemon(true);
+                    ret.setName("Download-" + this.count.getAndIncrement());
+                    ret.setUncaughtExceptionHandler((Thread thread, Throwable throwable) -> {
+                        LOGGER.error("Uncaught exception in thread {}", thread.getName(), throwable);
+                    });
+                    return ret;
+                }
+            };
+        }
+    }
+    // Leaf end - Virtual thread for download pool
+
     private static final DateTimeFormatter FILENAME_DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH.mm.ss", Locale.ROOT);
     public static final int LINEAR_LOOKUP_THRESHOLD = 8;
     private static final Set<String> ALLOWED_UNTRUSTED_LINK_PROTOCOLS = Set.of("http", "https");
