From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 17 Oct 2024 01:51:38 -0400
Subject: [PATCH] Remove stream and double iteration in enough deep sleeping
 player check


diff --git a/net/minecraft/server/players/SleepStatus.java b/net/minecraft/server/players/SleepStatus.java
index 51f69631603b90b194080aeecffafb4d613b2af8..c89a67611b6633d9eaf9a836059c32094df12d89 100644
--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -15,8 +15,22 @@ public class SleepStatus {
 
     public boolean areEnoughDeepSleeping(final int sleepPercentageNeeded, final List<ServerPlayer> players) {
         // CraftBukkit start
-        int deepSleepers = (int)players.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping || (player.level().purpurConfig.idleTimeoutCountAsSleeping && player.isAfk())).count(); // Purpur - AFK API
-        boolean anyDeepSleep = players.stream().anyMatch(Player::isSleepingLongEnough);
+        // Leaf start - Remove stream and double iteration in enough deep sleeping player check
+        int deepSleepers = 0;
+        boolean anyDeepSleep = false;
+
+        for (ServerPlayer player : players) {
+            final boolean isSleepingLongEnough = player.isSleepingLongEnough();
+
+            if (isSleepingLongEnough) {
+                anyDeepSleep = true;
+            }
+            if (isSleepingLongEnough || player.fauxSleeping || (player.level().purpurConfig.idleTimeoutCountAsSleeping && player.isAfk())) { // Purpur - AFK API
+                deepSleepers++;
+            }
+        }
+        // Leaf end - Remove stream and double iteration in enough deep sleeping player check
+
         return anyDeepSleep && deepSleepers >= this.sleepersNeeded(sleepPercentageNeeded);
         // CraftBukkit end
     }
