From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: pixelNightly <pixelnightly@proton.me>
Date: Mon, 4 Dec 2023 16:11:36 +0200
Subject: [PATCH] Fix sprint glitch


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index d59caf13e9066f2171c360e07a26122a540380a0..97098f05dda8a46716162dd0330ca60d30814999 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1419,7 +1419,8 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 player.setRealHealth(health);
             }
 
-            player.updateScaledHealth(false);
+            //player.updateScaledHealth(false); // Leaf - Fix sprint glitch - commented out
+            this.entityData.set(LivingEntity.DATA_HEALTH_ID, player.getScaledHealth()); // Leaf - Fix sprint glitch
             return;
         }
         // CraftBukkit end
