From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Sat, 26 Oct 2024 00:56:24 +0800
Subject: [PATCH] Remove stream in BehaviorUtils


diff --git a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
index c50d6ac77cae556aaa7ce8f3f1dc3a706ccd9d31..42c4ec0446ef5f78fba953fcaebae911de658aa4 100644
--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -126,10 +126,33 @@ public class BehaviorUtils {
 
     public static SectionPos findSectionClosestToVillage(final ServerLevel level, final SectionPos center, final int radius) {
         int distToVillage = level.sectionsToVillage(center);
-        return SectionPos.cube(center, radius)
-            .filter(s -> level.sectionsToVillage(s) < distToVillage)
-            .min(Comparator.comparingInt(level::sectionsToVillage))
-            .orElse(center);
+        // Leaf start - Remove stream in BehaviorUtils
+        SectionPos closestPos = center;
+        int closestDistance = distToVillage;
+
+        final int lowerX = center.getX() - radius;
+        final int lowerY = center.getY() - radius;
+        final int lowerZ = center.getZ() - radius;
+        final int upperX = center.getX() + radius;
+        final int upperY = center.getY() + radius;
+        final int upperZ = center.getZ() + radius;
+
+        for (int x = lowerX; x <= upperX; x++) {
+            for (int z = lowerZ; z <= upperZ; z++) {
+                for (int y = lowerY; y <= upperY; y++) {
+                    SectionPos pos = SectionPos.of(x, y, z);
+                    int distance = level.sectionsToVillage(pos);
+
+                    if (distance < closestDistance) {
+                        closestDistance = distance;
+                        closestPos = pos;
+                    }
+                }
+            }
+        }
+
+        return closestPos;
+        // Leaf end - Remove stream in BehaviorUtils
     }
 
     public static boolean isWithinAttackRange(final Mob body, final LivingEntity target, final int projectileAttackRangeMargin) {
