From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Mon, 29 Apr 2024 23:30:21 +0800
Subject: [PATCH] Don't save primed tnt entity


diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 61c6aa038e5d92661dabc411e611e88dfc227ad9..5b9010788875858dba14c809953b853957c68357 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -265,4 +265,11 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         return super.interact(player, hand, location);
     }
     // Purpur end - Shears can defuse TNT
+
+    // Leaf start - PMC - Don't save primed tnt entity
+    @Override
+    public boolean shouldBeSaved() {
+        return !org.dreeam.leaf.config.modules.opt.DontSaveEntity.dontSavePrimedTNT && super.shouldBeSaved();
+    }
+    // Leaf - PMC - Don't save primed tnt entity
 }
