From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Mon, 29 Apr 2024 23:30:21 +0800
Subject: [PATCH] Don't save primed tnt entity


diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 5dccf56d19dd217ec4b18a390779f0da0c428b98..28aed230b3a7da6ccedf1b1d9aab300fb154c26c 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -263,4 +263,11 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         return super.interact(player, hand);
     }
     // Purpur end - Shears can defuse TNT
+
+    // Leaf start - PMC - Don't save primed tnt entity
+    @Override
+    public boolean shouldBeSaved() {
+        return !org.dreeam.leaf.config.modules.opt.DontSaveEntity.dontSavePrimedTNT && super.shouldBeSaved();
+    }
+    // Leaf - PMC - Don't save primed tnt entity
 }
