From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Author: PureGero" <puregero@gmail.com>
Date: Thu, 1 Aug 2024 00:43:05 +0900
Subject: [PATCH] ShreddedPaper: Don't block main thread in
 Connection#syncAfterConfigurationChange


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 570a6b844319491b1e07df0a3e55d0b29533c649..5a8133497d9c1260a3bba7e9dc7f3b059bbe6b1a 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -325,6 +325,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
     private static void syncAfterConfigurationChange(ChannelFuture future) {
         try {
+            if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) net.minecraft.server.MinecraftServer.getServer().managedBlock(future::isDone); // ShreddedPaper - Don't block main thread in Connection#syncAfterConfigurationChange
             future.syncUninterruptibly();
         } catch (Exception var2) {
             if (var2 instanceof ClosedChannelException) {
