From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 17 Sep 2024 02:26:44 -0400
Subject: [PATCH] Remove stream in entity visible effects filter


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index d59575ca6dd4e8b4e74f88b6ef6a7881565f28b5..090139cecfeff2ee4867ef34a6a42adc4d044ef1 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1050,12 +1050,14 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     private void updateSynchronizedMobEffectParticles() {
-        List<ParticleOptions> visibleEffectParticles = this.activeEffects
-            .values()
-            .stream()
-            .filter(MobEffectInstance::isVisible)
-            .map(MobEffectInstance::getParticleOptions)
-            .toList();
+        // Leaf start - Remove stream in entity visible effects filter
+        List<ParticleOptions> visibleEffectParticles = new java.util.ArrayList<>();
+        for (MobEffectInstance effect : this.activeEffects.values()) {
+            if (effect.isVisible()) {
+                visibleEffectParticles.add(effect.getParticleOptions());
+            }
+        }
+        // Leaf end - Remove stream in entity visible effects filter
         this.entityData.set(DATA_EFFECT_PARTICLES, visibleEffectParticles);
         this.entityData.set(DATA_EFFECT_AMBIENCE_ID, areAllEffectsAmbient(this.activeEffects.values()));
     }
