From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 22 Mar 2025 11:22:23 +0100
Subject: [PATCH] Remove stream on updateConnectedPlayersWithinRange


diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 41154db05edbb777fbe08e1e79b4f30f12a3cc80..fe3a84af09907491f0d94c14918afbfe5bdf5138 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -67,11 +67,15 @@ public class VaultSharedData {
     void updateConnectedPlayersWithinRange(
         final ServerLevel serverLevel, final BlockPos pos, final VaultServerData serverData, final VaultConfig config, final double limit
     ) {
-        Set<UUID> currentConnectedPlayers = config.playerDetector()
-            .detect(serverLevel, config.entitySelector(), pos, limit, false)
-            .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+        // Leaf start - Remove stream on updateConnectedPlayersWithinRange
+        java.util.List<UUID> detectedPlayers = config.playerDetector().detect(serverLevel, config.entitySelector(), pos, limit, false);
+        Set<UUID> currentConnectedPlayers = new ObjectLinkedOpenHashSet<>();
+        for (UUID uuid : detectedPlayers) {
+            if (!serverData.getRewardedPlayers().contains(uuid)) {
+                currentConnectedPlayers.add(uuid);
+            }
+        }
+        // Leaf end - Remove stream on updateConnectedPlayersWithinRange
         if (!this.connectedPlayers.equals(currentConnectedPlayers)) {
             this.connectedPlayers = currentConnectedPlayers;
             this.markDirty();
