From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 7 Jun 2024 17:43:43 +0800
Subject: [PATCH] Remove stream in BlockBehaviour cache blockstate


diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index a26e2b62735e3aa7c22afa5d356a4e9b9bf0e7f6..4ad288dfa49abd60d9c74e317a493e2b532b97f6 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -1064,8 +1064,14 @@ public abstract class BlockBehaviour implements FeatureElement {
                         )
                     );
                 } else {
-                    this.largeCollisionShape = Arrays.stream(Direction.Axis.values())
-                        .anyMatch(dir -> this.collisionShape.min(dir) < 0.0 || this.collisionShape.max(dir) > 1.0);
+                    // Leaf start - Remove stream in BlockBehaviour cache blockstate
+                    for (Direction.Axis axis : Direction.Axis.values()) {
+                        if (this.collisionShape.min(axis) < 0.0 || this.collisionShape.max(axis) > 1.0) {
+                            this.largeCollisionShape = true;
+                            break;
+                        }
+                    }
+                    // Leaf end - Remove stream in BlockBehaviour cache blockstate
                     this.faceSturdy = new boolean[DIRECTIONS.length * SUPPORT_TYPE_COUNT];
 
                     for (Direction direction : DIRECTIONS) {
