From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Sat, 26 Oct 2024 14:04:54 +0800
Subject: [PATCH] Remove stream in YieldJobSite


diff --git a/net/minecraft/world/entity/ai/behavior/YieldJobSite.java b/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
index 09cc0f75c0dea7d9ef9caf9b1f14f46e4da84375..a3751f59d1a058d9be95cea7e3ae43a51447cc35 100644
--- a/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -41,23 +41,26 @@ public class YieldJobSite {
                             return true;
                         }
 
-                        i.<List<LivingEntity>>get(nearestEntities)
-                            .stream()
-                            .filter(v -> v instanceof Villager && v != body)
-                            .map(v -> (Villager)v)
-                            .filter(LivingEntity::isAlive)
-                            .filter(v -> nearbyWantsJobsite(poiType.get(), v, poiPos))
-                            .findFirst()
-                            .ifPresent(nearbyVillager -> {
-                                walkTarget.erase();
-                                lookTarget.erase();
-                                potentialJob.erase();
-                                if (nearbyVillager.getBrain().getMemory(MemoryModuleType.JOB_SITE).isEmpty()) {
-                                    BehaviorUtils.setWalkAndLookTargetMemories(nearbyVillager, poiPos, speedModifier, 1);
-                                    nearbyVillager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(level.dimension(), poiPos));
-                                    level.debugSynchronizers().updatePoi(poiPos);
+                        // Leaf start - Remove stream in YieldJobSite
+                        List<LivingEntity> mobsList = i.get(nearestEntities);
+                        for (LivingEntity v : mobsList) {
+                            if (v instanceof Villager nearbyVillager && v != body && v.isAlive()) {
+                                if (nearbyWantsJobsite(poiType.get(), nearbyVillager, poiPos)) {
+                                    walkTarget.erase();
+                                    lookTarget.erase();
+                                    potentialJob.erase();
+
+                                    if (nearbyVillager.getBrain().getMemory(MemoryModuleType.JOB_SITE).isEmpty()) {
+                                        BehaviorUtils.setWalkAndLookTargetMemories(nearbyVillager, poiPos, speedModifier, 1);
+                                        nearbyVillager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(level.dimension(), poiPos));
+                                        level.debugSynchronizers().updatePoi(poiPos);
+                                    }
+
+                                    break;
                                 }
-                            });
+                            }
+                        }
+                        // Leaf end - Remove stream in YieldJobSite
                         return true;
                     }
                 )
