From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Tue, 4 Mar 2025 11:07:42 +0100
Subject: [PATCH] Reduce polling on POI search


diff --git a/net/minecraft/world/entity/ai/village/poi/PoiTypes.java b/net/minecraft/world/entity/ai/village/poi/PoiTypes.java
index e2875cf82165143852d512d9f0476abd6a6815f2..7fc0f0005b085e4cedeeb4b379f4523cc45334c1 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiTypes.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiTypes.java
@@ -7,6 +7,8 @@ import java.util.Locale;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
 import net.minecraft.Util;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
@@ -66,7 +68,8 @@ public class PoiTypes {
         .stream()
         .flatMap(block -> block.getStateDefinition().getPossibleStates().stream())
         .collect(ImmutableSet.toImmutableSet());
-    private static final Map<BlockState, Holder<PoiType>> TYPE_BY_STATE = Maps.newHashMap();
+    // Replace HashMap with a more efficient map implementation
+    private static final Object2ObjectOpenHashMap<BlockState, Holder<PoiType>> TYPE_BY_STATE = new Object2ObjectOpenHashMap<>();
 
     private static Set<BlockState> getBlockStates(Block block) {
         return ImmutableSet.copyOf(block.getStateDefinition().getPossibleStates());
@@ -124,6 +127,7 @@ public class PoiTypes {
         register(registry, BEE_NEST, getBlockStates(Blocks.BEE_NEST), 0, 1);
         register(registry, NETHER_PORTAL, getBlockStates(Blocks.NETHER_PORTAL), 0, 1);
         register(registry, LODESTONE, getBlockStates(Blocks.LODESTONE), 0, 1);
+        TYPE_BY_STATE.trim();
         return register(registry, LIGHTNING_ROD, getBlockStates(Blocks.LIGHTNING_ROD), 0, 1);
     }
 }
