From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Mon, 12 May 2025 19:11:16 +0200
Subject: [PATCH] Remove stream in MobSensor


diff --git a/net/minecraft/world/entity/ai/sensing/MobSensor.java b/net/minecraft/world/entity/ai/sensing/MobSensor.java
index ac9c50d680814b5752f7c712633555d587633598..f314441d7a01145d8338a7b6b189b04f11fcd06c 100644
--- a/net/minecraft/world/entity/ai/sensing/MobSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/MobSensor.java
@@ -46,10 +46,15 @@ public class MobSensor<T extends LivingEntity> extends Sensor<T> {
     public void checkForMobsNearby(final T body) {
         Optional<List<LivingEntity>> livingEntitiesMemory = body.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!livingEntitiesMemory.isEmpty()) {
-            boolean mobPresent = livingEntitiesMemory.get().stream().anyMatch(entity -> this.mobTest.test(body, entity));
-            if (mobPresent) {
-                this.mobDetected(body);
+            // Leaf start - Remove stream in MobSensor
+            List<LivingEntity> entities = livingEntitiesMemory.get();
+            for (LivingEntity entity : entities) {
+                if (this.mobTest.test(body, entity)) {
+                    this.mobDetected(body);
+                    break;
+                }
             }
+            // Leaf end - Remove stream in MobSensor
         }
     }
 
