From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 22 Mar 2025 11:22:23 +0100
Subject: [PATCH] Remove stream on updateConnectedPlayersWithinRange


diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 036087f8085f05b43703ceca3413e0b8ac0a412b..514d8167ccbcc00de95d34e5ef2203dcaf824da1 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -65,11 +65,17 @@ public class VaultSharedData {
     }
 
     void updateConnectedPlayersWithinRange(ServerLevel level, BlockPos pos, VaultServerData serverData, VaultConfig config, double deactivationRange) {
-        Set<UUID> set = config.playerDetector()
-            .detect(level, config.entitySelector(), pos, deactivationRange, false)
-            .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+        // Leaf start - Remove stream on updateConnectedPlayersWithinRange
+        java.util.List<UUID> detectedPlayers = config.playerDetector().detect(level, config.entitySelector(), pos, deactivationRange, false);
+        Set<UUID> set = new ObjectLinkedOpenHashSet<>();
+
+        for (UUID uuid : detectedPlayers) {
+            if (!serverData.getRewardedPlayers().contains(uuid)) {
+                set.add(uuid);
+            }
+        }
+        // Leaf end - Remove stream on updateConnectedPlayersWithinRange
+
         if (!this.connectedPlayers.equals(set)) {
             this.connectedPlayers = set;
             this.markDirty();
