From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Thu, 6 Mar 2025 00:26:19 +0100
Subject: [PATCH] Track each world MSPT


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 8db073db493476911c2fefdc799e6efa83cd0366..a54a9dd8a50e62fed33e5f8ef80401371b5f5893 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1676,7 +1676,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // Leaf - SparklyPaper parallel world ticking mod (move level ticking logic out for branch convergence)
     private void tickLevel(ServerLevel serverLevel, BooleanSupplier hasTimeLeft) {
         try {
+            long i = Util.getNanos(); // SparklyPaper - track world's MSPT
             serverLevel.tick(hasTimeLeft);
+            // SparklyPaper start - track world's MSPT
+            long j = Util.getNanos() - i;
+
+            // These are from the "tickServer" function
+            serverLevel.tickTimes5s.add(this.tickCount, j);
+            serverLevel.tickTimes10s.add(this.tickCount, j);
+            serverLevel.tickTimes60s.add(this.tickCount, j);
+            // SparklyPaper end
         } catch (Throwable levelTickingException) {
             CrashReport crashReport = CrashReport.forThrowable(levelTickingException, "Exception ticking world");
             serverLevel.fillReportDetails(crashReport);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 0864a8a59e4e6a8c76fe33446847e9ec2b95c5a2..f7545417f6810bd8ad5748dc2d21e6c451d909c9 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -574,6 +574,12 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
     // Paper end - chunk tick iteration
 
+        // SparklyPaper start - track world's MSPT
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+    // SparklyPaper end
+
     public ServerLevel(
         MinecraftServer server,
         Executor dispatcher,
