From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Mon, 29 Apr 2024 23:30:21 +0800
Subject: [PATCH] Don't save primed tnt entity


diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index ef070f8a9ab3a4676e2141f7c0bc20a000d0cc3a..1acfc005260628cc70ebf063acc92fb2904d7794 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -266,4 +266,11 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         return super.interact(player, hand);
     }
     // Purpur end - Shears can defuse TNT
+
+    // Leaf start - PMC - Don't save primed tnt entity
+    @Override
+    public boolean shouldBeSaved() {
+        return !org.dreeam.leaf.config.modules.opt.DontSaveEntity.dontSavePrimedTNT && super.shouldBeSaved();
+    }
+    // Leaf - PMC - Don't save primed tnt entity
 }
