From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 29 Oct 2025 02:13:08 -0400
Subject: [PATCH] Remove stream in CraftWorld#spawnParticle


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 0817fd8c0006272f1ed0bce560f1200324ff42ec..29e303e39d234a1f203c1aa5e182301f6a06143e 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2257,7 +2257,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
         for (int i = 0; i < receivers.size(); i++) { // Paper - particle API
             ServerPlayer player = receivers.get(i); // Paper - particle API
-            if (sender != null && !player.getBukkitEntity().canSee(sender.getBukkitEntity())) continue; // CraftBukkit
+            if (sender != null && !player.getBukkitEntity().canSee(sender.getBukkitEntity())) continue; // CraftBukkit // Leaf - Remove stream in CraftWorld#spawnParticle - diff on change
             if (this.sendParticles(player, overrideLimiter, x, y, z, packet)) {
                 result++;
             }
@@ -2266,6 +2266,44 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return result;
     }
 
+    // Leaf start - Remove stream in CraftWorld#spawnParticle
+    // Copy from ServerLevel#sendParticlesSource
+    public <T extends ParticleOptions> void sendParticlesSourceBukkit(
+        final List<org.bukkit.entity.Player> bukkitReceivers,
+        final @Nullable Entity sender,
+        final T particle,
+        final boolean overrideLimiter,
+        final boolean alwaysShow,
+        final double x,
+        final double y,
+        final double z,
+        final int count,
+        final double xDist,
+        final double yDist,
+        final double zDist,
+        final double speed
+    ) {
+        ClientboundLevelParticlesPacket packet = new ClientboundLevelParticlesPacket(
+            particle, overrideLimiter, alwaysShow, x, y, z, (float) xDist, (float) yDist, (float) zDist, (float) speed, count
+        );
+
+        if (bukkitReceivers == null) {
+            for (int i = 0, size = players.size(); i < size; i++) { // Paper - particle API
+                ServerPlayer player = players.get(i); // Paper - particle API
+                if (sender != null && !player.getBukkitEntity().canSee(sender.getBukkitEntity())) continue; // CraftBukkit
+                this.sendParticles(player, overrideLimiter, x, y, z, packet);
+            }
+        } else {
+            for (int i = 0, size = bukkitReceivers.size(); i < size; i++) { // Paper - particle API
+                org.bukkit.entity.Player bukkitPlayer = bukkitReceivers.get(i); // Paper - particle API
+                ServerPlayer player = ((org.bukkit.craftbukkit.entity.CraftPlayer) bukkitPlayer).getHandle();
+                if (sender != null && !player.getBukkitEntity().canSee(sender.getBukkitEntity())) continue; // CraftBukkit
+                this.sendParticles(player, overrideLimiter, x, y, z, packet);
+            }
+        }
+    }
+    // Leaf end - Remove stream in CraftWorld#spawnParticle
+
     public <T extends ParticleOptions> boolean sendParticles(
         final ServerPlayer player,
         final T particle,
