From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 29 Mar 2025 13:40:46 +0100
Subject: [PATCH] Async target finding


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index b6af8da084c83ee38bb3ecea6a98feb0c1c74d2a..561723eeb1dd80f7fdd9aff33f6e1c4c16def0ff 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -40,8 +40,10 @@ public final class ChunkEntitySlices {
 
     private final EntityCollectionBySection allEntities;
     private final EntityCollectionBySection hardCollidingEntities;
-    private final Reference2ObjectOpenHashMap<Class<? extends Entity>, EntityCollectionBySection> entitiesByClass;
-    private final Reference2ObjectOpenHashMap<EntityType<?>, EntityCollectionBySection> entitiesByType;
+    // Leaf start - Async target finding
+    private final Reference2ObjectMap<Class<? extends Entity>, EntityCollectionBySection> entitiesByClass;
+    private final Reference2ObjectMap<EntityType<?>, EntityCollectionBySection> entitiesByType;
+    // Leaf end - Async target finding
     private final EntityList entities = new EntityList();
 
     public FullChunkStatus status;
@@ -67,9 +69,15 @@ public final class ChunkEntitySlices {
 
         this.allEntities = new EntityCollectionBySection(this);
         this.hardCollidingEntities = new EntityCollectionBySection(this);
-        this.entitiesByClass = new Reference2ObjectOpenHashMap<>();
-        this.entitiesByType = new Reference2ObjectOpenHashMap<>();
-
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            this.entitiesByClass = it.unimi.dsi.fastutil.objects.Reference2ObjectMaps.synchronize(new Reference2ObjectOpenHashMap<>());
+            this.entitiesByType = it.unimi.dsi.fastutil.objects.Reference2ObjectMaps.synchronize(new Reference2ObjectOpenHashMap<>());
+        } else {
+            this.entitiesByClass = new Reference2ObjectOpenHashMap<>();
+            this.entitiesByType = new Reference2ObjectOpenHashMap<>();
+        }
+        // Leaf end - Async target finding
         this.status = status;
         this.chunkData = chunkData;
     }
@@ -248,14 +256,26 @@ public final class ChunkEntitySlices {
             this.hardCollidingEntities.addEntity(entity, sectionIndex);
         }
 
-        for (final Iterator<Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection>> iterator =
-             this.entitiesByClass.reference2ObjectEntrySet().fastIterator(); iterator.hasNext();) {
-            final Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection> entry = iterator.next();
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            synchronized (this.entitiesByClass) {
+                for (final var entry : this.entitiesByClass.reference2ObjectEntrySet()) {
+                    if (entry.getKey().isInstance(entity)) {
+                        entry.getValue().addEntity(entity, sectionIndex);
+                    }
+                }
+            }
+        } else {
+            for (final Iterator<Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection>> iterator =
+                 this.entitiesByClass.reference2ObjectEntrySet().iterator(); iterator.hasNext(); ) {
+                final Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection> entry = iterator.next();
 
-            if (entry.getKey().isInstance(entity)) {
-                entry.getValue().addEntity(entity, sectionIndex);
+                if (entry.getKey().isInstance(entity)) {
+                    entry.getValue().addEntity(entity, sectionIndex);
+                }
             }
         }
+        // Leaf end - Async target finding
 
         EntityCollectionBySection byType = this.entitiesByType.get(entity.getType());
         if (byType != null) {
@@ -282,14 +302,27 @@ public final class ChunkEntitySlices {
             this.hardCollidingEntities.removeEntity(entity, sectionIndex);
         }
 
-        for (final Iterator<Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection>> iterator =
-             this.entitiesByClass.reference2ObjectEntrySet().fastIterator(); iterator.hasNext();) {
-            final Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection> entry = iterator.next();
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            synchronized (this.entitiesByClass) {
+                for (final var entry : this.entitiesByClass.reference2ObjectEntrySet()) {
+                    if (entry.getKey().isInstance(entity)) {
+                        entry.getValue().removeEntity(entity, sectionIndex);
+                    }
+                }
+            }
+        } else {
+            for (final Iterator<Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection>> iterator =
+                 this.entitiesByClass.reference2ObjectEntrySet().iterator(); iterator.hasNext();) {
+                final Reference2ObjectMap.Entry<Class<? extends Entity>, EntityCollectionBySection> entry = iterator.next();
 
-            if (entry.getKey().isInstance(entity)) {
-                entry.getValue().removeEntity(entity, sectionIndex);
+                if (entry.getKey().isInstance(entity)) {
+                    entry.getValue().removeEntity(entity, sectionIndex);
+                }
             }
         }
+        // Leaf end - Async target finding
+
 
         final EntityCollectionBySection byType = this.entitiesByType.get(entity.getType());
         byType.removeEntity(entity, sectionIndex);
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 24926aa7ed5c78b235659daf18b224b14beb744c..7841c300250d1fc6f964386bfa707391f5253f16 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -292,6 +292,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public gg.pufferfish.pufferfish.util.AsyncExecutor mobSpawnExecutor = new gg.pufferfish.pufferfish.util.AsyncExecutor("Leaf Async Mob Spawn Thread"); // Pufferfish - optimize mob spawning // Leaf - Fix Pufferfish and Purpur patches - Unify thread name
     public final Set<net.minecraft.world.entity.Entity> entitiesWithScheduledTasks = java.util.concurrent.ConcurrentHashMap.newKeySet(); // SparklyPaper - skip EntityScheduler's executeTick checks if there isn't any tasks to be run (concurrent because plugins may schedule tasks async)
     public java.util.concurrent.Semaphore serverLevelTickingSemaphore = null; // SparklyPaper - parallel world ticking
+    @Nullable public List<org.dreeam.leaf.async.ai.AsyncGoalThread> asyncGoalThread; // Leaf - Async target finding
 
     public static <S extends MinecraftServer> S spin(Function<Thread, S> threadFunction) {
         ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.init(); // Paper - rewrite data converter system
@@ -1088,6 +1089,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             org.dreeam.leaf.async.AsyncPlayerDataSaving.IO_POOL.awaitTermination(30, java.util.concurrent.TimeUnit.SECONDS);
         } catch (java.lang.InterruptedException ignored) {}
         // Leaf end - Async playerdata saving
+        // Leaf start - Async target finding
+        if (this.asyncGoalThread != null) {
+            for (var goalThread : this.asyncGoalThread) {
+                try {
+                    goalThread.join();
+                } catch (java.lang.InterruptedException ignored) {
+                }
+            }
+        }
+        // Leaf end - Async target finding
     }
 
     public String getLocalIp() {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 33dd16a26edd2974f04d9a868d3e58e8e3060032..7b395332fc049985a99204f561bd69302bc3b606 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -255,6 +255,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             DedicatedServer.LOGGER.info("Using " + serverLevelTickingSemaphore.availablePermits() + " permits for parallel world ticking"); // SparklyPaper - parallel world ticking
         }
         // Leaf end - SparklyPaper - parallel world ticking mod (make configurable)
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            asyncGoalThread = new java.util.ArrayList<>(org.dreeam.leaf.config.modules.async.AsyncTargetFinding.threadCount);
+            for (int i = org.dreeam.leaf.config.modules.async.AsyncTargetFinding.threadCount; i > 0; i--) {
+                asyncGoalThread.add(new org.dreeam.leaf.async.ai.AsyncGoalThread(this));
+            }
+        }
+        // Leaf end - Async target finding
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
 
         // Gale start - Pufferfish - SIMD support
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 9af7dafe03812d96aa477584d4147a68c240ab21..8f0e29e33262099debb5e31959a6fabb0fb86528 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -177,7 +177,16 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int EMPTY_TIME_NO_TICK = 300;
     private static final int MAX_SCHEDULED_TICKS_PER_TICK = 65536;
-    final List<ServerPlayer> players = Lists.newArrayList();
+    // Leaf start - Async target finding
+    final List<ServerPlayer> players;
+    {
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            this.players = Lists.newCopyOnWriteArrayList();
+        } else {
+            this.players = Lists.newArrayList();
+        }
+    }
+    // Leaf end - Async target finding
     public final ServerChunkCache chunkSource;
     private final MinecraftServer server;
     public final net.minecraft.world.level.storage.PrimaryLevelData serverLevelData; // CraftBukkit - type
@@ -218,6 +227,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
     public boolean hasRidableMoveEvent = false; // Purpur - Ridables
     final List<ServerPlayer> realPlayers; // Leaves - skip
+    public final @Nullable org.dreeam.leaf.async.ai.AsyncGoalExecutor asyncGoalExecutor; // Leaf - Async target finding
 
     public LevelChunk getChunkIfLoaded(int x, int z) {
         return this.chunkSource.getChunkAtIfLoadedImmediately(x, z); // Paper - Use getChunkIfLoadedImmediately
@@ -335,6 +345,12 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
     }
 
+    // Leaf start - Async target finding
+    public final void leafMidTickTasks() {
+        if (this.asyncGoalExecutor != null) this.asyncGoalExecutor.midTick();
+    }
+    // Leaf end - Async target finding
+
     @Override
     public final ChunkAccess moonrise$syncLoadNonFull(final int chunkX, final int chunkZ, final net.minecraft.world.level.chunk.status.ChunkStatus status) {
         return this.moonrise$getChunkTaskScheduler().syncLoadNonFull(chunkX, chunkZ, status);
@@ -711,6 +727,13 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         this.preciseTime = this.serverLevelData.getDayTime(); // Purpur - Configurable daylight cycle
         this.realPlayers = Lists.newArrayList(); // Leaves - skip
         this.tickExecutor = java.util.concurrent.Executors.newSingleThreadExecutor(new org.dreeam.leaf.async.world.SparklyPaperServerLevelTickExecutorThreadFactory(getWorld().getName())); // SparklyPaper - parallel world ticking
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            this.asyncGoalExecutor = new org.dreeam.leaf.async.ai.AsyncGoalExecutor(this);
+        } else {
+            this.asyncGoalExecutor = null;
+        }
+        // Leaf end - Async target finding
     }
 
     // Paper start
@@ -855,12 +878,14 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                                     }
                                     this.moonrise$midTickTasks(); // Paper - rewrite chunk system
                                     // Gale end - Airplane - remove lambda from ticking guard - copied from guardEntityTick
+                                    this.leafMidTickTasks(); // Leaf - Async target finding
                                 }
                             }
                         }
                     }
                 );
             this.tickBlockEntities();
+            if (this.asyncGoalExecutor != null) this.asyncGoalExecutor.tick(); // Leaf - Async target finding
         }
 
         // Paper - rewrite chunk system
@@ -1329,6 +1354,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         // Leaf end - SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
         // Paper end - rewrite chunk system
 
+        this.leafMidTickTasks(); // Leaf - Async target finding
     }
 
     private void tickBlock(BlockPos pos, Block block) {
@@ -1345,6 +1371,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         // Leaf end - SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
         // Paper end - rewrite chunk system
 
+        this.leafMidTickTasks(); // Leaf - Async target finding
     }
 
     // Paper start - log detailed entity tick information
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 6081bd14f30aa9d376c79ff06fbe8c7ae4a8e10a..899d6092a0433010cdee98e2a07dca5ca3f4b68c 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -225,12 +225,20 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         }
         // Paper end - Skip AI during inactive ticks for non-aware mobs
         boolean isThrottled = org.dreeam.leaf.config.modules.opt.ThrottleInactiveGoalSelectorTick.enabled && _pufferfish_inactiveTickDisableCounter++ % 20 != 0; // Pufferfish - throttle inactive goal selector ticking
+        // Leaf start - Async target finding
+        boolean running = this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1;
         if (this.goalSelector.inactiveTick(this.activatedPriority, true) && !isThrottled) { // Pufferfish - pass activated priroity // Pufferfish - throttle inactive goal selector ticking
             this.goalSelector.tick();
         }
         if (this.targetSelector.inactiveTick(this.activatedPriority, true)) { // Pufferfish - pass activated priority
             this.targetSelector.tick();
         }
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.goal) {
+            if (!running && (this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1)) {
+                ((ServerLevel) this.level()).asyncGoalExecutor.submit(this.getId());
+            }
+        }
+        // Leaf end - Async target finding
     }
     // Paper end
 
@@ -920,6 +928,8 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         // Paper end - Allow nerfed mobs to jump and float
         this.sensing.tick();
         int i = this.tickCount + this.getId();
+        // Leaf start - Async target finding
+        boolean running = this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1;
         if (i % 2 != 0 && this.tickCount > 1) {
             if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
             this.targetSelector.tickRunningGoals(false);
@@ -931,6 +941,12 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
             if (this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
             this.goalSelector.tick();
         }
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.goal) {
+            if (!running && (this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1)) {
+                ((ServerLevel) this.level()).asyncGoalExecutor.submit(this.getId());
+            }
+        }
+        // Leaf end - Async target finding
 
         this.navigation.tick();
         this.customServerAiStep((ServerLevel)this.level());
diff --git a/net/minecraft/world/entity/ai/control/LookControl.java b/net/minecraft/world/entity/ai/control/LookControl.java
index 8d71d7bf537a5cd676c24e5c8e4e2131fd71b8de..727b731269ca7ab4b5cf810d576c08b439d5788f 100644
--- a/net/minecraft/world/entity/ai/control/LookControl.java
+++ b/net/minecraft/world/entity/ai/control/LookControl.java
@@ -36,7 +36,7 @@ public class LookControl implements Control {
         this.setLookAt(x, y, z, this.mob.getHeadRotSpeed(), this.mob.getMaxHeadXRot());
     }
 
-    public void setLookAt(double x, double y, double z, float deltaYaw, float deltaPitch) {
+    public synchronized void setLookAt(double x, double y, double z, float deltaYaw, float deltaPitch) { // Leaf
         this.wantedX = x;
         this.wantedY = y;
         this.wantedZ = z;
@@ -45,7 +45,7 @@ public class LookControl implements Control {
         this.lookAtCooldown = 2;
     }
 
-    public void tick() {
+    public synchronized void tick() { // Leaf
         if (this.resetXRotOnTick()) {
             this.mob.setXRot(0.0F);
         }
diff --git a/net/minecraft/world/entity/ai/goal/BegGoal.java b/net/minecraft/world/entity/ai/goal/BegGoal.java
index 28ef40e8a645989ea181297069cf2bbe571f3082..451c736bc7b06107b2985baade46f2c7287273c9 100644
--- a/net/minecraft/world/entity/ai/goal/BegGoal.java
+++ b/net/minecraft/world/entity/ai/goal/BegGoal.java
@@ -59,10 +59,10 @@ public class BegGoal extends Goal {
         this.lookTime--;
     }
 
-    private boolean playerHoldingInteresting(Player player) {
+    private static boolean playerHoldingInteresting(Player player) { // Leaf start - Async Target Finding - static
         for (InteractionHand interactionHand : InteractionHand.values()) {
             ItemStack itemInHand = player.getItemInHand(interactionHand);
-            if (itemInHand.is(Items.BONE) || this.wolf.isFood(itemInHand)) {
+            if (itemInHand.is(Items.BONE) || itemInHand.is(net.minecraft.tags.ItemTags.WOLF_FOOD)) { // Leaf end - Async Target Finding
                 return true;
             }
         }
diff --git a/net/minecraft/world/entity/ai/goal/CatLieOnBedGoal.java b/net/minecraft/world/entity/ai/goal/CatLieOnBedGoal.java
index 4c46cd105cde3cbcde65a02ff691c3a8edd56c76..b3205c9f35687bc37124876198ec2d657ccaa96c 100644
--- a/net/minecraft/world/entity/ai/goal/CatLieOnBedGoal.java
+++ b/net/minecraft/world/entity/ai/goal/CatLieOnBedGoal.java
@@ -52,6 +52,13 @@ public class CatLieOnBedGoal extends MoveToBlockGoal {
 
     @Override
     protected boolean isValidTarget(LevelReader level, BlockPos pos) {
-        return level.isEmptyBlock(pos.above()) && level.getBlockState(pos).is(BlockTags.BEDS);
+        return level.isEmptyBlock(pos.above()) && level.getBlockState(pos).is(BlockTags.BEDS); // Leaf - Async search block - diff on change
     }
+
+    // Leaf start - Async search block
+    @Override
+    protected TypeToCheck typeToCheck() {
+        return TypeToCheck.CatLie;
+    }
+    // Leaf end - Async search block
 }
diff --git a/net/minecraft/world/entity/ai/goal/CatSitOnBlockGoal.java b/net/minecraft/world/entity/ai/goal/CatSitOnBlockGoal.java
index 9954f49bc364969c7ccb37f4186fa2ab8710f6ae..057090e3134048e75dbaefb703e8f2d35a172a3b 100644
--- a/net/minecraft/world/entity/ai/goal/CatSitOnBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/CatSitOnBlockGoal.java
@@ -44,14 +44,21 @@ public class CatSitOnBlockGoal extends MoveToBlockGoal {
 
     @Override
     protected boolean isValidTarget(LevelReader level, BlockPos pos) {
-        if (!level.isEmptyBlock(pos.above())) {
+        if (!level.isEmptyBlock(pos.above())) { // Leaf - Async search block - diff on change
             return false;
         } else {
-            BlockState blockState = level.getBlockState(pos);
+            BlockState blockState = level.getBlockState(pos); // Leaf - Async search block - diff on change
             return blockState.is(Blocks.CHEST)
                 ? ChestBlockEntity.getOpenCount(level, pos) < 1
                 : blockState.is(Blocks.FURNACE) && blockState.getValue(FurnaceBlock.LIT)
                     || blockState.is(BlockTags.BEDS, state -> state.getOptionalValue(BedBlock.PART).map(bedPart -> bedPart != BedPart.HEAD).orElse(true));
         }
     }
+
+    // Leaf start - Async search block
+    @Override
+    protected TypeToCheck typeToCheck() {
+        return TypeToCheck.CatSit;
+    }
+    // Leaf end - Async search block
 }
diff --git a/net/minecraft/world/entity/ai/goal/FollowFlockLeaderGoal.java b/net/minecraft/world/entity/ai/goal/FollowFlockLeaderGoal.java
index 1607d15e4a4b31bf44f2dcc65ac8d4835de28801..599678b649778bbac3a5b39c3afaeb060ab4ff56 100644
--- a/net/minecraft/world/entity/ai/goal/FollowFlockLeaderGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowFlockLeaderGoal.java
@@ -5,7 +5,7 @@ import java.util.List;
 import java.util.function.Predicate;
 import net.minecraft.world.entity.animal.AbstractSchoolingFish;
 
-public class FollowFlockLeaderGoal extends Goal {
+public class FollowFlockLeaderGoal extends Goal implements org.dreeam.leaf.async.ai.AsyncGoal { // Leaf
     private static final int INTERVAL_TICKS = 200;
     private final AbstractSchoolingFish mob;
     private int timeToRecalcPath;
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index e82e32407cec6109b9c3b0106295217f4a3f4aa2..32779f64c1e10f63ecd74259bc240147d88422f9 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -8,7 +8,7 @@ import java.util.Map;
 import java.util.Set;
 import java.util.function.Predicate;
 
-public class GoalSelector {
+public class GoalSelector implements org.dreeam.leaf.async.ai.VWaker { // Leaf
     private static final WrappedGoal NO_GOAL = new WrappedGoal(Integer.MAX_VALUE, new Goal() {
         @Override
         public boolean canUse() {
@@ -26,13 +26,23 @@ public class GoalSelector {
     private final ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<net.minecraft.world.entity.ai.goal.Goal.Flag> goalTypes = new ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<>(Goal.Flag.class); // Paper - remove streams from GoalSelector
     private int curRate; // Paper - EAR 2
 
+    // Leaf start - Async target finding
+    private boolean ctxGoalsDirty = true;
+    private WrappedGoal[] ctxGoals = {};
+    private int ctxIndex = 0;
+    public volatile int ctxState = -1;
+    private int[] ctxPriorityCache = new int[GOAL_FLAG_VALUES.length];
+    // Leaf end - Async target finding
+
     public void addGoal(int priority, Goal goal) {
         this.availableGoals.add(new WrappedGoal(priority, goal));
+        ctxGoalsDirty = true; // Leaf - Async target finding
     }
 
     @VisibleForTesting
     public void removeAllGoals(Predicate<Goal> filter) {
         this.availableGoals.removeIf(wrappedGoal -> filter.test(wrappedGoal.getGoal()));
+        ctxGoalsDirty = true; // Leaf - Async target finding
     }
 
     // Paper start - EAR 2
@@ -63,18 +73,19 @@ public class GoalSelector {
         }
 
         this.availableGoals.removeIf(wrappedGoal1 -> wrappedGoal1.getGoal() == goal);
+        ctxGoalsDirty = true; // Leaf - Async target finding
     }
 
     // Paper start - Perf: optimize goal types
     private static boolean goalContainsAnyFlags(WrappedGoal goal, ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<Goal.Flag> flags) {
-        return goal.getFlags().hasCommonElements(flags);
+        return goal.getFlags().hasCommonElements(flags); // Leaf - Async target finding - inline diff
     }
 
     private static boolean goalCanBeReplacedForAllFlags(WrappedGoal goal, Map<Goal.Flag, WrappedGoal> flag) {
         long flagIterator = goal.getFlags().getBackingSet();
         int wrappedGoalSize = goal.getFlags().size();
         for (int i = 0; i < wrappedGoalSize; ++i) {
-            final Goal.Flag flag1 = GOAL_FLAG_VALUES[Long.numberOfTrailingZeros(flagIterator)];
+            final Goal.Flag flag1 = GOAL_FLAG_VALUES[Long.numberOfTrailingZeros(flagIterator)]; // Leaf - Async target finding - inline diff
             flagIterator ^= ca.spottedleaf.concurrentutil.util.IntegerUtil.getTrailingBit(flagIterator);
             // Paper end - Perf: optimize goal types
             if (!flag.getOrDefault(flag1, NO_GOAL).canBeReplacedBy(goal)) {
@@ -85,7 +96,153 @@ public class GoalSelector {
         return true;
     }
 
+    // Leaf start - Async target finding
+    @Override
+    public final it.unimi.dsi.fastutil.objects.ReferenceArrayList<Goal> wake() {
+        var tickGoal = new it.unimi.dsi.fastutil.objects.ReferenceArrayList<Goal>();
+        if (ctxState == 0) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                // entity tempt
+                if (goal.isRunning() && (goalContainsAnyFlags(goal, this.goalTypes) || !goal.canContinueToUse())) {
+                    goal.stop();
+                }
+
+                ctxIndex++;
+            }
+
+            for (int i = 0; i < GOAL_FLAG_VALUES.length; i++) {
+                final Goal.Flag flag = GOAL_FLAG_VALUES[i];
+                WrappedGoal goal = lockedFlags.get(flag);
+                if (goal != null && !goal.isRunning()) {
+                    lockedFlags.remove(flag);
+                    ctxPriorityCache[i] = Integer.MAX_VALUE;
+                }
+            }
+
+            ctxIndex = 0;
+            ctxState = 1;
+        }
+        if (ctxState == 1) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                long flags = goal.getFlags().getBackingSet();
+                if (!goal.isRunning() && (this.goalTypes.getBackingSet() & flags) == 0L && unrolledIter1((int) flags, goal.getPriority(), ctxPriorityCache) && goal.canUse()) {
+                    unrolledIter2((int) flags, goal);
+                    goal.start();
+                }
+                ctxIndex++;
+            }
+
+            ctxIndex = 0;
+            ctxState = 2;
+        }
+        if (ctxState == 2) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                if (goal.isRunning()) {
+                    var unwrap = goal.getGoal();
+                    if (unwrap instanceof org.dreeam.leaf.async.ai.AsyncGoal asyncGoal && asyncGoal.tickAsync()) {
+                        unwrap.tick();
+                    } else {
+                        tickGoal.add(unwrap);
+                    }
+                }
+                ctxIndex++;
+            }
+
+            ctxState = -1;
+            ctxIndex = 0;
+        }
+        if (ctxState == 3) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                if (goal.isRunning() && goal.requiresUpdateEveryTick()) {
+                    var unwrap = goal.getGoal();
+                    if (unwrap instanceof org.dreeam.leaf.async.ai.AsyncGoal asyncGoal && asyncGoal.tickAsync()) {
+                        unwrap.tick();
+                    } else {
+                        tickGoal.add(unwrap);
+                    }
+                }
+                ctxIndex++;
+            }
+
+            ctxState = -1;
+            ctxIndex = 0;
+        }
+        return tickGoal;
+    }
+
+    private static boolean unrolledIter1(int iter, int priority, int[] arr) {
+        return iter == 0 || (((iter & 0b1) == 0 || priority < arr[0])
+            && ((iter & 0b10) == 0 || priority < arr[1])
+            && ((iter & 0b100) == 0 || priority < arr[2])
+            && ((iter & 0b1000) == 0 || priority < arr[3])
+            && ((iter & 0b10000) == 0 || priority < arr[4]));
+    }
+    private void unrolledIter2(int iter, WrappedGoal goal) {
+        if (iter == 0) {
+            return;
+        }
+        int priority = goal.isInterruptable() ? goal.getPriority() : -1;
+        if ((iter & 0b1) != 0) {
+            WrappedGoal prev = lockedFlags.put(GOAL_FLAG_VALUES[0], goal);
+            if (prev != null) prev.stop();
+            ctxPriorityCache[0] = priority;
+        }
+        if ((iter & 0b10) != 0) {
+            WrappedGoal prev = lockedFlags.put(GOAL_FLAG_VALUES[1], goal);
+            if (prev != null) prev.stop();
+            ctxPriorityCache[1] = priority;
+        }
+        if ((iter & 0b100) != 0) {
+            WrappedGoal prev = lockedFlags.put(GOAL_FLAG_VALUES[2], goal);
+            if (prev != null) prev.stop();
+            ctxPriorityCache[2] = priority;
+        }
+        if ((iter & 0b1000) != 0) {
+            WrappedGoal prev = lockedFlags.put(GOAL_FLAG_VALUES[3], goal);
+            if (prev != null) prev.stop();
+            ctxPriorityCache[3] = priority;
+        }
+        if ((iter & 0b10000) != 0) {
+            WrappedGoal prev = lockedFlags.put(GOAL_FLAG_VALUES[4], goal);
+            if (prev != null) prev.stop();
+            ctxPriorityCache[4] = priority;
+        }
+    }
+    // Leaf end - Async target finding
+
     public void tick() {
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.goal) {
+            if (ctxState == -1) {
+                if (ctxGoalsDirty) {
+                    this.ctxGoals = this.availableGoals.toArray(new WrappedGoal[0]);
+                    ctxGoalsDirty = false;
+                    if (ctxPriorityCache.length == 0) {
+                        ctxPriorityCache = new int[GOAL_FLAG_VALUES.length];
+                    }
+                    for (int i = 0; i < GOAL_FLAG_VALUES.length; i++) {
+                        final WrappedGoal goal = lockedFlags.get(GOAL_FLAG_VALUES[i]);
+                        if (goal != null) {
+                            if (goal.isInterruptable()) {
+                                ctxPriorityCache[i] = goal.getPriority();
+                            } else {
+                                ctxPriorityCache[i] = -1;
+                            }
+                        } else {
+                            ctxPriorityCache[i] = Integer.MAX_VALUE;
+                        }
+                    }
+                }
+                ctxState = 0;
+            }
+            return;
+        }
+        // Leaf end - Async target finding
+
         for (WrappedGoal wrappedGoal : this.availableGoals) {
             if (wrappedGoal.isRunning() && (goalContainsAnyFlags(wrappedGoal, this.goalTypes) || !wrappedGoal.canContinueToUse())) { // Paper - Perf: optimize goal types by removing streams
                 wrappedGoal.stop();
@@ -116,6 +273,33 @@ public class GoalSelector {
     }
 
     public void tickRunningGoals(boolean tickAllRunning) {
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.goal) {
+            if (ctxState == -1) {
+                if (ctxGoalsDirty) {
+                    this.ctxGoals = this.availableGoals.toArray(new WrappedGoal[0]);
+                    ctxGoalsDirty = false;
+                    if (ctxPriorityCache.length == 0) {
+                        ctxPriorityCache = new int[GOAL_FLAG_VALUES.length];
+                    }
+                    for (int i = 0; i < GOAL_FLAG_VALUES.length; i++) {
+                        final WrappedGoal goal = lockedFlags.get(GOAL_FLAG_VALUES[i]);
+                        if (goal != null) {
+                            if (goal.isInterruptable()) {
+                                ctxPriorityCache[i] = goal.getPriority();
+                            } else {
+                                ctxPriorityCache[i] = -1;
+                            }
+                        } else {
+                            ctxPriorityCache[i] = Integer.MAX_VALUE;
+                        }
+                    }
+                }
+                ctxState = tickAllRunning ? 2 : 3;
+            }
+            return;
+        }
+        // Leaf end - Async target finding
         for (WrappedGoal wrappedGoal : this.availableGoals) {
             if (wrappedGoal.isRunning() && (tickAllRunning || wrappedGoal.requiresUpdateEveryTick())) {
                 wrappedGoal.tick();
diff --git a/net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java b/net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java
index be59d0c27a83b329ec3f97c029cfb9c114e22472..ed27fe6fa9b99eb8df18a2481cd0816d0d1cba10 100644
--- a/net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java
+++ b/net/minecraft/world/entity/ai/goal/LlamaFollowCaravanGoal.java
@@ -22,7 +22,6 @@ public class LlamaFollowCaravanGoal extends Goal {
 
     @Override
     public boolean canUse() {
-        if (!this.llama.level().purpurConfig.llamaJoinCaravans || !this.llama.shouldJoinCaravan) return false; // Purpur - Llama API // Purpur - Config to disable Llama caravans
         if (!this.llama.isLeashed() && !this.llama.inCaravan()) {
             List<Entity> entities = this.llama.level().getEntities(this.llama, this.llama.getBoundingBox().inflate(9.0, 4.0, 9.0), entity1 -> {
                 EntityType<?> type = entity1.getType();
diff --git a/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java b/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
index 6463c3c9b08d6058f2843c225b08a40fc30a960b..772e69e5f7fadb6dac77d6c560f35195f8ebc80b 100644
--- a/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
@@ -11,7 +11,7 @@ import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.player.Player;
 
-public class LookAtPlayerGoal extends Goal {
+public class LookAtPlayerGoal extends Goal implements org.dreeam.leaf.async.ai.AsyncGoal { // Leaf
     public static final float DEFAULT_PROBABILITY = 0.02F;
     protected final Mob mob;
     @Nullable
diff --git a/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java b/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
index 3f080b15543bf8c5fa0774b62d7f12e13b82511a..a6f89adf86de16eeefdf4150d0092a0dca9519d0 100644
--- a/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
@@ -41,6 +41,43 @@ public abstract class MoveToBlockGoal extends Goal {
         this.setFlags(EnumSet.of(Goal.Flag.MOVE, Goal.Flag.JUMP));
     }
 
+    // Leaf start - Async search block
+    protected boolean findBlock() {
+        final var mob = this.mob;
+        final var serverLevel = getServerLevel(mob);
+        final TypeToCheck ty = this.typeToCheck();
+        if (ty == TypeToCheck.Unknown) {
+            return findNearestBlock();
+        }
+        final net.minecraft.world.level.block.Block toRemove;
+        if (this instanceof RemoveBlockGoal removeBlockGoal) {
+            toRemove = removeBlockGoal.blockToRemove;
+        } else {
+            toRemove = null;
+        }
+        final int verticalSearchStart = this.verticalSearchStart;
+        final int searchRange = this.searchRange;
+        final int verticalSearchRange = this.verticalSearchRange;
+        final BlockPos blockPos = mob.blockPosition();
+        final float restrictRadius = mob.getRestrictRadius();
+        final BlockPos restrictCenter = mob.getRestrictCenter();
+        return findNearestBlockAsync(ty, toRemove, mob, serverLevel, verticalSearchStart, searchRange, verticalSearchRange, blockPos, restrictRadius, restrictCenter);
+    }
+
+    protected enum TypeToCheck {
+        CatLie,
+        CatSit,
+        Drowned,
+        FoxEat,
+        RaidGarden,
+        RemoveBlock,
+        Strider,
+        TurtleToWater,
+        TurtleLay,
+        Unknown,
+    }
+    // Leaf end - Async search block
+
     @Override
     public boolean canUse() {
         if (this.nextStartTick > 0) {
@@ -48,7 +85,7 @@ public abstract class MoveToBlockGoal extends Goal {
             return false;
         } else {
             this.nextStartTick = this.nextStartTick(this.mob);
-            return this.findNearestBlock();
+            return this.findBlock(); // Leaf - Async search block
         }
     }
 
@@ -133,5 +170,108 @@ public abstract class MoveToBlockGoal extends Goal {
         return false;
     }
 
+    // Leaf start - Async search block
+    protected static boolean findNearestBlockAsync(
+        final TypeToCheck ty,
+        @org.jetbrains.annotations.Nullable final net.minecraft.world.level.block.Block toRemove,
+        final PathfinderMob mob,
+        final net.minecraft.server.level.ServerLevel serverLevel,
+        final int verticalSearchStart,
+        final int searchRange,
+        final int verticalSearchRange,
+        final BlockPos blockPos,
+        final float restrictRadius,
+        final BlockPos restrictCenter
+    ) {
+        BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
+        for (int i2 = verticalSearchStart; i2 <= verticalSearchRange; i2 = i2 > 0 ? -i2 : 1 - i2) {
+            for (int i3 = 0; i3 < searchRange; i3++) {
+                for (int i4 = 0; i4 <= i3; i4 = i4 > 0 ? -i4 : 1 - i4) {
+                    for (int i5 = i4 < i3 && i4 > -i3 ? i3 : 0; i5 <= i3; i5 = i5 > 0 ? -i5 : 1 - i5) {
+                        mutableBlockPos.setWithOffset(blockPos, i4, i2 - 1, i5);
+                        if (!serverLevel.hasChunkAt(mutableBlockPos)) continue;
+                        if (isWithinRestriction(restrictRadius, restrictCenter, mutableBlockPos)
+                            && isValidTargetAsync(ty, toRemove, mob, serverLevel, mutableBlockPos)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    private static boolean isWithinRestriction(float restrictRadius, BlockPos restrictCenter, BlockPos pos) {
+        return restrictRadius == -1.0F || restrictCenter.distSqr(pos) < restrictRadius * restrictRadius;
+    }
+    // Leaf end - Async search block
+
     protected abstract boolean isValidTarget(LevelReader level, BlockPos pos);
+
+    // Leaf start - Async search block
+    protected TypeToCheck typeToCheck() {
+        return TypeToCheck.Unknown;
+    }
+
+    private static boolean isValidTargetAsync(
+        TypeToCheck type,
+        @org.jetbrains.annotations.Nullable net.minecraft.world.level.block.Block blockToRemoveTy,
+        PathfinderMob mob,
+        LevelReader level,
+        BlockPos pos
+    ) {
+        switch (type) {
+            case CatLie -> {
+                return level.isEmptyBlock(pos.above()) && level.getBlockState(pos).is(net.minecraft.tags.BlockTags.BEDS);
+            }
+            case CatSit -> {
+                if (!level.isEmptyBlock(pos.above())) {
+                    return false;
+                } else {
+                    var blockState = level.getBlockState(pos);
+                    return blockState.is(net.minecraft.world.level.block.Blocks.CHEST)
+                        ? net.minecraft.world.level.block.entity.ChestBlockEntity.getOpenCount(level, pos) < 1
+                        : blockState.is(net.minecraft.world.level.block.Blocks.FURNACE) && blockState.getValue(net.minecraft.world.level.block.FurnaceBlock.LIT)
+                        || blockState.is(net.minecraft.tags.BlockTags.BEDS, state -> state.getOptionalValue(net.minecraft.world.level.block.BedBlock.PART).map(bedPart -> bedPart != net.minecraft.world.level.block.state.properties.BedPart.HEAD).orElse(true));
+                }
+            }
+            case Drowned -> {
+                BlockPos blockPos = pos.above();
+                return level.isEmptyBlock(blockPos) && level.isEmptyBlock(blockPos.above()) && level.getBlockState(pos).entityCanStandOn(level, pos, mob);
+            }
+            case FoxEat -> {
+                var blockState = level.getBlockState(pos);
+                return blockState.is(net.minecraft.world.level.block.Blocks.SWEET_BERRY_BUSH) && blockState.getValue(net.minecraft.world.level.block.SweetBerryBushBlock.AGE) >= 2 || net.minecraft.world.level.block.CaveVines.hasGlowBerries(blockState);
+            }
+            case RaidGarden -> {
+                var blockState = level.getBlockState(pos);
+                if (blockState.is(net.minecraft.world.level.block.Blocks.FARMLAND)) {
+                    blockState = level.getBlockState(pos.above());
+                    return blockState.getBlock() instanceof net.minecraft.world.level.block.CarrotBlock carrot && carrot.isMaxAge(blockState);
+                } else {
+                    return false;
+                }
+
+            }
+            case RemoveBlock -> {
+                var chunk = level.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4); // Paper - Prevent AI rules from loading chunks
+                return chunk != null
+                    && chunk.getBlockState(pos).is(blockToRemoveTy)
+                    && chunk.getBlockState(pos.above()).isAir()
+                    && chunk.getBlockState(pos.above(2)).isAir();
+            }
+            case Strider -> {
+                return level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.LAVA) && level.getBlockState(pos.above()).isPathfindable(net.minecraft.world.level.pathfinder.PathComputationType.LAND);
+            }
+            case TurtleToWater -> {
+                return level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.WATER);
+            }
+            case TurtleLay -> {
+                return level.isEmptyBlock(pos.above()) && net.minecraft.world.level.block.TurtleEggBlock.isSand(level, pos);
+            }
+            case Unknown -> throw new IllegalStateException();
+            case null -> throw new IllegalStateException();
+        }
+        // Leaf end - Async search block
+    }
 }
diff --git a/net/minecraft/world/entity/ai/goal/RandomLookAroundGoal.java b/net/minecraft/world/entity/ai/goal/RandomLookAroundGoal.java
index 71723bb4484084b4fd422e49230b6ee9b97e4c7e..1f07fce708aaf7db8f5e8c3f0553f23e40a7d620 100644
--- a/net/minecraft/world/entity/ai/goal/RandomLookAroundGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RandomLookAroundGoal.java
@@ -3,7 +3,7 @@ package net.minecraft.world.entity.ai.goal;
 import java.util.EnumSet;
 import net.minecraft.world.entity.Mob;
 
-public class RandomLookAroundGoal extends Goal {
+public class RandomLookAroundGoal extends Goal implements org.dreeam.leaf.async.ai.AsyncGoal { // Leaf
     private final Mob mob;
     private double relX;
     private double relZ;
diff --git a/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java b/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
index 28f770d40f5ada6d547e91cd3147969702e1ae75..f7599c94e76db96d634a54188383dcf6edbc91f4 100644
--- a/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
@@ -6,7 +6,7 @@ import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.phys.Vec3;
 
-public class RandomStrollGoal extends Goal {
+public class RandomStrollGoal extends Goal implements org.dreeam.leaf.async.ai.AsyncGoal { // Leaf
     public static final int DEFAULT_INTERVAL = 120;
     protected final PathfinderMob mob;
     protected double wantedX;
@@ -49,7 +49,7 @@ public class RandomStrollGoal extends Goal {
             }
 
             Vec3 position = this.getPosition();
-            if (position == null || (!this.mob.level().galeConfig().gameplayMechanics.entitiesCanRandomStrollIntoNonTickingChunks && !((net.minecraft.server.level.ServerLevel) this.mob.level()).isPositionEntityTicking(net.minecraft.core.BlockPos.containing(position)))) { // Gale - MultiPaper - prevent entities random strolling into non-ticking chunks
+            if (position == null) {
                 return false;
             } else {
                 this.wantedX = position.x;
diff --git a/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java b/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
index c67a88c9c77ece7c85ffb169ac96da4f28291228..064341c1baad4244aaf759bffa989df5dcbf5895 100644
--- a/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
@@ -151,8 +151,15 @@ public class RemoveBlockGoal extends MoveToBlockGoal {
     protected boolean isValidTarget(LevelReader level, BlockPos pos) {
         ChunkAccess chunk = level.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4); // Paper - Prevent AI rules from loading chunks
         return chunk != null
-            && chunk.getBlockState(pos).is(this.blockToRemove)
+            && chunk.getBlockState(pos).is(this.blockToRemove) // Leaf - Async search block - diff on change
             && chunk.getBlockState(pos.above()).isAir()
             && chunk.getBlockState(pos.above(2)).isAir();
     }
+
+    // Leaf start - Async search block
+    @Override
+    protected TypeToCheck typeToCheck() {
+        return TypeToCheck.RemoveBlock;
+    }
+    // Leaf end - Async search block
 }
diff --git a/net/minecraft/world/entity/ai/goal/TemptGoal.java b/net/minecraft/world/entity/ai/goal/TemptGoal.java
index f88f618d34fb343b31de3af1a875d6633703df71..ee8a7c953c798d5b0caf828e462cf12d7badd073 100644
--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -44,15 +44,6 @@ public class TemptGoal extends Goal {
         } else {
             this.player = getServerLevel(this.mob)
                 .getNearestPlayer(this.targetingConditions.range(this.mob.getAttributeValue(Attributes.TEMPT_RANGE)), this.mob);
-            // CraftBukkit start
-            if (this.player != null) {
-                org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(this.mob, this.player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TEMPT);
-                if (event.isCancelled()) {
-                    return false;
-                }
-                this.player = (event.getTarget() == null) ? null : ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle();
-            }
-            // CraftBukkit end
             return this.player != null;
         }
     }
diff --git a/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java b/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
index 4644f3f7af89623ca6218c0dd24bb6cd67db553b..b6c2b0949081e2c975c33e55be462a04d251c2a1 100644
--- a/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
@@ -16,7 +16,7 @@ public class DefendVillageTargetGoal extends TargetGoal {
     private final IronGolem golem;
     @Nullable
     private LivingEntity potentialTarget;
-    private final TargetingConditions attackTargeting = TargetingConditions.forCombat().range(64.0);
+    private static final TargetingConditions attackTargeting = TargetingConditions.forCombat().range(64.0); // Leaf - Async Target Finding - static
 
     public DefendVillageTargetGoal(IronGolem golem) {
         super(golem, false, true);
diff --git a/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java b/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
index 25fe78116ce01eeefe5c958423734195d27302eb..9e9f9d77dcedeff1e00cfd314a78c7aaac64e75f 100644
--- a/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
@@ -73,6 +73,69 @@ public class HurtByTargetGoal extends TargetGoal {
     protected void alertOthers() {
         double followDistance = this.getFollowDistance();
         AABB aabb = AABB.unitCubeFromLowerCorner(this.mob.position()).inflate(followDistance, 10.0, followDistance);
+
+        // Leaf start - Async alert other
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.goal) {
+            final var self = this.mob;
+            final var serverLevel = getServerLevel(self);
+            final var toIgnoreAlert = this.toIgnoreAlert;
+            var toAlert = new java.util.ArrayList<Mob>();
+            List<? extends Mob> entitiesOfClass = serverLevel
+                .getEntitiesOfClass(self.getClass(), aabb, EntitySelector.NO_SPECTATORS);
+            for (Mob mob : entitiesOfClass) {
+                if (self == mob
+                    || mob.getTarget() != null
+                    || (self instanceof TamableAnimal && ((TamableAnimal) self).getOwner() != ((TamableAnimal) mob).getOwner())
+                    || mob.isAlliedTo(self.getLastHurtByMob())) {
+                    continue;
+                }
+                if (toIgnoreAlert == null) {
+                    toAlert.add(mob);
+                    continue;
+                }
+                boolean flag = false;
+                for (Class<?> clazz : toIgnoreAlert) {
+                    if (mob.getClass() == clazz) {
+                        flag = true;
+                        break;
+                    }
+                }
+                if (!flag) {
+                    toAlert.add(mob);
+                }
+            }
+            if (!toAlert.isEmpty()) {
+                serverLevel.asyncGoalExecutor.post(() -> {
+                    LivingEntity lastHurtByMob = this.mob.getLastHurtByMob();
+                    if (lastHurtByMob.getType() == EntityType.PLAYER && getServerLevel(this.mob).getGameRules().getBoolean(GameRules.RULE_UNIVERSAL_ANGER)) {
+                        return;
+                    }
+                    for (Class<?> clazz : this.toIgnoreDamage) {
+                        if (clazz.isAssignableFrom(lastHurtByMob.getClass())) {
+                            return;
+                        }
+                    }
+                    if (this.mob.getTarget() == null) {
+                        return;
+                    }
+                    if (!canContinueToUse()) {
+                        return;
+                    }
+                    if (!this.canAttack(lastHurtByMob, HURT_BY_TARGETING)) {
+                        return;
+                    }
+                    for (var obj : toAlert) {
+                        Mob mob = (Mob) obj;
+                        if (mob.getTarget() == null) {
+                            alertOther(mob, lastHurtByMob);
+                        }
+                    }
+                });
+            }
+            return;
+        }
+        // Leaf end - Async alert other
+
         List<? extends Mob> entitiesOfClass = this.mob
             .level()
             .getEntitiesOfClass((Class<? extends Mob>)this.mob.getClass(), aabb, EntitySelector.NO_SPECTATORS);
@@ -87,7 +150,7 @@ public class HurtByTargetGoal extends TargetGoal {
 
                 mob = (Mob)var5.next();
                 if (this.mob != mob
-                    && mob.getTarget() == null
+                    && mob.getTarget() == null // Leaf - Async alert other - diff on change
                     && (!(this.mob instanceof TamableAnimal) || ((TamableAnimal)this.mob).getOwner() == ((TamableAnimal)mob).getOwner())
                     && !mob.isAlliedTo(this.mob.getLastHurtByMob())) {
                     if (this.toIgnoreAlert == null) {
@@ -96,7 +159,7 @@ public class HurtByTargetGoal extends TargetGoal {
 
                     boolean flag = false;
 
-                    for (Class<?> clazz : this.toIgnoreAlert) {
+                    for (Class<?> clazz : this.toIgnoreAlert) { // Leaf - Async alert other - diff on change
                         if (mob.getClass() == clazz) {
                             flag = true;
                             break;
diff --git a/net/minecraft/world/entity/ai/goal/target/NonTameRandomTargetGoal.java b/net/minecraft/world/entity/ai/goal/target/NonTameRandomTargetGoal.java
index abf57494950f55bbd75f335f26736cb9e703c197..efd2418f56c36e7850edde483a2a4906dd622441 100644
--- a/net/minecraft/world/entity/ai/goal/target/NonTameRandomTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/NonTameRandomTargetGoal.java
@@ -20,6 +20,15 @@ public class NonTameRandomTargetGoal<T extends LivingEntity> extends NearestAtta
 
     @Override
     public boolean canContinueToUse() {
-        return this.targetConditions != null ? this.targetConditions.test(getServerLevel(this.mob), this.mob, this.target) : super.canContinueToUse();
+        // Leaf start - Async target finding
+        if (this.target == null || !this.target.isAlive() || this.target.isRemoved()) {
+            return false;
+        }
+        var serverLevel = getServerLevel(this.mob);
+        if (serverLevel == null) {
+            return false;
+        }
+        return this.getTargetConditions().test(serverLevel, this.mob, this.target) && super.canContinueToUse();
+        // Leaf end - Async target finding
     }
 }
diff --git a/net/minecraft/world/entity/ai/goal/target/ResetUniversalAngerTargetGoal.java b/net/minecraft/world/entity/ai/goal/target/ResetUniversalAngerTargetGoal.java
index 61a7bb5f1760d47a13b6aafc123bcf3dff420860..238bb3544bfdbd2398e838121cb201921f4c909f 100644
--- a/net/minecraft/world/entity/ai/goal/target/ResetUniversalAngerTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/ResetUniversalAngerTargetGoal.java
@@ -37,6 +37,31 @@ public class ResetUniversalAngerTargetGoal<T extends Mob & NeutralMob> extends G
         this.lastHurtByPlayerTimestamp = this.mob.getLastHurtByMobTimestamp();
         this.mob.forgetCurrentTargetAndRefreshUniversalAnger();
         if (this.alertOthersOfSameType) {
+            // Leaf start - Async alert other
+            if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.goal) {
+                final var mob = this.mob;
+                final var serverLevel = getServerLevel(mob);
+                final double followRange = this.mob.getAttributeValue(Attributes.FOLLOW_RANGE);
+                final AABB aabb = AABB.unitCubeFromLowerCorner(this.mob.position()).inflate(followRange, 10.0, followRange);
+                var entities = serverLevel.getEntitiesOfClass(mob.getClass(), aabb, EntitySelector.NO_SPECTATORS);
+                List<NeutralMob> toStop = new java.util.ArrayList<>(entities.size());
+                for (Mob entity : entities) {
+                    if (entity != mob) {
+                        toStop.add((NeutralMob) entity);
+                    }
+                }
+                if (!toStop.isEmpty()) {
+                    serverLevel.asyncGoalExecutor.post(() -> {
+                        for (var neutralMob : toStop) {
+                            if (EntitySelector.NO_SPECTATORS.test((net.minecraft.world.entity.Entity) neutralMob)) {
+                                neutralMob.forgetCurrentTargetAndRefreshUniversalAnger();
+                            }
+                        }
+                    });
+                }
+                return;
+            }
+            // Leaf end - Async alert other
             this.getNearbyMobsOfSameType()
                 .stream()
                 .filter(mob -> mob != this.mob)
@@ -48,6 +73,7 @@ public class ResetUniversalAngerTargetGoal<T extends Mob & NeutralMob> extends G
     }
 
     private List<? extends Mob> getNearbyMobsOfSameType() {
+        // Leaf - Async alert other - diff on change
         double attributeValue = this.mob.getAttributeValue(Attributes.FOLLOW_RANGE);
         AABB aabb = AABB.unitCubeFromLowerCorner(this.mob.position()).inflate(attributeValue, 10.0, attributeValue);
         return this.mob.level().getEntitiesOfClass((Class<? extends Mob>)this.mob.getClass(), aabb, EntitySelector.NO_SPECTATORS);
diff --git a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
index ae585f10827f52db80f048b84d6096737d12fee2..6dd4488e3da284ef0293b190871d994e6f883d28 100644
--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -10,7 +10,7 @@ import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
 import net.minecraft.world.scores.Team;
 
-public abstract class TargetGoal extends Goal {
+public abstract class TargetGoal extends Goal implements org.dreeam.leaf.async.ai.AsyncGoal { // Leaf
     private static final int EMPTY_REACH_CACHE = 0;
     private static final int CAN_REACH_CACHE = 1;
     private static final int CANT_REACH_CACHE = 2;
diff --git a/net/minecraft/world/entity/ai/sensing/Sensing.java b/net/minecraft/world/entity/ai/sensing/Sensing.java
index 002d3c0d8b1107a275020d5c582c37e9a5c536ee..6fa0b8defbd1d06b3bf5d9b32ffd08f32c1fb707 100644
--- a/net/minecraft/world/entity/ai/sensing/Sensing.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensing.java
@@ -33,6 +33,17 @@ public class Sensing {
     }
 
     public void tick() {
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            synchronized (this) {
+                tickInternal();
+            }
+        } else {
+            tickInternal();
+        }
+    }
+    private void tickInternal() {
+    // Leaf end - Async target finding
         if (this.expiring == null) { // Gale - Petal - reduce line of sight updates
         this.seen.clear();
             // Gale start - Petal - reduce line of sight updates
@@ -63,6 +74,17 @@ public class Sensing {
     }
 
     public boolean hasLineOfSight(Entity entity) {
+        // Leaf start - Async target finding
+        if (org.dreeam.leaf.config.modules.async.AsyncTargetFinding.enabled) {
+            synchronized (this) {
+                return hasLineOfSightInternal(entity);
+            }
+        } else {
+            return hasLineOfSightInternal(entity);
+        }
+    }
+    private boolean hasLineOfSightInternal(Entity entity) {
+    // Leaf end - Async target finding
         int id = entity.getId();
         // Gale start - Petal - reduce line of sight cache lookups - merge sets
         int cached = this.seen.get(id);
diff --git a/net/minecraft/world/entity/animal/Dolphin.java b/net/minecraft/world/entity/animal/Dolphin.java
index 3ed9652510976770f5661dd7b317f27f046700d4..93017eec4b2e0984cf60d810b93df546e63511bf 100644
--- a/net/minecraft/world/entity/animal/Dolphin.java
+++ b/net/minecraft/world/entity/animal/Dolphin.java
@@ -715,7 +715,15 @@ public class Dolphin extends AgeableWaterCreature {
                 this.drop(itemBySlot);
                 Dolphin.this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
             } else if (!entitiesOfClass.isEmpty()) {
-                Dolphin.this.getNavigation().moveTo(entitiesOfClass.get(0), 1.2F);
+                // Leaf start
+                var entity = entitiesOfClass.get(0);
+                var exec = ((ServerLevel) Dolphin.this.level()).asyncGoalExecutor;
+                if (exec != null) {
+                    exec.post(() -> Dolphin.this.getNavigation().moveTo(entity, 1.2F));
+                } else {
+                    Dolphin.this.getNavigation().moveTo(entity, 1.2F);
+                }
+                // Leaf end
             }
         }
 
diff --git a/net/minecraft/world/entity/animal/Fox.java b/net/minecraft/world/entity/animal/Fox.java
index 90452f0945e761077608692877677f522d38bccd..54044ff16e0c2963b2220e4fb4932fe0802aa178 100644
--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -849,13 +849,18 @@ public class Fox extends Animal implements VariantHolder<Fox.Variant> {
                 return false;
             } else {
                 ServerLevel serverLevel = getServerLevel(Fox.this.level());
+                // Leaf start - Async target finding
+                if (serverLevel == null) {
+                    return false;
+                }
+                // Leaf end - Async target finding
 
                 for (UUID uuid : Fox.this.getTrustedUUIDs()) {
                     if (serverLevel.getEntity(uuid) instanceof LivingEntity livingEntity) {
                         this.trustedLastHurt = livingEntity;
                         this.trustedLastHurtBy = livingEntity.getLastHurtByMob();
                         int lastHurtByMobTimestamp = livingEntity.getLastHurtByMobTimestamp();
-                        return lastHurtByMobTimestamp != this.timestamp && this.canAttack(this.trustedLastHurtBy, this.targetConditions);
+                        return lastHurtByMobTimestamp != this.timestamp && this.canAttack(this.trustedLastHurtBy, this.getTargetConditions()); // Leaf - Async target finding
                     }
                 }
 
@@ -1033,7 +1038,7 @@ public class Fox extends Animal implements VariantHolder<Fox.Variant> {
         @Override
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
             BlockState blockState = level.getBlockState(pos);
-            return blockState.is(Blocks.SWEET_BERRY_BUSH) && blockState.getValue(SweetBerryBushBlock.AGE) >= 2 || CaveVines.hasGlowBerries(blockState);
+            return blockState.is(Blocks.SWEET_BERRY_BUSH) && blockState.getValue(SweetBerryBushBlock.AGE) >= 2 || CaveVines.hasGlowBerries(blockState); // Leaf - Async search block - diff on change
         }
 
         @Override
@@ -1097,6 +1102,13 @@ public class Fox extends Animal implements VariantHolder<Fox.Variant> {
             Fox.this.setSitting(false);
             super.start();
         }
+
+        // Leaf start - Async search block
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.FoxEat;
+        }
+        // Leaf end - Async search block
     }
 
     class FoxFloatGoal extends FloatGoal {
diff --git a/net/minecraft/world/entity/animal/Rabbit.java b/net/minecraft/world/entity/animal/Rabbit.java
index da5b32a17283e540615373097acc511d928aeff5..cdab995c86978f0a789c85d14151abf19d89b673 100644
--- a/net/minecraft/world/entity/animal/Rabbit.java
+++ b/net/minecraft/world/entity/animal/Rabbit.java
@@ -642,7 +642,13 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
                 this.wantsToRaid = this.rabbit.wantsMoreFood();
             }
 
-            return super.canUse();
+            // Leaf start - Async search block
+            if (this.wantsToRaid && !this.canRaid) {
+                return super.canUse();
+            } else {
+                return false;
+            }
+            // Leaf end - Async search block
         }
 
         @Override
@@ -685,7 +691,7 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
         @Override
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
             BlockState blockState = level.getBlockState(pos);
-            if (blockState.is(Blocks.FARMLAND) && this.wantsToRaid && !this.canRaid) {
+            if (blockState.is(Blocks.FARMLAND) && this.wantsToRaid && !this.canRaid) { // Leaf - Async search block - diff on change
                 blockState = level.getBlockState(pos.above());
                 if (blockState.getBlock() instanceof CarrotBlock && ((CarrotBlock)blockState.getBlock()).isMaxAge(blockState)) {
                     this.canRaid = true;
@@ -695,6 +701,13 @@ public class Rabbit extends Animal implements VariantHolder<Rabbit.Variant> {
 
             return false;
         }
+
+        // Leaf start - Async search block
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.RaidGarden;
+        }
+        // Leaf end - Async search block
     }
 
     public static enum Variant implements StringRepresentable {
diff --git a/net/minecraft/world/entity/animal/Turtle.java b/net/minecraft/world/entity/animal/Turtle.java
index 10477fea8fcd70bf0c1ba4b6e1113625be690e68..c5f626dcfe724f8ed19303d305f6eb5cfcc0f0af 100644
--- a/net/minecraft/world/entity/animal/Turtle.java
+++ b/net/minecraft/world/entity/animal/Turtle.java
@@ -527,8 +527,15 @@ public class Turtle extends Animal {
 
         @Override
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
-            return level.getBlockState(pos).is(Blocks.WATER);
+            return level.getBlockState(pos).is(Blocks.WATER); // Leaf - Async search block - diff on change
         }
+
+        // Leaf start - Async search block
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.TurtleToWater;
+        }
+        // Leaf end - Async search block
     }
 
     static class TurtleLayEggGoal extends MoveToBlockGoal {
@@ -584,8 +591,15 @@ public class Turtle extends Animal {
 
         @Override
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
-            return level.isEmptyBlock(pos.above()) && TurtleEggBlock.isSand(level, pos);
+            return level.isEmptyBlock(pos.above()) && TurtleEggBlock.isSand(level, pos); // Leaf - Async search block - diff on change
+        }
+
+        // Leaf start - Async search block
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.TurtleLay;
         }
+        // Leaf end - Async search block
     }
 
     static class TurtleMoveControl extends org.purpurmc.purpur.controller.MoveControllerWASD { // Purpur - Ridables
diff --git a/net/minecraft/world/entity/animal/Wolf.java b/net/minecraft/world/entity/animal/Wolf.java
index 7cb292de6b27fa4ba3c5fce526a4e939c576789f..68852158e2fd3c4ce241d722ac414d73ea3b8836 100644
--- a/net/minecraft/world/entity/animal/Wolf.java
+++ b/net/minecraft/world/entity/animal/Wolf.java
@@ -672,7 +672,7 @@ public class Wolf extends TamableAnimal implements NeutralMob, VariantHolder<Hol
 
     @Override
     public boolean isFood(ItemStack stack) {
-        return stack.is(ItemTags.WOLF_FOOD);
+        return stack.is(ItemTags.WOLF_FOOD); // Leaf - Async Target Finding - diff on change
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/Drowned.java b/net/minecraft/world/entity/monster/Drowned.java
index d44ed0d6a672a0b1eb0a8781e3e094096a2b753d..490c196419c94b75a540e64d513163df236eed0d 100644
--- a/net/minecraft/world/entity/monster/Drowned.java
+++ b/net/minecraft/world/entity/monster/Drowned.java
@@ -394,7 +394,7 @@ public class Drowned extends Zombie implements RangedAttackMob {
         @Override
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
             BlockPos blockPos = pos.above();
-            return level.isEmptyBlock(blockPos) && level.isEmptyBlock(blockPos.above()) && level.getBlockState(pos).entityCanStandOn(level, pos, this.drowned);
+            return level.isEmptyBlock(blockPos) && level.isEmptyBlock(blockPos.above()) && level.getBlockState(pos).entityCanStandOn(level, pos, this.drowned); // Leaf - Async search block - diff on change
         }
 
         @Override
@@ -408,6 +408,13 @@ public class Drowned extends Zombie implements RangedAttackMob {
         public void stop() {
             super.stop();
         }
+
+        // Leaf start - Async search block
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.Drowned;
+        }
+        // Leaf end - Async search block
     }
 
     static class DrownedGoToWaterGoal extends Goal {
diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index c7897532163d4fdf5a82982f7d24a47dd61e3dfa..3dde64d0c61c47d2911d2cd4b64b201c1bfaeba3 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -638,6 +638,13 @@ public class EnderMan extends Monster implements NeutralMob {
             }
         }
 
+        // Leaf start
+        @Override
+        public boolean tickAsync() {
+            return false;
+        }
+        // Leaf end
+
         @Override
         public void tick() {
             if (this.enderman.getTarget() == null) {
diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index 87dc32458412cded3109957eb28a38cc44024c55..52d870abd56360460444ff22c333d8ef7cb1d71f 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -570,8 +570,15 @@ public class Strider extends Animal implements ItemSteerable, Saddleable {
 
         @Override
         protected boolean isValidTarget(LevelReader level, BlockPos pos) {
-            return level.getBlockState(pos).is(Blocks.LAVA) && level.getBlockState(pos.above()).isPathfindable(PathComputationType.LAND);
+            return level.getBlockState(pos).is(Blocks.LAVA) && level.getBlockState(pos.above()).isPathfindable(PathComputationType.LAND); // Leaf - Async search block - diff on change
         }
+
+        // Leaf start - Async search block
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.Strider;
+        }
+        // Leaf end - Async search block
     }
 
     static class StriderPathNavigation extends GroundPathNavigation {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index ac7729d1caa80155697bfe0e8646e4eda5d1780e..2518ea7a9d2122ff55ab6546d21d7ed6cc933090 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1549,6 +1549,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
                 }
                 // Leaf end - SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks) - do not bother with condition work / make configurable
                 // Paper end - rewrite chunk system
+                ((ServerLevel) this).leafMidTickTasks(); // Leaf - Async target finding
             }
         }
         this.blockEntityTickers.removeMarkedEntries(); // SparklyPaper - optimize block entity removals
@@ -1817,9 +1818,11 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
     @Override
     public List<Entity> getEntities(@Nullable Entity entity, AABB boundingBox, Predicate<? super Entity> predicate) {
-        if (org.dreeam.leaf.config.modules.async.SparklyPaperParallelWorldTicking.enabled) // Leaf start - SparklyPaper - parallel world ticking mod (make configurable)
-            ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)this, boundingBox, "Cannot getEntities asynchronously"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
-        List<Entity> list = Lists.newArrayList();
+        // Leaf start - Async target finding
+        // if (org.dreeam.leaf.config.modules.async.SparklyPaperParallelWorldTicking.enabled) // Leaf start - SparklyPaper - parallel world ticking mod (make configurable)
+        //     ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)this, boundingBox, "Cannot getEntities asynchronously"); // SparklyPaper - parallel world ticking (additional concurrency issues logs)
+        // List<Entity> list = Lists.newArrayList(); // Leaf - Async target finding - unused
+        // Leaf end - Async target finding
 
         // Paper start - rewrite chunk system
         final List<Entity> ret = new java.util.ArrayList<>();
diff --git a/net/minecraft/world/level/pathfinder/PathfindingContext.java b/net/minecraft/world/level/pathfinder/PathfindingContext.java
index 7565fb33fbf5d51fac1fa61fe356ed0dad0ce839..0c77352721f6a4ec94111c7a2d1ce23fe1338a7c 100644
--- a/net/minecraft/world/level/pathfinder/PathfindingContext.java
+++ b/net/minecraft/world/level/pathfinder/PathfindingContext.java
@@ -16,7 +16,7 @@ public class PathfindingContext {
 
     public PathfindingContext(CollisionGetter level, Mob mob) {
         this.level = level;
-        if (mob.level() instanceof ServerLevel serverLevel && !org.dreeam.leaf.config.modules.async.AsyncTargetFinding.pathfinding) { // Leaf - Async pathfinding
+        if (mob.level() instanceof ServerLevel serverLevel && !org.dreeam.leaf.config.modules.async.AsyncTargetFinding.pathfinding && !org.dreeam.leaf.config.modules.async.AsyncTargetFinding.goal) { // Leaf - Async pathfinding // Leaf - Async target finding
             this.cache = serverLevel.getPathTypeCache();
         } else {
             this.cache = null;
