From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 4 Aug 2024 19:34:29 +0800
Subject: [PATCH] Configurable player knockback zombie


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 5286042fae6750275bf912e0928d65c11757ea83..cd275bb184ba817973d1b49e2e5f39811bbe48e4 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -2180,6 +2180,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public void knockback(double strength, double x, double z, @Nullable Entity attacker, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause eventCause) { // Paper - knockback events
+        if (!canKnockback(attacker, this)) return; // Leaf - Configurable player knockback zombie
         strength *= 1.0 - this.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE);
         if (true || !(strength <= 0.0)) { // CraftBukkit - Call event even when force is 0
             // this.needsSync = true; // CraftBukkit - Move down
@@ -2210,6 +2211,20 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         }
     }
 
+    // Leaf start - Configurable player knockback zombie
+    private boolean canKnockback(@Nullable Entity attacker, LivingEntity target) {
+        if (!org.dreeam.leaf.config.modules.gameplay.Knockback.canPlayerKnockbackZombie) {
+            if (attacker instanceof ServerPlayer && target.getType() == EntityType.ZOMBIE) { // Player -> Zombie
+                return false;
+            } else if (attacker instanceof Projectile projectile && projectile.getOwner() instanceof ServerPlayer && target.getType() == EntityType.ZOMBIE) { // Player -> projectile -> Zombie
+                return false;
+            }
+        }
+
+        return true;
+    }
+    // Leaf end - Configurable player knockback zombie
+
     public void indicateDamage(double xDistance, double zDistance) {
     }
 
