From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Mon, 14 Apr 2025 18:07:21 +0200
Subject: [PATCH] Optimise MobEffectUtil#getDigSpeedAmplification


diff --git a/net/minecraft/world/effect/MobEffectUtil.java b/net/minecraft/world/effect/MobEffectUtil.java
index cbf1b6af928aa439c3264b302e5f1a1ddd4c14f0..e4126bbfb46f2d3df9aac4c9362b9d81b3ecb713 100644
--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -27,17 +27,21 @@ public final class MobEffectUtil {
     }
 
     public static int getDigSpeedAmplification(LivingEntity entity) {
-        int i = 0;
-        int i1 = 0;
-        if (entity.hasEffect(MobEffects.DIG_SPEED)) {
-            i = entity.getEffect(MobEffects.DIG_SPEED).getAmplifier();
+        // Leaf start - Optimise MobEffectUtil#getDigSpeedAmplification
+        int digAmplifier = 0;
+        int conduitAmplifier = 0;
+        MobEffectInstance digEffect = entity.getEffect(MobEffects.DIG_SPEED);
+        if (digEffect != null) {
+            digAmplifier = digEffect.getAmplifier();
         }
 
-        if (entity.hasEffect(MobEffects.CONDUIT_POWER)) {
-            i1 = entity.getEffect(MobEffects.CONDUIT_POWER).getAmplifier();
+        MobEffectInstance conduitEffect = entity.getEffect(MobEffects.CONDUIT_POWER);
+        if (conduitEffect != null) {
+            conduitAmplifier = conduitEffect.getAmplifier();
         }
 
-        return Math.max(i, i1);
+        return Math.max(digAmplifier, conduitAmplifier);
+        // Leaf end - Optimise MobEffectUtil#getDigSpeedAmplification
     }
 
     public static boolean hasWaterBreathing(LivingEntity entity) {
