From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 4 Aug 2024 19:34:29 +0800
Subject: [PATCH] Configurable player knockback zombie


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index dce2ff4a740ea5a652b89f066934f480b14f25a3..a18f341cebe730adb4b121fb5eb56a519ddb225a 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -2166,6 +2166,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public void knockback(double power, double xd, double zd, final DamageSource source, final float damage, final boolean comesFromEffect, @Nullable Entity attacker, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause eventCause) { // Paper - knockback events
+        if (!canKnockback(attacker, this)) return; // Leaf - Configurable player knockback zombie
         power *= 1.0 - this.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE);
         if (true || !(power <= 0.0)) { // CraftBukkit - Call event even when force is 0
             // this.needsSync = true; // CraftBukkit - Move down
@@ -2206,6 +2207,20 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         // Paper end - knockback events
     }
 
+    // Leaf start - Configurable player knockback zombie
+    private boolean canKnockback(final @Nullable Entity attacker, final LivingEntity target) {
+        if (!org.dreeam.leaf.config.modules.gameplay.Knockback.canPlayerKnockbackZombie) {
+            if (attacker instanceof ServerPlayer && target.getType() == EntityTypes.ZOMBIE) { // Player -> Zombie
+                return false;
+            } else if (attacker instanceof Projectile projectile && projectile.getOwner() instanceof ServerPlayer && target.getType() == EntityTypes.ZOMBIE) { // Player -> projectile -> Zombie
+                return false;
+            }
+        }
+
+        return true;
+    }
+    // Leaf end - Configurable player knockback zombie
+
     public void indicateDamage(final double xd, final double zd) {
     }
 
