From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix MC-270656

Related MC issue: https://mojira.dev/MC-270656

diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 84ea591ecd4f42383f69525f32b748d15497cfed..6fa8c5bb6689b56b95c5f0a1c7d2691f12ac673c 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1079,6 +1079,13 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     public void trackStartFallingPosition() {
+        // Leaf start - Fix MC-270656
+        if (org.dreeam.leaf.config.modules.fixes.MCBugFix.mc270656) {
+            if (this.currentImpulseImpactPos != null && (this.getAbilities().flying || this.isFallFlying())) {
+                this.currentImpulseImpactPos = null;
+            }
+        }
+        // Leaf end - Fix MC-270656
         if (this.fallDistance > 0.0 && this.startingToFallPosition == null) {
             this.startingToFallPosition = this.position();
             if (this.currentImpulseImpactPos != null && this.currentImpulseImpactPos.y <= this.startingToFallPosition.y) {
