From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <mc@jvavav.com>
Date: Sat, 19 Apr 2025 22:12:19 +0800
Subject: [PATCH] Optimize BlockEntityType#isValid


diff --git a/net/minecraft/world/level/block/entity/BlockEntityType.java b/net/minecraft/world/level/block/entity/BlockEntityType.java
index a2a674f18d7f2c2e50a6b25f9ac3bf4534275976..6be69dcd9faa76ef999daaecc237ee3b4d04c182 100644
--- a/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -252,6 +252,14 @@ public class BlockEntityType<T extends BlockEntity> {
     private BlockEntityType(BlockEntityType.BlockEntitySupplier<? extends T> factory, Set<Block> validBlocks) {
         this.factory = factory;
         this.validBlocks = validBlocks;
+        // Leaf start - Optimize BlockEntityType#isValid
+        for (Block validBlock : validBlocks) {
+            if (validBlock.blockEntityType != null) {
+                throw new IllegalStateException("Duplicate block entity type");
+            }
+            validBlock.blockEntityType = this;
+        }
+        // Leaf end - Optimize BlockEntityType#isValid
     }
 
     @Nullable
@@ -260,7 +268,7 @@ public class BlockEntityType<T extends BlockEntity> {
     }
 
     public boolean isValid(BlockState state) {
-        return this.validBlocks.contains(state.getBlock());
+        return state.getBlock().blockEntityType == this; // Leaf - Optimize BlockEntityType#isValid - remove hash lookup
     }
 
     @Deprecated
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index d35211b0cae66b1a40e89539507e55973313f46f..3a49690672bc49259e863e5cde7f14d57c89fcd5 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -100,6 +100,7 @@ public abstract class BlockBehaviour implements FeatureElement {
     protected final BlockBehaviour.Properties properties;
     protected final Optional<ResourceKey<LootTable>> drops;
     protected final String descriptionId;
+    @Nullable public net.minecraft.world.level.block.entity.BlockEntityType blockEntityType = null; // Leaf - Optimize BlockEntityType#isValid
 
     public BlockBehaviour(BlockBehaviour.Properties properties) {
         this.hasCollision = properties.hasCollision;
