From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 23 Mar 2025 00:07:42 +0100
Subject: [PATCH] Optimize getNeighborSignal


diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 41802482875bd4d4b505eb758740140de0db415a..f8d2a2b052378b705777651536dc4b3362b0581a 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -150,25 +150,27 @@ public class PistonBaseBlock extends DirectionalBlock {
     }
 
     private boolean getNeighborSignal(SignalGetter signalGetter, BlockPos pos, Direction direction) {
-        for (Direction direction1 : Direction.values()) {
-            if (direction1 != direction && signalGetter.hasSignal(pos.relative(direction1), direction1)) {
+        for (Direction dir : Direction.Plane.HORIZONTAL) {
+            if (dir != direction && signalGetter.hasSignal(pos.relative(dir), dir)) {
                 return true;
             }
         }
-
+        if (Direction.UP != direction && signalGetter.hasSignal(pos.relative(Direction.UP), Direction.UP)) {
+            return true;
+        }
+        if (Direction.DOWN != direction && signalGetter.hasSignal(pos.relative(Direction.DOWN), Direction.DOWN)) {
+            return true;
+        }
         if (signalGetter.hasSignal(pos, Direction.DOWN)) {
             return true;
-        } else {
-            BlockPos blockPos = pos.above();
-
-            for (Direction direction2 : Direction.values()) {
-                if (direction2 != Direction.DOWN && signalGetter.hasSignal(blockPos.relative(direction2), direction2)) {
-                    return true;
-                }
+        }
+        BlockPos blockPosAbove = pos.above();
+        for (Direction dir : Direction.Plane.HORIZONTAL) {
+            if (signalGetter.hasSignal(blockPosAbove.relative(dir), dir)) {
+                return true;
             }
-
-            return false;
         }
+        return signalGetter.hasSignal(blockPosAbove.relative(Direction.UP), Direction.UP);
     }
 
     @Override
