From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Mon, 14 Apr 2025 18:07:21 +0200
Subject: [PATCH] Optimise MobEffectUtil#getDigSpeedAmplification


diff --git a/net/minecraft/world/effect/MobEffectUtil.java b/net/minecraft/world/effect/MobEffectUtil.java
index cbf1b6af928aa439c3264b302e5f1a1ddd4c14f0..c59a503ef8bc2dabcf9f7c85c8d93fb1fcadf71f 100644
--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -29,12 +29,14 @@ public final class MobEffectUtil {
     public static int getDigSpeedAmplification(LivingEntity entity) {
         int i = 0;
         int i1 = 0;
-        if (entity.hasEffect(MobEffects.DIG_SPEED)) {
-            i = entity.getEffect(MobEffects.DIG_SPEED).getAmplifier();
+        MobEffectInstance digEffect = entity.getEffect(MobEffects.DIG_SPEED);
+        if (digEffect != null) {
+            i = digEffect.getAmplifier();
         }
 
-        if (entity.hasEffect(MobEffects.CONDUIT_POWER)) {
-            i1 = entity.getEffect(MobEffects.CONDUIT_POWER).getAmplifier();
+        MobEffectInstance conduitEffect = entity.getEffect(MobEffects.CONDUIT_POWER);
+        if (conduitEffect != null) {
+            i1 = conduitEffect.getAmplifier();
         }
 
         return Math.max(i, i1);
