From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <mc@jvavav.com>
Date: Sat, 19 Apr 2025 19:21:50 +0800
Subject: [PATCH] Allowing endermen to despawn while holding mushroom


diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index c7897532163d4fdf5a82982f7d24a47dd61e3dfa..111f7008f73ed4c7b1be38ad2cf3d86f9391ad69 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -485,7 +485,20 @@ public class EnderMan extends Monster implements NeutralMob {
 
     @Override
     public boolean requiresCustomPersistence() {
-        return super.requiresCustomPersistence() || (!this.level().purpurConfig.endermanDespawnEvenWithBlock && this.getCarriedBlock() != null); // Purpur - Add config for allowing Endermen to despawn even while holding a block
+        // Leaf start - allowing endermen to despawn while holding mushroom
+        if (super.requiresCustomPersistence()) {
+            return true;
+        }
+        // Purpur - Add config for allowing Endermen to despawn even while holding a block
+        if (this.level().purpurConfig.endermanDespawnEvenWithBlock) {
+            return false;
+        }
+        var carried = this.getCarriedBlock();
+        if (carried == null) {
+            return false;
+        }
+        return !(org.dreeam.leaf.config.modules.fixes.DespawnEndermanWhileHoldingMushroom.despawn && (carried.is(Blocks.RED_MUSHROOM) || carried.is(Blocks.BROWN_MUSHROOM)));
+        // Leaf end
     }
 
     static class EndermanFreezeWhenLookedAt extends Goal {
diff --git a/net/minecraft/world/level/block/MushroomBlock.java b/net/minecraft/world/level/block/MushroomBlock.java
index afd952ddc8942818ec01d1c750413776b6537cc2..454181ae4168d87c4f3bf29cc9b6619dae2876e1 100644
--- a/net/minecraft/world/level/block/MushroomBlock.java
+++ b/net/minecraft/world/level/block/MushroomBlock.java
@@ -47,6 +47,12 @@ public class MushroomBlock extends BushBlock implements BonemealableBlock {
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        // Leaf start - disable mushroom growth in the end
+        if (org.dreeam.leaf.config.modules.fixes.DespawnEndermanWhileHoldingMushroom.noGrowthInTheEnd && level.isTheEnd()) {
+            return;
+        }
+        // Leaf end
+
         if (random.nextFloat() < (level.spigotConfig.mushroomModifier / (100.0f * 25))) { // Spigot - SPIGOT-7159: Better modifier resolution
             int i = 5;
             int i1 = 4;
