From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 14 Aug 2025 14:12:38 +0900
Subject: [PATCH] optimize attribute


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 5bb682f88019de563a3b000c863d9235cdca6d76..5e7f0429f8e6441b21994e92c4ecc67a4b71abd7 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -433,7 +433,7 @@ public class ServerEntity {
                 this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundUpdateAttributesPacket(this.entity.getId(), attributesToSync));
             }
 
-            attributesToSync.clear();
+            //attributesToSync.clear(); // Leaf - optimize attribute
         }
     }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 14bbf11c043fe92376b3f38937c5f8eb50fd7a76..04eacaded56998075282da8a5ce0c727afe0e838 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -379,7 +379,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             .add(Attributes.CAMERA_DISTANCE)
             .add(Attributes.WAYPOINT_TRANSMIT_RANGE);
     }
-    public boolean shouldSendAttribute(Attribute attribute) { return true; } // Purpur - Ridables
+    public boolean shouldSendAttribute(Attribute attribute) { return true; } // Purpur - Ridables // Leaf - optimize attribute - diff on change
 
     @Override
     protected void checkFallDamage(double y, boolean onGround, BlockState state, BlockPos pos) {
@@ -1370,13 +1370,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     private void refreshDirtyAttributes() {
-        Set<AttributeInstance> attributesToUpdate = this.getAttributes().getAttributesToUpdate();
+        // Leaf start - optimize attribute
+        int[] attributesToUpdate = this.getAttributes().getAttributesToUpdateIds();
 
-        for (AttributeInstance attributeInstance : attributesToUpdate) {
-            this.onAttributeUpdated(attributeInstance.getAttribute());
+        for (int attribute : attributesToUpdate) {
+            this.onAttributeUpdated(org.dreeam.leaf.util.RegistryTypeManager.ATTRIBUTE[attribute]);
         }
-
-        attributesToUpdate.clear();
+        // Leaf end - optimize attribute
     }
 
     protected void onAttributeUpdated(Holder<Attribute> attribute) {
diff --git a/net/minecraft/world/entity/ai/attributes/Attribute.java b/net/minecraft/world/entity/ai/attributes/Attribute.java
index 69d59650b26fcb1c654d7a784022ac7f4c9230ce..41ffe149588f4070113f6a0c60f7ba7f6a24882e 100644
--- a/net/minecraft/world/entity/ai/attributes/Attribute.java
+++ b/net/minecraft/world/entity/ai/attributes/Attribute.java
@@ -16,10 +16,15 @@ public class Attribute {
     private boolean syncable;
     private final String descriptionId;
     public Attribute.Sentiment sentiment = Attribute.Sentiment.POSITIVE;
+    // Leaf start - optimize attribute
+    public final int id;
+    private static int size = 0;
+    // Leaf end - optimize attribute
 
     protected Attribute(String descriptionId, double defaultValue) {
         this.defaultValue = defaultValue;
         this.descriptionId = descriptionId;
+        this.id = size++; // Leaf - optimize attribute
     }
 
     public double getDefaultValue() {
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
index e42e6d01a8bb1bca5dea440d3f002ead68590e11..136fffc3c6c75ee9e6c27151ce802c90b2b25736 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
@@ -50,16 +50,19 @@ public class AttributeInstance {
 
     @VisibleForTesting
     Map<Identifier, AttributeModifier> getModifiers(AttributeModifier.Operation operation) {
-        return this.modifiersByOperation.computeIfAbsent(operation, operation1 -> new Object2ObjectOpenHashMap<>());
+        return this.modifiersByOperation.computeIfAbsent(operation, operation1 -> new Object2ObjectArrayMap<>()); // Leaf - optimize attribute
     }
 
+    // Leaf start - optimize attribute
+    private static final AttributeModifier[] EMPTY_ATTRIBUTE_MODIFIER_ARRAY = {};
     public Set<AttributeModifier> getModifiers() {
-        return ImmutableSet.copyOf(this.modifierById.values());
+        return this.modifierById.isEmpty() ? Set.of() : it.unimi.dsi.fastutil.objects.ObjectArraySet.ofUnchecked(this.modifierById.values().toArray(EMPTY_ATTRIBUTE_MODIFIER_ARRAY));
     }
 
     public Set<AttributeModifier> getPermanentModifiers() {
-        return ImmutableSet.copyOf(this.permanentModifiers.values());
+        return this.permanentModifiers.isEmpty() ? Set.of() : it.unimi.dsi.fastutil.objects.ObjectArraySet.ofUnchecked(this.permanentModifiers.values().toArray(EMPTY_ATTRIBUTE_MODIFIER_ARRAY));
     }
+    // Leaf end - optimize attribute
 
     public @Nullable AttributeModifier getModifier(Identifier id) {
         return this.modifierById.get(id);
@@ -136,12 +139,15 @@ public class AttributeInstance {
     }
 
     public double getValue() {
-        if (this.dirty) {
-            this.cachedValue = this.calculateValue();
-            this.dirty = false;
+        // Leaf start - optimize attribute
+        if (!this.dirty) {
+            return this.cachedValue;
         }
-
-        return this.cachedValue;
+        double value = this.calculateValue();
+        this.cachedValue = value;
+        this.dirty = false;
+        return value;
+        // Leaf end - optimize attribute
     }
 
     private double calculateValue() {
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeMap.java b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
index 852f53a4705c3fa201135a89a2803f45f283fc08..ba090c219a1cf4797a068b0dcf25481984a2f784 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -13,10 +13,12 @@ import net.minecraft.core.Holder;
 import net.minecraft.resources.Identifier;
 import org.jspecify.annotations.Nullable;
 
-public class AttributeMap {
-    public final Map<Holder<Attribute>, AttributeInstance> attributes = new Object2ObjectOpenHashMap<>();
-    private final Set<AttributeInstance> attributesToSync = new ObjectOpenHashSet<>();
-    private final Set<AttributeInstance> attributesToUpdate = new ObjectOpenHashSet<>();
+public class AttributeMap implements java.util.function.Consumer<AttributeInstance> { // Leaf - optimize attribute
+    // Leaf start - optimize attribute
+    public final Map<Holder<Attribute>, AttributeInstance> attributes = new org.dreeam.leaf.util.map.AttributeInstanceArrayMap();
+    private final Set<AttributeInstance> attributesToSync = new org.dreeam.leaf.util.map.AttributeInstanceSet((org.dreeam.leaf.util.map.AttributeInstanceArrayMap) attributes);
+    private final Set<AttributeInstance> attributesToUpdate = new org.dreeam.leaf.util.map.AttributeInstanceSet((org.dreeam.leaf.util.map.AttributeInstanceArrayMap) attributes);
+    // Leaf end - optimize attribute
     private final AttributeSupplier supplier;
     private final net.minecraft.world.entity.LivingEntity entity; // Purpur - Ridables
 
@@ -25,33 +27,107 @@ public class AttributeMap {
         this(supplier, null);
     }
     public AttributeMap(AttributeSupplier defaultAttributes, net.minecraft.world.entity.LivingEntity entity) {
-        this.entity = entity;
+        this.entity = entity instanceof net.minecraft.world.entity.ambient.Bat ? entity : null; // Leaf - optimize attribute - only check bat
         // Purpur end - Ridables
         this.supplier = defaultAttributes;
     }
 
     private void onAttributeModified(AttributeInstance instance) {
-        this.attributesToUpdate.add(instance);
-        if (instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))) { // Purpur - Ridables
-            this.attributesToSync.add(instance);
+        // Leaf start - optimize attribute
+        Attribute attribute = instance.getAttribute().value();
+        ((org.dreeam.leaf.util.map.AttributeInstanceSet) this.attributesToUpdate).addAttribute(attribute);
+        if (attribute.isClientSyncable() && (entity == null || entity.shouldSendAttribute(attribute))) { // Purpur - Ridables
+            ((org.dreeam.leaf.util.map.AttributeInstanceSet) this.attributesToSync).addAttribute(attribute);
         }
+        // Leaf end - optimize attribute
     }
 
+    // Leaf start - optimize attribute
+    @Override
+    public void accept(AttributeInstance instance) {
+        this.onAttributeModified(instance);
+    }
+    private static final AttributeInstance[] EMPTY_ATTRIBUTE_INSTANCE = new AttributeInstance[0];
+    @Deprecated
     public Set<AttributeInstance> getAttributesToSync() {
-        return this.attributesToSync;
+        if (attributesToSync.isEmpty()) {
+            return Set.of();
+        }
+        it.unimi.dsi.fastutil.objects.ReferenceArraySet<AttributeInstance> clone = it.unimi.dsi.fastutil.objects.ReferenceArraySet.ofUnchecked(attributesToSync.toArray(EMPTY_ATTRIBUTE_INSTANCE));
+        this.attributesToSync.clear();
+        return clone;
     }
 
+    @Deprecated
     public Set<AttributeInstance> getAttributesToUpdate() {
-        return this.attributesToUpdate;
+        if (attributesToUpdate.isEmpty()) {
+            return Set.of();
+        }
+        it.unimi.dsi.fastutil.objects.ReferenceArraySet<AttributeInstance> clone = it.unimi.dsi.fastutil.objects.ReferenceArraySet.ofUnchecked(attributesToUpdate.toArray(EMPTY_ATTRIBUTE_INSTANCE));
+        this.attributesToUpdate.clear();
+        return clone;
+    }
+
+    public boolean attributeDirty() {
+        return !attributesToSync.isEmpty();
+    }
+
+    public int[] getAttributesToUpdateIds() {
+        int[] clone = ((org.dreeam.leaf.util.map.AttributeInstanceSet) attributesToUpdate).inner.toIntArray();
+        this.attributesToUpdate.clear();
+        return clone;
+    }
+
+    public int[] getAttributesToSyncIds() {
+        int[] clone = ((org.dreeam.leaf.util.map.AttributeInstanceSet) attributesToSync).inner.toIntArray();
+        this.attributesToSync.clear();
+        return clone;
     }
 
     public Collection<AttributeInstance> getSyncableAttributes() {
-        return this.attributes.values().stream().filter(instance -> instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))).collect(Collectors.toList()); // Purpur - Ridables
+        List<AttributeInstance> list = new ArrayList<>(this.attributes.size());
+        for (AttributeInstance instance : this.attributes.values()) {
+            if (instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))) { // Purpur - Ridables
+                list.add(instance);
+            }
+        }
+        return list;
+    }
+
+    public List<net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot> getSyncableAttributesPacket() {
+        List<net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot> list = new ArrayList<>(attributes.size());
+        if (attributes instanceof org.dreeam.leaf.util.map.AttributeInstanceArrayMap map) {
+            for (AttributeInstance instance : map.elements()) {
+                if (instance != null && instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))) { // Purpur - Ridables
+                    list.add(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot(
+                        instance.getAttribute(), instance.getBaseValue(), instance.getModifiers()
+                    ));
+                }
+            }
+        } else {
+            for (AttributeInstance instance : attributes.values()) {
+                if (instance != null && instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))) { // Purpur - Ridables
+                    list.add(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket.AttributeSnapshot(
+                        instance.getAttribute(), instance.getBaseValue(), instance.getModifiers()
+                    ));
+                }
+            }
+        }
+        return list;
     }
 
     public @Nullable AttributeInstance getInstance(Holder<Attribute> attribute) {
-        return this.attributes.computeIfAbsent(attribute, holder -> this.supplier.createInstance(this::onAttributeModified, (Holder<Attribute>)holder));
+        AttributeInstance v;
+        if ((v = this.attributes.get(attribute)) == null) {
+            AttributeInstance newValue;
+            if ((newValue = this.supplier.createInstance(this, attribute)) != null) {
+                attributes.put(attribute, newValue);
+                return newValue;
+            }
+        }
+        return v;
     }
+    // Leaf end - optimize attribute
 
     public boolean hasAttribute(Holder<Attribute> attribute) {
         return this.attributes.get(attribute) != null || this.supplier.hasAttribute(attribute);
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java b/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
index 8bf6b76277c529433dbf541dee23156bbb94c946..c309b381c8c7caf0bc9142e436c8b801e110df9f 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
@@ -11,7 +11,7 @@ public class AttributeSupplier {
     private final Map<Holder<Attribute>, AttributeInstance> instances;
 
     AttributeSupplier(Map<Holder<Attribute>, AttributeInstance> instances) {
-        this.instances = instances;
+        this.instances = new org.dreeam.leaf.util.map.AttributeInstanceArrayMap(instances); // Leaf - optimize attribute
     }
 
     public AttributeInstance getAttributeInstance(Holder<Attribute> attribute) {
