From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 25 Oct 2024 22:27:07 -0400
Subject: [PATCH] Cache tile entity position


diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 939541b5ccea19e3ed619befd51ec07309abee62..3afe47db8529a9b129828f288cc3a20acbf1972a 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -977,10 +977,12 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         private final T blockEntity;
         private final BlockEntityTicker<T> ticker;
         private boolean loggedInvalidBlockState;
+        private final BlockPos cachedPos; // Leaf - Cache tile entity position
 
         private BoundTickingBlockEntity(final T blockEntity, final BlockEntityTicker<T> ticker) {
             this.blockEntity = blockEntity;
             this.ticker = ticker;
+            this.cachedPos = this.blockEntity.getBlockPos(); // Leaf - Cache tile entity position
         }
 
         @Override
@@ -1031,7 +1033,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
         @Override
         public BlockPos getPos() {
-            return this.blockEntity.getBlockPos();
+            return this.cachedPos; // Leaf - Cache tile entity position
         }
 
         @Override
@@ -1058,13 +1060,16 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     private static class RebindableTickingBlockEntityWrapper implements TickingBlockEntity {
         private TickingBlockEntity ticker;
+        private BlockPos cachedPos; // Leaf - Cache tile entity position
 
         private RebindableTickingBlockEntityWrapper(final TickingBlockEntity ticker) {
             this.ticker = ticker;
+            this.cachedPos = this.ticker.getPos(); // Leaf - Cache tile entity position
         }
 
         private void rebind(final TickingBlockEntity ticker) {
             this.ticker = ticker;
+            this.cachedPos = this.ticker.getPos(); // Leaf - Cache tile entity position
         }
 
         @Override
@@ -1079,7 +1084,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
         @Override
         public BlockPos getPos() {
-            return this.ticker.getPos();
+            return this.cachedPos; // Leaf - Cache tile entity position
         }
 
         @Override
