From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Thu, 8 May 2025 10:08:54 +0200
Subject: [PATCH] reduce PlayerChunk Updates


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index abe79d07196de0a10a382d4c37161c7eb4a604ae..405b8da8b886b5caac7ed774472e106374c42185 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1367,6 +1367,10 @@ public class ServerGamePacketListenerImpl
                     this.resetPosition();
                 }
 
+                // Leaf start - ReduceChunkSourceUpdates
+                final net.minecraft.world.level.ChunkPos playerStartChunkPosition = this.player.chunkPosition();
+                // Leaf end - ReduceChunkSourceUpdates
+
                 if (!this.updateAwaitingTeleport() && this.player.hasClientLoaded()) {
                     double d = clampHorizontal(packet.getX(this.player.getX())); final double toX = d; // Paper - OBFHELPER
                     double d1 = clampVertical(packet.getY(this.player.getY())); final double toY = d1; // Paper - OBFHELPER
@@ -1638,7 +1642,9 @@ public class ServerGamePacketListenerImpl
                                     && !isFallFlying
                                     && !isAutoSpinAttack
                                     && this.noBlocksAround(this.player);
-                                this.player.serverLevel().getChunkSource().move(this.player);
+                                if (!org.dreeam.leaf.config.modules.opt.ReduceChunkSourceUpdates.enabled || this.player.serverLevel() != serverLevel || this.player.chunkPosition() == playerStartChunkPosition) {
+                                    this.player.serverLevel().getChunkSource().move(this.player);
+                                }
                                 Vec3 vec3 = new Vec3(this.player.getX() - x, this.player.getY() - y, this.player.getZ() - z);
                                 this.player.setOnGroundWithMovement(packet.isOnGround(), packet.horizontalCollision(), vec3);
                                 this.player.doCheckFallDamage(vec3.x, vec3.y, vec3.z, packet.isOnGround());
