From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 8 May 2025 04:56:30 +0900
Subject: [PATCH] send and flush position while knockback player


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index c86dfd6e08c40e26c73210adf32f9179ed889994..b1c1d2a2f4e719b6f5c14f7f6d2710517598c52c 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -214,6 +214,8 @@ public class Commands {
         TriggerCommand.register(this.dispatcher);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        org.dreeam.leaf.command.TestKB.register(this.dispatcher);
+
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 477455fdfcc591a89823e88983eb12dabb078d9b..082b1ee1a100606993ed5cdef1631a8c341584eb 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1411,6 +1411,30 @@ public abstract class Player extends LivingEntity {
                                     player.setVelocity(event.getVelocity());
                                 }
 
+
+                                // Leaf start
+                                if (!cancelled) {
+                                    ServerPlayer target1 = (ServerPlayer) target;
+                                    if (this instanceof ServerPlayer player1) {
+                                        if (org.dreeam.leaf.command.TestKB.doMoveTarget) {
+                                            target1.aiStep();
+                                        }
+                                        if (org.dreeam.leaf.command.TestKB.sendToTarget || org.dreeam.leaf.command.TestKB.sendToBoth) {
+                                            target1.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(this));
+                                        }
+                                        if (org.dreeam.leaf.command.TestKB.sendToSelf || org.dreeam.leaf.command.TestKB.sendToBoth) {
+                                            player1.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(target1));
+                                        }
+                                        if (org.dreeam.leaf.command.TestKB.sendTargetVelocityToSelf) {
+                                            player1.connection.send(new ClientboundSetEntityMotionPacket(target));
+                                        } else if (org.dreeam.leaf.command.TestKB.sendTargetVelocityNoKnockbackToSelf) {
+                                            player1.connection.send(new ClientboundSetEntityMotionPacket(target.getId(), deltaMovement));
+                                        }
+                                        player1.hasImpulse = true;
+                                        target1.hasImpulse = true;
+                                    }
+                                }
+                                // Leaf end
                                 if (!cancelled) {
                                 ((ServerPlayer)target).connection.send(new ClientboundSetEntityMotionPacket(target));
                                 target.hurtMarked = false;
@@ -1480,6 +1504,17 @@ public abstract class Player extends LivingEntity {
                             }
 
                             this.causeFoodExhaustion(this.level().spigotConfig.combatExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+                            if (org.dreeam.leaf.command.TestKB.flush) {
+                                if (target instanceof ServerPlayer target1 && this instanceof ServerPlayer player1) {
+                                    target1.connection.connection.flushChannel();
+                                    player1.connection.connection.flushChannel();
+                                }
+                            } else if (org.dreeam.leaf.command.TestKB.flushMulti) {
+                                if (target instanceof ServerPlayer target1 && this instanceof ServerPlayer player1) {
+                                    target1.connection.resumeFlushing();
+                                    player1.connection.resumeFlushing();
+                                }
+                            }
                         } else {
                             this.sendSoundEffect(this, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_ATTACK_NODAMAGE, this.getSoundSource(), 1.0F, 1.0F); // Paper - send while respecting visibility
                             // CraftBukkit start - resync on cancelled event
