From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Mon, 12 May 2025 19:11:16 +0200
Subject: [PATCH] Remove streams in MobSensor


diff --git a/net/minecraft/world/entity/ai/sensing/MobSensor.java b/net/minecraft/world/entity/ai/sensing/MobSensor.java
index bda210b4809a5aade7ab4d0f26fdda4d5f53f619..18c7fe0b39b366d3ac4fd83415953c063e5c562f 100644
--- a/net/minecraft/world/entity/ai/sensing/MobSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/MobSensor.java
@@ -40,7 +40,14 @@ public class MobSensor<T extends LivingEntity> extends Sensor<T> {
     public void checkForMobsNearby(T sensingEntity) {
         Optional<List<LivingEntity>> memory = sensingEntity.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!memory.isEmpty()) {
-            boolean flag = memory.get().stream().anyMatch(livingEntity -> this.mobTest.test(sensingEntity, livingEntity));
+            boolean flag = false;
+            List<LivingEntity> entities = memory.get();
+            for (LivingEntity livingEntity : entities) {
+                if (this.mobTest.test(sensingEntity, livingEntity)) {
+                    flag = true;
+                    break;
+                }
+            }
             if (flag) {
                 this.mobDetected(sensingEntity);
             }
