From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 10 May 2025 22:04:42 +0200
Subject: [PATCH] Smart sort items in NearestItemSensor


diff --git a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
index 09fd13e2d958da8326276c4dadf25bf488aff5ac..70036d073b644ff5f486491486c38db5df643107 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -6,6 +6,7 @@ import java.util.List;
 import java.util.Optional;
 import java.util.Set;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
@@ -16,6 +17,8 @@ public class NearestItemSensor extends Sensor<Mob> {
     private static final long Y_RANGE = 16L;
     public static final int MAX_DISTANCE_TO_WANTED_ITEM = 32;
 
+    private final org.dreeam.leaf.util.fastBitRadixSort itemSorter;
+    public NearestItemSensor() {this.itemSorter = new org.dreeam.leaf.util.fastBitRadixSort();}
     @Override
     public Set<MemoryModuleType<?>> requires() {
         return ImmutableSet.of(MemoryModuleType.NEAREST_VISIBLE_WANTED_ITEM);
@@ -24,12 +27,13 @@ public class NearestItemSensor extends Sensor<Mob> {
     @Override
     protected void doTick(ServerLevel level, Mob entity) {
         Brain<?> brain = entity.getBrain();
-        List<ItemEntity> entitiesOfClass = level.getEntitiesOfClass(ItemEntity.class, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), itemEntity -> itemEntity.closerThan(entity, MAX_DISTANCE_TO_WANTED_ITEM) && entity.wantsToPickUp(level, itemEntity.getItem())); // Paper - Perf: Move predicate into getEntities
-        entitiesOfClass.sort(Comparator.comparingDouble(entity::distanceToSqr));
+        List<ItemEntity> entitiesOfClass = level.getEntitiesOfClass(ItemEntity.class, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), itemEntity -> itemEntity.closerThan(entity, MAX_DISTANCE_TO_WANTED_ITEM) && entity.wantsToPickUp(level, itemEntity.getItem()));
+        ItemEntity[] sortedItems = this.itemSorter.sort(entitiesOfClass, entity, ItemEntity.class);
+
         // Paper start - Perf: remove streams from hot code
         ItemEntity nearest = null;
-        for (final ItemEntity itemEntity : entitiesOfClass) {
-            if (entity.hasLineOfSight(itemEntity)) { // Paper - Perf: Move predicate into getEntities
+        for (final ItemEntity itemEntity : sortedItems) {
+            if (entity.hasLineOfSight(itemEntity)) {
                 nearest = itemEntity;
                 break;
             }
