From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Thu, 14 Nov 2024 19:54:22 +0800
Subject: [PATCH] Reduce fluid status allocation in NoiseBasedChunkGenerator


diff --git a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
index 216ae83ee6bcd21e20507ff5cfd47507393fdf2a..6f9a95dd7f76c0b06d0faaf1627d9c46be14c66b 100644
--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -67,15 +67,12 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
     }
 
     private static Aquifer.FluidPicker createFluidPicker(NoiseGeneratorSettings settings) {
-        Aquifer.FluidStatus fluidStatus = new Aquifer.FluidStatus(-54, Blocks.LAVA.defaultBlockState());
         int seaLevel = settings.seaLevel();
-        Aquifer.FluidStatus fluidStatus1 = new Aquifer.FluidStatus(seaLevel, settings.defaultFluid());
-        Aquifer.FluidStatus fluidStatus2 = new Aquifer.FluidStatus(DimensionType.MIN_Y * 2, Blocks.AIR.defaultBlockState());
         return (x, y, z) -> {
             if (SharedConstants.DEBUG_DISABLE_FLUID_GENERATION) {
-                return fluidStatus2;
+                return new Aquifer.FluidStatus(DimensionType.MIN_Y * 2, Blocks.AIR.defaultBlockState()); // Leaf - Reduce fluid status allocation in NoiseBasedChunkGenerator
             } else {
-                return y < Math.min(-54, seaLevel) ? fluidStatus : fluidStatus1;
+                return y < Math.min(-54, seaLevel) ? new Aquifer.FluidStatus(-54, Blocks.LAVA.defaultBlockState()) : new Aquifer.FluidStatus(seaLevel, settings.defaultFluid()); // Leaf - Reduce fluid status allocation in NoiseBasedChunkGenerator
             }
         };
     }
