From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wling-art <wlingzhenyu@163.com>
Date: Sat, 17 May 2025 03:31:03 +0800
Subject: [PATCH] Optimize isEyeInFluid


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index e8a24178cb74afee749bfb067ac010f5178e070c..bc0ec2c752f08659a628863724ba81faf6142068 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -261,7 +261,9 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public boolean wasTouchingWater;
     protected Object2DoubleMap<TagKey<Fluid>> fluidHeight = new Object2DoubleArrayMap<>(2);
     protected boolean wasEyeInWater;
-    private final Set<TagKey<Fluid>> fluidOnEyes = new HashSet<>();
+    private static final int FLUID_WATER = 1; // Leaf - Optimize isEyeInFluid
+    private static final int FLUID_LAVA = 2;  // Leaf - Optimize isEyeInFluid
+    private int fluidCache = 0; // Leaf - Optimize isEyeInFluid
     public int invulnerableTime;
     protected boolean firstTick = true;
     protected final SynchedEntityData entityData;
@@ -932,8 +934,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             if (this instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
                 this.setSharedFlagOnFire(this.remainingFireTicks > 0 && (!this.level.galeConfig().gameplayMechanics.hideFlamesOnEntitiesWithFireResistance || !livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE)));
             } else {
-            // Gale end - Slice - hide flames on entities with fire resistance
-            this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+                // Gale end - Slice - hide flames on entities with fire resistance
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0);
             } // Gale - Slice - hide flames on entities with fire resistance
         }
 
@@ -955,7 +957,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             && (!(this instanceof Player player) || !player.getAbilities().invulnerable))) {
             // Paper end - Configurable nether ceiling damage
             if (this.level.purpurConfig.teleportOnNetherCeilingDamage && this.level.getWorld().getEnvironment() == org.bukkit.World.Environment.NETHER && this instanceof ServerPlayer player) player.teleport(io.papermc.paper.util.MCUtil.toLocation(this.level, this.level.getSharedSpawnPos())); else // Purpur - Add option to teleport to spawn on nether ceiling damage
-            this.onBelowWorld();
+                this.onBelowWorld();
         }
     }
 
@@ -1261,7 +1263,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
                 if (oldDeltaMovement.x < -1e-6 || oldDeltaMovement.x > 1e-6 || oldDeltaMovement.z < -1e-6 || oldDeltaMovement.z > 1e-6) {
                     // Gale end - skip negligible planar movement multiplication
-                float blockSpeedFactor = this.getBlockSpeedFactor();
+                    float blockSpeedFactor = this.getBlockSpeedFactor();
                     // Gale start - skip negligible planar movement multiplication
                     if (blockSpeedFactor < 1 - 1e-6 || blockSpeedFactor > 1 + 1e-6) {
                         this.setDeltaMovement(oldDeltaMovement.multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
@@ -1414,8 +1416,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             } else {
                 BlockState blockState = this.level().getBlockState(blockPos);
                 return (!(yOffset <= 0.5) || !blockState.is(BlockTags.FENCES))
-                        && !blockState.is(BlockTags.WALLS)
-                        && !(blockState.getBlock() instanceof FenceGateBlock)
+                    && !blockState.is(BlockTags.WALLS)
+                    && !(blockState.getBlock() instanceof FenceGateBlock)
                     ? blockPos.atY(Mth.floor(this.position.y - yOffset))
                     : blockPos;
             }
@@ -1981,7 +1983,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     private void updateFluidOnEyes() {
         this.wasEyeInWater = this.isEyeInFluid(FluidTags.WATER);
-        this.fluidOnEyes.clear();
+        fluidCache = 0; // Leaf - Optimize isEyeInFluid
+
         double eyeY = this.getEyeY();
         if (!(
             this.getVehicle() instanceof AbstractBoat abstractBoat
@@ -1993,7 +1996,14 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             FluidState fluidState = this.level().getFluidState(blockPos);
             double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
             if (d > eyeY) {
-                this.fluidOnEyes.addAll(fluidState.getTagsAsSet()); // Leaf - Remove stream in updateFluidOnEyes
+                // Leaf start - Optimize isEyeInFluid
+                if (fluidState.is(FluidTags.WATER)) {
+                    setFluidStatus(FluidTags.WATER, true);
+                }
+                if (fluidState.is(FluidTags.LAVA)) {
+                    setFluidStatus(FluidTags.LAVA, true);
+                }
+                // Leaf end - Optimize isEyeInFluid
             }
         }
     }
@@ -2073,10 +2083,34 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
+    // Leaf start - Optimize isEyeInFluid
     public boolean isEyeInFluid(TagKey<Fluid> fluidTag) {
-        return this.fluidOnEyes.contains(fluidTag);
+        if (fluidTag == FluidTags.WATER) {
+            return (fluidCache & FLUID_WATER) != 0;
+        } else if (fluidTag == FluidTags.LAVA) {
+            return (fluidCache & FLUID_LAVA) != 0;
+        }
+        return false;
+    }
+
+    public void setFluidStatus(TagKey<Fluid> fluidTag, boolean isInFluid) {
+        if (fluidTag == FluidTags.WATER) {
+            if (isInFluid) {
+                fluidCache |= FLUID_WATER;
+            } else {
+                fluidCache &= ~FLUID_WATER;
+            }
+        } else if (fluidTag == FluidTags.LAVA) {
+            if (isInFluid) {
+                fluidCache |= FLUID_LAVA;
+            } else {
+                fluidCache &= ~FLUID_LAVA;
+            }
+        }
     }
 
+    // Leaf start - Optimize isEyeInFluid
+
     public boolean isInLava() {
         return !this.firstTick && this.fluidHeight.getDouble(FluidTags.LAVA) > 0.0;
     }
@@ -2208,7 +2242,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     public void push(Entity entity) {
         if (!entity.noPhysics && !this.noPhysics) { // Gale - Akarin - collision physics check before vehicle check
-        if (!this.isPassengerOfSameVehicle(entity)) {
+            if (!this.isPassengerOfSameVehicle(entity)) {
                 if (this.level.paperConfig().collisions.onlyPlayersCollide && !(entity instanceof ServerPlayer || this instanceof ServerPlayer)) return; // Paper - Collision option for requiring a player participant
                 double d = entity.getX() - this.getX();
                 double d1 = entity.getZ() - this.getZ();
@@ -2446,7 +2480,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 } else {
                     compound.put("Pos", this.newDoubleList(this.getX(), this.getY(), this.getZ()));
                 }
-             }
+            }
             // CraftBukkit end
 
             this.setDeltaMovement(io.papermc.paper.util.MCUtil.sanitizeNanInf(this.deltaMovement, 0D)); // Paper - remove NaN values before usage in saving
@@ -3218,8 +3252,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             Entity orig = craft == null ? null : craft.getHandle();
             if (this.getBukkitEntity() instanceof org.bukkit.entity.Vehicle && passenger.getBukkitEntity() instanceof org.bukkit.entity.LivingEntity) {
                 org.bukkit.event.vehicle.VehicleExitEvent event = new org.bukkit.event.vehicle.VehicleExitEvent(
-                        (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
-                        (org.bukkit.entity.LivingEntity) passenger.getBukkitEntity(), !suppressCancellation // Paper - Force entity dismount during teleportation
+                    (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
+                    (org.bukkit.entity.LivingEntity) passenger.getBukkitEntity(), !suppressCancellation // Paper - Force entity dismount during teleportation
                 );
                 // Suppress during worldgen
                 if (this.valid) {
@@ -3331,7 +3365,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public void setAsInsidePortal(Portal portal, BlockPos pos) {
         if (this.isOnPortalCooldown()) {
             if (!(level().purpurConfig.playerFixStuckPortal && this instanceof Player && !pos.equals(this.portalPos))) // Purpur - Fix stuck in portals
-            this.setPortalCooldown();
+                this.setPortalCooldown();
         } else if (this.level.purpurConfig.entitiesCanUsePortals || this instanceof ServerPlayer) { // Purpur - Entities can use portals
             if (this.portalProcess == null || !this.portalProcess.isSamePortal(portal)) {
                 this.portalProcess = new PortalProcessor(portal, pos.immutable());
@@ -3621,10 +3655,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             pluginManager.callEvent(entityCombustEvent);
             if (!entityCombustEvent.isCancelled()) {
                 this.igniteForSeconds(entityCombustEvent.getDuration(), false);
-            // Paper start - fix EntityCombustEvent cancellation
+                // Paper start - fix EntityCombustEvent cancellation
             } else {
                 this.setRemainingFireTicks(this.remainingFireTicks - 1);
-            // Paper end - fix EntityCombustEvent cancellation
+                // Paper end - fix EntityCombustEvent cancellation
             }
             // CraftBukkit end
         }
@@ -3771,32 +3805,32 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         String string = this.level() == null ? "~NULL~" : this.level().toString();
         return this.removalReason != null
             ? String.format(
-                Locale.ROOT,
-                "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b, removed=%s]", // Paper - add more info
-                this.getClass().getSimpleName(),
-                this.getName().getString(),
-                this.id,
-                this.uuid, // Paper - add more info
-                string,
-                this.getX(),
-                this.getY(),
-                this.getZ(),
-                this.chunkPosition(), this.tickCount, this.valid, // Paper - add more info
-                this.removalReason
-            )
+            Locale.ROOT,
+            "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b, removed=%s]", // Paper - add more info
+            this.getClass().getSimpleName(),
+            this.getName().getString(),
+            this.id,
+            this.uuid, // Paper - add more info
+            string,
+            this.getX(),
+            this.getY(),
+            this.getZ(),
+            this.chunkPosition(), this.tickCount, this.valid, // Paper - add more info
+            this.removalReason
+        )
             : String.format(
-                Locale.ROOT,
-                "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b]", // Paper - add more info
-                this.getClass().getSimpleName(),
-                this.getName().getString(),
-                this.id,
-                this.uuid, // Paper - add more info
-                string,
-                this.getX(),
-                this.getY(),
-                this.getZ(),
-                this.chunkPosition(), this.tickCount, this.valid // Paper - add more info
-            );
+            Locale.ROOT,
+            "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b]", // Paper - add more info
+            this.getClass().getSimpleName(),
+            this.getName().getString(),
+            this.id,
+            this.uuid, // Paper - add more info
+            string,
+            this.getX(),
+            this.getY(),
+            this.getZ(),
+            this.chunkPosition(), this.tickCount, this.valid // Paper - add more info
+        );
     }
 
     public final boolean isInvulnerableToBase(DamageSource damageSource) {
@@ -4295,11 +4329,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         // Gale end - VMP - skip entity move if movement is zero
         // CraftBukkit start - block invalid bounding boxes
         double minX = bb.minX,
-                minY = bb.minY,
-                minZ = bb.minZ,
-                maxX = bb.maxX,
-                maxY = bb.maxY,
-                maxZ = bb.maxZ;
+            minY = bb.minY,
+            minZ = bb.minZ,
+            maxX = bb.maxX,
+            maxY = bb.maxY,
+            maxZ = bb.maxZ;
         double len = bb.maxX - bb.minX;
         if (len < 0) maxX = minX;
         if (len > 64) maxX = minX + 64.0;
