From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Wed, 4 Jun 2025 20:54:03 +0900
Subject: [PATCH] throttle mob spawning


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 7bfc636fb442036f742903c4e69a8a9fcd0e6fc2..762bdf1fe19546f89d34b9efdad66b00dab80006 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -166,6 +166,17 @@ public final class NaturalSpawner {
             // Paper start - Optional per player mob spawns
             final boolean canSpawn;
             int maxSpawns = Integer.MAX_VALUE;
+            // Leaf start
+            if (org.dreeam.leaf.config.modules.opt.ThrottleNaturalMobSpawning.enabled) {
+                int spawnChance = org.dreeam.leaf.config.modules.opt.ThrottleNaturalMobSpawning.spawnChance[mobCategory.ordinal()];
+                long failedAttempt = org.dreeam.leaf.config.modules.opt.ThrottleNaturalMobSpawning.failedAttempts[mobCategory.ordinal()];
+                if (failedAttempt >= 0L
+                    && chunk.failedSpawnAttempts[mobCategory.ordinal()] >= failedAttempt
+                    && (level.random.nextInt() & 1023) > spawnChance) {
+                    continue;
+                }
+            }
+            // Leaf end
             if (level.paperConfig().entities.spawning.perPlayerMobSpawns) {
                 // Copied from getFilteredSpawningCategories
                 int limit = mobCategory.getMaxInstancesPerChunk();
