From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Wed, 11 Jun 2025 20:51:36 +0200
Subject: [PATCH] Paper: Fix excess slot updates / inventory state id desync

Original Patch: https://github.com/PaperMC/Paper/pull/12654

diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index ff2ff95ec9d94e2e31e8174196b384c37d56f38a..2a49a0bdeb61c4fadddc241c8ebca908959d7e9c 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -553,7 +553,7 @@ public abstract class AbstractContainerMenu {
 
                 slot.setChanged();
                 // CraftBukkit start - Make sure the client has the right slot contents
-                if (player instanceof ServerPlayer serverPlayer && slot.getMaxStackSize() != 64) {
+                if (player instanceof ServerPlayer serverPlayer && slot.getMaxStackSize() != net.minecraft.world.Container.MAX_STACK) {
                     serverPlayer.connection.send(new ClientboundContainerSetSlotPacket(this.containerId, this.incrementStateId(), slot.index, slot.getItem()));
                     // Updating a crafting inventory makes the client reset the result slot, have to send it again
                     if (this.getBukkitView().getType() == org.bukkit.event.inventory.InventoryType.WORKBENCH || this.getBukkitView().getType() == org.bukkit.event.inventory.InventoryType.CRAFTING) {
