From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Mon, 14 Apr 2025 18:07:21 +0200
Subject: [PATCH] Optimise MobEffectUtil#getDigSpeedAmplification


diff --git a/net/minecraft/world/effect/MobEffectUtil.java b/net/minecraft/world/effect/MobEffectUtil.java
index f1c57bc4310784999784d77678ea88912d6e41db..581aea02f6ed53d072caf47ddc051b497798fa33 100644
--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -27,17 +27,21 @@ public final class MobEffectUtil {
     }
 
     public static int getDigSpeedAmplification(final LivingEntity mob) {
-        int a = 0;
-        int b = 0;
-        if (mob.hasEffect(MobEffects.HASTE)) {
-            a = mob.getEffect(MobEffects.HASTE).getAmplifier();
+        // Leaf start - Optimise MobEffectUtil#getDigSpeedAmplification
+        int digAmplifier = 0;
+        int conduitAmplifier = 0;
+        MobEffectInstance digEffect = mob.getEffect(MobEffects.HASTE);
+        if (digEffect != null) {
+            digAmplifier = digEffect.getAmplifier();
         }
 
-        if (mob.hasEffect(MobEffects.CONDUIT_POWER)) {
-            b = mob.getEffect(MobEffects.CONDUIT_POWER).getAmplifier();
+        MobEffectInstance conduitEffect = mob.getEffect(MobEffects.CONDUIT_POWER);
+        if (conduitEffect != null) {
+            conduitAmplifier = conduitEffect.getAmplifier();
         }
 
-        return Math.max(a, b);
+        return Math.max(digAmplifier, conduitAmplifier);
+        // Leaf end - Optimise MobEffectUtil#getDigSpeedAmplification
     }
 
     public static boolean hasWaterBreathing(final LivingEntity mob) {
