From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 26 Jun 2025 19:51:08 +0900
Subject: [PATCH] clear target and attacker when mob is removed


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 3aedae4c11bd33502e3ac2ba57d4f99c4c900a62..af17e0ff346c9f11c45fb2c707beba6c3afea935 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -768,6 +768,17 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
         super.remove(reason, eventCause); // CraftBukkit
         this.brain.clearMemories();
+
+        // Leaf start
+        this.setLastHurtByMob(null);
+        if (this instanceof Mob mob) {
+            mob.setTarget(null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, false);
+        }
+        this.lastHurtByPlayer = null;
+        this.lastHurtByPlayerTime = 0;
+        this.lastDamageSource = null;
+        this.lastDamageStamp = 0L;
+        // Leaf end
     }
 
     protected void triggerOnDeathMobEffects(ServerLevel level, Entity.RemovalReason removalReason) {
