From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 28 Jun 2025 06:57:17 +0900
Subject: [PATCH] fix MC-298464


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index df23d80d6b18e900414aa02e5c1812f0a10f0fb7..ca7d754bcca2fdf201610f25fdc4b62f9e834e82 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3965,6 +3965,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 leashable.dropLeash(); // Paper drop lead
             }
             // Paper end - Fix item duplication and teleport issues
+            // Leaf start
+            if (this instanceof Mob mob) {
+                mob.getBrain().clearMemories();
+            }
+            // Leaf end
             entityx.restoreFrom(this);
             this.removeAfterChangingDimensions();
             // CraftBukkit start - Forward the CraftEntity to the new entity
