From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 28 Jun 2025 14:25:48 +0900
Subject: [PATCH] remove NearestLivingEntitySensor


diff --git a/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java b/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
index af7f0de1618cfd0e9149d996eec28db78351bce1..6c318a393161902dde4a822057c6a6ec0ca26042 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
@@ -18,9 +18,10 @@ public class NearestLivingEntitySensor<T extends LivingEntity> extends Sensor<T>
         double attributeValue = entity.getAttributeValue(Attributes.FOLLOW_RANGE);
         AABB aabb = entity.getBoundingBox().inflate(attributeValue, attributeValue, attributeValue);
         // Leaf start - Smart sort entities in NearestLivingEntitySensor
-        level.moonrise$getEntityLookup().getEntities(LivingEntity.class, entity, aabb, org.dreeam.leaf.util.FastBitRadixSort.unsafeBuffer(), LivingEntity::isAlive);
-        LivingEntity[] sorted = org.dreeam.leaf.util.FastBitRadixSort.unsafeSort(new LivingEntity[0], entity, LivingEntity.class);
-        List<LivingEntity> sortedList = java.util.Arrays.asList(sorted);
+        //level.moonrise$getEntityLookup().getEntities(LivingEntity.class, entity, aabb, org.dreeam.leaf.util.FastBitRadixSort.unsafeBuffer(), LivingEntity::isAlive);
+        //LivingEntity[] sorted = org.dreeam.leaf.util.FastBitRadixSort.unsafeSort(new LivingEntity[0], entity, LivingEntity.class);
+        //List<LivingEntity> sortedList = java.util.Arrays.asList(sorted);
+        List<LivingEntity> sortedList = List.of();
         Brain<?> brain = entity.getBrain();
         brain.setMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES, sortedList);
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES, new NearestVisibleLivingEntities(level, entity, sortedList));
