From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 4 Jul 2025 03:22:38 +0900
Subject: [PATCH] optimize mob despawn


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 175c4ab639f1516de06e3a16e89e779c59754e1c..073dec29d48b5ab8dd379a1e66a0de6ef101ca1f 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -797,13 +797,19 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             }
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
+
+            if (org.dreeam.leaf.config.modules.opt.OptimizeDespawn.enabled && tickRateManager.runsNormally()) { despawnMap.tick(this, this.entityTickList); } // Leaf - optimize despawn
             this.entityTickList
                 .forEach(
                     entity -> {
                         entity.activatedPriorityReset = false; // Pufferfish - DAB
                         if (!entity.isRemoved()) {
                             if (!tickRateManager.isEntityFrozen(entity)) {
-                                entity.checkDespawn();
+                                // Leaf start - optimize despawn
+                                if (!org.dreeam.leaf.config.modules.opt.OptimizeDespawn.enabled) {
+                                    entity.checkDespawn();
+                                }
+                                // Leaf end - optimize despawn
                                 if (true) { // Paper - rewrite chunk system
                                     Entity vehicle = entity.getVehicle();
                                     if (vehicle != null) {
@@ -945,6 +951,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
     private int currentIceAndSnowTick = 0; protected void resetIceAndSnowTick() { this.currentIceAndSnowTick = this.simpleRandom.nextInt(16); } // Gale - Airplane - optimize random calls in chunk ticking
 
+    public final org.dreeam.leaf.world.DespawnMap despawnMap = new org.dreeam.leaf.world.DespawnMap(paperConfig()); // Leaf - optimize despawn
     public void tickChunk(LevelChunk chunk, int randomTickSpeed) {
         final net.minecraft.world.level.levelgen.BitRandomSource simpleRandom = this.simpleRandom; // Paper - optimise random ticking // Leaf - Faster random generator - upcasting
         ChunkPos pos = chunk.getPos();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 427b0b42ea1c6418a733824f4af0de1b1e901cb9..67556752551e1d30519e2533543de9b04342420b 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5099,6 +5099,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public void checkDespawn() {
     }
 
+    // Leaf start - optimize despawn
+    public void leaf$checkDespawn(org.dreeam.leaf.world.DespawnMap map) {
+    }
+    public void leaf$checkDespawnFallback(org.dreeam.leaf.world.DespawnMap map) {
+    }
+    // Leaf end - optimize despawn
+
     public Vec3[] getQuadLeashHolderOffsets() {
         return Leashable.createQuadLeashOffsets(this, 0.0, 0.5, 0.5, 0.0);
     }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index e96605d652ca7265352ace09fd5e1892b4bbcd65..29029c58637f5622cc542ed8826b807b15f20644 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -763,6 +763,53 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         }
     }
 
+    // Leaf start - optimize despawn
+    @Override
+    public void leaf$checkDespawn(org.dreeam.leaf.world.DespawnMap map) {
+        if (isRemoved()) {
+        } else if (map.difficultyIsPeaceful && this.shouldDespawnInPeaceful()) {
+            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        } else if (this.isPersistenceRequired() || this.requiresCustomPersistence()) {
+            this.noActionTime = 0;
+        } else {
+            map.checkDespawn(this);
+        }
+    }
+    @Override
+    public void leaf$checkDespawnFallback(org.dreeam.leaf.world.DespawnMap map) {
+        if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldActuallyDespawnInPeaceful()) { //Paper - allow changing despawnInPeaceful
+            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
+            Entity nearestPlayer = map.checkDespawnFallback(this); // Paper - Affects Spawning API
+            if (nearestPlayer != null) {
+                // Paper start - Configurable despawn distances
+                final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
+                final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
+                final double dy = Math.abs(nearestPlayer.getY() - this.getY());
+                final double dySqr = Mth.square(dy);
+                final double dxSqr = Mth.square(nearestPlayer.getX() - this.getX());
+                final double dzSqr = Mth.square(nearestPlayer.getZ() - this.getZ());
+                final double distanceSquared = dxSqr + dzSqr + dySqr;
+                // Despawn if hard/soft limit is exceeded
+                if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                }
+
+                if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
+                    if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                        this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                    }
+                } else {
+                    // Paper end - Configurable despawn distances
+                    this.noActionTime = 0;
+                }
+            }
+        } else {
+            this.noActionTime = 0;
+        }
+    }
+    // Leaf end - optimize despawn
+
     @Override
     protected final void serverAiStep() {
         this.noActionTime++;
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index de09a91b89661118e460842453e33f383ea08a94..e3cdff55261b2ff2c3d1cb1cf46b633a340458c9 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -956,6 +956,15 @@ public class EnderDragon extends Mob implements Enemy {
     public void checkDespawn() {
     }
 
+    // Leaf start - optimize despawn
+    @Override
+    public void leaf$checkDespawn(org.dreeam.leaf.world.DespawnMap map) {
+    }
+    @Override
+    public void leaf$checkDespawnFallback(org.dreeam.leaf.world.DespawnMap map) {
+    }
+    // Leaf end - optimize despawn
+
     public EnderDragonPart[] getSubEntities() {
         return this.subEntities;
     }
diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index e3c34d5f00ce64b3c08b10cfcb3dd14ebe8c1977..3171868523bc9b767cb9da5f58c43890fc8214da 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -695,6 +695,32 @@ public class WitherBoss extends Monster implements RangedAttackMob {
         }
     }
 
+
+    // Leaf start - optimize despawn
+    @Override
+    public void leaf$checkDespawn(org.dreeam.leaf.world.DespawnMap map) {
+        if (isRemoved()) {
+            return;
+        }
+        if (map.difficultyIsPeaceful && this.shouldDespawnInPeaceful()) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        } else {
+            this.noActionTime = 0;
+        }
+    }
+    @Override
+    public void leaf$checkDespawnFallback(org.dreeam.leaf.world.DespawnMap map) {
+        if (isRemoved()) {
+            return;
+        }
+        if (map.difficultyIsPeaceful && this.shouldDespawnInPeaceful()) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        } else {
+            this.noActionTime = 0;
+        }
+    }
+    // Leaf end - optimize despawn
+
     @Override
     public boolean addEffect(MobEffectInstance effectInstance, @Nullable Entity entity) {
         return false;
diff --git a/net/minecraft/world/entity/projectile/ShulkerBullet.java b/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 00154ba80175bcb07b3378f19514fec1700c94e9..ec0a30d76df2a20ceb895b2837b0f24bfe887aba 100644
--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -197,6 +197,22 @@ public class ShulkerBullet extends Projectile {
         }
     }
 
+
+    // Leaf start - optimize despawn
+    @Override
+    public void leaf$checkDespawn(org.dreeam.leaf.world.DespawnMap map) {
+        if (!isRemoved() && map.difficultyIsPeaceful) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        }
+    }
+    @Override
+    public void leaf$checkDespawnFallback(org.dreeam.leaf.world.DespawnMap map) {
+        if (!isRemoved() && map.difficultyIsPeaceful) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        }
+    }
+    // Leaf end - optimize despawn
+
     @Override
     protected double getDefaultGravity() {
         return 0.04;
