From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 4 Jul 2025 03:22:38 +0900
Subject: [PATCH] optimize mob despawn


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 7fb763e89e43698bfb2b9fcf6296705384e8624a..91752678f887559132921cff61697478e5f44e3d 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -797,13 +797,19 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             }
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
+
+            if (org.dreeam.leaf.config.modules.opt.OptimizeDespawn.enabled && tickRateManager.runsNormally()) { despawnMap.tick(this, this.entityTickList); } // Leaf - optimize despawn
             this.entityTickList
                 .forEach(
                     entity -> {
                         entity.activatedPriorityReset = false; // Pufferfish - DAB
                         if (!entity.isRemoved()) {
                             if (!tickRateManager.isEntityFrozen(entity)) {
-                                entity.checkDespawn();
+                                // Leaf start - optimize despawn
+                                if (!org.dreeam.leaf.config.modules.opt.OptimizeDespawn.enabled) {
+                                    entity.checkDespawn();
+                                }
+                                // Leaf end - optimize despawn
                                 if (true) { // Paper - rewrite chunk system
                                     Entity vehicle = entity.getVehicle();
                                     if (vehicle != null) {
@@ -945,6 +951,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
     private int currentIceAndSnowTick = 0; protected void resetIceAndSnowTick() { this.currentIceAndSnowTick = this.simpleRandom.nextInt(16); } // Gale - Airplane - optimize random calls in chunk ticking
 
+    public final org.dreeam.leaf.world.DespawnMap despawnMap = new org.dreeam.leaf.world.DespawnMap(paperConfig()); // Leaf - optimize despawn
     public void tickChunk(LevelChunk chunk, int randomTickSpeed) {
         final net.minecraft.world.level.levelgen.BitRandomSource simpleRandom = this.simpleRandom; // Paper - optimise random ticking // Leaf - Faster random generator - upcasting
         ChunkPos pos = chunk.getPos();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 428b2114749780770cc617d5e1aedea7f1b28b2d..ec061987ed6aa05475edc9ade6e4e572d9cd35a6 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5133,6 +5133,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public void checkDespawn() {
     }
 
+    // Leaf start - optimize despawn
+    public void leafCheckDespawn(org.dreeam.leaf.world.DespawnMap map) {
+    }
+    // Leaf end - optimize despawn
+
     public Vec3[] getQuadLeashHolderOffsets() {
         return Leashable.createQuadLeashOffsets(this, 0.0, 0.5, 0.5, 0.0);
     }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 751c8c17c3c2968fad358ae8807eaca54de82a34..bdd7348a5a3e0250f45549f145a04a97a5becfcf 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -762,6 +762,20 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         }
     }
 
+    // Leaf start - optimize despawn
+    @Override
+    public void leafCheckDespawn(org.dreeam.leaf.world.DespawnMap map) {
+        if (isRemoved()) {
+        } else if (map.difficultyIsPeaceful && this.shouldDespawnInPeaceful()) {
+            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        } else if (this.isPersistenceRequired() || this.requiresCustomPersistence()) {
+            this.noActionTime = 0;
+        } else {
+            map.checkDespawn(this);
+        }
+    }
+    // Leaf end - optimize despawn
+
     @Override
     protected final void serverAiStep() {
         this.noActionTime++;
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index de09a91b89661118e460842453e33f383ea08a94..70f6c350e3825d7a07ef7466d1f98df87daccc8b 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -956,6 +956,12 @@ public class EnderDragon extends Mob implements Enemy {
     public void checkDespawn() {
     }
 
+    // Leaf start - optimize despawn
+    @Override
+    public void leafCheckDespawn(org.dreeam.leaf.world.DespawnMap map) {
+    }
+    // Leaf end - optimize despawn
+
     public EnderDragonPart[] getSubEntities() {
         return this.subEntities;
     }
diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index e3c34d5f00ce64b3c08b10cfcb3dd14ebe8c1977..d70c5393df353af2bb4daf8d22b4af4634633496 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -695,6 +695,21 @@ public class WitherBoss extends Monster implements RangedAttackMob {
         }
     }
 
+
+    // Leaf start - optimize despawn
+    @Override
+    public void leafCheckDespawn(org.dreeam.leaf.world.DespawnMap map) {
+        if (isRemoved()) {
+            return;
+        }
+        if (map.difficultyIsPeaceful && this.shouldDespawnInPeaceful()) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        } else {
+            this.noActionTime = 0;
+        }
+    }
+    // Leaf end - optimize despawn
+
     @Override
     public boolean addEffect(MobEffectInstance effectInstance, @Nullable Entity entity) {
         return false;
diff --git a/net/minecraft/world/entity/projectile/ShulkerBullet.java b/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 00154ba80175bcb07b3378f19514fec1700c94e9..0f7ef763af55ce885106f3a66ba265526532403e 100644
--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -197,6 +197,16 @@ public class ShulkerBullet extends Projectile {
         }
     }
 
+
+    // Leaf start - optimize despawn
+    @Override
+    public void leafCheckDespawn(org.dreeam.leaf.world.DespawnMap map) {
+        if (!isRemoved() && map.difficultyIsPeaceful) {
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        }
+    }
+    // Leaf end - optimize despawn
+
     @Override
     protected double getDefaultGravity() {
         return 0.04;
