From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 23 Mar 2025 20:48:16 +0100
Subject: [PATCH] Use direct iteration on Sensing.tick


diff --git a/net/minecraft/world/entity/ai/sensing/Sensing.java b/net/minecraft/world/entity/ai/sensing/Sensing.java
index be7b5d1b671aaa60d190b649ebd9086f6b3ff9b3..fea91f2b1738ee82cc2da356851e9a91fa262d55 100644
--- a/net/minecraft/world/entity/ai/sensing/Sensing.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensing.java
@@ -40,7 +40,12 @@ public class Sensing {
         } else {
             var expiringNow = this.expiring[this.nextToExpireIndex];
 
-            expiringNow.forEach(this.seen::remove);
+            // Leaf start - Use direct iteration on Sensing.tick
+            var iterator = expiringNow.iterator();
+            while (iterator.hasNext()) {
+                this.seen.remove(iterator.nextInt());
+            }
+            // Leaf end - Use direct iteration on Sensing.tick
             expiringNow.clear();
 
             this.currentCacheAddIndex++;
