From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Wed, 2 Apr 2025 23:03:22 +0200
Subject: [PATCH] Null handling on MultifaceSpreader

WHYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY

diff --git a/net/minecraft/world/level/block/MultifaceSpreader.java b/net/minecraft/world/level/block/MultifaceSpreader.java
index 2d16bb0d6ee72567a33f0fd28f7b9370029d1fe3..11020eb46ac21bf1f4c85d44b5da7597d0f18571 100644
--- a/net/minecraft/world/level/block/MultifaceSpreader.java
+++ b/net/minecraft/world/level/block/MultifaceSpreader.java
@@ -172,6 +172,13 @@ public class MultifaceSpreader {
         }
 
         default boolean placeBlock(final LevelAccessor level, final MultifaceSpreader.SpreadPos spreadPos, final BlockState oldState, final boolean postProcess) {
+            // Leaf start - Null handling on MultifaceSpreader
+            // Check for null
+            if (spreadPos.source() == null || spreadPos.pos() == null) {
+                org.dreeam.leaf.config.LeafConfig.LOGGER.warn("Invalid SpreadPos with null source or position: {}", spreadPos);
+                return false;
+            }
+            // Leaf end - Null handling on MultifaceSpreader
             BlockState spreadState = this.getStateForPlacement(oldState, level, spreadPos.pos(), spreadPos.face());
             if (spreadState != null) {
                 if (postProcess) {
