From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:01:50 +0900
Subject: [PATCH] Optimize getScaledTrackingDistance


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 1205516c00335b5eacabad1b0ef022f9365fc439..20be4ace1c594565881c4829d10d940c3117841d 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -974,11 +974,15 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public int getScaledTrackingDistance(final int range) {
-        return this.entityBroadcastRangePercentage() * range / 100;
+        // Leaf start - Optimize getScaledTrackingDistance
+        int p = this.getProperties().entityBroadcastRangePercentage.get();
+        if (p == 100) return range;
+        return p * range / 100;
+        // Leaf end - Optimize getScaledTrackingDistance
     }
 
     public int entityBroadcastRangePercentage() {
-        return this.getProperties().entityBroadcastRangePercentage.get();
+        return this.getProperties().entityBroadcastRangePercentage.get(); // Leaf - Optimize getScaledTrackingDistance - diff on change
     }
 
     public void setEntityBroadcastRangePercentage(final int range) {
