From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:03:31 +0900
Subject: [PATCH] Optimize SynchedEntityData#packDirty


diff --git a/net/minecraft/network/syncher/SynchedEntityData.java b/net/minecraft/network/syncher/SynchedEntityData.java
index bdb4f9f5052fa5c919783b9cbe744c4c55cf9cae..33ed3ede938b0d737db7f5302b3748a6342f6963 100644
--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -84,7 +84,15 @@ public class SynchedEntityData {
         }
 
         this.isDirty = false;
-        List<SynchedEntityData.DataValue<?>> result = new ArrayList<>();
+        // Leaf start - Optimize SynchedEntityData#packDirty
+        int count = 0;
+        for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
+            if (dataItem.isDirty()) {
+                count++;
+            }
+        }
+        ArrayList<SynchedEntityData.DataValue<?>> result = new ArrayList<>(count);
+        // Leaf end - Optimize SynchedEntityData#packDirty
 
         for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
             if (dataItem.isDirty()) {
