From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 23 Mar 2025 20:48:16 +0100
Subject: [PATCH] Use direct iteration on Sensing.tick


diff --git a/net/minecraft/world/entity/ai/sensing/Sensing.java b/net/minecraft/world/entity/ai/sensing/Sensing.java
index 1eef00b8fe7d13554d16b7da1101f1f2daebddf3..dffe896ee16c88957d76adb716bf1ce11f7ca9ac 100644
--- a/net/minecraft/world/entity/ai/sensing/Sensing.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensing.java
@@ -41,7 +41,12 @@ public class Sensing {
         } else {
             var expiringNow = this.expiring[this.nextToExpireIndex];
 
-            expiringNow.forEach(this.seen::remove);
+            // Leaf start - Use direct iteration on Sensing.tick
+            var iterator = expiringNow.iterator();
+            while (iterator.hasNext()) {
+                this.seen.remove(iterator.nextInt());
+            }
+            // Leaf end - Use direct iteration on Sensing.tick
             expiringNow.clear();
 
             this.currentCacheAddIndex++;
