From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <mc@jvavav.com>
Date: Sat, 19 Apr 2025 22:12:19 +0800
Subject: [PATCH] Optimize BlockEntityType#isValid


diff --git a/net/minecraft/world/level/block/entity/BlockEntityType.java b/net/minecraft/world/level/block/entity/BlockEntityType.java
index 4c91a0cd674e23ae31e21ba7014fdbb15514cc45..d8e12059194985e5356153fc42ff21ed5b9bcf33 100644
--- a/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -17,6 +17,14 @@ public class BlockEntityType<T extends BlockEntity> {
     public BlockEntityType(final BlockEntityType.BlockEntitySupplier<? extends T> factory, final Set<Block> validBlocks) {
         this.factory = factory;
         this.validBlocks = validBlocks;
+        // Leaf start - Optimize BlockEntityType#isValid
+        for (Block validBlock : validBlocks) {
+            if (validBlock.blockEntityType != null) {
+                throw new IllegalStateException("Duplicate block entity type");
+            }
+            validBlock.blockEntityType = this;
+        }
+        // Leaf end - Optimize BlockEntityType#isValid
     }
 
     public T create(final BlockPos worldPosition, final BlockState blockState) {
@@ -24,7 +32,7 @@ public class BlockEntityType<T extends BlockEntity> {
     }
 
     public boolean isValid(final BlockState state) {
-        return this.validBlocks.contains(state.getBlock());
+        return state.getBlock().blockEntityType == this; // Leaf - Optimize BlockEntityType#isValid - remove hash lookup
     }
 
     @Deprecated
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 886b22fea779bfb432520956f596eefbebc8956c..24f9a6a3c48a72264c5edb4918cb7fabdb6a00d7 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -100,6 +100,7 @@ public abstract class BlockBehaviour implements FeatureElement {
     protected final BlockBehaviour.Properties properties;
     protected final Optional<ResourceKey<LootTable>> drops;
     protected final String descriptionId;
+    public @Nullable BlockEntityType blockEntityType = null; // Leaf - Optimize BlockEntityType#isValid
 
     public BlockBehaviour(final BlockBehaviour.Properties properties) {
         this.hasCollision = properties.hasCollision;
