From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Wed, 2 Apr 2025 23:03:22 +0200
Subject: [PATCH] Null handling on MultifaceSpreader

WHYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY

diff --git a/net/minecraft/world/level/block/MultifaceSpreader.java b/net/minecraft/world/level/block/MultifaceSpreader.java
index c856a2d2d56f78805f02c3bbb551cedfa040df7e..1dcb4634a489916b1c7a2b10ac16ad4a7962fb41 100644
--- a/net/minecraft/world/level/block/MultifaceSpreader.java
+++ b/net/minecraft/world/level/block/MultifaceSpreader.java
@@ -146,6 +146,14 @@ public class MultifaceSpreader {
         }
 
         default boolean placeBlock(LevelAccessor level, MultifaceSpreader.SpreadPos pos, BlockState state, boolean markForPostprocessing) {
+            // Leaf start - Null handling on MultifaceSpreader
+            // Check for null
+            if (pos.source() == null || pos.pos() == null) {
+                org.dreeam.leaf.config.LeafConfig.LOGGER.warn("Invalid SpreadPos with null source or position: {}", pos);
+                return false;
+            }
+            // Leaf end - Null handling on MultifaceSpreader
+
             BlockState stateForPlacement = this.getStateForPlacement(state, level, pos.pos(), pos.face());
             if (stateForPlacement != null) {
                 if (markForPostprocessing) {
