From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 6 May 2025 17:44:16 +0900
Subject: [PATCH] Protocol Core


diff --git a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
index 2f1e5c3de421bd07e30c417bf88935be85cf573c..702d368bb5cf11863197b92d17f03738d754ab35 100644
--- a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
+++ b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
@@ -46,6 +46,12 @@ public interface CustomPacketPayload {
                     return;
                 }
                 // Leaves end - protocol core
+                // Leaf start - Protocol core
+                if (value instanceof org.dreeam.leaf.protocol.LeafCustomPayload payload) {
+                    org.dreeam.leaf.protocol.Protocols.write(output, payload);
+                    return;
+                }
+                // Leaf end - Protocol core
                 this.writeCap(output, value.type(), value);
             }
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 442dcc3418341b004516368dd321989f408a4a8d..8ef0f2da02a7b1a89604c94f6546cd6c9d1337db 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2016,6 +2016,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         profiler.popPush("server gui refresh");
 
         org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handleTick(tickCount); // Leaves - protocol
+        org.dreeam.leaf.protocol.Protocols.tickServer(this); // Leaf - Protocol core
 
         for (Runnable tickable : this.tickables) {
             tickable.run();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 0020569d2bd01cf6fc96d78bcce082f18acd6213..ff370fb4ad700a48267164149d53ef1545186b5d 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -799,6 +799,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             }
         }
         // Purpur end - Ridables
+        org.dreeam.leaf.protocol.Protocols.tickPlayer(this); // Leaf - Protocol core
     }
 
     private void updatePlayerAttributes() {
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index f564aaf75ca18b53035bb434d8f5a8e7672e17c4..25866d0f9bcb9d9eb27ed7c7274c2e9b2ea03721 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -186,6 +186,12 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
         net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
 
+        // Leaf start - Protocol core
+        if (this instanceof ServerGamePacketListenerImpl gamePacketListener) {
+            org.dreeam.leaf.protocol.Protocols.handle(gamePacketListener.player, discardedPayload);
+        }
+        // Leaf end - Protocol core
+
         final net.minecraft.resources.Identifier identifier = packet.payload().type().id();
         final byte[] data = discardedPayload.data();
         try {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 3e1b3f759888eab888a040d2ae7975d3325c5867..061f77baa65bac052092ecc03fcaf14b0b41c163 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -611,6 +611,7 @@ public abstract class PlayerList {
         // Paper end - Fix kick event leave message not being sent
         org.purpurmc.purpur.task.BossBarTask.removeFromAll(player.getBukkitEntity()); // Purpur - Implement TPSBar
         org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handlePlayerLeave(player); // Leaves - protocol
+        org.dreeam.leaf.protocol.Protocols.disconnected(player); // Leaf - Protocol core
         net.minecraft.server.network.ServerGamePacketListenerImpl.afkCooldown.removeLong(player.getBukkitEntity().getUniqueId()); // Leaf - Improve Purpur AFK system
         ServerLevel level = player.level();
         player.awardStat(Stats.LEAVE_GAME);
