From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Mon, 14 Apr 2025 18:07:21 +0200
Subject: [PATCH] Optimise MobEffectUtil#getDigSpeedAmplification


diff --git a/net/minecraft/world/effect/MobEffectUtil.java b/net/minecraft/world/effect/MobEffectUtil.java
index 0ba2dbba9d6c13c167d15d7d186fa98b91f6392c..0e788a89b49be656279e0582b05312c35b8018ca 100644
--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -27,17 +27,21 @@ public final class MobEffectUtil {
     }
 
     public static int getDigSpeedAmplification(LivingEntity entity) {
-        int i = 0;
-        int i1 = 0;
-        if (entity.hasEffect(MobEffects.HASTE)) {
-            i = entity.getEffect(MobEffects.HASTE).getAmplifier();
+        // Leaf start - Optimise MobEffectUtil#getDigSpeedAmplification
+        int digAmplifier = 0;
+        int conduitAmplifier = 0;
+        MobEffectInstance digEffect = entity.getEffect(MobEffects.HASTE);
+        if (digEffect != null) {
+            digAmplifier = digEffect.getAmplifier();
         }
 
-        if (entity.hasEffect(MobEffects.CONDUIT_POWER)) {
-            i1 = entity.getEffect(MobEffects.CONDUIT_POWER).getAmplifier();
+        MobEffectInstance conduitEffect = entity.getEffect(MobEffects.CONDUIT_POWER);
+        if (conduitEffect != null) {
+            conduitAmplifier = conduitEffect.getAmplifier();
         }
 
-        return Math.max(i, i1);
+        return Math.max(digAmplifier, conduitAmplifier);
+        // Leaf end - Optimise MobEffectUtil#getDigSpeedAmplification
     }
 
     public static boolean hasWaterBreathing(LivingEntity entity) {
