From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 6 May 2025 17:44:16 +0900
Subject: [PATCH] Protocol Core


diff --git a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
index 2f1e5c3de421bd07e30c417bf88935be85cf573c..702d368bb5cf11863197b92d17f03738d754ab35 100644
--- a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
+++ b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
@@ -46,6 +46,12 @@ public interface CustomPacketPayload {
                     return;
                 }
                 // Leaves end - protocol core
+                // Leaf start - Protocol core
+                if (value instanceof org.dreeam.leaf.protocol.LeafCustomPayload payload) {
+                    org.dreeam.leaf.protocol.Protocols.write(output, payload);
+                    return;
+                }
+                // Leaf end - Protocol core
                 this.writeCap(output, value.type(), value);
             }
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index dfd5e5601ee6811241e641163be61769fb077143..68b11880849a1c41a33bab383f4ab7c3c368e6e1 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1993,6 +1993,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         profiler.popPush("server gui refresh");
 
         org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handleTick(tickCount); // Leaves - protocol
+        org.dreeam.leaf.protocol.Protocols.tickServer(this); // Leaf - Protocol core
 
         for (Runnable tickable : this.tickables) {
             tickable.run();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 286572cec2746ea41b34fcdb72f91dcfb8771793..e1b2022e7e063bc1e30a7139f8d452e0810771ac 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -790,6 +790,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             }
         }
         // Purpur end - Ridables
+        org.dreeam.leaf.protocol.Protocols.tickPlayer(this); // Leaf - Protocol core
     }
 
     private void updatePlayerAttributes() {
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 36cd6537f55ae713afbec15e17c409dee515600b..bfa522910ce7e9cb83668b69fa4c40daec8ab910 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -186,6 +186,12 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
         net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
 
+        // Leaf start - Protocol core
+        if (this instanceof ServerGamePacketListenerImpl gamePacketListener) {
+            org.dreeam.leaf.protocol.Protocols.handle(gamePacketListener.player, discardedPayload);
+        }
+        // Leaf end - Protocol core
+
         final net.minecraft.resources.Identifier identifier = packet.payload().type().id();
         final byte[] data = discardedPayload.data();
         try {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index e101e3f254adf74b66c603c4111a7d3d6609ccd3..6469a2b443373854b98912485b8592dc03abdcee 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -621,6 +621,7 @@ public abstract class PlayerList {
         // Paper end - Fix kick event leave message not being sent
         org.purpurmc.purpur.task.BossBarTask.removeFromAll(player.getBukkitEntity()); // Purpur - Implement TPSBar
         org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handlePlayerLeave(player); // Leaves - protocol
+        org.dreeam.leaf.protocol.Protocols.disconnected(player); // Leaf - Protocol core
         net.minecraft.server.network.ServerGamePacketListenerImpl.afkCooldown.removeLong(player.getBukkitEntity().getUniqueId()); // Leaf - Improve Purpur AFK system
         ServerLevel level = player.level();
         player.awardStat(Stats.LEAVE_GAME);
