From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 3 Jun 2025 15:41:12 +0900
Subject: [PATCH] optimize structure map


diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index 2ed9b81819bdfbe664096f817493770cdc1ac527..01a205eca9627736004a8e91e43b8aaa71a892f1 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -71,8 +71,8 @@ public abstract class ChunkAccess implements LightChunk, StructureAccess, BiomeR
     protected final @Nullable BlendingData blendingData;
     public final Map<Heightmap.Types, Heightmap> heightmaps = Maps.newEnumMap(Heightmap.Types.class);
     // Paper - rewrite chunk system
-    private final Map<Structure, StructureStart> structureStarts = Maps.newHashMap();
-    private final Map<Structure, LongSet> structureReferences = Maps.newHashMap();
+    private final Map<Structure, StructureStart> structureStarts = new it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap<>(); // Leaf - optimize structure map
+    private final Map<Structure, LongSet> structureReferences = new it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap<>(); // Leaf - optimize structure map
     protected final Map<BlockPos, CompoundTag> pendingBlockEntities = Maps.newHashMap();
     public final Map<BlockPos, BlockEntity> blockEntities = new Object2ObjectOpenHashMap<>();
     protected final LevelHeightAccessor levelHeightAccessor;
@@ -291,7 +291,7 @@ public abstract class ChunkAccess implements LightChunk, StructureAccess, BiomeR
     }
 
     public Map<Structure, StructureStart> getAllStarts() {
-        return Collections.unmodifiableMap(this.structureStarts);
+        return this.structureStarts; // Leaf - optimize structure map
     }
 
     public void setAllStarts(final Map<Structure, StructureStart> starts) {
@@ -307,13 +307,13 @@ public abstract class ChunkAccess implements LightChunk, StructureAccess, BiomeR
 
     @Override
     public void addReferenceForStructure(final Structure structure, final long reference) {
-        this.structureReferences.computeIfAbsent(structure, k -> new LongOpenHashSet()).add(reference);
+        this.structureReferences.computeIfAbsent(structure, k -> new it.unimi.dsi.fastutil.longs.LongArraySet()).add(reference); // Leaf - optimize structure map
         this.markUnsaved();
     }
 
     @Override
     public Map<Structure, LongSet> getAllReferences() {
-        return Collections.unmodifiableMap(this.structureReferences);
+        return this.structureReferences; // Leaf - optimize structure map
     }
 
     @Override
