From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 6 May 2025 17:44:16 +0900
Subject: [PATCH] Protocol Core


diff --git a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
index 1da034752064312962a4144f91be5265a9a08997..878e0e0f61a3020b120249d8b04a8e8c31da4976 100644
--- a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
+++ b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
@@ -40,6 +40,12 @@ public interface CustomPacketPayload {
 
             @Override
             public void encode(B buffer, CustomPacketPayload value) {
+                // Leaf start - Protocol core
+                if (value instanceof org.dreeam.leaf.protocol.LeafCustomPayload payload) {
+                    org.dreeam.leaf.protocol.Protocols.write(buffer, payload);
+                    return;
+                }
+                // Leaf end - Protocol core
                 this.writeCap(buffer, value.type(), value);
             }
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index e26178f8a50a902134141b6c6ab6243f34c0e3da..37da62fd44e6815a2451732d13a8edbe648777af 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1993,6 +1993,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         profilerFiller.popPush("server gui refresh");
 
+        org.dreeam.leaf.protocol.Protocols.tickServer(this); // Leaf - Protocol core
+
         for (Runnable runnable : this.tickables) {
             runnable.run();
         }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index bd97fe2058d236ba5c8e25e60d4b3216917fb5df..33e42bdce2a9b4dd7c627a953b55ff5a824f5e44 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -787,6 +787,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             }
         }
         // Purpur end - Ridables
+        org.dreeam.leaf.protocol.Protocols.tickPlayer(this); // Leaf - Protocol core
     }
 
     private void updatePlayerAttributes() {
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index f436a3f8b91325de2fbb9f96bdd9a01211f1e09e..0e1411d820caef8a6a80bca07d9c76e9615ff13e 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -178,6 +178,12 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
         net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
 
+        // Leaf start - Protocol core
+        if (this instanceof ServerGamePacketListenerImpl gamePacketListener) {
+            org.dreeam.leaf.protocol.Protocols.handle(gamePacketListener.player, discardedPayload);
+        }
+        // Leaf end - Protocol core
+
         final net.minecraft.resources.Identifier identifier = packet.payload().type().id();
         final byte[] data = discardedPayload.data();
         try {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 97cc842e147e604864c6fd2cd1ef0d8a63d41178..a6ab772731afa1b339c16fc91aa915cd35c71f4d 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -451,6 +451,7 @@ public abstract class PlayerList {
     }
 
     public net.kyori.adventure.text.@Nullable Component remove(ServerPlayer player) { // CraftBukkit - return string // Paper - return Component
+        org.dreeam.leaf.protocol.Protocols.disconnected(player); // Leaf - Protocol core
         // Paper start - Fix kick event leave message not being sent
         return this.remove(player, net.kyori.adventure.text.Component.translatable("multiplayer.player.left", net.kyori.adventure.text.format.NamedTextColor.YELLOW, io.papermc.paper.configuration.GlobalConfiguration.get().messages.useDisplayNameInQuitMessage ? player.getBukkitEntity().displayName() : io.papermc.paper.adventure.PaperAdventure.asAdventure(player.getDisplayName())));
     }
