From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <mc@jvavav.com>
Date: Sat, 19 Apr 2025 22:12:19 +0800
Subject: [PATCH] Optimize BlockEntityType#isValid


diff --git a/net/minecraft/world/level/block/entity/BlockEntityType.java b/net/minecraft/world/level/block/entity/BlockEntityType.java
index 8b392c168d9cfefa972aa7f1e9b68711f9683024..889ee600c6cc3ed49e24fcdcbb2164447edc7a4c 100644
--- a/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -295,6 +295,14 @@ public class BlockEntityType<T extends BlockEntity> {
     private BlockEntityType(BlockEntityType.BlockEntitySupplier<? extends T> factory, Set<Block> validBlocks) {
         this.factory = factory;
         this.validBlocks = validBlocks;
+        // Leaf start - Optimize BlockEntityType#isValid
+        for (Block validBlock : validBlocks) {
+            if (validBlock.blockEntityType != null) {
+                throw new IllegalStateException("Duplicate block entity type");
+            }
+            validBlock.blockEntityType = this;
+        }
+        // Leaf end - Optimize BlockEntityType#isValid
     }
 
     public T create(BlockPos pos, BlockState state) {
@@ -302,7 +310,7 @@ public class BlockEntityType<T extends BlockEntity> {
     }
 
     public boolean isValid(BlockState state) {
-        return this.validBlocks.contains(state.getBlock());
+        return state.getBlock().blockEntityType == this; // Leaf - Optimize BlockEntityType#isValid - remove hash lookup
     }
 
     @Deprecated
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 80775acc7500d76663b1bd7a135480cd96c4e295..47cff92e3244b556ee565b088003abd2784d4341 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -100,6 +100,7 @@ public abstract class BlockBehaviour implements FeatureElement {
     protected final BlockBehaviour.Properties properties;
     protected final Optional<ResourceKey<LootTable>> drops;
     protected final String descriptionId;
+    public net.minecraft.world.level.block.entity.@Nullable BlockEntityType blockEntityType = null; // Leaf - Optimize BlockEntityType#isValid
 
     public BlockBehaviour(BlockBehaviour.Properties properties) {
         this.hasCollision = properties.hasCollision;
