From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Fri, 9 May 2025 23:50:55 +0200
Subject: [PATCH] Only tick items at hand


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index dc1f4fcbd284735f8c98cd2f60095a892a875ce5..b7c83473cde3df58f966e183af92160e9fde0b51 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -851,12 +851,19 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 } // Gale - Hydrinity - increase time statistics in intervals
             }
 
+            // Leaf start - Only tick items at hand
+            if (org.dreeam.leaf.config.modules.opt.OptimizeItemTicking.onlyTickItemsInHand) {
+                this.synchronizeSpecialItemUpdates(this.getMainHandItem());
+                this.synchronizeSpecialItemUpdates(this.getOffhandItem());
+            } else {
             for (int i = 0; i < this.getInventory().getContainerSize(); i++) {
                 ItemStack itemStack = this.getInventory().getItem(i);
                 if (!itemStack.isEmpty()) {
                     this.synchronizeSpecialItemUpdates(itemStack);
                 }
             }
+            }
+            // Leaf end - Only tick items at hand
 
             if (this.getHealth() != this.lastSentHealth
                 || this.lastSentFood != this.foodData.getFoodLevel()
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index da86654718388f70e752627ade08db930e2278e4..013ccfa032723a78c99871f88252347a7c34c5d7 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -527,7 +527,13 @@ public abstract class Player extends Avatar implements ContainerUser {
         }
 
         this.tickRegeneration();
+        // Leaf start - Only tick items at hand
+        if (org.dreeam.leaf.config.modules.opt.OptimizeItemTicking.onlyTickItemsInHand) {
+            this.getMainHandItem().inventoryTick(this.level(), this, EquipmentSlot.MAINHAND);
+        } else {
         this.inventory.tick();
+        }
+        // Leaf end - Only tick items at hand
         if (this.abilities.flying && !this.isPassenger()) {
             this.resetFallDistance();
         }
