From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 11 May 2025 19:45:58 +0200
Subject: [PATCH] Flush location while knockback


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index ea7c260b29ee7872baf7a3d3d86fb0c6123644e2..d27a8d767711ea68c996c77bedc643b3c4f08698 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1118,6 +1118,12 @@ public abstract class Player extends Avatar implements ContainerUser {
                         this.itemAttackInteraction(entity, attackingItemStack, damageSource, true);
                         this.damageStatsAndHearts(entity, oldLivingEntityHealth);
                         this.causeFoodExhaustion(this.level().spigotConfig.combatExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+                        // Leaf start - Flush location while knockback
+                        if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback && this instanceof ServerPlayer attacker && entity instanceof ServerPlayer victim) {
+                            attacker.connection.connection.flushChannel();
+                            victim.connection.connection.flushChannel();
+                        }
+                        // Leaf end - Flush location while knockback
                     } else {
                         this.playServerSideSound(SoundEvents.PLAYER_ATTACK_NODAMAGE);
                     }
@@ -1274,7 +1280,17 @@ public abstract class Player extends Avatar implements ContainerUser {
             }
 
             if (!cancelled) {
-            ((ServerPlayer)entity).connection.send(new ClientboundSetEntityMotionPacket(entity));
+                // Leaf start - Flush location while knockback
+                if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback) {
+                    entity.needsSync = true;
+                    net.minecraft.server.level.ChunkMap.TrackedEntity targetTrackedEntity = entity.moonrise$getTrackedEntity();
+                    if (targetTrackedEntity != null) {
+                        targetTrackedEntity.serverEntity.sendChanges();
+                    }
+                } else {
+                    ((ServerPlayer) entity).connection.send(new ClientboundSetEntityMotionPacket(entity));
+                }
+                // Leaf end - Flush location while knockback
             entity.hurtMarked = false;
             entity.setDeltaMovement(oldMovement);
             }
@@ -1389,6 +1405,12 @@ public abstract class Player extends Avatar implements ContainerUser {
         this.itemAttackInteraction(target, weaponItem, damageSource, wasHurt);
         this.damageStatsAndHearts(target, oldLivingEntityHealth);
         this.causeFoodExhaustion(this.level().spigotConfig.combatExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+        // Leaf start - Flush location while knockback
+        if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback && this instanceof ServerPlayer attacker && target instanceof ServerPlayer victim) {
+            attacker.connection.connection.flushChannel();
+            victim.connection.connection.flushChannel();
+        }
+        // Leaf end - Flush location while knockback
         return true;
     }
 
