From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 10 May 2025 23:18:17 +0200
Subject: [PATCH] Optimise player movement checks


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 8ec27c522d4a974dc1e06d760af2b1bb113c98c4..60681cf62b6353226b7e1c299d5a887443f9d4f9 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1248,7 +1248,7 @@ public abstract class Entity
             }
             // Paper end
 
-            delta = this.maybeBackOffFromEdge(delta, moverType);
+            if (!org.dreeam.leaf.config.modules.opt.OptimizePlayerMovementProcessing.enabled) delta = this.maybeBackOffFromEdge(delta, moverType); // Leaf - Optimise player movement checks
             Vec3 movement = this.collide(delta);
             double movementLength = movement.lengthSqr();
             if (movementLength > 1.0E-7 || delta.lengthSqr() - movementLength < 1.0E-7) {
