From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 10 May 2025 23:18:17 +0200
Subject: [PATCH] Optimise player movement checks


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 02f02a8a36a4fcf28bf137393f6b6ac0cb8fd535..78efab6f31ca19e20b9847f4df7098029765f3d3 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1248,7 +1248,7 @@ public abstract class Entity
             }
             // Paper end
 
-            delta = this.maybeBackOffFromEdge(delta, moverType);
+            if (!org.dreeam.leaf.config.modules.opt.OptimizePlayerMovementProcessing.enabled) delta = this.maybeBackOffFromEdge(delta, moverType); // Leaf - Optimise player movement checks
             Vec3 movement = this.collide(delta);
             double movementLength = movement.lengthSqr();
             if (movementLength > 1.0E-7 || delta.lengthSqr() - movementLength < 1.0E-7) {
