From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 11 May 2025 19:45:58 +0200
Subject: [PATCH] Flush location while knockback


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 5948068ffdfcdc3d9a59211988df97521eff3a67..0d028c07c05a08fc543af186bb75cc70f5d98bdb 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1118,6 +1118,12 @@ public abstract class Player extends Avatar implements ContainerUser {
                         this.itemAttackInteraction(target, weaponItem, damageSource, true);
                         this.damageStatsAndHearts(target, f3);
                         this.causeFoodExhaustion(this.level().spigotConfig.combatExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+                        // Leaf start - Flush location while knockback
+                        if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback && this instanceof ServerPlayer player1 && target instanceof ServerPlayer player2) {
+                            player1.connection.connection.flushChannel();
+                            player2.connection.connection.flushChannel();
+                        }
+                        // Leaf end - Flush location while knockback
                     } else {
                         this.playServerSideSound(SoundEvents.PLAYER_ATTACK_NODAMAGE);
                     }
@@ -1270,6 +1276,13 @@ public abstract class Player extends Avatar implements ContainerUser {
             }
 
             if (!cancelled) {
+            // Leaf start - Flush location while knockback
+            if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback && target instanceof ServerPlayer targetPlayer && this instanceof ServerPlayer player1) {
+                targetPlayer.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(this));
+                player1.connection.send(new net.minecraft.network.protocol.game.ClientboundSetEntityMotionPacket(targetPlayer));
+                player1.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(targetPlayer));
+            }
+            // Leaf end - Flush location while knockback
             ((ServerPlayer)target).connection.send(new ClientboundSetEntityMotionPacket(target));
             target.hurtMarked = false;
             target.setDeltaMovement(currentMovement);
@@ -1380,6 +1393,12 @@ public abstract class Player extends Avatar implements ContainerUser {
                     this.itemAttackInteraction(target, itemBySlot, damageSource, flag);
                     this.damageStatsAndHearts(target, f2);
                     this.causeFoodExhaustion(this.level().spigotConfig.combatExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+                    // Leaf start - Flush location while knockback
+                    if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback && this instanceof ServerPlayer player1 && target instanceof ServerPlayer player2) {
+                        player1.connection.connection.flushChannel();
+                        player2.connection.connection.flushChannel();
+                    }
+                    // Leaf end - Flush location while knockback
                     return true;
                 }
             }
