From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Fri, 9 May 2025 23:50:55 +0200
Subject: [PATCH] Only tick items at hand


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index e1b2022e7e063bc1e30a7139f8d452e0810771ac..e59cbde71d74630dfcfa1ba79600f82496c86a00 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -851,12 +851,19 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 } // Gale - Hydrinity - increase time statistics in intervals
             }
 
+            // Leaf start - Only tick items at hand
+            if (org.dreeam.leaf.config.modules.opt.OptimizeItemTicking.onlyTickItemsInHand) {
+                this.synchronizeSpecialItemUpdates(this.getMainHandItem());
+                this.synchronizeSpecialItemUpdates(this.getOffhandItem());
+            } else {
             for (int i = 0; i < this.getInventory().getContainerSize(); i++) {
                 ItemStack itemStack = this.getInventory().getItem(i);
                 if (!itemStack.isEmpty()) {
                     this.synchronizeSpecialItemUpdates(itemStack);
                 }
             }
+            }
+            // Leaf end - Only tick items at hand
 
             if (this.getHealth() != this.lastSentHealth
                 || this.lastSentFood != this.foodData.getFoodLevel()
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 04d88c162a854c8bd48cbac5bce787654bf31545..2a0ae51cb0450f80abeaf3d858206512e9c6a4f3 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -527,7 +527,14 @@ public abstract class Player extends Avatar implements ContainerUser {
         }
 
         this.tickRegeneration();
+        // Leaf start - Only tick items at hand
+        if (org.dreeam.leaf.config.modules.opt.OptimizeItemTicking.onlyTickItemsInHand) {
+            this.getMainHandItem().inventoryTick(this.level(), this, EquipmentSlot.MAINHAND);
+            this.getOffhandItem().inventoryTick(this.level(), this, EquipmentSlot.OFFHAND);
+        } else {
         this.inventory.tick();
+        }
+        // Leaf end - Only tick items at hand
         if (this.abilities.flying && !this.isPassenger()) {
             this.resetFallDistance();
         }
