From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: adabugra <57899270+adabugra@users.noreply.github.com>
Date: Thu, 15 May 2025 19:09:04 +0300
Subject: [PATCH] Add configurable death item drop knockback settings


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index e59cbde71d74630dfcfa1ba79600f82496c86a00..ad5c46db4bbb42aca73831d8865d6b196886ad21 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1196,7 +1196,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         if (!keepInventory) {
             for (ItemStack item : this.getInventory().getContents()) {
                 if (!item.isEmpty() && !EnchantmentHelper.has(item, net.minecraft.world.item.enchantment.EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
-                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, false, null))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, org.dreeam.leaf.config.modules.gameplay.DeathItemDropKnockback.dropAround, false, false, null))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event) // Leaf - Add configurable death item drop knockback settings
                 }
             }
         }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 7589d704227f4497cb8d8129b22bf210df9dac83..e7b39414c7140715bbb996693588e1722e9a8f20 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4396,9 +4396,9 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         }
 
         if (randomly) {
-            float pow = this.random.nextFloat() * 0.5F;
+            float pow = this.random.nextFloat() * (float) org.dreeam.leaf.config.modules.gameplay.DeathItemDropKnockback.horizontalForce; // Leaf - Add configurable death item drop knockback settings
             float dir = this.random.nextFloat() * (float) (Math.PI * 2);
-            entity.setDeltaMovement(-Mth.sin(dir) * pow, 0.2F, Mth.cos(dir) * pow);
+            entity.setDeltaMovement(-Mth.sin(dir) * pow, (float) org.dreeam.leaf.config.modules.gameplay.DeathItemDropKnockback.verticalForce, Mth.cos(dir) * pow); // Leaf - Add configurable death item drop knockback settings
         } else {
             float pow = 0.3F;
             float sinX = Mth.sin(this.getXRot() * Mth.DEG_TO_RAD);
