From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:01:50 +0900
Subject: [PATCH] Optimize getScaledTrackingDistance


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index d0faf5c21e59430a1aab97c8a3fce9d0e0301a4f..8a16865a0e19603ff7a6a4fa274bec6013c0c68c 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -1016,11 +1016,15 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public int getScaledTrackingDistance(final int range) {
-        return this.entityBroadcastRangePercentage() * range / 100;
+        // Leaf start - Optimize getScaledTrackingDistance
+        int p = this.getProperties().entityBroadcastRangePercentage.get();
+        if (p == 100) return range;
+        return p * range / 100;
+        // Leaf end - Optimize getScaledTrackingDistance
     }
 
     public int entityBroadcastRangePercentage() {
-        return this.getProperties().entityBroadcastRangePercentage.get();
+        return this.getProperties().entityBroadcastRangePercentage.get(); // Leaf - Optimize getScaledTrackingDistance - diff on change
     }
 
     public void setEntityBroadcastRangePercentage(final int range) {
