From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 6 May 2025 17:44:16 +0900
Subject: [PATCH] Protocol Core


diff --git a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
index 081ca42b6a6644eb8c2b5ac9105fecf8ad39d501..8883d942b9c03982bd853ba5f90eb91f06cc8b55 100644
--- a/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
+++ b/net/minecraft/network/protocol/common/custom/CustomPacketPayload.java
@@ -46,6 +46,12 @@ public interface CustomPacketPayload {
                     return;
                 }
                 // Leaves end - protocol core
+                // Leaf start - Protocol core
+                if (value instanceof org.dreeam.leaf.protocol.LeafCustomPayload payload) {
+                    org.dreeam.leaf.protocol.Protocols.write(buffer, payload);
+                    return;
+                }
+                // Leaf end - Protocol core
                 this.writeCap(buffer, value.type(), value);
             }
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index adfc271bce540ad9aea908f3f2188e506b9a5e64..348d9040f21a3aa90a1376ec9fd3c3cafe5b9c7d 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1990,6 +1990,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         profilerFiller.popPush("server gui refresh");
 
         org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handleTick(tickCount); // Leaves - protocol
+        org.dreeam.leaf.protocol.Protocols.tickServer(this); // Leaf - Protocol core
 
         for (Runnable runnable : this.tickables) {
             runnable.run();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 1a023728f09adee06189d0f869aeb07ce6bec924..6dab07576f079db883f1212d7976a47342571a60 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -787,6 +787,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             }
         }
         // Purpur end - Ridables
+        org.dreeam.leaf.protocol.Protocols.tickPlayer(this); // Leaf - Protocol core
     }
 
     private void updatePlayerAttributes() {
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 655f1af6d27f320bad3a8d89cf9dc686f8684440..f919f11f5aaf3c0c80869cfa08521b2c263867df 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -192,6 +192,12 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
         net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
 
+        // Leaf start - Protocol core
+        if (this instanceof ServerGamePacketListenerImpl gamePacketListener) {
+            org.dreeam.leaf.protocol.Protocols.handle(gamePacketListener.player, discardedPayload);
+        }
+        // Leaf end - Protocol core
+
         final net.minecraft.resources.Identifier identifier = packet.payload().type().id();
         final byte[] data = discardedPayload.data();
         try {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index d4eefa2d1fa004e212e9b16093c92b6635a671bf..284940e73feb577c28848fa5cce2c541e3c5d02b 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -609,6 +609,7 @@ public abstract class PlayerList {
         // Paper end - Fix kick event leave message not being sent
         org.purpurmc.purpur.task.BossBarTask.removeFromAll(player.getBukkitEntity()); // Purpur - Implement TPSBar
         org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handlePlayerLeave(player); // Leaves - protocol
+        org.dreeam.leaf.protocol.Protocols.disconnected(player); // Leaf - Protocol core
         net.minecraft.server.network.ServerGamePacketListenerImpl.afkCooldown.remove(player.getBukkitEntity().getUniqueId()); // Leaf - Improve Purpur AFK system
         ServerLevel serverLevel = player.level();
         player.awardStat(Stats.LEAVE_GAME);
