From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:03:31 +0900
Subject: [PATCH] Optimize SynchedEntityData#packDirty


diff --git a/net/minecraft/network/syncher/SynchedEntityData.java b/net/minecraft/network/syncher/SynchedEntityData.java
index 0075164bf057271d63a8b97473aa5c41c03eac91..c45b34ddd3d1e9ac8b8f9dbcbefa770618e4c2d8 100644
--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -84,7 +84,15 @@ public class SynchedEntityData {
         }
 
         this.isDirty = false;
-        List<SynchedEntityData.DataValue<?>> result = new ArrayList<>();
+        // Leaf start - Optimize SynchedEntityData#packDirty
+        int count = 0;
+        for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
+            if (dataItem.isDirty()) {
+                count++;
+            }
+        }
+        ArrayList<SynchedEntityData.DataValue<?>> result = new ArrayList<>(count);
+        // Leaf end - Optimize SynchedEntityData#packDirty
 
         for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
             if (dataItem.isDirty()) {
