From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:01:50 +0900
Subject: [PATCH] Optimize getScaledTrackingDistance


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index a8890cbe24479fd65c0de1688187fd250c5ef2c8..0e77c717d3481aae5d591a08eed1eb5f1439cdbf 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -982,11 +982,17 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public int getScaledTrackingDistance(int trackingDistance) {
-        return this.entityBroadcastRangePercentage() * trackingDistance / 100;
+        // Leaf start - Optimize getScaledTrackingDistance
+        int p = this.getProperties().entityBroadcastRangePercentage.get();
+        if (p == 100) {
+            return trackingDistance;
+        }
+        return p * trackingDistance / 100;
+        // Leaf end - Optimize getScaledTrackingDistance
     }
 
     public int entityBroadcastRangePercentage() {
-        return this.getProperties().entityBroadcastRangePercentage.get();
+        return this.getProperties().entityBroadcastRangePercentage.get(); // Leaf - Optimize getScaledTrackingDistance - diff on change
     }
 
     public void setEntityBroadcastRangePercentage(int entityBroadcastRangePercentage) {
