From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Fri, 9 May 2025 23:50:55 +0200
Subject: [PATCH] Only tick items at hand


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 71c30caa61791ce7f758be66fd878157ac00084d..d6da48e85f90ab6952f9b1fca9a4fa6f14e54b26 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -848,12 +848,19 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 } // Gale - Hydrinity - increase time statistics in intervals
             }
 
+            // Leaf start - Only tick items at hand
+            if (org.dreeam.leaf.config.modules.opt.OptimizeItemTicking.onlyTickItemsInHand) {
+                this.synchronizeSpecialItemUpdates(this.getMainHandItem());
+                this.synchronizeSpecialItemUpdates(this.getOffhandItem());
+            } else {
             for (int i = 0; i < this.getInventory().getContainerSize(); i++) {
                 ItemStack item = this.getInventory().getItem(i);
                 if (!item.isEmpty()) {
                     this.synchronizeSpecialItemUpdates(item);
                 }
             }
+            }
+            // Leaf end - Only tick items at hand
 
             if (this.getHealth() != this.lastSentHealth
                 || this.lastSentFood != this.foodData.getFoodLevel()
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 94a8e31089ba4d4620ee18e71bb7dbdbba3b37d1..b03a730248b8ffda48750ac88eaa4d1864aac4c8 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -532,7 +532,14 @@ public abstract class Player extends Avatar implements ContainerUser {
         }
 
         this.tickRegeneration();
+        // Leaf start - Only tick items at hand
+        if (org.dreeam.leaf.config.modules.opt.OptimizeItemTicking.onlyTickItemsInHand) {
+            this.getMainHandItem().inventoryTick(this.level(), this, EquipmentSlot.MAINHAND);
+            this.getOffhandItem().inventoryTick(this.level(), this, EquipmentSlot.OFFHAND);
+        } else {
         this.inventory.tick();
+        }
+        // Leaf end - Only tick items at hand
         if (this.abilities.flying && !this.isPassenger()) {
             this.resetFallDistance();
         }
