From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrlingXD <90316914+wling-art@users.noreply.github.com>
Date: Sat, 17 May 2025 08:25:33 +0800
Subject: [PATCH] Optimize isEyeInFluid


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 23534448301c3ee7d7cb485c6ab19c4575fc59b2..cf22a560c331f08f7c1729ad5a40d966e1726a96 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2049,7 +2049,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     this.awardStat(Stats.SWIM_ONE_CM, rounded);
                     this.causeFoodExhaustion(this.level().spigotConfig.swimMultiplier * (float) rounded * 0.01F, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.SWIM); // CraftBukkit - EntityExhaustionEvent // Spigot
                 }
-            } else if (this.isEyeInFluid(FluidTags.WATER)) {
+            } else if (this.isEyeInWater()) { // Leaf - Optimize isEyeInFluid
                 int rounded = Math.round((float)Math.sqrt(dx * dx + dy * dy + dz * dz) * 100.0F);
                 if (rounded > 0) {
                     this.awardStat(Stats.WALK_UNDER_WATER_ONE_CM, rounded);
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 3daaf20ee8febfca8a3c9c76644cf8d6da8899c2..af0d37f13c63f60784dc803a1d328ca44fa8128d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -292,7 +292,14 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     public boolean wasTouchingWater;
     protected Object2DoubleMap<TagKey<Fluid>> fluidHeight = new Object2DoubleArrayMap<>(2);
     protected boolean wasEyeInWater;
-    private final Set<TagKey<Fluid>> fluidOnEyes = new HashSet<>();
+    // Leaf start - Optimize isEyeInFluid
+    // Remove original field since plugin should not direct access to it, and able to
+    // expose potential incompatibility asap.
+    // In paper api, if plugins have custom conditions, then It's more reasonable to
+    // use isEyeInFluid and their own conditions.
+    //private final Set<TagKey<Fluid>> fluidOnEyes = null;
+    private int isInWaterOrLava;
+    // Leaf end - Optimize isEyeInFluid
     public int invulnerableTime;
     protected boolean firstTick = true;
     protected final SynchedEntityData entityData;
@@ -2123,8 +2130,8 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     private void updateFluidOnEyes() {
-        this.wasEyeInWater = this.isEyeInFluid(FluidTags.WATER);
-        this.fluidOnEyes.clear();
+        this.wasEyeInWater = this.isInWaterOrLava == 1; // Leaf - Optimize isEyeInFluid
+        this.isInWaterOrLava = 0; // Leaf - Optimize isEyeInFluid - reset cache
         double eyeY = this.getEyeY();
         if (!(
             this.getVehicle() instanceof AbstractBoat abstractBoat
@@ -2136,7 +2143,9 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             FluidState fluidState = this.level().getFluidState(blockPos);
             double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
             if (d > eyeY) {
-                fluidState.getTags().forEach(this.fluidOnEyes::add);
+                // Leaf start - Optimize isEyeInFluid
+                this.isInWaterOrLava = fluidState.isEmpty() ? 0 : fluidState.is(FluidTags.WATER) ? 1 : 2;
+                // Leaf end - Optimize isEyeInFluid
             }
         }
     }
@@ -2216,9 +2225,25 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
     }
 
+    // Leaf start - Optimize isEyeInFluid
     public boolean isEyeInFluid(TagKey<Fluid> fluidTag) {
-        return this.fluidOnEyes.contains(fluidTag);
+        if (isInWaterOrLava == 0) {
+            return false;
+        }
+        if (fluidTag == FluidTags.WATER && isInWaterOrLava == 1) {
+            return true;
+        }
+        return fluidTag == FluidTags.LAVA && isInWaterOrLava == 2;
+    }
+
+    public boolean isEyeInWater() {
+        return isInWaterOrLava == 1;
+    }
+
+    public boolean isEyeInLava() {
+        return isInWaterOrLava == 2;
     }
+    // Leaf end - Optimize isEyeInFluid
 
     public boolean isInLava() {
         return !this.firstTick && this.fluidHeight.getDouble(FluidTags.LAVA) > 0.0;
diff --git a/net/minecraft/world/entity/ExperienceOrb.java b/net/minecraft/world/entity/ExperienceOrb.java
index c655689cdd2d4e655dfc872edd231dcdfb7c0995..7487d0d2c50fb86ceba3519269867e2488ee3bc5 100644
--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -150,7 +150,7 @@ public class ExperienceOrb extends Entity {
         } else {
             super.tick();
             boolean flag = !this.level().noCollision(this.getBoundingBox());
-            if (this.isEyeInFluid(FluidTags.WATER)) {
+            if (this.isEyeInWater()) { // Leaf - Optimize isEyeInFluid
                 this.setUnderwaterMovement();
             } else if (!flag) {
                 this.applyGravity();
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 7f2900001f12cd8f99bd75851e6539bd841ecee1..59054683591a1ff13f0a92ab7c0f4c9603916fef 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -470,7 +470,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
             }
 
-            if (this.isEyeInFluid(FluidTags.WATER)
+            if (this.isEyeInWater() // Leaf - Optimize isEyeInFluid
                 && !serverLevel1.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
                 boolean flag1 = !this.canBreatheUnderwater()
                     && !MobEffectUtil.hasWaterBreathing(this)
diff --git a/net/minecraft/world/entity/animal/equine/SkeletonHorse.java b/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
index 2502716b8efe4584fba9d9f9989ba9cba67568bc..d958741b7d218a4bc7e5a06beca308c7d6bb94ba 100644
--- a/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
+++ b/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
@@ -112,7 +112,7 @@ public class SkeletonHorse extends AbstractHorse {
 
     @Override
     public SoundEvent getAmbientSound() {
-        return this.isEyeInFluid(FluidTags.WATER) ? SoundEvents.SKELETON_HORSE_AMBIENT_WATER : SoundEvents.SKELETON_HORSE_AMBIENT;
+        return this.isEyeInWater() ? SoundEvents.SKELETON_HORSE_AMBIENT_WATER : SoundEvents.SKELETON_HORSE_AMBIENT; // Leaf - Optimize isEyeInFluid
     }
 
     @Override
diff --git a/net/minecraft/world/entity/animal/fish/AbstractFish.java b/net/minecraft/world/entity/animal/fish/AbstractFish.java
index 970904ce07ebc6aabc97166284cbb855d2976f09..3fd0b9e0c289650f6096ac857794accd05a370df 100644
--- a/net/minecraft/world/entity/animal/fish/AbstractFish.java
+++ b/net/minecraft/world/entity/animal/fish/AbstractFish.java
@@ -179,7 +179,7 @@ public abstract class AbstractFish extends WaterAnimal implements Bucketable {
 
         @Override
         public void vanillaTick() { // Purpur - Ridables
-            if (this.fish.isEyeInFluid(FluidTags.WATER)) {
+            if (this.fish.isEyeInWater()) { // Leaf - Optimize isEyeInFluid
                 this.fish.setDeltaMovement(this.fish.getDeltaMovement().add(0.0, 0.005, 0.0));
             }
 
diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index 433cd822f33b80096d34bef9675c24e022cf3a42..ee858baf44f000491f6a583bdb9ac3878ca69a58 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -390,7 +390,7 @@ public class Strider extends Animal implements ItemSteerable {
 
     @Override
     protected boolean canAddPassenger(Entity passenger) {
-        return !this.isVehicle() && !this.isEyeInFluid(FluidTags.LAVA);
+        return !this.isVehicle() && !this.isEyeInLava(); // Leaf - Optimize isEyeInFluid
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/Witch.java b/net/minecraft/world/entity/monster/Witch.java
index 97a427faa37ce9b0c558da40d582b16ac1e96da2..589169495c3d74660b5afa3f3b0be3e18f0d8e8e 100644
--- a/net/minecraft/world/entity/monster/Witch.java
+++ b/net/minecraft/world/entity/monster/Witch.java
@@ -179,7 +179,7 @@ public class Witch extends Raider implements RangedAttackMob {
                 }
             } else {
                 Holder<Potion> holder = null;
-                if (this.random.nextFloat() < 0.15F && this.isEyeInFluid(FluidTags.WATER) && !this.hasEffect(MobEffects.WATER_BREATHING)) {
+                if (this.random.nextFloat() < 0.15F && this.isEyeInWater() && !this.hasEffect(MobEffects.WATER_BREATHING)) { // Leaf - Optimize isEyeInFluid
                     holder = Potions.WATER_BREATHING;
                 } else if (this.random.nextFloat() < 0.15F
                     && (this.isOnFire() || this.getLastDamageSource() != null && this.getLastDamageSource().is(DamageTypeTags.IS_FIRE))
diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index 75d3449804655cdc3006b46643243a037cfe1963..823f3f360c53ec627df7fa46e591f490626d0bdd 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -294,7 +294,7 @@ public class Zombie extends Monster {
                     this.doUnderWaterConversion(serverLevel);
                 }
             } else if (this.convertsInWater()) {
-                if (this.isEyeInFluid(FluidTags.WATER)) {
+                if (this.isEyeInWater()) { // Leaf - Optimize isEyeInFluid
                     this.inWaterTime++;
                     if (this.inWaterTime >= 600) {
                         this.startUnderWaterConversion(300);
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 44c7f63d549d208decf1bb7fe7c9e751e3669192..223fb6395c9ebcea7a9ef31bbdf08c054dfd1aeb 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -330,7 +330,7 @@ public abstract class Player extends Avatar implements ContainerUser {
             this.lastItemInMainHand = mainHandItem.copy();
         }
 
-        if (!this.isEyeInFluid(FluidTags.WATER) && this.isEquipped(Items.TURTLE_HELMET)) {
+        if (!this.isEyeInWater() && this.isEquipped(Items.TURTLE_HELMET)) { // Leaf - Optimize isEyeInFluid
             this.turtleHelmetTick();
         }
 
@@ -370,8 +370,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     protected boolean updateIsUnderwater() {
-        this.wasUnderwater = this.isEyeInFluid(FluidTags.WATER);
-        return this.wasUnderwater;
+        return this.wasUnderwater = this.isEyeInWater(); // Leaf - Optimize isEyeInFluid
     }
 
     @Override
@@ -698,7 +697,7 @@ public abstract class Player extends Avatar implements ContainerUser {
         }
 
         destroySpeed *= (float)this.getAttributeValue(Attributes.BLOCK_BREAK_SPEED);
-        if (this.isEyeInFluid(FluidTags.WATER)) {
+        if (this.isEyeInWater()) { // Leaf - Optimize isEyeInFluid
             destroySpeed *= (float)this.getAttribute(Attributes.SUBMERGED_MINING_SPEED).getValue();
         }
 
diff --git a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
index e5854af03a58dc26a100feac59357862e46bfeed..33fb59123ce59b8fd5c9f84bc31e52d21ad060ff 100644
--- a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
@@ -787,7 +787,7 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
 
     @Override
     protected boolean canAddPassenger(Entity passenger) {
-        return this.getPassengers().size() < this.getMaxPassengers() && !this.isEyeInFluid(FluidTags.WATER);
+        return this.getPassengers().size() < this.getMaxPassengers() && !this.isEyeInWater(); // Leaf - Optimize isEyeInFluid
     }
 
     protected int getMaxPassengers() {
