From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Wed, 2 Apr 2025 23:03:22 +0200
Subject: [PATCH] Null handling on MultifaceSpreader

WHYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY

diff --git a/net/minecraft/world/level/block/MultifaceSpreader.java b/net/minecraft/world/level/block/MultifaceSpreader.java
index 5eb291396a83b8d98294c5f53d2e1f4915a0d84e..b5a78dc342bd2d2eecaaa3ab7cc89e391a91b2da 100644
--- a/net/minecraft/world/level/block/MultifaceSpreader.java
+++ b/net/minecraft/world/level/block/MultifaceSpreader.java
@@ -148,6 +148,14 @@ public class MultifaceSpreader {
         }
 
         default boolean placeBlock(LevelAccessor level, MultifaceSpreader.SpreadPos pos, BlockState state, boolean markForPostprocessing) {
+            // Leaf start - Null handling on MultifaceSpreader
+            // Check for null
+            if (pos.source() == null || pos.pos() == null) {
+                org.dreeam.leaf.config.LeafConfig.LOGGER.warn("Invalid SpreadPos with null source or position: {}", pos);
+                return false;
+            }
+            // Leaf end - Null handling on MultifaceSpreader
+
             BlockState stateForPlacement = this.getStateForPlacement(state, level, pos.pos(), pos.face());
             if (stateForPlacement != null) {
                 if (markForPostprocessing) {
