From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pascalpex <pascalpex@gmail.com>
Date: Wed, 4 Jun 2025 17:03:32 +0200
Subject: [PATCH] Add BlockExplosionHitEvent


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 87e899a988949813b1184ab79efb21aad0787847..5e37bd813ed989417bef5546c9c8a131aeb8a08f 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -602,7 +602,11 @@ public class ServerExplosion implements Explosion {
             }
             // CraftBukkit end
 
-            this.level.getBlockState(pos).onExplosionHit(this.level, pos, this, (stackx, position) -> addOrAppendStack(stacks, stackx, position));
+            // Leaf start - Add BlockExplosionHitEvent
+            if (org.dreeam.leaf.event.BlockExplosionHitEvent.getHandlerList().getRegisteredListeners().length == 0 || new org.dreeam.leaf.event.BlockExplosionHitEvent(org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, this.level.getWorld()).getBlock(), this.source == null ? null : this.source.getBukkitEntity(), org.bukkit.craftbukkit.CraftExplosionResult.toExplosionResult(this.blockInteraction)).callEvent()) {
+                this.level.getBlockState(pos).onExplosionHit(this.level, pos, this, (stackx, position) -> addOrAppendStack(stacks, stackx, position));
+            }
+            // Leaf end - Add BlockExplosionHitEvent
         }
 
         for (ServerExplosion.StackCollector stack : stacks) {
