From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 23 May 2025 12:01:42 +0900
Subject: [PATCH] Cache block state tags


diff --git a/net/minecraft/server/ReloadableServerResources.java b/net/minecraft/server/ReloadableServerResources.java
index d9cae732e2692eaa514ba9c9b8375b90253547a1..1e1f06fd4a1f2cd06ea13463aaefcdad8eab774c 100644
--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -110,5 +110,6 @@ public class ReloadableServerResources {
 
     public void updateStaticRegistryTags() {
         this.postponedTags.forEach(Registry.PendingTags::apply);
+        net.minecraft.world.level.block.Blocks.initPathType(); // Leaf - Cache block state tags
     }
 }
diff --git a/net/minecraft/world/entity/vehicle/DismountHelper.java b/net/minecraft/world/entity/vehicle/DismountHelper.java
index ae57c6f2d1c6f8b756efa77d97fca562b97aef35..0569fcce0de69ac35808989298491c4addf6a99b 100644
--- a/net/minecraft/world/entity/vehicle/DismountHelper.java
+++ b/net/minecraft/world/entity/vehicle/DismountHelper.java
@@ -54,7 +54,7 @@ public class DismountHelper {
 
     public static VoxelShape nonClimbableShape(BlockGetter level, BlockPos pos) {
         BlockState blockState = level.getBlockState(pos);
-        return !blockState.is(BlockTags.CLIMBABLE) && (!(blockState.getBlock() instanceof TrapDoorBlock) || !blockState.getValue(TrapDoorBlock.OPEN))
+        return ((blockState.tagFlag & (org.dreeam.leaf.util.BlockMasks.TRAP_DOOR_CL_OPEN | org.dreeam.leaf.util.BlockMasks.CLIMBABLE_TAG)) == 0) // Leaf - Cache block state tags
             ? blockState.getCollisionShape(level, pos)
             : Shapes.empty();
     }
diff --git a/net/minecraft/world/level/block/Blocks.java b/net/minecraft/world/level/block/Blocks.java
index 44e934a549be6d9169e900af617d24d65d328678..b9b0614b70e51cf34d0816b60ea0417eace04745 100644
--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -7301,4 +7301,14 @@ public class Blocks {
             }
         }
     }
+
+    // Leaf start - Cache block state tags
+    public static void initPathType() {
+        for (Block block : BuiltInRegistries.BLOCK) {
+            for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
+                blockState.initPathType();
+            }
+        }
+    }
+    // Leaf end - Cache block state tags
 }
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 47cff92e3244b556ee565b088003abd2784d4341..320df6752be17d81484a809ee511b06163327809 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -471,6 +471,8 @@ public abstract class BlockBehaviour implements FeatureElement {
         private boolean emptyCollisionShape;
         private boolean emptyConstantCollisionShape;
         private VoxelShape constantCollisionShape;
+        public byte pathType; // Leaf - Cache block state tags
+        public int tagFlag; // Leaf - Cache block state tags
 
         private static void initCaches(final VoxelShape shape, final boolean neighbours) {
             ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)shape).moonrise$isFullBlock();
@@ -637,6 +639,13 @@ public abstract class BlockBehaviour implements FeatureElement {
             // Paper end - optimise collisions
         }
 
+        // Leaf start - Cache block state tags
+        public void initPathType() {
+            pathType = (byte) net.minecraft.world.level.pathfinder.WalkNodeEvaluator.getPathTypeFromState(this.asState()).ordinal();
+            tagFlag = org.dreeam.leaf.util.BlockMasks.init(asState());
+        }
+        // Leaf end - Cache block state tags
+
         public Block getBlock() {
             return this.owner;
         }
diff --git a/net/minecraft/world/level/pathfinder/PathTypeCache.java b/net/minecraft/world/level/pathfinder/PathTypeCache.java
index 0d27cd18e5169852d4429d63a7a846cbb66c8147..a2401316d3baf832ba682dc16b2ea16ec9dbf2c7 100644
--- a/net/minecraft/world/level/pathfinder/PathTypeCache.java
+++ b/net/minecraft/world/level/pathfinder/PathTypeCache.java
@@ -23,7 +23,7 @@ public class PathTypeCache {
     }
 
     private PathType compute(BlockGetter level, BlockPos pos, int index, long packedPos) {
-        PathType pathTypeFromState = WalkNodeEvaluator.getPathTypeFromState(level, pos);
+        PathType pathTypeFromState = WalkNodeEvaluator.leafPathType(level, pos); // Leaf - Cache block state tags
         this.positions[index] = packedPos;
         this.pathTypes[index] = pathTypeFromState;
         return pathTypeFromState;
diff --git a/net/minecraft/world/level/pathfinder/PathfindingContext.java b/net/minecraft/world/level/pathfinder/PathfindingContext.java
index 57d60080810054ad791ac8537e105413576a666f..334afe90613ab5da12a9806b59ed4eb87c9eb09d 100644
--- a/net/minecraft/world/level/pathfinder/PathfindingContext.java
+++ b/net/minecraft/world/level/pathfinder/PathfindingContext.java
@@ -26,7 +26,7 @@ public class PathfindingContext {
 
     public PathType getPathTypeFromState(int x, int y, int z) {
         BlockPos blockPos = this.mutablePos.set(x, y, z);
-        return this.cache == null ? WalkNodeEvaluator.getPathTypeFromState(this.level, blockPos) : this.cache.getOrCompute(this.level, blockPos);
+        return this.cache == null ? WalkNodeEvaluator.leafPathType(this.level, blockPos) : this.cache.getOrCompute(this.level, blockPos); // Leaf - Cache block state tags
     }
 
     public BlockState getBlockState(BlockPos pos) {
diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index 876aef8dab7187a69bcb380692d0f6a583fb38bd..46130fd10265a2128abf1960f266a346c3a5bb27 100644
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -476,6 +476,17 @@ public class WalkNodeEvaluator extends NodeEvaluator {
         return pathType;
     }
 
+    // Leaf start - Cache block state tags
+    private static final PathType[] PATH_TYPES = PathType.values();
+    public static PathType leafPathType(BlockGetter level, BlockPos blockPos) {
+        BlockState blockState = level.getBlockStateIfLoaded(blockPos);
+        if (blockState == null) {
+            return PathType.BLOCKED;
+        }
+        return PATH_TYPES[blockState.pathType];
+    }
+    // Leaf end - Cache block state tags
+
     protected static PathType getPathTypeFromState(BlockGetter level, BlockPos pos) {
         // Paper start - Do not load chunks during pathfinding
         BlockState blockState = level.getBlockStateIfLoaded(pos);
@@ -483,6 +494,11 @@ public class WalkNodeEvaluator extends NodeEvaluator {
             return PathType.BLOCKED;
         }
         // Paper end
+        // Leaf start - Cache block state tags
+        return getPathTypeFromState(blockState);
+    }
+    public static PathType getPathTypeFromState(BlockState blockState) {
+        // Leaf end - Cache block state tags
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
