From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:03:31 +0900
Subject: [PATCH] Optimize SynchedEntityData#packDirty


diff --git a/net/minecraft/network/syncher/SynchedEntityData.java b/net/minecraft/network/syncher/SynchedEntityData.java
index 5d74c3e9938e5c6cc96f42b6df255e21aa925ed9..685f20282ac3dae127d614ee74041bb11d9c53aa 100644
--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -83,7 +83,15 @@ public class SynchedEntityData {
             return null;
         } else {
             this.isDirty = false;
-            List<SynchedEntityData.DataValue<?>> list = new ArrayList<>();
+            // Leaf start - Optimize SynchedEntityData#packDirty
+            int cap = 0;
+            for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
+                if (dataItem.isDirty()) {
+                    cap += 1;
+                }
+            }
+            ArrayList<SynchedEntityData.DataValue<?>> list = new ArrayList<>(cap);
+            // Leaf end - Optimize SynchedEntityData#packDirty
 
             for (SynchedEntityData.DataItem<?> dataItem : this.itemsById) {
                 if (dataItem.isDirty()) {
