From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 2 Jun 2025 03:29:20 +0800
Subject: [PATCH] Old Blast Protection explosion knockback


diff --git a/net/minecraft/world/entity/EquipmentSlot.java b/net/minecraft/world/entity/EquipmentSlot.java
index 63a9956f4aaf6feb6de2ef59a2211448f816e4fb..e5d67c834f214def5b04e75cc3decbcb05c6c2aa 100644
--- a/net/minecraft/world/entity/EquipmentSlot.java
+++ b/net/minecraft/world/entity/EquipmentSlot.java
@@ -22,6 +22,7 @@ public enum EquipmentSlot implements StringRepresentable {
     public static final int NO_COUNT_LIMIT = 0;
     public static final EquipmentSlot[] VALUES_ARRAY = values(); // Gale - JettPack - reduce array allocations
     public static final List<EquipmentSlot> VALUES = List.of(values());
+    public static final EquipmentSlot[] ARMOR_SLOTS = {EquipmentSlot.HEAD, EquipmentSlot.CHEST, EquipmentSlot.LEGS, EquipmentSlot.FEET}; // Leaf - Old Blast Protection explosion knockback
     public static final IntFunction<EquipmentSlot> BY_ID = ByIdMap.continuous(s -> s.id, values(), ByIdMap.OutOfBoundsStrategy.ZERO);
     public static final StringRepresentable.EnumCodec<EquipmentSlot> CODEC = StringRepresentable.fromEnum(EquipmentSlot::values);
     public static final StreamCodec<ByteBuf, EquipmentSlot> STREAM_CODEC = ByteBufCodecs.idMapper(BY_ID, s -> s.id);
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 5e37bd813ed989417bef5546c9c8a131aeb8a08f..deb52b739dc3d5ab6b45c94f958ce5aa31747f32 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -524,7 +524,7 @@ public class ServerExplosion implements Explosion {
                         double knockbackResistance = entity instanceof LivingEntity livingEntity
                             ? livingEntity.getAttributeValue(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE)
                             : 0.0;
-                        double knockbackPower = entity instanceof Player && this.level.paperConfig().environment.disableExplosionKnockback ? 0 : (1.0 - dist) * exposure * knockbackMultiplier * (1.0 - knockbackResistance); // Paper
+                        double knockbackPower = entity instanceof Player && this.level.paperConfig().environment.disableExplosionKnockback ? 0 : getExplosionKnockback(entity, (1.0 - dist) * exposure * knockbackMultiplier, knockbackResistance); // Paper // Leaf - Old Blast Protection explosion knockback
                         Vec3 knockback = direction.scale(knockbackPower);
                         // CraftBukkit start - Call EntityKnockbackEvent
                         if (entity instanceof LivingEntity) {
@@ -548,6 +548,54 @@ public class ServerExplosion implements Explosion {
         }
     }
 
+    // Leaf start - Old Blast Protection explosion knockback
+    private static double getExplosionKnockback(final Entity entity, final double velocity, final double knockbackResistance) {
+        if (!org.dreeam.leaf.config.modules.gameplay.Knockback.oldBlastProtectionKnockbackBehavior) {
+            return velocity * (1.0 - knockbackResistance);
+        }
+
+        // Only LivingEntity is affected by explosion knockback resistance attribute
+        // Others are `1.0-0=1` so the result is same as `velocity`
+        // Defined in `knockbackResistance` in `hurtEntities`
+        if (!(entity instanceof LivingEntity livingEntity)) {
+            return velocity;
+        }
+
+        // Old BLAST_PROTECTION logic
+        // BLAST_PROTECTION used ARMOR_SLOTS for slot types
+        // See 1.20.4's ProtectionEnchantment#getExplosionKnockbackAfterDampener,
+        // EnchantmentHelper#getEnchantmentLevel, Enchantment#getSlotItems,
+        // EnchantmentHelper#getItemEnchantmentLevel, Enchantments#BLAST_PROTECTION,
+        // these methods/fields for reference.
+        Map<net.minecraft.world.entity.EquipmentSlot, ItemStack> map = com.google.common.collect.Maps.newEnumMap(net.minecraft.world.entity.EquipmentSlot.class);
+
+        for (net.minecraft.world.entity.EquipmentSlot slot : net.minecraft.world.entity.EquipmentSlot.ARMOR_SLOTS) {
+            ItemStack itemStack = livingEntity.getItemBySlot(slot);
+            if (!itemStack.isEmpty()) {
+                map.put(slot, itemStack);
+            }
+        }
+
+        Iterable<ItemStack> items = map.values();
+        int i = 0;
+
+        if (items == null) return 0;
+
+        for (ItemStack itemStack : items) {
+            int enchantmentLevel = net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.BLAST_PROTECTION, itemStack);
+            if (enchantmentLevel > i) {
+                i = enchantmentLevel;
+            }
+        }
+
+        if (i > 0) {
+            return velocity * Mth.clamp(1.0 - (double) i * 0.15, 0.0, 1.0);
+        }
+
+        return velocity;
+    }
+    // Leaf end - Old Blast Protection explosion knockback
+
     private void interactWithBlocks(final List<BlockPos> targetBlocks) {
         List<ServerExplosion.StackCollector> stacks = new ArrayList<>();
         Util.shuffle(targetBlocks, this.level.random);
