From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 1 Aug 2026 08:56:01 +0900
Subject: [PATCH] cache biome zoom seed


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index e75e0fdc44f3ad3713f389e14fffc3f502028668..995b179ebfb5de617e4a800caa47f71ab3cf7aa5 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -717,8 +717,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         final ResourceKey<Level> dimensionKey = loading.info().dimensionKey();
         final SavedDataStorage savedDataStorage = new SavedDataStorage(this.storageSource.getDimensionPath(dimensionKey).resolve(LevelResource.DATA.id()), this.getFixerUpper(), this.registryAccess());
         savedDataStorage.set(WorldGenSettings.TYPE, new WorldGenSettings(worldDataAndGenSettings.genSettings().options(), worldDataAndGenSettings.genSettings().dimensions()));
-        long seed = worldOptions.seed();
-        long biomeZoomSeed = BiomeManager.obfuscateSeed(seed);
+        long biomeZoomSeed = worldOptions.biomeZoomSeed(); // Leaf - cache biome zoom seed
         List<CustomSpawner> overworldCustomSpawners = ImmutableList.of(
             new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(savedDataStorage) // Paper - save to world data
         );
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 94e45acbf67a3ff8139865d2befb9c89f54f10d8..e92af267d49c05eecf4b142e38d585a9c87ccbf9 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -3095,7 +3095,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         return new CommonPlayerSpawnInfo(
             level.dimensionTypeRegistration(),
             level.dimension(),
-            BiomeManager.obfuscateSeed(level.getSeed()),
+            level.worldGenSettings.options().biomeZoomSeed(), // Leaf - cache biome zoom seed
             this.gameMode.getGameModeForPlayer(),
             Optional.ofNullable(this.gameMode.getPreviousGameModeForPlayer()),
             level.isDebug(),
diff --git a/net/minecraft/server/level/WorldGenRegion.java b/net/minecraft/server/level/WorldGenRegion.java
index 98092b227c854d120cee27aca388e5a74f19383a..bbbab4415e687cf3f10ef916aafb18463c999c2c 100644
--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -131,7 +131,7 @@ public class WorldGenRegion implements WorldGenLevel {
         this.random = randomState.getOrCreateRandomFactory(WORLDGEN_REGION_RANDOM).at(this.center.getPos().getWorldPosition());
         this.dimensionType = level.dimensionType();
         this.uncachedBiomeResolver = level.uncachedBiomeResolver();
-        this.biomeManager = new BiomeManager(this, BiomeManager.obfuscateSeed(this.seed));
+        this.biomeManager = new BiomeManager(this, level.worldGenSettings.options().biomeZoomSeed()); // Leaf - cache biome zoom seed
         ChunkPos centerPos = center.getPos();
         this.centerChunkX = centerPos.x();
         this.centerChunkZ = centerPos.z();
diff --git a/net/minecraft/world/level/levelgen/WorldOptions.java b/net/minecraft/world/level/levelgen/WorldOptions.java
index b23dd531d284647e614f07f134c41e70593d5c04..4d58651cbaedb3035b7d8d8f367afc3f980da295 100644
--- a/net/minecraft/world/level/levelgen/WorldOptions.java
+++ b/net/minecraft/world/level/levelgen/WorldOptions.java
@@ -38,6 +38,7 @@ public class WorldOptions {
         : new WorldOptions("North Carolina".hashCode(), true, true);
     // Leaf end - Matter - Secure Seed
     private final long seed;
+    private final long biomeZoomSeed; // Leaf - cache biome zoom seed
     private long[] featureSeed = su.plo.matter.Globals.createRandomWorldSeed(); // Leaf - Matter - Secure Seed
     private final boolean generateStructures;
     private final boolean generateBonusChest;
@@ -79,8 +80,15 @@ public class WorldOptions {
         this.generateStructures = generateStructures;
         this.generateBonusChest = generateBonusChest;
         this.legacyCustomOptions = legacyCustomOptions;
+        this.biomeZoomSeed = net.minecraft.world.level.biome.BiomeManager.obfuscateSeed(seed); // Leaf - cache biome zoom seed
     }
 
+    // Leaf start - cache biome zoom seed
+    public long biomeZoomSeed() {
+        return this.biomeZoomSeed;
+    }
+    // Leaf end - cache biome zoom seed
+
     public long seed() {
         return this.seed;
     }
