From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 2 Jun 2025 03:29:20 +0800
Subject: [PATCH] Old Blast Protection explosion knockback


diff --git a/net/minecraft/world/entity/EquipmentSlot.java b/net/minecraft/world/entity/EquipmentSlot.java
index bf61513f45e19126e95ea7284446037bc019df07..cfb46db195951f23039881bf9be28f9512911132 100644
--- a/net/minecraft/world/entity/EquipmentSlot.java
+++ b/net/minecraft/world/entity/EquipmentSlot.java
@@ -23,6 +23,7 @@ public enum EquipmentSlot implements StringRepresentable {
     // Gale start - JettPack - reduce array allocations
     public static final EquipmentSlot[] VALUES_ARRAY = values();
     public static final List<EquipmentSlot> VALUES = List.of(VALUES_ARRAY);
+    public static final EquipmentSlot[] ARMOR_SLOTS = new EquipmentSlot[]{EquipmentSlot.HEAD, EquipmentSlot.CHEST, EquipmentSlot.LEGS, EquipmentSlot.FEET}; // Leaf - Old Blast Protection explosion knockback
     public static final IntFunction<EquipmentSlot> BY_ID = ByIdMap.continuous(equipmentSlot -> equipmentSlot.id, VALUES_ARRAY, ByIdMap.OutOfBoundsStrategy.ZERO);
     public static final StringRepresentable.EnumCodec<EquipmentSlot> CODEC = StringRepresentable.fromEnum(() -> VALUES_ARRAY);
     // Gale end - JettPack - reduce array allocations
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index af01b77a586258c492b542536c54eddaa2b2a6f0..c45d4646104565a98a4c9deee60ca5d799ee9cb7 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -531,7 +531,7 @@ public class ServerExplosion implements Explosion {
                         double d1 = entity instanceof LivingEntity livingEntity
                             ? livingEntity.getAttributeValue(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE)
                             : 0.0;
-                        double d2 = entity instanceof Player && this.level.paperConfig().environment.disableExplosionKnockback ? 0 : (1.0 - d) * f1 * knockbackMultiplier * (1.0 - d1); // Paper
+                        double d2 = entity instanceof Player && this.level.paperConfig().environment.disableExplosionKnockback ? 0 : getExplosionKnockback(entity, (1.0 - d) * f1 * knockbackMultiplier, d1); // Paper // Leaf - Old Blast Protection explosion knockback
                         Vec3 vec32 = vec31.scale(d2);
                         // CraftBukkit start - Call EntityKnockbackEvent
                         if (entity instanceof LivingEntity) {
@@ -555,6 +555,56 @@ public class ServerExplosion implements Explosion {
         }
     }
 
+    // Leaf start - Old Blast Protection explosion knockback
+    private static double getExplosionKnockback(Entity entity, double velocity, double knockbackResistance) {
+        if (!org.dreeam.leaf.config.modules.gameplay.Knockback.oldBlastProtectionKnockbackBehavior) {
+            return velocity * (1.0 - knockbackResistance);
+        }
+
+        // Only LivingEntity is affected by explosion knockback resistance attribute
+        // Others are `1.0-0=1` so the result is same as `velocity`
+        // Defined in `d1` in `hurtEntities`
+        if (!(entity instanceof LivingEntity livingEntity)) {
+            return velocity;
+        }
+
+        // Old BLAST_PROTECTION logic
+        // BLAST_PROTECTION used ARMOR_SLOTS for slot types
+        // See 1.20.4's ProtectionEnchantment#getExplosionKnockbackAfterDampener,
+        // EnchantmentHelper#getEnchantmentLevel, Enchantment#getSlotItems,
+        // EnchantmentHelper#getItemEnchantmentLevel, Enchantments#BLAST_PROTECTION,
+        // these methods/fields for reference.
+        Map<net.minecraft.world.entity.EquipmentSlot, ItemStack> map = com.google.common.collect.Maps.newEnumMap(net.minecraft.world.entity.EquipmentSlot.class);
+
+        for (net.minecraft.world.entity.EquipmentSlot slot : net.minecraft.world.entity.EquipmentSlot.ARMOR_SLOTS) {
+            ItemStack itemStack = livingEntity.getItemBySlot(slot);
+            if (!itemStack.isEmpty()) {
+                map.put(slot, itemStack);
+            }
+        }
+
+        Iterable<ItemStack> items = map.values();
+        int i = 0;
+
+        if (items == null) {
+            return 0;
+        }
+
+        for (ItemStack itemStack : items) {
+            int enchantmentLevel = net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.BLAST_PROTECTION, itemStack);
+            if (enchantmentLevel > i) {
+                i = enchantmentLevel;
+            }
+        }
+
+        if (i > 0) {
+            velocity *= Mth.clamp(1.0 - (double) i * 0.15, 0.0, 1.0);
+        }
+
+        return velocity;
+    }
+    // Leaf end - Old Blast Protection explosion knockback
+
     private void interactWithBlocks(List<BlockPos> blocks) {
         List<ServerExplosion.StackCollector> list = new ArrayList<>();
         Util.shuffle(blocks, this.level.random);
