From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:01:50 +0900
Subject: [PATCH] Optimize getScaledTrackingDistance


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 2139fc5c3c3e6b45e148d6f8798b9fbfcdef626c..ee20dd09bba40a4a98df501614d208552905c3cd 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -777,7 +777,13 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public int getScaledTrackingDistance(int trackingDistance) {
-        return this.getProperties().entityBroadcastRangePercentage * trackingDistance / 100;
+        // Leaf start - Optimize getScaledTrackingDistance
+        int p = this.getProperties().entityBroadcastRangePercentage;
+        if (p == 100) {
+            return trackingDistance;
+        }
+        return p * trackingDistance / 100;
+        // Leaf end - Optimize getScaledTrackingDistance
     }
 
     @Override
