From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 4 Jul 2025 03:13:33 +0900
Subject: [PATCH] optimize waypoint


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index d5839ab219f7b13b00bc4f00820c8d78ebbe74fc..6be0c272a66cd3575f466da9f5b495df6503ba4f 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5192,7 +5192,8 @@ public abstract class Entity
             int fx = Mth.floor(x);
             int fy = Mth.floor(y);
             int fz = Mth.floor(z);
-            if (fx != this.blockPosition.getX() || fy != this.blockPosition.getY() || fz != this.blockPosition.getZ()) {
+            boolean blockUpdated = fx != this.blockPosition.getX() || fy != this.blockPosition.getY() || fz != this.blockPosition.getZ(); // Leaf - optimize waypoint
+            if (blockUpdated) { // Leaf - optimize waypoint
                 this.blockPosition = new BlockPos(fx, fy, fz);
                 this.inBlockState = null;
                 if (SectionPos.blockToSectionCoord(fx) != this.chunkPosition.x() || SectionPos.blockToSectionCoord(fz) != this.chunkPosition.z()) {
@@ -5201,7 +5202,7 @@ public abstract class Entity
             }
 
             this.levelCallback.onMove();
-            if (!this.firstTick && this.level instanceof ServerLevel serverLevel && !this.isRemoved()) {
+            if ((!org.dreeam.leaf.config.modules.opt.OptimizeWaypoint.enabled || blockUpdated) && !this.firstTick && this.level instanceof ServerLevel serverLevel && !this.isRemoved()) { // Leaf - optimize waypoint
                 if (this instanceof WaypointTransmitter waypoint && waypoint.isTransmittingWaypoint()) {
                     serverLevel.getWaypointManager().updateWaypoint(waypoint);
                 }
