From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:46:30 +0200
Subject: [PATCH] Skip PreCreatureSpawnEvent if no listeners


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 9a8c8a0f361f5e76505e489a31079bee0d8a654a..f9c12fbb316a190880f5ed4b5531243b0e956368 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -543,6 +543,8 @@ public final class NaturalSpawner {
     ) {
         // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
                 org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
                 org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
@@ -553,6 +555,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = type.getCategory() != MobCategory.MISC
         // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
@@ -785,7 +789,9 @@ public final class NaturalSpawner {
     ) {
         // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
         final boolean hasPreSpawnEventListeners = com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0;
+        if (hasPreSpawnEventListeners) {
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
             org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
             org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
@@ -796,6 +802,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = type.getCategory() != MobCategory.MISC
             // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
