From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Replace EntitySelectorOptions map with optimized collection


diff --git a/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java b/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
index ad09c9ae2c06de3989e3740d37b3101c362f328a..cf1cf8a3d606ee28a0889c63fbe2e80b8ffd9ae0 100644
--- a/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
+++ b/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
@@ -336,7 +336,7 @@ public class EntitySelectorOptions {
             }, s -> true, Component.translatable("argument.entity.options.nbt.description"));
             register("scores", parser -> {
                 StringReader reader = parser.getReader();
-                Map<String, MinMaxBounds.Ints> expected = Maps.newHashMap();
+                it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<String, MinMaxBounds.Ints> expected = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // Leaf - Replace options map with optimized collection
                 reader.expect('{');
                 reader.skipWhitespace();
 
@@ -359,7 +359,7 @@ public class EntitySelectorOptions {
                     parser.addPredicate(entity -> {
                         Scoreboard scoreboard = entity.level().getServer().getScoreboard();
 
-                        for (Entry<String, MinMaxBounds.Ints> entry : expected.entrySet()) {
+                        for (Entry<String, MinMaxBounds.Ints> entry : expected.object2ObjectEntrySet()) { // Leaf - Replace options map with optimized collection
                             Objective objective = scoreboard.getObjective(entry.getKey());
                             if (objective == null) {
                                 return false;
