From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 4 Jul 2025 03:06:55 +0900
Subject: [PATCH] optimize no action time


diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index 81f1d4faebc38aa07a2c76c2c163584c8e5b624c..ccde7fb7da3698e4865ad12785a41113af3a8445 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -46,6 +46,12 @@ public abstract class Monster extends PathfinderMob implements Enemy {
     }
 
     protected void updateNoActionTime() {
+        // Leaf start - optimize no action time
+        if (org.dreeam.leaf.config.modules.opt.OptimizeNoActionTime.disableLightCheck) {
+            this.noActionTime++;
+            return;
+        }
+        // Leaf end - optimize no action time
         float br = this.getLightLevelDependentMagicValue();
         if (br > 0.5F) {
             this.noActionTime += 2;
