From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 4 Jul 2025 03:13:33 +0900
Subject: [PATCH] optimize waypoint


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 78efab6f31ca19e20b9847f4df7098029765f3d3..e3c66e5ae0e760cc03c7e976bd4f97df5eee8a1e 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5183,7 +5183,8 @@ public abstract class Entity
             int fx = Mth.floor(x);
             int fy = Mth.floor(y);
             int fz = Mth.floor(z);
-            if (fx != this.blockPosition.getX() || fy != this.blockPosition.getY() || fz != this.blockPosition.getZ()) {
+            boolean blockUpdated = fx != this.blockPosition.getX() || fy != this.blockPosition.getY() || fz != this.blockPosition.getZ(); // Leaf - optimize waypoint
+            if (blockUpdated) { // Leaf - optimize waypoint
                 this.blockPosition = new BlockPos(fx, fy, fz);
                 this.inBlockState = null;
                 if (SectionPos.blockToSectionCoord(fx) != this.chunkPosition.x() || SectionPos.blockToSectionCoord(fz) != this.chunkPosition.z()) {
@@ -5192,7 +5193,7 @@ public abstract class Entity
             }
 
             this.levelCallback.onMove();
-            if (!this.firstTick && this.level instanceof ServerLevel serverLevel && !this.isRemoved()) {
+            if ((!org.dreeam.leaf.config.modules.opt.OptimizeWaypoint.enabled || blockUpdated) && !this.firstTick && this.level instanceof ServerLevel serverLevel && !this.isRemoved()) { // Leaf - optimize waypoint
                 if (this instanceof WaypointTransmitter waypoint && waypoint.isTransmittingWaypoint()) {
                     serverLevel.getWaypointManager().updateWaypoint(waypoint);
                 }
