From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 16 Jul 2025 11:13:38 +0800
Subject: [PATCH] Re-route SetClosestHomeAsWalkTarget's poi finding to paper's
 faster logic

Also see findAnyPoiPositions in NearestBedSensor#doTick

diff --git a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
index 39ac9f6bb3badf32c09f268c4250fdec0f8161d1..4ba8f4502d258a72a2e6e2c62c31b59e5e8d4f49 100644
--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -85,11 +85,11 @@ public class SetClosestHomeAsWalkTarget {
                                 batchCache.put(key, lastUpdate + CACHE_TIMEOUT);
                                 return true;
                             };
-                            Set<Pair<Holder<PoiType>, BlockPos>> pois = poiManager.findAllWithType(
-                                    p -> p.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
-                                )
-                                .collect(Collectors.toSet());
-                            Path path = AcquirePoi.findPathToPois(body, pois);
+                            // Leaf start - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logic
+                            java.util.List<Pair<Holder<PoiType>, BlockPos>> pois = new java.util.ArrayList<>();
+                            ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiPositions(poiManager, p -> p.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY, false, Integer.MAX_VALUE, pois);
+                            Path path = AcquirePoi.findPathToPois(body, new java.util.HashSet<>(pois));
+                            // Leaf end - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logic
                             // Kaiiju start - petal - Async path processing
                             if (path instanceof org.dreeam.leaf.async.path.AsyncPath path1 && !path1.isProcessed()) {
                                 this.pending = path1;
