From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Replace EntitySelectorOptions map with optimized collection


diff --git a/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java b/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
index a244dae13c69dd72df1526692bfd8e54e6e042d7..e0dc44193b9e74140e3ff474603ae082328263d3 100644
--- a/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
+++ b/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
@@ -361,7 +361,7 @@ public class EntitySelectorOptions {
             }, parser -> true, Component.translatable("argument.entity.options.nbt.description"));
             register("scores", parser -> {
                 StringReader reader = parser.getReader();
-                Map<String, MinMaxBounds.Ints> map = Maps.newHashMap();
+                it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<String, MinMaxBounds.Ints> map = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // Leaf - Replace options map with optimized collection
                 reader.expect('{');
                 reader.skipWhitespace();
 
@@ -384,7 +384,7 @@ public class EntitySelectorOptions {
                     parser.addPredicate(entity -> {
                         Scoreboard scoreboard = entity.level().getServer().getScoreboard();
 
-                        for (Entry<String, MinMaxBounds.Ints> entry : map.entrySet()) {
+                        for (Entry<String, MinMaxBounds.Ints> entry : map.object2ObjectEntrySet()) { // Leaf - Replace options map with optimized collection
                             Objective objective = scoreboard.getObjective(entry.getKey());
                             if (objective == null) {
                                 return false;
