From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pascalpex <pascalpex@gmail.com>
Date: Wed, 4 Jun 2025 17:03:32 +0200
Subject: [PATCH] Add BlockExplosionHitEvent


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index bcbf7daa2c2b7941ab7cb333de15494969ca665c..af01b77a586258c492b542536c54eddaa2b2a6f0 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -609,9 +609,13 @@ public class ServerExplosion implements Explosion {
             }
             // CraftBukkit end
 
-            this.level
-                .getBlockState(blockPos)
-                .onExplosionHit(this.level, blockPos, this, (itemStack, blockPos1) -> addOrAppendStack(list, itemStack, blockPos1));
+            // Leaf start - Add BlockExplosionHitEvent
+            if (new org.dreeam.leaf.event.BlockExplosionHitEvent(CraftLocation.toBukkit(blockPos, this.level.getWorld()).getBlock(), this.source == null ? null : this.source.getBukkitEntity(), org.bukkit.craftbukkit.CraftExplosionResult.toExplosionResult(this.blockInteraction)).callEvent()) {
+                this.level
+                    .getBlockState(blockPos)
+                    .onExplosionHit(this.level, blockPos, this, (itemStack, blockPos1) -> addOrAppendStack(list, itemStack, blockPos1));
+            }
+            // Leaf end - Add BlockExplosionHitEvent
         }
 
         for (ServerExplosion.StackCollector stackCollector : list) {
