From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix Paper config fixClimbingBypassingCrammingRule

Waiting for Paper#12793

diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 59054683591a1ff13f0a92ab7c0f4c9603916fef..b37f6e54d37be405544362e96856be30f1562108 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4055,7 +4055,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         // Leaf start - Only player pushable
         final AABB box = this.getBoundingBox();
         List<Entity> pushableEntities = org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled
-            ? getNearbyPushablePlayers(this, box, EntitySelector.pushableBy(this))
+            ? getNearbyPushablePlayers(this, box, EntitySelector.pushable(this, this.level().paperConfig().collisions.fixClimbingBypassingCrammingRule)) // Leaf - Fix Paper config fixClimbingBypassingCrammingRule
             : this.level().getPushableEntities(this, box);
         // Leaf end - Only player pushable
         if (!pushableEntities.isEmpty()) {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 75165dff8bb8ccfd3cf14d890892df0bded4a1cc..0c6ded7fc7fb60e789a3c0a826f56b099b33935e 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1885,7 +1885,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
 
     public List<Entity> getPushableEntities(Entity entity, AABB boundingBox) {
-        return this.getEntities(entity, boundingBox, EntitySelector.pushableBy(entity));
+        return this.getEntities(entity, boundingBox, EntitySelector.pushable(entity, this.paperConfig().collisions.fixClimbingBypassingCrammingRule)); // Leaf - Fix Paper config fixClimbingBypassingCrammingRule
     }
 
     public abstract @Nullable Entity getEntity(int id);
