From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:12:14 +0900
Subject: [PATCH] optimize fluid state access


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 000d403ec41a64dccaeda194f606909a838b5a06..96204ee26dd7e65ed4df6392088fdba53058fb94 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1408,6 +1408,13 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return chunk.getFluidState(pos);
     }
 
+    // Leaf start - optimize fluid state access
+    public final @Nullable FluidState getFluidStateIfLoadedUnchecked(final int x, final int y, final int z) {
+        LevelChunk chunkAt = ((ServerLevel) this).getChunkSource().fullChunksNonSync.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x >> 4, z >> 4));
+        return chunkAt == null ? null : chunkAt.getFluidStateFinal(x, y, z);
+    }
+    // Leaf end - optimize fluid state access
+
     public boolean isBrightOutside() {
         return !this.dimensionType().hasFixedTime() && this.skyDarken < 4;
     }
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 560f3db2cf65a23548565be509184b34c914f82e..da7ea6618f051b79f22e7be93228f74aa2a0f47c 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -338,6 +338,21 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         // Leaf end - optimize LevelChunk#getBlockStateFinal
     }
 
+    // Leaf start - optimize fluid state access
+    public FluidState getFluidStateFinal(final int x, final int y, final int z) {
+        final int sectionIndex = (y >> 4) - this.minSection;
+        if (sectionIndex < 0 || sectionIndex >= this.sections.length) {
+            return AIR_FLUIDSTATE;
+        } else {
+            LevelChunkSection section = this.sections[sectionIndex];
+            if (section.fluidCount <= 0) {
+                return AIR_FLUIDSTATE;
+            }
+            return section.getStates().get((y & 15) << 8 | (z & 15) << 4 | x & 15).getFluidState();
+        }
+    }
+    // Leaf end - optimize fluid state access
+
     @Override
     public BlockState getBlockState(final BlockPos pos) {
         if (true) {
diff --git a/net/minecraft/world/level/chunk/LevelChunkSection.java b/net/minecraft/world/level/chunk/LevelChunkSection.java
index dc0270de6d13b7b6fc5de463fab05e9cca842480..c860b463b959f04545f4aae2a1787000e0353b86 100644
--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -11,7 +11,7 @@ import net.minecraft.world.level.material.FluidState;
 public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_counting.BlockCountingChunkSection { // Paper - block counting
     public static final int BIOME_CONTAINER_BITS = 2;
     short nonEmptyBlockCount; // Paper - package private
-    private short fluidCount;
+    public short fluidCount; // Leaf - optimize fluid state access - private -> public
     private short tickingBlockCount;
     private short tickingFluidCount;
     private final PalettedContainer<BlockState> states;
