From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 16 Jul 2025 11:13:38 +0800
Subject: [PATCH] Re-route SetClosestHomeAsWalkTarget's poi finding to paper's
 faster logic

Also see findAnyPoiPositions in NearestBedSensor#doTick

diff --git a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
index eb56a126091735444faefeb0b623da2cfa8f74de..d398dcd8a7b27d5cb7afb19699cc4282617b5ece 100644
--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -55,14 +55,14 @@ public class SetClosestHomeAsWalkTarget {
                                         return true;
                                     }
                                 };
-                                Set<Pair<Holder<PoiType>, BlockPos>> set = poiManager.findAllWithType(
-                                        poi -> poi.is(PoiTypes.HOME), predicate, mob.blockPosition(), 48, PoiManager.Occupancy.ANY
-                                    )
-                                    .collect(Collectors.toSet());
-                                // Kaiiju start - petal - async path processing
+                                // Leaf start - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logic
+                                java.util.List<Pair<Holder<PoiType>, BlockPos>> poiposes = new java.util.ArrayList<>();
+                                io.papermc.paper.util.PoiAccess.findAnyPoiPositions(poiManager, poi -> poi.is(PoiTypes.HOME), predicate, mob.blockPosition(), 48, PoiManager.Occupancy.ANY, false, Integer.MAX_VALUE, poiposes);
+                                // Leaf end - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logic
+                                // Kaiiju start - petal - Async path processing
                                 if (org.dreeam.leaf.config.modules.async.AsyncPathfinding.enabled) {
                                     // await on path async
-                                    Path possiblePath = AcquirePoi.findPathToPois(mob, set);
+                                    Path possiblePath = AcquirePoi.findPathToPois(mob, new java.util.HashSet<>(poiposes)); // Leaf - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logic
 
                                     // wait on the path to be processed
                                     org.dreeam.leaf.async.path.AsyncPathProcessor.awaitProcessing(possiblePath, path -> {
@@ -70,7 +70,7 @@ public class SetClosestHomeAsWalkTarget {
                                     });
                                 } else {
                                     // Kaiiju end
-                                    Path path = AcquirePoi.findPathToPois(mob, set);
+                                    Path path = AcquirePoi.findPathToPois(mob, new java.util.HashSet<>(poiposes)); // Leaf - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logi
                                     processPath(speedModifier, map, mutableLong, walkTarget, level, poiManager, mutableInt, path);
                                 } // Kaiiju - async path processing
 
