From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 16 Jul 2025 11:13:38 +0800
Subject: [PATCH] Re-route SetClosestHomeAsWalkTarget's poi finding to paper's
 faster logic

Also see findAnyPoiPositions in NearestBedSensor#doTick

diff --git a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
index f018e09c9b97908dee140016f079fde095bb94a8..2f8848a020ac9a280d69086a7f301b8bfcc5cd0a 100644
--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -78,11 +78,10 @@ public class SetClosestHomeAsWalkTarget {
                                         return true;
                                     }
                                 };
-                                Set<Pair<Holder<PoiType>, BlockPos>> set = poiManager.findAllWithType(
-                                        poi -> poi.is(PoiTypes.HOME), predicate, mob.blockPosition(), 48, PoiManager.Occupancy.ANY
-                                    )
-                                    .collect(Collectors.toSet());
-                                Path path = AcquirePoi.findPathToPois(mob, set);
+                                // Leaf start - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logic
+                                java.util.List<Pair<Holder<PoiType>, BlockPos>> poiposes = new java.util.ArrayList<>();
+                                io.papermc.paper.util.PoiAccess.findAnyPoiPositions(poiManager, poi -> poi.is(PoiTypes.HOME), predicate, mob.blockPosition(), 48, PoiManager.Occupancy.ANY, false, Integer.MAX_VALUE, poiposes);
+                                Path path = AcquirePoi.findPathToPois(mob, new java.util.HashSet<>(poiposes));
                                 // Leaf end - Re-route SetClosestHomeAsWalkTarget's poi finding to paper's faster logic
                                 // Kaiiju start - petal - Async path processing
                                 if (path != null && !path.isProcessed()) {
