From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 4 Jul 2025 03:06:55 +0900
Subject: [PATCH] optimize no action time


diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index 54033572e325f8486a438e31dde34f8407f9180a..bee0c24558e1c4c2765231c77805a969323fb13d 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -47,6 +47,12 @@ public abstract class Monster extends PathfinderMob implements Enemy {
     }
 
     protected void updateNoActionTime() {
+        // Leaf start - optimize no action time
+        if (org.dreeam.leaf.config.modules.opt.OptimizeNoActionTime.disableLightCheck) {
+            this.noActionTime++;
+            return;
+        }
+        // Leaf end - optimize no action time
         float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue();
         if (lightLevelDependentMagicValue > 0.5F) {
             this.noActionTime += 2;
