From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 10 May 2025 22:04:42 +0200
Subject: [PATCH] Smart sort items in NearestItemSensor


diff --git a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
index 09fd13e2d958da8326276c4dadf25bf488aff5ac..9bf3e8401ac2df3b116ab5d91a7a1cace5dd7d7d 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -24,11 +24,18 @@ public class NearestItemSensor extends Sensor<Mob> {
     @Override
     protected void doTick(ServerLevel level, Mob entity) {
         Brain<?> brain = entity.getBrain();
-        List<ItemEntity> entitiesOfClass = level.getEntitiesOfClass(ItemEntity.class, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), itemEntity -> itemEntity.closerThan(entity, MAX_DISTANCE_TO_WANTED_ITEM) && entity.wantsToPickUp(level, itemEntity.getItem())); // Paper - Perf: Move predicate into getEntities
-        entitiesOfClass.sort(Comparator.comparingDouble(entity::distanceToSqr));
+        // Leaf start - Smart sort items in NearestItemSensor
+        // List<ItemEntity> entitiesOfClass = level.getEntitiesOfClass(ItemEntity.class, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), itemEntity -> itemEntity.closerThan(entity, MAX_DISTANCE_TO_WANTED_ITEM) && entity.wantsToPickUp(level, itemEntity.getItem())); // Paper - Perf: Move predicate into getEntities
+        var pos = entity.position();
+        double entityX = pos.x();
+        double entityY = pos.y();
+        double entityZ = pos.z();
+        level.moonrise$getEntityLookup().getEntities(ItemEntity.class, null, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), org.dreeam.leaf.util.FastBitRadixSort.unsafeBuffer(), itemEntity -> itemEntity.position().distanceToSqr(entityX, entityY, entityZ) < (double) MAX_DISTANCE_TO_WANTED_ITEM * (double) MAX_DISTANCE_TO_WANTED_ITEM && entity.wantsToPickUp(level, itemEntity.getItem())); // Leaf - Smart sort items in NearestItemSensor
+        ItemEntity[] sortedItems = org.dreeam.leaf.util.FastBitRadixSort.unsafeSort(new ItemEntity[0], entity, ItemEntity.class); // Leaf - Smart sort items in NearestItemSensor
+        // Leaf end - Smart sort items in NearestItemSensor
         // Paper start - Perf: remove streams from hot code
         ItemEntity nearest = null;
-        for (final ItemEntity itemEntity : entitiesOfClass) {
+        for (final ItemEntity itemEntity : sortedItems) { // Leaf - Smart sort items in NearestItemSensor
             if (entity.hasLineOfSight(itemEntity)) { // Paper - Perf: Move predicate into getEntities
                 nearest = itemEntity;
                 break;
