From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 14:52:44 +0900
Subject: [PATCH] thread unsafe chunk map


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index dc3d0f0cefde5a9c0d47ccdcfaa5ea80ee03aa27..3acc875c7e5e70874f121d25fb5521589578b919 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -81,6 +81,7 @@ public final class ChunkHolderManager {
     final ChunkUnloadQueue unloadQueue;
 
     private final ConcurrentChainedLong2ReferenceHashTable<NewChunkHolder> chunkHolders = ConcurrentChainedLong2ReferenceHashTable.createWithCapacity(16384, 0.25f);
+    public final org.dreeam.leaf.world.ChunkCache<NewChunkHolder> chunkHoldersNoSync; // Leaf - thread unsafe chunk map
     private final ServerLevel world;
     private final ChunkTaskScheduler taskScheduler;
     private long currentTick;
@@ -116,6 +117,7 @@ public final class ChunkHolderManager {
         this.taskScheduler = taskScheduler;
         this.ticketLockArea = new ReentrantAreaLock(taskScheduler.getChunkSystemLockShift());
         this.unloadQueue = new ChunkUnloadQueue(((ChunkSystemServerLevel)world).moonrise$getRegionChunkShift());
+        this.chunkHoldersNoSync = new org.dreeam.leaf.world.ChunkCache<>(world.getChunkSource().mainThread); // Leaf - thread unsafe chunk map
     }
 
     public boolean processTicketUpdates(final int chunkX, final int chunkZ) {
@@ -383,6 +385,7 @@ public final class ChunkHolderManager {
                 if (current == null) {
                     // must create
                     current = ChunkHolderManager.this.createChunkHolder(key);
+                    ChunkHolderManager.this.chunkHoldersNoSync.put(key, current); // Leaf - thread unsafe chunk map
                     ChunkHolderManager.this.chunkHolders.put(key, current);
                     current.updateTicketLevel(newLevel);
                 } else {
@@ -911,11 +914,24 @@ public final class ChunkHolderManager {
     }
 
     public NewChunkHolder getChunkHolder(final int chunkX, final int chunkZ) {
-        return this.chunkHolders.get(CoordinateUtils.getChunkKey(chunkX, chunkZ));
+        // Leaf start - thread unsafe chunk map
+        final long position = CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        if (this.chunkHoldersNoSync.isSameThread()) {
+            return this.chunkHoldersNoSync.get(position);
+        } else {
+            return this.chunkHolders.get(position);
+        }
+        // Leaf end - thread unsafe chunk map
     }
 
     public NewChunkHolder getChunkHolder(final long position) {
-        return this.chunkHolders.get(position);
+        // Leaf start - thread unsafe chunk map
+        if (this.chunkHoldersNoSync.isSameThread()) {
+            return this.chunkHoldersNoSync.get(position);
+        } else {
+            return this.chunkHolders.get(position);
+        }
+        // Leaf end - thread unsafe chunk map
     }
 
     public void raisePriority(final int x, final int z, final Priority priority) {
@@ -974,6 +990,7 @@ public final class ChunkHolderManager {
         }
 
         current = this.createChunkHolder(position);
+        this.chunkHoldersNoSync.put(position, current); // Leaf - thread unsafe chunk map
         this.chunkHolders.put(position, current);
 
 
@@ -1147,7 +1164,11 @@ public final class ChunkHolderManager {
         this.autoSaveQueue.remove(holder);
         ((ChunkSystemServerChunkCache)this.world.getChunkSource()).moonrise$chunkHolderUnload(holder);
         PlatformHooks.get().onChunkHolderDelete(this.world, holder.vanillaChunkHolder);
-        this.chunkHolders.remove(CoordinateUtils.getChunkKey(holder.chunkX, holder.chunkZ));
+        // Leaf start - thread unsafe chunk map
+        long position = CoordinateUtils.getChunkKey(holder.chunkX, holder.chunkZ);
+        this.chunkHoldersNoSync.remove(position);
+        this.chunkHolders.remove(position);
+        // Leaf end - thread unsafe chunk map
     }
 
     // note: never call while inside the chunk system, this will absolutely break everything
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 102d15435c150d973032af2b2ec83361f17f15cd..2e18112a07ee3a9fc5ded8cd28c91ad7cb64ad3f 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1926,6 +1926,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 serverLevelTickingSemaphore.acquire();
                 tasks.add(
                     level.tickExecutor.submit(() -> {
+                        level.getChunkSource().fullChunksNonSync.setThread(); // Leaf - thread unsafe chunk map
+                        level.moonrise$getChunkTaskScheduler().chunkHolderManager.chunkHoldersNoSync.setThread(); // Leaf - thread unsafe chunk map
                         ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread currentThread = (ca.spottedleaf.moonrise.common.util.TickThread.ServerLevelTickThread) Thread.currentThread();
                         currentThread.currentTickingServerLevel = level;
                         try {
@@ -1944,7 +1946,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             level.explosionDensityCache.clear(); // Paper - Optimize explosions
         }
             while (!tasks.isEmpty()) {
-                tasks.poll().get();
+                final ServerLevel serverLevel = tasks.poll().get(); // Leaf - thread unsafe chunk map
+                serverLevel.getChunkSource().fullChunksNonSync.setThread(); // Leaf - thread unsafe chunk map
+                serverLevel.moonrise$getChunkTaskScheduler().chunkHolderManager.chunkHoldersNoSync.setThread(); // Leaf - thread unsafe chunk map
             }
         } catch (java.lang.InterruptedException | java.util.concurrent.ExecutionException e) {
             throw new RuntimeException(e); // Propagate exception
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 71a2b2f20993d05d336b68fc13c7af03266339aa..75bed9bea1279a6484b7bba64cfeee88a703e253 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -79,6 +79,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     private NaturalSpawner.@Nullable AsyncSpawnCostSampler asyncSpawnCostSampler; // Leaf - Pufferfish - async mob spawning
     // Paper start
     public final ca.spottedleaf.concurrentutil.map.concurrent.longs.ConcurrentChainedLong2ReferenceHashTable<net.minecraft.world.level.chunk.LevelChunk> fullChunks = new ca.spottedleaf.concurrentutil.map.concurrent.longs.ConcurrentChainedLong2ReferenceHashTable<>();
+    public final org.dreeam.leaf.world.ChunkCache<LevelChunk> fullChunksNonSync; // Leaf - thread unsafe chunk map
     public int getFullChunksCount() {
         return this.fullChunks.size();
     }
@@ -89,16 +90,27 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     @Override
     public final void moonrise$setFullChunk(final int chunkX, final int chunkZ, final LevelChunk chunk) {
         final long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        // Leaf start - thread unsafe chunk map
         if (chunk == null) {
+            this.fullChunksNonSync.remove(key);
             this.fullChunks.remove(key);
         } else {
+            this.fullChunksNonSync.put(key,chunk);
             this.fullChunks.put(key, chunk);
         }
+        // Leaf end - thread unsafe chunk map
     }
 
     @Override
     public final LevelChunk moonrise$getFullChunkIfLoaded(final int chunkX, final int chunkZ) {
-        return this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
+        // Leaf start - thread unsafe chunk map
+        long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        if (this.fullChunksNonSync.isSameThread()) {
+            return this.fullChunksNonSync.get(key);
+        } else {
+            return this.fullChunks.get(key);
+        }
+        // Leaf end - thread unsafe chunk map
     }
 
     @Override
@@ -207,6 +219,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         this.mainThreadProcessor = new ServerChunkCache.MainThreadExecutor(level);
         this.mainThread = Thread.currentThread();
         Path dataFolder = levelStorage.getDimensionPath(level.dimension()).resolve("data");
+        this.fullChunksNonSync = new org.dreeam.leaf.world.ChunkCache<>(mainThread); // Leaf - thread unsafe chunk map
 
         try {
             FileUtil.createDirectoriesSafe(dataFolder);
@@ -284,7 +297,14 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
     @Nullable
     public LevelChunk getChunkAtIfLoadedImmediately(int x, int z) {
-        return this.fullChunks.get(ChunkPos.pack(x, z));
+        // Leaf start - thread unsafe chunk map
+        long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z);
+        if (this.fullChunksNonSync.isSameThread()) {
+            return this.fullChunksNonSync.get(key);
+        } else {
+            return this.fullChunks.get(key);
+        }
+        // Leaf end - thread unsafe chunk map
     }
     // Paper end
 
@@ -313,7 +333,8 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     public @Nullable ChunkAccess getChunk(final int x, final int z, final ChunkStatus targetStatus, final boolean loadOrGenerate) {
         // Paper start - rewrite chunk system
         if (targetStatus == ChunkStatus.FULL) {
-            final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
+            long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z); // Leaf - thread unsafe chunk map
+            final LevelChunk ret = this.fullChunksNonSync.isSameThread() ? this.fullChunksNonSync.get(key) : this.fullChunks.get(key); // Leaf - thread unsafe chunk map
 
             if (ret != null) {
                 return ret;
@@ -329,7 +350,8 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     @Override
     public @Nullable LevelChunk getChunkNow(final int x, final int z) {
         // Paper start - rewrite chunk system
-        final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
+        long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z); // Leaf - thread unsafe chunk map
+        final LevelChunk ret = this.fullChunksNonSync.isSameThread() ? this.fullChunksNonSync.get(key) : this.fullChunks.get(key); // Leaf - thread unsafe chunk map
         if (!ca.spottedleaf.moonrise.common.PlatformHooks.get().hasCurrentlyLoadingChunk()) {
             return ret;
         }
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 0ac0d93df310f89546f6367325613bd0b6c65743..33fd6d86a1b920614ca56d00213a61bf532d4fa2 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1040,6 +1040,16 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         // Paper end - Perf: make sure loaded chunks get the inlined variant of this function
     }
 
+    // Leaf start - thread unsafe chunk map
+    public final @Nullable LevelChunk getChunkAtIfLoadedUnchecked(final int x, final int z) {
+        return ((ServerLevel) this).getChunkSource().fullChunksNonSync.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
+    }
+
+    public final @Nullable BlockState getBlockStateIfLoadedUnchecked(final int x, final int y, final int z) {
+        LevelChunk chunk = ((ServerLevel) this).getChunkSource().fullChunksNonSync.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x >> 4, z >> 4));
+        return chunk == null ? null : chunk.getBlockStateFinal(x, y, z);
+    }
+    // Leaf end - thread unsafe chunk map
     // Paper start - if loaded
     @Nullable
     @Override
