From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:30:37 +0900
Subject: [PATCH] optimize tickEffects


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 521ac5d20b740aa6c52c1e31326674fbf5e9a826..46518efbaaf5366155effcfce2c8c2595662d80b 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -962,6 +962,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     protected void tickEffects() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            if (!this.activeEffects.isEmpty()) { // Leaf - optimize tickEffects
             Iterator<Holder<MobEffect>> iterator = this.activeEffects.keySet().iterator();
 
             this.tickingEffects = true; // CraftBukkit
@@ -986,6 +987,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
             // CraftBukkit start
             this.tickingEffects = false;
+            } // Leaf - optimize tickEffects
             Runnable task;
             while ((task = this.deferredEffectUpdates.poll()) != null) {
                 task.run();
