From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:46:30 +0200
Subject: [PATCH] Skip PreCreatureSpawnEvent if no listeners


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index f38edb1b5773daa79a9e9517d2338c2e510f6f3d..324b94bb5ee3f8c41e9c2fd1079a5b2c55a8b77f 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -496,6 +496,8 @@ public final class NaturalSpawner {
     ) {
         // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
                 org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
                 org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
@@ -506,6 +508,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = type.getCategory() != MobCategory.MISC
         // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
@@ -781,6 +785,8 @@ public final class NaturalSpawner {
     ) {
         // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
             org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
             org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
@@ -791,6 +797,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = type.getCategory() != MobCategory.MISC
             // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
