From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 14:55:37 +0900
Subject: [PATCH] optimize get chunk


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 7b5b35c206798a0ac904f7a7389a0468018dd11c..a1d43377cbc6a37fe3199ecf886b511c879b9904 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1395,12 +1395,17 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             }
         }
         // CraftBukkit end
+        // Leaf start - optimize get chunk
+        int x = pos.getX();
+        int y = pos.getY();
+        int z = pos.getZ();
         if (!this.isInValidBounds(pos)) {
             return Blocks.VOID_AIR.defaultBlockState();
         }
 
-        ChunkAccess chunk = this.getChunk(pos.getX() >> 4, pos.getZ() >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
-        return chunk.getBlockState(pos);
+        ChunkAccess chunk = this.getChunk(x >> 4, z >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
+        return chunk.getBlockState(x, y, z);
+        // Leaf end - optimize get chunk
     }
 
     @Override
diff --git a/net/minecraft/world/level/block/SpreadingSnowyBlock.java b/net/minecraft/world/level/block/SpreadingSnowyBlock.java
index 513e23b693aa0f47bcf91e9087159a75471efd36..96abb55ee4de5f17423f63865b9acef8fb8313c1 100644
--- a/net/minecraft/world/level/block/SpreadingSnowyBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyBlock.java
@@ -64,7 +64,7 @@ public abstract class SpreadingSnowyBlock extends SnowyBlock {
         Optional<Block> baseBlock = blocks.getOptional(this.baseBlock);
         if (!baseBlock.isEmpty()) {
             // Paper start - Perf: optimize dirt and snow spreading
-            final net.minecraft.world.level.chunk.ChunkAccess cachedChunk = level.getChunkIfLoaded(pos);
+            final net.minecraft.world.level.chunk.LevelChunk cachedChunk = level.getChunkAtIfLoadedUnchecked(pos.getX() >> 4, pos.getZ() >> 4); // Leaf - optimize get chunk
             if (cachedChunk == null) { // Is this needed?
                 return;
             }
