From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:31:35 +0900
Subject: [PATCH] Only update frozen ticks if changed


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 46518efbaaf5366155effcfce2c8c2595662d80b..2f4eefa597e7a593f2fa85b521fee23c35db1d32 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3909,7 +3909,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         if (this.level() instanceof ServerLevel serverLevel) {
             profiler.push("freezing");
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // Leaf start - Only update frozen ticks if changed
+                int ticksFrozen = this.getTicksFrozen();
+                int newTicksFrozen = Math.max(0, ticksFrozen - 2);
+                if (ticksFrozen != newTicksFrozen) {
+                    this.setTicksFrozen(newTicksFrozen);
+                }
+                // Leaf end - Only update frozen ticks if changed
             }
 
             this.removeFrost();
