From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:46:30 +0200
Subject: [PATCH] Skip PreCreatureSpawnEvent if no listeners


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index bd33bf6f83bb8f8fd2fad70718622559806bb0d1..65bd78e70b3533d5a62744f6d0a0bedcec24a61c 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -499,6 +499,8 @@ public final class NaturalSpawner {
     ) {
         // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
                 org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
                 org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
@@ -509,6 +511,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = type.getCategory() != MobCategory.MISC
         // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
@@ -784,6 +788,8 @@ public final class NaturalSpawner {
     ) {
         // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
             org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
             org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
@@ -794,6 +800,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = type.getCategory() != MobCategory.MISC
             // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
