From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 9 Oct 2025 18:32:43 +0900
Subject: [PATCH] optimize canHoldAnyFluid


diff --git a/net/minecraft/world/level/material/FlowingFluid.java b/net/minecraft/world/level/material/FlowingFluid.java
index eed24d905e24b4107352ed3e1392c63ad6fb44e3..e2b6299f2b33513b46debb184c4de090b478c551 100644
--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -494,7 +494,7 @@ public abstract class FlowingFluid extends Fluid {
         final FluidState testFluidState
     ) {
         return !this.isSourceBlockOfThisType(testFluidState)
-            && canHoldAnyFluid(testState)
+            && (testState.tagFlag & org.dreeam.leaf.util.BlockMasks.CAN_HOLD_ANY_FLUID) != 0 // Leaf - optimize canHoldAnyFluid
             && canPassThroughWall(direction, level, sourcePos, sourceState, testPos, testState);
     }
 
@@ -566,7 +566,7 @@ public abstract class FlowingFluid extends Fluid {
     }
 
     private static boolean canHoldFluid(final BlockGetter level, final BlockPos pos, final BlockState state, final Fluid newFluid) {
-        return canHoldAnyFluid(state) && canHoldSpecificFluid(level, pos, state, newFluid);
+        return (state.tagFlag & org.dreeam.leaf.util.BlockMasks.CAN_HOLD_ANY_FLUID) != 0 && canHoldSpecificFluid(level, pos, state, newFluid); // Leaf - optimize canHoldAnyFluid
     }
 
     private static boolean canHoldSpecificFluid(final BlockGetter level, final BlockPos pos, final BlockState state, final Fluid newFluid) {
