From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:01:50 +0900
Subject: [PATCH] Optimize getScaledTrackingDistance


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 436e73086678e4afbf94f1b7bca9b0c74266f762..168fc4fea371dfe9ea1ffca3c9f290ba3547f0b9 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -783,7 +783,13 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public int getScaledTrackingDistance(int trackingDistance) {
-        return this.getProperties().entityBroadcastRangePercentage * trackingDistance / 100;
+        // Leaf start - Optimize getScaledTrackingDistance
+        int p = this.getProperties().entityBroadcastRangePercentage;
+        if (p == 100) {
+            return trackingDistance;
+        }
+        return p * trackingDistance / 100;
+        // Leaf end - Optimize getScaledTrackingDistance
     }
 
     @Override
