From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 7 Aug 2025 00:22:50 +0800
Subject: [PATCH] Add io_uring support


diff --git a/net/minecraft/server/network/EventLoopGroupHolder.java b/net/minecraft/server/network/EventLoopGroupHolder.java
index c602146d15f16007ee2499d0a68670d0d630bfbb..6ae69f9d14fbcb6f8a88e307357747a7756d9583 100644
--- a/net/minecraft/server/network/EventLoopGroupHolder.java
+++ b/net/minecraft/server/network/EventLoopGroupHolder.java
@@ -56,6 +56,20 @@ public abstract class EventLoopGroupHolder {
         }
     };
     // Paper end - Unix domain socket support
+    // Leaf start - Add io_uring support
+    private static final EventLoopGroupHolder IO_URING = new EventLoopGroupHolder("IO_Uring", io.netty.channel.uring.IoUringSocketChannel.class, io.netty.channel.uring.IoUringServerSocketChannel.class) {
+        @Override
+        protected IoHandlerFactory ioHandlerFactory() {
+            return io.netty.channel.uring.IoUringIoHandler.newFactory();
+        }
+    };
+    private static final EventLoopGroupHolder IO_URING_UNIX_DOMAIN = new EventLoopGroupHolder("IO_Uring Unix Domain Socket", io.netty.channel.uring.IoUringDomainSocketChannel.class, io.netty.channel.uring.IoUringServerDomainSocketChannel.class) {
+        @Override
+        protected IoHandlerFactory ioHandlerFactory() {
+            return io.netty.channel.uring.IoUringIoHandler.newFactory();
+        }
+    };
+    // Leaf end - Add io_uring support
     private final String type;
     private final Class<? extends Channel> channelCls;
     private final Class<? extends ServerChannel> serverChannelCls;
@@ -64,6 +78,12 @@ public abstract class EventLoopGroupHolder {
     @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - use variant with address param
     public static EventLoopGroupHolder remote(boolean tryNativeTransport) {
         if (tryNativeTransport) {
+            // Leaf start - Add io_uring support
+            if (org.dreeam.leaf.util.LeafConstants.ENABLE_IO_URING && io.netty.channel.uring.IoUring.isAvailable()) {
+                return IO_URING;
+            }
+            // Leaf end - Add io_uring support
+
             if (KQueue.isAvailable()) {
                 return KQUEUE;
             }
@@ -79,6 +99,16 @@ public abstract class EventLoopGroupHolder {
     // Paper start - Unix domain socket support
     public static EventLoopGroupHolder remote(java.net.SocketAddress address, boolean tryNativeTransport) {
         if (tryNativeTransport) {
+            // Leaf start - Add io_uring support
+            if (org.dreeam.leaf.util.LeafConstants.ENABLE_IO_URING && io.netty.channel.uring.IoUring.isAvailable()) {
+                if (address instanceof io.netty.channel.unix.DomainSocketAddress) {
+                    return IO_URING_UNIX_DOMAIN;
+                } else {
+                    return IO_URING;
+                }
+            }
+            // Leaf end - Add io_uring support
+
             if (KQueue.isAvailable()) {
                 return KQUEUE;
             }
