From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix Paper config fixClimbingBypassingCrammingRule

Waiting for Paper#12793

diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index de22fce4f98a600a26d1d78f3bbc1c5f5a8b5bfa..f62898ec36cea977d67da087bf741fb508b1f88d 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3846,7 +3846,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         // Leaf start - Only player pushable
         final AABB box = this.getBoundingBox();
         List<Entity> pushableEntities = org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled
-            ? getNearbyPushablePlayers(this, box, EntitySelector.pushableBy(this))
+            ? getNearbyPushablePlayers(this, box, EntitySelector.pushable(this, this.level().paperConfig().collisions.fixClimbingBypassingCrammingRule)) // Leaf - Fix Paper config fixClimbingBypassingCrammingRule
             : this.level().getPushableEntities(this, box);
         // Leaf end - Only player pushable
         if (!pushableEntities.isEmpty()) {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index ceba364a50fc96865cadf30674655a109baae06d..71f13ae7f3c4f65705dd278da8e4be55da9630c4 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1908,7 +1908,7 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
     // Paper end - rewrite chunk system
 
     public List<Entity> getPushableEntities(Entity entity, AABB boundingBox) {
-        return this.getEntities(entity, boundingBox, EntitySelector.pushableBy(entity));
+        return this.getEntities(entity, boundingBox, EntitySelector.pushable(entity, this.paperConfig().collisions.fixClimbingBypassingCrammingRule)); // Leaf - Fix Paper config fixClimbingBypassingCrammingRule
     }
 
     @Nullable
