From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:00 +0900
Subject: [PATCH] optimize applyMovementEmissionAndPlaySound


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 9ab84269f6d1c7e2de69b226efb1bea2e748ab40..de2897f8717ebacbdada75ebff821146c22f35c5 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1397,7 +1397,10 @@ public abstract class Entity
         float movedDistance = (float)(clippedMovement.length() * 0.6F);
         float horizontalMovedDistance = (float)(clippedMovement.horizontalDistance() * 0.6F);
         BlockPos supportingPos = this.getOnPos();
-        BlockState supportingState = this.level().getBlockState(supportingPos);
+        // Leaf start - optimize applyMovementEmissionAndPlaySound
+        BlockState supportingState = this.level.getBlockStateIfLoadedUnchecked(supportingPos.getX(), supportingPos.getY(), supportingPos.getZ());
+        if (supportingState == null) { supportingState = this.level.getBlockState(supportingPos); }
+        // Leaf end - optimize applyMovementEmissionAndPlaySound
         boolean climbing = this.isStateClimbable(supportingState);
         this.moveDist += climbing ? movedDistance : horizontalMovedDistance;
         this.flyDist += movedDistance;
