From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:00 +0900
Subject: [PATCH] optimize applyMovementEmissionAndPlaySound


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 26fa0765b0085f239caf7de14dbb99facac4b312..685718b730a012fe8a5de1e012843807a91c2913 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1400,7 +1400,10 @@ public abstract class Entity
         float movedDistance = (float)(clippedMovement.length() * 0.6F);
         float horizontalMovedDistance = (float)(clippedMovement.horizontalDistance() * 0.6F);
         BlockPos supportingPos = this.getOnPos();
-        BlockState supportingState = this.level().getBlockState(supportingPos);
+        // Leaf start - optimize applyMovementEmissionAndPlaySound
+        BlockState supportingState = this.level.getBlockStateIfLoadedUnchecked(supportingPos.getX(), supportingPos.getY(), supportingPos.getZ());
+        if (supportingState == null) { supportingState = this.level.getBlockState(supportingPos); }
+        // Leaf end - optimize applyMovementEmissionAndPlaySound
         boolean climbing = this.isStateClimbable(supportingState);
         this.moveDist += climbing ? movedDistance : horizontalMovedDistance;
         this.flyDist += movedDistance;
