From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 14:56:22 +0900
Subject: [PATCH] remove shouldTickBlocksAt check


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index f8b72a7a0f43f92a45ff183d66b436b5fbfd7aff..24f191a254c7d75c674082c6f77bd0f30969a942 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1580,7 +1580,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         // Paper end - Fix MC-117075 use removeAll
             if (ticker.isRemoved()) {
                 ((org.dreeam.leaf.util.list.BlockEntityTickersList) this.blockEntityTickers).markAsRemoved(tickerIndex); // Paper - Fix MC-117075; use removeAll // SparklyPaper - optimize block entity removals
-            } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getPos())) {
+            } else if (tickBlockEntities /*&& this.shouldTickBlocksAt(ticker.getPos())*/) { // Leaf - remove shouldTickBlocksAt check - duplicate at BoundTickingBlockEntity#tick
                 ticker.tick();
                 // Paper start - rewrite chunk system
                 // Leaf start - SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
