From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:06 +0900
Subject: [PATCH] optimize isStateClimbable


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 685718b730a012fe8a5de1e012843807a91c2913..c1b1ea4a2327bf311a2041fb4a52981e7620b9a8 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1512,7 +1512,7 @@ public abstract class Entity
     }
 
     private boolean isStateClimbable(final BlockState state) {
-        return state.is(BlockTags.CLIMBABLE) || state.is(Blocks.POWDER_SNOW);
+        return (state.tagFlag & org.dreeam.leaf.util.BlockMasks.IS_STATE_CLIMBABLE) != 0; // Leaf - optimize isStateClimbable
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(
