From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sun, 13 Jul 2025 19:40:10 +0900
Subject: [PATCH] fix MC-298464


diff --git a/net/minecraft/world/damagesource/CombatTracker.java b/net/minecraft/world/damagesource/CombatTracker.java
index aa6213ca382e4901363317df1e30332c5166f271..03b7397b303c883b14fbac7759ff296c4e55d778 100644
--- a/net/minecraft/world/damagesource/CombatTracker.java
+++ b/net/minecraft/world/damagesource/CombatTracker.java
@@ -47,6 +47,7 @@ public class CombatTracker {
     public void recordDamageAndCheckCombatState(final CombatEntry combatEntry) {
         final DamageSource source = combatEntry.source();
     // Paper end - Combat tracker API
+        if (org.dreeam.leaf.config.modules.fixes.MC298464.enabled) { this.entries.clear(); } // Leaf - Fix MC-298464
         this.entries.add(combatEntry);
         this.lastDamageTime = this.mob.tickCount;
         this.takingDamage = true;
