From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:12:14 +0900
Subject: [PATCH] optimize fluid state access


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index c748af5c1b443bcb4863dd14cc206f36bfef640b..8658c54d489a0f9561b19dd371346c0acb7c9fd5 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1418,6 +1418,13 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return chunk.getFluidState(pos);
     }
 
+    // Leaf start - optimize fluid state access
+    public final @Nullable FluidState getFluidStateIfLoadedUnchecked(final int x, final int y, final int z) {
+        LevelChunk chunkAt = ((ServerLevel) this).getChunkSource().fullChunksNonSync.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x >> 4, z >> 4));
+        return chunkAt == null ? null : chunkAt.getFluidStateFinal(x, y, z);
+    }
+    // Leaf end - optimize fluid state access
+
     public boolean isBrightOutside() {
         return !this.dimensionType().hasFixedTime() && this.skyDarken < 4;
     }
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index c1a6b7d2f782967a7a46c871a98a2b3deace4bca..7bc485a7b86dc6195588d29f7688f61db93ebaac 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -339,6 +339,21 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         // Leaf end - optimize LevelChunk#getBlockStateFinal
     }
 
+    // Leaf start - optimize fluid state access
+    public FluidState getFluidStateFinal(final int x, final int y, final int z) {
+        final int sectionIndex = (y >> 4) - this.minSection;
+        if (sectionIndex < 0 || sectionIndex >= this.sections.length) {
+            return AIR_FLUIDSTATE;
+        } else {
+            LevelChunkSection section = this.sections[sectionIndex];
+            if (section.fluidCount <= 0) {
+                return AIR_FLUIDSTATE;
+            }
+            return section.getStates().get((y & 15) << 8 | (z & 15) << 4 | x & 15).getFluidState();
+        }
+    }
+    // Leaf end - optimize fluid state access
+
     @Override
     public BlockState getBlockState(final BlockPos pos) {
         if (true) {
diff --git a/net/minecraft/world/level/chunk/LevelChunkSection.java b/net/minecraft/world/level/chunk/LevelChunkSection.java
index 705f6314d2550b0f1c8c5209dd65161d89cbfd7e..6a9bff7b23dbf7b8acb32e397704446440a3d630 100644
--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -12,7 +12,7 @@ import net.minecraft.world.level.material.FluidState;
 public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_counting.BlockCountingChunkSection { // Paper - block counting
     public static final int BIOME_CONTAINER_BITS = 2;
     short nonEmptyBlockCount; // Paper - package private
-    private short fluidCount;
+    public short fluidCount; // Leaf - optimize fluid state access - private -> public
     private short tickingBlockCount;
     private short tickingFluidCount;
     private final PalettedContainer<BlockState> states;
