From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:46:30 +0200
Subject: [PATCH] Skip PreCreatureSpawnEvent if no listeners


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 55580d94e22871edd217cceea34fc1d05d807811..ec6c727a685665b253b28df6a1bb48a97c9d6b8d 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -455,6 +455,8 @@ public final class NaturalSpawner {
         double distance
     ) {
         EntityType<?> entityType = data.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         // Paper start - PreCreatureSpawnEvent
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
                 org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
@@ -466,6 +468,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = entityType.getCategory() != MobCategory.MISC
         // Paper end - PreCreatureSpawnEvent
             && (
@@ -720,6 +724,8 @@ public final class NaturalSpawner {
         LevelChunk chunk
     ) {
         EntityType<?> entityType = data.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         // Paper start - PreCreatureSpawnEvent
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
             org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
@@ -731,6 +737,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = entityType.getCategory() != MobCategory.MISC
             // Paper end - PreCreatureSpawnEvent
             && (
