From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 4 Jul 2025 03:13:33 +0900
Subject: [PATCH] optimize waypoint


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2f9304c4fc7ac9f42493bc19748ee6c0f57208ee..5d4350dbe1d4189e17de77586a342c4ed0087c13 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5112,6 +5112,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             return;
         }
         // Paper end - Block invalid positions and bounding box
+        boolean blockUpdated; // Leaf - optimize waypoint
         if (this.position.x != x || this.position.y != y || this.position.z != z) {
             synchronized (this.posLock) { // Paper - detailed watchdog information
             this.position = new Vec3(x, y, z);
@@ -5119,7 +5120,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             int floor = Mth.floor(x);
             int floor1 = Mth.floor(y);
             int floor2 = Mth.floor(z);
-            if (floor != this.blockPosition.getX() || floor1 != this.blockPosition.getY() || floor2 != this.blockPosition.getZ()) {
+            blockUpdated = floor != this.blockPosition.getX() || floor1 != this.blockPosition.getY() || floor2 != this.blockPosition.getZ(); // Leaf - optimize waypoint
+            if (blockUpdated) { // Leaf - optimize waypoint
                 this.blockPosition = new BlockPos(floor, floor1, floor2);
                 this.inBlockState = null;
                 if (SectionPos.blockToSectionCoord(floor) != this.chunkPosition.x || SectionPos.blockToSectionCoord(floor2) != this.chunkPosition.z) {
@@ -5128,7 +5130,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             }
 
             this.levelCallback.onMove();
-            if (!this.firstTick && this.level instanceof ServerLevel serverLevel && !this.isRemoved()) {
+            if ((!org.dreeam.leaf.config.modules.opt.OptimizeWaypoint.enabled || blockUpdated) && !this.firstTick && this.level instanceof ServerLevel serverLevel && !this.isRemoved()) { // Leaf - optimize waypoint
                 if (this instanceof WaypointTransmitter waypointTransmitter && waypointTransmitter.isTransmittingWaypoint()) {
                     serverLevel.getWaypointManager().updateWaypoint(waypointTransmitter);
                 }
