From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix Paper config fixClimbingBypassingCrammingRule

Waiting for Paper#12793

diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 4ec5f27d771f98178c361c58b6929f6b0606af9b..c63c2d1b6bd005eda12c8c1f7cec2a29856fed14 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3844,7 +3844,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         // Leaf start - Only player pushable
         final AABB box = this.getBoundingBox();
         List<Entity> pushableEntities = org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled
-            ? getNearbyPushablePlayers(this, box, EntitySelector.pushableBy(this))
+            ? getNearbyPushablePlayers(this, box, EntitySelector.pushable(this, this.level().paperConfig().collisions.fixClimbingBypassingCrammingRule)) // Leaf - Fix Paper config fixClimbingBypassingCrammingRule
             : this.level().getPushableEntities(this, box);
         // Leaf end - Only player pushable
         if (!pushableEntities.isEmpty()) {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 38a6ca1c0789e50ca15f7a5e602da9be5f8371e9..afebf43067f13493d06e6c0620058c135eb9b880 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1907,7 +1907,7 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
     // Paper end - rewrite chunk system
 
     public List<Entity> getPushableEntities(Entity entity, AABB boundingBox) {
-        return this.getEntities(entity, boundingBox, EntitySelector.pushableBy(entity));
+        return this.getEntities(entity, boundingBox, EntitySelector.pushable(entity, this.paperConfig().collisions.fixClimbingBypassingCrammingRule)); // Leaf - Fix Paper config fixClimbingBypassingCrammingRule
     }
 
     @Nullable
