From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:00 +0900
Subject: [PATCH] optimize applyMovementEmissionAndPlaySound


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 5ab506486b86f3c4a407e91d4eb7f302e277f982..61415841531d0c2cba23270d24dd213ee1ba6006 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1396,7 +1396,10 @@ public abstract class Entity
         float movedDistance = (float)(clippedMovement.length() * 0.6F);
         float horizontalMovedDistance = (float)(clippedMovement.horizontalDistance() * 0.6F);
         BlockPos supportingPos = this.getOnPos();
-        BlockState supportingState = this.level().getBlockState(supportingPos);
+        // Leaf start - optimize applyMovementEmissionAndPlaySound
+        BlockState supportingState = this.level.getBlockStateIfLoadedUnchecked(supportingPos.getX(), supportingPos.getY(), supportingPos.getZ());
+        if (supportingState == null) { supportingState = this.level.getBlockState(supportingPos); }
+        // Leaf end - optimize applyMovementEmissionAndPlaySound
         boolean climbing = this.isStateClimbable(supportingState);
         this.moveDist += climbing ? movedDistance : horizontalMovedDistance;
         this.flyDist += movedDistance;
