From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:12:14 +0900
Subject: [PATCH] optimize fluid state access


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 24f191a254c7d75c674082c6f77bd0f30969a942..a7e2df85ca8d49245c51028cfec77243dd01f621 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1413,6 +1413,13 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return chunk.getFluidState(pos);
     }
 
+    // Leaf start - counting chunk section fluid
+    public final @Nullable FluidState getFluidStateIfLoadedUnchecked(final int x, final int y, final int z) {
+        LevelChunk chunkAt = ((ServerLevel) this).chunkSource.fullChunksNonSync.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x >> 4, z >> 4));
+        return chunkAt == null ? null : chunkAt.getFluidStateFinal(x, y, z);
+    }
+    // Leaf end - counting chunk section fluid
+
     public boolean isBrightOutside() {
         return !this.dimensionType().hasFixedTime() && this.skyDarken < 4;
     }
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 635d1d61705b4d929dc8fe7070f5d7c53fdc9ea8..bbad6066aba4b03614bb59314757c2713dfa61fa 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -339,6 +339,21 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         // Leaf end - optimize LevelChunk#getBlockStateFinal
     }
 
+    // Leaf start - counting chunk section fluid
+    public FluidState getFluidStateFinal(final int x, final int y, final int z) {
+        final int sectionIndex = (y >> 4) - this.minSection;
+        if (sectionIndex < 0 || sectionIndex >= this.sections.length) {
+            return AIR_FLUIDSTATE;
+        } else {
+            LevelChunkSection section = this.sections[sectionIndex];
+            if (section.fluidCount <= 0) {
+                return AIR_FLUIDSTATE;
+            }
+            return section.states.get((y & 15) << 8 | (z & 15) << 4 | x & 15).getFluidState();
+        }
+    }
+    // Leaf end - counting chunk section fluid
+
     @Override
     public BlockState getBlockState(final BlockPos pos) {
         if (true) {
diff --git a/net/minecraft/world/level/chunk/LevelChunkSection.java b/net/minecraft/world/level/chunk/LevelChunkSection.java
index 0f807317cc7fa113ee57ab31f91bb6b8df7c4ae0..40c3a4037061606d983ef9e268506ba19dc5449d 100644
--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -15,7 +15,7 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
     public static final int SECTION_SIZE = 4096;
     public static final int BIOME_CONTAINER_BITS = 2;
     short nonEmptyBlockCount; // Paper - package private
-    private short fluidCount;
+    public short fluidCount; // Leaf - counting chunk section fluid - private -> public
     private short tickingBlockCount;
     private short tickingFluidCount;
     public final PalettedContainer<BlockState> states;
