From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:46:30 +0200
Subject: [PATCH] Skip PreCreatureSpawnEvent if no listeners


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index cb72f8fbdbd7eb7c9e54fa9ee5c23d6dad04ef54..51f1241df3d8462634d161aa33a53fc468c65f52 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -457,6 +457,8 @@ public final class NaturalSpawner {
         double distance
     ) {
         EntityType<?> entityType = data.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         // Paper start - PreCreatureSpawnEvent
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
                 org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
@@ -468,6 +470,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = entityType.getCategory() != MobCategory.MISC
         // Paper end - PreCreatureSpawnEvent
             && (
@@ -722,6 +726,8 @@ public final class NaturalSpawner {
         LevelChunk chunk
     ) {
         EntityType<?> entityType = data.type();
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
         // Paper start - PreCreatureSpawnEvent
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
             org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
@@ -733,6 +739,8 @@ public final class NaturalSpawner {
             }
             return PreSpawnStatus.CANCELLED;
         }
+        }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
         final boolean success = entityType.getCategory() != MobCategory.MISC
             // Paper end - PreCreatureSpawnEvent
             && (
