From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:06 +0900
Subject: [PATCH] optimize isStateClimbable


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 61415841531d0c2cba23270d24dd213ee1ba6006..33a1a22426f7e55b434f1bd476d677124bfc9b10 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1508,7 +1508,7 @@ public abstract class Entity
     }
 
     private boolean isStateClimbable(final BlockState state) {
-        return state.is(BlockTags.CLIMBABLE) || state.is(Blocks.POWDER_SNOW);
+        return (state.tagFlag & org.dreeam.leaf.util.BlockMasks.IS_STATE_CLIMBABLE) != 0; // Leaf - optimize isStateClimbable
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(
