From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:17 +0900
Subject: [PATCH] optimize getOnPos


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 91a53e131398ce1cd0f8e6b4aa1e332eb1f50b1c..c5b065f2cf3f2f1eb38f78a11fd80171dbed7939 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1585,24 +1585,28 @@ public abstract class Entity
     }
 
     protected BlockPos getOnPos(final float offset) {
-        if (this.mainSupportingBlockPos.isPresent() && this.level().getChunkIfLoadedImmediately(this.mainSupportingBlockPos.get()) != null) { // Paper - ensure no loads
+        // Leaf start - optimize getOnPos
+        if (this.mainSupportingBlockPos.isPresent() /*&& this.level().getChunkIfLoadedImmediately(this.mainSupportingBlockPos.get()) != null*/) { // Paper - ensure no loads
             BlockPos getOnPos = this.mainSupportingBlockPos.get();
-            if (!(offset > 1.0E-5F)) {
-                return getOnPos;
+            net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(getOnPos);
+            if (chunk != null) {
+                if (!(offset > 1.0E-5F)) {
+                    return getOnPos;
+                }
+                BlockState belowState = chunk.getBlockStateFinal(getOnPos.getX(), getOnPos.getY(), getOnPos.getZ());
+                int flags = belowState.tagFlag;
+                return (!(offset <= 0.5) || (flags & org.dreeam.leaf.util.BlockMasks.FENCE_TAG) == 0)
+                    && (flags & org.dreeam.leaf.util.BlockMasks.WALL_TAG) == 0
+                    && (flags & org.dreeam.leaf.util.BlockMasks.FENCE_GATE_CL) == 0
+                    ? getOnPos.atY(Mth.floor(this.position.y - offset))
+                    : getOnPos;
             }
-
-            BlockState belowState = this.level().getBlockState(getOnPos);
-            return (!(offset <= 0.5) || !belowState.is(BlockTags.FENCES))
-                    && !belowState.is(BlockTags.WALLS)
-                    && !(belowState.getBlock() instanceof FenceGateBlock)
-                ? getOnPos.atY(Mth.floor(this.position.y - offset))
-                : getOnPos;
-        } else {
-            int xTruncated = Mth.floor(this.position.x);
-            int yTruncatedBelow = Mth.floor(this.position.y - offset);
-            int zTruncated = Mth.floor(this.position.z);
-            return new BlockPos(xTruncated, yTruncatedBelow, zTruncated);
         }
+        int xTruncated = Mth.floor(this.position.x);
+        int yTruncatedBelow = Mth.floor(this.position.y - offset);
+        int zTruncated = Mth.floor(this.position.z);
+        return new BlockPos(xTruncated, yTruncatedBelow, zTruncated);
+        // Leaf end - optimize getOnPos
     }
 
     protected float getBlockJumpFactor() {
