From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:06 +0900
Subject: [PATCH] optimize isStateClimbable


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index d799fd5b64e5ad932a50ffb38a62d3210376b963..7b430f18b55765240ecfc152272ab979dd9415d8 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1446,7 +1446,7 @@ public abstract class Entity
     }
 
     private boolean isStateClimbable(final BlockState state) {
-        return state.is(BlockTags.CLIMBABLE) || state.is(Blocks.POWDER_SNOW);
+        return (state.tagFlag & org.dreeam.leaf.util.BlockMasks.IS_STATE_CLIMBABLE) != 0; // Leaf - optimize isStateClimbable
     }
 
     private boolean vibrationAndSoundEffectsFromBlock(
