From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 7 Aug 2025 00:22:50 +0800
Subject: [PATCH] Bump netty to 4.2.x


diff --git a/net/minecraft/gametest/framework/GameTestServer.java b/net/minecraft/gametest/framework/GameTestServer.java
index 340d5487fa778277b9560250271c5143d80d9987..99543330da85ddecd91c954c8aa386c89a96c23f 100644
--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -333,6 +333,18 @@ public class GameTestServer extends MinecraftServer {
         return false;
     }
 
+    // Leaf start - Bump netty to 4.2.x
+    @Override
+    public boolean isIoUringEnabled() {
+        return false;
+    }
+
+    @Override
+    public boolean isKqueueEnabled() {
+        return false;
+    }
+    // Leaf end - Bump netty to 4.2.x
+
     @Override
     public boolean isCommandBlockEnabled() {
         return true;
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 4fbffd06d23860df912ff216446699e0ab5eb00b..8fc7e0d3ae4a3056c0e9e5e19af4261cfe4adfb0 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2055,6 +2055,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     public abstract boolean isEpollEnabled();
 
+    // Leaf start - Bump netty to 4.2.x
+    public abstract boolean isIoUringEnabled();
+
+    public abstract boolean isKqueueEnabled();
+    // Leaf end - Bump netty to 4.2.x
+
     public boolean isPvpAllowed() {
         return this.pvp;
     }
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index ee20dd09bba40a4a98df501614d208552905c3cd..89fb3276c50a96bdafb0f100d4eddc476aa4f61d 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -503,9 +503,21 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public boolean isEpollEnabled() {
-        return this.getProperties().useNativeTransport;
+        return this.getProperties().useNativeTransport && org.dreeam.leaf.NetworkIoModel.fromProperty() == org.dreeam.leaf.NetworkIoModel.EPOLL; // Leaf - Bump netty to 4.2.x
     }
 
+    // Leaf start - Bump netty to 4.2.x
+    @Override
+    public boolean isIoUringEnabled() {
+        return this.getProperties().useNativeTransport && org.dreeam.leaf.NetworkIoModel.fromProperty() == org.dreeam.leaf.NetworkIoModel.IO_URING;
+    }
+
+    @Override
+    public boolean isKqueueEnabled() {
+        return this.getProperties().useNativeTransport && org.dreeam.leaf.NetworkIoModel.fromProperty() == org.dreeam.leaf.NetworkIoModel.KQUEUE;
+    }
+    // Leaf end - Bump netty to 4.2.x
+
     @Override
     public DedicatedPlayerList getPlayerList() {
         return (DedicatedPlayerList)super.getPlayerList();
diff --git a/net/minecraft/server/network/ServerConnectionListener.java b/net/minecraft/server/network/ServerConnectionListener.java
index bd07e6a5aa1883786d789ea71711a0c0c0a95c26..ea5dfaa2efaccd643f9977cb60d982791439a597 100644
--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -48,12 +48,20 @@ import org.slf4j.Logger;
 
 public class ServerConnectionListener {
     private static final Logger LOGGER = LogUtils.getLogger();
+    // Leaf start - Bump netty to 4.2.x
+    /*
     public static final Supplier<NioEventLoopGroup> SERVER_EVENT_GROUP = Suppliers.memoize(
         () -> new NioEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Server IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()) // Paper
     );
     public static final Supplier<EpollEventLoopGroup> SERVER_EPOLL_EVENT_GROUP = Suppliers.memoize(
         () -> new EpollEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Epoll Server IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()) // Paper
     );
+    */
+    public static final Supplier<io.netty.channel.IoHandlerFactory> NIO_IO_HANDLE_FACTORY = io.netty.channel.nio.NioIoHandler::newFactory;
+    public static final Supplier<io.netty.channel.IoHandlerFactory> EPOLL_IO_HANDLE_FACTORY = io.netty.channel.epoll.EpollIoHandler::newFactory;
+    public static final Supplier<io.netty.channel.IoHandlerFactory> IO_URING_IO_HANDLE_FACTORY = io.netty.channel.uring.IoUringIoHandler::newFactory;
+    public static final Supplier<io.netty.channel.IoHandlerFactory> KQUEUE_IO_HANDLE_FACTORY = io.netty.channel.kqueue.KQueueIoHandler::newFactory;
+    // Leaf end - Bump netty to 4.2.x
     final MinecraftServer server;
     public volatile boolean running;
     private final List<ChannelFuture> channels = Collections.synchronizedList(Lists.newArrayList());
@@ -87,7 +95,29 @@ public class ServerConnectionListener {
         synchronized (this.channels) {
             Class<? extends io.netty.channel.ServerChannel> clazz; // Paper - Unix domain socket support
             EventLoopGroup eventLoopGroup;
-            if (Epoll.isAvailable() && this.server.isEpollEnabled()) {
+            // Leaf start - Bump netty to 4.2.x
+            if (io.netty.channel.uring.IoUring.isAvailable() && this.server.isIoUringEnabled()) {
+                // Paper start - Unix domain socket support
+                if (address instanceof io.netty.channel.unix.DomainSocketAddress) {
+                    clazz = io.netty.channel.uring.IoUringServerDomainSocketChannel.class;
+                } else {
+                    clazz = io.netty.channel.uring.IoUringServerSocketChannel.class;
+                }
+                // Paper end - Unix domain socket supp
+                eventLoopGroup = createEventLoopGroup("IoUring", IO_URING_IO_HANDLE_FACTORY);
+                LOGGER.info("Using io_uring channel type");
+            } else if (io.netty.channel.kqueue.KQueue.isAvailable() && this.server.isKqueueEnabled()) {
+                // Paper start - Unix domain socket support
+                if (address instanceof io.netty.channel.unix.DomainSocketAddress) {
+                    clazz = io.netty.channel.kqueue.KQueueServerDomainSocketChannel.class;
+                } else {
+                    clazz = io.netty.channel.kqueue.KQueueServerSocketChannel.class;
+                }
+                // Paper end - Unix domain socket support
+                eventLoopGroup = createEventLoopGroup("Kqueue", KQUEUE_IO_HANDLE_FACTORY);
+                LOGGER.info("Using kqueue channel type");
+            } else if (Epoll.isAvailable() && this.server.isEpollEnabled()) {
+                // Leaf end - Bump netty to 4.2.x
                 // Paper start - Unix domain socket support
                 if (address instanceof io.netty.channel.unix.DomainSocketAddress) {
                     clazz = io.netty.channel.epoll.EpollServerDomainSocketChannel.class;
@@ -95,12 +125,12 @@ public class ServerConnectionListener {
                 clazz = EpollServerSocketChannel.class;
                 }
                 // Paper end - Unix domain socket support
-                eventLoopGroup = SERVER_EPOLL_EVENT_GROUP.get();
+                eventLoopGroup = createEventLoopGroup("Epoll", EPOLL_IO_HANDLE_FACTORY); // Leaf - Bump netty to 4.2.x
                 LOGGER.info("Using epoll channel type");
             } else {
                 clazz = NioServerSocketChannel.class;
-                eventLoopGroup = SERVER_EVENT_GROUP.get();
-                LOGGER.info("Using default channel type");
+                eventLoopGroup = createEventLoopGroup("Nio", NIO_IO_HANDLE_FACTORY); // Leaf - Bump netty to 4.2.x
+                LOGGER.info("Using default nio channel type"); // Leaf - Bump netty to 4.2.x
             }
 
             // Paper start - Warn people with console access that HAProxy is in use.
@@ -195,6 +225,19 @@ public class ServerConnectionListener {
     }
     // CraftBukkit end
 
+    // Leaf start - Bump netty to 4.2.x
+    public EventLoopGroup createEventLoopGroup(final String name, final Supplier<io.netty.channel.IoHandlerFactory> ioHandlerFactorySupplier) {
+        return new io.netty.channel.MultiThreadIoEventLoopGroup(
+            0,
+            new ThreadFactoryBuilder().setNameFormat("Netty " + name + " Server IO #%d")
+                .setDaemon(true)
+                .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER))
+                .build(),
+            ioHandlerFactorySupplier.get()
+        );
+    }
+    // Leaf end - Bump netty to 4.2.x
+
     public SocketAddress startMemoryChannel() {
         ChannelFuture channelFuture;
         synchronized (this.channels) {
@@ -215,7 +258,7 @@ public class ServerConnectionListener {
                         }
                     }
                 )
-                .group(SERVER_EVENT_GROUP.get())
+                //.group(SERVER_EVENT_GROUP.get()) // Leaf - Bump netty to 4.2.x - not used, comment
                 .localAddress(LocalAddress.ANY)
                 .bind()
                 .syncUninterruptibly();
