From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:30:37 +0900
Subject: [PATCH] optimize tickEffects


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index f00d4e83a7ed59e08bf9a90f1bcab0381f59299e..d9d2f94e8cce69109923133caa098588ba84039f 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -965,6 +965,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     protected void tickEffects() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            if (!this.activeEffects.isEmpty()) { // Leaf - optimize tickEffects
             Iterator<Holder<MobEffect>> iterator = this.activeEffects.keySet().iterator();
 
             this.tickingEffects = true; // CraftBukkit
@@ -989,6 +990,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
             // CraftBukkit start
             this.tickingEffects = false;
+            } // Leaf - optimize tickEffects
             Runnable task;
             while ((task = this.deferredEffectUpdates.poll()) != null) {
                 task.run();
