From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 6 Jun 2025 05:44:21 +0800
Subject: [PATCH] do not log invalid flatten text component parse


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
index d5dfa9a64e2d3f9a37a5bdde60b92ed93bd89aed..83155bf0befab0207a4b756b19527d071037a870 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
@@ -254,7 +254,7 @@ public final class V4290 {
                         return ret;
                     }
                 } catch (final JsonParseException ex) {
-                    LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex);
+                    if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.invalidLegacyTextComponent) LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex); // Leaf - do not log invalid flatten text component parse
                 }
 
                 return null;
