From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Tue, 22 Jul 2025 22:31:58 +0200
Subject: [PATCH] Op lock


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index f0bdc2bbf358d4c5f113b395c6409ff9fa759e2f..203160ddf928b86d12a25c94550bf13b86598e03 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1202,6 +1202,9 @@ public abstract class PlayerList {
     }
 
     public void op(GameProfile profile) {
+        if (org.dreeam.leaf.config.modules.misc.OpLock.preventOpChanges) {
+            return;
+        }
         this.ops.add(new ServerOpListEntry(profile, this.server.getOperatorUserPermissionLevel(), this.ops.canBypassPlayerLimit(profile)));
         ServerPlayer player = this.getPlayer(profile.getId());
         if (player != null) {
@@ -1210,6 +1213,9 @@ public abstract class PlayerList {
     }
 
     public void deop(GameProfile profile) {
+        if (org.dreeam.leaf.config.modules.misc.OpLock.preventOpChanges) {
+            return;
+        }
         this.ops.remove(profile);
         ServerPlayer player = this.getPlayer(profile.getId());
         if (player != null) {
