From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix Paper config fixClimbingBypassingCrammingRule

Waiting for Paper#12793

diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 3a669d98db60bfe19aa0c4b55dc28d06d44f7b88..fdb6527c9d7e18c37e681ea2d8e4428ece49e2ff 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3850,7 +3850,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         // Leaf start - Only player pushable
         final AABB box = this.getBoundingBox();
         List<Entity> pushableEntities = org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled
-            ? getNearbyPushablePlayers(this, box, EntitySelector.pushableBy(this))
+            ? getNearbyPushablePlayers(this, box, EntitySelector.pushable(this, this.level().paperConfig().collisions.fixClimbingBypassingCrammingRule)) // Leaf - Fix Paper config
             : this.level().getPushableEntities(this, box);
         // Leaf end - Only player pushable
         if (!pushableEntities.isEmpty()) {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 9b4e6fdbedfa378ea436cdd5f1a36e2b3d94be17..645fad313ed3cd6ea5afe25c5d29dce828993371 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1909,7 +1909,7 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
     // Paper end - rewrite chunk system
 
     public List<Entity> getPushableEntities(Entity entity, AABB boundingBox) {
-        return this.getEntities(entity, boundingBox, EntitySelector.pushableBy(entity));
+        return this.getEntities(entity, boundingBox, EntitySelector.pushable(entity, this.paperConfig().collisions.fixClimbingBypassingCrammingRule)); // Leaf - Fix Paper config fixClimbingBypassingCrammingRule
     }
 
     @Nullable
