From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:00:57 +0900
Subject: [PATCH] optimize FluidState#is(TagKey)


diff --git a/net/minecraft/world/level/material/FluidState.java b/net/minecraft/world/level/material/FluidState.java
index 3e5ada4a84bc1da06102dfacb0d3d45a4fc6db8a..4a39a9b6401d615047d4573b3cc64a3e7361d89b 100644
--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -130,7 +130,16 @@ public final class FluidState extends StateHolder<Fluid, FluidState> implements
     }
 
     public boolean is(TagKey<Fluid> tag) {
-        return this.getType().builtInRegistryHolder().is(tag);
+        // Leaf start - optimize FluidState#is(TagKey)
+        Fluid fluid = this.getType();
+        if (fluid == Fluids.WATER || fluid == Fluids.FLOWING_WATER) {
+            return tag == net.minecraft.tags.FluidTags.WATER;
+        } else if (fluid == Fluids.LAVA || fluid == Fluids.FLOWING_LAVA) {
+            return tag == net.minecraft.tags.FluidTags.LAVA;
+        } else {
+            return false;
+        }
+        // Leaf end - optimize FluidState#is(TagKey)
     }
 
     public boolean is(HolderSet<Fluid> fluids) {
