From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:30:37 +0900
Subject: [PATCH] optimize tickEffects


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index d358043589f205cb1c2118709277aed50373f165..ecdc74bda848deb15cf2f85dd40a9c506e08b40d 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -963,6 +963,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     protected void tickEffects() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            if (!this.activeEffects.isEmpty()) { // Leaf - optimize tickEffects
             Iterator<Holder<MobEffect>> iterator = this.activeEffects.keySet().iterator();
 
             this.isTickingEffects = true; // CraftBukkit
@@ -988,6 +989,8 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
             // CraftBukkit start
             this.isTickingEffects = false;
+            } // Leaf - optimize tickEffects
+            if (!effectsToProcess.isEmpty()) { // Leaf - optimize tickEffects
             for (ProcessableEffect effect : this.effectsToProcess) {
                 if (effect.effect != null) {
                     this.addEffect(effect.effect, effect.cause);
@@ -996,6 +999,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
             }
             this.effectsToProcess.clear();
+            } // Leaf - optimize tickEffects
             // CraftBukkit end
         } else {
             for (MobEffectInstance effect : this.activeEffects.values()) {
