From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Tue, 22 Jul 2025 22:31:58 +0200
Subject: [PATCH] Op lock


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index d8bfb169b10bb2a742ecd0f4919db38da77998ad..aa5d6309b2194174ddc3163ce8b3839e59805f06 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1201,6 +1201,9 @@ public abstract class PlayerList {
     }
 
     public void op(GameProfile profile) {
+        if (org.dreeam.leaf.config.modules.misc.OpLock.preventOpChanges) {
+            return;
+        }
         this.ops.add(new ServerOpListEntry(profile, this.server.getOperatorUserPermissionLevel(), this.ops.canBypassPlayerLimit(profile)));
         ServerPlayer player = this.getPlayer(profile.getId());
         if (player != null) {
@@ -1209,6 +1212,9 @@ public abstract class PlayerList {
     }
 
     public void deop(GameProfile profile) {
+        if (org.dreeam.leaf.config.modules.misc.OpLock.preventOpChanges) {
+            return;
+        }
         this.ops.remove(profile);
         ServerPlayer player = this.getPlayer(profile.getId());
         if (player != null) {
