From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 11 Jul 2025 01:38:51 +0900
Subject: [PATCH] cache ItemStack max stack size

Co-authored-by: Taiyou06 <kaandindar21@gmail.com>

diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 7cf8894b1fd6283567dca0b7d0b945b2fbae1b8c..996a89636fd18112784fa7b57afe37e7f02518d0 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -158,6 +158,7 @@ public final class ItemStack implements DataComponentHolder {
     public static final ItemStack EMPTY = new ItemStack((Void)null);
     private static final Component DISABLED_ITEM_TOOLTIP = Component.translatable("item.disabled").withStyle(ChatFormatting.RED);
     private int count;
+    private int maxStackSize; // Leaf - cache ItemStack max stack size
     private int popTime;
     @Deprecated
     @Nullable
@@ -283,12 +284,14 @@ public final class ItemStack implements DataComponentHolder {
         this.item = item.asItem();
         this.count = count;
         this.components = components;
+        this.maxStackSize = this.getMaxStackSizeInternal(); // Leaf - cache ItemStack max stack size
         this.getItem().verifyComponentsAfterLoad(this);
     }
 
     private ItemStack(@Nullable Void unused) {
         this.item = null;
         this.components = new PatchedDataComponentMap(DataComponentMap.EMPTY);
+        this.maxStackSize = 1; // Leaf - cache ItemStack max stack size
     }
 
     public static DataResult<Unit> validateComponents(DataComponentMap components) {
@@ -579,10 +582,16 @@ public final class ItemStack implements DataComponentHolder {
         return itemStack;
     }
 
-    public int getMaxStackSize() {
+    // Leaf start - cache ItemStack max stack size
+    private int getMaxStackSizeInternal() {
         return this.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
     }
 
+    public int getMaxStackSize() {
+        return maxStackSize;
+    }
+    // Leaf end - cache ItemStack max stack size
+
     public boolean isStackable() {
         return this.getMaxStackSize() > 1 && (!this.isDamageableItem() || !this.isDamaged());
     }
@@ -979,12 +988,19 @@ public final class ItemStack implements DataComponentHolder {
     // CraftBukkit start
     public void restorePatch(DataComponentPatch datacomponentpatch) {
         this.components.restorePatch(datacomponentpatch);
+        this.maxStackSize = this.getMaxStackSizeInternal(); // Leaf - cache ItemStack max stack size
     }
     // CraftBukkit end
 
     @Nullable
     public <T> T set(DataComponentType<T> component, @Nullable T value) {
-        return this.components.set(component, value);
+        // Leaf start - cache ItemStack max stack size
+        T o = this.components.set(component, value);
+        if (component == DataComponents.MAX_STACK_SIZE) {
+            this.maxStackSize = getMaxStackSizeInternal();
+        }
+        return o;
+        // Leaf end - cache ItemStack max stack size
     }
 
     public <T> void copyFrom(DataComponentType<T> componentType, DataComponentGetter componentGetter) {
@@ -1010,10 +1026,12 @@ public final class ItemStack implements DataComponentHolder {
     public void applyComponentsAndValidate(DataComponentPatch components) {
         DataComponentPatch patch = this.components.asPatch();
         this.components.applyPatch(components);
+        this.maxStackSize = this.getMaxStackSizeInternal(); // Leaf - cache ItemStack max stack size
         Optional<Error<ItemStack>> optional = validateStrict(this).error();
         if (optional.isPresent()) {
             LOGGER.error("Failed to apply component patch '{}' to item: '{}'", components, optional.get().message());
             this.components.restorePatch(patch);
+            this.maxStackSize = this.getMaxStackSizeInternal(); // Leaf - cache ItemStack max stack size
         } else {
             this.getItem().verifyComponentsAfterLoad(this);
         }
@@ -1021,11 +1039,13 @@ public final class ItemStack implements DataComponentHolder {
 
     public void applyComponents(DataComponentPatch components) {
         this.components.applyPatch(components);
+        this.maxStackSize = this.getMaxStackSizeInternal(); // Leaf - cache ItemStack max stack size
         this.getItem().verifyComponentsAfterLoad(this);
     }
 
     public void applyComponents(DataComponentMap components) {
         this.components.setAll(components);
+        this.maxStackSize = this.getMaxStackSizeInternal(); // Leaf - cache ItemStack max stack size
         this.getItem().verifyComponentsAfterLoad(this);
     }
 
