From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:46:30 +0200
Subject: [PATCH] Skip PreCreatureSpawnEvent if no listeners


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 6c6d768d4aa5a8c90bff2433489d777c5e66069b..ec2f630b8eeed85e81b730e3a3d339b3419f39e8 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -519,17 +519,35 @@ public final class NaturalSpawner {
         LevelChunk chunk
     ) {
         EntityType<?> entityType = data.type();
-        // Paper start - PreCreatureSpawnEvent
-        com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
-            org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
-            org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(entityType), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
-        );
-        if (!event.callEvent()) {
-            if (event.shouldAbortSpawn()) {
-                return PreSpawnStatus.ABORT;
+
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        boolean shouldAbort = false;
+        boolean cancelled = false;
+
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            // Paper start - PreCreatureSpawnEvent
+            com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
+                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
+                org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(entityType),
+                org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
+            );
+            if (!event.callEvent()) {
+                if (event.shouldAbortSpawn()) {
+                    shouldAbort = true;
+                }
+                cancelled = true;
             }
+            // Paper end - PreCreatureSpawnEvent
+        }
+
+        if (shouldAbort) {
+            return PreSpawnStatus.ABORT;
+        }
+        if (cancelled) {
             return PreSpawnStatus.CANCELLED;
         }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
+
         final boolean success = entityType.getCategory() != MobCategory.MISC
             // Paper end - PreCreatureSpawnEvent
             && (
