From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:31:35 +0900
Subject: [PATCH] Only update frozen ticks if changed


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 33408a42f41b2b7b156b98c814236d67331ddf55..12fbe7b94fa9e89889bb12bd0d3b5c88d99ae989 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3921,7 +3921,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         if (this.level() instanceof ServerLevel serverLevel) {
             profiler.push("freezing");
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // Leaf start - Only update frozen ticks if changed
+                int ticksFrozen = this.getTicksFrozen();
+                int newTicksFrozen = Math.max(0, ticksFrozen - 2);
+                if (ticksFrozen != newTicksFrozen) {
+                    this.setTicksFrozen(newTicksFrozen);
+                }
+                // Leaf end - Only update frozen ticks if changed
             }
 
             this.removeFrost();
