From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Mon, 9 Feb 2026 12:37:26 +0300
Subject: [PATCH] Optimize map lookups with isEmpty check


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 74b85bc08dd8ed0f5a83eb612d7e24a026f8b05a..ccbbbabb936ef8d50401c159b3a09480ed01929d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -270,6 +270,7 @@ public abstract class Entity
     private int id = 0;
     public boolean blocksBuilding;
     public ImmutableList<Entity> passengers = ImmutableList.of();
+    public static final ImmutableList<Entity> EMPTY_PASSENGERS = ImmutableList.of(); // Leaf - Optimize map lookups with isEmpty check
     protected int boardingCooldown;
     private @Nullable Entity vehicle;
     private Level level;
@@ -490,7 +491,7 @@ public abstract class Entity
 
     @Override
     public final boolean moonrise$hasAnyPlayerPassengers() {
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             return false;
         }
         return this.getIndirectPassengersStream().anyMatch((entity) -> entity instanceof Player);
@@ -3699,7 +3700,7 @@ public abstract class Entity
             throw new IllegalStateException("Use x.startRiding(y), not y.addPassenger(x)");
         }
 
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             this.passengers = ImmutableList.of(passenger);
         } else {
             List<Entity> newPassengers = Lists.newArrayList(this.passengers);
@@ -3776,7 +3777,7 @@ public abstract class Entity
     }
 
     protected boolean canAddPassenger(final Entity passenger) {
-        return this.passengers.isEmpty();
+        return this.passengers == EMPTY_PASSENGERS; // Leaf - Optimize map lookups with isEmpty check
     }
 
     protected boolean couldAcceptPassenger() {
@@ -3983,7 +3984,7 @@ public abstract class Entity
     }
 
     public boolean isVehicle() {
-        return !this.passengers.isEmpty();
+        return this.passengers != EMPTY_PASSENGERS; // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean dismountsUnderwater() {
@@ -5081,11 +5082,11 @@ public abstract class Entity
     }
 
     public @Nullable Entity getFirstPassenger() {
-        return this.passengers.isEmpty() ? null : this.passengers.get(0);
+        return this.passengers == EMPTY_PASSENGERS ? null : this.passengers.get(0); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean hasPassenger(final Entity entity) {
-        return this.passengers.contains(entity);
+        return passengers != EMPTY_PASSENGERS && this.passengers.contains(entity); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean hasPassenger(final Predicate<Entity> test) {
@@ -5099,19 +5100,19 @@ public abstract class Entity
     }
 
     private Stream<Entity> getIndirectPassengersStream() {
-        if (this.passengers.isEmpty()) { return Stream.of(); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return this.passengers.stream().flatMap(Entity::getSelfAndPassengers);
     }
 
     @Override
     public Stream<Entity> getSelfAndPassengers() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return Stream.concat(Stream.of(this), this.getIndirectPassengersStream());
     }
 
     @Override
     public Stream<Entity> getPassengersAndSelf() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return Stream.concat(this.passengers.stream().flatMap(Entity::getPassengersAndSelf), Stream.of(this));
     }
 
@@ -5119,7 +5120,7 @@ public abstract class Entity
         // Paper start - optimise entity tracker
         final List<Entity> ret = new ArrayList<>();
 
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             return ret;
         }
 
@@ -5134,7 +5135,7 @@ public abstract class Entity
     }
 
     public boolean hasExactlyOnePlayerPassenger() {
-        if (this.passengers.isEmpty()) { return false; } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return false; } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return this.countPlayerPassengers() == 1;
     }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 7c2c24589e477a6515ac8da3ef51ac2580e73409..57c8f1089f8fd3cf581cf41f7bd8aa2c586f5137 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1168,7 +1168,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public boolean hasEffect(final Holder<MobEffect> effect) {
-        return this.activeEffects.containsKey(effect);
+        return !activeEffects.isEmpty() && this.activeEffects.containsKey(effect); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public @Nullable MobEffectInstance getEffect(final Holder<MobEffect> effect) {
@@ -4307,7 +4307,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     // CraftBukkit start - collidable API
     @Override
     public boolean canCollideWithBukkit(Entity entity) {
-        return this.isPushable() && this.collides != this.collidableExemptions.contains(entity.getUUID());
+        return this.isPushable() && this.collides != (!this.collidableExemptions.isEmpty() && this.collidableExemptions.contains(entity.getUUID())); // Leaf - Optimize map lookups with isEmpty check
     }
     // CraftBukkit end
 
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index cdf35d6e02df62334493b803e21f0c092675a93f..2f0e0d39941517d9ea69bbecc8610d2207791a02 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1768,7 +1768,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         if (!player.getItemInHand(hand).isEmpty()) {
             return InteractionResult.PASS;
         }
-        if (!passengers.isEmpty() || player.isPassenger()) {
+        if (passengers != Entity.EMPTY_PASSENGERS || player.isPassenger()) { // Leaf - Optimize map lookups with isEmpty check
             return InteractionResult.PASS;
         }
         if (this instanceof TamableAnimal tamable) {
diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index ce0f6871057e386ada6f5e12b4001f54d3105847..29f313ad83f60116ae03b4766c79614b2a31c738 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -471,7 +471,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
 
     private Entity getTopPassenger(final Entity vehicle, final int counter) {
         List<Entity> passengers = vehicle.getPassengers();
-        return counter != 1 && !passengers.isEmpty() ? this.getTopPassenger(passengers.getFirst(), counter - 1) : vehicle;
+        return counter != 1 && passengers != Entity.EMPTY_PASSENGERS ? this.getTopPassenger(passengers.getFirst(), counter - 1) : vehicle; // Leaf - Optimize map lookups with isEmpty check
     }
 
     @Override
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index a8c9aec27ab8190d86d59e40f278f0b394c02d16..7e6253a7979f3eabdf0b24c089ad50fedb6314ba 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -453,7 +453,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
         this.setDeltaMovement(Mth.clamp(movement.x, -maxSpeed, maxSpeed), movement.y, Mth.clamp(movement.z, -maxSpeed, maxSpeed));
 
         // Purpur start - Minecart settings and WASD controls
-        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && !passengers.isEmpty()) {
+        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && passengers != Entity.EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             Entity passenger = passengers.get(0);
             if (passenger instanceof net.minecraft.server.level.ServerPlayer player) {
                 net.minecraft.world.entity.player.Input lastClientInput = player.getLastClientInput();
