From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:24 +0900
Subject: [PATCH] cache eye block position


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index bd1b5ec838aa04c482d4bf186dd92214dadce4f0..d2be94132fafc92b969e1432447382eab4971f9d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -556,18 +556,13 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     // Purpur start - copied from Mob - API for any mob to burn daylight
     // Gale start - JettPack - optimize sun burn tick - cache eye blockpos
-    private BlockPos cached_eye_blockpos;
-    private net.minecraft.world.phys.Vec3 cached_position;
+    private BlockPos leaf$cached_eye_blockpos = BlockPos.ZERO;
+    private @Nullable Vec3 leaf$cached_position = null;
     // Gale end - JettPack - optimize sun burn tick - cache eye blockpos
     public boolean isSunBurnTick() {
         if (!this.level().isClientSide() && this.level().environmentAttributes().getValue(EnvironmentAttributes.MONSTERS_BURN, this.position())) {
             // Gale start - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
-            if (this.cached_position != this.position) {
-                this.cached_eye_blockpos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
-                this.cached_position = this.position;
-            }
-
-            float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue(cached_eye_blockpos); // Pass BlockPos to getBrightness
+            float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue();
 
             // Check brightness first
             if (lightLevelDependentMagicValue <= 0.5F) return false;
@@ -576,11 +571,28 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
             boolean flag = this.isInWater() || this.isInPowderSnow || this.wasInPowderSnow;
 
-            return !flag && this.level().canSeeSky(this.cached_eye_blockpos); // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
+            return !flag && this.level().canSeeSky(this.leaf$getEyeBlockPos()); // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
         }
 
         return false; // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos - diff on change
     }
+
+    // Leaf start - cache eye block pos
+    public BlockPos leaf$getEyeBlockPos() {
+        if (this.leaf$cached_position != this.position) {
+            double eyeY = this.getEyeY();
+            int x = Mth.floor(this.getX());
+            int y = Mth.floor(eyeY);
+            int z = Mth.floor(this.getZ());
+            BlockPos o = leaf$cached_eye_blockpos;
+            if (o.getX() != x || o.getY() != y || o.getZ() != z) {
+                this.leaf$cached_eye_blockpos = new BlockPos(x, y, z);
+            }
+            this.leaf$cached_position = this.position;
+        }
+        return this.leaf$cached_eye_blockpos;
+    }
+    // Leaf end - cache eye block pos
     // Purpur end - copied from Mob - API for any mob to burn daylight
 
     public Entity(EntityType<?> type, Level level) {
@@ -2145,8 +2157,9 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                 && abstractBoat.getBoundingBox().maxY >= eyeY
                 && abstractBoat.getBoundingBox().minY <= eyeY
         )) {
-            BlockPos blockPos = BlockPos.containing(this.getX(), eyeY, this.getZ());
-            FluidState fluidState = this.level().getFluidState(blockPos);
+            BlockPos blockPos = leaf$getEyeBlockPos(); // Leaf
+            FluidState fluidState = this.level().getFluidStateIfLoadedUnchecked(blockPos.getX(), blockPos.getY(), blockPos.getZ()); // Leaf
+            if (fluidState == null) { fluidState = this.level().getFluidState(blockPos); } // Leaf
             double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
             if (d > eyeY) {
                 // Leaf start - Optimize isEyeInFluid
@@ -2274,20 +2287,10 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     @Deprecated
     public float getLightLevelDependentMagicValue() {
-        return this.getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())); // Gale - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
-    }
-
-    // Gale start - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
-    /**
-     * @deprecated
-     */
-    @Deprecated
-    public float getLightLevelDependentMagicValue(BlockPos pos) {
-        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ())
-            ? this.level.getLightLevelDependentMagicValue(pos)
+        return this.level.hasChunkAt(this.getBlockX(), this.getBlockZ()) // Leaf
+            ? this.level.getLightLevelDependentMagicValue(leaf$getEyeBlockPos()) // Leaf
             : 0.0F;
     }
-    // Gale end - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
 
     public void absSnapTo(double x, double y, double z, float yRot, float xRot) {
         this.absSnapTo(x, y, z);
@@ -4541,6 +4544,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         EntityDimensions dimensions = this.getDimensions(pose);
         this.dimensions = dimensions;
         this.eyeHeight = dimensions.eyeHeight();
+        this.leaf$cached_position = null; // Leaf
     }
 
     public void refreshDimensions() {
@@ -4549,6 +4553,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         EntityDimensions dimensions = this.getDimensions(pose);
         this.dimensions = dimensions;
         this.eyeHeight = dimensions.eyeHeight();
+        this.leaf$cached_position = null; // Leaf
         this.reapplyPosition();
         boolean flag = dimensions.width() <= 4.0F && dimensions.height() <= 4.0F;
         if (!this.level.isClientSide()
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index a8bf96aafee76732cd49834e44235ab6620fa1c3..372a767721d23a4d5018f0d5ab3d8c96df4b3dbf 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -471,7 +471,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             if (this.isEyeInWater() // Leaf - Optimize isEyeInFluid
-                && !serverLevel1.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
+                && !serverLevel1.getBlockState(leaf$getEyeBlockPos()).is(Blocks.BUBBLE_COLUMN)) { // Leaf
                 boolean flag1 = !this.canBreatheUnderwater()
                     && !MobEffectUtil.hasWaterBreathing(this)
                     && (!flag || !((Player)this).getAbilities().invulnerable);
