From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 6 Jun 2025 05:44:21 +0800
Subject: [PATCH] do not log invalid flatten text component parse


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
index ca9e3667b420f54f42b358068c102fe0cae4102d..88901f4783a85cc764bb2db20aed69f96b2bc8c3 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
@@ -189,7 +189,7 @@ public final class V4290 {
                         return ret;
                     }
                 } catch (final JsonParseException ex) {
-                    LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex);
+                    if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.invalidLegacyTextComponent) LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex); // Leaf - do not log invalid flatten text component parse
                 }
 
                 return null;
