From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 6 Jun 2025 05:44:21 +0800
Subject: [PATCH] do not log invalid flatten text component parse


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
index 4197a5e89aec77f2893f0f118a927b96d5377ec1..de99a05bbe3e8dac2d6307eb53d954d77904a0c4 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
@@ -191,7 +191,7 @@ public final class V4290 {
                         return ret;
                     }
                 } catch (final JsonParseException ex) {
-                    LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex);
+                    if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.invalidLegacyTextComponent) LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex); // Leaf - do not log invalid flatten text component parse
                 }
 
                 return null;
