From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 9 Oct 2025 18:32:43 +0900
Subject: [PATCH] optimize canHoldAnyFluid


diff --git a/net/minecraft/world/level/material/FlowingFluid.java b/net/minecraft/world/level/material/FlowingFluid.java
index 26688d1401f76e5751d9f7149b5a2016c10d87ec..9f7d3e2499d699f1dd56674241d7189b3b40ded5 100644
--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -495,7 +495,7 @@ public abstract class FlowingFluid extends Fluid {
         final FluidState testFluidState
     ) {
         return !this.isSourceBlockOfThisType(testFluidState)
-            && canHoldAnyFluid(testState)
+            && (testState.tagFlag & org.dreeam.leaf.util.BlockMasks.CAN_HOLD_ANY_FLUID) != 0 // Leaf - optimize canHoldAnyFluid
             && canPassThroughWall(direction, level, sourcePos, sourceState, testPos, testState);
     }
 
@@ -577,7 +577,7 @@ public abstract class FlowingFluid extends Fluid {
     }
 
     private static boolean canHoldFluid(final BlockGetter level, final BlockPos pos, final BlockState state, final Fluid newFluid) {
-        return canHoldAnyFluid(state) && canHoldSpecificFluid(level, pos, state, newFluid);
+        return (state.tagFlag & org.dreeam.leaf.util.BlockMasks.CAN_HOLD_ANY_FLUID) != 0 && canHoldSpecificFluid(level, pos, state, newFluid); // Leaf - optimize canHoldAnyFluid
     }
 
     private static boolean canHoldSpecificFluid(final BlockGetter level, final BlockPos pos, final BlockState state, final Fluid newFluid) {
