From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 14:55:37 +0900
Subject: [PATCH] optimize get chunk


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 4e1748020c31b80b52f6f52fe90eb0f7099651d9..28fdc5b85c22fd03bc2e4b4c6ae3e3524fd039f6 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1409,12 +1409,17 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
             }
         }
         // CraftBukkit end
-        if (this.isOutsideBuildHeight(pos)) {
+        // Leaf start - optimize get chunk
+        int x = pos.getX();
+        int y = pos.getY();
+        int z = pos.getZ();
+        if (this.isOutsideBuildHeight(y)) {
             return Blocks.VOID_AIR.defaultBlockState();
         } else {
-            ChunkAccess chunk = this.getChunk(pos.getX() >> 4, pos.getZ() >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
-            return chunk.getBlockState(pos);
+            ChunkAccess chunk = this.getChunk(x >> 4, z >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
+            return chunk.getBlockState(x, y, z);
         }
+        // Leaf end - optimize get chunk
     }
 
     @Override
