From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:24 +0900
Subject: [PATCH] optimize eye block position


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index bd1b5ec838aa04c482d4bf186dd92214dadce4f0..67cc984d099b02172a88a1cecbbe0cd514f0e801 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -555,19 +555,13 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     // Purpur end - Add canSaveToDisk to Entity
 
     // Purpur start - copied from Mob - API for any mob to burn daylight
-    // Gale start - JettPack - optimize sun burn tick - cache eye blockpos
-    private BlockPos cached_eye_blockpos;
-    private net.minecraft.world.phys.Vec3 cached_position;
-    // Gale end - JettPack - optimize sun burn tick - cache eye blockpos
     public boolean isSunBurnTick() {
         if (!this.level().isClientSide() && this.level().environmentAttributes().getValue(EnvironmentAttributes.MONSTERS_BURN, this.position())) {
             // Gale start - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
-            if (this.cached_position != this.position) {
-                this.cached_eye_blockpos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
-                this.cached_position = this.position;
-            }
-
-            float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue(cached_eye_blockpos); // Pass BlockPos to getBrightness
+            // Leaf start - optimize eye block position
+            final BlockPos blockPos = new BlockPos(this.blockPosition.getX(), Mth.floor(this.getEyeY()), this.blockPosition.getZ());
+            float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue(blockPos);
+            // Leaf end - optimize eye block position
 
             // Check brightness first
             if (lightLevelDependentMagicValue <= 0.5F) return false;
@@ -576,7 +570,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
             boolean flag = this.isInWater() || this.isInPowderSnow || this.wasInPowderSnow;
 
-            return !flag && this.level().canSeeSky(this.cached_eye_blockpos); // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
+            return !flag && this.level().canSeeSky(blockPos); // Leaf - optimize eye block position
         }
 
         return false; // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos - diff on change
@@ -2145,9 +2139,15 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                 && abstractBoat.getBoundingBox().maxY >= eyeY
                 && abstractBoat.getBoundingBox().minY <= eyeY
         )) {
-            BlockPos blockPos = BlockPos.containing(this.getX(), eyeY, this.getZ());
-            FluidState fluidState = this.level().getFluidState(blockPos);
-            double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
+            // Leaf start - optimize eye block position
+            final int blockX = this.blockPosition.getX();
+            final int blockZ = this.blockPosition.getZ();
+            final int blockY = Mth.floor(eyeY);
+            final Level level = this.level();
+            FluidState fluidState = level.getFluidStateIfLoadedUnchecked(blockX, blockY, blockZ);
+            if (fluidState == null) { fluidState = level.getFluidState(new BlockPos(blockX, blockY, blockZ)); }
+            final double d = fluidState.isEmpty() ? blockY : blockY + fluidState.getHeight(level, new BlockPos(blockX, blockY, blockZ));
+            // Leaf end - optimize eye block position
             if (d > eyeY) {
                 // Leaf start - Optimize isEyeInFluid
                 this.isInWaterOrLava = fluidState.isEmpty() ? 0 : fluidState.is(FluidTags.WATER) ? 1 : 2;
@@ -2272,22 +2272,25 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
     }
 
+    // Leaf start - optimize eye block position
     @Deprecated
     public float getLightLevelDependentMagicValue() {
-        return this.getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())); // Gale - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
+        final Level level = this.level();
+        final int blockX = this.blockPosition.getX();
+        final int blockZ = this.blockPosition.getZ();
+        return level.hasChunkAt(blockX, blockZ)
+            ? level.getLightLevelDependentMagicValue(new BlockPos(blockX, Mth.floor(this.getEyeY()), blockZ))
+            : 0.0F;
     }
-
-    // Gale start - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
-    /**
-     * @deprecated
-     */
+    // Copied from getLightLevelDependentMagicValue, do not reuse common code
     @Deprecated
     public float getLightLevelDependentMagicValue(BlockPos pos) {
-        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ())
-            ? this.level.getLightLevelDependentMagicValue(pos)
+        final Level level = this.level();
+        return level.hasChunkAt(pos.getX(), pos.getZ())
+            ? level.getLightLevelDependentMagicValue(pos)
             : 0.0F;
     }
-    // Gale end - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
+    // Leaf end - optimize eye block position
 
     public void absSnapTo(double x, double y, double z, float yRot, float xRot) {
         this.absSnapTo(x, y, z);
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 19ffea9465b7674c278b2b1c75323704df5baaea..1848f2537502f8a3fecbd381ae022c10a9fa4971 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -471,7 +471,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             if (this.isEyeInWater() // Leaf - Optimize isEyeInFluid
-                && !serverLevel1.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
+                && !isEyeInBubbleColumn(serverLevel1)) { // Leaf - optimize eye block position
                 boolean flag1 = !this.canBreatheUnderwater()
                     && !MobEffectUtil.hasWaterBreathing(this)
                     && (!flag || !((Player)this).getAbilities().invulnerable);
@@ -539,6 +539,19 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         profilerFiller.pop();
     }
 
+    // Leaf start - optimize eye block position
+    private boolean isEyeInBubbleColumn(ServerLevel level) {
+        final int x = this.blockPosition.getX();
+        final int z = this.blockPosition.getZ();
+        final int y = Mth.floor(this.getEyeY());
+        BlockState blockState = level.getBlockStateIfLoadedUnchecked(x, y, z);
+        if (blockState == null) {
+            blockState = level.getBlockState(new BlockPos(x, y, z));
+        }
+        return blockState.is(Blocks.BUBBLE_COLUMN);
+    }
+    // Leaf end - optimize eye block position
+
     protected boolean shouldTakeDrowningDamage() {
         return this.getAirSupply() <= -this.level().purpurConfig.drowningDamageInterval; // Purpur - Drowning Settings
     }
