From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Overwrite987 <artem.korolev_2003@mail.ru>
Date: Mon, 28 Jul 2025 03:35:42 +0300
Subject: [PATCH] Do not create fire if explosion was cancelled


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 7426a7f4bf17277b0355185b58973140dab3c7b9..f6fbea16682e2a01bd0d2cdc6ed4a0934aa4ee7f 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -673,6 +673,9 @@ public class ServerExplosion implements Explosion {
     }
 
     private void createFire(List<BlockPos> blocks) {
+        if (this.wasCanceled) {
+            return;
+        }
         for (BlockPos blockPos : blocks) {
             if (this.level.random.nextInt(3) == 0 && this.level.getBlockState(blockPos).isAir() && this.level.getBlockState(blockPos.below()).isSolidRender()) {
                 // CraftBukkit start - Ignition by explosion
