From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 7 Aug 2025 00:22:50 +0800
Subject: [PATCH] Bump netty to 4.2.x


diff --git a/net/minecraft/server/network/ServerConnectionListener.java b/net/minecraft/server/network/ServerConnectionListener.java
index bd07e6a5aa1883786d789ea71711a0c0c0a95c26..a6396bf65bcf73899d35565f47950db0fb7c333e 100644
--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -48,12 +48,19 @@ import org.slf4j.Logger;
 
 public class ServerConnectionListener {
     private static final Logger LOGGER = LogUtils.getLogger();
+    // Leaf start - Bump netty to 4.2.x
+    /*
     public static final Supplier<NioEventLoopGroup> SERVER_EVENT_GROUP = Suppliers.memoize(
         () -> new NioEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Server IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()) // Paper
     );
     public static final Supplier<EpollEventLoopGroup> SERVER_EPOLL_EVENT_GROUP = Suppliers.memoize(
         () -> new EpollEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Epoll Server IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()) // Paper
     );
+    */
+    public static final Supplier<io.netty.channel.IoHandlerFactory> NIO_IO_HANDLE_FACTORY = io.netty.channel.nio.NioIoHandler::newFactory;
+    public static final Supplier<io.netty.channel.IoHandlerFactory> EPOLL_IO_HANDLE_FACTORY = io.netty.channel.epoll.EpollIoHandler::newFactory;
+    public static final Supplier<io.netty.channel.IoHandlerFactory> IO_URING_IO_HANDLE_FACTORY = io.netty.channel.uring.IoUringIoHandler::newFactory;
+    // Leaf end - Bump netty to 4.2.x
     final MinecraftServer server;
     public volatile boolean running;
     private final List<ChannelFuture> channels = Collections.synchronizedList(Lists.newArrayList());
@@ -87,7 +94,19 @@ public class ServerConnectionListener {
         synchronized (this.channels) {
             Class<? extends io.netty.channel.ServerChannel> clazz; // Paper - Unix domain socket support
             EventLoopGroup eventLoopGroup;
-            if (Epoll.isAvailable() && this.server.isEpollEnabled()) {
+            // Leaf start - Bump netty to 4.2.x
+            if (io.netty.channel.uring.IoUring.isAvailable() && org.dreeam.leaf.config.modules.network.NetworkIoModel.useIoUring) {
+                // Paper start - Unix domain socket support
+                if (address instanceof io.netty.channel.unix.DomainSocketAddress) {
+                    clazz = io.netty.channel.uring.IoUringServerDomainSocketChannel.class;
+                } else {
+                    clazz = io.netty.channel.uring.IoUringServerSocketChannel.class;
+                }
+                // Paper end - Unix domain socket supp
+                eventLoopGroup = createEventLoopGroup("IoUring", IO_URING_IO_HANDLE_FACTORY);
+                LOGGER.info("Using io_uring channel type");
+            } else if (Epoll.isAvailable() && org.dreeam.leaf.config.modules.network.NetworkIoModel.useEpoll) {
+                // Leaf end - Bump netty to 4.2.x
                 // Paper start - Unix domain socket support
                 if (address instanceof io.netty.channel.unix.DomainSocketAddress) {
                     clazz = io.netty.channel.epoll.EpollServerDomainSocketChannel.class;
@@ -95,12 +114,12 @@ public class ServerConnectionListener {
                 clazz = EpollServerSocketChannel.class;
                 }
                 // Paper end - Unix domain socket support
-                eventLoopGroup = SERVER_EPOLL_EVENT_GROUP.get();
+                eventLoopGroup = createEventLoopGroup("Epoll", EPOLL_IO_HANDLE_FACTORY); // Leaf - Bump netty to 4.2.x
                 LOGGER.info("Using epoll channel type");
             } else {
                 clazz = NioServerSocketChannel.class;
-                eventLoopGroup = SERVER_EVENT_GROUP.get();
-                LOGGER.info("Using default channel type");
+                eventLoopGroup = createEventLoopGroup("Nio", NIO_IO_HANDLE_FACTORY); // Leaf - Bump netty to 4.2.x
+                LOGGER.info("Using nio channel type (default)"); // Leaf - Bump netty to 4.2.x
             }
 
             // Paper start - Warn people with console access that HAProxy is in use.
@@ -195,6 +214,19 @@ public class ServerConnectionListener {
     }
     // CraftBukkit end
 
+    // Leaf start - Bump netty to 4.2.x
+    public EventLoopGroup createEventLoopGroup(final String name, final Supplier<io.netty.channel.IoHandlerFactory> ioHandlerFactorySupplier) {
+        return new io.netty.channel.MultiThreadIoEventLoopGroup(
+            0,
+            new ThreadFactoryBuilder().setNameFormat("Netty " + name + " Server IO #%d")
+                .setDaemon(true)
+                .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER))
+                .build(),
+            ioHandlerFactorySupplier.get()
+        );
+    }
+    // Leaf end - Bump netty to 4.2.x
+
     public SocketAddress startMemoryChannel() {
         ChannelFuture channelFuture;
         synchronized (this.channels) {
@@ -215,7 +247,7 @@ public class ServerConnectionListener {
                         }
                     }
                 )
-                .group(SERVER_EVENT_GROUP.get())
+                //.group(SERVER_EVENT_GROUP.get()) // Leaf - Bump netty to 4.2.x - not used, comment
                 .localAddress(LocalAddress.ANY)
                 .bind()
                 .syncUninterruptibly();
