From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 14:55:37 +0900
Subject: [PATCH] optimize get chunk


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 525ccfc7c83c60331c6e8405d944c16437874be4..aa32cf0bde154e16ac2e6499a43edd8e291e164c 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1409,12 +1409,17 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
             }
         }
         // CraftBukkit end
-        if (this.isOutsideBuildHeight(pos)) {
+        // Leaf start - optimize get chunk
+        int x = pos.getX();
+        int y = pos.getY();
+        int z = pos.getZ();
+        if (this.isOutsideBuildHeight(y)) {
             return Blocks.VOID_AIR.defaultBlockState();
         } else {
-            ChunkAccess chunk = this.getChunk(pos.getX() >> 4, pos.getZ() >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
-            return chunk.getBlockState(pos);
+            ChunkAccess chunk = this.getChunk(x >> 4, z >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
+            return chunk.getBlockState(x, y, z);
         }
+        // Leaf end - optimize get chunk
     }
 
     @Override
