From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:24:49 +0200
Subject: [PATCH] Inline randomBetweenInclusive


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 3a54fa25bf491902235392168a00c17378e01c0d..c750c7822808f26a515075bf2f05137066ad6dfb 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -661,7 +661,11 @@ public final class NaturalSpawner {
         int i = pos.getMinBlockX() + level.random.nextInt(16);
         int i1 = pos.getMinBlockZ() + level.random.nextInt(16);
         int i2 = chunk.getHeight(Heightmap.Types.WORLD_SURFACE, i, i1) + 1;
-        int i3 = Mth.randomBetweenInclusive(level.random, level.getMinY(), i2);
+        // Leaf start - Inline randomBetweenInclusive
+        int minY = level.getMinY();
+        int range = i2 - minY + 1;
+        int i3 = level.random.nextInt(range) + minY;
+        // Leaf end - Inline randomBetweenInclusive
         return new BlockPos(i, i3, i1);
     }
 
