From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Wed, 30 Jul 2025 23:18:41 +0200
Subject: [PATCH] Custom map for heightmap.type


diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index 585765134da9d99e994e9ac250c7c757d33b3698..d3021c4236657926d1464fb628c5d10db63e22ae 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -75,7 +75,7 @@ public abstract class ChunkAccess implements BiomeManager.NoiseBiomeSource, Ligh
     protected final UpgradeData upgradeData;
     @Nullable
     protected BlendingData blendingData;
-    public final Map<Heightmap.Types, Heightmap> heightmaps = Maps.newEnumMap(Heightmap.Types.class);
+    public final Map<Heightmap.Types, Heightmap> heightmaps = new org.dreeam.leaf.util.map.FastHeightmapMap();
     // Paper - rewrite chunk system
     private final Map<Structure, StructureStart> structureStarts = new it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap<>(); // Leaf - optimize structure map
     private final Map<Structure, LongSet> structuresRefences = new it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap<>(); // Leaf - optimize structure map
diff --git a/net/minecraft/world/level/levelgen/Heightmap.java b/net/minecraft/world/level/levelgen/Heightmap.java
index f7b8637680e654dc0b3e8850d65081a1c565bb69..86c563dfa3052484de310e212a3faa0a458a0834 100644
--- a/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/net/minecraft/world/level/levelgen/Heightmap.java
@@ -157,7 +157,7 @@ public class Heightmap {
         public static final Codec<Heightmap.Types> CODEC = StringRepresentable.fromEnum(Heightmap.Types::values);
         private static final IntFunction<Heightmap.Types> BY_ID = ByIdMap.continuous(types -> types.id, values(), ByIdMap.OutOfBoundsStrategy.ZERO);
         public static final StreamCodec<ByteBuf, Heightmap.Types> STREAM_CODEC = ByteBufCodecs.idMapper(BY_ID, types -> types.id);
-        private final int id;
+        public final int id;
         private final String serializationKey;
         private final Heightmap.Usage usage;
         private final Predicate<BlockState> isOpaque;
