From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 17 Aug 2025 14:46:30 +0200
Subject: [PATCH] Skip PreCreatureSpawnEvent if no listeners


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index c750c7822808f26a515075bf2f05137066ad6dfb..a6dc1a67a09510195ce4d0bb30a44df3409c8f01 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -541,17 +541,35 @@ public final class NaturalSpawner {
         LevelChunk chunk
     ) {
         EntityType<?> entityType = data.type();
-        // Paper start - PreCreatureSpawnEvent
-        com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
-            org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
-            org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(entityType), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
-        );
-        if (!event.callEvent()) {
-            if (event.shouldAbortSpawn()) {
-                return PreSpawnStatus.ABORT;
+
+        // Leaf start - Skip PreCreatureSpawnEvent if no listeners
+        boolean shouldAbort = false;
+        boolean cancelled = false;
+
+        if (com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            // Paper start - PreCreatureSpawnEvent
+            com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
+                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
+                org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(entityType),
+                org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
+            );
+            if (!event.callEvent()) {
+                if (event.shouldAbortSpawn()) {
+                    shouldAbort = true;
+                }
+                cancelled = true;
             }
+            // Paper end - PreCreatureSpawnEvent
+        }
+
+        if (shouldAbort) {
+            return PreSpawnStatus.ABORT;
+        }
+        if (cancelled) {
             return PreSpawnStatus.CANCELLED;
         }
+        // Leaf end - Skip PreCreatureSpawnEvent if no listeners
+
         final boolean success = entityType.getCategory() != MobCategory.MISC
             // Paper end - PreCreatureSpawnEvent
             && (
@@ -563,7 +581,7 @@ public final class NaturalSpawner {
             && SpawnPlacements.isSpawnPositionOk(entityType, level, pos)
             && SpawnPlacements.checkSpawnRules(entityType, level, EntitySpawnReason.NATURAL, pos, level.random)
             && level.noCollision(entityType.getSpawnAABB(pos.getX() + 0.5, pos.getY(), pos.getZ() + 0.5));
-        return success ? PreSpawnStatus.SUCCESS : PreSpawnStatus.FAIL; // Paper - PreCreatureSpawnEvent
+        return success ? PreSpawnStatus.SUCCESS : PreSpawnStatus.FAIL;
     }
     // Leaf end - optimize mob spawning
 
