From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Fri, 15 Aug 2025 13:35:07 +0200
Subject: [PATCH] batch heightmap updates


diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index c3ecdd80efad340c9fa0ea6913db9a3362b53ad4..c0b6f79a3be7a683a1f50875f9e9633d578cbd16 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -426,10 +426,9 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
                 }
                 // Leaf end - optimize random tick
                 Block block = state.getBlock();
-                this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR).update(i, y, i2, state);
-                this.heightmaps.get(Heightmap.Types.WORLD_SURFACE).update(i, y, i2, state);
+                // Leaf start - batch heighmap updates
+                org.dreeam.leaf.world.block.BatchedHeightMap.batchUpdate(this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING), this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES), this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR), this.heightmaps.get(Heightmap.Types.WORLD_SURFACE), this, i, y, i2, state);
+                // Leaf end - batch heighmap updates
                 boolean hasOnlyAir1 = section.hasOnlyAir();
                 if (hasOnlyAir != hasOnlyAir1) {
                     this.level.getChunkSource().getLightEngine().updateSectionStatus(pos, hasOnlyAir1);
diff --git a/net/minecraft/world/level/levelgen/Heightmap.java b/net/minecraft/world/level/levelgen/Heightmap.java
index 86c563dfa3052484de310e212a3faa0a458a0834..c0931ef29455169e6fc2b1300ccd1339ebe89964 100644
--- a/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/net/minecraft/world/level/levelgen/Heightmap.java
@@ -119,7 +119,7 @@ public class Heightmap {
         return this.data.get(index) + this.chunk.getMinY();
     }
 
-    private void setHeight(int x, int z, int value) {
+    public void setHeight(int x, int z, int value) {
         this.data.set(getIndex(x, z), value - this.chunk.getMinY());
     }
 
