From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:30:37 +0900
Subject: [PATCH] optimize tickEffects


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 6d7ac82f403de76cba3f7e13e63ad2735966d07e..9436b20485b3e864c316c650791c6470bc530e97 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -977,6 +977,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     protected void tickEffects() {
         if (this.level() instanceof ServerLevel serverLevel) {
+            if (!this.activeEffects.isEmpty()) { // Leaf - optimize tickEffects
             Iterator<Holder<MobEffect>> iterator = this.activeEffects.keySet().iterator();
 
             this.isTickingEffects = true; // CraftBukkit
@@ -1002,6 +1003,8 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
             // CraftBukkit start
             this.isTickingEffects = false;
+            } // Leaf - optimize tickEffects
+            if (!effectsToProcess.isEmpty()) { // Leaf - optimize tickEffects
             for (ProcessableEffect effect : this.effectsToProcess) {
                 if (effect.effect != null) {
                     this.addEffect(effect.effect, effect.cause);
@@ -1010,6 +1013,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
             }
             this.effectsToProcess.clear();
+            } // Leaf - optimize tickEffects
             // CraftBukkit end
         } else {
             for (MobEffectInstance mobEffectInstance1 : this.activeEffects.values()) {
