From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 21 Mar 2026 15:25:23 +0900
Subject: [PATCH] optimize PathNavigation#followThePath

To prevent unnecessary BlockPos allocation in `asBlockPos` under `getNextNodePos`

diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 54ac707d37e2701bb2563df06b9430a858dbb700..0336742e0886eb06beacaa2d8b1975d77b383142 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -325,14 +325,16 @@ public abstract class PathNavigation {
         if (!this.path.isProcessed()) return; // Kaiiju - petal - async path processing - skip if not processed
         Vec3 mobPos = this.getTempMobPos();
         this.maxDistanceToWaypoint = this.mob.getBbWidth() > 0.75F ? this.mob.getBbWidth() / 2.0F : 0.75F - this.mob.getBbWidth() / 2.0F;
-        Vec3i currentNodePos = this.path.getNextNodePos();
-        double xDistance = Math.abs(this.mob.getX() - (currentNodePos.getX() + 0.5));
-        double yDistance = Math.abs(this.mob.getY() - currentNodePos.getY());
-        double zDistance = Math.abs(this.mob.getZ() - (currentNodePos.getZ() + 0.5));
+        // Leaf start - optimize PathNavigation#followThePath
+        Node currentNode = this.path.getNextNode();
+        double xDistance = Math.abs(this.mob.getX() - (currentNode.x + 0.5));
+        double yDistance = Math.abs(this.mob.getY() - currentNode.y);
+        double zDistance = Math.abs(this.mob.getZ() - (currentNode.z + 0.5));
+        // Leaf end - optimize PathNavigation#followThePath
         boolean isCloseEnoughToCurrentNode = xDistance < this.maxDistanceToWaypoint
             && zDistance < this.maxDistanceToWaypoint
             && yDistance < this.getMaxVerticalDistanceToWaypoint();
-        if (isCloseEnoughToCurrentNode || this.canCutCorner(this.path.getNextNode().type) && this.shouldTargetNextNodeInDirection(mobPos)) {
+        if (isCloseEnoughToCurrentNode || this.canCutCorner(currentNode.type) && this.shouldTargetNextNodeInDirection(mobPos)) { // Leaf - optimize PathNavigation#followThePath
             this.path.advance();
         }
 
