From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:31:35 +0900
Subject: [PATCH] Only update frozen ticks if changed


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 5fa5e34587e9863ce95d9b3ad65742486a503d4e..bc8c3bfdc5c4d2a30453947a6e1a825ee14860d3 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3915,7 +3915,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         if (this.level() instanceof ServerLevel serverLevel) {
             profilerFiller.push("freezing");
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // Leaf start - Only update frozen ticks if changed
+                int ticksForzen = this.getTicksFrozen();
+                int newTicksForzen = Math.max(0, ticksForzen - 2);
+                if (ticksForzen != newTicksForzen) {
+                    this.setTicksFrozen(newTicksForzen);
+                }
+                // Leaf end - Only update frozen ticks if changed
             }
 
             this.removeFrost();
