From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Don't load POI for competitor scan


diff --git a/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java b/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
index c507cbb1b0d89f8fe5a00e67eea510f7516fe040..4aa67879f3e78f92e51e4c32a3aa69438b30ac63 100644
--- a/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
+++ b/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
@@ -20,7 +20,7 @@ public class PoiCompetitorScan {
                     (jobSite, nearestEntities) -> (level, body, timestamp) -> {
                         GlobalPos pos = i.get(jobSite);
                         level.getPoiManager()
-                            .getType(pos.pos())
+                            .getTypeIfLoaded(pos.pos()) // Leaf - Don't load POI for competitor scan
                             .ifPresent(
                                 // Paper start - Improve performance of PoiCompetitorScan by unrolling stream
                                 // The previous logic used Stream#reduce to simulate a form of single-iteration bubble sort
diff --git a/net/minecraft/world/entity/ai/village/poi/PoiManager.java b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
index 3b07e731e8d16c2b71cb55b85163029cf56fd02e..bfb995b4695847d2ea6e0c38069f633137ed458d 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -414,6 +414,12 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         return this.getOrLoad(SectionPos.asLong(pos)).flatMap(section -> section.getType(pos));
     }
 
+    // Leaf start - Don't load POI for competitor scan
+    public Optional<Holder<PoiType>> getTypeIfLoaded(final BlockPos pos) {
+        return this.get(SectionPos.asLong(pos)).flatMap(section -> section.getType(pos));
+    }
+    // Leaf end - Don't load POI for competitor scan
+
     @VisibleForDebug
     public @Nullable DebugPoiInfo getDebugPoiInfo(final BlockPos pos) {
         return this.getOrLoad(SectionPos.asLong(pos)).flatMap(section -> section.getDebugPoiInfo(pos)).orElse(null);
