From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Mon, 9 Feb 2026 12:37:26 +0300
Subject: [PATCH] Optimize map lookups with isEmpty check


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1f267f763a4686504634c62717fa8124090f25eb..2f57e12d0a74b3b044f128c93c85669ed4713b32 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -266,6 +266,7 @@ public abstract class Entity
     private int id = 0;
     public boolean blocksBuilding;
     public ImmutableList<Entity> passengers = ImmutableList.of();
+    public static final ImmutableList<Entity> EMPTY_PASSENGERS = ImmutableList.of(); // Leaf - Optimize map lookups with isEmpty check
     protected int boardingCooldown;
     private @Nullable Entity vehicle;
     private Level level;
@@ -484,7 +485,7 @@ public abstract class Entity
 
     @Override
     public final boolean moonrise$hasAnyPlayerPassengers() {
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             return false;
         }
         return this.getIndirectPassengersStream().anyMatch((entity) -> entity instanceof Player);
@@ -3576,7 +3577,7 @@ public abstract class Entity
             throw new IllegalStateException("Use x.startRiding(y), not y.addPassenger(x)");
         }
 
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             this.passengers = ImmutableList.of(passenger);
         } else {
             List<Entity> newPassengers = Lists.newArrayList(this.passengers);
@@ -3654,7 +3655,7 @@ public abstract class Entity
     }
 
     protected boolean canAddPassenger(final Entity passenger) {
-        return this.passengers.isEmpty();
+        return this.passengers == EMPTY_PASSENGERS; // Leaf - Optimize map lookups with isEmpty check
     }
 
     protected boolean couldAcceptPassenger() {
@@ -3843,7 +3844,7 @@ public abstract class Entity
     }
 
     public boolean isVehicle() {
-        return !this.passengers.isEmpty();
+        return this.passengers != EMPTY_PASSENGERS; // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean dismountsUnderwater() {
@@ -4886,11 +4887,11 @@ public abstract class Entity
     }
 
     public @Nullable Entity getFirstPassenger() {
-        return this.passengers.isEmpty() ? null : this.passengers.get(0);
+        return this.passengers == EMPTY_PASSENGERS ? null : this.passengers.get(0); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean hasPassenger(final Entity entity) {
-        return this.passengers.contains(entity);
+        return passengers != EMPTY_PASSENGERS && this.passengers.contains(entity); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean hasPassenger(final Predicate<Entity> test) {
@@ -4904,19 +4905,19 @@ public abstract class Entity
     }
 
     private Stream<Entity> getIndirectPassengersStream() {
-        if (this.passengers.isEmpty()) { return Stream.of(); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return this.passengers.stream().flatMap(Entity::getSelfAndPassengers);
     }
 
     @Override
     public Stream<Entity> getSelfAndPassengers() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return Stream.concat(Stream.of(this), this.getIndirectPassengersStream());
     }
 
     @Override
     public Stream<Entity> getPassengersAndSelf() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return Stream.concat(this.passengers.stream().flatMap(Entity::getPassengersAndSelf), Stream.of(this));
     }
 
@@ -4924,7 +4925,7 @@ public abstract class Entity
         // Paper start - optimise entity tracker
         final List<Entity> ret = new ArrayList<>();
 
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             return ret;
         }
 
@@ -4939,7 +4940,7 @@ public abstract class Entity
     }
 
     public boolean hasExactlyOnePlayerPassenger() {
-        if (this.passengers.isEmpty()) { return false; } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return false; } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return this.countPlayerPassengers() == 1;
     }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 42c9fdd09e933c04cbbbf40e9f7ab473236d2f9c..d5e860865ff1a0ba3a457397b6e6cfc2dd37ee13 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1177,7 +1177,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public boolean hasEffect(final Holder<MobEffect> effect) {
-        return this.activeEffects.containsKey(effect);
+        return !activeEffects.isEmpty() && this.activeEffects.containsKey(effect); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public @Nullable MobEffectInstance getEffect(final Holder<MobEffect> effect) {
@@ -4329,7 +4329,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     // CraftBukkit start - collidable API
     @Override
     public boolean canCollideWithBukkit(Entity entity) {
-        return this.isPushable() && this.collides != this.collidableExemptions.contains(entity.getUUID());
+        return this.isPushable() && this.collides != (!this.collidableExemptions.isEmpty() && this.collidableExemptions.contains(entity.getUUID())); // Leaf - Optimize map lookups with isEmpty check
     }
     // CraftBukkit end
 
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index e175bb2b4fd8f80996e14a409ea3b4bb6e1de10f..484348e23b8710d07a6a33817cb80293ee905646 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1754,7 +1754,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         if (!player.getItemInHand(hand).isEmpty()) {
             return InteractionResult.PASS;
         }
-        if (!passengers.isEmpty() || player.isPassenger()) {
+        if (passengers != Entity.EMPTY_PASSENGERS || player.isPassenger()) { // Leaf - Optimize map lookups with isEmpty check
             return InteractionResult.PASS;
         }
         if (this instanceof TamableAnimal tamable) {
diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index 2eadc3b6680066581a496a0fed38575428d741eb..403831747370295e801177ac925ffd2a470b0a4a 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -476,7 +476,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
 
     private Entity getTopPassenger(final Entity vehicle, final int counter) {
         List<Entity> passengers = vehicle.getPassengers();
-        return counter != 1 && !passengers.isEmpty() ? this.getTopPassenger(passengers.getFirst(), counter - 1) : vehicle;
+        return counter != 1 && passengers != Entity.EMPTY_PASSENGERS ? this.getTopPassenger(passengers.getFirst(), counter - 1) : vehicle; // Leaf - Optimize map lookups with isEmpty check
     }
 
     @Override
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index c635999ec8d5e0508cd66e0c4982692450963235..9b2f1834f364b30a48ba13198af485e2bf8a1cd1 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -447,7 +447,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
         this.setDeltaMovement(Mth.clamp(movement.x, -maxSpeed, maxSpeed), movement.y, Mth.clamp(movement.z, -maxSpeed, maxSpeed));
 
         // Purpur start - Minecart settings and WASD controls
-        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && !passengers.isEmpty()) {
+        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && passengers != Entity.EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             Entity passenger = passengers.get(0);
             if (passenger instanceof net.minecraft.server.level.ServerPlayer player) {
                 net.minecraft.world.entity.player.Input lastClientInput = player.getLastClientInput();
