From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:24 +0900
Subject: [PATCH] optimize updateFluidOnEyes


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 5a9fbb3be4ce10a4f068a37324cbbe7f9e6828fe..f201ef3e8ba8e96e74beb0ad0b5468e7ebbe67de 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2142,8 +2142,9 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 && abstractBoat.getBoundingBox().maxY >= eyeY
                 && abstractBoat.getBoundingBox().minY <= eyeY
         )) {
-            BlockPos blockPos = BlockPos.containing(this.getX(), eyeY, this.getZ());
-            FluidState fluidState = this.level().getFluidState(blockPos);
+            BlockPos blockPos = leaf$getEyeBlockPos(); // Leaf
+            FluidState fluidState = this.level().getFluidStateIfLoadedUnchecked(blockPos.getX(), blockPos.getY(), blockPos.getZ()); // Leaf
+            if (fluidState == null) { fluidState = this.level().getFluidState(blockPos); } // Leaf
             double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
             if (d > eyeY) {
                 // Leaf start - Optimize isEyeInFluid
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 5b01275395565648a1f4e2bbcb40346e32dcfeca..adc54d54f6214512ad5b3803d21ca0ca36b05a5c 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -464,7 +464,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             if (this.isEyeInWater() // Leaf - Optimize isEyeInFluid
-                && !serverLevel1.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
+                && !serverLevel1.getBlockState(leaf$getEyeBlockPos()).is(Blocks.BUBBLE_COLUMN)) { // Leaf
                 boolean flag1 = !this.canBreatheUnderwater()
                     && !MobEffectUtil.hasWaterBreathing(this)
                     && (!flag || !((Player)this).getAbilities().invulnerable);
