From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Mon, 11 Aug 2025 02:19:41 +0900
Subject: [PATCH] optimize collision shape


diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 9caac78f0dbadc838753e2db7b757b70cea2fae8..ef0849eb69deee974a8dd5ada628b469822755e5 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -623,6 +623,13 @@ public abstract class BlockBehaviour implements FeatureElement {
                 this.emptyConstantCollisionShape = this.constantCollisionShape != null && this.constantCollisionShape.isEmpty();
                 // init caches
                 initCaches(collisionShape, true);
+                // Leaf start - optimize collision shape
+                switch (getBlock()) {
+                    case net.minecraft.world.level.block.FireBlock fireBlock -> {}
+                    case net.minecraft.world.level.block.LiquidBlock liquidBlock -> {}
+                    default -> this.constantCollisionShape = collisionShape;
+                }
+                // Leaf end - optimize collision shape
                 if (this.constantCollisionShape != null) {
                     initCaches(this.constantCollisionShape, true);
                 }
