From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Cache world border


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index e41087a52c5c0521728357431afd182080ab1291..d1164dcb0a7381ab84a08bfadc9f07b6f65aef01 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -237,6 +237,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
     public boolean hasRidableMoveEvent = false; // Purpur - Ridables
     public final org.dreeam.leaf.async.tracker.AsyncTracker leaf$asyncTracker = new org.dreeam.leaf.async.tracker.AsyncTracker(this); // Leaf - Multithreaded tracker
+    private @Nullable WorldBorder worldBorder; // Leaf - Cache worldborder
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
@@ -2517,8 +2518,12 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
     @Override
     public WorldBorder getWorldBorder() {
+        if (this.worldBorder != null) {
+            return this.worldBorder;
+        }
         WorldBorder worldBorder = this.getDataStorage().computeIfAbsent(WorldBorder.TYPE);
         worldBorder.applyInitialSettings(this.levelData.getGameTime());
+        this.worldBorder = worldBorder;
         return worldBorder;
     }
 
