From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:24 +0900
Subject: [PATCH] optimize updateFluidOnEyes


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index c76174f177b672325107be67a760111d5b36efaf..0a93e8ca347c3827b6d16bc4c549d7db3ed286d1 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2134,7 +2134,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 && abstractBoat.getBoundingBox().minY <= eyeY
         )) {
             BlockPos blockPos = BlockPos.containing(this.getX(), eyeY, this.getZ());
-            FluidState fluidState = this.level().getFluidState(blockPos);
+            FluidState fluidState = this.level().getFluidStateIfLoadedUnchecked(blockPos.getX(), blockPos.getY(), blockPos.getZ());
+            if (fluidState == null) { fluidState = this.level().getFluidState(blockPos); }
             double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
             if (d > eyeY) {
                 // Leaf start - Optimize isEyeInFluid
