From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Wed, 13 Aug 2025 23:35:42 +0900
Subject: [PATCH] fix purpur attribute base patch


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 73029fced7a033c6bcb8051662c5f86d5b1f13c1..7ad7d944a334d300d6f6aa0b40a2279dbfb8f808 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -2954,6 +2954,16 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return this.getAttributes().getBaseValue(attribute);
     }
 
+    // Leaf start
+    public double setAttributeBaseValue(Holder<Attribute> attribute, double d) {
+        double old = this.getAttributes().getBaseValue(attribute);
+        if (old != d) {
+            this.getAttributes().getInstance(attribute).setBaseValue(d);
+        }
+        return old;
+    }
+    // Leaf end
+
     public AttributeMap getAttributes() {
         return this.attributes;
     }
diff --git a/net/minecraft/world/entity/ambient/Bat.java b/net/minecraft/world/entity/ambient/Bat.java
index b642f71d282773afb9fe273cbe65765afa2d5868..4232057fa3a5fe5707c07201a7df3e950b81be4a 100644
--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -97,15 +97,15 @@ public class Bat extends AmbientCreature {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.batMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.batScale);
-        this.getAttribute(Attributes.FOLLOW_RANGE).setBaseValue(this.level().purpurConfig.batFollowRange);
-        this.getAttribute(Attributes.KNOCKBACK_RESISTANCE).setBaseValue(this.level().purpurConfig.batKnockbackResistance);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.batMovementSpeed);
-        this.getAttribute(Attributes.FLYING_SPEED).setBaseValue(this.level().purpurConfig.batFlyingSpeed);
-        this.getAttribute(Attributes.ARMOR).setBaseValue(this.level().purpurConfig.batArmor);
-        this.getAttribute(Attributes.ARMOR_TOUGHNESS).setBaseValue(this.level().purpurConfig.batArmorToughness);
-        this.getAttribute(Attributes.ATTACK_KNOCKBACK).setBaseValue(this.level().purpurConfig.batAttackKnockback);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.batMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.batScale);
+        this.setAttributeBaseValue(Attributes.FOLLOW_RANGE, this.level().purpurConfig.batFollowRange);
+        this.setAttributeBaseValue(Attributes.KNOCKBACK_RESISTANCE, this.level().purpurConfig.batKnockbackResistance);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.batMovementSpeed);
+        this.setAttributeBaseValue(Attributes.FLYING_SPEED, this.level().purpurConfig.batFlyingSpeed);
+        this.setAttributeBaseValue(Attributes.ARMOR, this.level().purpurConfig.batArmor);
+        this.setAttributeBaseValue(Attributes.ARMOR_TOUGHNESS, this.level().purpurConfig.batArmorToughness);
+        this.setAttributeBaseValue(Attributes.ATTACK_KNOCKBACK, this.level().purpurConfig.batAttackKnockback);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 154a72fbc85431c59dc793041952a69de089055f..91e2eb319179e396c13309fac2b1890d0f938ad4 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -170,8 +170,8 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.allayMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.allayScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.allayMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.allayScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/armadillo/Armadillo.java b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
index ec678479ef33a6f1a832598fa0df5b396d491898..82ba640f65d6f3506112a6018d6c0c73b1191366 100644
--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -102,8 +102,8 @@ public class Armadillo extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.armadilloMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.armadilloScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.armadilloMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.armadilloScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/axolotl/Axolotl.java b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
index 9d5f4b5f756d84e54e9dbeb0d91eda53c09a6ddc..03b5fa8cee9d6c914711ce09731a115a97b1baed 100644
--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -141,8 +141,8 @@ public class Axolotl extends Animal implements Bucketable {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.axolotlMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.axolotlScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.axolotlMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.axolotlScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/bee/Bee.java b/net/minecraft/world/entity/animal/bee/Bee.java
index a003cdcc6f81305b906e4b2159912549487d8da4..ab9849f1f9117f40612474faa18a9ed103542ebc 100644
--- a/net/minecraft/world/entity/animal/bee/Bee.java
+++ b/net/minecraft/world/entity/animal/bee/Bee.java
@@ -480,8 +480,8 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.beeMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.beeScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.beeMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.beeScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/chicken/Chicken.java b/net/minecraft/world/entity/animal/chicken/Chicken.java
index bb3eeea32c5f52192e368ced55b4f4d880c94134..564dd797f0433f57ecfa2d11b1a92c5eba363f84 100644
--- a/net/minecraft/world/entity/animal/chicken/Chicken.java
+++ b/net/minecraft/world/entity/animal/chicken/Chicken.java
@@ -93,8 +93,8 @@ public class Chicken extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.chickenMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.chickenScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.chickenMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.chickenScale);
         // Purpur start - Chickens can retaliate
         if (level().purpurConfig.chickenRetaliate) {
             this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(2.0D);
diff --git a/net/minecraft/world/entity/animal/cow/Cow.java b/net/minecraft/world/entity/animal/cow/Cow.java
index 1463e94d3851614caee216b50e90d1980de75298..6415ca6c2d8a11f061000d6928cd37b0fe9c0e03 100644
--- a/net/minecraft/world/entity/animal/cow/Cow.java
+++ b/net/minecraft/world/entity/animal/cow/Cow.java
@@ -51,9 +51,9 @@ public class Cow extends AbstractCow {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.cowMaxHealth);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.SCALE).setBaseValue(this.level().purpurConfig.cowScale);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_DAMAGE).setBaseValue(this.level().purpurConfig.cowNaturallyAggressiveToPlayersDamage); // Purpur - Cows naturally aggressive to players chance
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.cowMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.SCALE, this.level().purpurConfig.cowScale);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_DAMAGE, this.level().purpurConfig.cowNaturallyAggressiveToPlayersDamage); // Purpur - Cows naturally aggressive to players chance
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/cow/MushroomCow.java b/net/minecraft/world/entity/animal/cow/MushroomCow.java
index 8019a13adaf4dcc71065a0949b29226dd66343d7..69b60131df92aab7f94edfe34d5d1ef7aa76be83 100644
--- a/net/minecraft/world/entity/animal/cow/MushroomCow.java
+++ b/net/minecraft/world/entity/animal/cow/MushroomCow.java
@@ -81,7 +81,7 @@ public class MushroomCow extends AbstractCow implements Shearable {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.mooshroomMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.mooshroomMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/dolphin/Dolphin.java b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
index ea3ccc1f39f70c8ce407ed94ab19a3caa3b90dc4..e5a5bb7ae387b9d7960804cb0b71ab340a4e424b 100644
--- a/net/minecraft/world/entity/animal/dolphin/Dolphin.java
+++ b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
@@ -155,8 +155,8 @@ public class Dolphin extends AgeableWaterCreature {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.dolphinMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.dolphinScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.dolphinMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.dolphinScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/feline/Cat.java b/net/minecraft/world/entity/animal/feline/Cat.java
index 59a4e8b5efa212571c73e7808faa65b1249b718d..bb295ac3a18cd3644f9b051c51cd918fcb7c9c10 100644
--- a/net/minecraft/world/entity/animal/feline/Cat.java
+++ b/net/minecraft/world/entity/animal/feline/Cat.java
@@ -123,8 +123,8 @@ public class Cat extends TamableAnimal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.catMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.catScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.catMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.catScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/feline/Ocelot.java b/net/minecraft/world/entity/animal/feline/Ocelot.java
index 912347613c61cc70375c3d90b3da95851e403f78..1fbf06bcc7fc047aa3b03e35a09207b850918fd4 100644
--- a/net/minecraft/world/entity/animal/feline/Ocelot.java
+++ b/net/minecraft/world/entity/animal/feline/Ocelot.java
@@ -86,8 +86,8 @@ public class Ocelot extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.ocelotMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.ocelotScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.ocelotMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.ocelotScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/fish/Cod.java b/net/minecraft/world/entity/animal/fish/Cod.java
index 05de579a16726454034dc4a913161676b5c73552..60793b01096c568b1224ba903593c9f08ccbd27c 100644
--- a/net/minecraft/world/entity/animal/fish/Cod.java
+++ b/net/minecraft/world/entity/animal/fish/Cod.java
@@ -28,7 +28,7 @@ public class Cod extends AbstractSchoolingFish {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.codMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.codMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/fish/Pufferfish.java b/net/minecraft/world/entity/animal/fish/Pufferfish.java
index 681eb25e45cec76f21cea9f77efe71290bdb547a..f4870da24cc2eec95143128415ffd7906c9f98ad 100644
--- a/net/minecraft/world/entity/animal/fish/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/fish/Pufferfish.java
@@ -62,7 +62,7 @@ public class Pufferfish extends AbstractFish {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.pufferfishMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.pufferfishMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/fish/Salmon.java b/net/minecraft/world/entity/animal/fish/Salmon.java
index b76b2c1e7ac93fe01bdf9ee6e3c811f448babcbb..f15898e7d30d7acc1fb9f520b3da42243af48ae6 100644
--- a/net/minecraft/world/entity/animal/fish/Salmon.java
+++ b/net/minecraft/world/entity/animal/fish/Salmon.java
@@ -55,7 +55,7 @@ public class Salmon extends AbstractSchoolingFish {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.salmonMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.salmonMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/fish/TropicalFish.java b/net/minecraft/world/entity/animal/fish/TropicalFish.java
index c7742f06cf2bed3496c28b541e45948e609c1999..1f2f51955e258adf8a2011ab4f06b7e3def7c792 100644
--- a/net/minecraft/world/entity/animal/fish/TropicalFish.java
+++ b/net/minecraft/world/entity/animal/fish/TropicalFish.java
@@ -92,7 +92,7 @@ public class TropicalFish extends AbstractSchoolingFish {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.tropicalFishMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.tropicalFishMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/fox/Fox.java b/net/minecraft/world/entity/animal/fox/Fox.java
index 98937a998f55ee462421a7477cc3e2fbf4e75ba3..ef9acaf4ca4757f35705b09ff93d19930bd7b267 100644
--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -192,8 +192,8 @@ public class Fox extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.foxMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.foxScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.foxMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.foxScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/golem/CopperGolem.java b/net/minecraft/world/entity/animal/golem/CopperGolem.java
index e9febadb3640261d16dbfe6b53197861cfa73a5c..62a3ec56ee673397e1e4b72da266aeb44e3c917a 100644
--- a/net/minecraft/world/entity/animal/golem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolem.java
@@ -133,10 +133,10 @@ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearab
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.copperGolemMaxHealth);
-        this.getAttribute(Attributes.STEP_HEIGHT).setBaseValue(this.level().purpurConfig.copperGolemStepHeight);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.copperGolemMovementSpeed);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.copperGolemScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.copperGolemMaxHealth);
+        this.setAttributeBaseValue(Attributes.STEP_HEIGHT, this.level().purpurConfig.copperGolemStepHeight);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.copperGolemMovementSpeed);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.copperGolemScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/golem/IronGolem.java b/net/minecraft/world/entity/animal/golem/IronGolem.java
index ddde4297b77d174034300165b591a224c0401743..3d49ab34480405bc4473bf1dc745fa418f2f02ad 100644
--- a/net/minecraft/world/entity/animal/golem/IronGolem.java
+++ b/net/minecraft/world/entity/animal/golem/IronGolem.java
@@ -94,8 +94,8 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.ironGolemMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.ironGolemScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.ironGolemMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.ironGolemScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/golem/SnowGolem.java b/net/minecraft/world/entity/animal/golem/SnowGolem.java
index c37c999671955eeaff0b145a890a512c78d642a8..697a436e43e7ca8ef6e5f4d157810d065e3ff22e 100644
--- a/net/minecraft/world/entity/animal/golem/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/golem/SnowGolem.java
@@ -82,8 +82,8 @@ public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackM
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.snowGolemMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.snowGolemScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.snowGolemMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.snowGolemScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
index 29de138dd245dac1491f8edd2c37ba8a7a8b7f49..c9361661c40c1be62d521ef1f6abcde017d03877 100644
--- a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
+++ b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
@@ -138,13 +138,13 @@ public class HappyGhast extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.happyGhastMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.happyGhastScale);
-        this.getAttribute(Attributes.TEMPT_RANGE).setBaseValue(this.level().purpurConfig.happyGhastTemptRange);
-        this.getAttribute(Attributes.FLYING_SPEED).setBaseValue(this.level().purpurConfig.happyGhastFlyingSpeed);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.happyGhastMovementSpeed);
-        this.getAttribute(Attributes.FOLLOW_RANGE).setBaseValue(this.level().purpurConfig.happyGhastFollowRange);
-        this.getAttribute(Attributes.CAMERA_DISTANCE).setBaseValue(this.level().purpurConfig.happyGhastCameraDistance);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.happyGhastMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.happyGhastScale);
+        this.setAttributeBaseValue(Attributes.TEMPT_RANGE, this.level().purpurConfig.happyGhastTemptRange);
+        this.setAttributeBaseValue(Attributes.FLYING_SPEED, this.level().purpurConfig.happyGhastFlyingSpeed);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.happyGhastMovementSpeed);
+        this.setAttributeBaseValue(Attributes.FOLLOW_RANGE, this.level().purpurConfig.happyGhastFollowRange);
+        this.setAttributeBaseValue(Attributes.CAMERA_DISTANCE, this.level().purpurConfig.happyGhastCameraDistance);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/nautilus/Nautilus.java b/net/minecraft/world/entity/animal/nautilus/Nautilus.java
index ab556f276ed899cde5bcd85246842966bbc586b5..816eb2ba7b24d0831d16de04c423f37b3b24ba04 100644
--- a/net/minecraft/world/entity/animal/nautilus/Nautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/Nautilus.java
@@ -24,11 +24,11 @@ public class Nautilus extends AbstractNautilus {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.nautilusMaxHealth);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.nautilusMovementSpeed);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_DAMAGE).setBaseValue(this.level().purpurConfig.nautilusAttackDamage);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.KNOCKBACK_RESISTANCE).setBaseValue(this.level().purpurConfig.nautilusKnockbackResistance);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.SCALE).setBaseValue(this.level().purpurConfig.nautilusScale);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.nautilusMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED, this.level().purpurConfig.nautilusMovementSpeed);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_DAMAGE, this.level().purpurConfig.nautilusAttackDamage);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.KNOCKBACK_RESISTANCE, this.level().purpurConfig.nautilusKnockbackResistance);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.SCALE, this.level().purpurConfig.nautilusScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java b/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
index 00795af5b4876bde266b113fc9e28199e07e5b7f..7e5836929e3ed16de65bcca012688b4e1450c45a 100644
--- a/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
@@ -46,11 +46,11 @@ public class ZombieNautilus extends AbstractNautilus {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.zombieNautilusMaxHealth);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.zombieNautilusMovementSpeed);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_DAMAGE).setBaseValue(this.level().purpurConfig.zombieNautilusAttackDamage);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.KNOCKBACK_RESISTANCE).setBaseValue(this.level().purpurConfig.zombieNautilusKnockbackResistance);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.SCALE).setBaseValue(this.level().purpurConfig.zombieNautilusScale);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.zombieNautilusMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED, this.level().purpurConfig.zombieNautilusMovementSpeed);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_DAMAGE, this.level().purpurConfig.zombieNautilusAttackDamage);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.KNOCKBACK_RESISTANCE, this.level().purpurConfig.zombieNautilusKnockbackResistance);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.SCALE, this.level().purpurConfig.zombieNautilusScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/panda/Panda.java b/net/minecraft/world/entity/animal/panda/Panda.java
index 20d203b7d738e1dd02783f9e3cc73e561c0bd1d7..7f2bf0df3ff5170852e626eeaa48e1fc81127d8d 100644
--- a/net/minecraft/world/entity/animal/panda/Panda.java
+++ b/net/minecraft/world/entity/animal/panda/Panda.java
@@ -137,8 +137,8 @@ public class Panda extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.pandaMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.pandaScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.pandaMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.pandaScale);
         setAttributes();
     }
     // Purpur end - Configurable entity base attributes
diff --git a/net/minecraft/world/entity/animal/parrot/Parrot.java b/net/minecraft/world/entity/animal/parrot/Parrot.java
index bbe3977939004b7d77e38b14d48c9c26e596d123..b2ce0a2f152dbd7f9b6128605e3dbfe5ca3ed105 100644
--- a/net/minecraft/world/entity/animal/parrot/Parrot.java
+++ b/net/minecraft/world/entity/animal/parrot/Parrot.java
@@ -202,8 +202,8 @@ public class Parrot extends ShoulderRidingEntity implements FlyingAnimal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.parrotMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.parrotScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.parrotMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.parrotScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/pig/Pig.java b/net/minecraft/world/entity/animal/pig/Pig.java
index a6a913b7aaca49eb59d4ec56050470c566271910..890c10fad3da49d4cc705eea9631d8458a7d6b36 100644
--- a/net/minecraft/world/entity/animal/pig/Pig.java
+++ b/net/minecraft/world/entity/animal/pig/Pig.java
@@ -83,8 +83,8 @@ public class Pig extends Animal implements ItemSteerable {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.pigMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.pigScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.pigMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.pigScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/polarbear/PolarBear.java b/net/minecraft/world/entity/animal/polarbear/PolarBear.java
index 592711e205ee959cc66de80933ad5a2dcec3092f..9d18f60e2b1e542fce349eedb71f1a48e291bf2a 100644
--- a/net/minecraft/world/entity/animal/polarbear/PolarBear.java
+++ b/net/minecraft/world/entity/animal/polarbear/PolarBear.java
@@ -121,8 +121,8 @@ public class PolarBear extends Animal implements NeutralMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.polarBearMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.polarBearScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.polarBearMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.polarBearScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/rabbit/Rabbit.java b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
index 2b294db92ce0884ed1f213b4e2a379677459daad..081fe3c94d0a789e50a9d9ba5e24a8d05f312aa6 100644
--- a/net/minecraft/world/entity/animal/rabbit/Rabbit.java
+++ b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
@@ -151,8 +151,8 @@ public class Rabbit extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.rabbitMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.rabbitScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.rabbitMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.rabbitScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/sheep/Sheep.java b/net/minecraft/world/entity/animal/sheep/Sheep.java
index ed2d2bfa599e89e6f20bb4902ea6ef48369b5b69..3d23c9cc41e7940e415986a422c9a36a0cf8019d 100644
--- a/net/minecraft/world/entity/animal/sheep/Sheep.java
+++ b/net/minecraft/world/entity/animal/sheep/Sheep.java
@@ -83,8 +83,8 @@ public class Sheep extends Animal implements Shearable {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.sheepMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.sheepScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.sheepMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.sheepScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index 8029e36436b3fc7e9930dea9f070be8502f88d12..af6523a4c2c5e5e8accf8f88537de2598af759c8 100644
--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -109,8 +109,8 @@ public class Sniffer extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.snifferMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.snifferScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.snifferMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.snifferScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/squid/GlowSquid.java b/net/minecraft/world/entity/animal/squid/GlowSquid.java
index dd1bd8b1a348a0cc398b123a77d3cb28dd1273ae..8a9be00ab346023789fcb1b7bd195b12b7bce039 100644
--- a/net/minecraft/world/entity/animal/squid/GlowSquid.java
+++ b/net/minecraft/world/entity/animal/squid/GlowSquid.java
@@ -53,7 +53,7 @@ public class GlowSquid extends Squid {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.glowSquidMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.glowSquidMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/squid/Squid.java b/net/minecraft/world/entity/animal/squid/Squid.java
index 6b5be8e58aa7e3acd14ff3cb1ba25d15f8d37410..6fde3d795c9de45ac60346e8623cd9fcac37cb4a 100644
--- a/net/minecraft/world/entity/animal/squid/Squid.java
+++ b/net/minecraft/world/entity/animal/squid/Squid.java
@@ -96,8 +96,8 @@ public class Squid extends AgeableWaterCreature {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.squidMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.squidScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.squidMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.squidScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/turtle/Turtle.java b/net/minecraft/world/entity/animal/turtle/Turtle.java
index 4fa4e755955c27c689661774d214f797ffa33697..3e67c7068326f38878994000750addda50dbe584 100644
--- a/net/minecraft/world/entity/animal/turtle/Turtle.java
+++ b/net/minecraft/world/entity/animal/turtle/Turtle.java
@@ -107,8 +107,8 @@ public class Turtle extends Animal {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.turtleMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.turtleScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.turtleMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.turtleScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/animal/wolf/Wolf.java b/net/minecraft/world/entity/animal/wolf/Wolf.java
index 7fbe841818d08fd930f7ce405e84e5ed4badb614..1f3b67945582ade7bb8544f495be0b6db6f11019 100644
--- a/net/minecraft/world/entity/animal/wolf/Wolf.java
+++ b/net/minecraft/world/entity/animal/wolf/Wolf.java
@@ -211,8 +211,8 @@ public class Wolf extends TamableAnimal implements NeutralMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.wolfMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.wolfScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.wolfMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.wolfScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index 335bff0b9537ea4ea9717e76e824b07642d7dd8e..8ca100f8e30899cb81e4994865404a740821df3a 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -166,7 +166,7 @@ public class EnderDragon extends Mob implements Enemy {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.enderDragonMaxHealth);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.enderDragonMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index eb6dfc7e765616fd26f7a22843e646933f6e933b..c98f904c73b8225428a5e7120aceee1406f5e0f6 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -216,8 +216,8 @@ public class WitherBoss extends Monster implements RangedAttackMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.witherMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.witherScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.witherMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.witherScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Blaze.java b/net/minecraft/world/entity/monster/Blaze.java
index 187da2d55cdb039b62abe708dadd6c05e13ac64f..cd8316949056624e60202fda2d439e40f8aec53a 100644
--- a/net/minecraft/world/entity/monster/Blaze.java
+++ b/net/minecraft/world/entity/monster/Blaze.java
@@ -79,8 +79,8 @@ public class Blaze extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.blazeMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.blazeScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.blazeMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.blazeScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Creeper.java b/net/minecraft/world/entity/monster/Creeper.java
index 608d787b73dce1726ccece9e0372acf995f261d4..a11ed0ab794a1d7dc27b01f5578faf853471311e 100644
--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -140,8 +140,8 @@ public class Creeper extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.creeperMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.creeperScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.creeperMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.creeperScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/ElderGuardian.java b/net/minecraft/world/entity/monster/ElderGuardian.java
index abb7496c293175828fde54ae2b4991aa3ff6b30f..d7b23479d3ceaedc32ae6f3d3cd5023d8ae9a873 100644
--- a/net/minecraft/world/entity/monster/ElderGuardian.java
+++ b/net/minecraft/world/entity/monster/ElderGuardian.java
@@ -46,8 +46,8 @@ public class ElderGuardian extends Guardian {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.elderGuardianMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.elderGuardianScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.elderGuardianMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.elderGuardianScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index 809933cf4eea9dcedae4195cfecfe392973e96b6..c038c190f4a88516919e6a00b1846f7d6314e415 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -109,8 +109,8 @@ public class EnderMan extends Monster implements NeutralMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.endermanMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.endermanScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.endermanMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.endermanScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Endermite.java b/net/minecraft/world/entity/monster/Endermite.java
index 4a778ad22f92e3fd95665a9b40ae932b427bac16..55435037007390388c1357f3042661bc9a1d7c09 100644
--- a/net/minecraft/world/entity/monster/Endermite.java
+++ b/net/minecraft/world/entity/monster/Endermite.java
@@ -67,8 +67,8 @@ public class Endermite extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.endermiteMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.endermiteScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.endermiteMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.endermiteScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Ghast.java b/net/minecraft/world/entity/monster/Ghast.java
index 3ab0aba57ed3282eb73fc501ded00558ec3d772a..e7c33e76998093360b38cb5e6037f628881d863e 100644
--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -78,8 +78,8 @@ public class Ghast extends Mob implements Enemy {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.ghastMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.ghastScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.ghastMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.ghastScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Giant.java b/net/minecraft/world/entity/monster/Giant.java
index 9a081135025f114d473c50e68a778e034f33ab82..d60422982afe4d7b748baab3097536ae5e4d5a3c 100644
--- a/net/minecraft/world/entity/monster/Giant.java
+++ b/net/minecraft/world/entity/monster/Giant.java
@@ -55,10 +55,10 @@ public class Giant extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     protected void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.giantMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.giantScale);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.giantMovementSpeed);
-        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(this.level().purpurConfig.giantAttackDamage);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.giantMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.giantScale);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.giantMovementSpeed);
+        this.setAttributeBaseValue(Attributes.ATTACK_DAMAGE, this.level().purpurConfig.giantAttackDamage);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Guardian.java b/net/minecraft/world/entity/monster/Guardian.java
index 60490d0611a352ba0546b39fd41259713a7f18b5..38353c0d68afbfe509aefa2d0f29205d8fbf88a7 100644
--- a/net/minecraft/world/entity/monster/Guardian.java
+++ b/net/minecraft/world/entity/monster/Guardian.java
@@ -92,8 +92,8 @@ public class Guardian extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.guardianMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.guardianScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.guardianMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.guardianScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Ravager.java b/net/minecraft/world/entity/monster/Ravager.java
index 5e6fd02985eef337f2ff8e48fbd904a064c0560a..ca3371eba812f0323b2f8fbf7cf13befa6d0835d 100644
--- a/net/minecraft/world/entity/monster/Ravager.java
+++ b/net/minecraft/world/entity/monster/Ravager.java
@@ -98,8 +98,8 @@ public class Ravager extends Raider {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.ravagerMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.ravagerScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.ravagerMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.ravagerScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Shulker.java b/net/minecraft/world/entity/monster/Shulker.java
index 12e3ba1a545d5ec208ed33af66429f56fcdbed7f..9e5789ac110044dbc9e82a44cb347f1030a5b1f5 100644
--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -128,8 +128,8 @@ public class Shulker extends AbstractGolem implements Enemy {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.shulkerMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.shulkerScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.shulkerMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.shulkerScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Silverfish.java b/net/minecraft/world/entity/monster/Silverfish.java
index 81976287402a701e3ef12c474cb68eb602cc2951..5bb1724b076a8cd8286512d030809e27763ed475 100644
--- a/net/minecraft/world/entity/monster/Silverfish.java
+++ b/net/minecraft/world/entity/monster/Silverfish.java
@@ -58,10 +58,10 @@ public class Silverfish extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.silverfishMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.silverfishScale);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.silverfishMovementSpeed);
-        this.getAttribute(Attributes.ATTACK_DAMAGE).setBaseValue(this.level().purpurConfig.silverfishAttackDamage);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.silverfishMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.silverfishScale);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.silverfishMovementSpeed);
+        this.setAttributeBaseValue(Attributes.ATTACK_DAMAGE, this.level().purpurConfig.silverfishAttackDamage);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index 6e55d5d049ae77afdbc372c4c925b487650ee01f..f13de81f8601bd58ea84983c3e7cd5240533adc8 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -112,8 +112,8 @@ public class Strider extends Animal implements ItemSteerable {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.striderMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.striderScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.striderMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.striderScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Vex.java b/net/minecraft/world/entity/monster/Vex.java
index acd196d435758a8730569ba4e772bb954c4f85b6..10d21d83cd60389fd055e4f69b65d06e13098df2 100644
--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -105,8 +105,8 @@ public class Vex extends Monster implements TraceableEntity {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.vexMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.vexScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.vexMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.vexScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Witch.java b/net/minecraft/world/entity/monster/Witch.java
index 589169495c3d74660b5afa3f3b0be3e18f0d8e8e..cecea55ee2bb894662c84e6a308f2282748b6a71 100644
--- a/net/minecraft/world/entity/monster/Witch.java
+++ b/net/minecraft/world/entity/monster/Witch.java
@@ -77,8 +77,8 @@ public class Witch extends Raider implements RangedAttackMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.witchMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.witchScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.witchMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.witchScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/Zoglin.java b/net/minecraft/world/entity/monster/Zoglin.java
index ec2151494223ffdbf96849a9d93ecafc42ea8903..65bd33c43a7c6c1b7e4f80f7817ef9a5903b1ff5 100644
--- a/net/minecraft/world/entity/monster/Zoglin.java
+++ b/net/minecraft/world/entity/monster/Zoglin.java
@@ -107,8 +107,8 @@ public class Zoglin extends Monster implements HoglinBase {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.zoglinMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.zoglinScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.zoglinMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.zoglinScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/creaking/Creaking.java b/net/minecraft/world/entity/monster/creaking/Creaking.java
index ffc1927832331f9e42f58a367b6ffba53e91bedf..963b51b8a50fab2485c10ea480cb4feab89892bb 100644
--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -129,8 +129,8 @@ public class Creaking extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.creakingMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.creakingScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.creakingMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.creakingScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/hoglin/Hoglin.java b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
index e37fdca11d427f272ea444cffb7eae5d185238be..bfaa40771ee40c3d0ee69d7a7a8e0f1a2a85f3a1 100644
--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -117,8 +117,8 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.hoglinMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.hoglinScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.hoglinMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.hoglinScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/illager/Evoker.java b/net/minecraft/world/entity/monster/illager/Evoker.java
index 1572f7653045902fa05f0499d920eb2b2ff46931..7c84f59f82fc9b2827483ea677f907870043bcc1 100644
--- a/net/minecraft/world/entity/monster/illager/Evoker.java
+++ b/net/minecraft/world/entity/monster/illager/Evoker.java
@@ -69,8 +69,8 @@ public class Evoker extends SpellcasterIllager {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.evokerMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.evokerScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.evokerMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.evokerScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/illager/Illusioner.java b/net/minecraft/world/entity/monster/illager/Illusioner.java
index ea9b61b6f8463edf3497d240b010f8da3e04f62f..c77a08005ea3008d55d3c71a44a3b3195d02060c 100644
--- a/net/minecraft/world/entity/monster/illager/Illusioner.java
+++ b/net/minecraft/world/entity/monster/illager/Illusioner.java
@@ -79,10 +79,10 @@ public class Illusioner extends SpellcasterIllager implements RangedAttackMob {
     // Purpur start - Configurable entity base attributes
     @Override
     protected void initAttributes() {
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.illusionerMovementSpeed);
-        this.getAttribute(Attributes.FOLLOW_RANGE).setBaseValue(this.level().purpurConfig.illusionerFollowRange);
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.illusionerMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.illusionerScale);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.illusionerMovementSpeed);
+        this.setAttributeBaseValue(Attributes.FOLLOW_RANGE, this.level().purpurConfig.illusionerFollowRange);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.illusionerMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.illusionerScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/illager/Pillager.java b/net/minecraft/world/entity/monster/illager/Pillager.java
index 1a283185ac03ba44bae531f9b234703fa213df1a..46e6c2023c173ca945f27a26eea2cfc8ae14436a 100644
--- a/net/minecraft/world/entity/monster/illager/Pillager.java
+++ b/net/minecraft/world/entity/monster/illager/Pillager.java
@@ -85,8 +85,8 @@ public class Pillager extends AbstractIllager implements CrossbowAttackMob, Inve
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.pillagerMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.pillagerScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.pillagerMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.pillagerScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/illager/Vindicator.java b/net/minecraft/world/entity/monster/illager/Vindicator.java
index 97bf22925df40a2f757b50f3fc7cbefc951443bb..2c7b700e0b0a46cf625930aad25707b18261c28a 100644
--- a/net/minecraft/world/entity/monster/illager/Vindicator.java
+++ b/net/minecraft/world/entity/monster/illager/Vindicator.java
@@ -77,8 +77,8 @@ public class Vindicator extends AbstractIllager {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.vindicatorMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.vindicatorScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.vindicatorMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.vindicatorScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index b2ad28c7491fbd06ac1a538aa3b5289d8280de62..be31ba9b13d97ff00974583019f14b9b0a7c22b1 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -161,8 +161,8 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.piglinMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.piglinScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.piglinMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.piglinScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/piglin/PiglinBrute.java b/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
index cef2cb8173fb2ebdc1433433c8030c65796544ec..c6cb848d49721679272dd30e9ffdd109f2ac4670 100644
--- a/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
@@ -85,8 +85,8 @@ public class PiglinBrute extends AbstractPiglin {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.piglinBruteMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.piglinBruteScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.piglinBruteMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.piglinBruteScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/skeleton/Bogged.java b/net/minecraft/world/entity/monster/skeleton/Bogged.java
index 59c986a28a5546a9070ec9bcf3f1994317b47fff..18cd9c6022433654f3b8a778fc722cff896ba399 100644
--- a/net/minecraft/world/entity/monster/skeleton/Bogged.java
+++ b/net/minecraft/world/entity/monster/skeleton/Bogged.java
@@ -61,8 +61,8 @@ public class Bogged extends AbstractSkeleton implements Shearable {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.boggedMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.boggedScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.boggedMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.boggedScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/skeleton/Skeleton.java b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
index 304e6e23a616ddc3312ef27ae5d94c927a52fb80..90cc74a3dc1232e26ba9a441692e28fd464938a8 100644
--- a/net/minecraft/world/entity/monster/skeleton/Skeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
@@ -46,7 +46,7 @@ public class Skeleton extends AbstractSkeleton {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.skeletonMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.skeletonMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/skeleton/Stray.java b/net/minecraft/world/entity/monster/skeleton/Stray.java
index 262bea9ca0daba064d2e74223b47eb1fbb761479..2813707c47f3dd8359f4cc7ce43f53f8db74ba8c 100644
--- a/net/minecraft/world/entity/monster/skeleton/Stray.java
+++ b/net/minecraft/world/entity/monster/skeleton/Stray.java
@@ -43,7 +43,7 @@ public class Stray extends AbstractSkeleton {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.strayMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.strayMaxHealth);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/skeleton/WitherSkeleton.java b/net/minecraft/world/entity/monster/skeleton/WitherSkeleton.java
index c01f182e65b7891dfdd7b92ff2c6e2cf3a8a0374..260920543a19c55eeaff3cc824297c1a1f973948 100644
--- a/net/minecraft/world/entity/monster/skeleton/WitherSkeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/WitherSkeleton.java
@@ -54,8 +54,8 @@ public class WitherSkeleton extends AbstractSkeleton {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.witherSkeletonMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.witherSkeletonScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.witherSkeletonMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.witherSkeletonScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/spider/CaveSpider.java b/net/minecraft/world/entity/monster/spider/CaveSpider.java
index fb4caabbbe5bd0befdaba49ee3cfe70c326b50a7..7595632a76af9dfcbfd80ad5ae777203ba7552d7 100644
--- a/net/minecraft/world/entity/monster/spider/CaveSpider.java
+++ b/net/minecraft/world/entity/monster/spider/CaveSpider.java
@@ -46,8 +46,8 @@ public class CaveSpider extends Spider {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.caveSpiderMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.caveSpiderScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.caveSpiderMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.caveSpiderScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/spider/Spider.java b/net/minecraft/world/entity/monster/spider/Spider.java
index 7993ea937772cabe848726c942b6df0ae52e74e2..9b7bfb322e49250fd4ec4bd6873d0fbcbbdb8f82 100644
--- a/net/minecraft/world/entity/monster/spider/Spider.java
+++ b/net/minecraft/world/entity/monster/spider/Spider.java
@@ -72,8 +72,8 @@ public class Spider extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.spiderMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.spiderScale);
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.spiderMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.spiderScale);
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/zombie/Drowned.java b/net/minecraft/world/entity/monster/zombie/Drowned.java
index c65831fc755423cdec2aa981d96641f3636f7376..399b3084e58182b4d12d62ac7e0d981167ad5312 100644
--- a/net/minecraft/world/entity/monster/zombie/Drowned.java
+++ b/net/minecraft/world/entity/monster/zombie/Drowned.java
@@ -99,9 +99,9 @@ public class Drowned extends Zombie implements RangedAttackMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.drownedMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.drownedScale);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.drownedMovementSpeed); // Leaf - Configurable movement speed of more entities - drowned
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.drownedMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.drownedScale);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.drownedMovementSpeed); // Leaf - Configurable movement speed of more entities - drowned
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/zombie/Husk.java b/net/minecraft/world/entity/monster/zombie/Husk.java
index 85e0cb3951955c8534e6e0b2d3c760d36dfac4ed..7b9f660d0918c338266a29fd0ca088f8533f23e1 100644
--- a/net/minecraft/world/entity/monster/zombie/Husk.java
+++ b/net/minecraft/world/entity/monster/zombie/Husk.java
@@ -50,8 +50,8 @@ public class Husk extends Zombie {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.huskMaxHealth);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.huskMovementSpeed); // Leaf - Configurable movement speed of more entities - husk
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.huskMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED, this.level().purpurConfig.huskMovementSpeed); // Leaf - Configurable movement speed of more entities - husk
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index 823f3f360c53ec627df7fa46e591f490626d0bdd..0d07934abd0b9f68bd70969f56964d034e503440 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -124,9 +124,9 @@ public class Zombie extends Monster {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.zombieMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.zombieScale);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.zombieMovementSpeed); // Leaf - Configurable movement speed of more entities - zombie
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.zombieMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.zombieScale);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.zombieMovementSpeed); // Leaf - Configurable movement speed of more entities - zombie
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/monster/zombie/ZombieVillager.java b/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
index 9249232f8d28ac86578cdf7bd877b54c90e5fd1e..60c1da1c024415254d78e2f38d66134b8d7baa03 100644
--- a/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
@@ -105,8 +105,8 @@ public class ZombieVillager extends Zombie implements VillagerDataHolder {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.zombieVillagerMaxHealth);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.zombieVillagerMovementSpeed); // Leaf - Configurable movement speed of more entities - zombieVillager
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.zombieVillagerMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MOVEMENT_SPEED, this.level().purpurConfig.zombieVillagerMovementSpeed); // Leaf - Configurable movement speed of more entities - zombieVillager
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java b/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
index 1449e9e3d457d2a67417af38fb07543f07a78ac7..843898743a789de45b4bf21a25c48e3b354d20e8 100644
--- a/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
@@ -84,9 +84,9 @@ public class ZombifiedPiglin extends Zombie implements NeutralMob {
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.zombifiedPiglinMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.zombifiedPiglinScale);
-        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.zombifiedPiglinMovementSpeed); // Leaf - Configurable movement speed of more entities - zombifiedPiglin
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.zombifiedPiglinMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.zombifiedPiglinScale);
+        this.setAttributeBaseValue(Attributes.MOVEMENT_SPEED, this.level().purpurConfig.zombifiedPiglinMovementSpeed); // Leaf - Configurable movement speed of more entities - zombifiedPiglin
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 03d2cf4145e188bdb9474f6c3a43083170536c45..0de56d252897988b4ec3043b446dfba9ffff4785 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -277,9 +277,9 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.villagerMaxHealth);
-        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.villagerScale);
-        this.getAttribute(Attributes.TEMPT_RANGE).setBaseValue(this.level().purpurConfig.villagerTemptRange); // Purpur - Villagers follow emerald blocks
+        this.setAttributeBaseValue(Attributes.MAX_HEALTH, this.level().purpurConfig.villagerMaxHealth);
+        this.setAttributeBaseValue(Attributes.SCALE, this.level().purpurConfig.villagerScale);
+        this.setAttributeBaseValue(Attributes.TEMPT_RANGE, this.level().purpurConfig.villagerTemptRange); // Purpur - Villagers follow emerald blocks
     }
     // Purpur end - Configurable entity base attributes
 
diff --git a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
index 94967cc21d76ce9abb51c5deda6cbe8c7b945af3..a79e0d337ba6c2ba17711179499ab02c09e81542 100644
--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
@@ -88,8 +88,8 @@ public class WanderingTrader extends AbstractVillager implements Consumable.Over
     // Purpur start - Configurable entity base attributes
     @Override
     public void initAttributes() {
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.wanderingTraderMaxHealth);
-        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.TEMPT_RANGE).setBaseValue(this.level().purpurConfig.wanderingTraderTemptRange); // Purpur - Villagers follow emerald blocks
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH, this.level().purpurConfig.wanderingTraderMaxHealth);
+        this.setAttributeBaseValue(net.minecraft.world.entity.ai.attributes.Attributes.TEMPT_RANGE, this.level().purpurConfig.wanderingTraderTemptRange); // Purpur - Villagers follow emerald blocks
     }
     // Purpur end - Configurable entity base attributes
 
