From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Mon, 9 Feb 2026 12:37:26 +0300
Subject: [PATCH] Optimize map lookups with isEmpty check


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 001a8685a0b40b4bacee38f13dab9d5077628494..8288ea265dc3ed39183825ae3e80b098b4ce169b 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -266,6 +266,7 @@ public abstract class Entity
     private int id = ENTITY_COUNTER.incrementAndGet();
     public boolean blocksBuilding;
     public ImmutableList<Entity> passengers = ImmutableList.of();
+    public static final ImmutableList<Entity> EMPTY_PASSENGERS = ImmutableList.of(); // Leaf - Optimize map lookups with isEmpty check
     protected int boardingCooldown;
     private @Nullable Entity vehicle;
     private Level level;
@@ -484,7 +485,7 @@ public abstract class Entity
 
     @Override
     public final boolean moonrise$hasAnyPlayerPassengers() {
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             return false;
         }
         return this.getIndirectPassengersStream().anyMatch((entity) -> entity instanceof Player);
@@ -3513,7 +3514,7 @@ public abstract class Entity
             throw new IllegalStateException("Use x.startRiding(y), not y.addPassenger(x)");
         }
 
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             this.passengers = ImmutableList.of(passenger);
         } else {
             List<Entity> newPassengers = Lists.newArrayList(this.passengers);
@@ -3591,7 +3592,7 @@ public abstract class Entity
     }
 
     protected boolean canAddPassenger(final Entity passenger) {
-        return this.passengers.isEmpty();
+        return this.passengers == EMPTY_PASSENGERS; // Leaf - Optimize map lookups with isEmpty check
     }
 
     protected boolean couldAcceptPassenger() {
@@ -3780,7 +3781,7 @@ public abstract class Entity
     }
 
     public boolean isVehicle() {
-        return !this.passengers.isEmpty();
+        return this.passengers != EMPTY_PASSENGERS; // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean dismountsUnderwater() {
@@ -4811,11 +4812,11 @@ public abstract class Entity
     }
 
     public @Nullable Entity getFirstPassenger() {
-        return this.passengers.isEmpty() ? null : this.passengers.get(0);
+        return this.passengers == EMPTY_PASSENGERS ? null : this.passengers.get(0); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean hasPassenger(final Entity entity) {
-        return this.passengers.contains(entity);
+        return passengers != EMPTY_PASSENGERS && this.passengers.contains(entity); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public boolean hasPassenger(final Predicate<Entity> test) {
@@ -4829,19 +4830,19 @@ public abstract class Entity
     }
 
     private Stream<Entity> getIndirectPassengersStream() {
-        if (this.passengers.isEmpty()) { return Stream.of(); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return this.passengers.stream().flatMap(Entity::getSelfAndPassengers);
     }
 
     @Override
     public Stream<Entity> getSelfAndPassengers() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return Stream.concat(Stream.of(this), this.getIndirectPassengersStream());
     }
 
     @Override
     public Stream<Entity> getPassengersAndSelf() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return Stream.concat(this.passengers.stream().flatMap(Entity::getPassengersAndSelf), Stream.of(this));
     }
 
@@ -4849,7 +4850,7 @@ public abstract class Entity
         // Paper start - optimise entity tracker
         final List<Entity> ret = new ArrayList<>();
 
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             return ret;
         }
 
@@ -4864,7 +4865,7 @@ public abstract class Entity
     }
 
     public boolean hasExactlyOnePlayerPassenger() {
-        if (this.passengers.isEmpty()) { return false; } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == EMPTY_PASSENGERS) { return false; } // Paper - Optimize indirect passenger iteration // Leaf - Optimize map lookups with isEmpty check
         return this.countPlayerPassengers() == 1;
     }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 13f642d6fc4e3dd3d2afe583435231e99e0e9fb6..539ee99c51a73efd1fb44e8924aaaefc5501a4b3 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1182,7 +1182,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public boolean hasEffect(final Holder<MobEffect> effect) {
-        return this.activeEffects.containsKey(effect);
+        return !activeEffects.isEmpty() && this.activeEffects.containsKey(effect); // Leaf - Optimize map lookups with isEmpty check
     }
 
     public @Nullable MobEffectInstance getEffect(final Holder<MobEffect> effect) {
@@ -4326,7 +4326,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     // CraftBukkit start - collidable API
     @Override
     public boolean canCollideWithBukkit(Entity entity) {
-        return this.isPushable() && this.collides != this.collidableExemptions.contains(entity.getUUID());
+        return this.isPushable() && this.collides != (!this.collidableExemptions.isEmpty() && this.collidableExemptions.contains(entity.getUUID())); // Leaf - Optimize map lookups with isEmpty check
     }
     // CraftBukkit end
 
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index ca905cc081235dd4c58f5231e6cbd66412f0c0b9..bee011da5222f8f462824d3494a21c152504c5a3 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1735,7 +1735,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         if (!player.getItemInHand(hand).isEmpty()) {
             return InteractionResult.PASS;
         }
-        if (!passengers.isEmpty() || player.isPassenger()) {
+        if (passengers != Entity.EMPTY_PASSENGERS || player.isPassenger()) { // Leaf - Optimize map lookups with isEmpty check
             return InteractionResult.PASS;
         }
         if (this instanceof TamableAnimal tamable) {
diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index 484e4d402b3a2fcf2ee1702654a7eed87b6698ff..9dab7178ce7ab030d9a0794f3bbebcd355495046 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -477,7 +477,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
 
     private Entity getTopPassenger(final Entity vehicle, final int counter) {
         List<Entity> passengers = vehicle.getPassengers();
-        return counter != 1 && !passengers.isEmpty() ? this.getTopPassenger(passengers.getFirst(), counter - 1) : vehicle;
+        return counter != 1 && passengers != Entity.EMPTY_PASSENGERS ? this.getTopPassenger(passengers.getFirst(), counter - 1) : vehicle; // Leaf - Optimize map lookups with isEmpty check
     }
 
     @Override
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index 41d82befd5d30a0bc1db083df9012e0b477239b4..593734ac811ee10875bcd45b0c8b459732e49cd7 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -447,7 +447,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
         this.setDeltaMovement(Mth.clamp(movement.x, -maxSpeed, maxSpeed), movement.y, Mth.clamp(movement.z, -maxSpeed, maxSpeed));
 
         // Purpur start - Minecart settings and WASD controls
-        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && !passengers.isEmpty()) {
+        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && passengers != Entity.EMPTY_PASSENGERS) { // Leaf - Optimize map lookups with isEmpty check
             Entity passenger = passengers.get(0);
             if (passenger instanceof net.minecraft.server.level.ServerPlayer player) {
                 net.minecraft.world.entity.player.Input lastClientInput = player.getLastClientInput();
