From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:24 +0900
Subject: [PATCH] optimize updateFluidOnEyes


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 9fdfd2c1307977f2f29c8e72e3921683ffcb198c..df648e11a3c5bc586ecb591bf838a28603c978a3 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2089,7 +2089,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 && abstractBoat.getBoundingBox().minY <= eyeY
         )) {
             BlockPos blockPos = BlockPos.containing(this.getX(), eyeY, this.getZ());
-            FluidState fluidState = this.level().getFluidState(blockPos);
+            FluidState fluidState = this.level().getFluidStateIfLoadedUnchecked(blockPos.getX(), blockPos.getY(), blockPos.getZ());
+            if (fluidState == null) { fluidState = this.level().getFluidState(blockPos); }
             double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
             if (d > eyeY) {
                 // Leaf start - Optimize isEyeInFluid
