From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Mon, 11 Aug 2025 02:19:41 +0900
Subject: [PATCH] optimize collision shape


diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 503420bbeaf3da0c568c28c55582f45469f40aeb..e186b435e7a25ba4c5a203a8748b1531dbf7b1b8 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -619,6 +619,18 @@ public abstract class BlockBehaviour implements FeatureElement {
                 this.emptyConstantCollisionShape = this.constantCollisionShape != null && this.constantCollisionShape.isEmpty();
                 // init caches
                 initCaches(collisionShape, true);
+                // Leaf start - optimize collision shape
+                switch (getBlock()) {
+                    case net.minecraft.world.level.block.LiquidBlock ignore -> {
+                        if (getFluidState().is(Fluids.WATER)) {
+                            this.constantCollisionShape = collisionShape;
+                        }
+                    }
+                    case net.minecraft.world.level.block.ScaffoldingBlock ignore -> {}
+                    case net.minecraft.world.level.block.PowderSnowBlock ignore -> {}
+                    default -> this.constantCollisionShape = collisionShape;
+                }
+                // Leaf end - optimize collision shape
                 if (this.constantCollisionShape != null) {
                     initCaches(this.constantCollisionShape, true);
                 }
