From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 31 Jan 2026 16:00:21 +0900
Subject: [PATCH] fix async save profile cache


diff --git a/net/minecraft/server/players/CachedUserNameToIdResolver.java b/net/minecraft/server/players/CachedUserNameToIdResolver.java
index a06275b08bc2a343260245a348958ed57fdd37d4..eba47e7eb856d04ceff87dbc055c7c3e30279e02 100644
--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -219,21 +219,26 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
         JsonArray jsonArray = new JsonArray();
         DateFormat dateFormat = createDateFormat();
         this.listTopMRUProfiles(org.spigotmc.SpigotConfig.userCacheCap).forEach(gameProfileInfo -> jsonArray.add(writeGameProfile(gameProfileInfo, dateFormat))); // Spigot // Paper - Fix GameProfileCache concurrency
-        String string = this.gson.toJson((JsonElement)jsonArray);
+        //String string = this.gson.toJson((JsonElement)jsonArray); // Leaf - fix async save profile cache
 
-        Runnable save = () -> { // Paper - Perf: Async GameProfileCache saving
+        java.util.concurrent.Callable<Void> save = () -> { // Paper - Perf: Async GameProfileCache saving // Leaf - fix async save profile cache
+        String string = this.gson.toJson((JsonElement)jsonArray); // Leaf - fix async save profile cache
         try (Writer writer = Files.newWriter(this.file, StandardCharsets.UTF_8)) {
             writer.write(string);
         } catch (IOException var9) {
         }
+        return null; // Leaf - fix async save profile cache
         // Paper start - Perf: Async GameProfileCache saving
         };
-        if (asyncSave) {
-            io.papermc.paper.util.MCUtil.scheduleAsyncTask(save);
-        } else {
-            save.run();
-        }
+        // Leaf start - fix async save profile cache
+        //if (asyncSave) {
+        //    io.papermc.paper.util.MCUtil.scheduleAsyncTask(save);
+        //} else {
+        //    save.run();
+        //}
         // Paper end - Perf: Async GameProfileCache saving
+        org.dreeam.leaf.async.AsyncPlayerDataSaving.submit2(save, this.file.toPath(), asyncSave, org.dreeam.leaf.config.modules.async.AsyncPlayerDataSave.profileCache);
+        // Leaf end - fix async save profile cache
     }
 
     private Stream<CachedUserNameToIdResolver.GameProfileInfo> getTopMRUProfiles(int limit) {
