From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 28 Feb 2026 14:42:29 +0900
Subject: [PATCH] reduce computeSpeed Vec3 allocations


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index b19fa0262eaf6a5bf0cc775ce1981b4479b5791b..d74b2c5156a0fe8c3838fc0f2241a66dabf22e91 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1089,7 +1089,7 @@ public abstract class Entity
             this.lastKnownPosition = this.position();
         }
 
-        this.lastKnownSpeed = this.position().subtract(this.lastKnownPosition);
+        this.lastKnownSpeed = this.position().equals(this.lastKnownPosition) ? Vec3.ZERO : this.position().subtract(this.lastKnownPosition); // Leaf - reduce computeSpeed Vec3 allocations
         this.lastKnownPosition = this.position();
     }
 
