From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:44:06 +0900
Subject: [PATCH] optimize updateFluidHeightAndDoFluidPushing


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 0a93e8ca347c3827b6d16bc4c549d7db3ed286d1..8d981522a89b8a964197c4ecebba8720f30d53aa 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4918,12 +4918,12 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         final int minChunkZ = minBlockZ >> 4;
         final int maxChunkZ = maxBlockZ >> 4;
 
-        final net.minecraft.world.level.chunk.ChunkSource chunkSource = world.getChunkSource();
+        final net.minecraft.server.level.ServerChunkCache chunkSource = ((ServerLevel) world).chunkSource; // Leaf - optimize get chunk
 
         for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
             for (int currChunkX = minChunkX; currChunkX <= maxChunkX; ++currChunkX) {
                 // Leaf start - Prevent double chunk retrieving in entity fluid pushing check and fluid height updating
-                final net.minecraft.world.level.chunk.ChunkAccess chunk = chunkSource.getChunk(currChunkX, currChunkZ, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, false);
+                final net.minecraft.world.level.chunk.LevelChunk chunk = chunkSource.getChunkAtIfLoadedUnchecked(currChunkX, currChunkZ); // Leaf - optimize get chunk
                 if (chunk == null) continue;
                 final net.minecraft.world.level.chunk.LevelChunkSection[] sections = chunk.getSections();
                 // Leaf end - Prevent double chunk retrieving in entity fluid pushing check and fluid height updating
@@ -4938,6 +4938,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     if (section.hasOnlyAir()) {
                         // empty
                         continue;
+                    } else if (fluid == FluidTags.WATER && section.waterFluidCount == 0) {
+                        continue;
+                    } else if (fluid == FluidTags.LAVA && section.lavaFluidCount == 0) {
+                        continue;
                     }
 
                     final net.minecraft.world.level.chunk.PalettedContainer<net.minecraft.world.level.block.state.BlockState> blocks = section.states;
@@ -4992,7 +4996,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
         this.fluidHeight.put(fluid, maxHeightDiff);
 
-        if (pushVector.lengthSqr() == 0.0) {
+        if (pushVector == Vec3.ZERO || pushVector.lengthSqr() == 0.0) {
             return inFluid;
         }
 
