From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 15:43:24 +0900
Subject: [PATCH] optimize updateFluidOnEyes


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 8c01317e1aa5cc588eb2d9e4d47de1da93d5146a..4ba145554af9c6aa580774f941f1713352d8078d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2152,8 +2152,9 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 && abstractBoat.getBoundingBox().maxY >= eyeY
                 && abstractBoat.getBoundingBox().minY <= eyeY
         )) {
-            BlockPos blockPos = BlockPos.containing(this.getX(), eyeY, this.getZ());
-            FluidState fluidState = this.level().getFluidState(blockPos);
+            BlockPos blockPos = leaf$getEyeBlockPos(); // Leaf
+            FluidState fluidState = this.level().getFluidStateIfLoadedUnchecked(blockPos.getX(), blockPos.getY(), blockPos.getZ()); // Leaf
+            if (fluidState == null) { fluidState = this.level().getFluidState(blockPos); } // Leaf
             double d = blockPos.getY() + fluidState.getHeight(this.level(), blockPos);
             if (d > eyeY) {
                 // Leaf start - Optimize isEyeInFluid
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 7c2bbd19c6f25e81e7c9967ea69376110a591561..ded9a9472e38fc2b25bb9d33e3d39f4f72191989 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -464,7 +464,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             if (this.isEyeInWater() // Leaf - Optimize isEyeInFluid
-                && !serverLevel1.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
+                && !serverLevel1.getBlockState(leaf$getEyeBlockPos()).is(Blocks.BUBBLE_COLUMN)) { // Leaf
                 boolean flag1 = !this.canBreatheUnderwater()
                     && !MobEffectUtil.hasWaterBreathing(this)
                     && (!flag || !((Player)this).getAbilities().invulnerable);
