From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 9 Aug 2025 16:44:28 +0900
Subject: [PATCH] optimize LivingEntity#travel


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index bb51c9a9afb1c4765869d87c31c36c49741cc2ce..c7c5338bc1b25a5ff0936946733bf56c786542e4 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3083,7 +3083,11 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public void travel(Vec3 travelVector) {
-        FluidState fluidState = this.level().getFluidState(this.blockPosition());
+        // Leaf start
+        BlockPos blockPos = this.blockPosition();
+        FluidState fluidState = this.level().getFluidStateIfLoadedUnchecked(blockPos.getX(), blockPos.getY(), blockPos.getZ());
+        if (fluidState == null) { this.level().getFluidState(blockPos); }
+        // Leaf end
         if ((this.isInWater() || this.isInLava()) && this.isAffectedByFluids() && !this.canStandOnFluid(fluidState)) {
             this.travelInFluid(travelVector);
         } else if (this.isFallFlying()) {
