From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix MC-270656

Related MC issue: https://mojira.dev/MC-270656

diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index cf22a560c331f08f7c1729ad5a40d966e1726a96..1aa47d5f25de08114abd9a179999b9993e3f0ef4 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1090,6 +1090,13 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     public void trackStartFallingPosition() {
+        // Leaf start - Fix MC-270656
+        if (org.dreeam.leaf.config.modules.fixes.MCBugFix.mc270656) {
+            if (this.currentImpulseImpactPos != null && (this.getAbilities().flying || this.isFallFlying())) {
+                this.currentImpulseImpactPos = null;
+            }
+        }
+        // Leaf end - Fix MC-270656
         if (this.fallDistance > 0.0 && this.startingToFallPosition == null) {
             this.startingToFallPosition = this.position();
             if (this.currentImpulseImpactPos != null && this.currentImpulseImpactPos.y <= this.startingToFallPosition.y) {
