From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sun, 10 Aug 2025 16:14:07 +0900
Subject: [PATCH] Replace fluid height map


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 8a43dfcb5b8c041a80bd01611bdc8865a754e8cd..7dda8c5db2bbbfe04db07b1a053489b4d19f3f09 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -286,7 +286,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public int tickCount;
     private int remainingFireTicks;
     public boolean wasTouchingWater;
-    protected Object2DoubleMap<TagKey<Fluid>> fluidHeight = new Object2DoubleArrayMap<>(2);
+    protected Object2DoubleMap<TagKey<Fluid>> fluidHeight = new org.dreeam.leaf.util.map.FluidHeightMap(); // Leaf - Replace fluid height map
     protected boolean wasEyeInWater;
     // Leaf start - Optimize isEyeInFluid
     // Remove original field since plugin should not direct access to it, and able to
