From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:31:35 +0900
Subject: [PATCH] Only update frozen ticks if changed


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index b50815f2c67ce8407f31d18af269a110fbb0b3d7..c0ad6c5b6b10534099e3fd5f7e0c18b3e2d2c26e 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3708,7 +3708,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
         if (this.level() instanceof ServerLevel serverLevel) {
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // Leaf start - Only update frozen ticks if changed
+                int ticksForzen = this.getTicksFrozen();
+                int newTicksForzen = Math.max(0, ticksForzen - 2);
+                if (ticksForzen != newTicksForzen) {
+                    this.setTicksFrozen(newTicksForzen);
+                }
+                // Leaf end - Only update frozen ticks if changed
             }
 
             this.removeFrost();
