From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 23 Jan 2026 18:47:34 +0900
Subject: [PATCH] fix skeleton horse trap NPE

Prevent removing the goal during the goals iteration.

diff --git a/net/minecraft/world/entity/animal/equine/SkeletonHorse.java b/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
index 7df95f57044a29d51af9d898535f1794941d3f31..e260669a389c554bdbc6ebc1e40c899872c7780d 100644
--- a/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
+++ b/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
@@ -205,8 +205,10 @@ public class SkeletonHorse extends AbstractHorse {
             this.isTrap = trap;
             if (trap) {
                 this.goalSelector.addGoal(1, this.skeletonTrapGoal);
+                this.skeletonTrapGoal.doTrap = true; // Leaf - fix skeleton horse trap NPE
             } else {
-                this.goalSelector.removeGoal(this.skeletonTrapGoal);
+                //this.goalSelector.removeGoal(this.skeletonTrapGoal); // Leaf - fix skeleton horse trap NPE
+                this.skeletonTrapGoal.doTrap = false; // Leaf - fix skeleton horse trap NPE
             }
         }
     }
diff --git a/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java b/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
index 686262974f067acdcd01e21dedcfababeeb72791..e551597d5f2371178c7a718e857fee3ca13eac79 100644
--- a/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
+++ b/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
@@ -19,6 +19,7 @@ import org.jspecify.annotations.Nullable;
 public class SkeletonTrapGoal extends Goal {
     private final SkeletonHorse horse;
     private java.util.List<org.bukkit.entity.HumanEntity> eligiblePlayers; // Paper
+    public boolean doTrap = true; // Leaf - fix skeleton horse trap NPE
 
     public SkeletonTrapGoal(final SkeletonHorse horse) {
         this.horse = horse;
@@ -26,7 +27,7 @@ public class SkeletonTrapGoal extends Goal {
 
     @Override
     public boolean canUse() {
-        return !(this.eligiblePlayers = this.horse.level().findNearbyBukkitPlayers(this.horse.getX(), this.horse.getY(), this.horse.getZ(), 10.0, net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING)).isEmpty(); // Paper - Affects Spawning API & SkeletonHorseTrapEvent
+        return doTrap && !(this.eligiblePlayers = this.horse.level().findNearbyBukkitPlayers(this.horse.getX(), this.horse.getY(), this.horse.getZ(), 10.0, net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING)).isEmpty(); // Paper - Affects Spawning API & SkeletonHorseTrapEvent // Leaf - fix skeleton horse trap NPE
     }
 
     @Override
@@ -102,4 +103,11 @@ public class SkeletonTrapGoal extends Goal {
         );
         skeleton.setItemSlot(slot, stack);
     }
+
+    // Leaf start - fix eligiblePlayers leak
+    @Override
+    public void stop() {
+        eligiblePlayers = null;
+    }
+    // Leaf end - fix eligiblePlayers leak
 }
