From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 12 Aug 2025 21:44:59 +0900
Subject: [PATCH] optimize get eye block position


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index a564635f274c95d1fe43a9485dd838ce0618e716..7ae9422903c8162b43d15480f26de50f8646ecf6 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -553,20 +553,14 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     // Leaf start - Fix Pufferfish and Purpur patches
     // Gale start - JettPack - optimize sun burn tick - cache eye blockpos
-    private BlockPos cached_eye_blockpos;
-    private net.minecraft.world.phys.Vec3 cached_position;
+    private BlockPos leaf$cached_eye_blockpos = BlockPos.ZERO;
+    private BlockPos leaf$cached_position = BlockPos.ZERO;
     // Gale end - JettPack - optimize sun burn tick - cache eye blockpos
     // Purpur start - copied from Mob - API for any mob to burn daylight
     public boolean isSunBurnTick() {
         if (this.level().isBrightOutside() && !this.level().isClientSide) {
             // Gale start - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
-            if (this.cached_position != this.position) {
-                this.cached_eye_blockpos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
-                this.cached_position = this.position;
-            }
-
-            float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue(cached_eye_blockpos); // Pass BlockPos to getBrightness
-
+            float lightLevelDependentMagicValue = this.getLightLevelDependentMagicValue();
             // Check brightness first
             if (lightLevelDependentMagicValue <= 0.5F) return false;
             if (this.random.nextFloat() * 30.0F >= (lightLevelDependentMagicValue - 0.4F) * 2.0F) return false;
@@ -574,11 +568,22 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
             boolean flag = this.isInWaterOrRain() || this.isInPowderSnow || this.wasInPowderSnow;
 
-            return !flag && this.level().canSeeSky(this.cached_eye_blockpos); // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
+            return !flag && this.level().canSeeSky(this.leaf$getEyeBlockPos()); // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos
         }
 
         return false; // Gale - JettPack - optimize sun burn tick - optimizations and cache eye blockpos - diff on change
     }
+
+    // Leaf start
+    public BlockPos leaf$getEyeBlockPos() {
+        if (this.leaf$cached_position != this.blockPosition) {
+            this.leaf$cached_eye_blockpos = this.blockPosition.atY(Mth.floor(this.getEyeY()));
+            this.leaf$cached_position = this.blockPosition;
+        }
+        return this.leaf$cached_eye_blockpos;
+    }
+    // Leaf end
+
     // Purpur end - copied from Mob - API for any mob to burn daylight
     // Leaf end - Fix Pufferfish and Purpur patches
 
@@ -2259,20 +2264,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     @Deprecated
     public float getLightLevelDependentMagicValue() {
-        return this.getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())); // Gale - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
-    }
-
-    // Gale start - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
-    /**
-     * @deprecated
-     */
-    @Deprecated
-    public float getLightLevelDependentMagicValue(BlockPos pos) {
-        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ())
-            ? this.level.getLightLevelDependentMagicValue(pos)
+        return this.level.hasChunkAt(this.getBlockX(), this.getBlockZ()) // Leaf
+            ? this.level.getLightLevelDependentMagicValue(leaf$getEyeBlockPos()) // Leaf
             : 0.0F;
     }
-    // Gale end - JettPack - optimize sun burn tick - allow passing BlockPos to getLightLevelDependentMagicValue
 
     public void absSnapTo(double x, double y, double z, float yRot, float xRot) {
         this.absSnapTo(x, y, z);
