From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Mon, 11 Aug 2025 02:19:59 +0900
Subject: [PATCH] cache collision list


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index a58ba94171b54feee82e4ec3d68d23f474507d8f..b9244ebe37f2cdd36d8c2188bd353826cb122cc2 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1110,6 +1110,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     public final org.dreeam.leaf.world.DespawnMap despawnMap = new org.dreeam.leaf.world.DespawnMap(paperConfig()); // Leaf - optimize despawn
     public final org.dreeam.leaf.world.NatureSpawnChunkMap natureSpawnChunkMap = new org.dreeam.leaf.world.NatureSpawnChunkMap(); // Leaf - optimize mob spawning
     public final org.dreeam.leaf.world.RandomTickSystem randomTickSystem = new org.dreeam.leaf.world.RandomTickSystem(); // Leaf - optimize random tick
+    public final org.dreeam.leaf.world.EntityCollisionCache entityCollisionCache = new org.dreeam.leaf.world.EntityCollisionCache(); // Leaf
+
     public void tickChunk(LevelChunk chunk, int randomTickSpeed) {
         final net.minecraft.world.level.levelgen.BitRandomSource simpleRandom = this.simpleRandom; // Paper - optimise random ticking // Leaf - Faster random generator - upcasting
         ChunkPos pos = chunk.getPos();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 9764e50b4bd7d82257ae17dab7b4193a032e4150..21a0e68dd3e5e273c27b8aed0d1f0e8601182b2f 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1617,8 +1617,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
         final AABB currentBox = this.getBoundingBox();
 
-        final List<VoxelShape> potentialCollisionsVoxel = new ArrayList<>();
-        final List<AABB> potentialCollisionsBB = new ArrayList<>();
+        // final List<VoxelShape> potentialCollisionsVoxel = new ArrayList<>(); // Leaf
+        // final List<AABB> potentialCollisionsBB = new ArrayList<>(); // Leaf
 
         final AABB initialCollisionBox;
         if (xZero & zZero) {
@@ -1630,17 +1630,17 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             initialCollisionBox = currentBox.expandTowards(movement);
         }
 
-        final List<AABB> entityAABBs = new ArrayList<>();
+        org.dreeam.leaf.world.EntityCollisionCache entityCollisionCache = ((ServerLevel) this.level).entityCollisionCache; // Leaf
         ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getEntityHardCollisions(
-            this.level, (Entity)(Object)this, initialCollisionBox, entityAABBs, 0, null
+            this.level, (Entity)(Object)this, initialCollisionBox, entityCollisionCache.entityAABBs(), 0, null // Leaf
         );
 
         ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
-            this.level, (Entity)(Object)this, initialCollisionBox, potentialCollisionsVoxel, potentialCollisionsBB,
+            this.level, (Entity)(Object)this, initialCollisionBox, entityCollisionCache.potentialCollisionsVoxel(), entityCollisionCache.potentialCollisionsBB(),
             ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER, null
         );
-        potentialCollisionsBB.addAll(entityAABBs);
-        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB);
+        entityCollisionCache.potentialCollisionsBB().addAll(entityCollisionCache.entityAABBs()); // Leaf
+        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, entityCollisionCache.potentialCollisionsVoxel(), entityCollisionCache.potentialCollisionsBB()); // Leaf
 
         final boolean collidedX = collided.x != movement.x;
         final boolean collidedY = collided.y != movement.y;
@@ -1651,6 +1651,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         final double stepHeight;
 
         if ((!collidedDownwards && !this.onGround) || (!collidedX && !collidedZ) || (stepHeight = (double)this.maxUpStep()) <= 0.0) {
+            entityCollisionCache.clear(); // Leaf
             return collided;
         }
 
@@ -1661,7 +1662,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
 
         final List<VoxelShape> stepVoxels = new ArrayList<>();
-        final List<AABB> stepAABBs = entityAABBs;
+        final List<AABB> stepAABBs = entityCollisionCache.entityAABBs(); // Leaf
 
         ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
             this.level, (Entity)(Object)this, stepRetrievalBox, stepVoxels, stepAABBs,
@@ -1675,6 +1676,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             }
         }
 
+        entityCollisionCache.clear(); // Leaf
         return collided;
         // Paper end - optimise collisions
     }
