From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sun, 31 Aug 2025 00:04:28 +0900
Subject: [PATCH] configurable goal update interval for mob ai


diff --git a/net/minecraft/world/entity/ai/goal/AvoidEntityGoal.java b/net/minecraft/world/entity/ai/goal/AvoidEntityGoal.java
index 7651676e72fcec52d7c1f9f7d7b6f9e585015c4d..e23eab1971932e2206ffe7bca05a4b844342143d 100644
--- a/net/minecraft/world/entity/ai/goal/AvoidEntityGoal.java
+++ b/net/minecraft/world/entity/ai/goal/AvoidEntityGoal.java
@@ -65,8 +65,10 @@ public class AvoidEntityGoal<T extends LivingEntity> extends Goal {
         this(mob, avoidClass, livingEntity -> true, maxDistance, walkSpeedModifier, sprintSpeedModifier, predicateOnAvoidEntity);
     }
 
+    private final int reciprocal = Math.max(1, reducedTickDelay(org.dreeam.leaf.config.modules.opt.EntityGoal.chanceAvoidEntity)); private int counter; // Leaf
     @Override
     public boolean canUse() {
+        counter += 1; if (counter < reciprocal) { return false; } else { counter = this.mob.random.nextInt(reciprocal); } // Leaf
         this.toAvoid = getServerLevel(this.mob)
             .getNearestEntity(
                 this.mob.level().getEntitiesOfClass(this.avoidClass, this.mob.getBoundingBox().inflate(this.maxDist, 3.0, this.maxDist), livingEntity -> true),
diff --git a/net/minecraft/world/entity/ai/goal/FollowParentGoal.java b/net/minecraft/world/entity/ai/goal/FollowParentGoal.java
index 3093f03d4f298bf39fec8bad2b6c22518774aea8..1e844ba7c2c9981ff0d446743080f6e58570ff4e 100644
--- a/net/minecraft/world/entity/ai/goal/FollowParentGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowParentGoal.java
@@ -19,8 +19,10 @@ public class FollowParentGoal extends Goal {
         this.speedModifier = speedModifier;
     }
 
+    private final int reciprocal = Math.max(1, reducedTickDelay(org.dreeam.leaf.config.modules.opt.EntityGoal.chanceFollowParent)); private int counter; // Leaf
     @Override
     public boolean canUse() {
+        counter += 1; if (counter < reciprocal) { return false; } else { counter = this.animal.random.nextInt(reciprocal); } // Leaf
         if (this.animal.getAge() >= 0) {
             return false;
         } else {
diff --git a/net/minecraft/world/entity/ai/goal/MoveThroughVillageGoal.java b/net/minecraft/world/entity/ai/goal/MoveThroughVillageGoal.java
index f3b1a34497dd5496c8a19279dcbaad673a804337..bd8de045836155ad288ca084751015c4c2a02125 100644
--- a/net/minecraft/world/entity/ai/goal/MoveThroughVillageGoal.java
+++ b/net/minecraft/world/entity/ai/goal/MoveThroughVillageGoal.java
@@ -44,8 +44,10 @@ public class MoveThroughVillageGoal extends Goal {
         }
     }
 
+    private final int reciprocal = Math.max(1, reducedTickDelay(org.dreeam.leaf.config.modules.opt.EntityGoal.chanceMoveThroughVillage)); private int counter; // Leaf
     @Override
     public boolean canUse() {
+        counter += 1; if (counter < reciprocal) { return false; } else { counter = this.mob.random.nextInt(reciprocal); } // Leaf
         if (!GoalUtils.hasGroundPathNavigation(this.mob)) {
             return false;
         } else {
diff --git a/net/minecraft/world/entity/ai/goal/TemptGoal.java b/net/minecraft/world/entity/ai/goal/TemptGoal.java
index 7cb94eb75a1de0cef02d4953ee5896bd5e3c7abd..5d7e4eb689f0411865e85675f1d66cd8c3a93393 100644
--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -51,8 +51,10 @@ public class TemptGoal extends Goal {
         this.targetingConditions = globalTemptationLookupIndex >= 0 ? TEMPT_TARGETING.copy() : TEMPT_TARGETING.copy().selector((entity, level) -> this.shouldFollow(entity)); // Paper - optimise temptation checks - skip selector if we have a lookup index.
     }
 
+    private final int reciprocal = Math.max(1, reducedTickDelay(org.dreeam.leaf.config.modules.opt.EntityGoal.chanceTempt)); private int counter; // Leaf
     @Override
     public boolean canUse() {
+        counter += 1; if (counter < reciprocal) { return false; } else { counter = this.mob.random.nextInt(reciprocal); } // Leaf
         if (this.calmDown > 0) {
             this.calmDown--;
             return false;
diff --git a/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java b/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
index 024792900c8ab716e91ef512d2da22548075044d..6c5efe75b9468d55afc94a9b7afee3557514a7ab 100644
--- a/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
@@ -36,7 +36,7 @@ public class NearestAttackableTargetGoal<T extends LivingEntity> extends TargetG
     ) {
         super(mob, mustSee, mustReach);
         this.targetType = targetType;
-        this.randomInterval = reducedTickDelay(interval);
+        this.randomInterval = org.dreeam.leaf.config.modules.opt.EntityGoal.chanceTarget > 0 && interval <= 10 ? reducedTickDelay(org.dreeam.leaf.config.modules.opt.EntityGoal.chanceTarget) : reducedTickDelay(interval); // Leaf
         this.setFlags(EnumSet.of(Goal.Flag.TARGET));
         this.targetConditions = TargetingConditions.forCombat().range(this.getFollowDistance()).selector(selector);
     }
