From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:31:35 +0900
Subject: [PATCH] Only update frozen ticks if changed


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 20147a79fd49a8921345bf9b0a8df28d87826d60..cf7c8a63fdc141628c18cf1cb44678201b4e3aa2 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3713,7 +3713,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
         if (this.level() instanceof ServerLevel serverLevel) {
             if ((!this.isInPowderSnow || !this.canFreeze()) && !this.freezeLocked) { // Paper - Freeze Tick Lock API
-                this.setTicksFrozen(Math.max(0, this.getTicksFrozen() - 2));
+                // Leaf start - Only update frozen ticks if changed
+                int ticksForzen = this.getTicksFrozen();
+                int newTicksForzen = Math.max(0, ticksForzen - 2);
+                if (ticksForzen != newTicksForzen) {
+                    this.setTicksFrozen(newTicksForzen);
+                }
+                // Leaf end - Only update frozen ticks if changed
             }
 
             this.removeFrost();
