From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:23:45 +0900
Subject: [PATCH] cache entity isAlive


diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index f127231587fef0cef6767a259e0e1294650e1330..3d692ca93d43cc24be487d6893a21b785e7534df 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -327,7 +327,7 @@ public final class ActivationRange {
         }
         // special case always immunities
         // immunize brand-new entities, dead entities, and portal scenarios
-        if (entity.defaultActivationState || entity.tickCount < 20 * 10 || !entity.isAlive() || (entity.portalProcess != null && !entity.portalProcess.hasExpired()) || entity.portalCooldown > 0) {
+        if (entity.defaultActivationState || entity.tickCount < 20 * 10 || !entity.leaf$isAlive || (entity.portalProcess != null && !entity.portalProcess.hasExpired()) || entity.portalCooldown > 0) { // Leaf - cache entity isAlive
             return true;
         }
         // immunize leashed entities
diff --git a/net/minecraft/world/entity/AgeableMob.java b/net/minecraft/world/entity/AgeableMob.java
index 04875840085541ebfc7014868beec49bb7ab9976..b09fd026448e62d18f727032f36a1749137c54d9 100644
--- a/net/minecraft/world/entity/AgeableMob.java
+++ b/net/minecraft/world/entity/AgeableMob.java
@@ -158,7 +158,7 @@ public abstract class AgeableMob extends PathfinderMob {
 
                 this.forcedAgeTimer--;
             }
-        } else if (this.isAlive()) {
+        } else if (this.leaf$isAlive) { // Leaf - cache entity isAlive
             int age = this.getAge();
             if (age < 0) {
                 this.setAge(++age);
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 722b771ce7e3befb145250005262fe919951e273..331ef55a9951c423de809f2e3c2975c764b8593c 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -382,6 +382,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    public boolean leaf$isAlive = true; // Leaf - cache entity isAlive
 
     public void inactiveTick() {
     }
@@ -900,7 +901,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     // CraftBukkit start
     public void postTick() {
         // No clean way to break out of ticking once the entity has been copied to a new world, so instead we move the portalling later in the tick cycle
-        if (!(this instanceof ServerPlayer) && this.isAlive()) { // Paper - don't attempt to teleport dead entities
+        if (!(this instanceof ServerPlayer) && this.leaf$isAlive) { // Paper - don't attempt to teleport dead entities // Leaf - cache entity isAlive
             this.handlePortal();
         }
     }
@@ -1839,7 +1840,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     this.debugBlockIntersection(pos, false, false);
                     return true;
                 }
-                if (!this.isAlive()) {
+                if (!this.leaf$isAlive) { // Leaf - cache entity isAlive
                     return false;
                 }
                 VoxelShape entityInsideCollisionShape = blockState.getEntityInsideCollisionShape(level, pos, this);
@@ -2527,7 +2528,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public boolean canBeHitByProjectile() {
-        return this.isAlive() && this.isPickable();
+        return this.leaf$isAlive && this.isPickable(); // Leaf - cache entity isAlive
     }
 
     public boolean isPickable() {
@@ -5344,6 +5345,15 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return this.removalReason != null;
     }
 
+
+    // Leaf start - cache entity isAlive
+    public final void leaf$onAliveDirty() {
+        if (!(this instanceof ServerPlayer) || this.getBukkitEntityRaw() != null) {
+            this.leaf$isAlive = isAlive();
+        }
+    }
+    // Leaf end - cache entity isAlive
+
     @Nullable
     public Entity.RemovalReason getRemovalReason() {
         return this.removalReason;
@@ -5362,6 +5372,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         if (this.removalReason == null) {
             this.removalReason = removalReason;
         }
+        this.leaf$onAliveDirty(); // Leaf - cache entity isAlive
 
         if (this.removalReason.shouldDestroy()) {
             this.stopRiding();
@@ -5380,6 +5391,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
     public void unsetRemoved() {
         this.removalReason = null;
+        this.leaf$onAliveDirty(); // Leaf - cache entity isAlive
     }
 
     // Paper start - Folia schedulers
@@ -5454,7 +5466,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public Vec3 getKnownMovement() {
-        return this.getControllingPassenger() instanceof Player player && this.isAlive() ? player.getKnownMovement() : this.getDeltaMovement();
+        return this.getControllingPassenger() instanceof Player player && this.leaf$isAlive ? player.getKnownMovement() : this.getDeltaMovement(); // Leaf - cache entity isAlive
     }
 
     @Nullable
diff --git a/net/minecraft/world/entity/EntitySelector.java b/net/minecraft/world/entity/EntitySelector.java
index a6a216c598978f719851856e4ea61e757ce47d7e..d3381bd8b27969c13b396444dbb84fd8ae53e5e4 100644
--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -7,10 +7,10 @@ import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.scores.Team;
 
 public final class EntitySelector {
-    public static final Predicate<Entity> ENTITY_STILL_ALIVE = Entity::isAlive;
-    public static final Predicate<Entity> LIVING_ENTITY_STILL_ALIVE = entity -> entity.isAlive() && entity instanceof LivingEntity;
-    public static final Predicate<Entity> ENTITY_NOT_BEING_RIDDEN = entity -> entity.isAlive() && !entity.isVehicle() && !entity.isPassenger();
-    public static final Predicate<Entity> CONTAINER_ENTITY_SELECTOR = entity -> entity instanceof Container && entity.isAlive();
+    public static final Predicate<Entity> ENTITY_STILL_ALIVE = e -> e.leaf$isAlive; // Leaf - cache entity isAlive
+    public static final Predicate<Entity> LIVING_ENTITY_STILL_ALIVE = entity -> entity.leaf$isAlive && entity instanceof LivingEntity; // Leaf - cache entity isAlive
+    public static final Predicate<Entity> ENTITY_NOT_BEING_RIDDEN = entity -> entity.leaf$isAlive && !entity.isVehicle() && !entity.isPassenger(); // Leaf - cache entity isAlive
+    public static final Predicate<Entity> CONTAINER_ENTITY_SELECTOR = entity -> entity instanceof Container && entity.leaf$isAlive; // Leaf - cache entity isAlive
     public static final Predicate<Entity> NO_CREATIVE_OR_SPECTATOR = entity -> !(
         entity instanceof Player player && (entity.isSpectator() || player.isCreative())
     );
@@ -32,7 +32,7 @@ public final class EntitySelector {
 
     // Paper start - Affects Spawning API
     public static final Predicate<Entity> PLAYER_AFFECTS_SPAWNING = (entity) -> {
-        return !entity.isSpectator() && entity.isAlive() && entity instanceof Player player && player.affectsSpawning; // Leaf - Optimize nearby alive players for spawning - diff on change
+        return !entity.isSpectator() && entity.leaf$isAlive && entity instanceof Player player && player.affectsSpawning; // Leaf - Optimize nearby alive players for spawning - diff on change // Leaf - cache entity isAlive
     };
     // Paper end - Affects Spawning API
 
diff --git a/net/minecraft/world/entity/InsideBlockEffectApplier.java b/net/minecraft/world/entity/InsideBlockEffectApplier.java
index eba3896b73c70f29732388373e89bf7ce3a07c92..61da4c7d7dcede051995ca23586dcbd9880bdf43 100644
--- a/net/minecraft/world/entity/InsideBlockEffectApplier.java
+++ b/net/minecraft/world/entity/InsideBlockEffectApplier.java
@@ -70,7 +70,7 @@ public interface InsideBlockEffectApplier {
                 return;
             }
 
-            if (!entity.isAlive()) {
+            if (!entity.leaf$isAlive) {
                 effects.clear();
                 this.lastStep = -1;
                 return;
@@ -82,15 +82,15 @@ public interface InsideBlockEffectApplier {
                 effects.get(i++).accept(entity);
                 effects.get(i++).accept(entity);
                 effects.get(i++).accept(entity);
-                if (!entity.isAlive()) {
+                if (!entity.leaf$isAlive) {
                     break;
                 }
             }
 
-            if (entity.isAlive()) {
+            if (entity.leaf$isAlive) {
                 for (; i < size; i++) {
                     effects.get(i).accept(entity);
-                    if (!entity.isAlive()) {
+                    if (!entity.leaf$isAlive) {
                         break;
                     }
                 }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 1799cdc0b6c1e585e7e1eeab3828ea0252ae2097..e6753df8a4dafe96df7d7dcd7a4551757b10ca1c 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -440,7 +440,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             this.clearFire();
         }
 
-        if (this.isAlive() && this.level() instanceof ServerLevel serverLevel1) {
+        if (this.leaf$isAlive && this.level() instanceof ServerLevel serverLevel1) { // Leaf - cache entity isAlive
             boolean flag = this instanceof Player;
             // Gale start - Pufferfish - reduce in wall checks
             long checkStuckInWallInterval = this.level().galeConfig().smallOptimizations.reducedIntervals.checkStuckInWall;
@@ -511,13 +511,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             this.lastHurtByPlayer = null;
         }
 
-        if (this.lastHurtMob != null && !this.lastHurtMob.isAlive()) {
+        if (this.lastHurtMob != null && !this.lastHurtMob.leaf$isAlive) { // Leaf - cache entity isAlive
             this.lastHurtMob = null;
         }
 
         LivingEntity lastHurtByMob = this.getLastHurtByMob();
         if (lastHurtByMob != null) {
-            if (!lastHurtByMob.isAlive()) {
+            if (!lastHurtByMob.leaf$isAlive) { // Leaf - cache entity isAlive
                 this.setLastHurtByMob(null);
             } else if (this.tickCount - this.lastHurtByMobTimestamp > 100) {
                 this.setLastHurtByMob(null);
@@ -1097,7 +1097,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public boolean canBeSeenByAnyone() {
-        return !this.isSpectator() && this.isAlive();
+        return !this.isSpectator() && this.leaf$isAlive; // Leaf - cache entity isAlive
     }
 
     public static boolean areAllEffectsAmbient(Collection<MobEffectInstance> potionEffects) {
@@ -1933,6 +1933,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 } else {
                     this.dead = false;
                     this.setHealth((float) deathEvent.getReviveHealth());
+                    this.leaf$onAliveDirty(); // Leaf - cache entity isAlive
                 }
                 // Paper end
                     this.createWitherRose(killCredit);
@@ -3759,7 +3760,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         }
 
         // Purpur start - copied from Zombie - API for any mob to burn daylight
-        if (this.isAlive()) {
+        if (this.leaf$isAlive) { // Leaf - cache entity isAlive
             boolean flag = this.shouldBurnInDay() && this.isSunBurnTick(); // Paper - shouldBurnInDay API // Purpur - use shouldBurnInDay() method to handle Phantoms properly - API for any mob to burn daylight
             if (flag) {
                 ItemStack itemBySlot = this.getItemBySlot(EquipmentSlot.HEAD);
@@ -4090,7 +4091,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     @Override
     public boolean isCollidable(boolean ignoreClimbing) {
-        return this.isAlive() && !this.isSpectator() && (ignoreClimbing || !this.onClimbable()) && this.collides; // CraftBukkit
+        return this.leaf$isAlive && !this.isSpectator() && (ignoreClimbing || !this.onClimbable()) && this.collides; // CraftBukkit // Leaf - cache entity isAlive
         // Paper end - Climbing should not bypass cramming gamerule
     }
 
@@ -4274,6 +4275,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     @Override
     public void onSyncedDataUpdated(EntityDataAccessor<?> key) {
         super.onSyncedDataUpdated(key);
+        if (DATA_HEALTH_ID.equals(key)) { this.leaf$onAliveDirty(); } else // Leaf - cache entity isAlive
         if (SLEEPING_POS_ID.equals(key)) {
             if (this.level().isClientSide) {
                 this.getSleepingPos().ifPresent(this::setPosToBed);
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 28523643d0789c6f472dfa92c3c0a32f0bcfc7b5..19ea4c589527a8ecacef5973ac23abcf734ff841 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -326,7 +326,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     @Override
     public void baseTick() {
         super.baseTick();
-        if (this.isAlive() && this.random.nextInt(1000) < this.ambientSoundTime++) {
+        if (this.leaf$isAlive && this.random.nextInt(1000) < this.ambientSoundTime++) { // Leaf - cache entity isAlive
             this.resetAmbientSoundTime();
             this.playAmbientSound();
         }
@@ -535,8 +535,8 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         super.aiStep();
         if (this.level() instanceof ServerLevel serverLevel
             && this.canPickUpLoot()
-            && this.isAlive()
-            && !this.dead
+            && this.leaf$isAlive // Leaf - cache entity isAlive
+            // && !this.dead // Leaf - cache entity isAlive
             && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING, serverLevel.purpurConfig.entitiesPickUpLootMobGriefingOverride)) { // Purpur - Add mobGriefing override to everything affected
             Vec3i pickupReach = this.getPickupReach();
 
diff --git a/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java b/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
index 51a6302ed5daf512acc6bed915f954ac76559270..540e08218da94fdca9e64d16ac6ab8af414f736c 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
@@ -19,7 +19,7 @@ public class NearestLivingEntitySensor<T extends LivingEntity> extends Sensor<T>
         AABB aabb = entity.getBoundingBox().inflate(attributeValue, attributeValue, attributeValue);
         // Leaf start - fast bit radix sort
         it.unimi.dsi.fastutil.objects.ObjectArrayList<LivingEntity> entitiesOfClass = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
-        ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel) level).moonrise$getEntityLookup().getEntities(LivingEntity.class, entity, aabb, entitiesOfClass, LivingEntity::isAlive);
+        ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel) level).moonrise$getEntityLookup().getEntities(LivingEntity.class, entity, aabb, entitiesOfClass, e -> e.leaf$isAlive); // Leaf - cache entity isAlive
         level.fastBitRadixSort.sort(entitiesOfClass.elements(), entitiesOfClass.size(), entity.position());
         // Leaf end - fast bit radix sort
         Brain<?> brain = entity.getBrain();
diff --git a/net/minecraft/world/entity/animal/AgeableWaterCreature.java b/net/minecraft/world/entity/animal/AgeableWaterCreature.java
index 1c244aaca4c29b79db7e8d6cda184e3b3cdcfced..a80530a7511b7fbe08a6ace4be65c65bebe5c36f 100644
--- a/net/minecraft/world/entity/animal/AgeableWaterCreature.java
+++ b/net/minecraft/world/entity/animal/AgeableWaterCreature.java
@@ -35,7 +35,7 @@ public abstract class AgeableWaterCreature extends AgeableMob {
     }
 
     protected void handleAirSupply(int airSupply) {
-        if (this.isAlive() && !this.isInWater()) {
+        if (this.leaf$isAlive && !this.isInWater()) { // Leaf - cache entity isAlive
             this.setAirSupply(airSupply - 1);
             if (this.getAirSupply() == -20) {
                 this.setAirSupply(0);
diff --git a/net/minecraft/world/entity/animal/Chicken.java b/net/minecraft/world/entity/animal/Chicken.java
index a5821ab8eab48fc2cd29eba857c9a624eb59290d..40bcd9b7b3ccc1a81dba330d6d1213c033d29b0a 100644
--- a/net/minecraft/world/entity/animal/Chicken.java
+++ b/net/minecraft/world/entity/animal/Chicken.java
@@ -171,7 +171,7 @@ public class Chicken extends Animal {
         }
 
         this.flap = this.flap + this.flapping * 2.0F;
-        if (this.level() instanceof ServerLevel serverLevel && this.isAlive() && !this.isBaby() && !this.isChickenJockey() && --this.eggTime <= 0) {
+        if (this.level() instanceof ServerLevel serverLevel && this.leaf$isAlive && !this.isBaby() && !this.isChickenJockey() && --this.eggTime <= 0) { // Leaf - cache entity isAlive
             this.forceDrops = true; // CraftBukkit
             if (this.dropFromGiftLootTable(serverLevel, BuiltInLootTables.CHICKEN_LAY, this::spawnAtLocation)) {
                 this.playSound(SoundEvents.CHICKEN_EGG, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
diff --git a/net/minecraft/world/entity/animal/Fox.java b/net/minecraft/world/entity/animal/Fox.java
index 619e7ec6ccfc076cd483a7215ab80e6bf4aaad83..5573c1726bc0c4667c0a41ef95e81f68a1f6446a 100644
--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -266,7 +266,7 @@ public class Fox extends Animal {
 
     @Override
     public void aiStep() {
-        if (!this.level().isClientSide && this.isAlive() && this.isEffectiveAi()) {
+        if (!this.level().isClientSide && this.leaf$isAlive && this.isEffectiveAi()) { // Leaf - cache entity isAlive
             this.ticksSinceEaten++;
             ItemStack itemBySlot = this.getItemBySlot(EquipmentSlot.MAINHAND);
             if (this.canEat(itemBySlot)) {
diff --git a/net/minecraft/world/entity/animal/HappyGhast.java b/net/minecraft/world/entity/animal/HappyGhast.java
index 788ee2632301c4a40b5c96e137d8cb0e96253a5f..3b44418618d8adbd3bce2b40fd15b78d3be8b61f 100644
--- a/net/minecraft/world/entity/animal/HappyGhast.java
+++ b/net/minecraft/world/entity/animal/HappyGhast.java
@@ -606,7 +606,7 @@ public class HappyGhast extends Animal {
     @Override
     public boolean canBeCollidedWith(@Nullable Entity entity) {
         return !this.isBaby()
-            && this.isAlive()
+            && this.leaf$isAlive // Leaf - cache entity isAlive
             && (
                 this.level().isClientSide() && entity instanceof Player && entity.position().y >= this.getBoundingBox().maxY
                     || this.isVehicle() && entity instanceof HappyGhast
diff --git a/net/minecraft/world/entity/animal/Pufferfish.java b/net/minecraft/world/entity/animal/Pufferfish.java
index 7c6ccac4ae5de741753a6e75936416b1402fc67d..ff432b44aa8cf7be4a72a66a9b3434a93fc2ba06 100644
--- a/net/minecraft/world/entity/animal/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/Pufferfish.java
@@ -128,7 +128,7 @@ public class Pufferfish extends AbstractFish {
 
     @Override
     public void tick() {
-        if (!this.level().isClientSide && this.isAlive() && this.isEffectiveAi()) {
+        if (!this.level().isClientSide && this.leaf$isAlive && this.isEffectiveAi()) { // Leaf - cache entity isAlive
             if (this.inflateCounter > 0) {
                 boolean increase = true; // Paper - Add PufferFishStateChangeEvent
                 if (this.getPuffState() == 0) {
@@ -170,10 +170,10 @@ public class Pufferfish extends AbstractFish {
     @Override
     public void aiStep() {
         super.aiStep();
-        if (this.level() instanceof ServerLevel serverLevel && this.isAlive() && this.getPuffState() > 0) {
+        if (this.level() instanceof ServerLevel serverLevel && this.leaf$isAlive && this.getPuffState() > 0) { // Leaf - cache entity isAlive
             for (Mob mob : this.level()
                 .getEntitiesOfClass(Mob.class, this.getBoundingBox().inflate(0.3), mob1 -> TARGETING_CONDITIONS.test(serverLevel, this, mob1))) {
-                if (mob.isAlive()) {
+                if (mob.leaf$isAlive) { // Leaf - cache entity isAlive
                     this.touch(serverLevel, mob);
                 }
             }
diff --git a/net/minecraft/world/entity/animal/Turtle.java b/net/minecraft/world/entity/animal/Turtle.java
index 59800c502c46c82cedde0f67e3848d7ba02cc479..66c020b2b279a7282e7cbfc772218a46fdd0473a 100644
--- a/net/minecraft/world/entity/animal/Turtle.java
+++ b/net/minecraft/world/entity/animal/Turtle.java
@@ -294,7 +294,7 @@ public class Turtle extends Animal {
     @Override
     public void aiStep() {
         super.aiStep();
-        if (this.isAlive() && this.isLayingEgg() && this.layEggCounter >= 1 && this.layEggCounter % 5 == 0) {
+        if (this.leaf$isAlive && this.isLayingEgg() && this.layEggCounter >= 1 && this.layEggCounter % 5 == 0) { // Leaf - cache entity isAlive
             BlockPos blockPos = this.blockPosition();
             if (TurtleEggBlock.onSand(this.level(), blockPos)) {
                 this.level().levelEvent(2001, blockPos, Block.getId(this.level().getBlockState(blockPos.below())));
diff --git a/net/minecraft/world/entity/animal/WaterAnimal.java b/net/minecraft/world/entity/animal/WaterAnimal.java
index ace8766a3640106cf471883029fa6a6272a890e7..22a732f157a79171f019feaf533fa4b879f99232 100644
--- a/net/minecraft/world/entity/animal/WaterAnimal.java
+++ b/net/minecraft/world/entity/animal/WaterAnimal.java
@@ -37,7 +37,7 @@ public abstract class WaterAnimal extends PathfinderMob {
     }
 
     protected void handleAirSupply(ServerLevel level, int airSupply) {
-        if (this.isAlive() && !this.isInWater()) {
+        if (this.leaf$isAlive && !this.isInWater()) { // Leaf - cache entity isAlive
             this.setAirSupply(airSupply - 1);
             if (this.getAirSupply() == -20) {
                 this.setAirSupply(0);
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 868dd1362112dd507ce2392023bfa9bc24686842..2314f9e835cb0c07a6e388eb3d95c578380a1b95 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -289,7 +289,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     @Override
     public void aiStep() {
         super.aiStep();
-        if (!this.level().isClientSide && this.isAlive() && this.tickCount % 10 == 0) {
+        if (!this.level().isClientSide && this.leaf$isAlive && this.tickCount % 10 == 0) { // Leaf - cache entity isAlive
             this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
         }
 
diff --git a/net/minecraft/world/entity/animal/armadillo/Armadillo.java b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
index 1a0ade86ca8d59d53b3d61be1417d0b840126b6c..90cc54cef40a4f0d214ac1a372d32dc5449b2bb0 100644
--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -169,7 +169,7 @@ public class Armadillo extends Animal {
         if (this.behaviorTick++ % this.activatedPriority == 0) // Leaf - Plazma - Add missing Pufferfish configurations
         ((Brain<Armadillo>)this.brain).tick(level, this);
         ArmadilloAi.updateActivity(this);
-        if (this.isAlive() && !this.isBaby() && --this.scuteTime <= 0) {
+        if (this.leaf$isAlive && !this.isBaby() && --this.scuteTime <= 0) { // Leaf - cache entity isAlive
             this.forceDrops = true; // CraftBukkit
             if (this.dropFromGiftLootTable(level, BuiltInLootTables.ARMADILLO_SHED, this::spawnAtLocation)) {
                 this.playSound(SoundEvents.ARMADILLO_SCUTE_DROP, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
diff --git a/net/minecraft/world/entity/animal/axolotl/Axolotl.java b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
index 45125fad187b101fbe72337c128bffc44dbb800c..b3a373e1ee3041ae5e78eee124f05ae69d5c3854 100644
--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -259,7 +259,7 @@ public class Axolotl extends Animal implements Bucketable {
     }
 
     protected void handleAirSupply(ServerLevel level, int airSupply) {
-        if (this.isAlive() && !this.isInWaterOrRain()) {
+        if (this.leaf$isAlive && !this.isInWaterOrRain()) { // Leaf - cache entity isAlive
             this.setAirSupply(airSupply - 1);
             if (this.getAirSupply() == -20) {
                 this.setAirSupply(0);
diff --git a/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index 3ce7e02b91d40170e040554cd601167b66ba602a..b182abbaef2e05570642d99685211a06a4a3fbe7 100644
--- a/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -590,7 +590,7 @@ public abstract class AbstractHorse extends Animal implements HasCustomInventory
         }
 
         super.aiStep();
-        if (this.level() instanceof ServerLevel serverLevel && this.isAlive()) {
+        if (this.level() instanceof ServerLevel serverLevel && this.leaf$isAlive) { // Leaf - cache entity isAlive
             if (this.random.nextInt(900) == 0 && this.deathTime == 0) {
                 this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
             }
diff --git a/net/minecraft/world/entity/animal/wolf/Wolf.java b/net/minecraft/world/entity/animal/wolf/Wolf.java
index 5d3fd6a6ce2eb3c462bfb10abdc817a114821ca1..fc70d4807f07b04bfb66c07de39c6011cf568f3c 100644
--- a/net/minecraft/world/entity/animal/wolf/Wolf.java
+++ b/net/minecraft/world/entity/animal/wolf/Wolf.java
@@ -445,7 +445,7 @@ public class Wolf extends TamableAnimal implements NeutralMob {
     @Override
     public void tick() {
         super.tick();
-        if (this.isAlive()) {
+        if (this.leaf$isAlive) { // Leaf - cache entity isAlive
             // Purpur start - Configurable chance for wolves to spawn rabid
             if (this.age % 300 == 0 && this.isRabid()) {
                 this.addEffect(new net.minecraft.world.effect.MobEffectInstance(net.minecraft.world.effect.MobEffects.NAUSEA, 400));
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index c99ddedaf25a009f0d19f97d01e7545b0c59aee2..a1c94dd67dd3830723806c6542064ee4f9c6ed2f 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -166,7 +166,7 @@ public class FallingBlockEntity extends Entity {
             }
             // Paper end - Configurable falling blocks height nerf
             this.handlePortal();
-            if (this.level() instanceof ServerLevel serverLevel && (this.isAlive() || this.forceTickAfterTeleportToDuplicate)) {
+            if (this.level() instanceof ServerLevel serverLevel && (this.leaf$isAlive || this.forceTickAfterTeleportToDuplicate)) { // Leaf - cache entity isAlive
                 BlockPos blockPos = this.blockPosition();
                 boolean flag = this.blockState.getBlock() instanceof ConcretePowderBlock;
                 boolean flag1 = flag && this.level().getFluidState(blockPos).is(FluidTags.WATER);
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index b4f72a15b117e1e5438f509a2d7392aa14886316..5315ae17e25d7b84a90a3c0dbbf537079c42d663 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -263,7 +263,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
             return;
         }
 
-        if (config.temporaryImmunity.duration > 0 && this.isAlive() && this.onGround && !this.isRemoved() && (config.temporaryImmunity.nearbyItemMaxAge == -1 || this.age <= config.temporaryImmunity.nearbyItemMaxAge) && this.age % Math.max(1, config.temporaryImmunity.checkForMinecartNearItemInterval) == 0 && config.temporaryImmunity.maxItemHorizontalDistance >= 0 && config.temporaryImmunity.maxItemVerticalDistance >= 0) {
+        if (config.temporaryImmunity.duration > 0 && this.leaf$isAlive && this.onGround && !this.isRemoved() && (config.temporaryImmunity.nearbyItemMaxAge == -1 || this.age <= config.temporaryImmunity.nearbyItemMaxAge) && this.age % Math.max(1, config.temporaryImmunity.checkForMinecartNearItemInterval) == 0 && config.temporaryImmunity.maxItemHorizontalDistance >= 0 && config.temporaryImmunity.maxItemVerticalDistance >= 0) { // Leaf - cache entity isAlive
             net.minecraft.world.phys.AABB aabb = this.getBoundingBox().inflate(config.temporaryImmunity.maxItemHorizontalDistance, config.temporaryImmunity.maxItemVerticalDistance, config.temporaryImmunity.maxItemHorizontalDistance);
             for (Entity entity : this.level().getEntities(this, aabb)) {
                 if (entity instanceof net.minecraft.world.entity.vehicle.MinecartHopper hopper) {
@@ -324,7 +324,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
 
     private boolean isMergable() {
         ItemStack item = this.getItem();
-        return this.isAlive() && this.pickupDelay != 32767 && this.age != -32768 && this.age < this.despawnRate && item.getCount() < item.getMaxStackSize(); // Paper - Alternative item-despawn-rate
+        return this.leaf$isAlive && this.pickupDelay != 32767 && this.age != -32768 && this.age < this.despawnRate && item.getCount() < item.getMaxStackSize(); // Paper - Alternative item-despawn-rate // Leaf - cache entity isAlive
     }
 
     private void tryToMerge(ItemEntity itemEntity) {
diff --git a/net/minecraft/world/entity/monster/Creeper.java b/net/minecraft/world/entity/monster/Creeper.java
index 9bcfd34ab0de8f3c12d01a9f26a97311384e3e72..c1a3595127f14c5e9cdee93891643b9f767248bf 100644
--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -210,7 +210,7 @@ public class Creeper extends Monster {
 
     @Override
     public void tick() {
-        if (this.isAlive()) {
+        if (this.leaf$isAlive) { // Leaf - cache entity isAlive
             this.oldSwell = this.swell;
             if (this.isIgnited()) {
                 this.setSwellDir(1);
@@ -370,6 +370,7 @@ public class Creeper extends Monster {
             if (!event.isCancelled()) {
             // CraftBukkit end
             this.dead = true;
+            this.leaf$onAliveDirty(); // Leaf - cache entity isAlive
             serverLevel.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), serverLevel.getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_MOBGRIEFING) && level().purpurConfig.creeperAllowGriefing ? Level.ExplosionInteraction.MOB : Level.ExplosionInteraction.NONE); // CraftBukkit // Paper - fix DamageSource API (revert to vanilla, no, just no, don't change this) // Purpur - Add enderman and creeper griefing controls
             this.spawnLingeringCloud();
             this.triggerOnDeathMobEffects(serverLevel, Entity.RemovalReason.KILLED);
diff --git a/net/minecraft/world/entity/monster/Guardian.java b/net/minecraft/world/entity/monster/Guardian.java
index 1ee5a2512dba74cf4949f5bf71c07dcf1275171f..4162e0fd1e3753ae2c5368ee330b5c99ebe6a312 100644
--- a/net/minecraft/world/entity/monster/Guardian.java
+++ b/net/minecraft/world/entity/monster/Guardian.java
@@ -227,7 +227,7 @@ public class Guardian extends Monster {
 
     @Override
     public void aiStep() {
-        if (this.isAlive()) {
+        if (this.leaf$isAlive) { // Leaf - cache entity isAlive
             if (this.level().isClientSide) {
                 this.clientSideTailAnimationO = this.clientSideTailAnimation;
                 if (!this.isInWater()) {
diff --git a/net/minecraft/world/entity/monster/Ravager.java b/net/minecraft/world/entity/monster/Ravager.java
index ac13a83d8b7fbbb57999b23869ac97d7fc006fd0..3e780df284d88830a37642193ea70c8370c849c3 100644
--- a/net/minecraft/world/entity/monster/Ravager.java
+++ b/net/minecraft/world/entity/monster/Ravager.java
@@ -182,7 +182,7 @@ public class Ravager extends Raider {
     @Override
     public void aiStep() {
         super.aiStep();
-        if (this.isAlive() && (getRider() == null || !this.isControllable())) { // Purpur - Ridables
+        if (this.leaf$isAlive && (getRider() == null || !this.isControllable())) { // Purpur - Ridables // Leaf - cache entity isAlive
             if (this.isImmobile()) {
                 this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(0.0);
             } else {
diff --git a/net/minecraft/world/entity/monster/Shulker.java b/net/minecraft/world/entity/monster/Shulker.java
index 9a0e634144ad852bb308ec957b21c1dc373e9d06..a653ca67cb9413b9919dbfbd7f6804137c8ec0f8 100644
--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -548,7 +548,7 @@ public class Shulker extends AbstractGolem implements Enemy {
 
     @Override
     public boolean canBeCollidedWith(@Nullable Entity entity) {
-        return this.isAlive();
+        return this.leaf$isAlive; // Leaf - cache entity isAlive
     }
 
     public Direction getAttachFace() {
diff --git a/net/minecraft/world/entity/monster/Skeleton.java b/net/minecraft/world/entity/monster/Skeleton.java
index 7683c6a2ab2a9775c7a7297f46b73724c44b4caa..18cbd7f1e6ea7e2b09a16687b13fd0c64572dfec 100644
--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -86,7 +86,7 @@ public class Skeleton extends AbstractSkeleton {
 
     @Override
     public void tick() {
-        if (!this.level().isClientSide && this.isAlive() && !this.isNoAi()) {
+        if (!this.level().isClientSide && this.leaf$isAlive && !this.isNoAi()) { // Leaf - cache entity isAlive
             if (this.isInPowderSnow) {
                 if (this.isFreezeConverting()) {
                     this.conversionTime--;
diff --git a/net/minecraft/world/entity/monster/Witch.java b/net/minecraft/world/entity/monster/Witch.java
index 5d0c0afc178cea8dd6fbd5e133693eb2c09e80ec..ac29ac1ed420f3c5be4a9318d1952626ffe309b5 100644
--- a/net/minecraft/world/entity/monster/Witch.java
+++ b/net/minecraft/world/entity/monster/Witch.java
@@ -149,7 +149,7 @@ public class Witch extends Raider implements RangedAttackMob {
 
     @Override
     public void aiStep() {
-        if (!this.level().isClientSide && this.isAlive()) {
+        if (!this.level().isClientSide && this.leaf$isAlive) { // Leaf - cache entity isAlive
             this.healRaidersGoal.decrementCooldown();
             if (this.healRaidersGoal.getCooldown() <= 0) {
                 this.attackPlayersGoal.setCanAttack(true);
diff --git a/net/minecraft/world/entity/monster/Zombie.java b/net/minecraft/world/entity/monster/Zombie.java
index 8a8b47366acf1d368d098a06d2996879a80df19a..243be5b98d304239406e06b6ad61bf6766c68060 100644
--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -279,7 +279,7 @@ public class Zombie extends Monster {
 
     @Override
     public void tick() {
-        if (!this.level().isClientSide && this.isAlive() && !this.isNoAi()) {
+        if (!this.level().isClientSide && this.leaf$isAlive && !this.isNoAi()) { // Leaf - cache entity isAlive
             if (this.isUnderWaterConverting()) {
                 this.conversionTime--;
                 if (this.conversionTime < 0) {
diff --git a/net/minecraft/world/entity/monster/ZombieVillager.java b/net/minecraft/world/entity/monster/ZombieVillager.java
index ced3734403c136f03a571e1e1c3afdde92cd2f6d..a1b4b3f15f632b5558aa067e69c28f63f7c17c17 100644
--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -175,7 +175,7 @@ public class ZombieVillager extends Zombie implements VillagerDataHolder {
 
     @Override
     public void tick() {
-        if (!this.level().isClientSide && this.isAlive() && this.isConverting()) {
+        if (!this.level().isClientSide && this.leaf$isAlive && this.isConverting()) { // Leaf - cache entity isAlive
             int conversionProgress = this.getConversionProgress();
             this.villagerConversionTime -= conversionProgress;
             if (this.villagerConversionTime <= 0) {
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 108ef0759e4060fc02d517175731f413c7e2532f..a9f85281e892351c929bb05eda5e3e0c77087835 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -367,7 +367,7 @@ public abstract class Player extends LivingEntity {
                 this.awardStat(Stats.PLAY_TIME, increaseTimeStatisticsInterval);
                 this.awardStat(Stats.TOTAL_WORLD_TIME, increaseTimeStatisticsInterval);
                 // Gale end - Hydrinity - increase time statistics in intervals
-            if (this.isAlive()) {
+            if (this.leaf$isAlive) { // Leaf - cache entity isAlive
                 this.awardStat(Stats.TIME_SINCE_DEATH, increaseTimeStatisticsInterval); // Gale - Hydrinity - increase time statistics in intervals
             }
 
diff --git a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index 4d6d843769e1ac9176fc1e90d429b8f0f2ec1f85..1f12276d3ce8e030f4efc1f27f641f20a88531c3 100644
--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -197,7 +197,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
         int sectionPosZ = SectionPos.blockToSectionCoord(this.position().z());
         Entity owner = this.getOwner();
         if (owner instanceof ServerPlayer serverPlayer
-            && !owner.isAlive()
+            && !owner.leaf$isAlive // Leaf - cache entity isAlive
             && serverPlayer.level().getGameRules().getBoolean(GameRules.RULE_ENDER_PEARLS_VANISH_ON_DEATH)) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
diff --git a/net/minecraft/world/entity/raid/Raid.java b/net/minecraft/world/entity/raid/Raid.java
index aaa990b01e886812cc1dbad819e922b23d08889c..0ea745dc800a3f7cb92233f5dd666edca7fd4004 100644
--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -223,7 +223,7 @@ public class Raid {
     private Predicate<ServerPlayer> validPlayer() {
         return player -> {
             BlockPos blockPos = player.blockPosition();
-            return player.isAlive() && player.level().getRaidAt(blockPos) == this;
+            return player.leaf$isAlive && player.level().getRaidAt(blockPos) == this; // Leaf - cache entity isAlive
         };
     }
 
diff --git a/net/minecraft/world/entity/raid/Raider.java b/net/minecraft/world/entity/raid/Raider.java
index 2254493c889b8967011c09dc448ba375d82e2035..a5daf048b9febd263ba6668723de7dbdc190e906 100644
--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -85,7 +85,7 @@ public abstract class Raider extends PatrollingMonster {
 
     @Override
     public void aiStep() {
-        if (this.level() instanceof ServerLevel serverLevel && this.isAlive()) {
+        if (this.level() instanceof ServerLevel serverLevel && this.leaf$isAlive) { // Leaf - cache entity isAlive
             Raid currentRaid = this.getCurrentRaid();
             if (this.canJoinRaid()) {
                 if (currentRaid == null) {
diff --git a/net/minecraft/world/entity/vehicle/MinecartHopper.java b/net/minecraft/world/entity/vehicle/MinecartHopper.java
index fb099517927f3c2699a348ea7692772c5420b8d1..2f863c912f060b137520f3ff7d97e3d29769f152 100644
--- a/net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ b/net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -96,7 +96,7 @@ public class MinecartHopper extends AbstractMinecartContainer implements Hopper
     }
 
     private void tryConsumeItems() {
-        if (!this.level().isClientSide && this.isAlive() && this.isEnabled() && !this.consumedItemThisFrame && this.suckInItems()) {
+        if (!this.level().isClientSide && this.leaf$isAlive && this.isEnabled() && !this.consumedItemThisFrame && this.suckInItems()) { // Leaf - cache entity isAlive
             this.consumedItemThisFrame = true;
             this.setChanged();
         }
diff --git a/net/minecraft/world/level/block/entity/BellBlockEntity.java b/net/minecraft/world/level/block/entity/BellBlockEntity.java
index ff63952c10d4642d5c14f470a18bea3066aa7ab0..9615d518d0c037f0e01fae1f07906fca60e08c4c 100644
--- a/net/minecraft/world/level/block/entity/BellBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BellBlockEntity.java
@@ -114,18 +114,18 @@ public class BellBlockEntity extends BlockEntity {
 
         if (!this.level.isClientSide) {
             for (LivingEntity livingEntity : this.nearbyEntities) {
-                if (livingEntity.isAlive() && !livingEntity.isRemoved() && blockPos.closerToCenterThan(livingEntity.position(), 32.0)) {
+                if (livingEntity.leaf$isAlive && !livingEntity.isRemoved() && blockPos.closerToCenterThan(livingEntity.position(), 32.0)) { // Leaf - cache entity isAlive
                     livingEntity.getBrain().setMemory(MemoryModuleType.HEARD_BELL_TIME, this.level.getGameTime());
                 }
             }
         }
 
-        this.nearbyEntities.removeIf(e -> !e.isAlive()); // Paper - Fix bell block entity memory leak
+        this.nearbyEntities.removeIf(e -> !e.leaf$isAlive); // Paper - Fix bell block entity memory leak // Leaf - cache entity isAlive
     }
 
     private static boolean areRaidersNearby(BlockPos pos, List<LivingEntity> raiders) {
         for (LivingEntity livingEntity : raiders) {
-            if (livingEntity.isAlive()
+            if (livingEntity.leaf$isAlive // Leaf - cache entity isAlive
                 && !livingEntity.isRemoved()
                 && pos.closerToCenterThan(livingEntity.position(), 32.0)
                 && livingEntity.getType().is(EntityTypeTags.RAIDERS)) {
@@ -167,7 +167,7 @@ public class BellBlockEntity extends BlockEntity {
     }
 
     private static boolean isRaiderWithinRange(BlockPos pos, LivingEntity raider) {
-        return raider.isAlive() && !raider.isRemoved() && pos.closerToCenterThan(raider.position(), 48.0) && raider.getType().is(EntityTypeTags.RAIDERS);
+        return raider.leaf$isAlive && !raider.isRemoved() && pos.closerToCenterThan(raider.position(), 48.0) && raider.getType().is(EntityTypeTags.RAIDERS); // Leaf - cache entity isAlive
     }
 
     @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - Add BellRevealRaiderEvent
