From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 22 Jan 2026 01:21:43 +0900
Subject: [PATCH] configurable ice and snow tick chance


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index f19fd8392e914d3cc7f92c173ca2192e18ec6762..8f179b23d038c6dec27fffacec8736151d7399b6 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1264,9 +1264,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         ProfilerFiller profiler = Profiler.get();
         profiler.push("iceandsnow");
 
+        final int rate = org.dreeam.leaf.config.modules.gameplay.IceAndSnowChance.iceAndSnowChance; // Leaf - configurable ice and snow tick chance
         if (!this.paperConfig().environment.disableIceAndSnow) { // Paper - Option to disable ice and snow
         for (int i = 0; i < tickSpeed; i++) {
-            if (simpleRandom.nextInt(48) == 0) { // Paper - optimise random ticking
+            if (simpleRandom.nextInt(rate) == 0) { // Paper - optimise random ticking // Leaf - configurable ice and snow tick chance
                 this.tickPrecipitation(this.getBlockRandomPos(minX, 0, minZ, 15));
             }
         }
