From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 30 Aug 2025 20:20:58 +0900
Subject: [PATCH] entity activation


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index a6c1f450f32fa688b102140963fd5bfb6b4e6ddf..236732210226b2d26a2cc449c12f73ee2a1efb3d 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -213,6 +213,17 @@ public final class ChunkEntitySlices {
         return collectedEntities;
     }
 
+    // Leaf start - entity activation
+    public void leaf$getAllEntities(final it.unimi.dsi.fastutil.objects.ObjectArrayList<Entity> into) {
+        final int len = this.entities.size();
+        if (len == 0) {
+            return;
+        }
+        final Entity[] rawData = this.entities.getRawData();
+        into.addElements(into.size(), rawData, 0, this.entities.size());
+    }
+    // Leaf end - entity activation
+
     public boolean isEmpty() {
         return this.entities.size() == 0;
     }
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
index 703bf9c2a56b262e2719a1787584de537b8f12e0..70bf88abe062338764ff908f185e8ce69a093887 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -616,6 +616,48 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         }
     }
 
+    // Leaf start - entity activation
+    public final void leaf$getEntities(final AABB box, final it.unimi.dsi.fastutil.longs.LongOpenHashSet chunks, final it.unimi.dsi.fastutil.objects.ObjectArrayList<Entity> into) {
+        final int minChunkX = (Mth.floor(box.minX) - 2) >> 4;
+        final int minChunkZ = (Mth.floor(box.minZ) - 2) >> 4;
+        final int maxChunkX = (Mth.floor(box.maxX) + 2) >> 4;
+        final int maxChunkZ = (Mth.floor(box.maxZ) + 2) >> 4;
+
+        final int minRegionX = minChunkX >> REGION_SHIFT;
+        final int minRegionZ = minChunkZ >> REGION_SHIFT;
+        final int maxRegionX = maxChunkX >> REGION_SHIFT;
+        final int maxRegionZ = maxChunkZ >> REGION_SHIFT;
+
+        for (int currRegionZ = minRegionZ; currRegionZ <= maxRegionZ; ++currRegionZ) {
+            final int minZ = currRegionZ == minRegionZ ? minChunkZ & REGION_MASK : 0;
+            final int maxZ = currRegionZ == maxRegionZ ? maxChunkZ & REGION_MASK : REGION_MASK;
+
+            for (int currRegionX = minRegionX; currRegionX <= maxRegionX; ++currRegionX) {
+                final ChunkSlicesRegion region = this.getRegion(currRegionX, currRegionZ);
+
+                if (region == null) {
+                    continue;
+                }
+
+                final int minX = currRegionX == minRegionX ? minChunkX & REGION_MASK : 0;
+                final int maxX = currRegionX == maxRegionX ? maxChunkX & REGION_MASK : REGION_MASK;
+
+                for (int currZ = minZ; currZ <= maxZ; ++currZ) {
+                    for (int currX = minX; currX <= maxX; ++currX) {
+                        final ChunkEntitySlices chunk = region.get(currX | (currZ << REGION_SHIFT));
+                        if (chunk == null || !chunk.status.isOrAfter(FullChunkStatus.FULL)) {
+                            continue;
+                        }
+                        if (chunks.add(CoordinateUtils.getChunkKey((currRegionX << REGION_SHIFT) | currX, (currRegionZ << REGION_SHIFT) | currZ))) {
+                            chunk.leaf$getAllEntities(into);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // Leaf end - entity activation
+
     public void getHardCollidingEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate) {
         final int minChunkX = (Mth.floor(box.minX) - 2) >> 4;
         final int minChunkZ = (Mth.floor(box.minZ) - 2) >> 4;
diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index 726bca1e9a3102ed13e6d11e011a8cceca24b65e..e4715f1a7c48d07d31e15362a7d575d72fc72f79 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -45,7 +45,7 @@ public final class ActivationRange {
     private ActivationRange() {
     }
 
-    static Activity[] VILLAGER_PANIC_IMMUNITIES = {
+    static final Activity[] VILLAGER_PANIC_IMMUNITIES = { // Leaf - entity activation - add final
         Activity.HIDE,
         Activity.PRE_RAID,
         Activity.RAID,
@@ -86,7 +86,7 @@ public final class ActivationRange {
     private static int getWakeUpDurationWithVariance(Entity entity, int wakeUpDuration) {
         double deviation = entity.level().galeConfig().gameplayMechanics.entityWakeUpDurationRatioStandardDeviation;
 
-        if (deviation <= 0) {
+        if (deviation <= 0.0) { // Leaf - entity activation
             return wakeUpDuration;
         }
 
@@ -94,8 +94,18 @@ public final class ActivationRange {
     }
     // Gale end - variable entity wake-up duration
 
+    // Leaf start - entity activation
+    @Deprecated
     static AABB maxBB = new AABB(0, 0, 0, 0, 0, 0);
 
+    private static final it.unimi.dsi.fastutil.objects.ObjectArrayList<Entity> entityListCache = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+    private static final it.unimi.dsi.fastutil.longs.LongOpenHashSet chunks = new it.unimi.dsi.fastutil.longs.LongOpenHashSet();
+    private static final org.dreeam.leaf.util.KDTree2D kdTree2 = new org.dreeam.leaf.util.KDTree2D();
+    private static final org.dreeam.leaf.util.KDTree3D kdTree3 = new org.dreeam.leaf.util.KDTree3D();
+    private static final Player[] EMPTY_PLAYERS = {};
+    private static final ActivationType[] ACTIVATION_TYPES = ActivationType.values();
+    // Leaf end - entity activation
+
     /**
      * These entities are excluded from Activation range checks.
      *
@@ -135,6 +145,12 @@ public final class ActivationRange {
      * @param world
      */
     public static synchronized void activateEntities(final Level world) { // Leaf
+        // Leaf start - entity activation
+        if (true) {
+            activateEntities1(world);
+            return;
+        }
+        // Leaf end - entity activation
         final int miscActivationRange = world.spigotConfig.miscActivationRange;
         final int raiderActivationRange = world.spigotConfig.raiderActivationRange;
         final int animalActivationRange = world.spigotConfig.animalActivationRange;
@@ -201,11 +217,144 @@ public final class ActivationRange {
         }
     }
 
+    // Leaf start - entity activation
+    private static void activateEntities1(final Level world) {
+        final int miscActivationRange = world.spigotConfig.miscActivationRange;
+        final int raiderActivationRange = world.spigotConfig.raiderActivationRange;
+        final int animalActivationRange = world.spigotConfig.animalActivationRange;
+        final int monsterActivationRange = world.spigotConfig.monsterActivationRange;
+        final int waterActivationRange = world.spigotConfig.waterActivationRange;
+        final int flyingActivationRange = world.spigotConfig.flyingMonsterActivationRange;
+        final int villagerActivationRange = world.spigotConfig.villagerActivationRange;
+        world.wakeupInactiveRemainingAnimals = Math.min(world.wakeupInactiveRemainingAnimals + 1, world.spigotConfig.wakeUpInactiveAnimals);
+        world.wakeupInactiveRemainingVillagers = Math.min(world.wakeupInactiveRemainingVillagers + 1, world.spigotConfig.wakeUpInactiveVillagers);
+        world.wakeupInactiveRemainingMonsters = Math.min(world.wakeupInactiveRemainingMonsters + 1, world.spigotConfig.wakeUpInactiveMonsters);
+        world.wakeupInactiveRemainingFlying = Math.min(world.wakeupInactiveRemainingFlying + 1, world.spigotConfig.wakeUpInactiveFlying);
+
+        int maxRange = Math.max(monsterActivationRange, animalActivationRange);
+        maxRange = Math.max(maxRange, raiderActivationRange);
+        maxRange = Math.max(maxRange, miscActivationRange);
+        maxRange = Math.max(maxRange, flyingActivationRange);
+        maxRange = Math.max(maxRange, waterActivationRange);
+        maxRange = Math.max(maxRange, villagerActivationRange);
+        maxRange = Math.min((world.spigotConfig.simulationDistance << 4) - 8, maxRange);
+
+        final double[] ranges = new double[ACTIVATION_TYPES.length];
+        ranges[ActivationType.WATER.ordinal()] = waterActivationRange;
+        ranges[ActivationType.FLYING_MONSTER.ordinal()] = flyingActivationRange;
+        ranges[ActivationType.VILLAGER.ordinal()] = villagerActivationRange;
+        ranges[ActivationType.MONSTER.ordinal()] = monsterActivationRange;
+        ranges[ActivationType.ANIMAL.ordinal()] = animalActivationRange;
+        ranges[ActivationType.RAIDER.ordinal()] = raiderActivationRange;
+        ranges[ActivationType.MISC.ordinal()] = miscActivationRange;
+        for (int i = 0; i < ranges.length; i++) {
+            if (ranges[i] > 0.0) {
+                ranges[i] = ranges[i] * ranges[i];
+            }
+        }
+        final long currentTick = MinecraftServer.currentTick; // cast to long
+        final it.unimi.dsi.fastutil.objects.ObjectArrayList<Entity> entities = entityListCache;
+        final it.unimi.dsi.fastutil.longs.LongOpenHashSet chunks = ActivationRange.chunks;
+        final ca.spottedleaf.moonrise.patches.chunk_system.level.entity.EntityLookup lookup = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel) world).moonrise$getEntityLookup();
+        final org.dreeam.leaf.util.KDTree2D kdTree2 = ActivationRange.kdTree2;
+        final org.dreeam.leaf.util.KDTree3D kdTree3 = ActivationRange.kdTree3;
+        final boolean tickMarkers = world.paperConfig().entities.markers.tick;
+        int j = 0;
+        final Player[] array = world.players().toArray(EMPTY_PLAYERS);
+        final double[] pxl = new double[array.length];
+        final double[] pyl = new double[array.length];
+        final double[] pzl = new double[array.length];
+        for (int i = 0; i < array.length; i++) {
+            final Player p = array[i];
+            p.activatedTick = currentTick;
+            if (world.spigotConfig.ignoreSpectatorActivation && p.isSpectator()) {
+                continue;
+            }
+            if (!world.purpurConfig.idleTimeoutTickNearbyEntities && p.isAfk()) {
+                continue; // Purpur - AFK API
+            }
+            array[j] = p;
+            pxl[j] = p.getX();
+            pyl[j] = p.getY();
+            pzl[j] = p.getZ();
+            j++;
+        }
+        final int[] indices = new int[j];
+        for (int k = 0; k < j; k++) {
+            indices[k] = k;
+        }
+        kdTree2.build(new double[][]{pxl, pzl}, indices);
+        for (int k = 0; k < j; k++) {
+            indices[k] = k;
+        }
+        kdTree3.build(new double[][]{pxl, pzl, pyl}, indices);
+        final double worldHeight = world.getHeight();
+        for (int i = 0; i < j; i++) {
+            final Player player = array[i];
+            final AABB box = player.getBoundingBox().inflate(maxRange, worldHeight, maxRange);
+            lookup.leaf$getEntities(box, chunks, entities);
+            ca.spottedleaf.moonrise.common.PlatformHooks.get().addToGetEntities(world, null, box, null, entities);
+        }
+        final Object[] raw = entities.elements();
+        final int size = entities.size();
+        final int minY = world.getMinY();
+        final int maxY = world.getMaxY();
+        for (int i = 0; i < size; i++) {
+            final Entity entity = (Entity) raw[i];
+            if (!tickMarkers && entity instanceof Marker) {
+                continue;
+            }
+            if (currentTick <= entity.activatedTick) {
+                continue;
+            }
+            // todo
+            /*
+            final int blockY = entity.getBlockY();
+            if (blockY < minY || blockY > maxY) {
+                continue;
+            }
+            */
+            final net.minecraft.world.phys.Vec3 p = entity.position;
+            if (entity.defaultActivationState) { // Pufferfish - diff on change
+                entity.activatedTick = currentTick;
+            } else {
+                // todo: check this works as intended
+                // bounding box -> position
+                final double max = ranges[entity.activationType.ordinal()];
+                // kdTree not empty here
+                final double near = kdTree2.nearestSqr(p.x, p.z, max);
+                if (near != max) { // Pufferfish - diff on change
+                    entity.activatedTick = currentTick;
+                }
+            }
+            // Pufferfish start
+            if (org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.enabled
+                && entity.getType().dabEnabled
+                && (!org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.dontEnableIfInWater || !entity.isInWater() || (entity instanceof net.minecraft.world.entity.animal.WaterAnimal || (entity instanceof final LivingEntity livingEntity && livingEntity.canBreatheUnderwater())))) { // Leaf - Option for dontEnableIfInWater
+                if (!entity.activatedPriorityReset) {
+                    entity.activatedPriorityReset = true;
+                    entity.activatedPriority = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.maximumActivationPrio;
+                }
+                final int squaredDistance = (int) kdTree3.nearestSqr(p.x, p.z, p.y, 16384.0);
+                entity.activatedPriority = squaredDistance > org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.startDistanceSquared ?
+                    Math.max(1, Math.min(squaredDistance >> org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.activationDistanceMod, entity.activatedPriority)) :
+                    1;
+            } else {
+                entity.activatedPriority = 1;
+            }
+            // Pufferfish end
+        }
+        entities.clear();
+        chunks.clear();
+    }
+    // Leaf end - entity activation
+
     /**
      * Tries to activate an entity.
      *
      * @param entity
      */
+    @Deprecated // Leaf - entity activation
     private static void activateEntity(final Entity entity) {
         if (MinecraftServer.currentTick > entity.activatedTick) {
             if (entity.defaultActivationState) { // Pufferfish - diff on change
@@ -355,13 +504,18 @@ public final class ActivationRange {
         return isActive;
     }
 
+    // Leaf start - entity activation
+    /*
     private static Set<EntityType<?>> ENTITIES_THAT_FLY = Set.of(
         EntityType.GHAST,
         EntityType.HAPPY_GHAST,
         EntityType.PHANTOM
     );
+    */
 
     private static boolean isEntityThatFlies(final Entity entity) {
-        return ENTITIES_THAT_FLY.contains(entity.getType());
+        EntityType<?> t = entity.getType();
+        return t == EntityType.GHAST || t == EntityType.HAPPY_GHAST || t == EntityType.PHANTOM;
     }
+    // Leaf end - entity activation
 }
