From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Thu, 16 Apr 2026 16:46:39 +0800
Subject: [PATCH] Skip item merge checks for full stacks


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index e3b854f572b1b7797d86facccd422da3a5feba3c..574887782776bb42ce3a9d6d43dcf04cc2b14ed4 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -273,8 +273,11 @@ public class ItemEntity extends Entity implements TraceableEntity {
                         }
                     }
                     // Paper end - Fix items merging through walls
-                    this.tryToMerge(entity);
-                    if (this.isRemoved()) {
+                    // Leaf start - Skip item merge checks for full stacks
+                    final ItemStack item = this.getItem();
+                    this.tryToMerge(entity, item, entity.getItem());
+                    if (this.isRemoved() || item.getCount() >= item.getMaxStackSize()) {
+                        // Leaf end - Skip item merge checks for full stacks
                         break;
                     }
                 }
@@ -290,6 +293,11 @@ public class ItemEntity extends Entity implements TraceableEntity {
     private void tryToMerge(final ItemEntity other) {
         ItemStack thisItemStack = this.getItem();
         ItemStack otherItemStack = other.getItem();
+        // Leaf start - Skip item merge checks for full stacks
+        tryToMerge(other, thisItemStack, otherItemStack);
+    }
+    private void tryToMerge(final ItemEntity other, final ItemStack thisItemStack, final ItemStack otherItemStack) {
+        // Leaf end - Skip item merge checks for full stacks
         if (Objects.equals(this.target, other.target) && areMergable(thisItemStack, otherItemStack)) {
             if (org.dreeam.leaf.config.modules.gameplay.UseSpigotItemMergingMech.enabled || otherItemStack.getCount() < thisItemStack.getCount()) { // Leaf - KeYi - Configurable spigot item merging mechanism
                 merge(this, thisItemStack, other, otherItemStack);
