From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Don't load POI for competitor scan


diff --git a/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java b/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
index 9b595322b21b83d38886d515a3b8dd04cde43cea..7e002bd744376b2d59a8dd0a4f58eabf2f6b1a75 100644
--- a/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
+++ b/net/minecraft/world/entity/ai/behavior/PoiCompetitorScan.java
@@ -20,7 +20,7 @@ public class PoiCompetitorScan {
                     (jobSite, nearestLivingEntities) -> (level, villager, gameTime) -> {
                         GlobalPos globalPos = instance.get(jobSite);
                         level.getPoiManager()
-                            .getType(globalPos.pos())
+                            .getTypeIfLoaded(globalPos.pos()) // Leaf - Don't load POI for competitor scan
                             .ifPresent(
                                 // Paper start - Improve performance of PoiCompetitorScan by unrolling stream
                                 // The previous logic used Stream#reduce to simulate a form of single-iteration bubble sort
diff --git a/net/minecraft/world/entity/ai/village/poi/PoiManager.java b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
index 95fa0d9487542ba5308c60d29095efa6c90e86d6..cb4329d42a7e260a8bdaae2bc9b3b6003424089b 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -342,6 +342,12 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         return this.getOrLoad(SectionPos.asLong(pos)).flatMap(poiSection -> poiSection.getType(pos));
     }
 
+    // Leaf start - Don't load POI for competitor scan
+    public Optional<Holder<PoiType>> getTypeIfLoaded(BlockPos pos) {
+        return this.get(SectionPos.asLong(pos)).flatMap(poiSection -> poiSection.getType(pos));
+    }
+    // Leaf end - Don't load POI for competitor scan
+
     @VisibleForDebug
     public @Nullable DebugPoiInfo getDebugPoiInfo(BlockPos pos) {
         return this.getOrLoad(SectionPos.asLong(pos)).flatMap(poiSection -> poiSection.getDebugPoiInfo(pos)).orElse(null);
