From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 19 Mar 2026 20:11:58 +0900
Subject: [PATCH] reduce player enchantment tick allocations


diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 05d9626aec8b66f5b02e875ad60b853d15c0e02a..d469337972028fa75a5779c3225ea43334630784 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -349,8 +349,14 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
     }
 
     public void tick(ServerLevel level, int enchantmentLevel, EnchantedItemInUse item, Entity entity) {
+        // Leaf start - reduce player enchantment tick allocations
+        List<ConditionalEffect<EnchantmentEntityEffect>> effects1 = this.getEffects(EnchantmentEffectComponents.TICK);
+        if (effects1.isEmpty()) {
+            return;
+        }
         applyEffects(
-            this.getEffects(EnchantmentEffectComponents.TICK),
+            effects1,
+        // Leaf end - reduce player enchantment tick allocations
             entityContext(level, enchantmentLevel, entity, entity.position()),
             enchantmentEntityEffect -> enchantmentEntityEffect.apply(level, enchantmentLevel, item, entity, entity.position())
         );
