From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 23 Jan 2026 18:47:34 +0900
Subject: [PATCH] fix skeleton horse trap NPE

Prevent removing the goal during the goals iteration.

diff --git a/net/minecraft/world/entity/animal/equine/SkeletonHorse.java b/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
index ed25701549c7453be446b29b9f08d2093388e679..0154b7d4ef53af7a912051647b088a6f5b368eeb 100644
--- a/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
+++ b/net/minecraft/world/entity/animal/equine/SkeletonHorse.java
@@ -204,8 +204,10 @@ public class SkeletonHorse extends AbstractHorse {
             this.isTrap = trap;
             if (trap) {
                 this.goalSelector.addGoal(1, this.skeletonTrapGoal);
+                this.skeletonTrapGoal.doTrap = true; // Leaf - fix skeleton horse trap NPE
             } else {
-                this.goalSelector.removeGoal(this.skeletonTrapGoal);
+                //this.goalSelector.removeGoal(this.skeletonTrapGoal); // Leaf - fix skeleton horse trap NPE
+                this.skeletonTrapGoal.doTrap = false; // Leaf - fix skeleton horse trap NPE
             }
         }
     }
diff --git a/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java b/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
index 4b95b4b4b3c8990b5b9f446781d5b78c2f48b802..a1aa40a2af59eab982a74cfafce94d585ee7976d 100644
--- a/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
+++ b/net/minecraft/world/entity/animal/equine/SkeletonTrapGoal.java
@@ -19,6 +19,7 @@ import org.jspecify.annotations.Nullable;
 public class SkeletonTrapGoal extends Goal {
     private final SkeletonHorse horse;
     private java.util.List<org.bukkit.entity.HumanEntity> eligiblePlayers; // Paper
+    public boolean doTrap = true; // Leaf - fix skeleton horse trap NPE
 
     public SkeletonTrapGoal(final SkeletonHorse horse) {
         this.horse = horse;
@@ -26,7 +27,7 @@ public class SkeletonTrapGoal extends Goal {
 
     @Override
     public boolean canUse() {
-        return !(this.eligiblePlayers = this.horse.level().findNearbyBukkitPlayers(this.horse.getX(), this.horse.getY(), this.horse.getZ(), 10.0, net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING)).isEmpty(); // Paper - Affects Spawning API & SkeletonHorseTrapEvent
+        return doTrap && !(this.eligiblePlayers = this.horse.level().findNearbyBukkitPlayers(this.horse.getX(), this.horse.getY(), this.horse.getZ(), 10.0, net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING)).isEmpty(); // Paper - Affects Spawning API & SkeletonHorseTrapEvent // Leaf - fix skeleton horse trap NPE
     }
 
     @Override
@@ -102,4 +103,11 @@ public class SkeletonTrapGoal extends Goal {
         );
         skeleton.setItemSlot(slot, stack);
     }
+
+    // Leaf start - fix eligiblePlayers leak
+    @Override
+    public void stop() {
+        eligiblePlayers = null;
+    }
+    // Leaf end - fix eligiblePlayers leak
 }
