From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 30 Oct 2025 10:01:57 +0900
Subject: [PATCH] resend velocity on dimension change


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 083d0086d1afdee5961401bcb10ad6e7811dc12f..ea64e6fb54f5b3be175f0983343521f87351ab4c 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1503,6 +1503,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     this.stopUsingItem();
                 }
                 // Paper end - Reset shield blocking on dimension change
+                this.markHurt(); // Leaf - resend velocity on dimension change
                 return this;
             }
         }
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 178085c71bb5fee0818a8ce88bf1e36a107a4dd1..5ec0a1bc3fc0b79efd1eb99259d42ee5f0a2a08c 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1024,6 +1024,7 @@ public abstract class PlayerList {
         // Paper end - Add PlayerPostRespawnEvent
 
         // CraftBukkit end
+        serverPlayer.markHurt(); // Leaf - resend velocity on dimension change
 
         return serverPlayer;
     }
