From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 21 Mar 2026 15:25:23 +0900
Subject: [PATCH] optimize PathNavigation#followThePath

To prevent unnecessary BlockPos allocation in `asBlockPos` under `getNextNodePos`

diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 6893ac65206ffa2357aa5f725227d5be4b3364a3..0663cb18c289f16ea76a9c1790ef4e46e5ef4de7 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -320,12 +320,14 @@ public abstract class PathNavigation {
         if (!this.path.isProcessed()) return; // Kaiiju - petal - async path processing - skip if not processed
         Vec3 tempMobPos = this.getTempMobPos();
         this.maxDistanceToWaypoint = this.mob.getBbWidth() > 0.75F ? this.mob.getBbWidth() / 2.0F : 0.75F - this.mob.getBbWidth() / 2.0F;
-        Vec3i nextNodePos = this.path.getNextNodePos();
-        double abs = Math.abs(this.mob.getX() - (nextNodePos.getX() + 0.5));
-        double abs1 = Math.abs(this.mob.getY() - nextNodePos.getY());
-        double abs2 = Math.abs(this.mob.getZ() - (nextNodePos.getZ() + 0.5));
+        // Leaf start - optimize PathNavigation#followThePath
+        Node nextNode = this.path.getNextNode();
+        double abs = Math.abs(this.mob.getX() - (nextNode.x + 0.5));
+        double abs1 = Math.abs(this.mob.getY() - nextNode.y);
+        double abs2 = Math.abs(this.mob.getZ() - (nextNode.z + 0.5));
+        // Leaf end - optimize PathNavigation#followThePath
         boolean flag = abs < this.maxDistanceToWaypoint && abs2 < this.maxDistanceToWaypoint && abs1 < 1.0;
-        if (flag || this.canCutCorner(this.path.getNextNode().type) && this.shouldTargetNextNodeInDirection(tempMobPos)) {
+        if (flag || this.canCutCorner(nextNode.type) && this.shouldTargetNextNodeInDirection(tempMobPos)) { // Leaf - optimize PathNavigation#followThePath
             this.path.advance();
         }
 
