From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 19 Mar 2026 20:11:58 +0900
Subject: [PATCH] reduce player enchantment tick allocations

To reduce allocation of creating entity loot contexts

diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 05d9626aec8b66f5b02e875ad60b853d15c0e02a..94285f13be555d92ce273cd70b2bf10840232bae 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -349,8 +349,14 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
     }
 
     public void tick(ServerLevel level, int enchantmentLevel, EnchantedItemInUse item, Entity entity) {
+        // Leaf start - reduce player enchantment tick allocations
+        List<ConditionalEffect<EnchantmentEntityEffect>> tickEffects = this.getEffects(EnchantmentEffectComponents.TICK);
+        if (tickEffects.isEmpty()) {
+            return;
+        }
         applyEffects(
-            this.getEffects(EnchantmentEffectComponents.TICK),
+            tickEffects,
+        // Leaf end - reduce player enchantment tick allocations
             entityContext(level, enchantmentLevel, entity, entity.position()),
             enchantmentEntityEffect -> enchantmentEntityEffect.apply(level, enchantmentLevel, item, entity, entity.position())
         );
