From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LeavesMC <leaves@leavesmc.org>
Date: Mon, 25 Nov 2025 00:00:00 +0000
Subject: [PATCH] Optimize Redstone Wire


diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index 0000000000000000000000000000000000000000..0000000000000000000000000000000000000000 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.level.block;
 
+import org.leavesmc.leaves.redstone.RedstoneWireTurbo; // Leaves - Optimize Redstone Wire
 import com.google.common.collect.Sets;
 import java.util.Map;
 import java.util.Set;
@@ -360,21 +361,8 @@ public class RedStoneWireBlock extends Block {
     }
 
     private void updatePowerStrength(Level level, BlockPos pos, BlockState state) {
-        int i = this.calculateTargetStrength(level, pos);
-        if (state.getValue(POWER) != i) {
-            if (level.getBlockState(pos) == state) {
-                level.setBlock(pos, (BlockState) state.setValue(POWER, i), 2);
-            }
-
-            Set<BlockPos> set = Sets.newHashSet();
-            set.add(pos);
-
-            for(Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
-                set.add(pos.relative(direction));
-            }
-
-            for(BlockPos blockPos : set) {
-                level.updateNeighborsAt(blockPos, this);
-            }
-        }
+        // Leaves start - Optimize Redstone Wire
+        RedstoneWireTurbo.updatePowerStrength(this, level, pos, state);
+        // Leaves end - Optimize Redstone Wire
     }
 
     private int calculateTargetStrength(Level level, BlockPos pos) {
@@ -383,6 +371,10 @@ public class RedStoneWireBlock extends Block {
             this.shouldSignalTL.set(false);
             int i = this.getBlockSignal(level, pos);
             this.shouldSignalTL.set(true);
+            return i;
+        }
+        // Leaf end - SparklyPaper - parallel world ticking
+        return this.getBlockSignal(level, pos);
+    }
+
+    private int getBlockSignal(Level level, BlockPos pos) {