From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Rewrite entity despawn time


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f6d619709d4e5b0e6d1b943226579d7388835cdb..8ace428b7eda64508accffe80120148424eaecf0 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -373,6 +373,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public boolean fixedPose = false; // Paper - Expand Pose API
     private final int despawnTime; // Paper - entity despawn time limit
     public int totalEntityAge; // Paper - age-like counter for all entities
+    private int lastTickTicks = -1; // Leaf - Rewrite entity despawn time
     public boolean activatedPriorityReset = false; // Pufferfish - DAB
     public int activatedPriority = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.maximumActivationPrio; // Pufferfish - DAB (golf score)
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
@@ -384,6 +385,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
 
     public void inactiveTick() {
+        this.lastTickTicks = net.minecraft.server.MinecraftServer.currentTick; // Leaf - Rewrite entity despawn time
     }
     // Paper end - EAR 2
     // CraftBukkit end
@@ -887,15 +889,43 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public void tick() {
+        // Leaf start - Rewrite entity despawn time
+        if (this.detectDespawnTime()) {
+            return;
+        }
+        /*
         // Paper start - entity despawn time limit
         if (this.despawnTime >= 0 && this.totalEntityAge >= this.despawnTime) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
             return;
         }
         // Paper end - entity despawn time limit
+         */
+        this.lastTickTicks = net.minecraft.server.MinecraftServer.currentTick;
+        // Leaf end - Rewrite entity despawn time
         this.baseTick();
     }
 
+    // Leaf start - Rewrite entity despawn time
+    private boolean detectDespawnTime() {
+        if (this.despawnTime >= 0) {
+            int tickInterval = net.minecraft.server.MinecraftServer.currentTick - this.lastTickTicks;
+            if (this.lastTickTicks != -1 && tickInterval > 1) {
+                if (this.totalEntityAge + tickInterval >= this.despawnTime) {
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                    return true;
+                }
+            } else {
+                if (this.totalEntityAge >= this.despawnTime) {
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    // Leaf end - Rewrite entity despawn time
+
     // CraftBukkit start
     public void postTick() {
         // No clean way to break out of ticking once the entity has been copied to a new world, so instead we move the portalling later in the tick cycle
