From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 24 Apr 2026 02:41:08 -0400
Subject: [PATCH] Replace entity equipment items to array


diff --git a/net/minecraft/world/entity/EntityEquipment.java b/net/minecraft/world/entity/EntityEquipment.java
index 3ed39d9df87bcd066c30c6fa1c52a591a9ce4bc6..9795d70ed8276ea76f225effe7b9910b80458b0a 100644
--- a/net/minecraft/world/entity/EntityEquipment.java
+++ b/net/minecraft/world/entity/EntityEquipment.java
@@ -13,11 +13,19 @@ public class EntityEquipment implements net.caffeinemc.mods.lithium.common.entit
         map.putAll((Map<? extends EquipmentSlot, ? extends ItemStack>)items);
         return new EntityEquipment(map);
     }, equipment -> {
-        Map<EquipmentSlot, ItemStack> items = new EnumMap<>(equipment.items);
-        items.values().removeIf(ItemStack::isEmpty);
+        // Leaf start - Replace entity equipment items to array
+        Map<EquipmentSlot, ItemStack> items = new EnumMap<>(EquipmentSlot.class);
+        final ItemStack[] stacks = equipment.items;
+        for (int i = 0; i < stacks.length; i++) {
+            ItemStack stack = stacks[i];
+            if (!stack.isEmpty()) {
+                items.put(EquipmentSlot.VALUES_ARRAY[i], stack);
+            }
+        }
+        // Leaf end - Replace entity equipment items to array
         return items;
     });
-    private final EnumMap<EquipmentSlot, ItemStack> items;
+    private final ItemStack[] items; // Leaf - Replace entity equipment items to array
     // Leaf start - Lithium - equipment tracking
     boolean shouldTickEnchantments = false;
     ItemStack recheckEnchantmentForStack = null;
@@ -26,7 +34,16 @@ public class EntityEquipment implements net.caffeinemc.mods.lithium.common.entit
     // Leaf end - Lithium - equipment tracking
 
     private EntityEquipment(final EnumMap<EquipmentSlot, ItemStack> items) {
-        this.items = items;
+        // Leaf start - Replace entity equipment items to array
+        ItemStack[] self = new ItemStack[EquipmentSlot.VALUES_ARRAY.length];
+        java.util.Arrays.fill(self, ItemStack.EMPTY);
+        if (!items.isEmpty()) {
+            for (Entry<EquipmentSlot, ItemStack> e : items.entrySet()) {
+                self[e.getKey().ordinal()] = e.getValue();
+            }
+        }
+        this.items = self;
+        // Leaf end - Replace entity equipment items to array
     }
 
     public EntityEquipment() {
@@ -35,18 +52,23 @@ public class EntityEquipment implements net.caffeinemc.mods.lithium.common.entit
 
     public ItemStack set(final EquipmentSlot slot, final ItemStack itemStack) {
         // Leaf start - Lithium - equipment tracking
-        ItemStack oldStack = Objects.requireNonNullElse(this.items.put(slot, itemStack), ItemStack.EMPTY);
+        // Leaf start - Replace entity equipment items to array
+        final ItemStack[] items = this.items;
+        final int slotIndex = slot.ordinal();
+        ItemStack oldStack = items[slotIndex];
+        items[slotIndex] = itemStack == null ? ItemStack.EMPTY : itemStack;
+        // Leaf end - Replace entity equipment items to array
         if (inLevel) this.onEquipmentReplaced(oldStack, itemStack);
         return oldStack;
         // Leaf end - Lithium - equipment tracking
     }
 
     public ItemStack get(final EquipmentSlot slot) {
-        return this.items.getOrDefault(slot, ItemStack.EMPTY);
+        return this.items[slot.ordinal()]; // Leaf - Replace entity equipment items to array
     }
 
     public boolean isEmpty() {
-        for (ItemStack item : this.items.values()) {
+        for (ItemStack item : this.items) { // Leaf - Replace entity equipment items to array
             if (!item.isEmpty()) {
                 return false;
             }
@@ -56,23 +78,25 @@ public class EntityEquipment implements net.caffeinemc.mods.lithium.common.entit
     }
 
     public void tick(final Entity owner) {
-        for (Entry<EquipmentSlot, ItemStack> entry : this.items.entrySet()) {
-            ItemStack item = entry.getValue();
+        // Leaf start - Replace entity equipment items to array
+        final ItemStack[] items = this.items;
+        for (int i = 0; i < items.length; i++) {
+            ItemStack item = items[i];
             if (!item.isEmpty()) {
-                item.inventoryTick(owner.level(), owner, entry.getKey());
+                item.inventoryTick(owner.level(), owner, EquipmentSlot.VALUES_ARRAY[i]);
             }
         }
+        // Leaf end - Replace entity equipment items to array
     }
 
     public void setAll(final EntityEquipment equipment) {
         if (this.inLevel) this.invalidateData(); // Leaf - Lithium - equipment tracking
-        this.items.clear();
-        this.items.putAll(equipment.items);
+        System.arraycopy(equipment.items, 0, this.items, 0, this.items.length); // Leaf - Replace entity equipment items to array
         if (this.inLevel) this.initializeData(); // Leaf - Lithium - equipment tracking
     }
 
     public void dropAll(final LivingEntity dropper) {
-        for (ItemStack item : this.items.values()) {
+        for (ItemStack item : this.items) { // Leaf - Replace entity equipment items to array
             dropper.drop(item, true, false);
         }
 
@@ -80,14 +104,14 @@ public class EntityEquipment implements net.caffeinemc.mods.lithium.common.entit
     }
 
     public void clear() {
-        this.items.replaceAll((s, v) -> ItemStack.EMPTY);
+        java.util.Arrays.fill(this.items, ItemStack.EMPTY); // Leaf - Replace entity equipment items to array
         if (inLevel) this.invalidateData(); // Leaf - Lithium - equipment tracking
     }
 
     // Paper start - EntityDeathEvent
     // Needed to not set ItemStack.EMPTY to not existent slot.
     public boolean has(final EquipmentSlot slot) {
-        return this.items.containsKey(slot);
+        return this.items[slot.ordinal()] != ItemStack.EMPTY; // Leaf - Replace entity equipment items to array
     }
     // Paper end - EntityDeathEvent
 
@@ -210,7 +234,7 @@ public class EntityEquipment implements net.caffeinemc.mods.lithium.common.entit
         this.recheckEnchantmentForStack = null;
         this.hasUnsentEquipmentChanges = true;
 
-        for (ItemStack oldStack : this.items.values()) {
+        for (ItemStack oldStack : this.items) { // Leaf - Replace entity equipment items to array
             if (!oldStack.isEmpty()) {
                 //noinspection unchecked
                 oldStack.lithium$unsubscribeWithData(this, 0);
@@ -223,7 +247,7 @@ public class EntityEquipment implements net.caffeinemc.mods.lithium.common.entit
         this.recheckEnchantmentForStack = null;
         this.hasUnsentEquipmentChanges = true;
 
-        for (ItemStack newStack : this.items.values()) {
+        for (ItemStack newStack : this.items) { // Leaf - Replace entity equipment items to array
             if (!newStack.isEmpty()) {
                 if (!this.shouldTickEnchantments) {
                     this.shouldTickEnchantments = stackHasTickableEnchantment(newStack);
