From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Github Actions <artyom030803@gmail.com>
Date: Fri, 27 Mar 2026 06:47:13 +0300
Subject: [PATCH] Do not process VehicleEntityCollisionEvent if no listeners


diff --git a/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java
index 1ce8a8513e2fb27254dcfe920b1ea8babfb1cd3c..b2bb3d0a378bb8100e0fc24f4e6d4852d846df44 100644
--- a/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java
@@ -10,9 +10,11 @@ import net.minecraft.world.phys.Vec3;
 
 public abstract class MinecartBehavior {
     protected final AbstractMinecart minecart;
+    protected final boolean hasVehicleEntityCollisionEvent;
 
     protected MinecartBehavior(AbstractMinecart minecart) {
         this.minecart = minecart;
+        this.hasVehicleEntityCollisionEvent = org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0; // Leaf - Do not process VehicleEntityCollisionEvent if no listeners
     }
 
     public InterpolationHandler getInterpolation() {
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 4b336bb2690244b5c99bb7dd566e4fd84e031e79..253eeb94044b3e1616fa59e60fec119df6d3306b 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -532,13 +532,15 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
-                        // CraftBukkit start
-                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                            entity.getBukkitEntity()
-                        );
-                        if (!collisionEvent.callEvent()) continue;
-                        // CraftBukkit end
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (hasVehicleEntityCollisionEvent) {
+                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                            );
+                            if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
+                        }
+                        // Leaf end
                         boolean flag = entity.startRiding(this.minecart);
                         if (flag) {
                             return true;
@@ -564,13 +566,15 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         || entity.isPassenger()) {
                         // CraftBukkit start
                         if (!this.minecart.isPassengerOfSameVehicle(entity)) {
-                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                                entity.getBukkitEntity()
-                            );
-                            if (!collisionEvent.callEvent()) {
-                                continue;
+                            // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                            if (hasVehicleEntityCollisionEvent) {
+                                org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                    (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                                );
+                                if (!collisionEvent.callEvent()) continue;
+                                // CraftBukkit end
                             }
+                            // Leaf end
                         }
                         // CraftBukkit end
                         entity.push(this.minecart);
@@ -581,15 +585,15 @@ public class NewMinecartBehavior extends MinecartBehavior {
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, box)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
-                    // CraftBukkit start
-                    org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                        (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                        entity1.getBukkitEntity()
-                    );
-                    if (!collisionEvent.callEvent()) {
-                        continue;
+                    // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                    if (hasVehicleEntityCollisionEvent) {
+                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
+                        );
+                        if (!collisionEvent.callEvent()) continue;
+                        // CraftBukkit end
                     }
-                    // CraftBukkit end
+                    // Leaf end
                     entity1.push(this.minecart);
                     flag = true;
                 }
diff --git a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
index 7ab2542ddd7494585f240657906ccee80cd7944e..eda17f4b11c85e5ae918aa6d05b9b19b63cb9347 100644
--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -377,21 +377,26 @@ public class OldMinecartBehavior extends MinecartBehavior {
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
                         // CraftBukkit start
-                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
-                        );
-                        if (!collisionEvent.callEvent()) continue;
-                        // CraftBukkit end
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (hasVehicleEntityCollisionEvent) {
+                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                            );
+                            if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
+                        }
+                        // Leaf end
                         entity.startRiding(this.minecart);
                     } else {
-                        // CraftBukkit start
-                        if (!this.minecart.isPassengerOfSameVehicle(entity)) {
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (hasVehicleEntityCollisionEvent) {
                             org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                                 (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
                             );
                             if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
                         }
-                        // CraftBukkit end
+                        // Leaf end
                         entity.push(this.minecart);
                     }
                 }
@@ -399,12 +404,15 @@ public class OldMinecartBehavior extends MinecartBehavior {
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, aabb)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
-                    // CraftBukkit start
-                    org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                        (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
-                    );
-                    if (!collisionEvent.callEvent()) continue;
-                    // CraftBukkit end
+                    // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                    if (hasVehicleEntityCollisionEvent) {
+                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
+                        );
+                        if (!collisionEvent.callEvent()) continue;
+                        // CraftBukkit end
+                    }
+                    // Leaf end
                     entity1.push(this.minecart);
                 }
             }
