From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Github Actions <artyom030803@gmail.com>
Date: Fri, 27 Mar 2026 06:47:13 +0300
Subject: [PATCH] Do not process VehicleEntityCollisionEvent if no listeners


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 98313e336f1e10d11d6e58bba80c42038c66aed9..fbcc6080dbe83665533620a5319a92956a201147 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -251,6 +251,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     private int id = ENTITY_COUNTER.incrementAndGet();
     public boolean blocksBuilding;
     public ImmutableList<Entity> passengers = ImmutableList.of();
+    public static final ImmutableList<Entity> EMPTY_PASSENGERS = ImmutableList.of(); // Leaf - optimize empty check for passengers
     protected int boardingCooldown;
     private @Nullable Entity vehicle;
     private Level level;
@@ -477,7 +478,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     @Override
     public final boolean moonrise$hasAnyPlayerPassengers() {
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == Entity.EMPTY_PASSENGERS) { // Leaf - optimize empty check for passengers
             return false;
         }
         return this.getIndirectPassengersStream().anyMatch((entity) -> entity instanceof Player);
@@ -3511,7 +3512,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         if (passenger.getVehicle() != this) {
             throw new IllegalStateException("Use x.startRiding(y), not y.addPassenger(x)");
         } else {
-            if (this.passengers.isEmpty()) {
+            if (this.passengers == Entity.EMPTY_PASSENGERS) { // Leaf - optimize empty check for passengers
                 this.passengers = ImmutableList.of(passenger);
             } else {
                 List<Entity> list = Lists.newArrayList(this.passengers);
@@ -3590,7 +3591,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     protected boolean canAddPassenger(Entity passenger) {
-        return this.passengers.isEmpty();
+        return this.passengers == Entity.EMPTY_PASSENGERS; // Leaf - optimize empty check for passengers
     }
 
     protected boolean couldAcceptPassenger() {
@@ -3778,7 +3779,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public boolean isVehicle() {
-        return !this.passengers.isEmpty();
+        return this.passengers != Entity.EMPTY_PASSENGERS; // Leaf - optimize empty check for passengers
     }
 
     public boolean dismountsUnderwater() {
@@ -4788,11 +4789,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public @Nullable Entity getFirstPassenger() {
-        return this.passengers.isEmpty() ? null : this.passengers.get(0);
+        return this.passengers == Entity.EMPTY_PASSENGERS ? null : this.passengers.get(0); // Leaf - optimize empty check for passengers
     }
 
     public boolean hasPassenger(Entity entity) {
-        return !passengers.isEmpty() && this.passengers.contains(entity); // Leaf - Optimize map lookups with isEmpty check
+        return passengers != Entity.EMPTY_PASSENGERS && this.passengers.contains(entity); // Leaf - Optimize passengers lookups with emptiness check
     }
 
     public boolean hasPassenger(Predicate<Entity> predicate) {
@@ -4806,19 +4807,19 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     private Stream<Entity> getIndirectPassengersStream() {
-        if (this.passengers.isEmpty()) { return Stream.of(); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == Entity.EMPTY_PASSENGERS) { return Stream.of(); } // Paper - Optimize indirect passenger iteration // Leaf - optimize empty check for passengers
         return this.passengers.stream().flatMap(Entity::getSelfAndPassengers);
     }
 
     @Override
     public Stream<Entity> getSelfAndPassengers() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == Entity.EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - optimize empty check for passengers
         return Stream.concat(Stream.of(this), this.getIndirectPassengersStream());
     }
 
     @Override
     public Stream<Entity> getPassengersAndSelf() {
-        if (this.passengers.isEmpty()) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == Entity.EMPTY_PASSENGERS) { return Stream.of(this); } // Paper - Optimize indirect passenger iteration // Leaf - optimize empty check for passengers
         return Stream.concat(this.passengers.stream().flatMap(Entity::getPassengersAndSelf), Stream.of(this));
     }
 
@@ -4826,7 +4827,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         // Paper start - optimise entity tracker
         final List<Entity> ret = new ArrayList<>();
 
-        if (this.passengers.isEmpty()) {
+        if (this.passengers == Entity.EMPTY_PASSENGERS) { // Leaf - optimize empty check for passengers
             return ret;
         }
 
@@ -4841,7 +4842,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public boolean hasExactlyOnePlayerPassenger() {
-        if (this.passengers.isEmpty()) { return false; } // Paper - Optimize indirect passenger iteration
+        if (this.passengers == Entity.EMPTY_PASSENGERS) { return false; } // Paper - Optimize indirect passenger iteration // Leaf - optimize empty check for passengers
         return this.countPlayerPassengers() == 1;
     }
 
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 10045bdd3c9692fb0a56f140b4891c67cf64eeea..cfb1b055f4f12b65b62dae47cf06256ed26913c8 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1722,7 +1722,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         if (!player.getItemInHand(hand).isEmpty()) {
             return InteractionResult.PASS;
         }
-        if (!passengers.isEmpty() || player.isPassenger()) {
+        if (passengers != Entity.EMPTY_PASSENGERS || player.isPassenger()) { // Leaf - optimize empty check for passengers
             return InteractionResult.PASS;
         }
         if (this instanceof TamableAnimal tamable) {
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index 48099ae66045b6e78ec52832e3b972ac9c9bc246..20c0aab8ebcca0890e40cc310a260259b33c118c 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -111,6 +111,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
     // Purpur end - Minecart settings and WASD controls
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
     // CraftBukkit end
+    protected final boolean hasVehicleEntityCollisionEvent;
 
     protected AbstractMinecart(EntityType<?> type, Level level) {
         super(type, level);
@@ -125,6 +126,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
             maxSpeed = storedMaxSpeed = level.purpurConfig.minecartMaxSpeed;
             // Purpur end - Minecart settings and WASD controls
         }
+        this.hasVehicleEntityCollisionEvent = org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0; // Leaf - Do not process VehicleEntityCollisionEvent if no listeners
     }
 
     protected AbstractMinecart(EntityType<?> type, Level level, double x, double y, double z) {
@@ -180,8 +182,14 @@ public abstract class AbstractMinecart extends VehicleEntity {
             return false;
         }
 
-        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent((org.bukkit.entity.Vehicle) this.getBukkitEntity(), entity.getBukkitEntity());
-        return collisionEvent.callEvent();
+        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+        if (hasVehicleEntityCollisionEvent) {
+            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent((org.bukkit.entity.Vehicle) this.getBukkitEntity(), entity.getBukkitEntity());
+            return collisionEvent.callEvent();
+        } else {
+            return true;
+        }
+        // Leaf end
         // Paper end - fix VehicleEntityCollisionEvent not called when colliding with player
     }
 
@@ -434,7 +442,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
         this.setDeltaMovement(Mth.clamp(deltaMovement.x, -maxSpeed, maxSpeed), deltaMovement.y, Mth.clamp(deltaMovement.z, -maxSpeed, maxSpeed));
 
         // Purpur start - Minecart settings and WASD controls
-        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && !passengers.isEmpty()) {
+        if (level().purpurConfig.minecartControllable && !isInWater() && !isInLava() && passengers != Entity.EMPTY_PASSENGERS) { // Leaf - optimize empty check for passengers
             Entity passenger = passengers.get(0);
             if (passenger instanceof net.minecraft.server.level.ServerPlayer player) {
                 net.minecraft.world.entity.player.Input lastClientInput = player.getLastClientInput();
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 4b336bb2690244b5c99bb7dd566e4fd84e031e79..aa760d2c73b97568213c426d614746b597c6ed3b 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -532,13 +532,15 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
-                        // CraftBukkit start
-                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                            entity.getBukkitEntity()
-                        );
-                        if (!collisionEvent.callEvent()) continue;
-                        // CraftBukkit end
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (this.minecart.hasVehicleEntityCollisionEvent) {
+                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                            );
+                            if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
+                        }
+                        // Leaf end
                         boolean flag = entity.startRiding(this.minecart);
                         if (flag) {
                             return true;
@@ -564,13 +566,15 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         || entity.isPassenger()) {
                         // CraftBukkit start
                         if (!this.minecart.isPassengerOfSameVehicle(entity)) {
-                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                                entity.getBukkitEntity()
-                            );
-                            if (!collisionEvent.callEvent()) {
-                                continue;
+                            // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                            if (this.minecart.hasVehicleEntityCollisionEvent) {
+                                org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                    (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                                );
+                                if (!collisionEvent.callEvent()) continue;
+                                // CraftBukkit end
                             }
+                            // Leaf end
                         }
                         // CraftBukkit end
                         entity.push(this.minecart);
@@ -581,15 +585,15 @@ public class NewMinecartBehavior extends MinecartBehavior {
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, box)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
-                    // CraftBukkit start
-                    org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                        (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
-                        entity1.getBukkitEntity()
-                    );
-                    if (!collisionEvent.callEvent()) {
-                        continue;
+                    // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                    if (this.minecart.hasVehicleEntityCollisionEvent) {
+                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
+                        );
+                        if (!collisionEvent.callEvent()) continue;
+                        // CraftBukkit end
                     }
-                    // CraftBukkit end
+                    // Leaf end
                     entity1.push(this.minecart);
                     flag = true;
                 }
diff --git a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
index 7ab2542ddd7494585f240657906ccee80cd7944e..fa9cf4518bf6f5a2f53846f5896c0b273c23d376 100644
--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -377,21 +377,26 @@ public class OldMinecartBehavior extends MinecartBehavior {
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
                         // CraftBukkit start
-                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
-                        );
-                        if (!collisionEvent.callEvent()) continue;
-                        // CraftBukkit end
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (this.minecart.hasVehicleEntityCollisionEvent) {
+                            org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                                (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
+                            );
+                            if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
+                        }
+                        // Leaf end
                         entity.startRiding(this.minecart);
                     } else {
-                        // CraftBukkit start
-                        if (!this.minecart.isPassengerOfSameVehicle(entity)) {
+                        // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                        if (this.minecart.hasVehicleEntityCollisionEvent) {
                             org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                                 (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
                             );
                             if (!collisionEvent.callEvent()) continue;
+                            // CraftBukkit end
                         }
-                        // CraftBukkit end
+                        // Leaf end
                         entity.push(this.minecart);
                     }
                 }
@@ -399,12 +404,15 @@ public class OldMinecartBehavior extends MinecartBehavior {
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, aabb)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
-                    // CraftBukkit start
-                    org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
-                        (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
-                    );
-                    if (!collisionEvent.callEvent()) continue;
-                    // CraftBukkit end
+                    // Leaf start - Do not process VehicleEntityCollisionEvent if no listeners
+                    if (this.minecart.hasVehicleEntityCollisionEvent) {
+                        org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                            (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
+                        );
+                        if (!collisionEvent.callEvent()) continue;
+                        // CraftBukkit end
+                    }
+                    // Leaf end
                     entity1.push(this.minecart);
                 }
             }
