From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Reduce debug subscribers overhead


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 0ad8c7f45580129477e6cfe960c8ea28875c7192..b651f781dc3cd6c7fc0abd76971218e679ae9772 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1991,7 +1991,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         profiler.popPush("players");
         this.playerList.tick();
         profiler.popPush("debugSubscribers");
-        this.debugSubscribers.tick();
+        if (!org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) this.debugSubscribers.tick(); // Leaf - Reduce debug subscribers overhead
         if (this.tickRateManager.runsNormally()) {
             profiler.popPush("gameTests");
             GameTestTicker.SINGLETON.tick();
diff --git a/net/minecraft/util/debug/ServerDebugSubscribers.java b/net/minecraft/util/debug/ServerDebugSubscribers.java
index ffbc6e6b88336e692c94586de8181a8c33b002f3..79019d6e8eb4ae53a394313d763670305bb914f8 100644
--- a/net/minecraft/util/debug/ServerDebugSubscribers.java
+++ b/net/minecraft/util/debug/ServerDebugSubscribers.java
@@ -13,17 +13,24 @@ import net.minecraft.server.players.NameAndId;
 
 public class ServerDebugSubscribers {
     private final MinecraftServer server;
-    private final Map<DebugSubscription<?>, List<ServerPlayer>> enabledSubscriptions = new HashMap<>();
+    // Leaf start - Reduce debug subscribers overhead
+    private final Map<DebugSubscription<?>, List<ServerPlayer>> enabledSubscriptions = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<>();
+    private int tickCount;
+    private static final Set<DebugSubscription<?>> EMPTY_SUBSCRIPTIONS = Set.of();
+    private static final List<ServerPlayer> EMPTY_PLAYERS = List.of();
+    // Leaf end - Reduce debug subscribers overhead
 
     public ServerDebugSubscribers(final MinecraftServer server) {
         this.server = server;
     }
 
     private List<ServerPlayer> getSubscribersFor(final DebugSubscription<?> subscription) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return EMPTY_PLAYERS; // Leaf - Reduce debug subscribers overhead
         return this.enabledSubscriptions.getOrDefault(subscription, List.of());
     }
 
     public void tick() {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE || tickCount++ % 20 != 0) return; // Leaf - Reduce debug subscribers overhead
         this.enabledSubscriptions.values().forEach(List::clear);
 
         for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
@@ -36,20 +43,24 @@ public class ServerDebugSubscribers {
     }
 
     public void broadcastToAll(final DebugSubscription<?> subscription, final Packet<?> packet) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return; // Leaf - Reduce debug subscribers overhead
         for (ServerPlayer player : this.getSubscribersFor(subscription)) {
             player.connection.send(packet);
         }
     }
 
     public Set<DebugSubscription<?>> enabledSubscriptions() {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return EMPTY_SUBSCRIPTIONS; // Leaf - Reduce debug subscribers overhead
         return Set.copyOf(this.enabledSubscriptions.keySet());
     }
 
     public boolean hasAnySubscriberFor(final DebugSubscription<?> subscription) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return false; // Leaf - Reduce debug subscribers overhead
         return !this.getSubscribersFor(subscription).isEmpty();
     }
 
     public boolean hasRequiredPermissions(final ServerPlayer player) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return false; // Leaf - Reduce debug subscribers overhead
         NameAndId nameAndId = player.nameAndId();
         return SharedConstants.IS_RUNNING_IN_IDE && this.server.isSingleplayerOwner(nameAndId) || this.server.getPlayerList().isOp(nameAndId);
     }
