From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Tue, 31 Mar 2026 09:30:12 +0300
Subject: [PATCH] Skip VehicleEntityCollisionEvent if no listeners


diff --git a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
index 33fb59123ce59b8fd5c9f84bc31e52d21ad060ff..a3dc24f96c0f52eede08c1310e974367c269b467 100644
--- a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
@@ -189,11 +189,13 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
             if (entity.getBoundingBox().minY < this.getBoundingBox().maxY) {
                 // CraftBukkit start
                 if (!this.isPassengerOfSameVehicle(entity)) {
+                    if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                         (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
                         entity.getBukkitEntity()
                     );
                     if (!event.callEvent()) return;
+                    } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                 }
                 // CraftBukkit end
                 super.push(entity);
@@ -201,11 +203,13 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
         } else if (entity.getBoundingBox().minY <= this.getBoundingBox().minY) {
             // CraftBukkit start
             if (!this.isPassengerOfSameVehicle(entity)) {
+                if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                 org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                     (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
                     entity.getBukkitEntity()
                 );
                 if (!event.callEvent()) return;
+                } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
             }
             // CraftBukkit end
             super.push(entity);
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index 48099ae66045b6e78ec52832e3b972ac9c9bc246..37d1555ecd6820b77e7e70463d7ee277a42caac7 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -180,9 +180,15 @@ public abstract class AbstractMinecart extends VehicleEntity {
             return false;
         }
 
+        // Leaf start - Skip VehicleEntityCollisionEvent if no listeners
+        if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) {
         org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent((org.bukkit.entity.Vehicle) this.getBukkitEntity(), entity.getBukkitEntity());
         return collisionEvent.callEvent();
         // Paper end - fix VehicleEntityCollisionEvent not called when colliding with player
+        } else {
+            return true;
+        }
+        // Leaf end - Skip VehicleEntityCollisionEvent if no listeners
     }
 
     @Override
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 4b336bb2690244b5c99bb7dd566e4fd84e031e79..13f13f44052ac7c84afdabf69012e4a5f81d603f 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -532,6 +532,7 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
+                        if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                         // CraftBukkit start
                         org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                             (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
@@ -539,6 +540,7 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         );
                         if (!collisionEvent.callEvent()) continue;
                         // CraftBukkit end
+                        } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                         boolean flag = entity.startRiding(this.minecart);
                         if (flag) {
                             return true;
@@ -564,6 +566,7 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         || entity.isPassenger()) {
                         // CraftBukkit start
                         if (!this.minecart.isPassengerOfSameVehicle(entity)) {
+                            if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                             org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                                 (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
                                 entity.getBukkitEntity()
@@ -571,6 +574,7 @@ public class NewMinecartBehavior extends MinecartBehavior {
                             if (!collisionEvent.callEvent()) {
                                 continue;
                             }
+                            } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                         }
                         // CraftBukkit end
                         entity.push(this.minecart);
@@ -581,6 +585,7 @@ public class NewMinecartBehavior extends MinecartBehavior {
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, box)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
+                    if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                     // CraftBukkit start
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                         (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(),
@@ -590,6 +595,7 @@ public class NewMinecartBehavior extends MinecartBehavior {
                         continue;
                     }
                     // CraftBukkit end
+                    } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                     entity1.push(this.minecart);
                     flag = true;
                 }
diff --git a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
index 7ab2542ddd7494585f240657906ccee80cd7944e..26049ea9cb8c55e0ee4738240c97d5e21360b5e9 100644
--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -376,20 +376,24 @@ public class OldMinecartBehavior extends MinecartBehavior {
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
+                        if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                         // CraftBukkit start
                         org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                             (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
                         );
                         if (!collisionEvent.callEvent()) continue;
                         // CraftBukkit end
+                        } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                         entity.startRiding(this.minecart);
                     } else {
                         // CraftBukkit start
                         if (!this.minecart.isPassengerOfSameVehicle(entity)) {
+                            if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                             org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                                 (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity()
                             );
                             if (!collisionEvent.callEvent()) continue;
+                            } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                         }
                         // CraftBukkit end
                         entity.push(this.minecart);
@@ -399,12 +403,14 @@ public class OldMinecartBehavior extends MinecartBehavior {
         } else {
             for (Entity entity1 : this.level().getEntities(this.minecart, aabb)) {
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
+                    if (org.bukkit.event.vehicle.VehicleEntityCollisionEvent.getHandlerList().getRegisteredListeners().length > 0) { // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                     // CraftBukkit start
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
                         (org.bukkit.entity.Vehicle) this.minecart.getBukkitEntity(), entity1.getBukkitEntity()
                     );
                     if (!collisionEvent.callEvent()) continue;
                     // CraftBukkit end
+                    } // Leaf - Skip VehicleEntityCollisionEvent if no listeners
                     entity1.push(this.minecart);
                 }
             }
