From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 21 May 2026 14:05:19 +0800
Subject: [PATCH] Pre authenticate event


diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index b279be372f33dd20c77110895723547c83266820..871e8e88c85eacf8ae5522abb04c04171aaac758 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -193,7 +193,10 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
         if (singleplayerProfile != null && this.requestedUsername.equalsIgnoreCase(singleplayerProfile.name())) {
             this.startClientVerification(singleplayerProfile);
         } else {
-            if (this.server.usesAuthentication() && !this.connection.isMemoryConnection()) {
+            // Leaf start - pre authenticate event
+            boolean shouldAuthenticate = new org.dreeam.leaf.event.AsyncPreAuthenticateEvent(requestedUsername, requestedUuid, connection.getRemoteAddress(), !server.usesAuthentication()).callEvent();
+            if (shouldAuthenticate && !this.connection.isMemoryConnection()) {
+                // Leaf end - pre authenticate event
                 this.state = ServerLoginPacketListenerImpl.State.KEY;
                 this.connection.send(new ClientboundHelloPacket("", this.server.getKeyPair().getPublic().getEncoded(), this.challenge, true));
             } else {
