From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Add read-only mode for Linear v2


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
index f950472f371af3581af82c5bcdd7a800191cf051..0ca348d43143d11ba6dd0e6df33b5ee90ad8c381 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
@@ -1705,7 +1705,7 @@ public final class NewChunkHolder {
 
     private boolean saveChunk(final ChunkAccess chunk, final boolean unloading, final Completable<CompoundTag>[] chunkSave) {
         // Leaf start - disable world data saving
-        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world)) {
+        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world) || org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) { // Leaf - Add read-only mode for Linear v2
             if (chunk.isUnsaved()) {
                 chunk.tryMarkSaved();
             }
@@ -1763,6 +1763,7 @@ public final class NewChunkHolder {
     private boolean lastEntitySaveNull;
     private boolean saveEntities(final ChunkEntitySlices entities, final boolean unloading, final Completable<CompoundTag>[] entitySave) {
         if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world)) return false; // Leaf - disable world data saving
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return false; // Leaf - Add read-only mode for Linear v2
         try {
             if (entities.isTransient()) {
                 if (!unloading) {
@@ -1820,7 +1821,7 @@ public final class NewChunkHolder {
     private boolean lastPoiSaveNull;
     private boolean savePOI(final PoiChunk poi, final boolean unloading, final Completable<CompoundTag>[] poiSave) {
         // Leaf start - disable world data saving
-        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world)) {
+        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world) || org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) { // Leaf - Add read-only mode for Linear v2
             if (poi.isDirty()) {
                 poi.setDirty(false);
             }
diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 5a8937740f7779f971c2a13ec580eed7bcc26395..98b48a8a6f6372114ee1f71b2c69bff97af4ef64 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -130,6 +130,7 @@ public class PlayerAdvancements {
 
     public void save() {
         if (org.spigotmc.SpigotConfig.disableAdvancementSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         JsonElement json = this.codec.encodeStart(JsonOps.INSTANCE, this.asData()).getOrThrow();
 
         try {
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 3c6bde03919d8580b444c5b1851b15936899bbe5..908adae851185b48b8dc4ba2e7efad02402fba04 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1758,6 +1758,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     // Paper start - Incremental chunk and player saving
     public void saveIncrementally(final boolean doFull) {
         if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this)) return; // Leaf - disable world data saving
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         if (doFull) {
             org.bukkit.Bukkit.getPluginManager().callEvent(new org.bukkit.event.world.WorldSaveEvent(this.getWorld()));
             this.saveLevelData(false);
@@ -1774,7 +1775,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         // Paper end - add close param
         ServerChunkCache chunkSource = this.getChunkSource();
         // Leaf start - disable world data saving
-        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this)) {
+        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this) || org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) { // Leaf - Add read-only mode for Linear v2
             if (close) {
                 try {
                     chunkSource.close(false);
diff --git a/net/minecraft/stats/ServerStatsCounter.java b/net/minecraft/stats/ServerStatsCounter.java
index 84a3d9c0073eb389fcb74cffa3c55c9cf79d2fd8..1df6efebeda39097d75590c751ec635962b6e626 100644
--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -90,6 +90,7 @@ public class ServerStatsCounter extends StatsCounter {
 
     public void save() {
         if (org.spigotmc.SpigotConfig.disableStatSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         try {
             FileUtil.createDirectoriesSafe(this.file.getParent());
 
@@ -104,6 +105,7 @@ public class ServerStatsCounter extends StatsCounter {
     @Override
     public void setValue(final Player player, final Stat<?> stat, final int count) {
         if (org.spigotmc.SpigotConfig.disableStatSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         if (stat.getType() == Stats.CUSTOM && stat.getValue() instanceof final net.minecraft.resources.Identifier key && org.spigotmc.SpigotConfig.forcedStats.get(key) != null) return; // Paper - disable saving forced stats
         super.setValue(player, stat, count);
         this.dirty.add(stat);
diff --git a/net/minecraft/world/level/storage/PlayerDataStorage.java b/net/minecraft/world/level/storage/PlayerDataStorage.java
index 30a19e85f9f2a480443e1eb27b613c53b9a245bc..68e63a5ee014d87c503858cc1a1ca63c9d4904c0 100644
--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -33,6 +33,7 @@ public class PlayerDataStorage {
 
     public void save(final Player player) {
         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         // Leaf start - Async playerdata saving
         CompoundTag dataToStore;
         try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(player.problemPath(), LOGGER)) {
