From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 28 Feb 2026 14:42:29 +0900
Subject: [PATCH] reduce computeSpeed Vec3 allocations


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 557e7ac7f6ca421b4fef6090dfa6270cc37ee06d..07503e07a9736b25272999c6f2e3ca42d1f7ccbd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1046,7 +1046,7 @@ public abstract class Entity
             this.lastKnownPosition = this.position();
         }
 
-        this.lastKnownSpeed = this.position().subtract(this.lastKnownPosition);
+        this.lastKnownSpeed = this.position().equals(this.lastKnownPosition) ? Vec3.ZERO : this.position().subtract(this.lastKnownPosition); // Leaf - reduce computeSpeed Vec3 allocations
         this.lastKnownPosition = this.position();
     }
 
