From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 28 Feb 2026 14:42:29 +0900
Subject: [PATCH] reduce computeSpeed Vec3 allocations


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1d2afab2205f6c670abedec69462174ee324cb9f..3abef66dc0558b5fede9362d01ca8a01068f9c77 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1047,7 +1047,7 @@ public abstract class Entity
             this.lastKnownPosition = this.position();
         }
 
-        this.lastKnownSpeed = this.position().subtract(this.lastKnownPosition);
+        this.lastKnownSpeed = this.position().equals(this.lastKnownPosition) ? Vec3.ZERO : this.position().subtract(this.lastKnownPosition); // Leaf - reduce computeSpeed Vec3 allocations
         this.lastKnownPosition = this.position();
     }
 
