From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 20 Jul 2026 03:35:39 +0800
Subject: [PATCH] Replace Lithium tracker list with array


diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 119d27783833d01e9aea31270305836eafe55ace..f32ae978bac5419e60395c503047d82733f58a07 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -868,21 +868,18 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     //Mod count used to avoid transfer attempts that are known to fail (no change since last attempt)
     private long insertStackListModCount, extractStackListModCount;
 
-    @Nullable
-    private List<net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionItemEntityMovementTracker> collectItemEntityTracker;
+    private net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionItemEntityMovementTracker @Nullable [] collectItemEntityTracker; // Leaf - Replace Lithium tracker list with array
     private boolean collectItemEntityTrackerWasEmpty;
     @Nullable
     private AABB collectItemEntityBox;
     private long collectItemEntityAttemptTime;
 
-    @Nullable
-    private List<net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker> extractInventoryEntityTracker;
+    private net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker @Nullable [] extractInventoryEntityTracker; // Leaf - Replace Lithium tracker list with array
     @Nullable
     private AABB extractInventoryEntityBox;
     private long extractInventoryEntityFailedSearchTime;
 
-    @Nullable
-    private List<net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker> insertInventoryEntityTracker;
+    private net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker @Nullable [] insertInventoryEntityTracker; // Leaf - Replace Lithium tracker list with array
     @Nullable
     private AABB insertInventoryEntityBox;
     private long insertInventoryEntityFailedSearchTime;
@@ -964,19 +961,19 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this.insertBlockInventory).listenForContentChangesOnce(this.insertStackList, this);
         }
         if (listenToInsertEntities) {
-            if (this.insertInventoryEntityTracker == null || this.insertInventoryEntityTracker.isEmpty()) {
+            if (this.insertInventoryEntityTracker == null || this.insertInventoryEntityTracker.length == 0) { // Leaf - Replace Lithium tracker list with array
                 return;
             }
             net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.listenToEntityMovementOnce(this, insertInventoryEntityTracker);
         }
         if (listenToExtractEntities) {
-            if (this.extractInventoryEntityTracker == null || this.extractInventoryEntityTracker.isEmpty()) {
+            if (this.extractInventoryEntityTracker == null || this.extractInventoryEntityTracker.length == 0) { // Leaf - Replace Lithium tracker list with array
                 return;
             }
             net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.listenToEntityMovementOnce(this, extractInventoryEntityTracker);
         }
         if (listenToItemEntities) {
-            if (this.collectItemEntityTracker == null || this.collectItemEntityTracker.isEmpty()) {
+            if (this.collectItemEntityTracker == null || this.collectItemEntityTracker.length == 0) { // Leaf - Replace Lithium tracker list with array
                 return;
             }
             net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.listenToEntityMovementOnce(this, collectItemEntityTracker);
@@ -1111,7 +1108,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private void invalidateInsertionData() {
         if (this.level instanceof net.minecraft.server.level.ServerLevel) {
             if (this.insertInventoryEntityTracker != null) {
-                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.insertInventoryEntityTracker);
+                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.insertInventoryEntityTracker); // Leaf - Replace Lithium tracker list with array
                 this.insertInventoryEntityTracker = null;
                 this.insertInventoryEntityBox = null;
                 this.insertInventoryEntityFailedSearchTime = 0L;
@@ -1139,13 +1136,13 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private void invalidateExtractionData() {
         if (this.level instanceof net.minecraft.server.level.ServerLevel) {
             if (this.extractInventoryEntityTracker != null) {
-                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.extractInventoryEntityTracker);
+                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.extractInventoryEntityTracker); // Leaf - Replace Lithium tracker list with array
                 this.extractInventoryEntityTracker = null;
                 this.extractInventoryEntityBox = null;
                 this.extractInventoryEntityFailedSearchTime = 0L;
             }
             if (this.collectItemEntityTracker != null) {
-                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.collectItemEntityTracker);
+                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.collectItemEntityTracker); // Leaf - Replace Lithium tracker list with array
                 this.collectItemEntityTracker = null;
                 this.collectItemEntityBox = null;
                 this.collectItemEntityTrackerWasEmpty = false;
@@ -1281,7 +1278,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         if (this.insertInventoryEntityTracker == null) {
             this.initInsertInventoryTracker(world);
         }
-        if (net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(this.insertInventoryEntityFailedSearchTime, this.insertInventoryEntityTracker)) {
+        if (net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(this.insertInventoryEntityFailedSearchTime, this.insertInventoryEntityTracker)) { // Leaf - Replace Lithium tracker list with array
             this.insertInventoryEntityFailedSearchTime = this.tickedGameTime;
             return null;
         }
@@ -1346,7 +1343,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         if (this.extractInventoryEntityTracker == null) {
             this.initExtractInventoryTracker(level);
         }
-        if (net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(this.extractInventoryEntityFailedSearchTime, this.extractInventoryEntityTracker)) {
+        if (net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(this.extractInventoryEntityFailedSearchTime, this.extractInventoryEntityTracker)) { // Leaf - Replace Lithium tracker list with array
             this.extractInventoryEntityFailedSearchTime = this.tickedGameTime;
             return null;
         }
@@ -1466,7 +1463,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         long modCount = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(hopperBlockEntity).getModCount();
 
         if ((hopperBlockEntity.collectItemEntityTrackerWasEmpty || hopperBlockEntity.myModCountAtLastItemCollect == modCount) &&
-            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(hopperBlockEntity.collectItemEntityAttemptTime, hopperBlockEntity.collectItemEntityTracker)) {
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(hopperBlockEntity.collectItemEntityAttemptTime, hopperBlockEntity.collectItemEntityTracker)) { // Leaf - Replace Lithium tracker list with array
             hopperBlockEntity.collectItemEntityAttemptTime = hopperBlockEntity.tickedGameTime;
             return java.util.Collections.emptyList();
         }
