From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 28 Feb 2026 14:42:29 +0900
Subject: [PATCH] reduce computeSpeed Vec3 allocations


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 32522db0ef77c07cb3a67bc8dab13db6a36bb8e2..2a96f21b8438330aeea75df63ee293ddaa73902a 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1046,7 +1046,7 @@ public abstract class Entity
             this.lastKnownPosition = this.position();
         }
 
-        this.lastKnownSpeed = this.position().subtract(this.lastKnownPosition);
+        this.lastKnownSpeed = this.position().equals(this.lastKnownPosition) ? Vec3.ZERO : this.position().subtract(this.lastKnownPosition); // Leaf - reduce computeSpeed Vec3 allocations
         this.lastKnownPosition = this.position();
     }
 
