From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Sun, 12 Apr 2026 23:52:06 +0300
Subject: [PATCH] remove duplicated purpur code


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index b686e9abacdd18d913d642a5911be0f6f8ce2001..ea8d350c94d36e213d91af75d4d3567172162e1a 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1868,18 +1868,6 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
             // Purpur end - Totems work in inventory
 
-            // Purpur start - Totems work in inventory
-            if (level().purpurConfig.totemOfUndyingWorksInInventory && this instanceof ServerPlayer player && (itemStack == null || itemStack.getItem() != Items.TOTEM_OF_UNDYING) && player.getBukkitEntity().hasPermission("purpur.inventory_totem")) {
-                for (ItemStack item : player.getInventory().getNonEquipmentItems()) {
-                    if (item.getItem() == Items.TOTEM_OF_UNDYING) {
-                        itemInHand = item;
-                        itemStack = item.copy();
-                        break;
-                    }
-                }
-            }
-            // Purpur end - Totems work in inventory
-
             final org.bukkit.inventory.EquipmentSlot handSlot = (hand != null) ? org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand) : null;
             final EntityResurrectEvent event = new EntityResurrectEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), handSlot);
             event.setCancelled(itemStack == null);
