From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sun, 10 Aug 2025 16:14:07 +0900
Subject: [PATCH] Replace entity tracker by fluid map


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2b9d5f3c759a6ce2b6176f088f9d756d0dfe2a56..69598b0c0d3ef096c6d9d3316972f414c9e19a0c 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -308,7 +308,7 @@ public abstract class Entity
     public final RandomSource random; // Paper - Share random for entities to make them more random // Purpur - Add toggle for RNG manipulation
     public int tickCount;
     private int remainingFireTicks;
-    private final EntityFluidInteraction fluidInteraction = new EntityFluidInteraction(FLUIDS_WITH_CURRENT);
+    private final EntityFluidInteraction fluidInteraction = new EntityFluidInteraction(); // Leaf - Replace entity tracker by fluid map
     protected boolean wasTouchingWater;
     protected boolean wasEyeInWater;
     private int invulnerableTime;
diff --git a/net/minecraft/world/entity/EntityFluidInteraction.java b/net/minecraft/world/entity/EntityFluidInteraction.java
index ed6aa849a7ba718f673026c74a454f993f791ee5..87efdfcebb3a622d92e809d894a14ba0c037b29d 100644
--- a/net/minecraft/world/entity/EntityFluidInteraction.java
+++ b/net/minecraft/world/entity/EntityFluidInteraction.java
@@ -26,7 +26,7 @@ import org.jspecify.annotations.Nullable;
 
 public class EntityFluidInteraction {
     private final List<EntityFluidInteraction.Tracker> fluidTrackers = new ArrayList<>();
-    private final Reference2ObjectMap<TagKey<Fluid>, EntityFluidInteraction.CurrentAccumulator> currentAccumulators = new Reference2ObjectArrayMap<>();
+    private final Reference2ObjectMap<TagKey<Fluid>, EntityFluidInteraction.CurrentAccumulator> currentAccumulators = new org.dreeam.leaf.util.map.TrackerByFluidMap(); // Leaf - Replace entity tracker by fluid map
 
     public EntityFluidInteraction(final Set<TagKey<Fluid>> fluidsWithCurrent) {
         for (TagKey<Fluid> fluid : fluidsWithCurrent) {
@@ -34,9 +34,13 @@ public class EntityFluidInteraction {
         }
     }
 
+    public EntityFluidInteraction() {
+        ((org.dreeam.leaf.util.map.TrackerByFluidMap) this.currentAccumulators).init();
+    }
+
     public boolean update(final Entity entity, final boolean ignoreCurrent) {
         this.fluidTrackers.removeIf(EntityFluidInteraction.Tracker::reset);
-        this.currentAccumulators.values().forEach(EntityFluidInteraction.CurrentAccumulator::reset);
+        ((org.dreeam.leaf.util.map.TrackerByFluidMap) this.currentAccumulators).reset(); // Leaf - Replace entity tracker by fluid map
         AABB box = entity.getFluidInteractionBox();
         if (box == null) {
             return false;
@@ -154,11 +158,13 @@ public class EntityFluidInteraction {
     }
 
     private EntityFluidInteraction.@Nullable CurrentAccumulator getCurrentAccumulatorFor(final Holder<Fluid> fluid) {
-        for (Entry<TagKey<Fluid>, EntityFluidInteraction.CurrentAccumulator> entry : Reference2ObjectMaps.fastIterable(this.currentAccumulators)) {
-            if (fluid.is(entry.getKey())) {
-                return entry.getValue();
-            }
+        // Leaf start - Replace entity tracker by fluid map
+        if (fluid.is(net.minecraft.tags.FluidTags.WATER)) {
+            return this.currentAccumulators.get(net.minecraft.tags.FluidTags.WATER);
+        } else if (fluid.is(net.minecraft.tags.FluidTags.LAVA)) {
+            return this.currentAccumulators.get(net.minecraft.tags.FluidTags.LAVA);
         }
+        // Leaf end - Replace entity tracker by fluid map
 
         return null;
     }
@@ -196,7 +202,7 @@ public class EntityFluidInteraction {
         return false;
     }
 
-    private static class CurrentAccumulator {
+    public static class CurrentAccumulator { // Leaf - Replace entity tracker by fluid map - private -> public
         private double height;
         private Vec3 accumulatedCurrent = Vec3.ZERO;
         private int currentCount;
