From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Wed, 29 Jul 2026 12:52:50 +0900
Subject: [PATCH] optimize ClientboundCommandsPacket#enumerateNodes


diff --git a/net/minecraft/network/protocol/game/ClientboundCommandsPacket.java b/net/minecraft/network/protocol/game/ClientboundCommandsPacket.java
index d8c5069c30d4453c2f71c22196610510ab315643..fc6ce6a8b3305405cd6a827d4e4e63b95b3122fd 100644
--- a/net/minecraft/network/protocol/game/ClientboundCommandsPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundCommandsPacket.java
@@ -85,14 +85,17 @@ public class ClientboundCommandsPacket implements Packet<ClientGamePacketListene
 
         CommandNode<S> node;
         while ((node = queue.poll()) != null) {
-            if (!nodeToId.containsKey(node)) {
-                int id = nodeToId.size();
-                nodeToId.put(node, id);
-                queue.addAll(node.getChildren());
-                if (node.getRedirect() != null) {
-                    queue.add(node.getRedirect());
+            // Leaf start - optimize ClientboundCommandsPacket#enumerateNodes
+            int id = nodeToId.size();
+            nodeToId.computeIfAbsent(node, node1 -> {
+                CommandNode<S> node2 = (CommandNode<S>) node1;
+                queue.addAll(node2.getChildren());
+                if (node2.getRedirect() != null) {
+                    queue.add(node2.getRedirect());
                 }
-            }
+                return id;
+            });
+            // Leaf end - optimize ClientboundCommandsPacket#enumerateNodes
         }
 
         return nodeToId;
