From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sun, 10 Aug 2025 16:14:07 +0900
Subject: [PATCH] Replace entity tracker by fluid map


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1b0e7f7eb2b55ba7e548aebc92f9ace799bd3586..1abf0ac4d615db75637fce237dec01d944c10315 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -305,7 +305,7 @@ public abstract class Entity
     protected final RandomSource random; // Paper - Share random for entities to make them more random  // Purpur - Add toggle for RNG manipulation
     public int tickCount;
     private int remainingFireTicks;
-    private final EntityFluidInteraction fluidInteraction = new EntityFluidInteraction(Set.of(FluidTags.WATER, FluidTags.LAVA));
+    private final EntityFluidInteraction fluidInteraction = new EntityFluidInteraction(); // Leaf - Replace entity tracker by fluid map
     protected boolean wasTouchingWater;
     protected boolean wasEyeInWater;
     public int invulnerableTime;
diff --git a/net/minecraft/world/entity/EntityFluidInteraction.java b/net/minecraft/world/entity/EntityFluidInteraction.java
index 817ec969b758f9902e9dc19d3be2d801740d207b..d896eac6d9641f87dcd5357bbeaaef535fac4712 100644
--- a/net/minecraft/world/entity/EntityFluidInteraction.java
+++ b/net/minecraft/world/entity/EntityFluidInteraction.java
@@ -21,7 +21,7 @@ import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
 public class EntityFluidInteraction {
-    private final Map<TagKey<Fluid>, EntityFluidInteraction.Tracker> trackerByFluid = new Reference2ObjectArrayMap<>();
+    private final Map<TagKey<Fluid>, EntityFluidInteraction.Tracker> trackerByFluid = new org.dreeam.leaf.util.map.TrackerByFluidMap(); // Leaf - Replace entity tracker by fluid map
 
     public EntityFluidInteraction(final Set<TagKey<Fluid>> fluids) {
         for (TagKey<Fluid> fluid : fluids) {
@@ -29,8 +29,12 @@ public class EntityFluidInteraction {
         }
     }
 
+    public EntityFluidInteraction() {
+        ((org.dreeam.leaf.util.map.TrackerByFluidMap) this.trackerByFluid).init();
+    }
+
     public void update(final Entity entity, final boolean ignoreCurrent) {
-        this.trackerByFluid.values().forEach(EntityFluidInteraction.Tracker::reset);
+        ((org.dreeam.leaf.util.map.TrackerByFluidMap) this.trackerByFluid).reset(); // Leaf - Replace entity tracker by fluid map
         AABB box = entity.getFluidInteractionBox();
         if (box != null) {
             int x0 = Mth.floor(box.minX);
@@ -124,12 +128,13 @@ public class EntityFluidInteraction {
     }
 
     private EntityFluidInteraction.@Nullable Tracker getTrackerFor(final Fluid fluid) {
-        for (Entry<TagKey<Fluid>, EntityFluidInteraction.Tracker> entry : this.trackerByFluid.entrySet()) {
-            TagKey<Fluid> tag = entry.getKey();
-            if (fluid.is(tag)) {
-                return entry.getValue();
-            }
+        // Leaf start - Replace entity tracker by fluid map
+        if (fluid.is(net.minecraft.tags.FluidTags.WATER)) {
+            return this.trackerByFluid.get(net.minecraft.tags.FluidTags.WATER);
+        } else if (fluid.is(net.minecraft.tags.FluidTags.LAVA)) {
+            return this.trackerByFluid.get(net.minecraft.tags.FluidTags.LAVA);
         }
+        // Leaf end - Replace entity tracker by fluid map
 
         return null;
     }
@@ -155,7 +160,7 @@ public class EntityFluidInteraction {
         return tracker != null && tracker.eyesInside;
     }
 
-    private static class Tracker {
+    public static class Tracker { // Leaf - Replace entity tracker by fluid map - private -> public
         private double height;
         private boolean eyesInside;
         private Vec3 accumulatedCurrent = Vec3.ZERO;
