From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrlingXD <90316914+wling-art@users.noreply.github.com>
Date: Sat, 8 Aug 2026 17:09:43 +0800
Subject: [PATCH] optimize PathNavigation#shouldRecomputePath


diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 0336742e0886eb06beacaa2d8b1975d77b383142..bf47b60064d5253748cc94798c3d6032d4af5bc3 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -487,8 +487,10 @@ public abstract class PathNavigation {
             return false;
         } else if (this.path != null && this.path.isProcessed() && !this.path.isDone() && this.path.getNodeCount() != 0) { // Kaiiju - petal - async path processing - Skip if not processed
             Node target = this.path.getEndNode();
-            Vec3 middlePos = new Vec3((target.x + this.mob.getX()) / 2.0, (target.y + this.mob.getY()) / 2.0, (target.z + this.mob.getZ()) / 2.0);
-            return pos.closerToCenterThan(middlePos, this.path.getNodeCount() - this.path.getNextNodeIndex());
+            // Leaf start - optimize PathNavigation#shouldRecomputePath - avoid Vec3 alloc on this hot path
+            return pos.distToCenterSqr((target.x + this.mob.getX()) / 2.0, (target.y + this.mob.getY()) / 2.0, (target.z + this.mob.getZ()) / 2.0)
+                < Mth.square(this.path.getNodeCount() - this.path.getNextNodeIndex());
+            // Leaf end - optimize PathNavigation#shouldRecomputePath
         } else {
             return false;
         }
