From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Avoid registry access lookup on every random tick


diff --git a/net/minecraft/world/level/block/SpreadingSnowyBlock.java b/net/minecraft/world/level/block/SpreadingSnowyBlock.java
index 0cd8ac43984acf79e4efed2214a47f8b31dcb09c..9baee897cc10f0e11002c2340c69b63c54bdd6ec 100644
--- a/net/minecraft/world/level/block/SpreadingSnowyBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyBlock.java
@@ -60,7 +60,7 @@ public abstract class SpreadingSnowyBlock extends SnowyBlock {
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (this instanceof GrassBlock && level.paperConfig().tickRates.grassSpread != 1 && (level.paperConfig().tickRates.grassSpread < 1 || (net.minecraft.server.MinecraftServer.currentTick + pos.hashCode()) % level.paperConfig().tickRates.grassSpread != 0)) { return; } // Paper - Configurable random tick rates for blocks
-        Registry<Block> blocks = level.registryAccess().lookupOrThrow(Registries.BLOCK);
+        Registry<Block> blocks = net.minecraft.core.registries.BuiltInRegistries.BLOCK; // Leaf - avoid registry access lookup on every random tick
         Optional<Block> baseBlock = blocks.getOptional(this.baseBlock);
         if (!baseBlock.isEmpty()) {
             // Paper start - Perf: optimize dirt and snow spreading
