From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 22 Jan 2026 01:21:43 +0900
Subject: [PATCH] configurable ice and snow tick chance


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 090432a62540174dca6f46bbec5e27c6611a074c..375a7b31cbd319f79cc31a3c046b7e85a5e18dcb 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1231,9 +1231,10 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("iceandsnow");
 
+        final int rate = org.dreeam.leaf.config.modules.gameplay.IceAndSnowChance.iceAndSnowChance; // Leaf - configurable ice and snow tick chance
         if (!this.paperConfig().environment.disableIceAndSnow) { // Paper - Option to disable ice and snow
         for (int i = 0; i < randomTickSpeed; i++) {
-            if (simpleRandom.nextInt(48) == 0) {  // Paper - optimise random ticking
+            if (simpleRandom.nextInt(rate) == 0) {  // Paper - optimise random ticking // Leaf - configurable ice and snow tick chance
                 this.tickPrecipitation(this.getBlockRandomPos(minBlockX, 0, minBlockZ, 15));
             }
         }
