From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Thu, 16 Apr 2026 16:46:39 +0800
Subject: [PATCH] Skip item merge checks for full stacks


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index c1e7c04b048099dab7bb4cabdcb53add2201247f..abcb9fb10887aea6356fe39ab45f0ebc483988f0 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -295,7 +295,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
                     }
                     // Paper end - Fix items merging through walls
                     this.tryToMerge(itemEntity);
-                    if (this.isRemoved()) {
+                    if (this.isRemoved() || this.getItem().getCount() >= this.getItem().getMaxStackSize()) { // Leaf - Skip item merge checks for full stacks
                         break;
                     }
                 }
