From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Replace ServerClockManager map with optimized collection


diff --git a/net/minecraft/world/clock/ServerClockManager.java b/net/minecraft/world/clock/ServerClockManager.java
index 2cadb4fca189d6060a4eabba6099a52f84f50b1d..3c296009ccb6c93a4e36d41dc5da990bc6da6bfe 100644
--- a/net/minecraft/world/clock/ServerClockManager.java
+++ b/net/minecraft/world/clock/ServerClockManager.java
@@ -31,7 +31,7 @@ public class ServerClockManager extends SavedData implements ClockManager {
     private final PackedClockStates packedClockStates;
     private MinecraftServer server;
     private @org.jspecify.annotations.Nullable ServerLevel level; // Paper - per-world time
-    private final Map<Holder<WorldClock>, ServerClockManager.ClockInstance> clocks = new HashMap<>();
+    private final Map<Holder<WorldClock>, ServerClockManager.ClockInstance> clocks = new it.unimi.dsi.fastutil.objects.Object2ReferenceLinkedOpenHashMap<>(); // Leaf - Replace ServerClockManager map with optimized collection
 
     private ServerClockManager(final PackedClockStates packedClockStates) {
         this.packedClockStates = packedClockStates;
