From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Skip firing GenericGameEvent if no listeners


diff --git a/net/minecraft/world/level/gameevent/GameEventDispatcher.java b/net/minecraft/world/level/gameevent/GameEventDispatcher.java
index 1bdb36266e718bce07eae8355538a8c7db86c3c8..007453a29656a50dc067cb61d63e2ad6d22a155a 100644
--- a/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -22,6 +22,7 @@ public class GameEventDispatcher {
     public void post(final Holder<GameEvent> gameEvent, final Vec3 position, final GameEvent.Context context) {
         int radius = gameEvent.value().notificationRadius();
         BlockPos center = BlockPos.containing(position);
+        if (org.bukkit.event.world.GenericGameEvent.getHandlerList().getRegisteredListeners().length != 0) { // Leaf - Skip firing GenericGameEvent if no listeners
         // CraftBukkit start
         org.bukkit.event.world.GenericGameEvent apiEvent = new org.bukkit.event.world.GenericGameEvent(
             org.bukkit.craftbukkit.CraftGameEvent.minecraftHolderToBukkit(gameEvent), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(center, this.level),
@@ -31,6 +32,7 @@ public class GameEventDispatcher {
         }
         radius = apiEvent.getRadius();
         // CraftBukkit end
+        } // Leaf - Skip firing GenericGameEvent if no listeners
         int sectionMinX = SectionPos.blockToSectionCoord(center.getX() - radius);
         int sectionMinY = SectionPos.blockToSectionCoord(center.getY() - radius);
         int sectionMinZ = SectionPos.blockToSectionCoord(center.getZ() - radius);
