From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Thu, 16 Apr 2026 16:46:39 +0800
Subject: [PATCH] Skip item merge checks for full stacks


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index f3be1e846bb3ee03d1dfc891823418ebd1de4887..d27940451ad2147a24f81f26318918506838bd95 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -262,8 +262,11 @@ public class ItemEntity extends Entity implements TraceableEntity {
                         }
                     }
                     // Paper end - Fix items merging through walls
-                    this.tryToMerge(itemEntity);
-                    if (this.isRemoved()) {
+                    // Leaf start - Skip item merge checks for full stacks
+                    final ItemStack item = this.getItem();
+                    this.tryToMerge(itemEntity, item, itemEntity.getItem());
+                    if (this.isRemoved() || item.getCount() >= item.getMaxStackSize()) {
+                        // Leaf end - Skip item merge checks for full stacks
                         break;
                     }
                 }
@@ -279,6 +282,11 @@ public class ItemEntity extends Entity implements TraceableEntity {
     private void tryToMerge(ItemEntity itemEntity) {
         ItemStack item = this.getItem();
         ItemStack item1 = itemEntity.getItem();
+        // Leaf start - Skip item merge checks for full stacks
+        tryToMerge(itemEntity, item, item1);
+    }
+    private void tryToMerge(ItemEntity itemEntity, ItemStack item, ItemStack item1) {
+        // Leaf end - Skip item merge checks for full stacks
         if (Objects.equals(this.target, itemEntity.target) && areMergable(item, item1)) {
             if (org.dreeam.leaf.config.modules.gameplay.UseSpigotItemMergingMech.enabled || item1.getCount() < item.getCount()) { // Leaf - KeYi - Configurable spigot item merging mechanism
                 merge(this, item, itemEntity, item1);
