From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 12 Oct 2022 10:42:15 -0400
Subject: [PATCH] Leaf config changes


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
index e1b2447e3a8655b79deaef0163159e9681acb9a7..75c53c02dec32d8bc02e5a841e53319cd6df2d86 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V4290.java
@@ -254,7 +254,7 @@ public final class V4290 {
                         return ret;
                     }
                 } catch (final JsonParseException ex) {
-                    if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.invalidLegacyTextComponent) LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex); // Leaf - Do not log invalid flatten text component parse
+                    if (org.dreeam.leaf.config.modules.misc.global.LogToConsole.invalidLegacyTextComponent) LOGGER.error("Failed to convert json to nbt: " + unparsedJson, ex); // Leaf - Do not log invalid flatten text component parse
                 }
 
                 return null;
diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index 4d5c38c0fb6d18026f84be5eccf87d78608decaa..c05103092c25b82e72cfe990969836cc32a66265 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -83,7 +83,7 @@ public final class ActivationRange {
 
     // Gale start - variable entity wake-up duration
     private static int getWakeUpDurationWithVariance(Entity entity, int wakeUpDuration) {
-        final double deviation = entity.level().galeConfig().gameplayMechanics.entityWakeUpDurationRatioStandardDeviation;
+        final double deviation = entity.level().leafConfig().entityWakeUpDuration.ratioStandardDeviation;
         final org.dreeam.leaf.util.math.random.FasterRandomSource wakeUpDurationRandom = org.dreeam.leaf.util.math.random.FasterRandomSource.SHARED_INSTANCE;
 
         if (deviation <= 0) {
diff --git a/net/minecraft/network/chat/SignedMessageChain.java b/net/minecraft/network/chat/SignedMessageChain.java
index 3aa85202ed8fc2f45ca37504834ad2dcc141507a..dfa19e677e554f5abf9e6a01314d18c9b343fa8e 100644
--- a/net/minecraft/network/chat/SignedMessageChain.java
+++ b/net/minecraft/network/chat/SignedMessageChain.java
@@ -49,7 +49,7 @@ public class SignedMessageChain {
                     throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.CHAIN_BROKEN);
                 }
 
-                if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().misc.verifyChatOrder && body.timeStamp().isBefore(SignedMessageChain.this.lastTimeStamp)) { // Gale - Pufferfish - make chat order verification configurable
+                if (org.dreeam.leaf.config.modules.network.global.ChatOrderVerification.enabled && body.timeStamp().isBefore(SignedMessageChain.this.lastTimeStamp)) { // Gale - Pufferfish - make chat order verification configurable
                     this.setChainBroken();
                     throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.OUT_OF_ORDER_CHAT, org.bukkit.event.player.PlayerKickEvent.Cause.OUT_OF_ORDER_CHAT); // Paper - kick event causes
                 }
@@ -61,7 +61,7 @@ public class SignedMessageChain {
                     throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.INVALID_SIGNATURE);
                 }
 
-                if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.chat.expiredMessageWarning && unpacked.hasExpiredServer(Instant.now())) { // Gale - do not log expired message warnings
+                if (org.dreeam.leaf.config.modules.misc.global.Chat.expiredMessageWarning && unpacked.hasExpiredServer(Instant.now())) { // Gale - do not log expired message warnings
                     SignedMessageChain.LOGGER.warn("Received expired chat: '{}'. Is the client/server system time unsynchronized?", body.content());
                 }
 
diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index cd117cafbe9c3fe6bcfda09abe8afc8bd6629b66..c8d0b0b0fbbc17cc7c904a6f637727b80bbd9317 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -129,7 +129,7 @@ public class Main {
             Thread.ofPlatform().daemon().name("DataFixers init thread").start(DataFixers::getDataFixer);
             // Paper end - Perf: Init DataConverter asynchronously
             Util.startTimerHackThread();
-            org.dreeam.leaf.config.ConfigModule.loadAfterBootstrap(); // Leaf - Leaf config - post load
+            org.dreeam.leaf.config.LeafConfig.loadAfterBootstrap(); // Leaf - Leaf config - post load
             Path settingsFile = Paths.get("server.properties");
             DedicatedServerSettings settings = new DedicatedServerSettings(options); // CraftBukkit - CLI argument support
             settings.forceSave();
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 3083f7da4bd04db8724f8e457559bef4469d7675..a9be9cbf19448c7012fc562d1bebe42a013770c1 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -699,6 +699,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             collideTeam.setSeeFriendlyInvisibles(false); // Because we want to mimic them not being on a team at all
         }
         // Paper end - Configurable player collision; Handle collideRule team for player collision toggle
+        org.dreeam.leaf.config.migration.gale.GaleConfigMigration.finalizeMigration(this); // Leaf - Gale config migration
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
         this.server.spark.registerCommandBeforePlugins(this.server); // Paper - spark
         this.server.spark.enableAfterPlugins(this.server); // Paper - spark
diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 98b48a8a6f6372114ee1f71b2c69bff97af4ef64..6cdeccd4004feebd86a8637025b535f7cd5fc0f5 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -149,7 +149,7 @@ public class PlayerAdvancements {
             AdvancementHolder advancement = manager.get(id);
             if (advancement == null) {
                 if (!id.getNamespace().equals(Identifier.DEFAULT_NAMESPACE)) return; // CraftBukkit
-                if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.ignoredAdvancements) LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", id, this.playerSavePath); // Gale - Purpur - do not log ignored advancements
+                if (org.dreeam.leaf.config.modules.misc.global.LogToConsole.ignoredAdvancements) LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", id, this.playerSavePath); // Gale - Purpur - do not log ignored advancements
             } else {
                 this.startProgress(advancement, progress);
                 this.progressChanged.add(advancement); // Leaf - SparklyPaper - parallel world ticking - Always add to non-concurrent set during load, flushDirty will handle sync
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 738e3885297e65080cd514127d6ac19c42efd486..b8d319681de201583a54d9dfbfe2abe7c8d9063f 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -248,10 +248,6 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
-        // Gale start - Gale configuration
-        galeConfigurations.initializeGlobalConfiguration(this.registryAccess());
-        galeConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
-        // Gale end - Gale configuration
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index ba589ac63ebba99679e4d20a5f455da85941a44f..5cd55e01ff5e7f4bc264603e2f91534550c3e5e1 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -51,7 +51,6 @@ import net.minecraft.network.protocol.game.ClientboundBlockEventPacket;
 import net.minecraft.network.protocol.game.ClientboundDamageEventPacket;
 import net.minecraft.network.protocol.game.ClientboundEntityEventPacket;
 import net.minecraft.network.protocol.game.ClientboundExplodePacket;
-import net.minecraft.network.protocol.game.ClientboundGameEventPacket;
 import net.minecraft.network.protocol.game.ClientboundLevelEventPacket;
 import net.minecraft.network.protocol.game.ClientboundLevelParticlesPacket;
 import net.minecraft.network.protocol.game.ClientboundSoundEntityPacket;
@@ -70,7 +69,6 @@ import net.minecraft.util.CsvOutput;
 import net.minecraft.util.Mth;
 import net.minecraft.util.ProgressListener;
 import net.minecraft.util.Util;
-import net.minecraft.util.datafix.DataFixTypes;
 import net.minecraft.util.debug.DebugSubscriptions;
 import net.minecraft.util.debug.LevelDebugSynchronizers;
 import net.minecraft.util.profiling.Profiler;
@@ -134,18 +132,14 @@ import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.LevelChunkSection;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
-import net.minecraft.world.level.chunk.storage.EntityStorage;
 import net.minecraft.world.level.chunk.storage.RegionStorageInfo;
-import net.minecraft.world.level.chunk.storage.SimpleRegionStorage;
 import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.dimension.end.EnderDragonFight;
-import net.minecraft.world.level.entity.EntityPersistentStorage;
 import net.minecraft.world.level.entity.EntityTickList;
 import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.level.entity.LevelCallback;
 import net.minecraft.world.level.entity.LevelEntityGetter;
-import net.minecraft.world.level.entity.PersistentEntitySectionManager;
 import net.minecraft.world.level.gameevent.DynamicGameEventListener;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.gameevent.GameEventDispatcher;
@@ -171,7 +165,6 @@ import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
 import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.level.storage.SavedDataStorage;
-import net.minecraft.world.level.storage.ServerLevelData;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.BooleanOp;
@@ -681,7 +674,24 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         savedDataStorage.set(io.papermc.paper.world.saveddata.PaperWorldPDC.TYPE, loadedWorldData.pdc() == null ? io.papermc.paper.world.saveddata.PaperWorldPDC.TYPE.constructor().get() : loadedWorldData.pdc());
         final GameRules gameRules = new GameRules(server.getWorldData().enabledFeatures(), savedDataStorage.computeIfAbsent(net.minecraft.world.level.gamerules.GameRuleMap.TYPE));
         this.gameRules = gameRules;
-        super(levelData, dimension, server.registryAccess(), levelStem.type(), false, isDebug, biomeZoomSeed, server.getMaxChainedNeighborUpdates(), loadedWorldData.bukkitName(), gen, biomeProvider, env, spigotConfig -> server.paperConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), spigotConfig -> server.galeConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), executor); // Paper - create paper world configs // Paper - Anti-Xray - Pass executor // Gale - Gale configuration
+        super(
+            levelData,
+            dimension,
+            server.registryAccess(),
+            levelStem.type(),
+            false,
+            isDebug,
+            biomeZoomSeed,
+            server.getMaxChainedNeighborUpdates(),
+            loadedWorldData.bukkitName(),
+            gen,
+            biomeProvider,
+            env,
+            spigotConfig -> server.paperConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), // Paper - create paper world configs
+            spigotConfig -> new org.galemc.gale.configuration.GaleWorldConfiguration(spigotConfig, dimension.identifier()), // Gale - Gale configuration
+            spigotConfig -> org.dreeam.leaf.config.LeafConfig.initWorldConfig(server.storageSource.getDimensionPath(dimension)), // Leaf - per-world configuration
+            executor
+        );
         this.weatherData = savedDataStorage.computeIfAbsent(WeatherData.TYPE);
         this.weatherData.setLevel(this);
         this.typeKey = typeKey;
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 4030776c3d12f8e797720180e141322b5d2cc548..4e2562d81775b94e7b10f1b64fddde5a358f52ca 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2403,7 +2403,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public void crit(final Entity entity) {
         // Gale start - MultiPaper - broadcast crit animations as the entity being critted
         var level = this.level();
-        level.getChunkSource().sendToTrackingPlayersAndSelf(level.galeConfig().gameplayMechanics.fixes.broadcastCritAnimationsAsTheEntityBeingCritted ? entity : this, new ClientboundAnimatePacket(entity, ClientboundAnimatePacket.CRITICAL_HIT));
+        level.getChunkSource().sendToTrackingPlayersAndSelf(level.leafConfig().fixes.broadcastCritAnimationsAsTheEntityBeingCritted ? entity : this, new ClientboundAnimatePacket(entity, ClientboundAnimatePacket.CRITICAL_HIT));
         // Gale end - MultiPaper - broadcast crit animations as the entity being critted
     }
 
@@ -2411,7 +2411,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public void magicCrit(final Entity entity) {
         // Gale start - MultiPaper - broadcast crit animations as the entity being critted
         var level = this.level();
-        level.getChunkSource().sendToTrackingPlayersAndSelf(level.galeConfig().gameplayMechanics.fixes.broadcastCritAnimationsAsTheEntityBeingCritted ? entity : this, new ClientboundAnimatePacket(entity, ClientboundAnimatePacket.MAGIC_CRITICAL_HIT));
+        level.getChunkSource().sendToTrackingPlayersAndSelf(level.leafConfig().fixes.broadcastCritAnimationsAsTheEntityBeingCritted ? entity : this, new ClientboundAnimatePacket(entity, ClientboundAnimatePacket.MAGIC_CRITICAL_HIT));
         // Gale end - MultiPaper - broadcast crit animations as the entity being critted
     }
 
diff --git a/net/minecraft/server/level/WorldGenRegion.java b/net/minecraft/server/level/WorldGenRegion.java
index 1049e948496fcbab5f3d5789ec1e975692309a36..96594a636d6c871b007e671daf290684d743936f 100644
--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -327,7 +327,7 @@ public class WorldGenRegion implements WorldGenLevel {
             int chunkX = SectionPos.blockToSectionCoord(pos.getX());
             int chunkZ = SectionPos.blockToSectionCoord(pos.getZ());
             // Paper start - Buffer OOB setBlock calls
-            if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.setBlockInFarChunk && !hasSetFarWarned) { // Gale - Purpur - do not log setBlock in far chunks
+            if (org.dreeam.leaf.config.modules.misc.global.LogToConsole.setBlockInFarChunk && !hasSetFarWarned) { // Gale - Purpur - do not log setBlock in far chunks
             Util.logAndPauseIfInIde(
                 "Detected setBlock in a far chunk ["
                     + chunkX
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 1c0b616aa34e93e015267fc2415d00431ad494d7..5c7272f449b569d3cdefc296088808b9b94b37c4 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -115,7 +115,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         long now = System.nanoTime();
         io.papermc.paper.util.KeepAlive.PendingKeepAlive pending = this.keepAlive.pendingKeepAlives.peek();
         // Gale start - Purpur - send multiple keep-alive packets
-        if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().misc.keepalive.sendMultiple) {
+        if (org.dreeam.leaf.config.modules.network.global.Keepalive.sendMultiple) {
             if (this.keepAlivePending && !keepAlives.isEmpty() && keepAlives.contains(packet.getId())) {
                 int ping = (int) (Util.getMillis() - packet.getId());
                 int updatedLatency = (this.latency * 3 + ping) / 4;
@@ -306,7 +306,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         Profiler.get().push("keepAlive");
         long now = Util.getMillis();
         // Gale start - Purpur - send multiple keep-alive packets
-        if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().misc.keepalive.sendMultiple) {
+        if (org.dreeam.leaf.config.modules.network.global.Keepalive.sendMultiple) {
             if (this.checkIfClosed(now)) {
                 long currTime = System.nanoTime();
                 if ((currTime - this.keepAlive.lastKeepAliveTx) >= java.util.concurrent.TimeUnit.SECONDS.toNanos(1L)) { // 1 second
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 118f4df4599768488d5fd86e97b0b024bea00365..c7eed64b8446dca61c134ed18283ceb12d6b52c9 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1359,12 +1359,12 @@ public class ServerGamePacketListenerImpl
 
     @Override
     public void handleEditBook(final ServerboundEditBookPacket packet) {
-        // Gale start - Pufferfish - make book writing configurable
+        // Leaf start - Pufferfish - make book writing configurable
         final org.bukkit.craftbukkit.entity.CraftPlayer craftPlayer = this.player.getBukkitEntity();
-        if (!(org.galemc.gale.configuration.GaleGlobalConfiguration.get().gameplayMechanics.enableBookWriting || craftPlayer.hasPermission(org.bukkit.craftbukkit.util.permissions.CraftDefaultPermissions.writeBooks) || craftPlayer.hasPermission("pufferfish.usebooks"))) {
+        if (!(org.dreeam.leaf.config.modules.gameplay.global.GameplayMechanics.enableBookWriting || craftPlayer.hasPermission(org.bukkit.craftbukkit.util.permissions.CraftDefaultPermissions.writeBooks) || craftPlayer.hasPermission("pufferfish.usebooks"))) {
             return;
         }
-        // Gale end - Pufferfish - make book writing configurable
+        // Leaf end - Pufferfish - make book writing configurable
         // Paper start - Book size limits
         final io.papermc.paper.configuration.type.number.IntOr.Disabled pageMax = io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.bookSize.pageMax;
         if (!this.cserver.isPrimaryThread() && pageMax.enabled()) {
@@ -2745,7 +2745,7 @@ public class ServerGamePacketListenerImpl
         // CraftBukkit start
         String rawMessage = message.signedContent();
         if (rawMessage.isEmpty()) {
-            if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.chat.emptyMessageWarning) LOGGER.warn("{} tried to send an empty message", this.player.getScoreboardName()); // Gale - do not log empty message warnings
+            if (org.dreeam.leaf.config.modules.misc.global.Chat.emptyMessageWarning) LOGGER.warn("{} tried to send an empty message", this.player.getScoreboardName()); // Gale - do not log empty message warnings
         } else if (this.getCraftPlayer().isConversing()) {
             final String conversationInput = rawMessage;
             this.server.processQueue.add(() -> ServerGamePacketListenerImpl.this.getCraftPlayer().acceptConversationInput(conversationInput));
diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 4bbf4cd3ca3d412b70e3a851c2873240baff13c6..77d0e507458ef2bdc4c863860322910e468efdb3 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -106,7 +106,7 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
     public void tickTimeout() {
     // Paper end - login cookie API
         // Gale start - make slow login timeout configurable
-        int slowLoginTimeout = org.galemc.gale.configuration.GaleGlobalConfiguration.get().misc.premiumAccountSlowLoginTimeout;
+        int slowLoginTimeout = org.dreeam.leaf.config.modules.network.global.PremiumAccountSlowLoginTimeout.ticks;
         if (this.tick++ >= (slowLoginTimeout < 1 ? MAX_TICKS_BEFORE_LOGIN : slowLoginTimeout)) {
             // Gale end - make slow login timeout configurable
             this.disconnectAsync(Component.translatable("multiplayer.disconnect.slow_login")); // Paper
@@ -139,7 +139,7 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
     public void disconnect(final Component component) {
         try {
             // Gale start - Pufferfish - do not log disconnections with null id
-            if (!org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.nullIdDisconnections && this.authenticatedProfile != null && this.authenticatedProfile.id() == null) {
+            if (!org.dreeam.leaf.config.modules.misc.global.LogToConsole.nullIdDisconnections && this.authenticatedProfile != null && this.authenticatedProfile.id() == null) {
                 String reasonString = component.getString();
                 if ("Disconnected".equals(reasonString) || Component.translatable("multiplayer.disconnect.generic").getString().equals(reasonString)) {
                     return;
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 8f451c9116d0e464d01e85449891c7270bb1b0d7..6ebd0a86a07ae0a21976614fe6d0084b11226a9d 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -450,7 +450,7 @@ public abstract class PlayerList {
         }
         // Paper end - Configurable player collision
         org.purpurmc.purpur.task.BossBarTask.addToAll(player); // Purpur - Implement TPSBar
-        if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.playerLoginLocations) { // Gale - JettPack - make logging login location configurable
+        if (org.dreeam.leaf.config.modules.misc.global.LogToConsole.playerLoginLocations) { // Gale - JettPack - make logging login location configurable
         // CraftBukkit start - moved down
         LOGGER.info(
             "{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", // Paper - add world identifier
@@ -1420,7 +1420,7 @@ public abstract class PlayerList {
     public void broadcastChatMessage(final PlayerChatMessage message, final Predicate<ServerPlayer> isFiltered, final @Nullable ServerPlayer senderPlayer, final ChatType.Bound chatType, final @Nullable Function<net.kyori.adventure.audience.Audience, Component> unsignedFunction) {
         // Paper end
         boolean trusted = this.verifyChatTrusted(message);
-        this.server.logChatMessage((unsignedFunction == null ? message.decoratedContent() : unsignedFunction.apply(this.server.console)), chatType, trusted || !org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled || !org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.chat.notSecureMarker ? null : "Not Secure"); // Paper // Gale - do not log Not Secure marker // Leaf - Mirai - Configurable chat message signatures
+        this.server.logChatMessage((unsignedFunction == null ? message.decoratedContent() : unsignedFunction.apply(this.server.console)), chatType, trusted || !org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled || !org.dreeam.leaf.config.modules.misc.global.Chat.notSecureMarker ? null : "Not Secure"); // Paper // Gale - do not log Not Secure marker // Leaf - Mirai - Configurable chat message signatures
         OutgoingChatMessage tracked = OutgoingChatMessage.create(message);
         boolean wasFullyFiltered = false;
 
diff --git a/net/minecraft/stats/ServerRecipeBook.java b/net/minecraft/stats/ServerRecipeBook.java
index 95a50f25455ec86f7662e9abd039be17609ed1e4..1ed76c10465bc14b5c50b8394992a48478d64d1b 100644
--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -110,7 +110,7 @@ public class ServerRecipeBook extends RecipeBook {
     ) {
         for (ResourceKey<Recipe<?>> recipe : recipes) {
             if (!validator.test(recipe)) {
-                if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.unrecognizedRecipes) LOGGER.error("Tried to load unrecognized recipe: {} removed now.", recipe); // Gale - Purpur - do not log unrecognized recipes
+                if (org.dreeam.leaf.config.modules.misc.global.LogToConsole.unrecognizedRecipes) LOGGER.error("Tried to load unrecognized recipe: {} removed now.", recipe); // Gale - Purpur - do not log unrecognized recipes
             } else {
                 recipeAddingMethod.accept(recipe);
             }
diff --git a/net/minecraft/stats/ServerStatsCounter.java b/net/minecraft/stats/ServerStatsCounter.java
index 1df6efebeda39097d75590c751ec635962b6e626..b9097219282a028e1c3ac02fb28b2bab4e7c3efa 100644
--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -127,7 +127,7 @@ public class ServerStatsCounter extends StatsCounter {
         this.stats
             .putAll(
                 STATS_CODEC.parse(data.get("stats").orElseEmptyMap())
-                    .resultOrPartial(error -> {if (org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.invalidStatistics) LOGGER.error("Failed to parse statistics for {}: {}", this.file, error);}) // Gale - EMC - do not log invalid statistics
+                    .resultOrPartial(error -> {if (org.dreeam.leaf.config.modules.misc.global.LogToConsole.invalidStatistics) LOGGER.error("Failed to parse statistics for {}: {}", this.file, error);}) // Gale - EMC - do not log invalid statistics
                     .orElse(Map.of())
             );
     }
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 650a382301321196b5f547f500869f2906311260..7df5dd2be9315c118e8feb1011a876bdfae918e3 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1039,7 +1039,7 @@ public abstract class Entity
         if (!this.level().isClientSide()) {
             // Gale start - Slice - hide flames on entities with fire resistance
             if (this instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
-                this.setSharedFlagOnFire(this.remainingFireTicks > 0 && (!this.level.galeConfig().gameplayMechanics.hideFlamesOnEntitiesWithFireResistance || !livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE)));
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0 && (!this.level.leafConfig().hideFlamesOnEntitiesWithFireResistance.enabled || !livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE)));
             } else {
             // Gale end - Slice - hide flames on entities with fire resistance
             this.setSharedFlagOnFire(this.remainingFireTicks > 0);
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 14cf550c114df95fd43a5e002715c34e92be3c2a..e7b2ee8b275cb701490d865be72bdf0dd7164fa5 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -463,7 +463,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         if (this.isAlive() && this.level() instanceof ServerLevel level) {
             boolean isPlayer = this instanceof Player;
             // Gale start - Pufferfish - reduce in wall checks
-            long checkStuckInWallInterval = this.level().galeConfig().smallOptimizations.reducedIntervals.checkStuckInWall;
+            long checkStuckInWallInterval = this.level().leafConfig().reducedIntervals.checkStuckInWall;
             if ((checkStuckInWallInterval <= 1 || (tickCount % checkStuckInWallInterval == 0 && couldPossiblyBeHurt(1.0F))) && this.isInWall()) {
                 // Gale end - Pufferfish - reduce in wall checks
                 this.hurtServer(level, this.damageSources().inWall(), 1.0F);
diff --git a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
index 42c4ec0446ef5f78fba953fcaebae911de658aa4..36acbcd056b09c5d97a93e65fc77b4e266e6f41f 100644
--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -103,7 +103,7 @@ public class BehaviorUtils {
         itemEntity.setDeltaMovement(throwVector);
         // Gale start - EMC - reduce villager item re-pickup
         if (thrower instanceof net.minecraft.world.entity.npc.villager.Villager) {
-            int repickupDelay = thrower.level().galeConfig().smallOptimizations.reducedIntervals.villagerItemRepickup;
+            int repickupDelay = thrower.level().leafConfig().reducedIntervals.villagerItemRepickup;
 
             if (repickupDelay <= -1) {
                 itemEntity.setDefaultPickUpDelay();
diff --git a/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java b/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
index bc8fdce935f26c4ddb5c1b7e95e6317816fb9e88..df5fcee117314cbccfab6c385cec1346c5e055de 100644
--- a/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
@@ -50,7 +50,7 @@ public class RandomStrollGoal extends Goal {
         }
 
         Vec3 pos = this.getPosition();
-        if (pos == null || (!this.mob.level().galeConfig().gameplayMechanics.entitiesCanRandomStrollIntoNonTickingChunks && !((net.minecraft.server.level.ServerLevel) this.mob.level()).isPositionEntityTicking(net.minecraft.core.BlockPos.containing(pos)))) { // Gale - MultiPaper - prevent entities random strolling into non-ticking chunks
+        if (pos == null || (!this.mob.level().leafConfig().randomStrollIntoNonTickingChunks.enabled && !((net.minecraft.server.level.ServerLevel) this.mob.level()).isPositionEntityTicking(net.minecraft.core.BlockPos.containing(pos)))) { // Gale - MultiPaper - prevent entities random strolling into non-ticking chunks
             return false;
         }
 
diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 1b2edb6ebf41041bf1d3aabcfe8ea348c0dcd469..e364d82ac613d6bc5e3997537a8348538c1e0047 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -121,7 +121,7 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
                 this.mob.lookAt(target, 30.0F, 30.0F);
                 // Gale start - Purpur - fix MC-121706
             }
-            if (!hasStrafingTime || this.mob.level().galeConfig().gameplayMechanics.fixes.mc121706) {
+            if (!hasStrafingTime || this.mob.level().leafConfig().fixes.mc121706) {
                 // Gale end - Purpur - fix MC-121706
                 this.mob.getLookControl().setLookAt(target, 30.0F, 30.0F);
             }
diff --git a/net/minecraft/world/entity/ai/sensing/Sensing.java b/net/minecraft/world/entity/ai/sensing/Sensing.java
index fea91f2b1738ee82cc2da356851e9a91fa262d55..3a891ff9d466b2597545b8d91098c9bf4a23a086 100644
--- a/net/minecraft/world/entity/ai/sensing/Sensing.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensing.java
@@ -20,7 +20,7 @@ public class Sensing {
     public Sensing(final Mob mob) {
         this.mob = mob;
         // Gale start - Petal - reduce line of sight updates - expiring entity id lists
-        int updateLineOfSightInterval = org.galemc.gale.configuration.GaleGlobalConfiguration.get().smallOptimizations.reducedIntervals.updateEntityLineOfSight;
+        int updateLineOfSightInterval = org.dreeam.leaf.config.modules.opt.global.ReducedIntervals.updateEntityLineOfSight;
         if (updateLineOfSightInterval <= 1) {
             this.expiring = null;
         } else {
diff --git a/net/minecraft/world/entity/animal/fish/WaterAnimal.java b/net/minecraft/world/entity/animal/fish/WaterAnimal.java
index 9bbeb92efe6eae57114792578c935242b4fd753f..3d26752e6ed23b5a43dad89434b6e7558e7394d3 100644
--- a/net/minecraft/world/entity/animal/fish/WaterAnimal.java
+++ b/net/minecraft/world/entity/animal/fish/WaterAnimal.java
@@ -81,7 +81,7 @@ public abstract class WaterAnimal extends PathfinderMob {
         minSpawnLevel = level.getMinecraftWorld().paperConfig().entities.spawning.wateranimalSpawnHeight.minimum.or(minSpawnLevel);
         // Paper end - Make water animal spawn height configurable
         // Gale start - Purpur - fix MC-238526
-        boolean shouldFixMC238526 = spawnReason == EntitySpawnReason.SPAWNER && level.getMinecraftWorld().galeConfig().gameplayMechanics.fixes.mc238526;
+        boolean shouldFixMC238526 = spawnReason == EntitySpawnReason.SPAWNER && level.getMinecraftWorld().leafConfig().fixes.mc238526;
         boolean isAllowedHeight = pos.getY() >= minSpawnLevel && pos.getY() <= seaLevel;
         return (shouldFixMC238526 || isAllowedHeight)
             // Gale end - Purpur - fix MC-238526
diff --git a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 571da17706bf21ef955117e9e2bbda24ed144bba..8f232847b3944607c2afefd06f98fe1ecd9c1003 100644
--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -351,10 +351,10 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         return DoubleDoubleImmutablePair.of(dx, dz);
     }
 
-    // Gale start - EMC - make saving fireworks configurable
+    // Leaf start - EMC - make saving fireworks configurable
     @Override
     public boolean shouldBeSaved() {
-        return this.level().galeConfig().smallOptimizations.saveFireworks && super.shouldBeSaved();
+        return this.level().leafConfig().saveFireworks.enabled && super.shouldBeSaved();
     }
-    // Gale end - EMC - make saving fireworks configurable
+    // Leaf end - EMC - make saving fireworks configurable
 }
diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index 8f0113a7f74cc207fc9e3ebe487af58aa0e3424d..4446ed8a1a75315386886593e5113f8f76e3c041 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -72,19 +72,19 @@ public abstract class Projectile extends Entity implements TraceableEntity {
             boolean isLoaded = ((net.minecraft.server.level.ServerChunkCache) this.level().getChunkSource()).getChunkAtIfLoadedImmediately(newX, newZ) != null;
 
             if (!isLoaded) {
-                var maxProjectileChunkLoadsConfig = this.level().galeConfig().smallOptimizations.maxProjectileChunkLoads;
+                var maxProjectileChunkLoadsConfig = this.level().leafConfig().maxProjectileChunkLoads;
                 int maxChunkLoadsPerTick = maxProjectileChunkLoadsConfig.perTick;
 
                 if (maxChunkLoadsPerTick >= 0 && chunksLoadedThisTick > maxChunkLoadsPerTick) {
                     return;
                 }
 
-                int maxChunkLoadsPerProjectile = maxProjectileChunkLoadsConfig.perProjectile.max;
+                int maxChunkLoadsPerProjectile = maxProjectileChunkLoadsConfig.perProjectileMax;
 
                 if (maxChunkLoadsPerProjectile >= 0 && this.chunksLoadedByProjectile >= maxChunkLoadsPerProjectile) {
-                    if (maxProjectileChunkLoadsConfig.perProjectile.removeFromWorldAfterReachLimit) {
+                    if (maxProjectileChunkLoadsConfig.perProjectileRemoveFromWorldAfterReachLimit) {
                         this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD); // Leaf - Fix Pufferfish and Purpur patches - Purpur
-                    } else if (maxProjectileChunkLoadsConfig.perProjectile.resetMovementAfterReachLimit) {
+                    } else if (maxProjectileChunkLoadsConfig.perProjectileResetMovementAfterReachLimit) {
                         this.setDeltaMovement(0, this.getDeltaMovement().y, 0);
                     }
 
diff --git a/net/minecraft/world/item/DyeColor.java b/net/minecraft/world/item/DyeColor.java
index 35b14be6dc363f1727133e985a49ef262da88373..77c6c009e65d2e1ca7f2821b82bca5b8a4edb702 100644
--- a/net/minecraft/world/item/DyeColor.java
+++ b/net/minecraft/world/item/DyeColor.java
@@ -212,7 +212,7 @@ public enum DyeColor implements StringRepresentable {
 
     public static DyeColor getMixedColor(final ServerLevel level, final DyeColor dyeColor1, final DyeColor dyeColor2) {
         // Gale start - carpet-fixes - optimize sheep offspring color
-        if (level.galeConfig().smallOptimizations.useOptimizedSheepOffspringColor) {
+        if (level.leafConfig().optimizedSheepOffspringColor.enabled) {
             DyeColor col = properDye(dyeColor1, dyeColor2);
             if (col == null) col = level.simpleRandom.nextBoolean() ? dyeColor1 : dyeColor2;
             return col;
diff --git a/net/minecraft/world/item/EndCrystalItem.java b/net/minecraft/world/item/EndCrystalItem.java
index b2b8efad9da71fddb1a1011c05b7d6c7fe71bffb..1675a20199068972be743251feeb783e0cc823d9 100644
--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -52,7 +52,7 @@ public class EndCrystalItem extends Item {
             // CraftBukkit end
             level.addFreshEntity(crystal);
             level.gameEvent(context.getPlayer(), GameEvent.ENTITY_PLACE, above);
-            if (level.galeConfig().gameplayMechanics.tryRespawnEnderDragonAfterEndCrystalPlace) { // Gale - Pufferfish - make ender dragon respawn attempt after placing end crystals configurable
+            if (level.leafConfig().enderDragonRespawn.tryAfterEndCrystalPlace) { // Gale - Pufferfish - make ender dragon respawn attempt after placing end crystals configurable
             EnderDragonFight fight = serverLevel.getDragonFight();
             if (fight != null) {
                 fight.tryRespawn(above); // Paper - Perf: Do crystal-portal proximity check before entity lookup
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index dcc783fe5aa1b4790db5873e22b32ea4517d68df..435792ec17f7b1171825af22406a83fe21804452 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -179,6 +179,16 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
     // Gale end - Gale configuration
 
+    // Leaf start - per-world configuration
+    private volatile org.dreeam.leaf.config.LeafWorldConfig leafConfig;
+    public org.dreeam.leaf.config.LeafWorldConfig leafConfig() {
+        return this.leafConfig;
+    }
+    public void setLeafConfig(org.dreeam.leaf.config.LeafWorldConfig leafConfig) {
+        this.leafConfig = leafConfig;
+    }
+    // Leaf end - per-world configuration
+
     public final org.purpurmc.purpur.PurpurWorldConfig purpurConfig; // Purpur - Purpur config files
     public final io.papermc.paper.redstone.RedstoneWireTurbo turbo; // Leaf - SparklyPaper - parallel world ticking - moved to world
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
@@ -895,6 +905,8 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         io.papermc.paper.configuration.WorldConfiguration> paperWorldConfigCreator, // Paper - create paper world config
         java.util.function.Function<org.spigotmc.SpigotWorldConfig, // Gale - Gale configuration
         org.galemc.gale.configuration.GaleWorldConfiguration> galeWorldConfigCreator, // Gale - Gale configuration
+        java.util.function.Function<org.spigotmc.SpigotWorldConfig, // Leaf - per-world configuration
+        org.dreeam.leaf.config.LeafWorldConfig> leafWorldConfigCreator, // Leaf - per-world configuration
         java.util.concurrent.Executor executor // Paper - Anti-Xray
     ) {
         // Paper start - getblock optimisations - cache world height/sections
@@ -911,6 +923,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
         this.purpurConfig = new org.purpurmc.purpur.PurpurWorldConfig(bukkitName, environment, worldKey); // Purpur - Purpur config files
         this.galeConfig = galeWorldConfigCreator.apply(this.spigotConfig); // Gale - Gale configuration
+        this.leafConfig = leafWorldConfigCreator.apply(this.spigotConfig); // Leaf - per-world configuration
         this.playerBreedingCooldowns = this.getNewBreedingCooldownCache(); // Purpur - Add adjustable breeding cooldown to config
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, worldKey, biomeProvider, environment);
diff --git a/net/minecraft/world/level/levelgen/PhantomSpawner.java b/net/minecraft/world/level/levelgen/PhantomSpawner.java
index ba9a21ab3188a8ff0681e7d368eb4d2161d2e2c7..a30bdbdfd77c31ce330bad6be956325654059e94 100644
--- a/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -54,7 +54,7 @@ public class PhantomSpawner implements CustomSpawner {
                                                 .south(-10 + random.nextInt(21));
                                             // Gale start - MultiPaper - don't load chunks to spawn phantoms
                                             BlockState blockState;
-                                            if (level.galeConfig().smallOptimizations.loadChunks.toSpawnPhantoms) {
+                                            if (level.leafConfig().loadChunks.toSpawnPhantoms) {
                                                 blockState = level.getBlockState(spawnPos);
                                             } else {
                                                 blockState = level.getBlockStateIfLoaded(spawnPos);
