From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Optimize tickPrecipitation


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index ba589ac63ebba99679e4d20a5f455da85941a44f..069320a28d824531e93e66e5d325477fc32ce1fc 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1269,7 +1269,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         if (!this.paperConfig().environment.disableIceAndSnow) { // Paper - Option to disable ice and snow
         for (int i = 0; i < tickSpeed; i++) {
             if (simpleRandom.nextInt(rate) == 0) { // Paper - optimise random ticking // Leaf - configurable ice and snow tick chance
-                this.tickPrecipitation(this.getBlockRandomPos(minX, 0, minZ, 15));
+                this.tickPrecipitation(this.getBlockRandomPos(minX, 0, minZ, 15), chunk); // Leaf - Optimize tickPrecipitation
             }
         }
         } // Paper - Option to disable ice and snow
@@ -1327,10 +1327,16 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     @VisibleForTesting
+    // Leaf start - Optimize tickPrecipitation
     public void tickPrecipitation(final BlockPos pos) {
-        BlockPos topPos = this.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, pos);
+        this.tickPrecipitation(pos, null);
+    }
+
+    public void tickPrecipitation(final BlockPos pos, final @Nullable LevelChunk chunk) {
+        BlockPos topPos = chunk == null ? this.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, pos) : new BlockPos(pos.getX(), chunk.getHeight(Heightmap.Types.MOTION_BLOCKING, pos.getX(), pos.getZ()) + 1, pos.getZ());
+        // Leaf end - Optimize tickPrecipitation
         BlockPos belowPos = topPos.below();
-        Biome biome = this.getBiome(topPos).value();
+        Biome biome = this.getBiomeCached(chunk, topPos).value(); // Leaf - Optimize tickPrecipitation
         if (biome.shouldFreeze(this, belowPos)) {
             org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this, belowPos, Blocks.ICE.defaultBlockState(), Block.UPDATE_ALL, null); // CraftBukkit
         }
