From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 12 Oct 2022 10:42:15 -0400
Subject: [PATCH] Leaf config changes


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index ba589ac63ebba99679e4d20a5f455da85941a44f..3739eaf02c4aa74108b732f8a9c7bcbd0fdbbae2 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -681,7 +681,24 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         savedDataStorage.set(io.papermc.paper.world.saveddata.PaperWorldPDC.TYPE, loadedWorldData.pdc() == null ? io.papermc.paper.world.saveddata.PaperWorldPDC.TYPE.constructor().get() : loadedWorldData.pdc());
         final GameRules gameRules = new GameRules(server.getWorldData().enabledFeatures(), savedDataStorage.computeIfAbsent(net.minecraft.world.level.gamerules.GameRuleMap.TYPE));
         this.gameRules = gameRules;
-        super(levelData, dimension, server.registryAccess(), levelStem.type(), false, isDebug, biomeZoomSeed, server.getMaxChainedNeighborUpdates(), loadedWorldData.bukkitName(), gen, biomeProvider, env, spigotConfig -> server.paperConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), spigotConfig -> server.galeConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), executor); // Paper - create paper world configs // Paper - Anti-Xray - Pass executor // Gale - Gale configuration
+        super(
+            levelData,
+            dimension,
+            server.registryAccess(),
+            levelStem.type(),
+            false,
+            isDebug,
+            biomeZoomSeed,
+            server.getMaxChainedNeighborUpdates(),
+            loadedWorldData.bukkitName(),
+            gen,
+            biomeProvider,
+            env,
+            spigotConfig -> server.paperConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), // Paper - create paper world configs
+            spigotConfig -> server.galeConfigurations.createWorldConfig(io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(server.storageSource.getDimensionPath(dimension), dimension.identifier(), spigotConfig, server.registryAccess(), gameRules)), // Gale - Gale configuration
+            spigotConfig -> org.dreeam.leaf.config.LeafConfig.createWorldConfig(server.storageSource.getDimensionPath(dimension)), // Leaf - per-world configuration
+            executor
+        );
         this.weatherData = savedDataStorage.computeIfAbsent(WeatherData.TYPE);
         this.weatherData.setLevel(this);
         this.typeKey = typeKey;
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index dcc783fe5aa1b4790db5873e22b32ea4517d68df..048f4d0f0ce0166059781835c1590ac7ed49f934 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -179,6 +179,13 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
     // Gale end - Gale configuration
 
+    // Leaf start - per-world configuration
+    private final org.dreeam.leaf.config.LeafWorldConfig leafConfig;
+    public org.dreeam.leaf.config.LeafWorldConfig leafConfig() {
+        return this.leafConfig;
+    }
+    // Leaf end - per-world configuration
+
     public final org.purpurmc.purpur.PurpurWorldConfig purpurConfig; // Purpur - Purpur config files
     public final io.papermc.paper.redstone.RedstoneWireTurbo turbo; // Leaf - SparklyPaper - parallel world ticking - moved to world
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
@@ -895,6 +902,8 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         io.papermc.paper.configuration.WorldConfiguration> paperWorldConfigCreator, // Paper - create paper world config
         java.util.function.Function<org.spigotmc.SpigotWorldConfig, // Gale - Gale configuration
         org.galemc.gale.configuration.GaleWorldConfiguration> galeWorldConfigCreator, // Gale - Gale configuration
+        java.util.function.Function<org.spigotmc.SpigotWorldConfig, // Leaf - per-world configuration
+        org.dreeam.leaf.config.LeafWorldConfig> leafWorldConfigCreator, // Leaf - per-world configuration
         java.util.concurrent.Executor executor // Paper - Anti-Xray
     ) {
         // Paper start - getblock optimisations - cache world height/sections
@@ -911,6 +920,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
         this.purpurConfig = new org.purpurmc.purpur.PurpurWorldConfig(bukkitName, environment, worldKey); // Purpur - Purpur config files
         this.galeConfig = galeWorldConfigCreator.apply(this.spigotConfig); // Gale - Gale configuration
+        this.leafConfig = leafWorldConfigCreator.apply(this.spigotConfig); // Leaf - per-world configuration
         this.playerBreedingCooldowns = this.getNewBreedingCooldownCache(); // Purpur - Add adjustable breeding cooldown to config
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, worldKey, biomeProvider, environment);
