From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Reduce debug subscribers overhead


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index a3290b777531d866c9b49de4447924cb121ca371..94b14dc0bf9d33273c9559bfd11f9846c7cb00b9 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1987,7 +1987,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         profilerFiller.popPush("players");
         this.playerList.tick();
         profilerFiller.popPush("debugSubscribers");
-        this.debugSubscribers.tick();
+        if (!org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) this.debugSubscribers.tick(); // Leaf - Reduce debug subscribers overhead
         if (this.tickRateManager.runsNormally()) {
             profilerFiller.popPush("gameTests");
             GameTestTicker.SINGLETON.tick();
diff --git a/net/minecraft/util/debug/ServerDebugSubscribers.java b/net/minecraft/util/debug/ServerDebugSubscribers.java
index a700521a74b7e8b0c3000df18dce85592f01e568..5bdd43ef13d85005d969602cdfd4861e441e7b22 100644
--- a/net/minecraft/util/debug/ServerDebugSubscribers.java
+++ b/net/minecraft/util/debug/ServerDebugSubscribers.java
@@ -13,17 +13,24 @@ import net.minecraft.server.players.NameAndId;
 
 public class ServerDebugSubscribers {
     private final MinecraftServer server;
-    private final Map<DebugSubscription<?>, List<ServerPlayer>> enabledSubscriptions = new HashMap<>();
+    // Leaf start - Reduce debug subscribers overhead
+    private final Map<DebugSubscription<?>, List<ServerPlayer>> enabledSubscriptions = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<>();
+    private int tickCount;
+    private static final Set<DebugSubscription<?>> EMPTY_SUBSCRIPTIONS = Set.of();
+    private static final List<ServerPlayer> EMPTY_PLAYERS = List.of();
+    // Leaf end - Reduce debug subscribers overhead
 
     public ServerDebugSubscribers(MinecraftServer server) {
         this.server = server;
     }
 
     private List<ServerPlayer> getSubscribersFor(DebugSubscription<?> subscription) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return EMPTY_PLAYERS; // Leaf - Reduce debug subscribers overhead
         return this.enabledSubscriptions.getOrDefault(subscription, List.of());
     }
 
     public void tick() {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE || tickCount++ % 20 != 0) return; // Leaf - Reduce debug subscribers overhead
         this.enabledSubscriptions.values().forEach(List::clear);
 
         for (ServerPlayer serverPlayer : this.server.getPlayerList().getPlayers()) {
@@ -36,20 +43,24 @@ public class ServerDebugSubscribers {
     }
 
     public void broadcastToAll(DebugSubscription<?> subscription, Packet<?> packet) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return; // Leaf - Reduce debug subscribers overhead
         for (ServerPlayer serverPlayer : this.getSubscribersFor(subscription)) {
             serverPlayer.connection.send(packet);
         }
     }
 
     public Set<DebugSubscription<?>> enabledSubscriptions() {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return EMPTY_SUBSCRIPTIONS; // Leaf - Reduce debug subscribers overhead
         return Set.copyOf(this.enabledSubscriptions.keySet());
     }
 
     public boolean hasAnySubscriberFor(DebugSubscription<?> subscription) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return false; // Leaf - Reduce debug subscribers overhead
         return !this.getSubscribersFor(subscription).isEmpty();
     }
 
     public boolean hasRequiredPermissions(ServerPlayer player) {
+        if (org.dreeam.leaf.util.LeafConstants.DISABLE_VANILLA_DEBUG_FEATURE) return false; // Leaf - Reduce debug subscribers overhead
         NameAndId nameAndId = player.nameAndId();
         return SharedConstants.IS_RUNNING_IN_IDE && this.server.isSingleplayerOwner(nameAndId) || this.server.getPlayerList().isOp(nameAndId);
     }
