From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Add read-only mode for Linear v2


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
index f950472f371af3581af82c5bcdd7a800191cf051..0ca348d43143d11ba6dd0e6df33b5ee90ad8c381 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
@@ -1705,7 +1705,7 @@ public final class NewChunkHolder {
 
     private boolean saveChunk(final ChunkAccess chunk, final boolean unloading, final Completable<CompoundTag>[] chunkSave) {
         // Leaf start - disable world data saving
-        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world)) {
+        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world) || org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) { // Leaf - Add read-only mode for Linear v2
             if (chunk.isUnsaved()) {
                 chunk.tryMarkSaved();
             }
@@ -1763,6 +1763,7 @@ public final class NewChunkHolder {
     private boolean lastEntitySaveNull;
     private boolean saveEntities(final ChunkEntitySlices entities, final boolean unloading, final Completable<CompoundTag>[] entitySave) {
         if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world)) return false; // Leaf - disable world data saving
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return false; // Leaf - Add read-only mode for Linear v2
         try {
             if (entities.isTransient()) {
                 if (!unloading) {
@@ -1820,7 +1821,7 @@ public final class NewChunkHolder {
     private boolean lastPoiSaveNull;
     private boolean savePOI(final PoiChunk poi, final boolean unloading, final Completable<CompoundTag>[] poiSave) {
         // Leaf start - disable world data saving
-        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world)) {
+        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this.world) || org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) { // Leaf - Add read-only mode for Linear v2
             if (poi.isDirty()) {
                 poi.setDirty(false);
             }
diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 1d6644566947a7222dbb06470fecfd2927ba71f5..9f8dc80f279e2fa6e4c7a292c6239dc5fe9ee15f 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -130,6 +130,7 @@ public class PlayerAdvancements {
 
     public void save() {
         if (org.spigotmc.SpigotConfig.disableAdvancementSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         JsonElement jsonElement = this.codec.encodeStart(JsonOps.INSTANCE, this.asData()).getOrThrow();
 
         try {
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 3fb0c67855de020b219790f542d25975803d7720..0e382269d5c0ae85ccf424104e9fb404d9cc8ce0 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1715,6 +1715,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     // Paper start - Incremental chunk and player saving
     public void saveIncrementally(boolean doFull) {
         if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this)) return; // Leaf - disable world data saving
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         if (doFull) {
             org.bukkit.Bukkit.getPluginManager().callEvent(new org.bukkit.event.world.WorldSaveEvent(this.getWorld()));
         }
@@ -1741,7 +1742,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         // Paper end - add close param
         ServerChunkCache chunkSource = this.getChunkSource();
         // Leaf start - disable world data saving
-        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this)) {
+        if (org.dreeam.leaf.config.modules.misc.DisableWorldDataSaving.shouldSkipSave(this) || org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) { // Leaf - Add read-only mode for Linear v2
             if (close) {
                 try {
                     chunkSource.close(false);
diff --git a/net/minecraft/stats/ServerStatsCounter.java b/net/minecraft/stats/ServerStatsCounter.java
index e8af6a4877d01cc56ef0347013978ba9d583aa7e..a697c066efd6b63ec5e14144ed050430197ae889 100644
--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -90,6 +90,7 @@ public class ServerStatsCounter extends StatsCounter {
 
     public void save() {
         if (org.spigotmc.SpigotConfig.disableStatSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         try {
             FileUtil.createDirectoriesSafe(this.file.getParent());
 
@@ -104,6 +105,7 @@ public class ServerStatsCounter extends StatsCounter {
     @Override
     public void setValue(Player player, Stat<?> stat, int value) {
         if (org.spigotmc.SpigotConfig.disableStatSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         if (stat.getType() == Stats.CUSTOM && stat.getValue() instanceof final net.minecraft.resources.Identifier key && org.spigotmc.SpigotConfig.forcedStats.get(key) != null) return; // Paper - disable saving forced stats
         super.setValue(player, stat, value);
         this.dirty.add(stat);
diff --git a/net/minecraft/world/level/storage/PlayerDataStorage.java b/net/minecraft/world/level/storage/PlayerDataStorage.java
index 4099f4ebf76c6bf74384aa5b9f5e889d53ebcfa9..1d5666b7b5167b607b13874830fa7b33f22c945a 100644
--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -33,6 +33,7 @@ public class PlayerDataStorage {
 
     public void save(Player player) {
         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
+        if (org.dreeam.leaf.config.modules.misc.RegionFormatConfig.isReadOnlyMode()) return; // Leaf - Add read-only mode for Linear v2
         // Leaf start - Async playerdata saving
         CompoundTag compoundTag;
         try (ProblemReporter.ScopedCollector scopedCollector = new ProblemReporter.ScopedCollector(player.problemPath(), LOGGER)) {
