From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: vhs03 <vhs03.vhs03@outlook.de>
Date: Wed, 10 Jun 2026 23:46:16 +0200
Subject: [PATCH] configurable shield blocking delay


diff --git a/net/minecraft/world/item/Items.java b/net/minecraft/world/item/Items.java
index 310e61a9bbece157d81095130ecb3131deef9288..6942bd422dc1941fe0a2689e0673a407e1fb46fe 100644
--- a/net/minecraft/world/item/Items.java
+++ b/net/minecraft/world/item/Items.java
@@ -1958,7 +1958,7 @@ public class Items {
             .component(
                 DataComponents.BLOCKS_ATTACKS,
                 new BlocksAttacks(
-                    0.25F,
+                    org.dreeam.leaf.config.modules.gameplay.ShieldBlockingDelay.shieldBlockingDelay / 20.0F,
                     1.0F,
                     List.of(new BlocksAttacks.DamageReduction(90.0F, Optional.empty(), 0.0F, 1.0F)),
                     new BlocksAttacks.ItemDamageFunction(3.0F, 1.0F, 1.0F),
