--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -185,6 +_,8 @@
         public CompressionFormat compressionFormat = CompressionFormat.ZLIB;
         @Comment("This setting controls if equipment should be updated when handling certain player actions.")
         public boolean updateEquipmentOnPlayerActions = true;
+        @Comment("Only checks an item's amount and type instead of its full data during inventory desync checks.")
+        public boolean simplifyRemoteItemMatching = false;
 
         public enum CompressionFormat {
             GZIP,
