From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 21 May 2024 11:43:49 +0800
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index bdc8d830f54e4567ec0b03041221fdcd016ce76c..957f15a91f6aa1235c12b82079e32bb42449c78d 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +25,13 @@ public interface ServerBuildInfo {
     Key BRAND_GALE_ID = Key.key("galemc", "gale");
     // Gale end - branding changes
 
+    // Leaf start
+    /**
+     * The brand id for Leaf.
+     */
+    Key BRAND_LEAF_ID = Key.key("winds-studio", "leaf");
+    // Leaf end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
