From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 11 Aug 2022 04:13:27 +0800
Subject: [PATCH] Leaf config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2204336d8800311b65e894739ab1b27273e7c6f2..39808c0cc23a97f707366378f28ba64b80e6acbb 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1971,6 +1971,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Leaf start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLeafConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Leaf end
+
         /**
          * Sends the component to the player
          *
