From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Sun, 12 Jun 2022 10:57:51 -0500
Subject: [PATCH] Leaf config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1ba84f8db9d279f5fb2b4be10c02919bb27de30a..f304f40d69b2b6059d57115ee4a043ceaac0632f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1934,6 +1934,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Leaf start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLeafConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Leaf end
+
         /**
          * Sends the component to the player
          *
