From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 29 Oct 2022 02:35:35 -0400
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index 8e64ccdb928bc38cb703e60c8ec31339a6f6d894..848e2f7d1695a87b1227e38ef9a8de2559881211 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -23,15 +23,15 @@ configurations.api {
 
 dependencies {
     // api dependencies are listed transitively to API consumers
-    api("com.google.guava:guava:31.0.1-jre")
-    api("com.google.code.gson:gson:2.8.9")
+    api("com.google.guava:guava:31.1-jre")
+    api("com.google.code.gson:gson:2.10")
     api("net.md-5:bungeecord-chat:1.16-R0.4-deprecated+build.6") // Paper
     api("org.yaml:snakeyaml:1.33")
     // Paper start
     api("com.googlecode.json-simple:json-simple:1.1.1") {
         isTransitive = false // includes junit
     }
-    api("it.unimi.dsi:fastutil:8.5.6")
+    api("it.unimi.dsi:fastutil:8.5.9")
     apiAndDocs(platform("net.kyori:adventure-bom:$adventureVersion"))
     apiAndDocs("net.kyori:adventure-api")
     apiAndDocs("net.kyori:adventure-text-minimessage")
@@ -39,18 +39,18 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-legacy")
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
-    api("org.apache.logging.log4j:log4j-api:2.17.1")
-    api("org.slf4j:slf4j-api:1.8.0-beta4")
-    api("io.sentry:sentry:5.4.0") // Pufferfish
+    api("org.apache.logging.log4j:log4j-api:2.19.0")
+    api("org.slf4j:slf4j-api:1.8.0-beta4") // Leaf TODO - Bump later, need to fix breaking compatibility
+    api("io.sentry:sentry:6.8.0") // Pufferfish
 
-    implementation("org.ow2.asm:asm:9.2")
-    implementation("org.ow2.asm:asm-commons:9.2")
+    implementation("org.ow2.asm:asm:9.4")
+    implementation("org.ow2.asm:asm-commons:9.4")
     // Paper end
 
     compileOnly("org.apache.maven:maven-resolver-provider:3.8.5")
-    compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
-    compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
-    compileOnly("com.google.code.findbugs:jsr305:1.3.9") // Paper
+    compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.8.2")
+    compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.8.2")
+    compileOnly("com.google.code.findbugs:jsr305:3.0.2") // Paper
 
     val annotations = "org.jetbrains:annotations:23.0.0" // Paper - we don't want Java 5 annotations...
     compileOnly(annotations)
@@ -64,8 +64,8 @@ dependencies {
 
     testImplementation("org.apache.commons:commons-lang3:3.12.0")
     testImplementation("junit:junit:4.13.2")
-    testImplementation("org.hamcrest:hamcrest-library:1.3")
-    testImplementation("org.ow2.asm:asm-tree:9.3")
+    testImplementation("org.hamcrest:hamcrest-library:2.2")
+    testImplementation("org.ow2.asm:asm-tree:9.4")
 }
 
 configure<PublishingExtension> {
@@ -149,6 +149,9 @@ val scanJar = tasks.register("scanJarForBadCalls", io.papermc.paperweight.tasks.
     jarToScan.set(tasks.jar.flatMap { it.archiveFile })
     classpath.from(configurations.compileClasspath)
 }
+repositories {
+    mavenCentral()
+}
 tasks.check {
     dependsOn(scanJar)
 }
