From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 11 Aug 2022 04:13:27 +0800
Subject: [PATCH] Leaf config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 007e23a9383ab8eda12c6dffb385256215356040..c890e9655551865f7cea3c92782ee6a09e90f88e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2047,6 +2047,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
 
         // Paper end
 
+        // Leaf start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLeafConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Leaf end
+
         /**
          * Sends the component to the player
          *
