From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 28 Oct 2022 17:24:16 -0400
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index 8af9efe2dbc9563a2c327002cf05c5d2fd8fc68b..d784da2b478862070956e52884538946bd7562fe 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -24,15 +24,17 @@ configurations.api {
 dependencies {
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:31.1-jre")
-    api("com.google.code.gson:gson:2.10")
-    api("net.md-5:bungeecord-chat:1.16-R0.4-deprecated+build.9") // Paper
+    // Leaf start - Bump Dependencies
+    api("com.google.code.gson:gson:2.10.1")
+    api("net.md-5:bungeecord-chat:1.19-R0.1-SNAPSHOT") // Paper
+    // Leaf end
     api("org.yaml:snakeyaml:1.33")
     api("org.joml:joml:1.10.5")
     // Paper start
     api("com.googlecode.json-simple:json-simple:1.1.1") {
         isTransitive = false // includes junit
     }
-    api("it.unimi.dsi:fastutil:8.5.6")
+    api("it.unimi.dsi:fastutil:8.5.12") // Leaf - Bump Dependencies
     apiAndDocs(platform("net.kyori:adventure-bom:$adventureVersion"))
     apiAndDocs("net.kyori:adventure-api")
     apiAndDocs("net.kyori:adventure-text-minimessage")
@@ -40,32 +42,34 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-legacy")
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
-    api("org.apache.logging.log4j:log4j-api:2.17.1")
-    api("org.slf4j:slf4j-api:1.8.0-beta4")
+    // Leaf start - Bump Dependencies
+    api("org.apache.logging.log4j:log4j-api:2.20.0")
+    api("org.slf4j:slf4j-api:2.0.6")
 
     implementation("org.ow2.asm:asm:9.4")
     implementation("org.ow2.asm:asm-commons:9.4")
     // Paper end
 
-    api("org.apache.maven:maven-resolver-provider:3.8.5") // Paper, expose
-    compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
-    compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
-    compileOnly("com.google.code.findbugs:jsr305:1.3.9") // Paper
+    compileOnly("org.apache.maven:maven-resolver-provider:3.9.0")
+    compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.7")
+    compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.7")
+    compileOnly("com.google.code.findbugs:jsr305:3.0.2") // Paper
 
-    val annotations = "org.jetbrains:annotations:23.0.0" // Paper - we don't want Java 5 annotations...
+    val annotations = "org.jetbrains:annotations:24.0.1" // Paper - we don't want Java 5 annotations...
     compileOnly(annotations)
     testCompileOnly(annotations)
 
     // Paper start - add checker
-    val checkerQual = "org.checkerframework:checker-qual:3.21.0"
+    val checkerQual = "org.checkerframework:checker-qual:3.32.0"
     compileOnlyApi(checkerQual)
     testCompileOnly(checkerQual)
     // Paper end
-    testImplementation("org.mockito:mockito-core:4.9.0") // Paper - add mockito
+    testImplementation("org.mockito:mockito-core:5.2.0") // Paper - add mockito
 
     testImplementation("org.apache.commons:commons-lang3:3.12.0")
     testImplementation("junit:junit:4.13.2")
-    testImplementation("org.hamcrest:hamcrest-library:1.3")
+    testImplementation("org.hamcrest:hamcrest-library:2.2")
+    // Leaf end
     testImplementation("org.ow2.asm:asm-tree:9.4")
 }
 
@@ -112,9 +116,11 @@ tasks.withType<Javadoc> {
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
-        "https://guava.dev/releases/31.0.1-jre/api/docs/",
+        // Leaf - Bump Dependencies
+        "https://guava.dev/releases/31.1-jre/api/docs/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/1.33/",
-        "https://javadoc.io/doc/org.jetbrains/annotations/23.0.0/", // Paper - we don't want Java 5 annotations
+        "https://javadoc.io/doc/org.jetbrains/annotations/24.0.1/", // Paper - we don't want Java 5 annotations
+        // Leaf end
         // Paper start
         //"https://javadoc.io/doc/net.md-5/bungeecord-chat/1.16-R0.4/", // don't link to bungee chat
         "https://jd.advntr.dev/api/$adventureVersion/",
@@ -154,6 +160,9 @@ val scanJar = tasks.register("scanJarForBadCalls", io.papermc.paperweight.tasks.
     jarToScan.set(tasks.jar.flatMap { it.archiveFile })
     classpath.from(configurations.compileClasspath)
 }
+repositories {
+    mavenCentral()
+}
 tasks.check {
     dependsOn(scanJar)
 }
