From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 11 Aug 2022 04:13:27 +0800
Subject: [PATCH] Leaf config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 23d80bdb7e082b0921ca7b75a43a2ccc0381bc2d..415437e57e4d5a464bdcd30dff8970c5f7b08085 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2072,6 +2072,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Leaf start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLeafConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Leaf end
+
         /**
          * Sends the component to the player
          *
