From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 4 Dec 2022 15:58:51 -0500
Subject: [PATCH] Paper Upstream


diff --git a/src/main/java/com/destroystokyo/paper/MaterialTags.java b/src/main/java/com/destroystokyo/paper/MaterialTags.java
index f5ae132cef01666bacda11422c0b73dd585fea27..da47931bdb3b9a9257e670804aa3add5b4fd081f 100644
--- a/src/main/java/com/destroystokyo/paper/MaterialTags.java
+++ b/src/main/java/com/destroystokyo/paper/MaterialTags.java
@@ -507,6 +507,12 @@ public class MaterialTags {
         .endsWith("_BOOTS")
         .ensureSize("BOOTS", 6).lock();
 
+    /**
+     * Covers all variants of armor.
+     */
+    public static final MaterialSetTag ARMOR = new MaterialSetTag(keyFor("armor")).add(HELMETS, CHESTPLATES, LEGGINGS, BOOTS)
+            .ensureSize("ARMOR", 25).lock();
+
     /**
      * Covers the variants of bows.
      */
diff --git a/src/main/java/org/bukkit/WorldCreator.java b/src/main/java/org/bukkit/WorldCreator.java
index 355f9f27d29c65efebf099a72c37892a309edef1..dc92ce261b9e67979da3f82cb75292ff980ad87e 100644
--- a/src/main/java/org/bukkit/WorldCreator.java
+++ b/src/main/java/org/bukkit/WorldCreator.java
@@ -592,10 +592,13 @@ public class WorldCreator {
      * configuration, otherwise, will act as an override towards this setting
      *
      * @param keepSpawnLoaded the new value
+     * @return This object, for chaining
      */
-    public void keepSpawnLoaded(@NotNull net.kyori.adventure.util.TriState keepSpawnLoaded) {
+    @NotNull
+    public WorldCreator keepSpawnLoaded(@NotNull net.kyori.adventure.util.TriState keepSpawnLoaded) {
         java.util.Objects.requireNonNull(keepSpawnLoaded, "keepSpawnLoaded");
         this.keepSpawnLoaded = keepSpawnLoaded;
+        return this;
     }
 
     // Paper end
