From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 17 Jul 2023 07:45:39 +0800
Subject: [PATCH] Revert Purpur Fire Immunity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 98de85d1382fe84cdc2e2c9db04bf1b4f157291c..52867495d0f746ff40e802c4f1018511e58fd03e 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1086,18 +1086,5 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @return True if in daylight
      */
     boolean isInDaylight();
-
-    /**
-     * Checks if the entity is fire immune
-     *
-     * @return True if fire immune
-     */
-    boolean isImmuneToFire();
-
-    /**
-     * Sets if the entity is fire immune
-     * Set this to null to restore the entity type default
-     */
-    void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end
 }
