From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 17 Jul 2023 07:45:39 +0800
Subject: [PATCH] Revert Purpur Fire Immunity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index af421c1e9ad155e190dc3fe67ed24e96857d9e33..964ec7bd70fcdfcffd51b310cdea84c7c5db6c95 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1002,18 +1002,5 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @return True if in daylight
      */
     boolean isInDaylight();
-
-    /**
-     * Checks if the entity is fire immune
-     *
-     * @return True if fire immune
-     */
-    boolean isImmuneToFire();
-
-    /**
-     * Sets if the entity is fire immune
-     * Set this to null to restore the entity type default
-     */
-    void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end
 }
