From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam-qwq <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 1 Jul 2022 18:01:01 +0800
Subject: [PATCH] Remove Mojang username check


diff --git a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 38ac9aebd3f0687c57d9ed8186cb90e489655d47..b9335a9fb0998c1c7e7e99ebd7a15aa6ad60a408 100644
--- a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -278,7 +278,8 @@ public class ServerLoginPacketListenerImpl implements TickablePacketListener, Se
     @Override
     public void handleHello(ServerboundHelloPacket packet) {
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.HELLO, "Unexpected hello packet", new Object[0]);
-        Validate.validState(ServerLoginPacketListenerImpl.isValidUsername(packet.name()), "Invalid characters in username", new Object[0]);
+        // Leaf start - Remove Mojang's username check
+        // Leaf end - Remove Mojang's username check
         // Paper start - validate usernames
         if (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode() && io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.performUsernameValidation) {
             if (!this.iKnowThisMayNotBeTheBestIdeaButPleaseDisableUsernameValidation && !validateUsername(packet.name())) {
