From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam-qwq <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 1 Jul 2022 18:01:01 +0800
Subject: [PATCH] Remove Mojang username check


diff --git a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 855eac5e6f3a99a2fc31ad0e02c496d4b5c60622..8e55795d8ce05707460a1383104158dd0e22a737 100644
--- a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -270,7 +270,8 @@ public class ServerLoginPacketListenerImpl implements TickablePacketListener, Se
     @Override
     public void handleHello(ServerboundHelloPacket packet) {
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.HELLO, "Unexpected hello packet", new Object[0]);
-        Validate.validState(ServerLoginPacketListenerImpl.isValidUsername(packet.name()), "Invalid characters in username", new Object[0]);
+        // Leaf start - Remove Mojang's username check
+        // Leaf end
         // Paper start - validate usernames
         if (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode() && io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.performUsernameValidation) {
             if (!this.iKnowThisMayNotBeTheBestIdeaButPleaseDisableUsernameValidation && !validateUsername(packet.name())) {
