From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 28 Oct 2022 17:48:45 -0400
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index c745122af630d0f79bed7f0c03baa66eeb2e5da0..6ffb4052614f7b1a8fa3630d1d543d22d7981541 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -17,7 +17,7 @@ dependencies {
     // Gale end - project setup
     implementation("io.projectreactor.tools:blockhound:1.0.7.RELEASE") // Gale - base thread pool - watch for blocking base threads
     // Paper start
-    implementation("org.jline:jline-terminal-jansi:3.21.0")
+    implementation("org.jline:jline-terminal-jansi:3.22.0") // Leaf - Bump Dependencies
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     /*
           Required to add the missing Log4j2Plugins.dat file from log4j-core
@@ -25,29 +25,31 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - implementation
-    annotationProcessor("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - Needed to generate meta for our Log4j plugins
-    implementation("io.netty:netty-codec-haproxy:4.1.87.Final") // Paper - Add support for proxy protocol
+    // Leaf start - Bump Dependencies
+    implementation("org.apache.logging.log4j:log4j-core:2.20.0") // Paper - implementation
+    annotationProcessor("org.apache.logging.log4j:log4j-core:2.20.0") // Paper - Needed to generate meta for our Log4j plugins // Leaf - Bump Dependencies
+    implementation("io.netty:netty-codec-haproxy:4.1.89.Final") // Paper - Add support for proxy protocol // Leaf - Bump Dependencies
     // Paper end
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.19.0") // Paper - remove exclusion
-    implementation("org.ow2.asm:asm:9.3")
-    implementation("org.ow2.asm:asm-commons:9.3") // Paper - ASM event executor generation
-    testImplementation("org.mockito:mockito-core:4.9.0") // Paper - switch to mockito
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.20.0") // Paper - remove exclusion
+    implementation("org.ow2.asm:asm:9.4")
+    implementation("org.ow2.asm:asm-commons:9.4") // Paper - ASM event executor generation
+    // Leaf end
+    testImplementation("org.mockito:mockito-core:5.1.1") // Paper - switch to mockito
     implementation("org.spongepowered:configurate-yaml:4.1.2") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
     implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
     // Paper start - Use Velocity cipher
-    implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.2.0-SNAPSHOT") { // Leaf - Bump Dependencies
         isTransitive = false
     }
     // Paper end
-    runtimeOnly("org.xerial:sqlite-jdbc:3.36.0.3")
-    runtimeOnly("mysql:mysql-connector-java:8.0.29")
+    runtimeOnly("org.xerial:sqlite-jdbc:3.41.0.0") // Leaf - Bump Dependencies
+    runtimeOnly("mysql:mysql-connector-java:8.0.32") // Leaf - Bump Dependencies
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
 
-    runtimeOnly("org.apache.maven:maven-resolver-provider:3.8.5")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
+    runtimeOnly("org.apache.maven:maven-resolver-provider:3.8.7") // Leaf TODO - Bump to 4.0.0 later, need to fix breaking compatibility
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3") // Leaf TODO - Bump later, need to fix breaking compatibility
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3") // Leaf TODO - Bump later, need to fix breaking compatibility
 
     // Pufferfish start
     implementation("org.yaml:snakeyaml:1.33")
@@ -56,11 +58,11 @@ dependencies {
     }
     // Pufferfish end
 
-    testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
+    testImplementation("io.github.classgraph:classgraph:4.8.154") // Paper - mob goal test // Leaf - Bump Dependencies
     testImplementation("junit:junit:4.13.2")
-    testImplementation("org.hamcrest:hamcrest-library:1.3")
+    testImplementation("org.hamcrest:hamcrest-library:2.2") // Leaf - Bump Dependencies
 
-    implementation("io.netty:netty-all:4.1.87.Final"); // Paper - Bump netty
+    implementation("io.netty:netty-all:4.1.89.Final"); // Paper - Bump netty // Leaf - Bump Dependencies
 }
 
 val craftbukkitPackageVersion = "1_19_R2" // Paper
@@ -228,3 +230,6 @@ tasks.registerRunTask("runDev") {
     description = "Spin up a non-relocated Mojang-mapped test server"
     classpath(sourceSets.main.map { it.runtimeClasspath })
 }
+repositories {
+    mavenCentral()
+}
