From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam-qwq <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 1 Jul 2022 18:06:03 +0800
Subject: [PATCH] Remove Spigot check for broken BungeeCord configurations


diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index 08b21f5a7a07f44f8044f56991fb6723cd8f3eea..a8806d9235e125bab8b0cf223c253a97e3b53e19 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -138,12 +138,9 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                         {
                             connection.spoofedProfile = ServerHandshakePacketListenerImpl.gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
                         }
-                    } else if ( ( split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) {
-                        Component chatmessage = Component.literal("Unknown data in login hostname, did you forget to enable BungeeCord in spigot.yml?");
-                        this.connection.send(new ClientboundLoginDisconnectPacket(chatmessage));
-                        this.connection.disconnect(chatmessage);
-                        return;
                     }
+                    // Leaf start - Remove Spigot check for broken BungeeCord configurations
+                    // Leaf end - Remove Spigot check for broken BungeeCord configurations
                     // Spigot End
                     ((ServerLoginPacketListenerImpl) this.connection.getPacketListener()).hostname = packet.hostName + ":" + packet.port; // CraftBukkit - set hostname
                 }
