From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam-qwq <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 1 Jul 2022 18:06:03 +0800
Subject: [PATCH] Remove Spigot check for broken BungeeCord configurations


diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index e7ff7ad3bf4dd17fdd34202ec3aef8e9512bc36d..7274af2cd668206bb2dbcc5ca5bc2b700d2e7ae9 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -139,12 +139,9 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                         {
                             connection.spoofedProfile = ServerHandshakePacketListenerImpl.gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
                         }
-                    } else if ( ( split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) {
-                        Component chatmessage = Component.literal("Unknown data in login hostname, did you forget to enable BungeeCord in spigot.yml?");
-                        this.connection.send(new ClientboundLoginDisconnectPacket(chatmessage));
-                        this.connection.disconnect(chatmessage);
-                        return;
                     }
+                    // Leaf start - Remove Spigot check for broken BungeeCord configurations
+                    // Leaf end - Remove Spigot check for broken BungeeCord configurations
                     // Spigot End
                 }
                 break;
