From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 31 Jul 2024 22:05:21 +0800
Subject: [PATCH] Leaf Bootstrap

org.bukkit.craftbukkit.Main#main -> LeafBootstrap -> PaperBootstrap -> ...

diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index c1e88c31910e96ef07cece05046c0b55e708b52d..c63dfc0b740ce2e544ad920188d7d7dfc2ae3e04 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -278,7 +278,8 @@ public class Main {
 
                 //System.out.println("Loading libraries, please wait...");
                 //net.minecraft.server.Main.main(options);
-                io.papermc.paper.PaperBootstrap.boot(options);
+                //io.papermc.paper.PaperBootstrap.boot(options); // Leaf - LeafBoostrap - diff on change
+                org.dreeam.leaf.LeafBootstrap.boot(options); // Leaf - LeafBoostrap
             } catch (Throwable t) {
                 t.printStackTrace();
             }
diff --git a/src/main/java/org/dreeam/leaf/LeafBootstrap.java b/src/main/java/org/dreeam/leaf/LeafBootstrap.java
new file mode 100644
index 0000000000000000000000000000000000000000..316654051b80ac0fd62cf3b7a0e1b91010ec24b7
--- /dev/null
+++ b/src/main/java/org/dreeam/leaf/LeafBootstrap.java
@@ -0,0 +1,16 @@
+package org.dreeam.leaf;
+
+import io.papermc.paper.PaperBootstrap;
+import joptsimple.OptionSet;
+
+public class LeafBootstrap {
+
+    public static void boot(final OptionSet options) {
+        runPreBootTasks();
+
+        PaperBootstrap.boot(options);
+    }
+
+    private static void runPreBootTasks() {
+    }
+}
