From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BillyGalbreath <blake.galbreath@gmail.com>
Date: Fri, 16 Sep 2022 01:10:40 -0400
Subject: [PATCH] log suppression for sent expired chat


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index d3dd458dc1d3eb7f426565e38d63bb26c48cec4b..b7c299c1e164346516d70e481e472a4ebfad99ac 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2604,10 +2604,8 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
             }
         }
 
-        if (playerchatmessage.hasExpiredServer(Instant.now())) {
-            ServerGamePacketListenerImpl.LOGGER.warn("{} sent expired chat: '{}'. Is the client/server system time unsynchronized? c: {} s: {}", this.player.getName().getString(), playerchatmessage.signedContent().plain(), playerchatmessage.timeStamp().getEpochSecond(), Instant.now().getEpochSecond()); // Paper
-        }
-
+        // Purpur start - log suppression for sent expired chat
+        // Purpur end
         return true;
     }
 
