From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 12 Oct 2022 14:36:58 -0400
Subject: [PATCH] Remove Mojang username check


diff --git a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 3d505329642f4483b55c0f93110c1857bc4bcad1..b6309029f2ae8509303a47c8d650b88cba24fd29 100644
--- a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -36,6 +36,7 @@ import net.minecraft.util.Crypt;
 import net.minecraft.util.CryptException;
 import net.minecraft.world.entity.player.Player;
 import org.apache.commons.lang3.Validate;
+import org.dreeam.leaf.LeafConfig;
 import org.slf4j.Logger;
 import org.bukkit.craftbukkit.util.Waitable;
 import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
@@ -255,7 +256,9 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener
     public void handleHello(ServerboundHelloPacket packet) {
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.HELLO, "Unexpected hello packet", me.titaniumtown.Constants.EMPTY_object_arr); // JettPack
         this.gameProfile = packet.getGameProfile();
-        Validate.validState(ServerLoginPacketListenerImpl.isValidUsername(this.gameProfile.getName()), "Invalid characters in username", me.titaniumtown.Constants.EMPTY_object_arr); // JettPack
+        // Leaf start - Remove Mojang's username check
+        if (!LeafConfig.removeMojangUsernameCheck) Validate.validState(ServerLoginPacketListenerImpl.isValidUsername(this.gameProfile.getName()), "Invalid characters in username", me.titaniumtown.Constants.EMPTY_object_arr); // JettPack
+        // Leaf end - Remove Mojang's username check
         // Paper start - validate usernames
         if (com.destroystokyo.paper.PaperConfig.isProxyOnlineMode() && com.destroystokyo.paper.PaperConfig.performUsernameValidation) {
             if (!this.iKnowThisMayNotBeTheBestIdeaButPleaseDisableUsernameValidation && !validateUsername(this.gameProfile.getName())) {
diff --git a/src/main/java/org/dreeam/leaf/LeafConfig.java b/src/main/java/org/dreeam/leaf/LeafConfig.java
index 2e6cc6ed30fc2ba0913913bd19bf1b83c40fcce5..ba3b9d101763b3885d3552cb579ab906b5d21ecc 100644
--- a/src/main/java/org/dreeam/leaf/LeafConfig.java
+++ b/src/main/java/org/dreeam/leaf/LeafConfig.java
@@ -131,4 +131,12 @@ public class LeafConfig {
         ensureDefault(key, defaultValue, comment);
         return config.getStringList(key);
     }
+
+    public static boolean removeMojangUsernameCheck;
+
+    private static void removeconfig() {
+        removeMojangUsernameCheck = getBoolean("remove-Mojang-username-check", true,
+                "Remove username check of Mojang",
+                "enabling all characters as username");
+    }
 }
