From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 12 Oct 2022 14:48:45 -0400
Subject: [PATCH] Remove Spigot Check for Broken BungeeCord Configurations


diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index e8a79277f3e57bc82d67ada6543c7f8665311bd3..db19a38aa5f3aa27f966a5497b3176b3134f1d06 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -10,8 +10,8 @@ import net.minecraft.network.protocol.handshake.ClientIntentionPacket;
 import net.minecraft.network.protocol.handshake.ServerHandshakePacketListener;
 import net.minecraft.network.protocol.login.ClientboundLoginDisconnectPacket;
 import net.minecraft.server.MinecraftServer;
+import org.dreeam.leaf.LeafConfig;
 
-// CraftBukkit start
 import java.net.InetAddress;
 import java.util.HashMap;
 // CraftBukkit end
@@ -139,7 +139,9 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                         {
                             connection.spoofedProfile = ServerHandshakePacketListenerImpl.gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
                         }
-                    } else if ( ( split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) {
+                        // Leaf start - Remove Spigot check for broken BungeeCord configurations
+                    } else if ((split.length == 3 || split.length == 4) && (ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher(split[1]).matches()) && !(LeafConfig.removeSpigotCheckBungeeConfig)) {
+                        // Leaf end - Remove Spigot check for broken BungeeCord configurations
                         TranslatableComponent chatmessage = new TranslatableComponent("Unknown data in login hostname, did you forget to enable BungeeCord in spigot.yml?");
                         this.connection.send(new ClientboundLoginDisconnectPacket(chatmessage));
                         this.connection.disconnect(chatmessage);
diff --git a/src/main/java/org/dreeam/leaf/LeafConfig.java b/src/main/java/org/dreeam/leaf/LeafConfig.java
index ba3b9d101763b3885d3552cb579ab906b5d21ecc..08737d56fa50d17b3231952d131a9fefe203103a 100644
--- a/src/main/java/org/dreeam/leaf/LeafConfig.java
+++ b/src/main/java/org/dreeam/leaf/LeafConfig.java
@@ -133,10 +133,12 @@ public class LeafConfig {
     }
 
     public static boolean removeMojangUsernameCheck;
+    public static boolean removeSpigotCheckBungeeConfig;
 
     private static void removeconfig() {
         removeMojangUsernameCheck = getBoolean("remove-Mojang-username-check", true,
                 "Remove username check of Mojang",
                 "enabling all characters as username");
+        removeSpigotCheckBungeeConfig = getBoolean("remove-Spigot-check-bungee-config", true);
     }
 }
