From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 29 Oct 2022 02:36:19 -0400
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index ffec1b240af1e453a46ef6e3c877a06424301816..d7da9801096ca9557497839157713870535d2ed2 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -26,12 +26,12 @@ dependencies {
      */
     implementation("org.apache.logging.log4j:log4j-core:2.17.1") // Paper - implementation
     annotationProcessor("org.apache.logging.log4j:log4j-core:2.17.1") // Paper - Needed to generate meta for our Log4j plugins
-    implementation("io.netty:netty-codec-haproxy:4.1.77.Final")
+    implementation("io.netty:netty-codec-haproxy:4.1.84.Final")
     // Paper end
     implementation("org.apache.logging.log4j:log4j-iostreams:2.17.1") // Paper
     implementation("org.apache.logging.log4j:log4j-slf4j18-impl:2.17.1") // Paper
-    implementation("org.ow2.asm:asm:9.3")
-    implementation("org.ow2.asm:asm-commons:9.3") // Paper - ASM event executor generation
+    implementation("org.ow2.asm:asm:9.4")
+    implementation("org.ow2.asm:asm-commons:9.4") // Paper - ASM event executor generation
     implementation("org.spongepowered:configurate-yaml:4.1.2") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
     implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
@@ -40,29 +40,29 @@ dependencies {
         isTransitive = false
     }
     // Paper end
-    runtimeOnly("org.xerial:sqlite-jdbc:3.36.0.3")
-    runtimeOnly("mysql:mysql-connector-java:8.0.29")
+    runtimeOnly("org.xerial:sqlite-jdbc:3.39.3.0")
+    runtimeOnly("mysql:mysql-connector-java:8.0.30")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
 
-    implementation("cat.inspiracio:rhino-js-engine:1.7.7.1") // Purpur
+    implementation("cat.inspiracio:rhino-js-engine:1.7.10") // Purpur
     implementation("dev.omega24:upnp4j:1.0") // Purpur
 
     runtimeOnly("org.apache.maven:maven-resolver-provider:3.8.5")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.8.2")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.8.2")
 
     // Pufferfish start
-    implementation("org.yaml:snakeyaml:1.32")
+    implementation("org.yaml:snakeyaml:1.33")
     implementation ("me.carleslc.Simple-YAML:Simple-Yaml:1.8.2") {
         exclude(group="org.yaml", module="snakeyaml")
     }
     // Pufferfish end
 
-    testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
+    testImplementation("io.github.classgraph:classgraph:4.8.149") // Paper - mob goal test
     testImplementation("junit:junit:4.13.2")
-    testImplementation("org.hamcrest:hamcrest-library:1.3")
+    testImplementation("org.hamcrest:hamcrest-library:2.2")
 
-    implementation("io.netty:netty-all:4.1.77.Final"); // Paper - Bump netty
+    implementation("io.netty:netty-all:4.1.84.Final"); // Paper - Bump netty
 }
 
 val craftbukkitPackageVersion = "1_19_R1" // Paper
@@ -238,3 +238,6 @@ tasks.registerRunTask("runDev") {
     description = "Spin up a non-relocated Mojang-mapped test server"
     classpath(sourceSets.main.map { it.runtimeClasspath })
 }
+repositories {
+    mavenCentral()
+}
