From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 28 Oct 2022 17:48:45 -0400
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index 586d7998918fdb8b1db3c433d10234b6d7943115..c068173108818b8b042e1c1f518cf60bbf25c9b3 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -11,13 +11,13 @@ dependencies {
     // Depend on own API
     implementation(project(":leaf-api")) // Leaf
     // Depend on Paper MojangAPI
-    implementation("io.papermc.paper:paper-mojangapi:1.19.3-R0.1-SNAPSHOT") {
+    implementation("io.papermc.paper:paper-mojangapi:1.19.4-R0.1-SNAPSHOT") { // Leaf - Bump Dependencies
         exclude("io.papermc.paper", "paper-api")
     }
     // Gale end - project setup
-    implementation("io.projectreactor.tools:blockhound:1.0.7.RELEASE") // Gale - base thread pool - watch for blocking base threads
+    implementation("io.projectreactor.tools:blockhound:1.0.8.RELEASE") // Gale - base thread pool - watch for blocking base threads // Leaf - Bump Dependencies
     // Paper start
-    implementation("org.jline:jline-terminal-jansi:3.21.0")
+    implementation("org.jline:jline-terminal-jansi:3.23.0") // Leaf - Bump Dependencies
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     /*
           Required to add the missing Log4j2Plugins.dat file from log4j-core
@@ -25,14 +25,16 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - implementation
-    annotationProcessor("org.apache.logging.log4j:log4j-core:2.14.1") // Paper - Needed to generate meta for our Log4j plugins
-    implementation("io.netty:netty-codec-haproxy:4.1.87.Final") // Paper - Add support for proxy protocol
+    // Leaf start - Bump Dependencies
+    implementation("org.apache.logging.log4j:log4j-core:2.20.0") // Paper - implementation
+    annotationProcessor("org.apache.logging.log4j:log4j-core:2.20.0") // Paper - Needed to generate meta for our Log4j plugins
+    implementation("io.netty:netty-codec-haproxy:4.1.91.Final") // Paper - Add support for proxy protocol
     // Paper end
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.19.0") // Paper - remove exclusion
-    implementation("org.ow2.asm:asm:9.4")
-    implementation("org.ow2.asm:asm-commons:9.4") // Paper - ASM event executor generation
-    testImplementation("org.mockito:mockito-core:4.9.0") // Paper - switch to mockito
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.20.0") // Paper - remove exclusion
+    implementation("org.ow2.asm:asm:9.5")
+    implementation("org.ow2.asm:asm-commons:9.5") // Paper - ASM event executor generation
+    testImplementation("org.mockito:mockito-core:5.2.0") // Paper - switch to mockito
+    // Leaf end
     implementation("org.spongepowered:configurate-yaml:4.1.2") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
     implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
@@ -40,7 +42,7 @@ dependencies {
     runtimeOnly("com.mysql:mysql-connector-j:8.0.32")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
     // Paper start - Use Velocity cipher
-    implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.2.0-SNAPSHOT") { // Leaf - Bump Dependencies
         isTransitive = false
     }
     // Paper end
@@ -49,22 +51,25 @@ dependencies {
     implementation("org.mozilla:rhino-engine:1.7.14") // Purpur
     implementation("dev.omega24:upnp4j:1.0") // Purpur
 
-    runtimeOnly("org.apache.maven:maven-resolver-provider:3.8.5")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
+    runtimeOnly("org.apache.maven:maven-resolver-provider:3.9.1")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.7")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.7")
+    // Leaf end
 
     // Pufferfish start
     implementation("org.yaml:snakeyaml:1.33")
-    implementation("me.carleslc.Simple-YAML:Simple-Yaml:1.8.3") {
+    // Leaf start - Bump Dependencies
+    implementation("me.carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
         exclude(group = "org.yaml", module = "snakeyaml")
     }
     // Pufferfish end
 
-    testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
+    testImplementation("io.github.classgraph:classgraph:4.8.157") // Paper - mob goal test
     testImplementation("junit:junit:4.13.2")
-    testImplementation("org.hamcrest:hamcrest-library:1.3")
+    testImplementation("org.hamcrest:hamcrest-library:2.2")
 
-    implementation("io.netty:netty-all:4.1.87.Final"); // Paper - Bump netty
+    implementation("io.netty:netty-all:4.1.91.Final"); // Paper - Bump netty
+    // Leaf end
 }
 
 val craftbukkitPackageVersion = "1_19_R3" // Paper
@@ -231,3 +236,6 @@ tasks.registerRunTask("runDev") {
     description = "Spin up a non-relocated Mojang-mapped test server"
     classpath(sourceSets.main.map { it.runtimeClasspath })
 }
+repositories {
+    mavenCentral()
+}
