From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 28 Oct 2022 17:48:45 -0400
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index ab5aee09874e2258d05b22ca1be3c273cb1afb34..238637b9bd22121124a4c4664ce1ac417a80bafa 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -22,7 +22,7 @@ dependencies {
     }
     // Gale end - project setup
     // Paper start
-    implementation("org.jline:jline-terminal-jansi:3.21.0")
+    implementation("org.jline:jline-terminal-jansi:3.24.1") // Leaf - Bump Dependencies
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     implementation("net.kyori:adventure-text-serializer-ansi:4.14.0") // Keep in sync with adventureVersion from Paper-API build file
     implementation("net.kyori:ansi:1.0.3") // Manually bump beyond above transitive dep
@@ -32,30 +32,34 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - implementation
-    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - Needed to generate meta for our Log4j plugins
+    // Leaf start - Bump Dependencies
+    implementation("org.apache.logging.log4j:log4j-core:2.22.0") // Paper - implementation
+    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.22.0") // Paper - Needed to generate meta for our Log4j plugins
     runtimeOnly(log4jPlugins.output)
     alsoShade(log4jPlugins.output)
-    implementation("io.netty:netty-codec-haproxy:4.1.97.Final") // Paper - Add support for proxy protocol
+    implementation("io.netty:netty-codec-haproxy:4.1.101.Final") // Paper - Add support for proxy protocol
     // Paper end
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.19.0") // Paper - remove exclusion
-    implementation("org.ow2.asm:asm:9.5")
-    implementation("org.ow2.asm:asm-commons:9.5") // Paper - ASM event executor generation
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.22.0") // Paper - remove exclusion
+    implementation("org.ow2.asm:asm:9.6")
+    implementation("org.ow2.asm:asm-commons:9.6") // Paper - ASM event executor generation
     implementation("org.spongepowered:configurate-yaml:4.2.0-SNAPSHOT") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
-    implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
-    runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.1")
-    runtimeOnly("com.mysql:mysql-connector-j:8.1.0")
-    runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
+    implementation("net.fabricmc:mapping-io:0.5.0") // Paper - needed to read mappings for stacktrace deobfuscation
+    runtimeOnly("org.xerial:sqlite-jdbc:3.44.0.0")
+    runtimeOnly("com.mysql:mysql-connector-j:8.2.0")
+    runtimeOnly("com.lmax:disruptor:3.4.4") // Paper // Leaf - Bump Dependencies - Waiting Log4j 3.x that supported disruptor 4.0.0
+    // Leaf end
     // Paper start - Use Velocity cipher
-    implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.2.0-SNAPSHOT") { // Leaf - Bump Dependencies
         isTransitive = false
     }
     // Paper end
 
-    runtimeOnly("org.apache.maven:maven-resolver-provider:3.8.5")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
+    // Leaf start - Bump Dependencies
+    runtimeOnly("org.apache.maven:maven-resolver-provider:3.9.5")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.17")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.17")
+    // Leaf end
 
     // Pufferfish start
     implementation("org.yaml:snakeyaml:2.2")
@@ -68,10 +72,14 @@ dependencies {
     implementation("org.mozilla:rhino-engine:1.7.14") // Purpur
     implementation("dev.omega24:upnp4j:1.0") // Purpur
 
-    testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
-    testImplementation("org.junit.jupiter:junit-jupiter:5.10.0")
+    // Leaf start - Bump Dependencies
+    testImplementation("io.github.classgraph:classgraph:4.8.165") // Paper - mob goal test
+    testImplementation("org.junit.jupiter:junit-jupiter:5.10.1")
     testImplementation("org.hamcrest:hamcrest:2.2")
-    testImplementation("org.mockito:mockito-core:5.5.0")
+    testImplementation("org.mockito:mockito-core:5.7.0")
+
+    implementation("io.netty:netty-all:4.1.101.Final")
+    // Leaf end
 }
 
 val craftbukkitPackageVersion = "1_20_R2" // Paper
@@ -263,3 +271,7 @@ sourceSets {
     }
 }
 // Gale end - package license into jar
+
+repositories {
+    mavenCentral()
+}
diff --git a/src/main/java/io/papermc/paper/util/ObfHelper.java b/src/main/java/io/papermc/paper/util/ObfHelper.java
index b8b17d046f836c8652ab094db00ab1af84971b2c..e856d7ce39e388f8bf45d394fd2fb586e3370c27 100644
--- a/src/main/java/io/papermc/paper/util/ObfHelper.java
+++ b/src/main/java/io/papermc/paper/util/ObfHelper.java
@@ -93,7 +93,7 @@ public enum ObfHelper {
                 return null;
             }
             final MemoryMappingTree tree = new MemoryMappingTree();
-            MappingReader.read(new InputStreamReader(mappingsInputStream, StandardCharsets.UTF_8), MappingFormat.TINY_2, tree);
+            MappingReader.read(new InputStreamReader(mappingsInputStream, StandardCharsets.UTF_8), MappingFormat.TINY_2_FILE, tree); // Leaf - Use mapping-io 0.5.0-beta.1
             final Set<ClassMapping> classes = new HashSet<>();
 
             final StringPool pool = new StringPool();
