From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 14 Jan 2024 05:14:09 -0500
Subject: [PATCH] Configurable server GUI name


diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 8f74c2ec5252b6265549589310d742337c91cb2c..8f951e3b6241d040b28d3c8b8e7d7f96ab4d2b8d 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -56,7 +56,7 @@ public class MinecraftServerGui extends JComponent {
             ;
         }
 
-        final JFrame jframe = new JFrame("Purpur Minecraft server"); // Purpur
+        final JFrame jframe = new JFrame(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName); // Purpur // Leaf - Configurable server GUI name
         final MinecraftServerGui servergui = new MinecraftServerGui(server);
 
         jframe.setDefaultCloseOperation(2);
@@ -64,7 +64,7 @@ public class MinecraftServerGui extends JComponent {
         jframe.pack();
         jframe.setLocationRelativeTo((Component) null);
         jframe.setVisible(true);
-        jframe.setName("Purpur Minecraft server"); // Paper - Improve ServerGUI // Purpur
+        jframe.setName(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName); // Paper - Improve ServerGUI // Purpur // Leaf - Configurable server GUI name
 
         // Paper start - Improve ServerGUI
         try {
@@ -76,7 +76,7 @@ public class MinecraftServerGui extends JComponent {
         jframe.addWindowListener(new WindowAdapter() {
             public void windowClosing(WindowEvent windowevent) {
                 if (!servergui.isClosing.getAndSet(true)) {
-                    jframe.setTitle("Purpur Minecraft server - shutting down!"); // Purpur
+                    jframe.setTitle(org.dreeam.leaf.config.modules.misc.ServerBrand.serverGUIName + " - shutting down!"); // Purpur // Leaf - Configurable server GUI name
                     server.halt(true);
                     servergui.runFinalizers();
                 }
diff --git a/src/main/java/org/dreeam/leaf/config/modules/misc/ServerBrand.java b/src/main/java/org/dreeam/leaf/config/modules/misc/ServerBrand.java
index 7f7fe2930b2e26dd9ef4ceb0a22b005a8cc477b6..5c608079850c5483943669a98e2ba7e0d5a03b23 100644
--- a/src/main/java/org/dreeam/leaf/config/modules/misc/ServerBrand.java
+++ b/src/main/java/org/dreeam/leaf/config/modules/misc/ServerBrand.java
@@ -18,4 +18,7 @@ public class ServerBrand implements IConfigModule {
 
     @ConfigInfo(baseName = "server-mod-name")
     public static String serverModName = io.papermc.paper.ServerBuildInfo.buildInfo().brandName();
+
+    @ConfigInfo(baseName = "server-gui-name")
+    public static String serverGUIName = "Leaf Console";
 }
