From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 28 Oct 2022 17:48:45 -0400
Subject: [PATCH] Bump Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index f1e1db576f08c1813bf9c7c7b216b3d91187a28f..2c7f11c51379826322015d72ab4d3ffe17c4676b 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -22,9 +22,9 @@ dependencies {
     }
     // Gale end - project setup
     // Paper start
-    implementation("org.jline:jline-terminal-jansi:3.21.0")
+    implementation("org.jline:jline-terminal-jansi:3.25.1") // Leaf - Bump Dependencies
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
-    implementation("net.kyori:adventure-text-serializer-ansi:4.14.0") // Keep in sync with adventureVersion from Paper-API build file
+    implementation("net.kyori:adventure-text-serializer-ansi:4.15.0") // Keep in sync with adventureVersion from Paper-API build file // Leaf - Bump Dependencies
     implementation("net.kyori:ansi:1.0.3") // Manually bump beyond above transitive dep
     /*
           Required to add the missing Log4j2Plugins.dat file from log4j-core
@@ -32,24 +32,26 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - implementation
-    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - Needed to generate meta for our Log4j plugins
+    // Leaf start - Bump Dependencies
+    implementation("org.apache.logging.log4j:log4j-core:2.22.1") // Paper - implementation
+    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.22.1") // Paper - Needed to generate meta for our Log4j plugins
     runtimeOnly(log4jPlugins.output)
     alsoShade(log4jPlugins.output)
-    implementation("io.netty:netty-codec-haproxy:4.1.97.Final") // Paper - Add support for proxy protocol
+    implementation("io.netty:netty-codec-haproxy:4.1.107.Final") // Paper - Add support for proxy protocol
     // Paper end
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.19.0") // Paper - remove exclusion
-    implementation("org.ow2.asm:asm-commons:9.5")
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.22.1") // Paper - remove exclusion
+    implementation("org.ow2.asm:asm-commons:9.6")
     implementation("org.spongepowered:configurate-yaml:4.2.0-SNAPSHOT") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
-    implementation("net.fabricmc:mapping-io:0.5.0") // Paper - needed to read mappings for stacktrace deobfuscation
-    runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.1")
-    runtimeOnly("com.mysql:mysql-connector-j:8.2.0")
-    runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
+    implementation("net.fabricmc:mapping-io:0.5.1") // Paper - needed to read mappings for stacktrace deobfuscation
+    runtimeOnly("org.xerial:sqlite-jdbc:3.45.1.0") // Leaf - Bump Dependencies
+    runtimeOnly("com.mysql:mysql-connector-j:8.3.0")
+    runtimeOnly("com.lmax:disruptor:3.4.4") // Paper // Leaf - Bump Dependencies - Waiting Log4j 3.x to support disruptor 4.0.0
     // Paper start - Use Velocity cipher
-    implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.3.0-SNAPSHOT") { // Leaf - Bump Dependencies
         isTransitive = false
     }
+    // Leaf end
     // Paper end - Use Velocity cipher
 
     runtimeOnly("org.apache.maven:maven-resolver-provider:3.9.6")
@@ -67,10 +69,14 @@ dependencies {
     implementation("org.mozilla:rhino-engine:1.7.14") // Purpur
     implementation("dev.omega24:upnp4j:1.0") // Purpur
 
-    testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
-    testImplementation("org.junit.jupiter:junit-jupiter:5.10.0")
+    // Leaf start - Bump Dependencies
+    testImplementation("io.github.classgraph:classgraph:4.8.165") // Paper - mob goal test
+    testImplementation("org.junit.jupiter:junit-jupiter:5.10.2")
     testImplementation("org.hamcrest:hamcrest:2.2")
-    testImplementation("org.mockito:mockito-core:5.5.0")
+    testImplementation("org.mockito:mockito-core:5.10.0")
+
+    implementation("io.netty:netty-all:4.1.107.Final")
+    // Leaf end
 }
 
 val craftbukkitPackageVersion = "1_20_R3" // Paper
@@ -262,3 +268,7 @@ sourceSets {
     }
 }
 // Gale end - package license into jar
+
+repositories {
+    mavenCentral()
+}
