From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 10 Jun 2023 11:28:24 +0800
Subject: [PATCH] Fix compile error


diff --git a/src/main/java/io/papermc/paper/plugin/provider/configuration/serializer/PermissionConfigurationSerializer.java b/src/main/java/io/papermc/paper/plugin/provider/configuration/serializer/PermissionConfigurationSerializer.java
index d1088e4b7fa5f8e689f23b150b83645ce1ae5a0e..f951f4024745503e9cdfa7ff17b9313ac6d7b4c4 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/configuration/serializer/PermissionConfigurationSerializer.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/configuration/serializer/PermissionConfigurationSerializer.java
@@ -39,7 +39,7 @@ public class PermissionConfigurationSerializer {
                     try {
                         result.add(Permission.loadPermission(entry.getKey().toString(), (Map<?, ?>) entry.getValue(), permissionDefault, result));
                     } catch (Throwable ex) {
-                        throw new SerializationException(null, "Error loading permission %s".formatted(entry.getKey()), ex);
+                        throw new SerializationException((Type) null, "Error loading permission %s".formatted(entry.getKey()), ex);
                     }
                 }
             }
