From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 9 Oct 2023 21:33:08 -0400
Subject: [PATCH] Faster Random for xaeroMapServerID generation


diff --git a/src/main/java/org/dreeam/leaf/LeafConfig.java b/src/main/java/org/dreeam/leaf/LeafConfig.java
index 245bf3d0c17e91c24ff28e53f0ee73b66b4d5e50..c9e57d0b97797aef9008cbdeeb4ab2a8bf51ea59 100644
--- a/src/main/java/org/dreeam/leaf/LeafConfig.java
+++ b/src/main/java/org/dreeam/leaf/LeafConfig.java
@@ -1,7 +1,7 @@
 package org.dreeam.leaf;
 
 import com.google.common.collect.ImmutableMap;
-import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
 import net.minecraft.server.MinecraftServer;
 import org.bukkit.Bukkit;
 import org.bukkit.configuration.ConfigurationSection;
@@ -206,7 +206,7 @@ public class LeafConfig {
     public static boolean jadeProtocol = false;
     public static boolean appleskinProtocol = false;
     public static boolean xaeroMapProtocol = false;
-    public static int xaeroMapServerID = new Random().nextInt();
+    public static int xaeroMapServerID = ThreadLocalRandom.current().nextInt(); // Leaf - Faster Random
     private static void network() {
         jadeProtocol = getBoolean("network.protocol.jade-protocol", jadeProtocol);
         appleskinProtocol = getBoolean("network.protocol.appleskin-protocol", appleskinProtocol);
