From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 22 Jul 2023 16:15:37 +0800
Subject: [PATCH] Optimize Vine Spread


diff --git a/src/main/java/net/minecraft/world/level/block/VineBlock.java b/src/main/java/net/minecraft/world/level/block/VineBlock.java
index 10b63e9a4253f5f572cbc717edc6322a98e9e0eb..629b9404683eeff298287568fa1538ddd12aa1bc 100644
--- a/src/main/java/net/minecraft/world/level/block/VineBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/VineBlock.java
@@ -295,11 +295,8 @@ public class VineBlock extends Block {
         boolean flag = true;
         Iterable<BlockPos> iterable = BlockPos.betweenClosed(pos.getX() - 4, pos.getY() - 1, pos.getZ() - 4, pos.getX() + 4, pos.getY() + 1, pos.getZ() + 4);
         int i = 5;
-        Iterator iterator = iterable.iterator();
-
-        while (iterator.hasNext()) {
-            BlockPos blockposition1 = (BlockPos) iterator.next();
 
+        for (BlockPos blockposition1 : iterable) { // Leaf - Optimize Vine Spread
             if (world.getBlockState(blockposition1).is((Block) this)) {
                 --i;
                 if (i <= 0) {
