From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 11 Mar 2023 04:28:34 -0500
Subject: [PATCH] =?UTF-8?q?=E2=80=9C=C3=A2Fix=20build=20system=20issues?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit


diff --git a/build.gradle.kts b/build.gradle.kts
index 3401058ba444cc18691cb5eb1e324a739ea14bc4..86812efc78480170a4a8d490b41f9a7e3c1cd29e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -199,15 +199,14 @@ fun TaskContainer.registerRunTask(
     systemProperty("gale.detect.thread.blocks", true) // Gale - base thread pool - watch for blocking base threads
 
     val memoryGb = providers.gradleProperty("paper.runMemoryGb").getOrElse("2")
-    val modifiedJvmArgs = jvmArgs ?: arrayListOf()
-    modifiedJvmArgs.addAll(listOf("-Xms${memoryGb}G", "-Xmx${memoryGb}G"))
+    minHeapSize = "${memoryGb}G"
+    maxHeapSize = "${memoryGb}G"
     // Gale start - enable virtual threads for development runs
-    modifiedJvmArgs.add("--enable-preview")
-    modifiedJvmArgs.addAll(listOf("--add-opens=java.base/java.lang=ALL-UNNAMED"))
+    args("--enable-preview")
+    args(listOf("--add-opens=java.base/java.lang=ALL-UNNAMED"))
     // Gale end - enable virtual threads for development runs
-    modifiedJvmArgs.add("--add-modules=jdk.incubator.vector") // Gale - Pufferfish - SIMD support
-    modifiedJvmArgs.add("-XX:+AllowRedefinitionToAddDeleteMethods") // Gale - base thread pool - watch for blocking base threads
-    jvmArgs = modifiedJvmArgs
+    args("--add-modules=jdk.incubator.vector") // Gale - Pufferfish - SIMD support
+    args("-XX:+AllowRedefinitionToAddDeleteMethods") // Gale - base thread pool - watch for blocking base threads
 
     doFirst {
         workingDir.mkdirs()
