From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 11 Mar 2023 04:28:34 -0500
Subject: [PATCH] Fix build system issues


diff --git a/build.gradle.kts b/build.gradle.kts
index 63a819196ddcad8106c7358172176c41031c70cd..3ef66e60ef5c8645ed6b88328a33ef332bb1ea27 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -198,14 +198,8 @@ fun TaskContainer.registerRunTask(
     minHeapSize = "${memoryGb}G"
     maxHeapSize = "${memoryGb}G"
     // Gale start - modify JVM args for development runs
-    val modifiedJvmArgs = ArrayList(jvmArgs) ?: arrayListOf()
-    // Gale start - enable virtual threads for development runs
-    modifiedJvmArgs.add("--enable-preview")
-    modifiedJvmArgs.addAll(listOf("--add-opens=java.base/java.lang=ALL-UNNAMED"))
-    // Gale end - enable virtual threads for development runs
-    modifiedJvmArgs.add("--add-modules=jdk.incubator.vector") // Gale - Pufferfish - SIMD support
-    modifiedJvmArgs.add("-XX:+AllowRedefinitionToAddDeleteMethods") // Gale - base thread pool - watch for blocking base threads
-    jvmArgs = modifiedJvmArgs
+    args("--add-modules=jdk.incubator.vector") // Gale - Pufferfish - SIMD support // Leaf - Fix build system issues
+    args("-XX:+AllowRedefinitionToAddDeleteMethods") // Gale - base thread pool - watch for blocking base threads // Leaf - Fix build system issues
     // Gale end - modify JVM args for development runs
 
     doFirst {
