From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 24 Oct 2023 19:50:01 -0400
Subject: [PATCH] Update method use


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java
index 6eb0e250101d10064fe150f03d98086afeba67ca..adff3b47b283998ae5137b1336b818a332bfd09d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftItem.java
@@ -155,42 +155,42 @@ public class CraftItem extends CraftEntity implements Item {
     // Purpur start
     @Override
     public void setImmuneToCactus(boolean immuneToCactus) {
-        item.immuneToCactus = immuneToCactus;
+        this.getHandle().immuneToCactus = immuneToCactus;
     }
 
     @Override
     public boolean isImmuneToCactus() {
-        return item.immuneToCactus;
+        return this.getHandle().immuneToCactus;
     }
 
     @Override
     public void setImmuneToExplosion(boolean immuneToExplosion) {
-        item.immuneToExplosion = immuneToExplosion;
+        this.getHandle().immuneToExplosion = immuneToExplosion;
     }
 
     @Override
     public boolean isImmuneToExplosion() {
-        return item.immuneToExplosion;
+        return this.getHandle().immuneToExplosion;
     }
 
     @Override
     public void setImmuneToFire(boolean immuneToFire) {
-        item.immuneToFire = immuneToFire;
+        this.getHandle().immuneToFire = immuneToFire;
     }
 
     @Override
     public boolean isImmuneToFire() {
-        return item.immuneToFire;
+        return this.getHandle().immuneToFire;
     }
 
     @Override
     public void setImmuneToLightning(boolean immuneToLightning) {
-        item.immuneToLightning = immuneToLightning;
+        this.getHandle().immuneToLightning = immuneToLightning;
     }
 
     @Override
     public boolean isImmuneToLightning() {
-        return item.immuneToLightning;
+        return this.getHandle().immuneToLightning;
     }
     // Purpur end
 }
