From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: pixelNightly <pixelnightly@proton.me>
Date: Mon, 4 Dec 2023 16:11:36 +0200
Subject: [PATCH] Fix sprint glitch


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index a37244bdd21c27d15e0ce6731c28f8797fdddf76..3c16a7d29454b23efeb36eb7fecae742249d64ac 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1439,7 +1439,8 @@ public abstract class LivingEntity extends Entity implements Attackable {
                 player.setRealHealth(health);
             }
 
-            player.updateScaledHealth(false);
+            //player.updateScaledHealth(false); // Leaf - Commented out to fix sprint glitch
+            this.entityData.set(LivingEntity.DATA_HEALTH_ID, player.getScaledHealth()); // Leaf - Sprint glitch fixed by pafias
             return;
         }
         // CraftBukkit end
