From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lilingfengdev <lilingfengvvv@outlook.com>
Date: Tue, 16 Jan 2024 19:37:01 +0800
Subject: [PATCH] Fix MC Bug


diff --git a/src/main/java/net/minecraft/world/entity/animal/Wolf.java b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
index 5d49ed7ddf44a3d549b178ae548664194967776b..8ff325a6f8e70be3df989296044d00931814a10a 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Wolf.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
@@ -655,14 +655,17 @@ public class Wolf extends TamableAnimal implements NeutralMob {
     public boolean canMate(Animal other) {
         if (other == this) {
             return false;
-        } else if (!this.isTame()) {
-            return false;
         } else if (!(other instanceof Wolf)) {
             return false;
         } else {
-            Wolf entitywolf = (Wolf) other;
-
-            return !entitywolf.isTame() ? false : (entitywolf.isInSittingPose() ? false : this.isInLove() && entitywolf.isInLove());
+            // DivineMC start - Fix MC-93018
+            Wolf wolf = (Wolf) other;
+            if (wolf.isInSittingPose()) {
+                return false;
+            } else {
+                return this.isInLove() && wolf.isInLove();
+            }
+            // DivineMC end
         }
     }
 
