From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lilingfengdev <lilingfengvvv@outlook.com>
Date: Tue, 16 Jan 2024 20:15:11 +0800
Subject: [PATCH] Fix MC-266334


diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index 1ea348b6007d8aebf1e0f29018a03eb10c445369..9fff19c3daf8112499246dd489a5f03842e8da5c 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -48,7 +48,9 @@ import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.entity.Player;
 import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
 // CraftBukkit end
-
+// Leaf start
+import org.dreeam.leaf.LeafConfig;
+// Leaf end
 public class ArmorStand extends LivingEntity {
 
     public static final int WOBBLE_TIME = 5;
@@ -514,7 +516,9 @@ public class ArmorStand extends LivingEntity {
                 } else {
                     boolean flag = source.is(DamageTypeTags.CAN_BREAK_ARMOR_STAND);
                     boolean flag1 = source.is(DamageTypeTags.ALWAYS_KILLS_ARMOR_STANDS);
-
+                    if (DivineConfig.fixMc266334 && !flag && source.getDirectEntity() instanceof net.minecraft.world.entity.projectile.ShulkerBullet){
+                        return false;
+                    }
                     if (!flag && !flag1) {
                         return false;
                     } else {
diff --git a/src/main/java/org/dreeam/leaf/LeafConfig.java b/src/main/java/org/dreeam/leaf/LeafConfig.java
index c51b54bcb96d06197d6265055e2e9e44858cd224..e4bfb584635fa497279fd6c14211a420b1ae1914 100644
--- a/src/main/java/org/dreeam/leaf/LeafConfig.java
+++ b/src/main/java/org/dreeam/leaf/LeafConfig.java
@@ -313,4 +313,9 @@ public class LeafConfig {
         extraHorizontal = getDouble("playerKnockback.extraHorizontal", extraHorizontal);
         extraVertical = getDouble("playerKnockback.extraVertical", extraVertical);
     }
+
+    public static boolean fixMc266334 = false;
+    private static void fixMc266334() {
+        fixMc266334 = getBoolean("bugfix.mc-266334", fixMc266334,"Fix for MC-266334");
+    }
 }
