From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lilingfengdev <lilingfengvvv@outlook.com>
Date: Tue, 16 Jan 2024 20:22:55 +0800
Subject: [PATCH] --amend


diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index 9fff19c3daf8112499246dd489a5f03842e8da5c..54bb5e19938f1f11f0b0729ff6435215ae7d609b 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -516,7 +516,7 @@ public class ArmorStand extends LivingEntity {
                 } else {
                     boolean flag = source.is(DamageTypeTags.CAN_BREAK_ARMOR_STAND);
                     boolean flag1 = source.is(DamageTypeTags.ALWAYS_KILLS_ARMOR_STANDS);
-                    if (DivineConfig.fixMc266334 && !flag && source.getDirectEntity() instanceof net.minecraft.world.entity.projectile.ShulkerBullet){
+                    if (LeafConfig.fixMc266334 && !flag && source.getDirectEntity() instanceof net.minecraft.world.entity.projectile.ShulkerBullet){
                         return false;
                     }
                     if (!flag && !flag1) {
