From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 19 Feb 2024 13:10:16 -0500
Subject: [PATCH] Redirect vanilla getProfiler in PathNavigationRegion


diff --git a/src/main/java/net/minecraft/world/level/PathNavigationRegion.java b/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
index a1a4b99167919bedb8a45c3b81889f58f2abfbcc..30d19ae31fe29436482b3fbe146dc6238b1fd04b 100644
--- a/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
+++ b/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
@@ -150,4 +150,10 @@ public class PathNavigationRegion implements BlockGetter, CollisionGetter {
     public int getHeight() {
         return this.level.getHeight();
     }
+
+    // Leaf start - Redirect to fix plugin incompatibility
+    public net.minecraft.util.profiling.ProfilerFiller getProfiler() {
+        return net.minecraft.util.profiling.InactiveProfiler.INSTANCE; // Gale - Purpur - remove vanilla profiler
+    }
+    // Leaf end - Redirect to fix plugin incompatibility
 }
