From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <fsjk947@gmail.com>
Date: Wed, 13 Mar 2024 03:33:08 +0800
Subject: [PATCH] Fix MC-65198

Mojang issues: https://bugs.mojang.com/browse/MC-65198

diff --git a/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java b/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java
index 7215af6cc91f48b040c23c54536d4aac8d293497..53002139c8ba4e4886cf73404c3bbd152421053a 100644
--- a/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -141,6 +141,7 @@ public abstract class ItemCombinerMenu extends AbstractContainerMenu {
             ItemStack itemstack1 = slot1.getItem();
 
             itemstack = itemstack1.copy();
+            ItemStack itemStack2 = itemstack.copy(); // Leaf - Fix MC-65198
             int j = this.getInventorySlotStart();
             int k = this.getUseRowEnd();
 
@@ -179,7 +180,7 @@ public abstract class ItemCombinerMenu extends AbstractContainerMenu {
             }
 
             this.activeQuickItem = itemstack; // Purpur
-            slot1.onTake(player, itemstack1);
+            slot1.onTake(player, itemStack2); // Leaf - Fix MC-65198
             this.activeQuickItem = null; // Purpur
         }
 
diff --git a/src/main/java/net/minecraft/world/inventory/ResultSlot.java b/src/main/java/net/minecraft/world/inventory/ResultSlot.java
index f17ff5988d826c8fad68f6bf7bac1d06edae01ae..02337174c26d7532c3ab99e269ec7ce8793b7f13 100644
--- a/src/main/java/net/minecraft/world/inventory/ResultSlot.java
+++ b/src/main/java/net/minecraft/world/inventory/ResultSlot.java
@@ -45,7 +45,7 @@ public class ResultSlot extends Slot {
     @Override
     protected void checkTakeAchievements(ItemStack stack) {
         if (this.removeCount > 0) {
-            stack.onCraftedBy(this.player.level(), this.player, this.removeCount);
+            stack.onCraftedBy(this.player.level(), this.player, stack.getCount()); // Leaf - Fix MC-65198
         }
 
         if (this.container instanceof RecipeCraftingHolder recipeCraftingHolder) {
diff --git a/src/main/java/net/minecraft/world/inventory/StonecutterMenu.java b/src/main/java/net/minecraft/world/inventory/StonecutterMenu.java
index eade15820dd9db38b6af2a5c4314acfb14ca03e9..5fef8a6bc746a34a0b742c51f169a502bc9b6ce6 100644
--- a/src/main/java/net/minecraft/world/inventory/StonecutterMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/StonecutterMenu.java
@@ -259,6 +259,7 @@ public class StonecutterMenu extends AbstractContainerMenu {
             Item item = itemstack1.getItem();
 
             itemstack = itemstack1.copy();
+            ItemStack itemStack2 = itemstack.copy(); // Leaf - Fix MC-65198
             if (slot == 1) {
                 item.onCraftedBy(itemstack1, player.level(), player);
                 if (!this.moveItemStackTo(itemstack1, 2, 38, true)) {
@@ -291,7 +292,7 @@ public class StonecutterMenu extends AbstractContainerMenu {
                 return ItemStack.EMPTY;
             }
 
-            slot1.onTake(player, itemstack1);
+            slot1.onTake(player, itemStack2); // Leaf - Fix MC-65198
             this.broadcastChanges();
         }
 
