From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:27:48 -0500
Subject: [PATCH] Fix NPE during creating GUI graph


diff --git a/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java b/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java
index 7c32c60100cf2bd109eb8762efa856c10b704fdd..81d40895cfa84676f5b200f53b128d9d6c5bc0bd 100644
--- a/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java
+++ b/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java
@@ -59,7 +59,14 @@ public class RAMDetails extends JList<String> {
         GraphData data = RAMGraph.DATA.peekLast();
         Vector<String> vector = new Vector<>();
 
-        double[] tps = org.bukkit.Bukkit.getTPSIncluding5SecondAverage(); // Gale - Purpur - 5 second TPS average
+        // Leaf - Revert to fix potential NPE during creating GUI graph
+        double[] tps = new double[] {
+                server.tps5s.getAverage(),
+                server.tps1.getAverage(),
+                server.tps5.getAverage(),
+                server.tps15.getAverage()
+        }; // Gale - Purpur - 5 second TPS average
+        // Leaf end - Revert to fix potential NPE during creating GUI graph
         String[] tpsAvg = new String[tps.length];
 
         for ( int g = 0; g < tps.length; g++) {
