From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lilingfengdev <145678359+lilingfengdev@users.noreply.github.com>
Date: Thu, 18 Jan 2024 13:30:02 +0800
Subject: [PATCH] Polpot: Make egg and snowball can knockback player

Original project: https://github.com/PolpotMC/Polpot

diff --git a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
index 440d3d72d8b2dac14f83a83caa5ae9dbf3e979b6..cb430ae1628cbf48a002991fd1c5674e4d3f63ac 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
@@ -3,6 +3,7 @@ package net.minecraft.world.entity.projectile;
 import net.minecraft.core.particles.ItemParticleOption;
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
@@ -55,6 +56,12 @@ public class Snowball extends ThrowableItemProjectile {
         Entity entity = entityHitResult.getEntity();
         int i = entity.level().purpurConfig.snowballDamage >= 0 ? entity.level().purpurConfig.snowballDamage : entity instanceof Blaze ? 3 : 0; // Purpur
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), (float)i);
+        // Leaf - Polpot start - make snowball can knockback player
+        if (org.dreeam.leaf.LeafConfig.snowballCanKnockback && entity instanceof ServerPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            ((ServerPlayer) entity).knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ(), this);
+        }
+        // Leaf - Polpot end
     }
 
     // Purpur start - borrowed and modified code from ThrownPotion#onHitBlock and ThrownPotion#dowseFire
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
index b64ecadae45c2126b92963ac8d118dde76126ddd..be46cc7423886900344b2ece0efd6615cd32b9ee 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -45,7 +45,14 @@ public class ThrownEgg extends ThrowableItemProjectile {
     @Override
     protected void onHitEntity(EntityHitResult entityHitResult) {
         super.onHitEntity(entityHitResult);
+        Entity entity = entityHitResult.getEntity(); // Polpot - make egg can knockback player
         entityHitResult.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Leaf - Polpot start - make egg can knockback player
+        if (org.dreeam.leaf.LeafConfig.eggCanKnockback && entity instanceof ServerPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            ((ServerPlayer) entity).knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ(), this);
+        }
+        // Leaf - Polpot end - make egg can knockback player
     }
 
     @Override
diff --git a/src/main/java/org/dreeam/leaf/LeafConfig.java b/src/main/java/org/dreeam/leaf/LeafConfig.java
index 3368e4d398132c64969b018e4fe8b7bbe8f2f364..f41f3da6021d71ad7291d9401282c473aaca7e33 100644
--- a/src/main/java/org/dreeam/leaf/LeafConfig.java
+++ b/src/main/java/org/dreeam/leaf/LeafConfig.java
@@ -314,6 +314,8 @@ public class LeafConfig {
     public static double verticalLimit = 0.4000000059604645D;
     public static double extraHorizontal = 0.5D;
     public static double extraVertical = 0.1D;
+    public static boolean snowballCanKnockback = false;
+    public static boolean eggCanKnockback = false;
     private static void knockback() {
         friction = getDouble("playerKnockback.friction", friction);
         horizontal = getDouble("playerKnockback.horizontal", horizontal);
@@ -321,5 +323,7 @@ public class LeafConfig {
         verticalLimit = getDouble("playerKnockback.verticalLimit", verticalLimit);
         extraHorizontal = getDouble("playerKnockback.extraHorizontal", extraHorizontal);
         extraVertical = getDouble("playerKnockback.extraVertical", extraVertical);
+        snowballCanKnockback = getBoolean("playerKnockback.snowball-knockback-players", snowballCanKnockback, "Make snowball can knockback players");
+        eggCanKnockback = getBoolean("playerKnockback.egg-knockback-players", eggCanKnockback, "Make egg can knockback players");
     }
 }
