From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index d44d95e19cecab0f76a2c7e01849c3de9a1d97f9..0d6236637b2fc27b2aa5d20fe404b6202e3fb012 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1532,6 +1532,8 @@ public abstract class PlayerList {
             File file = this.server.getWorldPath(LevelResource.PLAYER_STATS_DIR).toFile();
             File file1 = new File(file, String.valueOf(uuid) + ".json");
 
+            // Leaf start - Remove useless creating stats json bases on player name logic
+            /*
             if (!file1.exists()) {
                 File file2 = new File(file, displayName + ".json"); // CraftBukkit
                 Path path = file2.toPath();
@@ -1540,6 +1542,8 @@ public abstract class PlayerList {
                     file2.renameTo(file1);
                 }
             }
+             */
+            // Leaf end - Remove useless creating stats json bases on player name logic
 
             serverstatisticmanager = new ServerStatsCounter(this.server, file1);
             // this.stats.put(uuid, serverstatisticmanager); // CraftBukkit
