From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 1 Feb 2024 12:48:58 -0500
Subject: [PATCH] Redirect to Gale's method to fix plugin incompatibility


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index d988b3364674c384665e3a5c3a451f0ef3a81e2c..0883923acbcc661442e9cd97382420a46699d93f 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -335,6 +335,8 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
     }
     // Gale end - make max interaction distance configurable
 
+    public double MAX_INTERACTION_DISTANCE = ServerGamePacketListenerImpl.getMaxInteractionDistanceSquared(this.connection.getPlayer().level()); // Leaf - redirect to Gale's `getMaxInteractionDistanceSquared` method to fix incompatibility with ModelEngine plugin
+
     // Purpur start
     private final com.google.common.cache.LoadingCache<CraftPlayer, Boolean> kickPermissionCache = com.google.common.cache.CacheBuilder.newBuilder()
             .maximumSize(1000)
