From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 24 Feb 2024 01:16:07 -0500
Subject: [PATCH] Including 5s in getTPS()


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9a677b5dac3474ad4f6605c6aaa3e675d94537df..1b9685f759481282ba4be599169a60aba2b2012a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3115,6 +3115,8 @@ public final class CraftServer implements Server {
 
     @Override
     public double[] getTPS() {
+        if (org.dreeam.leaf.config.modules.misc.Including5sIngetTPS.enabled) return getTPSIncluding5SecondAverage(); // Leaf - Including 5s in getTPS()
+
         return new double[] {
             net.minecraft.server.MinecraftServer.getServer().tps1.getAverage(),
             net.minecraft.server.MinecraftServer.getServer().tps5.getAverage(),
diff --git a/src/main/java/org/dreeam/leaf/config/modules/misc/Including5sIngetTPS.java b/src/main/java/org/dreeam/leaf/config/modules/misc/Including5sIngetTPS.java
new file mode 100644
index 0000000000000000000000000000000000000000..8a7eb25b407296bb0e4c1fcd33c6f4caa27735ec
--- /dev/null
+++ b/src/main/java/org/dreeam/leaf/config/modules/misc/Including5sIngetTPS.java
@@ -0,0 +1,18 @@
+package org.dreeam.leaf.config.modules.misc;
+
+import org.dreeam.leaf.config.ConfigModules;
+import org.dreeam.leaf.config.EnumConfigCategory;
+
+public class Including5sIngetTPS extends ConfigModules {
+
+    public String getBasePath() {
+        return EnumConfigCategory.MISC.getBaseKeyName();
+    }
+
+    public static boolean enabled = true;
+
+    @Override
+    public void onLoaded() {
+        enabled = config.getBoolean(getBasePath() + ".including-5s-in-get-tps", enabled);
+    }
+}
