From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index c410f781371a87ecc2849144c747a6ac3182a65a..b87fa81ccee6792b3a202b2c193299dbd2cce112 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1580,6 +1580,8 @@ public abstract class PlayerList {
             File file = this.server.getWorldPath(LevelResource.PLAYER_STATS_DIR).toFile();
             File file1 = new File(file, String.valueOf(uuid) + ".json");
 
+            // Leaf start - Remove useless creating stats json bases on player name logic
+            /*
             if (!file1.exists()) {
                 File file2 = new File(file, displayName + ".json"); // CraftBukkit
                 Path path = file2.toPath();
@@ -1588,6 +1590,8 @@ public abstract class PlayerList {
                     file2.renameTo(file1);
                 }
             }
+             */
+            // Leaf end - Remove useless creating stats json bases on player name logic
 
             serverstatisticmanager = new ServerStatsCounter(this.server, file1);
             // this.stats.put(uuid, serverstatisticmanager); // CraftBukkit
