From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 19 Feb 2024 13:10:16 -0500
Subject: [PATCH] Redirect vanilla profiler


diff --git a/src/main/java/net/minecraft/world/level/PathNavigationRegion.java b/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
index 9c2321ebb1237b4ecd3e7f7053e5039b800d89ff..009716c07f5c6b5571792fe87b0a07a4e65fc43d 100644
--- a/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
+++ b/src/main/java/net/minecraft/world/level/PathNavigationRegion.java
@@ -153,4 +153,8 @@ public class PathNavigationRegion implements BlockGetter, CollisionGetter {
     public int getHeight() {
         return this.level.getHeight();
     }
+
+    public net.minecraft.util.profiling.ProfilerFiller getProfiler() {
+        return net.minecraft.util.profiling.InactiveProfiler.INSTANCE; // Gale - Purpur - remove vanilla profiler // Leaf - Redirect to fix plugin incompatibility
+    }
 }
