From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 2 May 2024 20:21:29 -0400
Subject: [PATCH] Ignore terminal provider warning


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d26f85511bb091a54d2059ef8f59f5df3093ccde..29a26b45f2dea70a3b72f08d88882c62b47e3382 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -324,6 +324,9 @@ public class Main {
                 // Paper end - Log Java and OS versioning to help with debugging plugin issues
 
                 printlnStartupInfoToSystemOut("Loading libraries, please wait..."); // Gale - include time in startup logs
+
+                System.setProperty("org.jline.terminal.disableDeprecatedProviderWarning", "true"); // Leaf - Ignore terminal provider warning
+
                 net.minecraft.server.Main.main(options);
             } catch (Throwable t) {
                 t.printStackTrace();
