From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <fsjk947@gmail.com>
Date: Wed, 13 Mar 2024 04:21:38 +0800
Subject: [PATCH] Fix-MC-93826


diff --git a/src/main/java/net/minecraft/world/entity/animal/Animal.java b/src/main/java/net/minecraft/world/entity/animal/Animal.java
index a64ab2058d4e3439bf6ee418f3192b83c346eb85..25c0ed1ada04ab3b6d95135d2a8e7fc21929121b 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Animal.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Animal.java
@@ -41,6 +41,7 @@ public abstract class Animal extends AgeableMob {
     public int inLove;
     @Nullable
     public UUID loveCause;
+    private Object level; // Leaf - Fix MC-93826
     public ItemStack breedItem; // CraftBukkit - Add breedItem variable
     public abstract int getPurpurBreedTime(); // Purpur
 
@@ -73,7 +74,11 @@ public abstract class Animal extends AgeableMob {
                 double d1 = this.random.nextGaussian() * 0.02D;
                 double d2 = this.random.nextGaussian() * 0.02D;
 
-                this.level().addParticle(ParticleTypes.HEART, this.getRandomX(1.0D), this.getRandomY() + 0.5D, this.getRandomZ(1.0D), d0, d1, d2);
+                // Leaf start - Fix MC-93826
+                if (this.level instanceof ServerLevel serverLevel) {
+                    serverLevel.sendParticles(ParticleTypes.HEART, this.getRandomX(1.0D), this.getRandomY() + 0.5D, this.getRandomZ(1.0D), 1, d0, d1, d2, 0);
+                }
+                // Leaf end
             }
         }
 
