From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Tue, 12 Mar 2024 01:10:54 +0800
Subject: [PATCH] Use a shadow fork that supports Java 21


diff --git a/build.gradle.kts b/build.gradle.kts
index 9d7cc530187263c6a4ed99df2c79d66d76749cc6..6497e2ce6720afb583dbfac5be2b7a14f8f2b632 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,7 +3,7 @@ import io.papermc.paperweight.util.*
 plugins {
     java
     `maven-publish`
-    id("com.github.johnrengelman.shadow")
+    id("io.github.goooler.shadow") // Leaf - use a shadow fork that supports Java 21
 }
 
 val log4jPlugins = sourceSets.create("log4jPlugins")
