From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nostalfinals <yuu8583@proton.me>
Date: Tue, 12 Mar 2024 01:10:54 +0800
Subject: [PATCH] Use a shadow fork that supports Java 21


diff --git a/build.gradle.kts b/build.gradle.kts
index 5c14f2a186509e9f4a7a525320cc7492be939b7c..6eaa616acb8c6f21c6554b0bead67ac7cf613933 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,7 +3,7 @@ import io.papermc.paperweight.util.*
 plugins {
     java
     `maven-publish`
-    id("com.github.johnrengelman.shadow")
+    id("io.github.goooler.shadow") // Leaf - use a shadow fork that supports Java 21
 }
 
 val log4jPlugins = sourceSets.create("log4jPlugins")
