From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 23 Jun 2024 11:26:20 +0800
Subject: [PATCH] Use caffeine cache kickPermission instead of using
 google.common.cache


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 1dcf131335fbeb19f6e01dee921587075df1231c..b362ebbf164c46764c7b918b24cbe1721afc078f 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -338,17 +338,12 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
     // CraftBukkit end
 
     // Purpur start
-    private final com.google.common.cache.LoadingCache<CraftPlayer, Boolean> kickPermissionCache = com.google.common.cache.CacheBuilder.newBuilder()
+    // Leaf start - Use caffeine cache kickPermission instead of using google.common.cache
+    private final com.github.benmanes.caffeine.cache.LoadingCache<CraftPlayer, Boolean> kickPermissionCache = com.github.benmanes.caffeine.cache.Caffeine.newBuilder()
             .maximumSize(1000)
             .expireAfterWrite(1, java.util.concurrent.TimeUnit.MINUTES)
-            .build(
-                    new com.google.common.cache.CacheLoader<>() {
-                        @Override
-                        public Boolean load(CraftPlayer player) {
-                            return player.hasPermission("purpur.bypassIdleKick");
-                        }
-                    }
-            );
+            .build(player -> player.hasPermission("purpur.bypassIdleKick"));
+    // Leaf - Use caffeine cache kickPermission instead of using google.common.cache
     // Purpur end
 
     @Override
@@ -420,7 +415,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
         if (this.player.getLastActionTime() > 0L && this.server.getPlayerIdleTimeout() > 0 && Util.getMillis() - this.player.getLastActionTime() > (long) this.server.getPlayerIdleTimeout() * 1000L * 60L && !this.player.wonGame) { // Paper - Prevent AFK kick while watching end credits
             // Purpur start
             this.player.setAfk(true);
-            if (!this.player.level().purpurConfig.idleTimeoutKick || (!Boolean.parseBoolean(System.getenv("PURPUR_FORCE_IDLE_KICK")) && kickPermissionCache.getUnchecked(this.player.getBukkitEntity()))) {
+            if (!this.player.level().purpurConfig.idleTimeoutKick || (!Boolean.parseBoolean(System.getenv("PURPUR_FORCE_IDLE_KICK")) && kickPermissionCache.get(this.player.getBukkitEntity()))) { // Leaf - Use caffeine cache kickPermission instead of using google.common.cache
                 return;
             }
             // Purpur end
