From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Sun, 7 Jul 2024 01:59:11 +0800
Subject: [PATCH] Fix-MC-223153

Related MC issue: https://bugs.mojang.com/browse/MC-223153

diff --git a/src/main/java/net/minecraft/world/level/block/Blocks.java b/src/main/java/net/minecraft/world/level/block/Blocks.java
index 7d58a95f7ae8983b466b275f4f82597d38762af0..6a2faa69e86123aeb902c2792bb225a6eaf8ba53 100644
--- a/src/main/java/net/minecraft/world/level/block/Blocks.java
+++ b/src/main/java/net/minecraft/world/level/block/Blocks.java
@@ -7535,6 +7535,7 @@ public class Blocks {
                 .mapColor(MapColor.COLOR_ORANGE)
                 .instrument(NoteBlockInstrument.BASEDRUM)
                 .requiresCorrectToolForDrops()
+                .sound(SoundType.COPPER) // Leaf - Fix MC-223153
                 .strength(5.0F, 6.0F)
         )
     );
