From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 23 Aug 2024 20:22:00 -0400
Subject: [PATCH] Remove stream in RecipeManager getRecipeFor


diff --git a/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java b/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
index febf87b14125925f548393360e89077329a6c522..de7537c4eacf6fa549f74fd329102a2e4865194a 100644
--- a/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
@@ -118,9 +118,16 @@ public class RecipeManager extends SimpleJsonResourceReloadListener {
             return Optional.of(recipe);
         } else {
             // CraftBukkit start
-            List<RecipeHolder<T>> list = this.byType(type).stream().filter((recipeholder1) -> {
-                return recipeholder1.value().matches(input, world);
-            }).toList();
+            // Leaf start - Remove streams in RecipeManager getRecipeFor
+            List<RecipeHolder<T>> list = new java.util.ArrayList<>();
+
+            for (RecipeHolder<T> recipeholder1 : this.byType(type)) {
+                if (recipeholder1.value().matches(input, world)) {
+                    list.add(recipeholder1);
+                }
+            }
+            // Leaf end - Remove streams in RecipeManager getRecipeFor
+
             return list.isEmpty() ? Optional.empty() : Optional.of(list.getLast()); // CraftBukkit - SPIGOT-4638: last recipe gets priority
             // CraftBukkit end
         }
