From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Sun, 7 Jul 2024 01:29:57 +0800
Subject: [PATCH] Fix-MC-200418

Related MC issue: https://bugs.mojang.com/browse/MC-200418

diff --git a/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java b/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java
index 7ac5f6fc44fd5f901c30fe93856ed6b04b86a265..f6545e1703d6a957a051915effa4084427ff06fe 100644
--- a/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -335,6 +335,12 @@ public class ZombieVillager extends Zombie implements VillagerDataHolder {
                 world.levelEvent((Player) null, 1027, this.blockPosition(), 0);
             }
 
+            // Leaf start - Fix MC-200418
+            if (entityvillager.isPassenger() && entityvillager.getVehicle() instanceof net.minecraft.world.entity.animal.Chicken && entityvillager.isBaby()) {
+                entityvillager.removeVehicle();
+            }
+            // Leaf end
+
         }
     }
 
