From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: alazeprt <nono135246@126.com>
Date: Sun, 15 Sep 2024 21:00:57 +0800
Subject: [PATCH] fix bugs on replay api


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 8f188b137247289770b0a663124f6345a902cd8a..11949d56094523968fd47c3f6f25a627681acf97 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -161,6 +161,8 @@ import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.ScoreAccess;
 import net.minecraft.world.scores.ScoreHolder;
+import org.leavesmc.leaves.replay.ServerPhotographer;
+import org.leavesmc.leaves.replay.ServerPhotographerGameMode;
 import org.slf4j.Logger;
 import net.minecraft.world.Container;
 import net.minecraft.world.InteractionHand;
@@ -425,7 +427,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
             public void dataChanged(AbstractContainerMenu handler, int property, int value) {}
         };
         this.textFilter = server.createTextFilterForPlayer(this);
-        this.gameMode = server.createGameModeForPlayer(this);
+        this.gameMode = this instanceof ServerPhotographer ? new ServerPhotographerGameMode((ServerPhotographer) this) : server.createGameModeForPlayer(this);
         this.server = server;
         this.stats = server.getPlayerList().getPlayerStats(this);
         this.advancements = server.getPlayerList().getPlayerAdvancements(this);
diff --git a/src/main/java/org/leavesmc/leaves/bot/BotStatsCounter.java b/src/main/java/org/leavesmc/leaves/bot/BotStatsCounter.java
new file mode 100644
index 0000000000000000000000000000000000000000..1dfbda8b2439e3f21fea953292aa0e3e853b22e0
--- /dev/null
+++ b/src/main/java/org/leavesmc/leaves/bot/BotStatsCounter.java
@@ -0,0 +1,38 @@
+package org.leavesmc.leaves.bot;
+
+import com.mojang.datafixers.DataFixer;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.stats.ServerStatsCounter;
+import net.minecraft.stats.Stat;
+import net.minecraft.world.entity.player.Player;
+
+import java.io.File;
+
+public class BotStatsCounter extends ServerStatsCounter {
+
+    private static final File UNKOWN_FILE = new File("BOT_STATS_REMOVE_THIS");
+
+    public BotStatsCounter(MinecraftServer server) {
+        super(server, UNKOWN_FILE);
+    }
+
+    @Override
+    public void save() {
+
+    }
+
+    @Override
+    public void setValue(Player player, Stat<?> stat, int value) {
+
+    }
+
+    @Override
+    public void parseLocal(DataFixer dataFixer, String json) {
+
+    }
+
+    @Override
+    public int getValue(Stat<?> stat) {
+        return 0;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/org/leavesmc/leaves/replay/ServerPhotographer.java b/src/main/java/org/leavesmc/leaves/replay/ServerPhotographer.java
index 17cd14449bde1ee7d2b5938a04ee3eebf516d581..2bcfe08870e006210b9d1457e983547c4079be4a 100644
--- a/src/main/java/org/leavesmc/leaves/replay/ServerPhotographer.java
+++ b/src/main/java/org/leavesmc/leaves/replay/ServerPhotographer.java
@@ -5,6 +5,7 @@ import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ClientInformation;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.level.ServerPlayerGameMode;
 import net.minecraft.stats.ServerStatsCounter;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.phys.Vec3;
@@ -13,6 +14,7 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.CraftWorld;
 import org.jetbrains.annotations.NotNull;
 import org.leavesmc.leaves.LeavesLogger;
+import org.leavesmc.leaves.bot.BotStatsCounter;
 import org.leavesmc.leaves.entity.CraftPhotographer;
 import org.leavesmc.leaves.entity.Photographer;
 
@@ -37,7 +39,6 @@ public class ServerPhotographer extends ServerPlayer {
 
     private ServerPhotographer(MinecraftServer server, ServerLevel world, GameProfile profile) {
         super(server, world, profile, ClientInformation.createDefault());
-        this.gameMode = new ServerPhotographerGameMode(this);
         this.followPlayer = null;
         this.stats = new BotStatsCounter(server);
         this.lastPos = this.position();
