From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Om Rotat <lolelol42069@gmail.com>
Date: Fri, 18 Apr 2025 11:37:25 +0700
Subject: [PATCH] 	modified:  
 src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a2f329e958fe72d262da9a4cee050efb69294df4..006c608f6527fbbb046702deb62238b2ea4e673c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2484,7 +2484,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     private void sendCustomPayload(ResourceLocation id, byte[] message) {
-        ClientboundCustomPayloadPacket packet = new ClientboundCustomPayloadPacket(new DiscardedPayload(id, Unpooled.wrappedBuffer(message)));
+        ClientboundCustomPayloadPacket packet = new ClientboundCustomPayloadPacket(new DiscardedPayload(id, message));
         this.getHandle().connection.send(packet);
     }
 
