From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Om Rotat <lolelol42069@gmail.com>
Date: Fri, 18 Apr 2025 20:26:08 +0700
Subject: [PATCH] 	modified:  
 src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java


diff --git a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index f1941dabe9a130d0656479da3e3048d5b034932c..828c35f07aeff62d94a261479eacf87ea281da7a 100644
--- a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -117,7 +117,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     @Override
     public void onPacketError(Packet packet, Exception exception) throws ReportedException {
         ServerCommonPacketListener.super.onPacketError(packet, exception);
-        ((MinecraftServer)this.server).reportPacketHandlingException(exception, packet.type());
+
     }
 
     @Override
